---
  title: "GCT_nat_comms_2022_R1"
author: "Thomas R W Oliver"
output: 
  html_document:
  toc: true
toc_depth: 2
number_sections: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#GCT_nat_comms_2022_R1

#00_SUPPLEMENTARY_DATA
##WGS and mRNA metadata

Reading in the metadata tables for the DNA and RNA samples with updated histological annotation.

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA')

wgs = read.table('WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)
mrna = read.table('mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #descriptive column contains apostrophes

#mrna$Normal.Tumour = ifelse(mrna$Updated.Description == "Seminiferous_tubule", "Normal", "Tumour") #2 cuts were re-annotated from normal to GCNIS so their normal/tumour annotation needs updating 20/1/22
#write.table(mrna, 'mRNA_summary_metadata.txt', col.names = T, row.names = F, quote = F , sep = '\t')

```

##Histological counts

Counts for each histological feature per tumour, divided by DNA and RNA

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA')

#read in metadata for microbiopsies
wgs = read.table('WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)
mrna = read.table('mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '')

#subset to columns of interest
wgs.s = wgs[wgs$Experimental_arm == 'LCM', c('Case.ID', 'Histology')]
wgs.s$type = 'DNA'
mrna.s = mrna[, c('Case.ID', 'Updated.Description')]
mrna.s$type = 'RNA'

colnames(mrna.s) = colnames(wgs.s)

#combined dataframe of all cuts
cuts_combined = rbind(wgs.s, mrna.s)

histo.counts = data.frame(matrix(nrow = length(unique(cuts_combined$Case.ID)), ncol = length(unique(cuts_combined$Histology)) * 2))
row.names(histo.counts) = unique(cuts_combined$Case.ID)
colnames(histo.counts) = c(paste0(unique(cuts_combined$Histology), '_DNA'), paste0(unique(cuts_combined$Histology), '_RNA'))

histo.counts[is.na(histo.counts)] <- 0 #use 0, not NA

for(i in row.names(histo.counts)){
  if(nrow(cuts_combined[cuts_combined$type == 'DNA' & cuts_combined$Case.ID == i,]) > 0){ #if we have WGS for that patient
    sample_dna = cuts_combined[cuts_combined$type == 'DNA' & cuts_combined$Case.ID == i,]
    for(j in unique(cuts_combined$Histology)){
      histo.counts[i, paste0(j, '_DNA')] = nrow(sample_dna[sample_dna$Histology == j,]) #how many WGS we have for that feature
    }
  }
  if(nrow(cuts_combined[cuts_combined$type == 'RNA' & cuts_combined$Case.ID == i,]) > 0){
    sample_rna = cuts_combined[cuts_combined$type == 'RNA' & cuts_combined$Case.ID == i,]
    for(j in unique(cuts_combined$Histology)){
      histo.counts[i, paste0(j, '_RNA')] = nrow(sample_rna[sample_rna$Histology == j,])
    }
  }
}

#double check the final counts
sum(histo.counts[,grepl("DNA", colnames(histo.counts))]) #131
sum(histo.counts[,grepl("RNA", colnames(histo.counts))]) #416

write.table(histo.counts, 'histo_counts_20211222.txt', quote = F, sep = '\t', row.names = T, col.names = T)

```

#01_DNA_PROCESSING

##Stage bam files

Where offpipe variant calling and analyses are required.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/staged_bulk_bam_files

lfs setstripe . -c -1
mkdir logs
module load dataImportExport
stageBam.pl -p 2127 -s PD50654a,PD50654b,PD50655a,PD50655b,PD50656a,PD50656b,PD50657a,PD50657b,PD50659a,PD50659b,PD50660a,PD50660b,PD50661a,PD50661b -o . -lo logs

```

```{bash}

cd /lustre/scratch117/casm/team274/to3/gct_bams

lfs setstripe . -c -1
mkdir logs
module load dataImportExport

INPUT="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt"
for f in PD42036 PD43296 PD43298 PD43299 PD45544 PD46269 PD46966 PD46969;
do
grep $f $INPUT | awk -F "\t" '{ if ($23 >= 5) { print $1 } }' > ${f}_samples.txt
grep $f $INPUT | awk -F "\t" '{ if ($23 >= 5) { print $3 } }' | uniq >> ${f}_samples.txt
done

for f in PD43296 PD43298 PD43299 PD45544 PD46269 PD46966 PD46969;
do
SAMPLES=$(cat ${f}_samples.txt | tr '\n' ',' | sed 's/.$//')
stageBam.pl -p 2127 -s $SAMPLES -o . -lo logs -m 10000
#echo $SAMPLES
done

#stageBam.pl -p 2127 -s PD46969d_lo0001,PD46969d_lo0003,PD46969d_lo0005,PD46969e_lo0001,PD46969e_lo0002,PD46969e_lo0005,PD46969f -o . -lo logs -m 10000

for f in PD42036;
do
SAMPLES=$(cat ${f}_samples.txt | tr '\n' ',' | sed 's/.$//')
stageBam.pl -p 1951 -s $SAMPLES -o . -lo logs -m 10000
#echo $SAMPLES
done

```

##QC

###Estimate level of contamination

####Run conpair

Symlink reference fasta file to lustre space so genome.fa.dict can be renamed to genome.dict. Also need to load a different version of java before starting.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/conpair/pileup

#needs this version of java to run
module load conpair
module load jre/1.8.0_131

#pileup will fail as genome.fa.dict is supposed to be called genome.dict to be read correctly. We can avoid this by symlinking to the original files without needing to download new versions of the files
mkdir reference_fa_files
cd reference_fa_files
ln -sf /nfs/cancer_ref02/Homo_sapiens/GRCh37d5/genome.fa. .
ln -sf /nfs/cancer_ref02/Homo_sapiens/GRCh37d5/genome.fa.fai .
ln -sf /nfs/cancer_ref02/Homo_sapiens/GRCh37d5/genome.fa.dict genome.dict

```

#####Pileup

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/conpair/pileup

#run gatk pileup for tumor - takes ~5 minutes
infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/supplementary_data/GCT_DNA_supplementary_data.txt" #LCM data only at this point
grep -v 'Sample' $infile | while IFS=$'\t' read -r -a tmp; 
do
#echo "${tmp[0]}" #sample id
#echo "${tmp[2]}" #project id
bsub -o ${tmp[0]}_pileup.out -e ${tmp[0]}_pileup.err -q small -R "span[hosts=1]" -M10000 -R 'select[mem>10000] rusage[mem=10000]' \
run_gatk_pileup_for_sample.py \
-B /lustre/scratch117/casm/team294/rs30/GCT_analyses/Staged_BAM/${tmp[2]}/${tmp[0]}/mapped_sample/${tmp[0]}.sample.dupmarked.bam  \
-O ${tmp[0]}_pileup \
--reference /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/conpair/pileup/reference_fa_files/genome.fa
done

#run gatk pileup for normal - takes ~5 minutes
grep -v 'Sample' $infile | cut -f3,4 | sort | uniq | while IFS=$'\t' read -r -a tmp; 
do
#echo "${tmp[0]}" #project id
#echo "${tmp[1]}" #normal id
bsub -o ${tmp[1]}_pileup.out -e ${tmp[1]}_pileup.err -q small -R "span[hosts=1]" -M10000 -R 'select[mem>10000] rusage[mem=10000]' \
run_gatk_pileup_for_sample.py \
-B /lustre/scratch117/casm/team294/rs30/GCT_analyses/Staged_BAM/${tmp[0]}/${tmp[1]}/mapped_sample/${tmp[1]}.sample.dupmarked.bam  \
-O ${tmp[1]}_pileup \
--reference /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/conpair/pileup/reference_fa_files/genome.fa
done

#run gatk pileup for bulk samples
for f in $(ls /lustre/scratch119/casm/team294rr/to3/testes/tumour/staged_bulk_bam_files/2127/PD*/mapped_sample/PD*bam);
do
#echo $f
filename=$(echo $f | tr '\n' '\0' | xargs -0 -n 1 basename)
#echo $filename
sample="${filename%.sample.dupmarked.bam}"
echo $sample
bsub -o ${sample}_pileup.out -e ${sample}_pileup.err -q small -R "span[hosts=1]" -M10000 -R 'select[mem>10000] rusage[mem=10000]' \
run_gatk_pileup_for_sample.py \
-B /lustre/scratch119/casm/team294rr/to3/testes/tumour/staged_bulk_bam_files/2127/${sample}/mapped_sample/${sample}.sample.dupmarked.bam  \
-O ${sample}_pileup \
--reference /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/conpair/pileup/reference_fa_files/genome.fa
done

```

#####Contamination

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/conpair/contamination

#check for contamination of matched normal and tumor
infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/supplementary_data/GCT_DNA_supplementary_data.txt"
grep -v 'Sample' $infile | while IFS=$'\t' read -r -a tmp; 
do
#echo "${tmp[0]}" #sample id
#echo "${tmp[2]}" #project id
#echo "${tmp[3]}" #normal id
bsub -o ${tmp[0]}_${tmp[3]}_contamination.out -e ${tmp[0]}_${tmp[3]}_contamination.err -q small -R "span[hosts=1]" -M1000 -R 'select[mem>1000] rusage[mem=1000]' \
estimate_tumor_normal_contamination.py \
-T ../pileup/${tmp[0]}_pileup \
-N ../pileup/${tmp[3]}_pileup \
--outfile ${tmp[0]}_${tmp[3]}_contamination \
--min_mapping_quality 30
done

for f in PD50654 PD50655 PD50656 PD50657 PD50659 PD50660 PD50661;
do
bsub -o ${f}a_${f}b_contamination.out -e ${f}a_${f}b_contamination.err -q small -R "span[hosts=1]" -M1000 -R 'select[mem>1000] rusage[mem=1000]' \
estimate_tumor_normal_contamination.py \
-T ../pileup/${f}a_pileup \
-N ../pileup/${f}b_pileup \
--outfile ${f}a_${f}b_contamination \
--min_mapping_quality 30
done

```

####Append contamination estimates to metadata

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/conpair/contamination')

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', header = T, sep = '\t', stringsAsFactors = F)

contam_files = list.files('.', 'contamination$')

#check normal samples
contam_df = c()
for(i in contam_files){
  sample = paste(unlist(strsplit(i, '_'))[1:2], collapse = '_')
  file = read.table(i, header = F, sep = '\t', stringsAsFactors = F)
  contam_frac = as.numeric(unlist(strsplit(unlist(strsplit(file[1,], 'Normal sample contamination level: '))[2], '%')))/100 #previously expressed as percentage
  contam_df = rbind(contam_df, c(sample, contam_frac))
}

contam_df = data.frame(contam_df)
contam_df$X2 = as.numeric(contam_df$X2)
contam_df[contam_df$X2 > 0.005, ] #no samples have over an estimated 0.5% contamination

#check tumour samples
contam_df = c()
for(i in contam_files){
  sample = paste(unlist(strsplit(i, '_'))[1:2], collapse = '_')
  file = read.table(i, header = F, sep = '\t', stringsAsFactors = F)
  contam_frac = as.numeric(unlist(strsplit(unlist(strsplit(file[2,], 'Tumor sample contamination level: '))[2], '%')))/100 #previously expressed as percentage
  contam_df = rbind(contam_df, c(sample, contam_frac))
}

contam_df = data.frame(contam_df)
contam_df$X2 = as.numeric(contam_df$X2)
names(contam_df) = c('sample', 'est_conpair_contam')
contam_df$sample = sapply(strsplit(contam_df$sample, '_PD'), '[[', 1) #naming different for bulk YST samples

contam_df[contam_df$est_conpair_contam > 0.005, ] #no samples have over an estimated 0.5% contamination

#assign contamination estimate back to supplementary data
manifest$conpair_est = NA
for(i in 1:nrow(manifest)){
  manifest$conpair_est[i] = contam_df[contam_df$sample == manifest$Sample[i],]$est_conpair_contam
}

#manifest = manifest[, c(1:14, 38, 15:37)] #move column to new position

write.table(manifest, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', sep = '\t', quote = F, col.names = T, row.names = F)

```

###Estimate total callable length of genome

####Run mosdepth for whole genome

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mosdepth

ls -d /lustre/scratch117/casm/team294/rs30/GCT_analyses/Staged_BAM/*/*/*/*bam >> gct_bam_list.txt
ls -d /lustre/scratch119/casm/team294rr/to3/testes/tumour/staged_bulk_bam_files/*/*/*/*bam >> gct_bam_list.txt

while read line;
do
set $line
sampleID=$(basename $1 | sed "s/[.].*//")
echo $1
echo $sampleID
/nfs/users/nfs_t/to3/scripts/run_mosdepth.sh $sampleID $1 "0:1:4:100:" /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mosdepth/ ; 
done < gct_bam_list.txt

module load bedtools

ls *.quantized.quantized.bed.gz | while read FILE ; do zgrep -P "\tCALLABLE|HIGH_COVERAGE$" "$FILE" | bedtools intersect -a - -b /lustre/scratch119/casm/team294rr/rs30/References/GRCh37d5/chrom.sizes.bed -u | bedtools subtract -a - -b /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/caveman/flagging/simple_repeats.bed | bedtools subtract -a - -b /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/caveman/flagging/hi_seq_depth.bed | bedtools subtract -a - -b /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/caveman/flagging/centromeric_repeats.bed > ${FILE%.bed.gz}.callable.calledContigs.subtracted_ignored_regions.bed ; done

ls *.quantized.quantized.callable.calledContigs.subtracted_ignored_regions.bed | while read FILE ; do perl /nfs/users/nfs_r/rs30/repositories/Baylor/scripts/BED_stats/get_BED-stats.v2.pl "$FILE" ; done

```

####Run mosdepth for autosomal genome

Used as a standardised way across sexes to estimate mutations per Mb

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mosdepth

for f in $(ls *.bed);
do
filename=$(basename -- "$f")
filename="${filename%.quantized.quantized.callable.calledContigs.subtracted_ignored_regions.bed}"
echo $filename
awk -F "\t" '{ if(($1 != "X") && ($1 != "Y")) { print } }' ${filename}.quantized.quantized.callable.calledContigs.subtracted_ignored_regions.bed > ${filename}.quantized.quantized.callable.calledContigs.subtracted_ignored_regions.autosomalchr_only.bed
done

ls *.quantized.quantized.callable.calledContigs.subtracted_ignored_regions.autosomalchr_only.bed | while read FILE ; do perl /nfs/users/nfs_r/rs30/repositories/Baylor/scripts/BED_stats/get_BED-stats.v2.pl "$FILE" ; done

```

####Put callable lengths in metadata

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mosdepth')

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', header = T, sep = '\t', stringsAsFactors = F)

for(i in 1:nrow(manifest)){
  bed_stats = read.table(paste0(manifest$Sample[i], '.quantized.quantized.callable.calledContigs.subtracted_ignored_regions.autosomalchr_only.bedStats'), header = F, sep = '\t', stringsAsFactors = F)
  manifest$Autosomal.callable.region.length[i] = unlist(strsplit(bed_stats[5,], 'Total Size: | bases'))[2]
}

manifest$Autosomal.callable.region.length = as.numeric(manifest$Autosomal.callable.region.length)

write.table(manifest, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', col.names = T, row.names = F, sep = '\t', quote = F)

```

##VARIANT CALLING AND FILTERING (microbiopsies)

###-Substitutions

####CaVEMan

Substitutions called using CaVEMan against matched normal sample (blood or normal testis).

#####Check all files are matched correctly

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/subs/lcm/01_caveman

grep "##SAMPLE=<ID=NORMAL" *_complete.vcf > matched_normal_check.txt
grep "##SAMPLE=<ID=TUMOUR" *_complete.vcf > matched_tumour_check.txt
paste --delimiters='\t' matched_tumour_check.txt matched_normal_check.txt > combined_matched_check.txt

```

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/subs/lcm/01_caveman")

cv_file = read.table("combined_matched_check.txt", header = F, sep = '\t', stringsAsFactors = F, comment.char = "")
cv_file$V1 = substr(lapply(strsplit(cv_file$V1, "SampleName="), "[[", 2), 0, 7)
cv_file$V2 = substr(lapply(strsplit(cv_file$V2, "SampleName="), "[[", 2), 0, 7)

cv_file[cv_file$V1 != cv_file$V2,] #tumour and normal match in each vcf

```

#####Check files not truncated

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/subs/lcm/01_caveman

ls *_complete.vcf | while read FILE ; do zgrep "^#" -v "$FILE" | cut -f1 | uniq -c > ../00_qc_chrom/"$FILE".sub.counts.txt ; done

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/subs/lcm/00_qc_chrom
wc -l *.sub.counts.txt

#some samples don't have Y variants called but otherwise don't seem truncated 22/12/21

```

####Low input pipeline artefact filter ("ms44 filter")

This incorporates the PASS, ASMD and CLPM filters used in the bulk filtering strategy too.

```{bash}

#22/03/20
/nfs/users/nfs_t/to3/scripts/runLCMFiltering_by_sample_project_v3-RS.sh \
-pt 1951,1951,1951,1951,1951,1951,1951,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127 \
-st PD42034b_lo0078,PD42034b_lo0079,PD42034b_lo0081,PD42034b_lo0082,PD42034b_lo0084,PD42036b_lo0046,PD42036b_lo0047,PD43296a_lo0088,PD43296a_lo0089,PD43296a_lo0092,PD43296a_lo0095,PD43296a_lo0097,PD43296a_lo0119,PD43296a_lo0121,PD43296a_lo0124,PD43296a_lo0126,PD43296a_lo0128,PD43296a_lo0132,PD43296a_lo0134,PD43298a_lo0028,PD43298a_lo0032,PD43298a_lo0034,PD43298a_lo0037,PD43298a_lo0040,PD43298a_lo0045,PD43298a_lo0049,PD43298a_lo0053,PD43298a_lo0055,PD43298a_lo0075,PD43298a_lo0086,PD43298a_lo0091,PD43298a_lo0093,PD43298a_lo0100,PD46269a_lo0024,PD46269a_lo0025,PD46269a_lo0028,PD46269a_lo0030,PD46269a_lo0033,PD46269a_lo0034,PD46269a_lo0035,PD46269a_lo0040,PD46269a_lo0044,PD46269a_lo0046,PD46270a_lo0002,PD46270a_lo0007,PD46270c_lo0001,PD46270c_lo0003,PD46270c_lo0005,PD46271a_lo0001,PD46271a_lo0005 \
-pn 1951,1951,1951,1951,1951,1951,1951,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127,2127 \
-sn PD42034d,PD42034d,PD42034d,PD42034d,PD42034d,PD42036c,PD42036c,PD43296b,PD43296b,PD43296b,PD43296b,PD43296b,PD43296b,PD43296b,PD43296b,PD43296b,PD43296b,PD43296b,PD43296b,PD43298b,PD43298b,PD43298b,PD43298b,PD43298b,PD43298b,PD43298b,PD43298b,PD43298b,PD43298b,PD43298b,PD43298b,PD43298b,PD43298b,PD46269b,PD46269b,PD46269b,PD46269b,PD46269b,PD46269b,PD46269b,PD46269b,PD46269b,PD46269b,PD46270b,PD46270b,PD46270b,PD46270b,PD46270b,PD46271b,PD46271b \
-o /lustre/scratch119/casm/team294rr/to3/testes/tumour/ms44

/nfs/users/nfs_t/to3/scripts/runLCMFiltering_by_sample_project_v3-RS.sh \
-pt 2127,2127,2127,2127 \
-st PD43296a_lo0119,PD43298a_lo0028,PD43298a_lo0040,PD46269a_lo0024 \
-pn 2127,2127,2127,2127 \
-sn PD43296b,PD43298b,PD43298b,PD46269b \
-o /lustre/scratch119/casm/team294rr/to3/testes/tumour/file_repo/SSMs

```

First run had contamination estimate of 0, now going to use 0.1 and see if it is any different.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/subs/lcm/02_lcm/normal_contam_0.1

infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt"
grep -v 'PD506\|Sample' $infile | while IFS=$'\t' read -r -a tmp ; 
do
echo "${tmp[0]}" >> samples.txt #sample id
echo "${tmp[2]}" >> normals.txt #matched normal id
done

grep "PD420" $infile | while IFS=$'\t' read -r -a tmp ;
do
echo 1951 >> projectIDs.txt
done

grep -v "PD506\|Sample\|PD420" $infile | while IFS=$'\t' read -r -a tmp ;
do
echo 2127 >> projectIDs.txt
done

cp projectIDs.txt Normal_projectIDs.txt #same location for sample and matche normal

#run in screen
ST=$(cat samples.txt | tr '\n' ','| sed 's/.$//')
PT=$(cat projectIDs.txt | tr '\n' ','| sed 's/.$//')
SN=$(cat normals.txt | tr '\n' ','| sed 's/.$//')
PN=$(cat Normal_projectIDs.txt | tr '\n' ','| sed 's/.$//')

#echo $ST
#echo $PT
#echo $SN
#echo $PN

/nfs/users/nfs_t/to3/scripts/runLCMFiltering_by_sample_project_v3_farm5_TO3_20220207.sh  \
        -pt $PT \
        -st $ST \
        -pn $PN \
        -sn $SN \
        -o .  \
        --threads 2

```

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/subs/lcm/02_lcm/normal_contam_0.1

for f in $(ls *_complete_final_allinfo_4.txt);
do
sample=${f%_complete_final_allinfo_4.txt}
wc -l $f
wc -l ../${sample}_complete_final_allinfo_4.txt
done

```

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/subs/lcm/02_lcm/normal_contam_0.1")

lcm_calls_0.1 = list.files(".", "_complete_final_allinfo_4.txt")
samples = unlist(strsplit(lcm_calls_0.1, "_complete_final_allinfo_4.txt"))

caveman_comparison = data.frame()
old_uniq = data.frame()
new_uniq = data.frame()
for(sample in samples){
  new = read.table(paste0(sample, "_complete_final_allinfo_4.txt"), header = T, sep = '\t', stringsAsFactors = F)
  new$ID = paste(new$Chr, new$End, new$Ref, new$Alt, sep = "_")
  new$sample = sample
  colnames(new)[58] = "PD.sample.dupmarked.sec.depth"
  colnames(new)[79] = "PD.matched.normal.dupmarked.sec.depth"
  old = read.table(paste0("../", sample, "_complete_final_allinfo_4.txt"), header = T, sep = '\t', stringsAsFactors = F)
  old$ID = paste(old$Chr, old$End, old$Ref, old$Alt, sep = "_")
  old$sample = sample
  colnames(old)[58] = "PD.sample.dupmarked.sec.depth"
  colnames(old)[79] = "PD.matched.normal.dupmarked.sec.depth"
  caveman_comparison = rbind(caveman_comparison, c(sample, nrow(new) - nrow(old), mean(c(nrow(old[!old$ID %in% new$ID,]), nrow(new[!new$ID %in% old$ID,]))), mean(c(nrow(new), nrow(old)))))
  old_uniq = rbind(old_uniq, old[!old$ID %in% new$ID,])
  new_uniq = rbind(new_uniq, new[!new$ID %in% old$ID,])
}

names(caveman_comparison) = c("sample", "extra_subs_0.1", "mean_difference_uniq_subs", "mean_post_ms44_filter_burden")
caveman_comparison$extra_subs_0.1 = as.numeric(caveman_comparison$extra_subs_0.1)
caveman_comparison$mean_difference_uniq_subs = as.numeric(caveman_comparison$mean_difference_uniq_subs)
caveman_comparison$mean_post_ms44_filter_burden = as.numeric(caveman_comparison$mean_post_ms44_filter_burden)
caveman_comparison$frac_calls = caveman_comparison$mean_difference_uniq_subs / caveman_comparison$mean_post_ms44_filter_burden

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)
caveman_comparison$purity = NA

for(i in 1:nrow(caveman_comparison)){
  caveman_comparison$purity[i] = manifest[manifest$Sample == caveman_comparison$sample[i],]$Purity
}

library(ggplot2)

ggplot(caveman_comparison) + geom_histogram(mapping = aes(x = extra_subs_0.1), bins = 20) + theme_linedraw() + xlab("Extra subs post-ms44 filters with caveman normal contamination set to 0.1")
ggplot(caveman_comparison) + geom_histogram(mapping = aes(x = abs(frac_calls)), bins = 20) + theme_linedraw() + xlab("Mean discordant burden / mean total burden between 0.0 and 0.1 normal contamination")
ggplot(caveman_comparison) + geom_point(mapping = aes(x = extra_subs_0.1, y = purity)) + theme_linedraw() + xlab("Extra subs post-ms44 filters with caveman normal contamination set to 0.1")
ggplot(caveman_comparison) + geom_point(mapping = aes(x = abs(frac_calls), y = purity)) + theme_linedraw() + xlab("Mean discordant burden / mean total burden between 0.0 and 0.1 normal contamination")

new_uniq[new_uniq$Func.refGene == "exonic",]
#Chr     Start       End Ref Alt Func.refGene Gene.refGene GeneDetail.refGene ExonicFunc.refGene
#10 108427543 108427543   C   T       exonic       SORCS1                     nonsynonymous SNV
#15  93521588  93521588   C   G       exonic         CHD2                     nonsynonymous SNV
#15  59344529  59344529   G   T       exonic       RNF111                        synonymous SNV
#18  43253742  43253742   C   T       exonic      SLC14A2                        synonymous SNV
old_uniq[old_uniq$Func.refGene == "exonic",] #none

ssms = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/gct_vagrent_ssm_all_mutations_plus_depth20210915.txt', header = T,  sep = '\t', stringsAsFactors = F)
ssms$ID = paste(ssms$Chr, ssms$Position, ssms$Wildtype, ssms$Mutant, sep = "_")

#how many of these apparently new calls are actually detected in our final sub set when we recall all unique variants per tumour across each tumour sample again?
new_uniq$final_called = "No"
for(i in 1:nrow(new_uniq)){
  if(nrow(ssms[ssms$Sample == new_uniq$sample[i] & ssms$ID == new_uniq$ID[i],]) > 0) new_uniq$final_called[i] = "Yes"
}

nrow(new_uniq[new_uniq$final_called == "Yes",]) / nrow(new_uniq) #70.4% of the 491 variants

mean(caveman_comparison$extra_subs_0.1) #1.656489
mean(caveman_comparison$frac_calls) #0.002679737

```


####Pileup across all samples per tumour

Symlink to all tumour and normal bam files.

Create bed file for each patient within the directory with the VCFs.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/file_repo/SSMs/matched

for patientID in $(ls *_complete_final_retained_4.vcf | cut -c1-7 | uniq);
do
  cut -f 1,2,4,5 ${patientID}*_complete_final_retained_4.vcf >> /lustre/scratch119/casm/team294rr/to3/testes/tumour/cgpVAF/SNVs/${patientID}_SSM.snps.tmp.bed
  sort -k1,1 -k2,2n -k3,4 /lustre/scratch119/casm/team294rr/to3/testes/tumour/cgpVAF/SNVs/${patientID}_SSM.snps.tmp.bed | uniq > /lustre/scratch119/casm/team294rr/to3/testes/tumour/cgpVAF/SNVs/${patientID}_SSM.snps.bed
  grep -v "#" /lustre/scratch119/casm/team294rr/to3/testes/tumour/cgpVAF/SNVs/${patientID}_SSM.snps.bed > /lustre/scratch119/casm/team294rr/to3/testes/tumour/cgpVAF/SNVs/${patientID}_SSM_nohash.snps.bed
  mv /lustre/scratch119/casm/team294rr/to3/testes/tumour/cgpVAF/SNVs/${patientID}_SSM_nohash.snps.bed /lustre/scratch119/casm/team294rr/to3/testes/tumour/cgpVAF/SNVs/${patientID}_SSM.snps.bed
done

```

#####Run

Create txt file with a list of the matched normals then submit the following to cgpfarm.

```{bash}

cd /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2

while read line;
do
  set $line
  patientID=$(echo $1 | cut -c1-7)
  #echo $patientID
  declare -a sampleArray=()
  #echo $1
  
  for sample in $patientID*lo*.sample.dupmarked.bam;
  do
    filename=$(basename -- "$sample")
    sampleID="${filename%.sample.dupmarked.bam}"
    #echo $sampleID
    sampleArray+=("$sampleID")
    
  done
  
  #echo "${sampleArray[@]}"
  
  mkdir -p "$patientID"
 
  bsub  -o ${patientID}/%J.$patientID.stdout -e ${patientID}/%J.$patientID.stderr -q normal -n 4 -R'span[hosts=1] select[mem>=8000] rusage[mem=8000]' -M8000  perl-5.16.3 -I /software/CGP/projects/vafCorrect/lib/perl5 /software/CGP/projects/vafCorrect/bin/cgpVaf.pl -d $PWD -o "$patientID" -a snp -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be ".sample.dupmarked.bam" -hdr /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b ${patientID}_SSM.snps.bed -nn $1 -tn "${sampleArray[@]}" -bq 25 -mq 30
  
done < matched_normals_nolcm.txt

#separate for PD46966 as it uses an LCM cut as the matched normal
patientID='PD46966'
MATCHED_NORMAL='PD46966b_lo0004'
TUMOUR_SAMPLES='PD46966a_lo0001 PD46966a_lo0002 PD46966a_lo0003 PD46966a_lo0005 PD46966a_lo0006 PD46966a_lo0009 PD46966a_lo0010 PD46966a_lo0012 PD46966a_lo0013'

bsub  -o ${patientID}/%J.$patientID.stdout -e ${patientID}/%J.$patientID.stderr -q normal -n 4 -R'span[hosts=1] select[mem>=8000] rusage[mem=8000]' -M8000  perl-5.16.3 -I /software/CGP/projects/vafCorrect/lib/perl5 /software/CGP/projects/vafCorrect/bin/cgpVaf.pl -d $PWD -o "$patientID" -a snp -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be ".sample.dupmarked.bam" -hdr /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b ${patientID}_SSM.snps.bed -nn ${MATCHED_NORMAL} -tn ${TUMOUR_SAMPLES} -bq 25 -mq 30

```

#####Edit output

Remove the hashed lines from the cgpVaf output
```{bash}
cd /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2

for f in $(ls /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/*/*_snp_vaf.tsv);
do
filename=$(basename -- "$f")
patient=${filename:0:7}
#echo $patient
filename="${filename%_snp_vaf.tsv}"
grep -v "##" $f > ${patient}/${filename}_snp_vaf_nohash.tsv
done

```

####Generate new VCF files per sample with 1+ mt read

This will capture variants that pass filters in other samples but miss the threshold in this particular sample. Improves sensitivity for mutation calling. We can then filter this file for remaining artefacts and variants who frequency does not statistically exceed the noise threshold.

```{bash}

#farm5
cd /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/
module load bcftools

#now run to separate all samples out of genotyped vcfs with enriched variant calls
while read line;
do
  set $line
  patientID=$(echo $1 | cut -c1-7)
  #echo $patientID
  echo $1 #matched normal
  
  for sample in $patientID*lo*.sample.dupmarked.bam;
  do
    filename=$(basename -- "$sample")
    sampleID="${filename%.sample.dupmarked.bam}"
    #echo $sampleID
    bcftools view -s $1,$sampleID $patientID/${1}*_snp_vaf.vcf.gz > $patientID/${sampleID}_snps_geno.vcf #subset vcf to include only the matched normal and a single tumour sample
    bcftools filter -i 'FORMAT/MTR[1] > 0' $patientID/${sampleID}_snps_geno.vcf > $patientID/${sampleID}_snps_geno.final.vcf #second sample filtering as first is the matched normal
    
  done
  
done < matched_normals.txt

```

####Pileup across all tumours against unrelated normal LCM samples

Using an unrelated collection of LCM placenta and testis samples, I will be able to generate a locus specific error rate that will be used to inform whether any of my remaining mutations should be removed.

First, generate a bed file of all mutations called across all samples, similar to above but all variants across all tumours combined.

Create bed file for each patient within the directory with the VCFs.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/file_repo/SSMs/matched

cut -f 1,2,4,5 *_snps_geno.final.vcf >> /lustre/scratch117/casm/team294/to3/testes/tumour/allelecount/20200511/GCT_SSM.snvs.tmp.bed
sort -k1,1 -k2,2n -k3,4 /lustre/scratch117/casm/team294/to3/testes/tumour/allelecount/20200511/GCT_SSM.snvs.tmp.bed | uniq > /lustre/scratch117/casm/team294/to3/testes/tumour/allelecount/20200511/GCT_SSM.snvs.bed
rm /lustre/scratch117/casm/team294/to3/testes/tumour/allelecount/20200511/GCT_SSM.snvs.tmp.bed
grep -v "#" /lustre/scratch117/casm/team294/to3/testes/tumour/allelecount/20200511/GCT_SSM.snvs.bed > /lustre/scratch117/casm/team294/to3/testes/tumour/allelecount/20200511/GCT_SSM.snvs_nohash.bed
mv /lustre/scratch117/casm/team294/to3/testes/tumour/allelecount/20200511/GCT_SSM.snvs_nohash.bed /lustre/scratch117/casm/team294/to3/testes/tumour/allelecount/20200511/GCT_SSM.snvs.bed

wc -l GCT_SSM.snvs.bed #32354 mutations

```

Now to gather a list of unmatched normal testis and placenta samples against which to generate the error rates. Also need a list of my own tumour samples.

```{bash}

wc -l /lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/placenta_samples.txt #125 placenta WGS
wc -l /lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/normal_reference_testis_samples.txt #125 normal, unrelated testis samples
wc -l /lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/gct_samples.txt #131 tumour samples

```

Create HTS files for each project

```{bash}

cd /lustre/scratch117/casm/team294/to3/testes/tumour/allelecount/20200511

cat /lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/placenta_samples.txt | while read line;
do
set $line
echo "/nfs/cancer_ref01/nst_links/live/2058/${1}/${1}.sample.dupmarked.bam"
done > 2058_bam_list.txt

cat /lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/normal_reference_testis_samples.txt | while read line;
do
set $line
echo "/nfs/cancer_ref01/nst_links/live/1951/${1}/${1}.sample.dupmarked.bam"
done > 1951_bam_list.txt

cat /lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/gct_samples.txt | while read line;
do
set $line
echo "/nfs/cancer_ref01/nst_links/live/2127/${1}/${1}.sample.dupmarked.bam"
done > 2127_bam_list.txt #edit relevant files to be 1951

```

#####Run

Run alleleCounter

```{bash}

cd /lustre/scratch117/casm/team294/to3/testes/tumour/allelecount/20200511

while read i;
do
sampleID=$(basename $i | sed "s/[.].*//")
bsub -J"${sampleID}" -o ${sampleID}.o -e ${sampleID}.e -q normal -n 5 -R"select[mem>10000] rusage[mem=10000] span[hosts=1]" -M10000 "/software/CGP/canpipe/live/bin/canpipe_live alleleCounter -l GCT_SSM.snvs.bed -b $i -m 25 -q 30 -o /lustre/scratch117/casm/team294/to3/testes/tumour/allelecount/20200511/output/$sampleID.txt"
done < 2058_bam_list.txt 

while read i;
do
sampleID=$(basename $i | sed "s/[.].*//")
bsub -J"${sampleID}" -o ${sampleID}.o -e ${sampleID}.e -q normal -n 5 -R"select[mem>10000] rusage[mem=10000] span[hosts=1]" -M10000 "/software/CGP/canpipe/live/bin/canpipe_live alleleCounter -l GCT_SSM.snvs.bed -b $i -m 25 -q 30 -o /lustre/scratch117/casm/team294/to3/testes/tumour/allelecount/20200511/output/$sampleID.txt"
done < 1951_bam_list.txt 

while read i;
do
sampleID=$(basename $i | sed "s/[.].*//")
bsub -J"${sampleID}" -o ${sampleID}.o -e ${sampleID}.e -q normal -n 5 -R"select[mem>10000] rusage[mem=10000] span[hosts=1]" -M10000 "/software/CGP/canpipe/live/bin/canpipe_live alleleCounter -l GCT_SSM.snvs.bed -b $i -m 25 -q 30 -o /lustre/scratch117/casm/team294/to3/testes/tumour/allelecount/20200511/output/$sampleID.txt"
done < 2127_bam_list.txt 

```

####Shearwater-like sub filtering v2

Original code missed some variants found at a universally low depth (just under 1% of calls), likely regions of low mapping quality.

Also removes variants whose average coverage across all tumour samples is less than 10.

#####Function

```{r}

#-------------------------------------------------
# Libraries
#-------------------------------------------------

library("GenomicRanges")
library("deepSNV")
library("Rsamtools")

logbb = deepSNV:::logbb
dbetabinom = VGAM::dbetabinom

#-------------------------------------------------
# Functions
#-------------------------------------------------

estimateRho_gridml = function(x, mu) {
  # Estimate rho by MLE grid approach
  rhovec = 10^seq(-6,-0.5,by=0.05) # rho will be bounded within 1e-6 and 0.32
  mm = x[,2]
  cov = c(x[,1])
  ll = sapply(rhovec, function(rhoj) sum(dbetabinom(x=mm, size=cov, rho=rhoj, prob=mu, log=T)))
  rhovec[ll==max(ll)][1]
}

shearwater_probability=function(patient, save=NULL, path_prefix='', rho=10^-3){
  #Function to calculate probability of presence of mutation based on Shearwarer
  #'patient' is the name of the patient-specific subfolder
  #'path_prefix' is any prefix to the path necessary to find that subfolder
  #'rho' is the constant for the overdispersion parameter. If rho=NULL, calculate
  # it from the data (much slower)
  #'save' is path for output. If NULL, returns matrix
  # output of this function will be a matrix (muts by samples) of p values
  
  
  #A file of mutations in patient subdirectory (format: Chr_Ref_Pos_Alt)
  Muts_patient = read.table(paste0(path_prefix,patient,"/All_mutations_filtered.txt"))[,1] 
  #A file of with the sample names belonging to this patient
  case_samples=read.table(paste0(path_prefix,patient,"/samples.txt"))[,1]
  normal_panel = normal_samples
  norm_all_counts = all_counts[normal_panel,,]
  coords_proj = substr(Muts_patient,1,nchar(Muts_patient)-4)
  case_all_counts = all_counts[case_samples,,]
 
  avg_depth_list = c()

  for (i in coords_proj){
    mean_d = sum(case_all_counts[,i,]) / length(case_samples)
    avg_depth_list = c(avg_depth_list, mean_d)
  }
  
  Muts_patient = Muts_patient[avg_depth_list > 10] #Removes remaining globally low coverage variants which were the result of mapping issues
  coords_proj = coords_proj[avg_depth_list > 10]

  #Set up pval matrix
  pval_mat = matrix(1,ncol=length(case_samples),nrow=length(Muts_patient))
  rownames(pval_mat)=Muts_patient
  colnames(pval_mat)=case_samples
  
  Alt=substr(Muts_patient,nchar(Muts_patient),nchar(Muts_patient))
  Ref=substr(Muts_patient,nchar(Muts_patient)-2,nchar(Muts_patient)-2)
  
  for (s in case_samples){
    rho_est=rep(NA,length(coords_proj))
    test_counts = all_counts[s,coords_proj,]
    for (k in 1:length(coords_proj)) {
      n = sum(test_counts[coords_proj[k],])
      x = test_counts[coords_proj[k],Alt[k]]
      
      N_indiv = rowSums(norm_all_counts[,coords_proj[k],])
      X_indiv = norm_all_counts[,coords_proj[k],c("A","C","G","T")!=Ref[k]]
      pseudo = .Machine$double.eps    
      N=sum(N_indiv)
      X=sum(X_indiv)
      
      mu = max(X,pseudo)/max(N,pseudo)
      counts = cbind(N,X)
      if(is.null(rho)) rho = estimateRho_gridml(counts,mu)
      rdisp = (1 - rho)/rho
      
      prob0 = (X + x)/(N + n); prob0[prob0==0] = pseudo; prob0[prob0==1] = 1 - pseudo
      prob1s = x/(n + pseudo); prob1s[prob1s==0] = pseudo; prob1s[prob1s==1] = 1 - pseudo
      prob1c = X/(N + pseudo); prob1c[prob1c==0] = pseudo; prob1c[prob1c==1] = 1 - pseudo
      
      prob1s = pmax(prob1s,prob1c) # Min error rate is that of the population (one-sided test)
      nu0 = prob0 * rdisp; nu1s = prob1s * rdisp; nu1c = prob1c * rdisp; 
      
      # Likelihood-Ratio Tests
      LL = logbb(x, n, nu0, rdisp) + logbb(X, N, nu0, rdisp) - logbb(x, n, nu1s, rdisp) - logbb(X, N, nu1c, rdisp)
      pvals = pchisq(-2*LL, df=1, lower.tail=F)/2 # We divide by 2 as we are performing a 1-sided test
      # Saving the result
      pval_mat[k,s] = pvals
    } 
  }
  if(is.null(save)){
    return(pval_mat)
  }else{
    write.table(pval_mat,save)
  }
}

```

#####Input

```{r}

setwd('/lustre/scratch117/casm/team294/to3/testes/tumour/allelecount/20200511')
#options(stringsAsFactors = F)

# Vector of normal reference samples, taken from a separate cohort of patients
normal_samples = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/reference_lcm_normal_panel.txt")[,1] 

# Vector of all samples, including the reference panel and samples of interest
tumour_samples = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/gct_samples.txt")[,1]
samples <- c(tumour_samples, normal_samples)

# "Bed" file of all mutations to be considered (across all patients)
# Format: Chr Ref Pos Alt
muts = read.table("GCT_SSM.snvs.bed") 
coords = paste(muts$V1, muts$V2, sep="_")

# Read in data from AlleleCounter
all_counts = array(0,dim=c(length(samples), length(coords), 4),
                   dimnames = list(samples, coords, c("A","C","G","T")))

print(length(samples)) #381 samples, 250 reference samples, 131 GCT genomes

for (k in 1:length(samples)){
  #Read in allele counts per sample
  if(file.exists(paste0("output/", samples[k],".txt"))){
    data=read.table(paste0("output/", samples[k],".txt"),comment.char = '',header=T)
    muts_data=paste(data$X.CHR,data$POS,sep="_")
    data=data[!duplicated(muts_data),]
    muts_data=muts_data[!duplicated(muts_data)]
    rownames(data)=muts_data
    all_counts[k,,]=as.matrix(data[coords,3:6])
  }
  print(k)
}

#create list of study patients
setwd('/lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2')
sample.list <- read.table('patients_all.txt', header = F)[,1]

#generate mutations per patient list
patient.list <- unique(substr(sample.list, 0, 7))

for(i in patient.list){
  data <- read.table(paste0(i, '/', list.files(paste0(i, '/'), pattern = '_snp_vaf_nohash.tsv')[1]), comment.char = "", header = T)
  Muts = paste(data$Chrom, data$Pos, data$Ref, data$Alt, sep="_")
  write.table(Muts, paste0(i, '/All_mutations_filtered.txt'), col.names = F, row.names = F, quote = F)
}

#generate a list of samples per tumour
for(i in patient.list){
  write.table(unlist(strsplit(list.files(paste0(i, '/'), pattern = '_snps_geno.final.vcf'), '_snps_geno.final.vcf')), paste0(i, '/samples.txt'), col.names = F, row.names = F, quote = F)
}

# read in the list of study patients
patients = read.table("/lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/patients_all.txt")[,1]

```

#####Run

```{r}

for (patient in patients){
  shearwater_probability(patient=patient, save=paste0("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/sub_shearwater/", patient, "/shearwater_snv_pval_mat.txt"))
  print(patient)
}

#get an idea of the mutation burden per sample after this filtering step
post_swater_mb <- c()
for (patient in patients){
  pval_mat = read.table(paste0("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/sub_shearwater/", patient, "/shearwater_snv_pval_mat.txt"), row.names = 1, sep = ' ', header = T)
  qval_mat = apply(pval_mat,2,function(x) p.adjust(x,method="BH", n = length(as.matrix(pval_mat))))
  post_swater_mb <- c(post_swater_mb, colSums(qval_mat < 0.001, na.rm = T))
  }

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/sub_shearwater/')

#BH correction for multiple testing
for (patient in patients){
  pval_mat = read.table(paste0(patient, "/shearwater_snv_pval_mat.txt"), row.names = 1, sep = ' ', header = T)
  qval_mat = apply(pval_mat,2,function(x) p.adjust(x,method="BH", n = length(as.matrix(pval_mat))))
  write.table(qval_mat, paste0(patient, "/shearwater_snv_qval_mat.txt"), row.names = T, sep = '\t', quote = F)
  }

library(VariantAnnotation)

for (patient in patients){
  qval_mat = read.table(paste0(patient, "/shearwater_snv_qval_mat.txt"), row.names = 1, sep = '\t', header = T)
  samples = read.table(paste0(patient, "/samples.txt"), sep = '\t', header = F)[,1]
  for (sample in samples){
    data = readVcf(paste0(patient, '/', sample, '_snps_geno.final.vcf'))
    reads = cbind((geno(data)$FAZ)[,2]+(geno(data)$RAZ)[,2],(geno(data)$FCZ)[,2]+(geno(data)$RCZ)[,2],
                (geno(data)$FGZ)[,2]+(geno(data)$RGZ)[,2],(geno(data)$FTZ)[,2]+(geno(data)$RTZ)[,2])
    Var = data.frame(Chr=as.character(seqnames(rowRanges(data))),
                   Pos=start(ranges(rowRanges(data))),
                   Ref=as.character(ref(data)))
    Alt_tmp = CharacterList(alt(data))
    Var$Alt = as.character(unlist(Alt_tmp))
    Var$NR=rowSums(reads)
    Var$NV=NA
    colnames(reads)=c("A","C","G","T")
    for (k in c("A","C","G","T")){
      Var$NV[Var$Alt==k] = reads[Var$Alt==k,k]
    }
    Var$VAF=Var$NV/Var$NR
    Var$ID = paste(Var$Chr, Var$Pos, Var$Ref, Var$Alt, sep = '_')
    qval_patient_sig_muts = row.names(qval_mat[qval_mat[,sample] < 0.001,])
    swater_filtered_muts = Var[Var$ID %in% qval_patient_sig_muts,]
    write.table(swater_filtered_muts, paste0(patient, '/', sample, '_post_swater_final_snvs.txt'), sep = '\t', quote = F, row.names = F)
  }
}

```

#####Apply to genotyped VCF

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/sub_shearwater')

patients = read.table("/lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/patients_all.txt")[,1]

library(VariantAnnotation)

for (patient in patients){
  samples = read.table(paste0(patient, "/samples.txt"), sep = '\t', header = F)[,1]
  for (sample in samples){
    data = readVcf(paste0(patient, '/', sample, '_snps_geno.final.vcf'))
    filtered_muts = read.table(paste0(patient, '/', sample, '_post_swater_final_snvs.txt'), header = T, sep = '\t', stringsAsFactors = F)[,8]
    chr = as.character(seqnames(rowRanges(data)))
    pos = start(ranges(rowRanges(data)))
    ref = as.character(ref(data))
    alt = unlist(CharacterList(alt(data)))
    info(data)$mut_ID = paste(chr, pos, ref, alt, sep = '_')
    info(header(data)) = rbind(info(header(data)), DataFrame(Number = '.', Type = 'String', Description = 'Unique mutation ID', row.names = 'mut_ID'))
    filtered_data = data[info(data)$mut_ID %in% filtered_muts]
    writeVcf(filtered_data, filename = paste0(patient, '/',sample, '_post_swater_final_snvs.vcf'))
  }
}

```

####VAGrENT annotation

Vagrent annotations were originally applied to original caveman file.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/sub_shearwater

cp PD*/*_post_swater_final_snvs.vcf vagrent_annotation/
  
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/sub_shearwater/vagrent_annotation/

module load vagrent

for f in $(ls *_post_swater_final_snvs.vcf);
do
filename=$(basename -- "$f")
sample="${filename%_post_swater_final_snvs.vcf}"
AnnotateVcf.pl -i ${sample}_post_swater_final_snvs.vcf -o ${sample}_post_swater_final_snvs.annot.vcf -c /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/vagrent/e75/vagrent.cache.gz --species Human --assembly GRCh37d5
done

```

####Convert to text file

#####Functions

```{r}

library(VariantAnnotation)

annotated_subs_text_file = function(vcf_file, type){
  if(type == 'lcm'){
    
    vcf = readVcf(vcf_file)
    sample = unlist(strsplit(vcf_file, '_post_swater_final_snvs.annot.vcf'))
    case = substr(sample, 0, 7)
    reads = cbind((geno(vcf)$FAZ)[,2]+(geno(vcf)$RAZ)[,2],(geno(vcf)$FCZ)[,2]+(geno(vcf)$RCZ)[,2],
                (geno(vcf)$FGZ)[,2]+(geno(vcf)$RGZ)[,2],(geno(vcf)$FTZ)[,2]+(geno(vcf)$RTZ)[,2])
    var.df = data.frame(Chr = as.character(seqnames(rowRanges(vcf))),
                    Position = start(ranges(rowRanges(vcf))),
                    Wildtype = as.character(ref(vcf)),
                    Mutant = unlist(CharacterList(alt(vcf))))
    var.df$Total_read_depth=rowSums(reads)
    var.df$Variant_read_depth=NA
    var.df$Case_ID = case
    var.df$Sample = sample
    var.df = var.df[, c(7, 8, 1:6)]
    colnames(reads)=c("A","C","G","T")
    for (k in c("A","C","G","T")){
      var.df$Variant_read_depth[var.df$Mutant==k] = reads[var.df$Mutant==k,k]
    }
  
    var.df$VAF = var.df$Variant_read_depth/ var.df$Total_read_depth
    var.df$Mutation = info(vcf)$VT
    vagrent_annots = lapply(strsplit(info(vcf)$VW, '\\|'), `length<-`, max(lengths(strsplit(info(vcf)$VW, '\\|')))) #make all elements in list the same length for subsetting purposes
    var.df$Gene = sapply(vagrent_annots, '[[', 1)
    var.df$Transcript = sapply(vagrent_annots, '[[', 3)
    var.df$Codon = sapply(vagrent_annots, '[[', 4)
    var.df$Protein = sapply(vagrent_annots, '[[', 5)
    var.df$Consequence = unlist(lapply(info(vcf)$VC, function(x) if(identical(x, character(0))) NA_character_ else x))
    write.table(var.df, paste0(sample, '_post_swater_final_snvs.annot.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
    
  }
  
  if(type == 'bulk'){
    
    vcf = readVcf(vcf_file)
    sample = unlist(strsplit(vcf_file, '.caveman_c.annot_filtered.vcf'))
    case = substr(sample, 0, 7)
    reads = cbind((geno(vcf)$FAZ)[,2]+(geno(vcf)$RAZ)[,2],(geno(vcf)$FCZ)[,2]+(geno(vcf)$RCZ)[,2],
                (geno(vcf)$FGZ)[,2]+(geno(vcf)$RGZ)[,2],(geno(vcf)$FTZ)[,2]+(geno(vcf)$RTZ)[,2])
    var.df = data.frame(Chr = as.character(seqnames(rowRanges(vcf))),
                    Position = start(ranges(rowRanges(vcf))),
                    Wildtype = as.character(ref(vcf)),
                    Mutant = unlist(CharacterList(alt(vcf))))
    var.df$Total_read_depth=rowSums(reads)
    var.df$Variant_read_depth=NA
    var.df$Case_ID = case
    var.df$Sample = sample
    var.df = var.df[, c(7, 8, 1:6)]
    colnames(reads)=c("A","C","G","T")
    for (k in c("A","C","G","T")){
      var.df$Variant_read_depth[var.df$Mutant==k] = reads[var.df$Mutant==k,k]
    }
  
    var.df$VAF = var.df$Variant_read_depth/ var.df$Total_read_depth
    var.df$Mutation = info(vcf)$VT
    vagrent_annots = lapply(strsplit(info(vcf)$VW, '\\|'), `length<-`, max(lengths(strsplit(info(vcf)$VW, '\\|')))) #make all elements in list the same length for subsetting purposes
    var.df$Gene = sapply(vagrent_annots, '[[', 1)
    var.df$Transcript = sapply(vagrent_annots, '[[', 3)
    var.df$Codon = sapply(vagrent_annots, '[[', 4)
    var.df$Protein = sapply(vagrent_annots, '[[', 5)
    var.df$Consequence = unlist(lapply(info(vcf)$VC, function(x) if(identical(x, character(0))) NA_character_ else x))
    write.table(var.df, paste0(sample, '.caveman_c.annot_filtered.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
    
    }
}
  

```

#####Run

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/sub_shearwater/vagrent_annotation')

subs_vcf = list.files('.', '.annot.vcf')
sample_type = ifelse(grepl('PD506', subs_vcf), 'bulk', 'lcm')

for(i in 1:length(subs_vcf)){
  annotated_subs_text_file(vcf_file = subs_vcf[i], type = sample_type[i])
}

```

###-Indels

####Pindel

Indels called using pindel against matched normal sample (blood or normal testis).

#####Check all files are matched correctly

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/indels/lcm/99_live_check

infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt"
grep -v 'PD506\|Sample' $infile | while IFS=$'\t' read -r -a tmp ; 
do
#echo "${tmp[0]}" #sample id
cp /nfs/cancer_ref01/nst_links/live/2127/${tmp[0]}/${tmp[0]}.pindel.annot.vcf.gz .
cp /nfs/cancer_ref01/nst_links/live/1951/${tmp[0]}/${tmp[0]}.pindel.annot.vcf.gz .
done

zgrep "##SAMPLE=<ID=NORMAL" *.pindel.annot.vcf.gz > matched_normal_check.txt
zgrep "##SAMPLE=<ID=TUMOUR" *.pindel.annot.vcf.gz > matched_tumour_check.txt
paste --delimiters='\t' matched_tumour_check.txt matched_normal_check.txt > combined_matched_check.txt

```

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/indels/lcm/99_live_check")

pin_file = read.table("combined_matched_check.txt", header = F, sep = '\t', stringsAsFactors = F, comment.char = "")
pin_file$V1 = substr(lapply(strsplit(pin_file$V1, "SampleName="), "[[", 2), 0, 7)
pin_file$V2 = substr(lapply(strsplit(pin_file$V2, "SampleName="), "[[", 2), 0, 7)

pin_file[pin_file$V1 != pin_file$V2,]

#        V1      V2
#        PD46966 PDv37is
#        PD46966 PDv37is
#        PD46966 PDv37is
#        PD46966 PDv37is

```


```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/indels/lcm/01_pindel

grep "##SAMPLE=<ID=NORMAL" *.pindel.annot.vcf > matched_normal_check.txt
grep "##SAMPLE=<ID=TUMOUR" *.pindel.annot.vcf > matched_tumour_check.txt
paste --delimiters='\t' matched_tumour_check.txt matched_normal_check.txt > combined_matched_check.txt

```

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/indels/lcm/01_pindel")

pin_file = read.table("combined_matched_check.txt", header = F, sep = '\t', stringsAsFactors = F, comment.char = "")
pin_file$V1 = substr(lapply(strsplit(pin_file$V1, "SampleName="), "[[", 2), 0, 7)
pin_file$V2 = substr(lapply(strsplit(pin_file$V2, "SampleName="), "[[", 2), 0, 7)

pin_file[pin_file$V1 != pin_file$V2,] #tumour and normal match in each vcf

```

#####Check files not truncated

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/indels/lcm/01_pindel

ls *.pindel.annot.vcf | while read FILE ; do zgrep "^#" -v "$FILE" | cut -f1 | uniq -c > ../00_qc_chrom/"$FILE".indel.counts.txt ; done

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/indels/lcm/00_qc_chrom
wc -l *.indel.counts.txt

#looks fine 23/12/21

```

####Pass & quality filtering

Generate files containing on mutations that have passed pindel filters and have a minimum QUAL threshold of 300.
```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/file_repo/SSMs/matched

module load bcftools

for i in $(ls *.pindel.annot.vcf);
do
filename=$(basename -- "$i")
SAMPLE="${filename%.pindel.annot.vcf}"
bcftools filter -i 'QUAL >= 300 & FILTER="PASS"' ${SAMPLE}.pindel.annot.vcf > ${SAMPLE}.pindel.HQ.vcf
done

```

####Pileup across all samples per tumour

Symlink to all tumour and normal bam files.

Create bed file for each patient within the directory with the VCFs.
```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/file_repo/SSMs/matched

for patientID in $(ls *.pindel.HQ.vcf | cut -c1-7 | uniq);
do
cut -f 1,2,4,5 ${patientID}*.pindel.HQ.vcf >> /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/${patientID}_SSM.indels.tmp.bed
sort -k1,1 -k2,2n -k3,4 /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/${patientID}_SSM.indels.tmp.bed | uniq > /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/${patientID}_SSM.indels.bed
rm /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/${patientID}_SSM.indels.tmp.bed
grep -v "#" /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/${patientID}_SSM.indels.bed > /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/${patientID}_SSM_nohash.indels.bed
mv /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/${patientID}_SSM_nohash.indels.bed /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/${patientID}_SSM.indels.bed
done

```

#####Run per chromosome for most heavily sampled tumours

Create txt file with a list of the matched normals then submit the following to cgpfarm. Will need to remove the PD46966b_lo0004 which this code duplicates as the matched normal and a LCM sample.

Submit per chromosome for the two patients with the most samples to speed things up.

```{bash}

#depending on where the free lustre space is
cd /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2

while read line;
do
set $line
patientID=$(echo $1 | cut -c1-7)
#echo $patientID
declare -a sampleArray=()
#echo $1

for sample in $patientID*lo*.sample.dupmarked.bam;
do
filename=$(basename -- "$sample")
sampleID="${filename%.sample.dupmarked.bam}"
#echo $sampleID
sampleArray+=("$sampleID")

done

#echo "${sampleArray[@]}"
declare -a arr=("1" "2" "3" "4" "5" "6" "7" "8" "9" "10" "11" "12" "13" "14" "15" "16" "17" "18" "19" "20" "21" "22" "X" "Y")

#echo "${arr[@]}"
#mkdir -p "$patientID"
for f in $(echo ${arr[@]});
do
#echo $f
bsub -J cgpvaf"${f}" -o ${patientID}/%J.$patientID.stdout -e ${patientID}/%J.$patientID.stderr -q long -n 6 -R"span[hosts=1] select[mem>=3000] rusage[mem=3000]" -M3000 perl-5.16.3 -I /software/CGP/projects/vafCorrect/lib/perl5 /software/CGP/projects/vafCorrect/bin/cgpVaf.pl -d $PWD -o "$patientID" -a indel -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be ".sample.dupmarked.bam" -hdr /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b ${patientID}_SSM.indels.bed -nn $1 -tn "${sampleArray[@]}" -chr "${f}" -bq 25 -mq 30
done
done < matched_normals_nolcm_long.txt

```

Once submitted on a per chromosome basis, the results will need to be concatenated

```{bash}

for patientID in PD43296;
do
echo $patientID
#echo ${patientID}b
MATCHED_NORMAL='PD43296b'
declare -a sampleArray=()

for sample in $patientID*lo*.sample.dupmarked.bam;
do
filename=$(basename -- "$sample")
sampleID="${filename%.sample.dupmarked.bam}"
echo $sampleID
sampleArray+=("$sampleID")

done

#sampleArray+=("$MATCHED_NORMAL")
echo "${sampleArray[@]}"

bsub -J cgpvaf"${patientID}" -o ${patientID}/%J.$patientID.stdout -e ${patientID}/%J.$patientID.stderr -q long -n 6 -R"span[hosts=1] select[mem>=4000] rusage[mem=4000]" -M4000 perl-5.16.3 -I /software/CGP/projects/vafCorrect/lib/perl5 /software/CGP/projects/vafCorrect/bin/cgpVaf.pl -d /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/PD43296/tmpvaf_PD43296a_lo0026 -o "$patientID" -a indel -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be ".sample.dupmarked.bam" -nn ${MATCHED_NORMAL} -tn "${sampleArray[@]}" -ct 1 -bo 1 -b ${patientID}_SSM.indels.bed -hdr /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz
done

for patientID in PD43298;
do
echo $patientID
#echo ${patientID}b
MATCHED_NORMAL='PD43298b'
declare -a sampleArray=()

for sample in $patientID*lo*.sample.dupmarked.bam;
do
filename=$(basename -- "$sample")
sampleID="${filename%.sample.dupmarked.bam}"
echo $sampleID
sampleArray+=("$sampleID")

done

#sampleArray+=("$MATCHED_NORMAL")
echo "${sampleArray[@]}"

bsub -J cgpvaf"${patientID}" -o ${patientID}/%J.$patientID.stdout -e ${patientID}/%J.$patientID.stderr -q long -n 6 -R"span[hosts=1] select[mem>=4000] rusage[mem=4000]" -M4000 perl-5.16.3 -I /software/CGP/projects/vafCorrect/lib/perl5 /software/CGP/projects/vafCorrect/bin/cgpVaf.pl -d /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/PD43298/tmpvaf_PD43298a_lo0003 -o "$patientID" -a indel -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be ".sample.dupmarked.bam" -nn ${MATCHED_NORMAL} -tn "${sampleArray[@]}" -ct 1 -bo 1 -b ${patientID}_SSM.indels.bed -hdr /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz
done

```

#####Run for remaining tumours

```{bash}

#depending on where the free lustre space is
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/cgpVAF/indels
cd /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2

while read line;
do
set $line
patientID=$(echo $1 | cut -c1-7)
#echo $patientID
declare -a sampleArray=()
#echo $1

for sample in $patientID*lo*.sample.dupmarked.bam;
do
filename=$(basename -- "$sample")
sampleID="${filename%.sample.dupmarked.bam}"
#echo $sampleID
sampleArray+=("$sampleID")

done

#echo "${sampleArray[@]}"

mkdir -p "$patientID"

bsub  -o ${patientID}/%J.$patientID.stdout -e ${patientID}/%J.$patientID.stderr -q long -n 6 -R'span[hosts=1] select[mem>=20000] rusage[mem=20000]' -M20000  perl-5.16.3 -I /software/CGP/projects/vafCorrect/lib/perl5 /software/CGP/projects/vafCorrect/bin/cgpVaf.pl -d $PWD -o "$patientID" -a indel -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be ".sample.dupmarked.bam" -hdr /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b ${patientID}_SSM.indels.bed -nn $1 -tn "${sampleArray[@]}" -bq 25 -mq 30

done < matched_normals_nolcm_short.txt #run case with LCM matched normal separately

#separate for PD46966 as it uses an LCM cut as the matched normal
patientID='PD46966'
MATCHED_NORMAL='PD46966b_lo0004'
TUMOUR_SAMPLES='PD46966a_lo0001 PD46966a_lo0002 PD46966a_lo0003 PD46966a_lo0005 PD46966a_lo0006 PD46966a_lo0009 PD46966a_lo0010 PD46966a_lo0012 PD46966a_lo0013'

bsub  -o ${patientID}/%J.$patientID.stdout -e ${patientID}/%J.$patientID.stderr -q long -n 6 -R'span[hosts=1] select[mem>=8000] rusage[mem=8000]' -M8000  perl-5.16.3 -I /software/CGP/projects/vafCorrect/lib/perl5 /software/CGP/projects/vafCorrect/bin/cgpVaf.pl -d $PWD -o "$patientID" -a indel -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be ".sample.dupmarked.bam" -hdr /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b ${patientID}_SSM.indels.bed -nn ${MATCHED_NORMAL} -tn ${TUMOUR_SAMPLES} -bq 25 -mq 30

```

#####Edit output

Remove the hashed lines from the cgpVaf output then check wc -l to make sure output is a sensible length
```{bash}

for f in $(ls /lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/PD*/PD*_indel_vaf.tsv);
do
filename=$(basename -- "$f")
filename="${filename%_indel_vaf.tsv}"
grep -v "##" $f > ${filename}_indel_vaf_nohash.tsv
done

```

####Generate new VCF files per sample with 1+ mt read

This will capture variants that pass filters in other samples but miss the threshold in this particular sample. Improves sensitivity for mutation calling. We can then filter this file for remaining artefacts and variants who frequency does not statistically exceed the noise threshold.

```{bash}

#farm5
module load bcftools

#now run to separate all samples out of genotyped vcfs with enriched variant calls
while read line;
do
set $line
patientID=$(echo $1 | cut -c1-7)
#echo $patientID
#echo $1 #matched normal

for sample in $patientID*lo*.sample.dupmarked.bam;
do
filename=$(basename -- "$sample")
sampleID="${filename%.sample.dupmarked.bam}"
#echo $sampleID
bcftools view -s $1,$sampleID $patientID/${1}*_indel_vaf.vcf.gz > $patientID/${sampleID}_indels_geno.vcf
bcftools filter -i 'FORMAT/MTR[1] > 0' $patientID/${sampleID}_indels_geno.vcf > $patientID/${sampleID}_indels_geno.final.vcf #second sample filtering as first is the matched normal

done

done < matched_normals.txt

```

####Pileup across all tumours against unrelated normal LCM samples

Using an unrelated collection of LCM placenta and testis samples, I will be able to generate a locus specific error rate that will be used to inform whether any of my remaining mutations should be removed. Exonerate takes a lot longer to run for the indels than the allelecount SNV pileup so I will use a reduced normal panel of 100 samples (50 randomly sampled from both the testis and placenta datasets).

Copy the previous indel bed files over to the new directory.

```{bash}
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/file_repo/SSMs/matched

cut -f 1,2,4,5 *_indels_geno.final.vcf >> /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200513/GCT_SSM.indels.tmp.bed
sort -k1,1 -k2,2n -k3,4 /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200513/GCT_SSM.indels.tmp.bed | uniq > /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200513/GCT_SSM.indels.bed
rm /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200513/GCT_SSM.indels.tmp.bed
grep -v "#" /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200513/GCT_SSM.indels.bed > /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200513/GCT_SSM.indels_nohash.bed
mv /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200513/GCT_SSM.indels_nohash.bed /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200513/GCT_SSM.indels.bed

wc -l GCT_SSM.indels.bed #47103 mutations

```

Now to gather a list of unmatched normal testis and placenta samples against which to generate the error rates. Also need a list of my own tumour samples.

```{bash}

wc -l /lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/placenta_samples.txt #125 placenta WGS
wc -l /lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/normal_reference_testis_samples.txt #125 normal, unrelated testis samples
wc -l /lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/gct_samples.txt #131 tumour samples

shuf -n 50 /lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/placenta_samples.txt > rand_placenta_samples.txt
shuf -n 50 /lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/normal_reference_testis_samples.txt > rand_norm_testis_samples.txt
cat rand_placenta_samples.txt rand_norm_testis_samples.txt > rand_norm_reference.txt
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/gct_samples.txt .

```

Symlink to bam files including matched normal as before.

First run across tumour + unrelated normal samples. Second batch includes the matched normal samples.

#####Run

```{bash}

cd /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200513
cat  rand_norm_reference.txt gct_samples.txt > all_samples.txt

declare -a sampleArray=()

while read line;
do
set $line
sampleArray+=("$1")
done < all_samples.txt

declare -a arr=("1" "2" "3" "4" "5" "6" "7" "8" "9" "10" "11" "12" "13" "14" "15" "16" "17" "18" "19" "20" "21" "22" "X" "Y")

for f in $(echo ${arr[@]});
do
bsub -J cgpvaf"${f}" -o output/%J.stdout -e output/%J.stderr -q basement -n 6 -R"span[hosts=1] select[mem>=6000] rusage[mem=6000]" -M6000 perl-5.16.3 -I /software/CGP/projects/vafCorrect/lib/perl5 /software/CGP/projects/vafCorrect/bin/cgpVaf.pl -d $PWD -o output -a indel -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be ".sample.dupmarked.bam" -hdr /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b GCT_SSM.indels.bed -nn PDv37is -tn "${sampleArray[@]}" -chr "${f}" -bq 25 -mq 30
done

cd /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200516
#cat  rand_norm_reference.txt gct_samples.txt > all_samples.txt

declare -a sampleArray=()

while read line;
do
set $line
sampleArray+=("$1")
done < matched_normals.txt

declare -a arr=("1" "2" "3" "4" "5" "6" "7" "8" "9" "10" "11" "12" "13" "14" "15" "16" "17" "18" "19" "20" "21" "22" "X" "Y")

for f in $(echo ${arr[@]});
do
bsub -J cgpvaf"${f}" -o output/%J.stdout -e output/%J.stderr -q long -n 6 -R"span[hosts=1] select[mem>=6000] rusage[mem=6000]" -M6000 perl-5.16.3 -I /software/CGP/projects/vafCorrect/lib/perl5 /software/CGP/projects/vafCorrect/bin/cgpVaf.pl -d $PWD -o output -a indel -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be ".sample.dupmarked.bam" -hdr /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b GCT_SSM.indels.bed -nn PDv37is -tn "${sampleArray[@]}" -chr "${f}" -bq 25 -mq 30
done

```

Run per chromosome because of length of time it takes exonerate to do this across 100s of samples. Therefore, output must be concatenated.

```{bash}

cd /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200513
#most samples
declare -a sampleArray=()

while read line;
do
set $line
sampleArray+=("$1")
done < all_samples.txt

bsub -J cgpvaf"_concat" -o output/%J.stdout -e output/%J.stderr -q long -n 4 -R"span[hosts=1] select[mem>=4000] rusage[mem=4000]" -M4000 perl-5.16.3 -I /software/CGP/projects/vafCorrect/lib/perl5 /software/CGP/projects/vafCorrect/bin/cgpVaf.pl -d /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200513/output/tmpvaf_PD42142b4_lo0019 -o output -a indel -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be ".sample.dupmarked.bam" -nn PDv37is -tn "${sampleArray[@]}" -ct 1 -bo 1 -b GCT_SSM.indels.bed -hdr /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz

#matched normal
cd /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200516
#cat  rand_norm_reference.txt gct_samples.txt > all_samples.txt

declare -a sampleArray=()

while read line;
do
set $line
sampleArray+=("$1")
done < matched_normals.txt

bsub -J cgpvaf"_concat" -o output/%J.stdout -e output/%J.stderr -q long -n 6 -R"span[hosts=1] select[mem>=4000] rusage[mem=4000]" -M4000 perl-5.16.3 -I /software/CGP/projects/vafCorrect/lib/perl5 /software/CGP/projects/vafCorrect/bin/cgpVaf.pl -d /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200516/output/tmpvaf_PD42034d -o output -a indel -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be ".sample.dupmarked.bam" -nn PDv37is -tn "${sampleArray[@]}" -ct 1 -bo 1 -b GCT_SSM.indels.bed -hdr /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz

```

#####Edit output

Remove hashed lines from cgpvaf output
```{bash}

cd /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200523 #moved the tsv files into here from two previous runs

for f in $(ls /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200523/PD*_indel_vaf.tsv);
do
filename=$(basename -- "$f")
filename="${filename%_indel_vaf.tsv}"
grep -v "##" $f > ${filename}_indel_vaf_nohash.tsv
done

```

#####Combined output from both runs

Combine the exonerate counts from the matched normals and all other samples.

```{r}

setwd('/lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200523')

tumour_samples = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/reference_datasets/gct_samples.txt', header = F, sep = '\t', stringsAsFactors = F)[,1]

samples = read.table('PDv37is_PD42142b4_lo0019_indel_vaf_nohash.tsv', comment.char="",header=T)
Muts = paste(samples$Chrom,samples$Pos,samples$Ref,samples$Alt,sep="_")
Genotype = samples[,grepl("VAF",colnames(samples)) & !grepl("PDv37is",colnames(samples))]
NR = samples[,grepl("DEP",colnames(samples)) & !grepl("PDv37is",colnames(samples))]
NV = samples[,grepl("MTR",colnames(samples)) & !grepl("PDv37is",colnames(samples))]
AMB = samples[,grepl("AMB",colnames(samples)) & !grepl("PDv37is",colnames(samples))]
UNK = samples[,grepl("UNK",colnames(samples)) & !grepl("PDv37is",colnames(samples))]
colnames(Genotype)=colnames(NR)=colnames(NV)=colnames(AMB)=colnames(UNK)=gsub("_VAF","",colnames(Genotype))
rownames(Genotype)=rownames(NV)=rownames(NR)=rownames(AMB)=rownames(UNK)=Muts

normals = read.table('PDv37is_PD42034d_indel_vaf_nohash.tsv', comment.char="",header=T)
Muts1 = paste(normals$Chrom,normals$Pos,normals$Ref,normals$Alt,sep="_")
Genotype1 = normals[,grepl("VAF",colnames(normals)) & !grepl("PDv37is",colnames(normals))]
NR1 = normals[,grepl("DEP",colnames(normals)) & !grepl("PDv37is",colnames(normals))]
NV1 = normals[,grepl("MTR",colnames(normals)) & !grepl("PDv37is",colnames(normals))]
colnames(Genotype1)=colnames(NR1)=colnames(NV1)=gsub("_VAF","",colnames(Genotype1))
rownames(Genotype1)=rownames(NV1)=rownames(NR1)=Muts1

NR1 = NR1[rownames(NR),]
NV1 = NV1[rownames(NR),]

write.table(cbind(NR, NR1), 'NR_all_samples_all_muts.txt', row.names = T, col.names = T, quote = F, sep = '\t')
write.table(cbind(NV, NV1), 'NV_all_samples_all_muts.txt', row.names = T, col.names = T, quote = F, sep = '\t')

```

```{bash}

while read line;
do
set $line
mkdir -p ${1}
done < /lustre/scratch119/casm/team294rr/to3/testes/tumour/cgpVAF/SNVs/patients_all.txt

```

Create list of mutations called per patient
```{bash}

cd /lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200523

for f in $(ls /lustre/scratch119/casm/team294rr/to3/testes/tumour/file_repo/SSMs/matched/*_indels_geno.final.vcf | cut -c76-82 | uniq);
do
cut -f 1,2,4,5 /lustre/scratch119/casm/team294rr/to3/testes/tumour/file_repo/SSMs/matched/${f}*_indels_geno.final.vcf >> ${f}/${f}.indels.tmp.bed
sort -k1,1 -k2,2n -k3,4 ${f}/${f}.indels.tmp.bed | uniq > ${f}/${f}.indels.bed
rm ${f}/${f}.indels.tmp.bed
grep -v "#" ${f}/${f}.indels.bed > ${f}/${f}.indels_nohash.bed
mv ${f}/${f}.indels_nohash.bed ${f}/${f}.indels.bed
done

```

####Shearwater-like indel filtering v1

Similar to above but, rather than using allelecounter, we used cgpVAF/exonerate for the pileup. It was also run on a per patient basis rather than a single pileup across the cohort. Also filters on minimum variant depth and total coverage, as well as maximum VAF in normal samples.

#####Functions

```{r}

#-------------------------------------------------
# Libraries
#-------------------------------------------------

library("GenomicRanges")
library("deepSNV")
library("Rsamtools")

logbb = deepSNV:::logbb
dbetabinom = VGAM::dbetabinom

#-------------------------------------------------
# Functions
#-------------------------------------------------

estimateRho_gridml = function(x, mu) {
  # Estimate rho by MLE grid approach
  rhovec = 10^seq(-6,-0.5,by=0.05) # rho will be bounded within 1e-6 and 0.32
  mm = x[,2]
  cov = c(x[,1])
  ll = sapply(rhovec, function(rhoj) sum(dbetabinom(x=mm, size=cov, rho=rhoj, prob=mu, log=T)))
  rhovec[ll==max(ll)][1]
}

shearwater_indel_probability=function(patient, save=NULL, path_prefix='', rho=10^-3){
  #Function to calculate probability of presence of mutation based on Shearwater
  #'patient' is the name of the patient-specific subfolder
  #'path_prefix' is any prefix to the path necessary to find that subfolder
  #'rho' is the constant for the overdispersion parameter. If rho=NULL, calculate
  # it from the data (much slower)
  #'save' is path for output. If NULL, returns matrix
  # output of this function will be a matrix (muts by samples) of p values
  
  Muts_patient = read.table(paste0(path_prefix, patient, "/All_mutations_filtered.txt"))[,1] #A file of mutations in patient subdirectory (format: Chr_Ref_Pos_Alt)
  Muts_patient = unique(Muts_patient) #to avoid duplicating variants
  case_samples = read.table(paste0(path_prefix, patient,"/samples.txt"))[,1] #A file of with the sample names belonging to this patient
  NR = read.table(paste0(path_prefix, 'NR_all_samples_all_muts.txt'), row.names = 1, header = T, sep = '\t') #total read depth at each variant locus called in all patients across all samples
  NR[NR == 0] = 1 #to prevent NAs
  NV = read.table(paste0(path_prefix, 'NV_all_samples_all_muts.txt'), row.names = 1, header = T, sep = '\t') #variant read depth per variant per sample in all samples
  matched_normals = read.table(paste0(path_prefix, 'matched_normals.txt'), header = F, sep = '\t')[,1]
  matched_normal = matched_normals[grepl(patient, matched_normals)] #picks only this patient's matched normal
  
  NR = NR[Muts_patient,] #only mutations relevant to patient
  NV = NV[Muts_patient,]
  
  Depth_filter = rowMeans(NR[, case_samples]) > 10 #Removes remaining globally low coverage variants which were the result of mapping issues
  
  NR = NR[Depth_filter,]
  NV = NV[Depth_filter,]
  
  Min_var_depth = rowSums(NV[, case_samples] >= 5) > 0 #removes low VAF sequencing artefacts. This ensures only variants called in a sample to at least 5 variant reads are kept.
  
  NR = NR[Min_var_depth,]
  NV = NV[Min_var_depth,]
  
  Muts_patient = row.names(NR) #new, slimmer list to run shearwater-like approach on
  
  #Select the normal panel not belonging to this patient with the pileup of sites where variants called in patient's tumour samples
  normal_panel = normal_samples
  norm_NR = NR[, normal_panel]
  norm_NV = NV[, normal_panel]
  
  #Set up pval matrix
  pval_mat = matrix(1,ncol=length(case_samples),nrow=length(Muts_patient))
  rownames(pval_mat)=Muts_patient
  colnames(pval_mat)=case_samples
  
  coords_proj = paste(matrix(unlist(strsplit(Muts_patient, '_')), ncol = 4, byrow = T)[,1], matrix(unlist(strsplit(Muts_patient, '_')), ncol = 4, byrow = T)[,2], sep = '_')
  
  Alt = matrix(unlist(strsplit(Muts_patient, '_')), ncol = 4, byrow = T)[,4]
  Ref = matrix(unlist(strsplit(Muts_patient, '_')), ncol = 4, byrow = T)[,3]
  
  for (s in case_samples){
    rho_est = rep(NA,length(coords_proj))
    test_counts_NR = NR[, s]
    test_counts_NV = NV[, s]
    
    for (k in 1:length(coords_proj)) {
      n = test_counts_NR[k]
      x = test_counts_NV[k]
      N_indiv = norm_NR[k,]
      X_indiv = norm_NV[k,]
      pseudo = .Machine$double.eps    
      N = sum(N_indiv)
      X = sum(X_indiv)
      mu = max(X,pseudo)/max(N,pseudo)
      counts = cbind(N,X)
      
      if(is.null(rho)) rho = estimateRho_gridml(counts,mu)
      rdisp = (1 - rho)/rho
      
      prob0 = (X + x)/(N + n); prob0[prob0==0] = pseudo; prob0[prob0==1] = 1 - pseudo
      prob1s = x/(n + pseudo); prob1s[prob1s==0] = pseudo; prob1s[prob1s==1] = 1 - pseudo
      prob1c = X/(N + pseudo); prob1c[prob1c==0] = pseudo; prob1c[prob1c==1] = 1 - pseudo
      
      prob1s = pmax(prob1s,prob1c) # Min error rate is that of the population (one-sided test)
      nu0 = prob0 * rdisp; nu1s = prob1s * rdisp; nu1c = prob1c * rdisp; 
      
      # Likelihood-Ratio Tests
      LL = logbb(x, n, nu0, rdisp) + logbb(X, N, nu0, rdisp) - logbb(x, n, nu1s, rdisp) - logbb(X, N, nu1c, rdisp)
      pvals = pchisq(-2*LL, df=1, lower.tail=F) / 2 # We divide by 2 as we are performing a 1-sided test
      # Saving the result
      pval_mat[k,s] = pvals
    } 
  }
  pval_mat = pval_mat[((rowSums(norm_NV) / rowSums(norm_NR)) < 0.2) & (NV[, matched_normal]/NR[, matched_normal] < 0.2),] #remove variants seen at a VAF of >0.2 in the normal reference panel or patient's matched normal, these end up being artefactual, even if they are seen at a significantly higher VAF in the tumour samples
  if(is.null(save)){
    return(pval_mat)
  }else{
    write.table(pval_mat, save)
  }
}


```

#####Input

```{r}

setwd('/lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200523')
#options(stringsAsFactors = F)

#create list of patients
patient.list <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/cgpVAF/SNVs/patients_all.txt', header = F)[,1]

# Vector of normal reference samples, no GCT
normal_samples = read.table("/lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200513/rand_norm_reference.txt")[,1] 

# Vector of all tumour samples, used to create a per patient sample list
tumour_samples = read.table("/lustre/scratch117/casm/team294/to3/testes/tumour/shearwater_like_indels/20200513/gct_samples.txt")[,1]

for (i in unique(substr(tumour_samples, 0, 7))){
  patient.samples = tumour_samples[grepl(paste0(i), tumour_samples)]
  patient.samples = data.frame(patient.samples)
  write.table(patient.samples, paste0(i, '/samples.txt'), sep = '\t', col.names = F, row.names = F, quote = F)
}

#generate a list of mutations per patient in the format Chr_Ref_Pos_Alt
#patient_muts_input = unlist(list.files('/lustre/scratch117/casm/team294/to3/testes/tumour/cgpVAF_phase2/', pattern = paste0('_indel_vaf_nohash.tsv')))

for (i in unique(substr(tumour_samples, 0, 7))){
  data <- read.table(paste0(i, '/', i, '.indels.bed'), header = F, sep = '\t')
  Muts = paste(data$V1,data$V2,data$V3,data$V4,sep="_")
  write.table(Muts, paste0(i, '/All_mutations_filtered.txt'), sep = '\t', col.names = F, row.names = F, quote = F)
}

# read in the list of study patients
patient.list <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/cgpVAF/SNVs/patients_all.txt', header = F)[,1]

```

#####Run

```{r}
#-------------------------------------------------
# Run
#-------------------------------------------------

for (patient in patient.list){
  shearwater_indel_probability(patient=patient,save=paste0(patient, "/shearwater_pval_indel_mat.txt"), rho = NULL) 
  print(patient)
}

#get an idea of the mutation burden per sample after this filtering step
post_swater_mb <- c()
for (patient in patient.list){
  pval_mat = read.table(paste0(patient, "/shearwater_pval_indel_mat.txt"), sep = ' ', row.names = 1, header = T)
  qval_mat = apply(pval_mat, 2, function(x) p.adjust(x, method="BH", n = length(as.matrix(pval_mat))))
  post_swater_mb <- c(post_swater_mb, colSums(qval_mat <= 0.001, na.rm = T))
}

#BH correction for multiple testing
for (patient in patient.list){
  pval_mat = read.table(paste0(patient, "/shearwater_pval_indel_mat.txt"), row.names = 1, sep = ' ', header = T)
  qval_mat = apply(pval_mat,2,function(x) p.adjust(x,method="BH",n = length(as.matrix(pval_mat))))
  write.table(qval_mat, paste0(patient, "/shearwater_qval_indel_mat.txt"), row.names = T, sep = '\t', quote = F)
}

#filter vcf to leave only variants that pass the site specific error threshold

library(VariantAnnotation)

for (patient in patient.list){
  qval_mat = read.table(paste0(patient, "/shearwater_qval_indel_mat.txt"), row.names = 1, sep = '\t', header = T)
  #qval_mat[is.na(qval_mat)] = 0.5 #replaces NA values should no longer be required
  samples = read.table(paste0(patient, "/samples.txt"), sep = '\t', header = F)[,1]
  for (sample in samples){
    data = readVcf(paste0('/lustre/scratch119/casm/team294rr/to3/testes/tumour/file_repo/SSMs/matched/', sample, '_indels_geno.final.vcf'))
    reads = cbind((geno(data)$MTR)[,2], (geno(data)$WTR)[,2])
    Var = data.frame(Chr = as.character(seqnames(rowRanges(data))), Pos=start(ranges(rowRanges(data))),
                     Ref=as.character(ref(data)))
    Alt_tmp = CharacterList(alt(data))
    Var$Alt = as.character(unlist(Alt_tmp))
    Var$NR = rowSums(reads)
    Var$NV = reads[,1]
    Var$VAF=Var$NV/Var$NR
    Var$ID = paste(Var$Chr, Var$Pos, Var$Ref, Var$Alt, sep = '_')
    qval_patient_sig_muts = row.names(qval_mat[qval_mat[,sample] <= 0.001,])
    swater_filtered_muts = Var[Var$ID %in% qval_patient_sig_muts,]
    write.table(swater_filtered_muts, paste0(patient, '/', sample, '_post_swater_final_indels.txt'), sep = '\t', quote = F, row.names = F)
  }
}

```

####Low quality indel site filter

Based on the number of reads annotated as AMB or UNK at those sites.

####Homopolymer filter (7+)

This filter and the the subsequent ones are employed in the bulk samples too.

Copy over relevant files into the directory.

```{bash}

#copy over original tumour vcfs
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/file_repo/SSMs/matched/*.pindel.annot.vcf /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering

#copy over vcfs generated by genotyping all indels that pass in at least one sample across the tumour on a per sample basis
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/file_repo/SSMs/matched/*indels_geno.final.vcf /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering

#copy over all post swater + rs30 filter files
cp /lustre/scratch117/casm/team294/rs30/GCT_analyses/Indel_filtering/Coverage_filtered/PD*/*_post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.txt /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering

```

Filter original pindel files to only include those that pass pindel and are not 1bp events occurring on homopolymer runs longer than 6bp.

```{bash}

cd /nfs/users/nfs_t/to3/scripts/pindel-filtering

infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt"
grep -v 'PD506\|Sample' $infile | while IFS=$'\t' read -r -a tmp ; 
do
#echo "${tmp[0]}" #sample id
#echo "${tmp[2]}" #project id
#echo "${tmp[3]}" #matched normal id
./bsub_homopolymer_filter_args_lustre.sh /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering/ 7 ${tmp[0]} PASS
done

```

Create bed file per patient of variants that pass the homopolymer filter within our per sample vcfs.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering/

infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt"
grep -v 'Sample\|PD506' $infile | cut -f2 | sort | uniq | while IFS=$'\t' read -r -a tmp;
do
grep -v "#" -h ${tmp[0]}*.pindel.HP.7.vcf | cut -f 1,2,4,5 | sort -k1,1 -k2,2n -k3,4 | uniq > ${tmp[0]}.pindel.pass.hp.7.muts.txt
done

for f in $(ls *.pindel.pass.hp.7.muts.txt);
do
filename=$(basename -- "$f")
sample="${filename%.pindel.pass.hp.7.muts.txt}"
awk 'BEGIN{FS=OFS="\t"}{print $1, $2, $3, $4, $1"_"$2"_"$3"_"$4}' $f > ${sample}.pindel.pass.hp.7.muts.appended.txt
done

```

Subset post_swater + rs30 filtering file to keep those not found at homopolymer runs

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering')

txt_files = list.files('.', "_post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.txt")

for(file in txt_files){
  
  put_muts = read.table(file, header = F, sep = '\t', stringsAsFactors = F)
  sample = unlist(strsplit(file, "_post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.txt"))
  patient = substr(sample, 0, 7)
  non_hp_calls = read.table(paste0(patient, ".pindel.pass.hp.7.muts.appended.txt"), header = F, sep = '\t', stringsAsFactors = F)
  put_muts = put_muts[put_muts$V8 %in% non_hp_calls$V5,]
  write.table(put_muts, paste0(sample, "_post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.txt"), quote = F, sep = '\t', col.names = F, row.names = F)
  
}

```

#####Apply to genotypes VCF

Filter genotyped vcf to only contain variants that passed above homopolymer filtering.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering

module load bcftools

for f in $(ls *_post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.txt);
do
filename=$(basename -- "$f")
sample="${filename%_post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.txt}" #get sample name
awk 'BEGIN{FS=OFS="\t"}{print $1, $2, ".", $3, $4, ".", ".", "."}' $f > ${sample}_post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.vcf #get columns for new vcf
sed -i "1i##fileformat=VCFv4.2\n#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO" ${sample}_post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.vcf 
#gzip and index original vcf and the newly filtered variants
bgzip -c ${sample}_post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.vcf > ${sample}_post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.vcf.gz
bgzip -c ${sample}_indels_geno.final.vcf > ${sample}_indels_geno.final.vcf.gz
bcftools index ${sample}_post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.vcf.gz
bcftools index ${sample}_indels_geno.final.vcf.gz
#intersect old genotype vcf and new calls, should find all new calls in old call file
bcftools isec -n=2 -w1 -Ov ${sample}_indels_geno.final.vcf.gz ${sample}_post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.vcf.gz -o ${sample}_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.vcf
done

#check same number of variants called in both files
for f in $(ls *_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.vcf);
do
filename=$(basename -- "$f")
sample="${filename%_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.vcf}"
echo $sample
grep -v "#" $f | wc -l
grep -v "#" ${sample}_post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.txt | wc -l
done

#looks good 28/6/21

```

####SNP and matched normal pindel filtering

Copy matched normal pindel calls over.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering

cp /lustre/scratch117/casm/team294/rs30/GCT_analyses/Indel_filtering/SNP_filtered/snp_file_Normal_sample.irods_list.matched.txt .

infile='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering/snp_file_Normal_sample.irods_list.matched.txt'

grep '1951' $infile |  cut -f2 | uniq | while IFS=$'\t' read -r -a tmp; 
do
cp /nfs/cancer_ref01/nst_links/live/1951/${tmp[0]}/${tmp[0]}.pindel.annot.vcf.gz .
done

grep '2127' $infile |  cut -f2 | uniq | while IFS=$'\t' read -r -a tmp; 
do
cp /nfs/cancer_ref01/nst_links/live/2127/${tmp[0]}/${tmp[0]}.pindel.annot.vcf.gz .
done

for f in PD46270b PD46271b;
do
cp /nfs/cancer_ref01/nst_links/live/2127/${f}/${f}.pindel.annot.vcf.gz .
done

```

Copy over SNP files

```{bash}

infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt"
grep 'PD420' $infile | while IFS=$'\t' read -r -a tmp;
do
cp /nfs/cancer_ref01/nst_links/live/1951/${tmp[0]}/${tmp[0]}.caveman_c.snps.vcf.gz .
done

infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt"
grep -v 'PD420\|PD506\|Sample' $infile | while IFS=$'\t' read -r -a tmp;
do
cp /nfs/cancer_ref01/nst_links/live/2127/${tmp[0]}/${tmp[0]}.caveman_c.snps.vcf.gz .
done

#those which has been repeated unmatched since of other analyses need another version of caveman to call on
infile='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering/snp_file_Normal_sample.irods_list.matched.txt'
grep -v '#' $infile |  while IFS=$'\t' read -r -a tmp; 
do
filename=$(basename -- "${tmp[2]}")
sample="${filename%.v*.caveman_c.snps.vcf.gz}"
echo $sample
cp ${tmp[2]} ./${sample}.caveman_c.snps.vcf.gz
done

```

#####Check SNP files are matched

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering

zgrep "##SAMPLE=<ID=NORMAL" *.caveman_c.snps.vcf.gz > matched_normal_snp_check.txt
zgrep "##SAMPLE=<ID=TUMOUR" *.caveman_c.snps.vcf.gz > matched_tumour_snp_check.txt
paste --delimiters='\t' matched_tumour_snp_check.txt matched_normal_snp_check.txt > combined_matched_snp_check.txt

```

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering")

cv_file = read.table("combined_matched_snp_check.txt", header = F, sep = '\t', stringsAsFactors = F, comment.char = "")
cv_file$V1 = substr(lapply(strsplit(cv_file$V1, "SampleName="), "[[", 2), 0, 7)
cv_file$V2 = substr(lapply(strsplit(cv_file$V2, "SampleName="), "[[", 2), 0, 7)

cv_file[cv_file$V1 != cv_file$V2,] #tumour and normal match in each vcf

```

#####Check indel files from matched normal were run unmatched

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering

infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt"
grep -v 'Sample\|PD506' $infile | cut -f3 | uniq |while IFS=$'\t' read -r -a tmp;
do
echo ${tmp[0]} #matched normal ID

zgrep "##SAMPLE=<ID=NORMAL" ${tmp[0]}*.pindel.annot.vcf.gz >> matched_normal_indel_check.txt
zgrep "##SAMPLE=<ID=TUMOUR" ${tmp[0]}*.pindel.annot.vcf.gz >> matched_tumour_indel_check.txt

done

paste --delimiters='\t' matched_tumour_indel_check.txt matched_normal_indel_check.txt > combined_matched_indel_check.txt

```

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering")

pin_file = read.table("combined_matched_indel_check.txt", header = F, sep = '\t', stringsAsFactors = F, comment.char = "")
pin_file$V1 = substr(lapply(strsplit(pin_file$V1, "SampleName="), "[[", 2), 0, 7)
pin_file$V2 = substr(lapply(strsplit(pin_file$V2, "SampleName="), "[[", 2), 0, 7)

pin_file[pin_file$V1 == pin_file$V2,] #each matched normal run against in silico in each vcf

```

#####Generate bed files

Generate bed files for snps and the unmatched pindel calls.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering

#generate bed file of indel regions to avoid
for f in $(ls *.pindel.annot.vcf.gz);
do
filename=$(basename -- "$f")
sample="${filename%.pindel.annot.vcf.gz}"
zless ${sample}.pindel.annot.vcf.gz | awk '! /\#/' | awk '{if(length($4) > length($5)) print $1"\t"($2-6)"\t"($2+length($4)+4); else print $1"\t"($2-6)"\t"($2+length($5)+4)}' > ${sample}.all_indels_5bp_window.bed
done

#generate bed file of SNP sites to avoid
for f in $(ls *.caveman_c.snps.vcf.gz);
do
filename=$(basename -- "$f")
sample="${filename%.caveman_c.snps.vcf.gz}"
echo $sample
zless $f | awk '! /\#/' | awk '{print $1"\t"($2-2)"\t"($2+1)}' > ${sample}.caveman_c.snps_1bp_window.bed
done

```

#####Intersect with VCFs

Remove variants that are called in matched normal or at SNP sites.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering

infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt"
grep -v 'Sample\|PD506' $infile | while IFS=$'\t' read -r -a tmp;
do
echo ${tmp[0]} #sample ID
echo ${tmp[2]} #matched normal ID
bedtools intersect -v -header -a ${tmp[0]}_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.vcf -b ${tmp[2]}.all_indels_5bp_window.bed > ${tmp[0]}_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.vcf
bedtools intersect -v -header -a ${tmp[0]}_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.vcf -b ${tmp[0]}.caveman_c.snps_1bp_window.bed > ${tmp[0]}_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.nosnp_1bp_window.annot.vcf
rm ${tmp[0]}_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.vcf
done

```

####VAGrENT annotation

```{bash}

module load vagrent

for f in $(ls *_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.nosnp_1bp_window.annot.vcf);
do
filename=$(basename -- "$f")
sample="${filename%_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.nosnp_1bp_window.annot.vcf}"
AnnotateVcf.pl -i ${sample}_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.nosnp_1bp_window.annot.vcf -o ${sample}_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.nosnp_1bp_window.annot.vagrent.annot.vcf -c /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/vagrent/e75/vagrent.cache.gz --species Human --assembly GRCh37d5
done

```

####Convert to text file

#####Functions

```{r}

#convert putative indels vcf into text file
#input file is *_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.nosnp_1bp_window.annot.vagrent.annot.vcf
lcm_pindel_vcf_to_text_no_filter = function(vcf_file){
  vcf = readVcf(vcf_file)
  sample = unlist(strsplit(vcf_file, '_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.nosnp_1bp_window.annot.vagrent.annot.vcf'))
  var.df = data.frame(Chr=as.character(seqnames(rowRanges(vcf))),
                  Position=start(ranges(rowRanges(vcf))),
                  Wildtype=as.character(ref(vcf)),
                  Mutant=as.character(unlist(CharacterList(alt(vcf)))),
                  Variant_read_depth=geno(vcf)$MTR[,2],
                  Total_read_depth=geno(vcf)$MTR[,2]+geno(vcf)$WTR[,2])
  var.df$VAF = var.df$Variant_read_depth / var.df$Total_read_depth
  var.df$Mutation = info(vcf)$VT
  vagrent_annots = lapply(strsplit(info(vcf)$VW, '\\|'), `length<-`, max(lengths(strsplit(info(vcf)$VW, '\\|')))) #make all elements in list the same length for subsetting purposes
  var.df$Gene = sapply(vagrent_annots, '[[', 1)
  var.df$Transcript = sapply(vagrent_annots, '[[', 3)
  var.df$Codon = sapply(vagrent_annots, '[[', 4)
  var.df$Protein = sapply(vagrent_annots, '[[', 5)
  
  conseq_list = lapply(info(vcf)$VC, function(x) if(identical(x, character(0))) NA_character_ else x)
  conseq_vec = c()
  for(i in 1:length(conseq_list)){
    conseq_vec = c(conseq_vec, paste(unlist(conseq_list[i]), collapse = ';'))
  }
  
  var.df$Consequence = conseq_vec
  var.df$Case_ID = substr(sample, 0, 7)
  var.df$Sample = sample
  
  write.table(var.df[, c("Case_ID", "Sample", "Chr", "Position", "Wildtype", "Mutant", "Total_read_depth", "Variant_read_depth", "VAF", "Mutation", "Gene",      "Transcript", "Codon", "Protein", "Consequence")], paste0(sample, '_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.nosnp_1bp_window.annot.vagrent.annot.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
  
}

```

#####Run

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/')

vcf_list = list.files('.', pattern = 'vcf')

for(i in vcf_list[!grepl('PD506', vcf_list)]){
  lcm_pindel_vcf_to_text_no_filter(i)
}

```

###-Structural variants

####BRASS

Initially run against matched normal (blood or normal testis).

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/01_BRASS

#original run
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/structural_variants/20200822/output/*_original_annotated.bed .

#re-run two samples
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/rerun_check/output/*_original_annotated.bed .

```

####Check that the BRASS files were matched

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/structural_variants/20200822/output

for f in $(ls *original_annotated.bed);
do
sed -n '9,10p' $f >> matched_run_check_20211231.txt
done

#PD40234b_lo0078 and PD42034b_lo0084 used the in silico normal instead, need re-running

```

PD40234b_lo0078 and PD42034b_lo0084 re-run

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/01_BRASS

for f in $(ls *original_annotated.bed);
do
sed -n '9,10p' $f >> matched_run_check_20220117.txt
done

```

####ms44 filters

#####Step 1 (annotate)

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/structural_variants/20200822

declare -a ponArray=()

while read line;
do
set $line
ponArray+=("$1")
done < pon_all.txt #tab file containing location of unrelated reference panel bam files

pon_list=$(echo "${ponArray[@]}" | tr ' ' ,)

while read line;
do
set $line
#echo $1 #sample ID
#echo $2 #project
#echo $3 #matched normal
bsub -G team294-vwork -J$1 -q small -M25000 -R"select[mem>25000] rusage[mem=25000] span[hosts=1]" -n 10 -e $1.stderr -o $1.stdout \
"/lustre/scratch116/casm/cgp/users/ms44/scripts/annotateBRASS_v4.sh /nfs/cancer_ref01/nst_links/live/$2/$1/$1.sample.dupmarked.bam /nfs/cancer_ref01/nst_links/live/$2/$1/$1.mt.brass.brm.bam /nfs/cancer_ref01/nst_links/live/$2/$1/$1.brass.annot.bedpe.gz ./output/${1}_original_annotated.bed /nfs/cancer_ref01/nst_links/live/$2/$3/$3.sample.dupmarked.bam $pon_list 10 /lustre/scratch119/casm/team274sb/to3/References/GRCh37d5/genome.fa /lustre/scratch119/casm/team294rr/to3/testes/tumour/structural_variants/20200822/common_all_20180423.vcf.gz"
done < ms44_sv_filter_input.txt #tab file with SAMPLE PROJECT MATCHED_NORMAL as three columns

#time limit hit for PD42036a_lo0001 and PD42036a_lo0007
declare -a ponArray=()

while read line;
do
set $line
ponArray+=("$1")
done < pon_all.txt

pon_list=$(echo "${ponArray[@]}" | tr ' ' ,)

for f in PD42036a_lo0001 PD42036a_lo0007;
do
#echo $1 #sample ID
#echo $2 #project
#echo $3 #matched normal
bsub -G team294-vwork -J${f} -q normal -M25000 -R"select[mem>25000] rusage[mem=25000] span[hosts=1]" -n 10 -e ${f}.stderr -o ${f}.stdout \
"/lustre/scratch116/casm/cgp/users/ms44/scripts/annotateBRASS_v4.sh /nfs/cancer_ref01/nst_links/live/1951/${f}/${f}.sample.dupmarked.bam /nfs/cancer_ref01/nst_links/live/1951/${f}/${f}.mt.brass.brm.bam /nfs/cancer_ref01/nst_links/live/1951/${f}/${f}.brass.annot.bedpe.gz ./output/${f}_original_annotated.bed /nfs/cancer_ref01/nst_links/live/1951/PD42036c/PD42036c.sample.dupmarked.bam $pon_list 10 /lustre/scratch119/casm/team274sb/to3/References/GRCh37d5/genome.fa /lustre/scratch119/casm/team294rr/to3/testes/tumour/structural_variants/20200822/common_all_20180423.vcf.gz"
done

```

Re-run for two samples

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/rerun_check

module unload java openjdk
module load openjdk/17.0.1

declare -a ponArray=()

while read line;
do
set $line
ponArray+=("$1")
done < /lustre/scratch119/casm/team294rr/to3/testes/tumour/structural_variants/20200822/pon_all.txt

pon_list=$(echo "${ponArray[@]}" | tr ' ' ,)

for f in PD42034b_lo0078 PD42034b_lo0084;
do
#echo $1 #sample ID
#echo $2 #project
#echo $3 #matched normal
bsub -G team294-vwork -J${f} -q normal -M25000 -R"select[mem>25000] rusage[mem=25000] span[hosts=1]" -n 10 -e ${f}.stderr -o ${f}.stdout \
"/lustre/scratch116/casm/cgp/users/ms44/scripts/annotateBRASS_v4.sh /nfs/cancer_ref01/nst_links/live/1951/${f}/${f}.sample.dupmarked.bam /nfs/cancer_ref01/nst_links/live/1951/${f}/${f}.mt.brass.brm.bam /nfs/cancer_ref01/nst_links/live/1951/${f}/${f}.brass.annot.bedpe.gz ./output/${f}_original_annotated.bed /nfs/cancer_ref01/nst_links/live/1951/PD42034d/PD42034d.sample.dupmarked.bam $pon_list 10 /lustre/scratch119/casm/team274sb/to3/References/GRCh37d5/genome.fa /lustre/scratch119/casm/team294rr/to3/testes/tumour/structural_variants/20200822/common_all_20180423.vcf.gz"
done

```

#####Step 2 (filter)

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/structural_variants/20200822/output

for f in *_annotated.bed; do grep "^#" $f > ${f/_original_annotated/_retained}; grep -v "^#" $f | awk -F$'\t' '{if(($49 >=4 && $50 >= 4 && $67 >= 3 && $68 >= 3) && ($51!~/NaN/ && $51 > 0 && $52!~/NaN/ && $52 > 0) && ($53 < 20 && $54 < 20 && ($53 + $54) <= 30) && ((($59/$47) <= 0.5 && ($61/$47) <= 0.5) || (($60/$48) <= 0.5 && ($62/$48) <= 0.5)) && ($63 < 0.075 && $64 < 0.075) && (($65 == 0 || (($71/$65) < 0.5 && ($73/$65) < 0.5 && $67 >= 3)) || ($66 == 0 || (($72/$66) < 0.5 && ($74/$66) < 0.5 && $68 >= 3))) && ($75~/false/ || $76 <= 1) && ($77~/false/ || (($79/$78) < (1/3)) && $79 == 0) && (($84 < (0.15) || $84~NaN) && ($85 < (0.15) || $85~NaN)) && ($65 > 0 && ($86/$65) < 0.5) && ($66 > 0 && ($87/$66) < 0.5)){print}else{}}' >> ${f/_original_annotated/_retained}; done

```

Re-run for two samples

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/rerun_check/output

for f in *_annotated.bed; do grep "^#" $f > ${f/_original_annotated/_retained}; grep -v "^#" $f | awk -F$'\t' '{if(($49 >=4 && $50 >= 4 && $67 >= 3 && $68 >= 3) && ($51!~/NaN/ && $51 > 0 && $52!~/NaN/ && $52 > 0) && ($53 < 20 && $54 < 20 && ($53 + $54) <= 30) && ((($59/$47) <= 0.5 && ($61/$47) <= 0.5) || (($60/$48) <= 0.5 && ($62/$48) <= 0.5)) && ($63 < 0.075 && $64 < 0.075) && (($65 == 0 || (($71/$65) < 0.5 && ($73/$65) < 0.5 && $67 >= 3)) || ($66 == 0 || (($72/$66) < 0.5 && ($74/$66) < 0.5 && $68 >= 3))) && ($75~/false/ || $76 <= 1) && ($77~/false/ || (($79/$78) < (1/3)) && $79 == 0) && (($84 < (0.15) || $84~NaN) && ($85 < (0.15) || $85~NaN)) && ($65 > 0 && ($86/$65) < 0.5) && ($66 > 0 && ($87/$66) < 0.5)){print}else{}}' >> ${f/_original_annotated/_retained}; done

```

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/02_LCM

#original run
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/structural_variants/20200822/output/*_retained.bed .

#re-run two samples
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/rerun_check/output/*_retained.bed .

```

#####Generate variant text file for further manual curation + annotation

First, we need to gather all structural variants that passed BRASS and the ms44 filters into a single file.

```{bash}

#cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/structural_variants/20200822/output

#for f in $(ls *_retained.bed);
#do
#grep -v '#' $f 
#done >> post_ms44_svs.bed

#grep '# chr1' PD46969e_lo0005_retained.bed > post_ms44_svs.txt
#sort -k1,2 post_ms44_svs.bed >> post_ms44_svs.txt

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/02_LCM

for f in $(ls *_retained.bed);
do
grep -v '#' $f 
done >> post_ms44_svs.bed

grep '# chr1' PD46969e_lo0005_retained.bed > post_ms44_svs.txt
sort -k1,2 post_ms44_svs.bed >> post_ms44_svs.txt #remove hash from header 20220117

```

Review the variants called across all samples
```{r}

#setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/structural_variants/20200822/output')
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/02_LCM')

reassembled <- read.table('post_ms44_svs.txt', header = T, sep = '\t', stringsAsFactors = F, comment.char = '') #2375 as of 6/9/20, 2377 as of 17/1/22

#annotate genes contained within deletions and tandem-duplications
##generate bed file

annot_sv <- reassembled[reassembled$svclass %in% c('deletion', 'tandem-duplication'),] #1437 SVs
#write.table(annot_sv[, c('chr1', 'start1', 'end2')], 'gct_dup_del_coord_20200906.bed', col.names = F, row.names = F, quote = F, sep = '\t')
write.table(annot_sv[, c('chr1', 'start1', 'end2')], 'gct_dup_del_coord_20220117.bed', col.names = F, row.names = F, quote = F, sep = '\t')

```

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/02_LCM

module load bedtools
#~/scripts/intersect_intv.sh gct_dup_del_coord_20200906.bed gct_sv_genes_20200906.txt
~/scripts/intersect_intv.sh gct_dup_del_coord_20220117.bed gct_sv_genes_20220117.txt

```

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/02_LCM")

sv_genes <- read.table('gct_sv_genes_20220117.txt', header = T, stringsAsFactors = F, sep = '\t')
#sv_genes <- read.table('gct_sv_genes_20200906.txt', header = T, stringsAsFactors = F, sep = '\t')
sv_genes <- sv_genes[!(duplicated(sv_genes)),]
sv_genes$Chr <- unlist(lapply(strsplit(sv_genes$Chromosome, ':'), "[[", 1))
#sv_genes$Chr <- t(data.frame(strsplit(sv_genes$Chromosome, ':')))[,1]
sv_genes$Gene.Start <- as.numeric(unlist(lapply(strsplit(unlist(lapply(strsplit(sv_genes$Chromosome, ':'), "[[", 2)), "-"), "[[", 1)))
sv_genes$Gene.End <- as.numeric(unlist(lapply(strsplit(unlist(lapply(strsplit(sv_genes$Chromosome, ':'), "[[", 2)), "-"), "[[", 2)))
#sv_genes$Gene.Start <- as.numeric(t(data.frame(strsplit(t(data.frame(strsplit(sv_genes$Chromosome, ':')))[,2], '-')))[,1])
#sv_genes$Gene.End <- as.numeric(t(data.frame(strsplit(t(data.frame(strsplit(sv_genes$Chromosome, ':')))[,2], '-')))[,2])


reassembled$Genes.in.dup.del <- NA
annot_sv$Genes.in.dup.del <- NA

for(i in 1:nrow(sv_genes)){
  if(length(annot_sv[annot_sv$chr1 == sv_genes$Chr[i] & annot_sv$start1 < sv_genes$Gene.End[i] & annot_sv$end2 > sv_genes$Gene.Start[i],]$Genes.in.dup.del)){
    annot_sv[annot_sv$chr1 == sv_genes$Chr[i] & annot_sv$start1 < sv_genes$Gene.End[i] & annot_sv$end2 > sv_genes$Gene.Start[i],]$Genes.in.dup.del = paste(annot_sv[annot_sv$chr1 == sv_genes$Chr[i] & annot_sv$start1 < sv_genes$Gene.End[i] & annot_sv$end2 > sv_genes$Gene.Start[i],]$Genes.in.dup.del, sv_genes$GeneSymbol[i], sep = ',')
  }
}

##now remove aberrant NAs after paste function
max(nchar(annot_sv$Genes.in.dup.del), na.rm = T) #5816 characters at longest
annot_sv[!is.na(annot_sv$Genes.in.dup.del),]$Genes.in.dup.del <- substr(annot_sv[!is.na(annot_sv$Genes.in.dup.del),]$Genes.in.dup.del, 4, 5817) 

##add back to main SV list
#for(i in 1:nrow(reassembled)){
#  if(reassembled$assembled.readnames[i] %in% annot_sv$assembled.readnames){
#    reassembled$Genes.in.dup.del[i] = annot_sv[annot_sv$assembled.readnames == reassembled$assembled.readnames[i],]$Genes.in.dup.del
#  }
#}

for(i in 1:nrow(reassembled)){
  if(reassembled$chr1[i] %in% annot_sv$chr1 & reassembled$start1[i] %in% annot_sv$start1 & reassembled$end1[i] %in% annot_sv$end1 & reassembled$chr2[i] %in% annot_sv$chr2 & reassembled$start2[i] %in% annot_sv$start2 & reassembled$end2[i] %in% annot_sv$end2 & reassembled$sample[i] %in% annot_sv$sample){
    reassembled$Genes.in.dup.del[i] = annot_sv[annot_sv$chr1 == reassembled$chr1[i] & annot_sv$start1 == reassembled$start1[i] & annot_sv$end1 == reassembled$end1[i] & annot_sv$chr2 == reassembled$chr2[i] & annot_sv$start2 == reassembled$start2[i] & annot_sv$end2 == reassembled$end2[i] & annot_sv$sample == reassembled$sample[i],]$Genes.in.dup.del[1]
  }
}

#check if breakpoints generally or regions within intrachromosomal SVs are cancer genes
##Intersect with COSMIC v94 consensus cancer gene list, previously v90
cosmic_v94 <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/cosmic_v94_cancer_gene_census.csv', sep = ',', header = T, stringsAsFactors = F)
cancer_genes <- unlist(strsplit(paste(cosmic_v94$Synonyms, collapse = ","), ","))

#cosmic_v90 <- read.table('/lustre/scratch119/casm/team274sb/to3/placenta/reference_datasets/COSMICv90_cancer_gene_consensus.tsv', sep = '\t', header = T, stringsAsFactors = F)
#all_sv_genes <- c(reassembled$gene1, reassembled$gene2, sv_genes$GeneSymbol)
#cancer_genes <- unique(all_sv_genes[all_sv_genes %in% cosmic_v90$Gene.Symbol])
#cancer_genes <- cosmic_v90$Gene.Symbol


##add flagged cancer genes to list 
#reassembled = read.table('post_ms44_svs_final_20200518.txt', header = T, stringsAsFactors = F, sep = '\t')
reassembled$Cancer.gene <- NA
#for(i in cancer_genes){
#  reassembled[grepl(pattern = paste0(i), x = reassembled$gene1) | grepl(pattern = paste0(i), x = reassembled$gene2) | grepl(pattern = paste0(i), x = reassembled$Genes.in.dup.del),]$Cancer.gene = paste(reassembled[grepl(pattern = paste0(i), x = reassembled$gene1) | grepl(pattern = paste0(i), x = reassembled$gene2) | grepl(pattern = paste0(i), x = reassembled$Genes.in.dup.del),]$Cancer.gene, i, sep = ',')
#}

for(i in 1:nrow(reassembled)){
  gene_list = unlist(strsplit(reassembled$Genes.in.dup.del[i], ','))
  gene_list = c(gene_list, reassembled$gene1[i], reassembled$gene2[i])
  gene_list = unique(gene_list)
  gene_list = cancer_genes[which(as.vector(cancer_genes) %in% as.vector(gene_list))]
  if(length(gene_list)){
    reassembled$Cancer.gene[i] = paste(gene_list, collapse = ',')
  }
  else {
    reassembled$Cancer.gene[i] = NA
  }
}

reassembled$Cancer.gene

#annotate whether the cancer genes are dominant or recessive
reassembled$Dom.Rec.COSMICv94 <- NA
for(i in 1:nrow(reassembled)){
  if(!is.na(reassembled$Cancer.gene[i])){
    gene.list <- unlist(strsplit(reassembled$Cancer.gene[i], ','))
    reassembled$Dom.Rec.COSMICv94[i] = paste(cosmic_v94[cosmic_v94$Gene.Symbol %in% gene.list, ]$Molecular.Genetics, collapse = ',')
  }
}

#fetch the 20bp upstream of the lower co-ordinate and downstream of higher co-ordinates, unfortunately brass uses 0-based co-ordinates at the start and 1-based at the end...
head(reassembled)
bed_coord <- cbind(reassembled$chr1, (reassembled$start1 - 10), reassembled$start1 + 10)
bed_coord <- rbind(bed_coord, cbind(reassembled$chr1, (reassembled$end1 - 10), reassembled$end1 + 10))
bed_coord <- rbind(bed_coord, cbind(reassembled$chr2, (reassembled$start2 - 10), (reassembled$start2 + 10)))
bed_coord <- rbind(bed_coord, cbind(reassembled$chr2, (reassembled$end2 - 10), (reassembled$end2 + 10)))

#write.table(bed_coord, '20bp_NT_context_gct_SVs_20200906.bed', quote = F, sep = '\t', row.names = F, col.names = F)
write.table(bed_coord, '20bp_NT_context_gct_SVs_20220117.bed', quote = F, sep = '\t', row.names = F, col.names = F)

```


```{bash}
#use bedtools to get context from reference genome

module load bedtools
#bedtools getfasta -fi /lustre/scratch119/casm/team274sb/to3/References/GRCh37d5/genome.fa -fo 20bp_NT_context_gct_SVs_20200906.fa -bed 20bp_NT_context_gct_SVs_20200906.bed 
bedtools getfasta -fi /lustre/scratch119/casm/team274sb/to3/References/GRCh37d5/genome.fa -fo 20bp_NT_context_gct_SVs_20220117.fa -bed 20bp_NT_context_gct_SVs_20220117.bed 

```

```{r}

#read in fasta file
library(seqinr)
#fasta <- read.fasta('/lustre/scratch119/casm/team294rr/to3/testes/tumour/structural_variants/20200822/output/20bp_NT_context_gct_SVs_20200906.fa')
fasta <- read.fasta('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/02_LCM/20bp_NT_context_gct_SVs_20220117.fa')

#convert to dataframe
context <- data.frame(Fragments=names(fasta), Seqs=toupper(unlist(getSequence(fasta, as.string=T))), stringsAsFactors = F)
context$Chr <- t(data.frame(strsplit(context$Fragments, ':')))[,1]
context$Start <- as.numeric(t(data.frame(strsplit(t(data.frame(strsplit(context$Fragments, ':')))[,2], '-')))[,1])
context$End <- as.numeric(t(data.frame(strsplit(t(data.frame(strsplit(context$Fragments, ':')))[,2], '-')))[,2])

#add in the 20bp context to the BRASS co-ordinates
reassembled$start1_20bp_nt_context <- NA
reassembled$end1_20bp_nt_context <- NA
reassembled$start2_20bp_nt_context <- NA
reassembled$end2_20bp_nt_context <- NA

for(i in 1:nrow(reassembled)){
  reassembled$start1_20bp_nt_context[i] = context[context$Chr == reassembled$chr1[i] & context$End == (reassembled$start1[i] + 10),]$Seqs[1]
  reassembled$end1_20bp_nt_context[i] = context[context$Chr == reassembled$chr1[i] & context$End == (reassembled$end1[i] + 10),]$Seqs[1]
  reassembled$start2_20bp_nt_context[i] = context[context$Chr == reassembled$chr2[i] & context$Start == (reassembled$start2[i] - 10),]$Seqs[1]
  reassembled$end2_20bp_nt_context[i] = context[context$Chr == reassembled$chr2[i] & context$Start == (reassembled$end2[i] - 10),]$Seqs[1]
}

reassembled$sample_nomatched <- as.vector(t(data.frame(strsplit(reassembled$sample, ',')))[,1])
reassembled$patient <- substr(reassembled$sample_nomatched, 0, 7)

#write.table(reassembled, 'post_ms44_svs_for_inspection_20200906.txt', sep = '\t', quote = F, row.names = F)
write.table(reassembled, 'post_ms44_svs_for_inspection_20220117.txt', sep = '\t', quote = F, row.names = F)

```

Review edge cases and small structural variants
```{r}

#write.table(reassembled[(reassembled$bkdist != -1 & reassembled$bkdist < 1000) | (reassembled$Unique.left.reads < 6 & (reassembled$Mismatched.Supplementary.Alignment.BRASS.left == 'true' | reassembled$Incongruent.Supplementary.Alignment.BRASS.left == 'true')) | (reassembled$Unique.right.reads < 6 & (reassembled$Mismatched.Supplementary.Alignment.BRASS.right == 'true' | reassembled$Incongruent.Supplementary.Alignment.BRASS.right == 'true')), ], 'post_ms44_filter_edge_cases_20200906.txt', sep = '\t', quote = F, row.names = F, col.names = T)
write.table(reassembled[(reassembled$bkdist != -1 & reassembled$bkdist < 1000) | (reassembled$Unique.left.reads < 6 & (reassembled$Mismatched.Supplementary.Alignment.BRASS.left == 'true' | reassembled$Incongruent.Supplementary.Alignment.BRASS.left == 'true')) | (reassembled$Unique.right.reads < 6 & (reassembled$Mismatched.Supplementary.Alignment.BRASS.right == 'true' | reassembled$Incongruent.Supplementary.Alignment.BRASS.right == 'true')), ], 'post_ms44_filter_edge_cases_20220117.txt', sep = '\t', quote = F, row.names = F, col.names = T)

```

Compare the size of the original file and the current one.

```{r}

old = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/structural_variants/20200822/output/post_ms44_filter_edge_cases_20200906.txt", header = T, sep = '\t', stringsAsFactors = F)
new = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/02_LCM/post_ms44_filter_edge_cases_20220117.txt", header = T, sep = '\t', stringsAsFactors = F)

old$mutID = paste(old$chr1, old$start1, old$end1, old$chr2, old$start2, old$end2, sep = "_")
new$mutID = paste(new$chr1, new$start1, new$end1, new$chr2, new$start2, new$end2, sep = "_")

sum(!new$mutID %in% old$mutID) #0 all previously identified as edge cases, meaning the two new variants pass the borderline filter
old[!old$mutID %in% new$mutID, 1:6] #none

```

Filter out artefactual edge cases

```{r}

#library(dplyr)
#setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/final_variant_calls/SV/pre_filtering')
#reviewed = read.table('post_ms44_filter_edge_cases_20200906.txt', header = T, sep = '\t', stringsAsFactors = F)
#all_filtered_svs = read.table('post_ms44_svs_for_inspection_20200906.txt', header = T, sep = '\t', stringsAsFactors = F)
#to_remove = unique(reviewed[reviewed$keep == 'N', 1:6])
#final_svs = anti_join(all_filtered_svs, to_remove, by = c("chr1", "start1", "end1", "chr2", "start2", "end2"))
#write.table(final_svs, 'filtered_reviewed_svs_20200906.txt', quote = F, col.names = T, row.names = F, sep = '\t')

library(dplyr)

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/02_LCM')

reviewed = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/final_variant_calls/SV/pre_filtering/post_ms44_filter_edge_cases_20200906.txt', header = T, sep = '\t', stringsAsFactors = F) #same as the most recent one 20220117
all_filtered_svs = read.table('post_ms44_svs_for_inspection_20220117.txt', header = T, sep = '\t', stringsAsFactors = F)

to_remove = unique(reviewed[reviewed$keep == 'N', 1:6])

final_svs = anti_join(all_filtered_svs, to_remove, by = c("chr1", "start1", "end1", "chr2", "start2", "end2"))

#write.table(final_svs, 'filtered_reviewed_svs_20200906.txt', quote = F, col.names = T, row.names = F, sep = '\t')
write.table(final_svs, 'filtered_reviewed_svs_20220117.txt', quote = F, col.names = T, row.names = F, sep = '\t')

```

Check that only 2 additional variants in new file
wc -l /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/02_LCM/filtered_reviewed_svs_20220117.txt 2318
head -1 /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/02_LCM/filtered_reviewed_svs_20220117.txt
wc -l /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/final_variant_calls/SV/filtered_reviewed_svs_20200906.txt 2315
head -1 /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/final_variant_calls/SV/filtered_reviewed_svs_20200906.txt

1 more than expected... why?

```{r}

old = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/final_variant_calls/SV/filtered_reviewed_svs_20200906.txt", header = T, sep = '\t', stringsAsFactors = F)
new = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/02_LCM/filtered_reviewed_svs_20220117.txt", header = T, sep = '\t', stringsAsFactors = F)

dim(old) #2315
dim(new) #2317, looks correct, wc -l for old file off by one...

old$mutID = paste(old$chr1, old$start1, old$end1, old$chr2, old$start2, old$end2, sep = "_")
new$mutID = paste(new$chr1, new$start1, new$end1, new$chr2, new$start2, new$end2, sep = "_")

anti_join(new[,1:6], old[,1:6]) #other differences due to slight different in the start and end coordinates when run matched
#new[!new$mutID %in% old$mutID, 1:6]

new[!new$mutID %in% old$mutID,]

```

####Retrotransposition filter

Remove variants that who have a breakpoint within 100bp of a retrotranposition event.

```{r}

rt_coord = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/retrotransposition_coordinates_20211231.txt", header = T, sep = '\t', stringsAsFactors = F) #coordinates of all rt breakpoints
#rt_coord = read.table('bulk_yst_rt_coordinates.txt', header = T, sep = '\t', stringsAsFactors = F) 

standard_sv = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/02_LCM/filtered_reviewed_svs_20220117.txt', header = T, sep = '\t', stringsAsFactors = F) #some of these in BRASS may have been mistakenly called as standard structural variants

#remove all SVs that occur with 100bp of a called RT
for(j in 1:nrow(rt_coord)){
  if(nrow(standard_sv[standard_sv$sample_nomatched == rt_coord$sample[j] & standard_sv$chr1 == rt_coord$chr[j] & standard_sv$start1 >= (rt_coord$breakpoint[j] - 100) & standard_sv$start1 <= (rt_coord$breakpoint[j] + 100), ])){
    standard_sv[standard_sv$sample_nomatched == rt_coord$sample[j] & standard_sv$chr1 == rt_coord$chr[j] & standard_sv$start1 >= (rt_coord$breakpoint[j] - 100) & standard_sv$start1 <= (rt_coord$breakpoint[j] + 100), ]$svclass = 'retrotransposition'
  }
  if(nrow(standard_sv[standard_sv$sample_nomatched == rt_coord$sample[j] & standard_sv$chr1 == rt_coord$chr[j] & standard_sv$end1 >= (rt_coord$breakpoint[j] - 100) & standard_sv$end1 <= (rt_coord$breakpoint[j] + 100), ])){
    standard_sv[standard_sv$sample_nomatched == rt_coord$sample[j] & standard_sv$chr1 == rt_coord$chr[j] & standard_sv$end1 >= (rt_coord$breakpoint[j] - 100) & standard_sv$end1 <= (rt_coord$breakpoint[j] + 100), ]$svclass = 'retrotransposition'
  }
  if(nrow(standard_sv[standard_sv$sample_nomatched == rt_coord$sample[j] & standard_sv$chr2 == rt_coord$chr[j] & standard_sv$start2 >= (rt_coord$breakpoint[j] - 100) & standard_sv$start2 <= (rt_coord$breakpoint[j] + 100), ])){
    standard_sv[standard_sv$sample_nomatched == rt_coord$sample[j] & standard_sv$chr2 == rt_coord$chr[j] & standard_sv$start2 >= (rt_coord$breakpoint[j] - 100) & standard_sv$start2 <= (rt_coord$breakpoint[j] + 100), ]$svclass = 'retrotransposition'
  }
  if(nrow(standard_sv[standard_sv$sample_nomatched == rt_coord$sample[j] & standard_sv$chr2 == rt_coord$chr[j] & standard_sv$end2 >= (rt_coord$breakpoint[j] - 100) & standard_sv$end2 <= (rt_coord$breakpoint[j] + 100), ])){
    standard_sv[standard_sv$sample_nomatched == rt_coord$sample[j] & standard_sv$chr2 == rt_coord$chr[j] & standard_sv$end2 >= (rt_coord$breakpoint[j] - 100) & standard_sv$end2 <= (rt_coord$breakpoint[j] + 100), ]$svclass = 'retrotransposition'
  }
}

no_rt = standard_sv[standard_sv$svclass != 'retrotransposition',] #17 labelled as RT
write.table(no_rt, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/03_rt/final_lcm_gct_brass_calls_20220118.txt', sep ='\t', quote = F, col.names = T, row.names = F)

```

##VARIANT CALLING AND FILTERING (bulk)

###-Substitutions

####CaVEMan

Substitutions called using CaVEMan against matched normal sample (blood).

#####Check all files are matched correctly

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/subs/bulk/01_caveman

zgrep "##SAMPLE=<ID=NORMAL" *.caveman_c.annot.vcf.gz > matched_normal_check.txt
zgrep "##SAMPLE=<ID=TUMOUR" *.caveman_c.annot.vcf.gz > matched_tumour_check.txt
paste --delimiters='\t' matched_tumour_check.txt matched_normal_check.txt > combined_matched_check.txt

```

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/subs/bulk/01_caveman")

cv_file = read.table("combined_matched_check.txt", header = F, sep = '\t', stringsAsFactors = F, comment.char = "")
cv_file$V1 = substr(lapply(strsplit(cv_file$V1, "SampleName="), "[[", 2), 0, 7)
cv_file$V2 = substr(lapply(strsplit(cv_file$V2, "SampleName="), "[[", 2), 0, 7)

cv_file[cv_file$V1 != cv_file$V2,] #tumour and normal match in each vcf

```

#####Check files not truncated

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/subs/bulk/01_caveman

ls *.caveman_c.annot.vcf.gz | while read FILE ; do zgrep "^#" -v "$FILE" | cut -f1 | uniq -c > ../00_qc_chrom/"$FILE".sub.counts.txt ; done

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/subs/bulk/00_qc_chrom
wc -l *.sub.counts.txt

#don't seem truncated 28/12/21

```

####Standard filters

#####Functions

4+ variant reads, 10+ coverage, PASS filter, 140+ ASMD, 0 CLPM.

```{r}

library(VariantAnnotation)

#convert putative subs vcf into text file
#input file is *.caveman_c.annot.vcf.gz

bulk_caveman_vcf_to_text_no_filter = function(vcf_file){
  vcf = readVcf(vcf_file)
  sample = unlist(strsplit(vcf_file, '.caveman_c.annot.vcf.gz'))
  reads = cbind((geno(vcf)$FAZ)[,2]+(geno(vcf)$RAZ)[,2],(geno(vcf)$FCZ)[,2]+(geno(vcf)$RCZ)[,2],
                (geno(vcf)$FGZ)[,2]+(geno(vcf)$RGZ)[,2],(geno(vcf)$FTZ)[,2]+(geno(vcf)$RTZ)[,2])
  var.df = data.frame(chr = as.character(seqnames(rowRanges(vcf))),
                    pos = start(ranges(rowRanges(vcf))),
                    ref = as.character(ref(vcf)),
                    alt = unlist(CharacterList(alt(vcf))))
  var.df$NR=rowSums(reads)
  var.df$NV=NA
  colnames(reads)=c("A","C","G","T")
  for (k in c("A","C","G","T")){
    var.df$NV[var.df$alt==k] = reads[var.df$alt==k,k]
  }
  var.df$VAF = var.df$NV/ var.df$NR
  var.df$ASMD = info(vcf)$ASMD
  var.df$CLPM = info(vcf)$CLPM
  var.df$FILTER = rowRanges(vcf)$FILTER
  var.df$vagrent_default = info(vcf)$VD
  var.df$vagrent_most_del = info(vcf)$VW
  var.df$variant_type = info(vcf)$VT
  var.df$variant_consequence = unlist(lapply(info(vcf)$VC, function(x) if(identical(x, character(0))) NA_character_ else x))
  write.table(var.df, paste0(sample, '.caveman_c.annot_nofilter.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
}

#apply main filters to putative subs
#input file is *.caveman_c.annot_nofilter.txt
 
bulk_caveman_txt_filter = function(txt_file){
  muts = read.table(txt_file, header = T, sep = '\t', stringsAsFactors = F)
  sample = unlist(strsplit(txt_file, '.caveman_c.annot_nofilter.txt'))
  row.names(muts) = paste(muts$chr, muts$pos, muts$ref, muts$alt, sep = '_')
  filtered_subs = muts[(muts$NV >= 4) & (muts$NR >= 10) & (muts$ASMD >= 140) & (muts$CLPM == 0) & (muts$FILTER == 'PASS'),]
  write.table(filtered_subs, paste0(sample, '.caveman_c.annot_filtered.txt'), col.names = T, row.names = T, sep = '\t', quote = F)
}

```

#####Run

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/subs')

vcf_files = list.files('.', '.caveman_c.annot.vcf.gz$')

for(vcf_file in vcf_files){
  sample = unlist(strsplit(vcf_file, '.caveman_c.annot.vcf.gz'))
  bulk_caveman_vcf_to_text_no_filter(paste0(sample, '.caveman_c.annot.vcf.gz'))
  bulk_caveman_txt_filter(paste0(sample, '.caveman_c.annot_nofilter.txt'))
}

```

Generate the vcf filtered to contain only mutations that passed new soft filters.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/subs

#generate a bed file to intersect with the pre-filtered vcf
for file in $(ls *.caveman_c.annot_filtered.txt);
do
filename=$(basename -- "$file")
SAMPLE="${filename%.caveman_c.annot_filtered.txt}"
cp $file ${SAMPLE}.caveman_c.annot_filtered.bed
sed -i '1d' ${SAMPLE}.caveman_c.annot_filtered.bed
awk 'BEGIN{FS=OFS="\t"}{print $2, ($3 - 1), $3}' ${SAMPLE}.caveman_c.annot_filtered.bed > ${SAMPLE}.caveman_c.annot_filtered_bedtools.bed
mv ${SAMPLE}.caveman_c.annot_filtered_bedtools.bed ${SAMPLE}.caveman_c.annot_filtered.bed
done

#intersect this bed file with the relevant vcf
module load bedtools

for file in $(ls *.caveman_c.annot_filtered.bed);
do
filename=$(basename -- "$file")
SAMPLE="${filename%.caveman_c.annot_filtered.bed}"
bedtools intersect -a ${SAMPLE}.caveman_c.annot.vcf.gz -b ${SAMPLE}.caveman_c.annot_filtered.bed -u -header > ${SAMPLE}.caveman_c.annot_filtered.vcf
done

```

Check the text and vcf files are the same length.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/subs/bulk/02_standard

for f in $(ls *);
do
#filename=$(basename -- "$f")
echo $f
grep -v "#" $f | wc -l
done

#looks good 28/12/21, only the header of the text file is extra

```

###-Indels

####Pindel

Indels called using pindel against matched normal sample (blood).

#####Check all files are matched correctly

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/indels/bulk/01_pindel

zgrep "##SAMPLE=<ID=NORMAL" *.pindel.annot.vcf.gz > matched_normal_check.txt
zgrep "##SAMPLE=<ID=TUMOUR" *.pindel.annot.vcf.gz > matched_tumour_check.txt
paste --delimiters='\t' matched_tumour_check.txt matched_normal_check.txt > combined_matched_check.txt

```

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/indels/bulk/01_pindel")

pin_file = read.table("combined_matched_check.txt", header = F, sep = '\t', stringsAsFactors = F, comment.char = "")
pin_file$V1 = substr(lapply(strsplit(pin_file$V1, "SampleName="), "[[", 2), 0, 7)
pin_file$V2 = substr(lapply(strsplit(pin_file$V2, "SampleName="), "[[", 2), 0, 7)

pin_file[pin_file$V1 != pin_file$V2,] #tumour and normal match in each vcf

```

#####Check files not truncated

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/indels/bulk/01_pindel

ls *.pindel.annot.vcf.gz | while read FILE ; do zgrep "^#" -v "$FILE" | cut -f1 | uniq -c > ../00_qc_chrom/"$FILE".indel.counts.txt ; done

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/indels/bulk/00_qc_chrom
wc -l *.indel.counts.txt

#looks fine 28/12/21

```

####Homopolymer filtering

Filter original pindel files to only include those that pass pindel and are not 1bp events occurring on homopolymer runs longer than 6bp.

```{bash}

cd /nfs/users/nfs_t/to3/scripts/pindel-filtering

infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt"
grep 'PD506' $infile | while IFS=$'\t' read -r -a tmp ; 
do
#echo "${tmp[0]}" #sample id
#echo "${tmp[2]}" #project id
#echo "${tmp[3]}" #matched normal id
./bsub_homopolymer_filter_args.sh 2127 /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/indels/ 7 PASS ${tmp[0]}
done

```

####Apply standard filters

Only keep variants that have have a total read depth of 10 and variant read depth of at least 5 with a QUAL score of at least 300. Also added the requirement to have at least one supporting variant read in both directions.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/indels

module load bcftools

for i in $(ls *.pindel.HP.7.vcf);
do
filename=$(basename -- "$i")
SAMPLE="${filename%.pindel.HP.7.vcf}"
bcftools filter -i 'QUAL >= 300 & (FORMAT/NU[1] + FORMAT/PU[1]) >= 5 & (FORMAT/NR[1] + FORMAT/PR[1]) >= 10 & FORMAT/NU[1] >= 1 & FORMAT/PU[1] >= 1' ${SAMPLE}.pindel.HP.7.vcf > ${SAMPLE}.pindel.pass.hp7_threshold.qual.depth.direction.filtered.vcf
done

```

####SNP and matched normal pindel filtering

#####Check SNP files are matched

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/indels

zgrep "##SAMPLE=<ID=NORMAL" *.caveman_c.snps.vcf.gz > matched_normal_snp_check.txt
zgrep "##SAMPLE=<ID=TUMOUR" *.caveman_c.snps.vcf.gz > matched_tumour_snp_check.txt
paste --delimiters='\t' matched_tumour_snp_check.txt matched_normal_snp_check.txt > combined_matched_snp_check.txt

```

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering")

cv_file = read.table("combined_matched_snp_check.txt", header = F, sep = '\t', stringsAsFactors = F, comment.char = "")
cv_file$V1 = substr(lapply(strsplit(cv_file$V1, "SampleName="), "[[", 2), 0, 7)
cv_file$V2 = substr(lapply(strsplit(cv_file$V2, "SampleName="), "[[", 2), 0, 7)

cv_file[cv_file$V1 != cv_file$V2,] #tumour and normal match in each vcf

```

#####Check indel files from matched normal were run unmatched

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering

infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt"
grep -v 'Sample\|PD506' $infile | cut -f3 | uniq |while IFS=$'\t' read -r -a tmp;
do
echo ${tmp[0]} #matched normal ID

zgrep "##SAMPLE=<ID=NORMAL" ${tmp[0]}*.pindel.annot.vcf.gz >> matched_normal_indel_check.txt
zgrep "##SAMPLE=<ID=TUMOUR" ${tmp[0]}*.pindel.annot.vcf.gz >> matched_tumour_indel_check.txt

done

paste --delimiters='\t' matched_tumour_indel_check.txt matched_normal_indel_check.txt > combined_matched_indel_check.txt

```

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering")

pin_file = read.table("combined_matched_indel_check.txt", header = F, sep = '\t', stringsAsFactors = F, comment.char = "")
pin_file$V1 = substr(lapply(strsplit(pin_file$V1, "SampleName="), "[[", 2), 0, 7)
pin_file$V2 = substr(lapply(strsplit(pin_file$V2, "SampleName="), "[[", 2), 0, 7)

pin_file[pin_file$V1 == pin_file$V2,] #each matched normal run against in silico in each vcf

```

#####Generate bed files

Generate bed files for snps and the unmatched pindel calls from the matched sample.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/indels

#generate bed file of indel regions to avoid
for f in $(ls *b.pindel.annot.vcf.gz);
do
filename=$(basename -- "$f")
sample="${filename%b.pindel.annot.vcf.gz}"
zless ${sample}b.pindel.annot.vcf.gz | awk '! /\#/' | awk '{if(length($4) > length($5)) print $1"\t"($2-6)"\t"($2+length($4)+4); else print $1"\t"($2-6)"\t"($2+length($5)+4)}' > ${sample}b.all_indels_5bp_window.bed
done

#generate bed file of SNP sites to avoid
for f in $(ls *.caveman_c.snps.vcf.gz);
do
filename=$(basename -- "$f")
sample="${filename%.caveman_c.snps.vcf.gz}"
zless ${sample}.caveman_c.snps.vcf.gz | awk '! /\#/' | awk '{print $1"\t"($2-2)"\t"($2+1)}' > ${sample}.caveman_c.snps_1bp_window.bed
done

```

#####Intersect with VCFs

Now filter out remaining indels that lie across indel calls from matched blood file or SNPs.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/indels

for f in $(ls *.pindel.pass.hp7_threshold.qual.depth.direction.filtered.vcf);
do
filename=$(basename -- "$f")
sample="${filename%a.pindel.pass.hp7_threshold.qual.depth.direction.filtered.vcf}"
bedtools intersect -v -header -a ${sample}a.pindel.pass.hp7_threshold.qual.depth.direction.filtered.vcf -b ${sample}b.all_indels_5bp_window.bed > ${sample}a.pindel.pass.hp7_threshold.qual.depth.direction.filtered.nogermline.5bp_window.annot.vcf
bedtools intersect -v -header -a ${sample}a.pindel.pass.hp7_threshold.qual.depth.direction.filtered.nogermline.5bp_window.annot.vcf -b ${sample}a.caveman_c.snps_1bp_window.bed > ${sample}a.pindel.pass.hp7_threshold.qual.depth.direction.filtered.nogermline.5bp_window.nosnp_1bp_window.annot.vcf
rm ${sample}a.pindel.pass.hp7_threshold.qual.depth.direction.filtered.nogermline.5bp_window.annot.vcf
done

```

###Structural variants

####BRASS

#####Check all files are matched correctly

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/bulk/01_BRASS

for f in $(ls *.brass.annot.bedpe.gz);
do
zless $f | sed -n '9,10p' >> matched_run_check_20211231.txt
done

#look good 31/12/21

```

####Score filter

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/structural_variants

zless /nfs/cancer_ref01/nst_links/live/2127/PD506*a/*brass.annot.bedpe.gz | grep score | grep -v "#" > bulk_yst_reassembled_brass_calls_20210705_noheader.txt
zless /nfs/cancer_ref01/nst_links/live/2127/PD506*a/*brass.annot.bedpe.gz | grep "# chr1" | sort | uniq > bulk_yst_reassembled_brass_calls_20210705_header.txt

cat bulk_yst_reassembled_brass_calls_20210705_header.txt bulk_yst_reassembled_brass_calls_20210705_noheader.txt > bulk_yst_reassembled_brass_calls_20210705.txt

rm bulk_yst_reassembled_brass_calls_20210705_noheader.txt bulk_yst_reassembled_brass_calls_20210705_header.txt

#removed hash from header 05/07/21

```

Review the variants called across all samples
```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/structural_variants')

reassembled <- read.table('bulk_yst_reassembled_brass_calls_20210705.txt', header = T, sep = '\t', stringsAsFactors = F) #458 as of 5/7/21

#fragile_sites <- read.table('/lustre/scratch119/casm/team274sb/to3/placenta/structural_variants/Fragile_sites.txt', header = T, sep = '\t', stringsAsFactors = F)

#fragile site annotation
#reassembled$fragile.site.chr1 <- NA
#reassembled$fragile.site.chr2 <- NA

#for(k in 1:nrow(fragile_sites)){
#  if(length(reassembled[reassembled$chr1 == fragile_sites$Chr[k] & reassembled$start1 < fragile_sites$HumGen_37_end[k] & reassembled$end1 > fragile_sites$HumGen_37_start[k],]$fragile.site.chr1)){
#   reassembled[reassembled$chr1 == fragile_sites$Chr[k] & reassembled$start1 < fragile_sites$HumGen_37_end[k] & reassembled$end1 > fragile_sites$HumGen_37_start[k],]$fragile.site.chr1 = fragile_sites$Fragile_site[k]
#  }
#  if(length(reassembled[reassembled$chr2 == fragile_sites$Chr[k] & reassembled$start2 < fragile_sites$HumGen_37_end[k] & reassembled$end2 > fragile_sites$HumGen_37_start[k],]$fragile.site.chr2)){
#    reassembled[reassembled$chr2 == fragile_sites$Chr[k] & reassembled$start2 < fragile_sites$HumGen_37_end[k] & reassembled$end2 > fragile_sites$HumGen_37_start[k],]$fragile.site.chr2 = fragile_sites$Fragile_site[k]
#  }
#}

#annotate genes contained within deletions, tandem-duplications and inversions
##generate bed file

annot_sv <- reassembled[reassembled$svclass %in% c('deletion', 'tandem-duplication', 'inversion'),] #405 SVs
write.table(annot_sv[, c('chr1', 'start1', 'end2')], 'bulk_yst_intrachrom_coord_20210705.bed', col.names = F, row.names = F, quote = F, sep = '\t')

```

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/structural_variants

module load bedtools
~/scripts/intersect_intv.sh bulk_yst_intrachrom_coord_20210705.bed bulk_yst_sv_genes_20210705.txt

```

```{r}

sv_genes <- read.table('bulk_yst_sv_genes_20210705.txt', header = T, stringsAsFactors = F, sep = '\t')
sv_genes <- sv_genes[!(duplicated(sv_genes)),]
sv_genes$Chr <- t(data.frame(strsplit(sv_genes$Chromosome, ':')))[,1]
sv_genes$Gene.Start <- as.numeric(t(data.frame(strsplit(t(data.frame(strsplit(sv_genes$Chromosome, ':')))[,2], '-')))[,1])
sv_genes$Gene.End <- as.numeric(t(data.frame(strsplit(t(data.frame(strsplit(sv_genes$Chromosome, ':')))[,2], '-')))[,2])

reassembled$Genes.in.dup.del <- NA
annot_sv$Genes.in.dup.del <- NA

for(i in 1:nrow(sv_genes)){
  if(length(annot_sv[annot_sv$chr1 == sv_genes$Chr[i] & annot_sv$start1 < sv_genes$Gene.End[i] & annot_sv$end2 > sv_genes$Gene.Start[i],]$Genes.in.dup.del)){
    annot_sv[annot_sv$chr1 == sv_genes$Chr[i] & annot_sv$start1 < sv_genes$Gene.End[i] & annot_sv$end2 > sv_genes$Gene.Start[i],]$Genes.in.dup.del = paste(annot_sv[annot_sv$chr1 == sv_genes$Chr[i] & annot_sv$start1 < sv_genes$Gene.End[i] & annot_sv$end2 > sv_genes$Gene.Start[i],]$Genes.in.dup.del, sv_genes$GeneSymbol[i], sep = ',')
  }
}

##remove aberrant NAs after paste function
annot_sv$Genes.in.dup.del = gsub('NA,', '', annot_sv$Genes.in.dup.del)

#add back to original data frame
for(i in 1:nrow(reassembled)){
  if(reassembled$chr1[i] %in% annot_sv$chr1 & reassembled$start1[i] %in% annot_sv$start1 & reassembled$end1[i] %in% annot_sv$end1 & reassembled$chr2[i] %in% annot_sv$chr2 & reassembled$start2[i] %in% annot_sv$start2 & reassembled$end2[i] %in% annot_sv$end2 & reassembled$sample[i] %in% annot_sv$sample){
    reassembled$Genes.in.dup.del[i] = annot_sv[annot_sv$chr1 == reassembled$chr1[i] & annot_sv$start1 == reassembled$start1[i] & annot_sv$end1 == reassembled$end1[i] & annot_sv$chr2 == reassembled$chr2[i] & annot_sv$start2 == reassembled$start2[i] & annot_sv$end2 == reassembled$end2[i] & annot_sv$sample == reassembled$sample[i],]$Genes.in.dup.del[1]
  }
}

#check if breakpoints generally or regions within intrachromosomal SVs are cancer genes
##Intersect with COSMIC v94 consensus cancer gene list
cosmic_v94 <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/cosmic_v94_cancer_gene_census.csv', sep = ',', header = T, stringsAsFactors = F)
cancer_genes <- unlist(strsplit(paste(cosmic_v94$Synonyms, collapse = ","), ","))
#cancer_genes <- cosmic_v94$Gene.Symbol

##add flagged cancer genes to list 
reassembled$Cancer.gene <- NA

for(i in 1:nrow(reassembled)){
  gene_list = unlist(strsplit(reassembled$Genes.in.dup.del[i], ','))
  gene_list = c(gene_list, reassembled$gene1[i], reassembled$gene2[i])
  gene_list = unique(gene_list)
  gene_list = cancer_genes[which(as.vector(cancer_genes) %in% as.vector(gene_list))]
  if(length(gene_list)){
    reassembled$Cancer.gene[i] = paste(gene_list, collapse = ',')
  }
  else {
    reassembled$Cancer.gene[i] = NA
  }
  }

reassembled$Cancer.gene

#annotate whether the cancer genes are dominant or recessive
reassembled$Dom.Rec.COSMICv94 <- NA
for(i in 1:nrow(reassembled)){
  if(!is.na(reassembled$Cancer.gene[i])){
    gene.list <- unlist(strsplit(reassembled$Cancer.gene[i], ','))
    reassembled$Dom.Rec.COSMICv94[i] = paste(cosmic_v94[cosmic_v94$Gene.Symbol %in% gene.list, ]$Molecular.Genetics, collapse = ',')
  }
}

reassembled$Dom.Rec.COSMICv94

#fetch the 10bp upstream of the lower co-ordinate and downstream of higher co-ordinates, unfortunately brass uses 0-based co-ordinates at the start and 1-based at the end...
head(reassembled)
bed_coord <- cbind(reassembled$chr1, (reassembled$start1 - 10), reassembled$start1 + 10)
bed_coord <- rbind(bed_coord, cbind(reassembled$chr1, (reassembled$end1 - 10), reassembled$end1 + 10))
bed_coord <- rbind(bed_coord, cbind(reassembled$chr2, (reassembled$start2 - 10), (reassembled$start2 + 10)))
bed_coord <- rbind(bed_coord, cbind(reassembled$chr2, (reassembled$end2 - 10), (reassembled$end2 + 10)))

#write.table(bed_coord, '20bp_NT_context_bulk_yst_SVs_20210705.bed', quote = F, sep = '\t', row.names = F, col.names = F)
write.table(bed_coord, '20bp_NT_context_bulk_yst_SVs_20220118.bed', quote = F, sep = '\t', row.names = F, col.names = F)

```


```{bash}

#use bedtools to get context from reference genome

module load bedtools
#bedtools getfasta -fi /lustre/scratch119/casm/team274sb/to3/References/GRCh37d5/genome.fa -fo 20bp_bulk_yst_context_20210509.fa -bed 20bp_NT_context_bulk_yst_SVs_20210705.bed 
bedtools getfasta -fi /lustre/scratch119/casm/team274sb/to3/References/GRCh37d5/genome.fa -fo 20bp_bulk_yst_context_20220118.fa -bed 20bp_NT_context_bulk_yst_SVs_20220118.bed 

```

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/structural_variants/")

#read in fasta file
library(seqinr)
#fasta <- read.fasta('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/structural_variants/20bp_bulk_yst_context_20210509.fa')
fasta <- read.fasta('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/structural_variants/20bp_bulk_yst_context_20220118.fa')

#convert to dataframe
context <- data.frame(Fragments=names(fasta), Seqs=toupper(unlist(getSequence(fasta, as.string=T))), stringsAsFactors = F)
context$Chr <- t(data.frame(strsplit(context$Fragments, ':')))[,1]
context$Start <- as.numeric(t(data.frame(strsplit(t(data.frame(strsplit(context$Fragments, ':')))[,2], '-')))[,1])
context$End <- as.numeric(t(data.frame(strsplit(t(data.frame(strsplit(context$Fragments, ':')))[,2], '-')))[,2])

#add in the 20bp context to the BRASS co-ordinates
reassembled$start1_20bp_nt_context <- NA
reassembled$end1_20bp_nt_context <- NA
reassembled$start2_20bp_nt_context <- NA
reassembled$end2_20bp_nt_context <- NA

for(i in 1:nrow(reassembled)){
  reassembled$start1_20bp_nt_context[i] = context[context$Chr == reassembled$chr1[i] & context$End == (reassembled$start1[i] + 10),]$Seqs[1]
  reassembled$end1_20bp_nt_context[i] = context[context$Chr == reassembled$chr1[i] & context$End == (reassembled$end1[i] + 10),]$Seqs[1]
  reassembled$start2_20bp_nt_context[i] = context[context$Chr == reassembled$chr2[i] & context$Start == (reassembled$start2[i] - 10),]$Seqs[1]
  reassembled$end2_20bp_nt_context[i] = context[context$Chr == reassembled$chr2[i] & context$Start == (reassembled$end2[i] - 10),]$Seqs[1]
}

#reassembled$sample_nomatched <- as.vector(t(data.frame(strsplit(reassembled$sample, ',')))[,1])
#reassembled$patient <- substr(reassembled$sample_nomatched, 0, 7)

#write.table(reassembled, 'bulk_yst_reassembled_variants_annot_20200705.txt', sep = '\t', quote = F, row.names = F)
write.table(reassembled, 'bulk_yst_reassembled_variants_annot_20220118.txt', sep = '\t', quote = F, row.names = F)

```

```{r}

cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/structural_variants/bulk_yst_reassembled_variants_annot_20220118.txt /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/bulk/02_score

```


####Matched normal filter

Some variants are still called in the matched normal and tumour samples, these are removed.

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/bulk/02_score")

#bulk_brass_calls = read.table('bulk_yst_reassembled_variants_annot_20200705.txt', header = T, sep = '\t', stringsAsFactors = F)
bulk_brass_calls = read.table('bulk_yst_reassembled_variants_annot_20220118.txt', header = T, sep = '\t', stringsAsFactors = F)
no_matched_calls = bulk_brass_calls[!grepl(",", bulk_brass_calls$sample),] #samples are comma separated where this is relevant

#write.table(no_matched_calls, "/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/bulk/03_matched/bulk_yst_reassembled_nomatched_variants_annot_20211231.txt", sep = '\t', col.names = T, row.names = F, quote = F)
write.table(no_matched_calls, "/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/bulk/03_matched/bulk_yst_reassembled_nomatched_variants_annot_20220118.txt", sep = '\t', col.names = T, row.names = F, quote = F)

```

####Retrotransposition filter

Remove variants that who have a breakpoint within 100bp of a retrotranposition event.

```{r}

rt_coord = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/retrotransposition_coordinates_20211231.txt", header = T, sep = '\t', stringsAsFactors = F) #coordinates of all rt breakpoints
#rt_coord = read.table('bulk_yst_rt_coordinates.txt', header = T, sep = '\t', stringsAsFactors = F) 

standard_sv = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/bulk/03_matched/bulk_yst_reassembled_nomatched_variants_annot_20220118.txt', header = T, sep = '\t', stringsAsFactors = F) #some of these in BRASS may have been mistakenly called as standard structural variants

#remove all SVs that occur with 100bp of a called RT
for(j in 1:nrow(rt_coord)){
  if(nrow(standard_sv[standard_sv$sample == rt_coord$sample[j] & standard_sv$chr1 == rt_coord$chr[j] & standard_sv$start1 >= (rt_coord$breakpoint[j] - 100) & standard_sv$start1 <= (rt_coord$breakpoint[j] + 100), ])){
    standard_sv[standard_sv$sample == rt_coord$sample[j] & standard_sv$chr1 == rt_coord$chr[j] & standard_sv$start1 >= (rt_coord$breakpoint[j] - 100) & standard_sv$start1 <= (rt_coord$breakpoint[j] + 100), ]$svclass = 'retrotransposition'
  }
  if(nrow(standard_sv[standard_sv$sample == rt_coord$sample[j] & standard_sv$chr1 == rt_coord$chr[j] & standard_sv$end1 >= (rt_coord$breakpoint[j] - 100) & standard_sv$end1 <= (rt_coord$breakpoint[j] + 100), ])){
    standard_sv[standard_sv$sample == rt_coord$sample[j] & standard_sv$chr1 == rt_coord$chr[j] & standard_sv$end1 >= (rt_coord$breakpoint[j] - 100) & standard_sv$end1 <= (rt_coord$breakpoint[j] + 100), ]$svclass = 'retrotransposition'
  }
  if(nrow(standard_sv[standard_sv$sample == rt_coord$sample[j] & standard_sv$chr2 == rt_coord$chr[j] & standard_sv$start2 >= (rt_coord$breakpoint[j] - 100) & standard_sv$start2 <= (rt_coord$breakpoint[j] + 100), ])){
    standard_sv[standard_sv$sample == rt_coord$sample[j] & standard_sv$chr2 == rt_coord$chr[j] & standard_sv$start2 >= (rt_coord$breakpoint[j] - 100) & standard_sv$start2 <= (rt_coord$breakpoint[j] + 100), ]$svclass = 'retrotransposition'
  }
  if(nrow(standard_sv[standard_sv$sample == rt_coord$sample[j] & standard_sv$chr2 == rt_coord$chr[j] & standard_sv$end2 >= (rt_coord$breakpoint[j] - 100) & standard_sv$end2 <= (rt_coord$breakpoint[j] + 100), ])){
    standard_sv[standard_sv$sample == rt_coord$sample[j] & standard_sv$chr2 == rt_coord$chr[j] & standard_sv$end2 >= (rt_coord$breakpoint[j] - 100) & standard_sv$end2 <= (rt_coord$breakpoint[j] + 100), ]$svclass = 'retrotransposition'
  }
}

no_rt = standard_sv[standard_sv$svclass != 'retrotransposition',]
write.table(no_rt, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/bulk/04_rt/final_bulk_gct_brass_calls_20220118.txt', sep ='\t', quote = F, col.names = T, row.names = F)

```

##VARIANT CALLING AND FILTERING (LCM and bulk)

###-Copy number aberrations

Battenberg depends on BAF to determine copy number states. This can be problematic as most of our samples are highly pure and have frequent LOH events which makes determining the exact copy number configuration challenging. Our solution to this is to splice the matched normal reads into the tumour and re-run copy number calling. If this improves the fit, as determined by manual review and use of DPClust, then the "merged" version, rather than the "original" is used.

In each "merged" sample we use the ploidy and copy number calls from the "merged" run but the purity from the "original" run.

####Original battenberg run

Configure variables

```{bash}

#LCM setup
# set this to your lustre area
export WORK=/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/original_file_bb
# this defines the root of the reference area for the supported human build
export REF_BASE=/lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5
# paths to the BAM files on lustre for LCM samples
export MT_BAM=/lustre/scratch117/casm/team294/rs30/GCT_analyses/Staged_BAM/
export WT_BAM=/lustre/scratch117/casm/team294/rs30/GCT_analyses/Staged_BAM/

#bulk YST sample setup
#export MT_BAM=/lustre/scratch119/casm/team294rr/to3/testes/tumour/staged_bulk_bam_files/
#export WT_BAM=/lustre/scratch119/casm/team294rr/to3/testes/tumour/staged_bulk_bam_files/

# how many CPUs you want to use
export CPU_TO_USE=4
# gender - can be XX or XY or L (determined from the normal bam file)
export GENDER="L"
export SPECIES="Human"
export ASSEMBLY="GRCh37d5"

```

Battenberg may need to be run across batches of samples due to the large volume of intermediate files that can fill up the lustre space.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/original_file_bb

module load cgpbattenberg

infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt"
grep -v 'Sample' $infile | while IFS=$'\t' read -r -a tmp ; do
echo "${tmp[0]}" #sample id
echo "${tmp[2]}" #project id
echo "${tmp[3]}" #matched normal id
mkdir -p "${tmp[0]}"
bsub -o $WORK/${tmp[0]}/${tmp[0]}.out -e $WORK/${tmp[0]}/${tmp[0]}.err -n$CPU_TO_USE -R "span[hosts=1]" -M40000 -R 'select[mem>40000] rusage[mem=40000]' \
battenberg.pl \
-o $WORK/${tmp[0]} \
-r $REF_BASE/genome.fa.fai \
-tb $MT_BAM/${tmp[2]}/${tmp[0]}/mapped_sample/${tmp[0]}.sample.dupmarked.bam \
-nb $WT_BAM/${tmp[2]}/${tmp[3]}/mapped_sample/${tmp[3]}.sample.dupmarked.bam \
-e $REF_BASE/battenberg/impute/impute_info.txt \
-u $REF_BASE/battenberg/1000genomesloci \
-ig $REF_BASE/battenberg/ignore_contigs.txt \
-c $REF_BASE/battenberg/probloci.txt \
-gc $REF_BASE/battenberg/battenberg_wgs_gc_correction_1000g \
-ge $GENDER \
-gl $REF_BASE/gender.tsv \
-rs "$SPECIES" \
-ra "$ASSEMBLY" \
-t $CPU_TO_USE
done

```

####Merged battenberg run

#####Tumour BAM file header retrieval

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/final_variant_calls/BB_rerun/merged_bams

infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt"
grep -v 'Sample' $infile | while IFS=$'\t' read -r -a tmp ; do
echo "${tmp[0]}" #sample id
echo "${tmp[2]}" #project id
echo "${tmp[3]}" #matched normal id
echo "${tmp[0]}_${tmp[3]}.bam" #new name
samtools view -H /lustre/scratch117/casm/team294/rs30/GCT_analyses/Staged_BAM/${tmp[2]}/${tmp[0]}/mapped_sample/${tmp[0]}.sample.dupmarked.bam > ${tmp[0]}.tumour.header.sam
done

```

#####Create new, merged BAM files

Basic samtools merge, sort, reheader and index to create new BAM file

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/final_variant_calls/BB_rerun/merged_bams

#run_v2.sh = for LCM data
samtools merge -@ 10 - /lustre/scratch117/casm/team294/rs30/GCT_analyses/Staged_BAM/${2}/${1}/mapped_sample/${1}.sample.dupmarked.bam /lustre/scratch117/casm/team294/rs30/GCT_analyses/Staged_BAM/${2}/${3}/mapped_sample/${3}.sample.dupmarked.bam | samtools sort -m 2G -@ 10  - | samtools reheader ${1}.tumour.header.sam - > ${1}_${3}.sorted.reheader.bam
samtools index -@ 5 ${1}_${3}.sorted.reheader.bam

#run_v3.sh = for bulk YST data
samtools merge -@ 10 - /lustre/scratch119/casm/team294rr/to3/testes/tumour/staged_bulk_bam_files/${2}/${1}/mapped_sample/${1}.sample.dupmarked.bam /lustre/scratch119/casm/team294rr/to3/testes/tumour/staged_bulk_bam_files/${2}/${3}/mapped_sample/${3}.sample.dupmarked.bam | samtools sort -m 2G -@ 10  - | samtools reheader ${1}.tumour.header.sam - > ${1}_${3}.sorted.reheader.bam
samtools index -@ 5 ${1}_${3}.sorted.reheader.bam

#LCM
infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt"
grep 'PD4' $infile | while IFS=$'\t' read -r -a tmp ; do
echo "${tmp[0]}" #sample id
echo "${tmp[2]}" #project id
echo "${tmp[3]}" #matched normal id
echo "${tmp[0]}_${tmp[3]}.bam" #new name
bsub -o ${tmp[0]}_${tmp[3]}.sorted.reheader.bam.out -e ${tmp[0]}_${tmp[3]}.sorted.reheader.bam.err -n30 -R "span[hosts=1]" -M30000 -R 'select[mem>30000] rusage[mem=30000]' bash run_v2.sh ${tmp[0]} ${tmp[2]} ${tmp[3]}
done

#Bulk YST
infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt"
grep 'PD5' $infile | while IFS=$'\t' read -r -a tmp ; do
echo "${tmp[0]}" #sample id
echo "${tmp[2]}" #project id
echo "${tmp[3]}" #matched normal id
echo "${tmp[0]}_${tmp[3]}.bam" #new name
bsub -o ${tmp[0]}_${tmp[3]}.sorted.reheader.bam.out -e ${tmp[0]}_${tmp[3]}.sorted.reheader.bam.err -n30 -R "span[hosts=1]" -M30000 -R 'select[mem>30000] rusage[mem=30000]' bash run_v3.sh ${tmp[0]} ${tmp[2]} ${tmp[3]}
done

```

#####Run Battenberg

Configure variables

```{bash}

# set this to your lustre area
export WORK=/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/merged_files_bb
# this defines the root of the reference area for the supported human build
export REF_BASE=/lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5
# paths to the BAM files on lustre
export MT_BAM=/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/final_variant_calls/BB_rerun/merged_bams
export WT_BAM=/lustre/scratch117/casm/team294/rs30/GCT_analyses/Staged_BAM
# how many CPUs you want to use
export CPU_TO_USE=10
# gender - can be XX or XY or L (determined from the normal bam file)
export GENDER="L"
export SPECIES="Human"
export ASSEMBLY="GRCh37d5"

```

May need to be run in batches, as before.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/merged_files_bb

module load cgpbattenberg

infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/supplementary_data/GCT_DNA_supplementary_data.txt"
grep -v 'Sample' $infile | while IFS=$'\t' read -r -a tmp ; 
do
echo "${tmp[0]}" #sample id
echo "${tmp[2]}" #project id
echo "${tmp[3]}" #matched normal id
mkdir -p "${tmp[0]}"
bsub -o $WORK/${tmp[0]}/${tmp[0]}.out -e $WORK/${tmp[0]}/${tmp[0]}.err -n$CPU_TO_USE -R "span[hosts=1]" -M40000 -R 'select[mem>40000] rusage[mem=40000]' \
battenberg.pl \
-o $WORK/${tmp[0]} \
-r $REF_BASE/genome.fa.fai \
-tb $MT_BAM/${tmp[0]}_${tmp[3]}.sorted.reheader.bam \
-nb $WT_BAM/${tmp[2]}/${tmp[3]}/mapped_sample/${tmp[3]}.sample.dupmarked.bam \
-e $REF_BASE/battenberg/impute/impute_info.txt \
-u $REF_BASE/battenberg/1000genomesloci \
-ig $REF_BASE/battenberg/ignore_contigs.txt \
-c $REF_BASE/battenberg/probloci.txt \
-gc $REF_BASE/battenberg/battenberg_wgs_gc_correction_1000g \
-ge $GENDER \
-gl $REF_BASE/gender.tsv \
-rs "$SPECIES" \
-ra "$ASSEMBLY" \
-t $CPU_TO_USE
done

```

####Move relevant files to new directory

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/original_file_bb/')

samples = list.dirs(path = ".", full.names = F, recursive = F)

for(sample in samples){
  purity_file = read.table(paste0(sample, '/', sample, '_cellularity_ploidy.txt'), header = T, sep = '\t', stringsAsFactors = F)
  if(purity_file$cellularity < 0.8) print(paste0(sample, '_', purity_file$cellularity))
  
}

```

Each output has been reviewed and the decision to use the original or merged Battenberg file taken.

First, copy over according to whether we take the original file or merged file. For those where we use the merged file, take the purity and rho values from the original run.

```{r}

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', header = T, sep = '\t', stringsAsFactors = F)

for(i in 1:nrow(manifest)){
  if(manifest$Battenberg.used[i] == 'original'){
      system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/original_file_bb/', manifest$Sample[i], '/', manifest$Sample[i], '_rho_and_psi.txt /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input'))
      system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/original_file_bb/', manifest$Sample[i], '/', manifest$Sample[i], '_cellularity_ploidy.txt /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input'))
      system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/original_file_bb/', manifest$Sample[i], '/', manifest$Sample[i], '_subclones.txt.gz /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input'))
  }
  if(manifest$Battenberg.used[i] == 'merged'){
      system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/merged_file_bb/', manifest$Sample[i], '/', manifest$Sample[i], '_rho_and_psi.txt /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input'))
      system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/merged_file_bb/', manifest$Sample[i], '/', manifest$Sample[i], '_cellularity_ploidy.txt /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input'))
      system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/merged_file_bb/', manifest$Sample[i], '/', manifest$Sample[i], '_subclones.txt.gz /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input'))
  }
}

for(i in 1:nrow(manifest)){
  if(manifest$Battenberg.used[i] == 'merged'){
    #read in original purity files
    original_rho_file = read.table(paste0('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/original_file_bb/', manifest$Sample[i], '/', manifest$Sample[i], '_rho_and_psi.txt'), header = T, sep = '\t', stringsAsFactors = F)
    original_purity_file = read.table(paste0('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/original_file_bb/', manifest$Sample[i], '/', manifest$Sample[i], '_cellularity_ploidy.txt'), header = T, sep = '\t', stringsAsFactors = F) 
    
    #read in new purity files
    new_rho_file = read.table(paste0('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/merged_file_bb/', manifest$Sample[i], '/', manifest$Sample[i], '_rho_and_psi.txt'), header = T, sep = '\t', stringsAsFactors = F)
    new_purity_file = read.table(paste0('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/merged_file_bb/', manifest$Sample[i], '/', manifest$Sample[i], '_cellularity_ploidy.txt'), header = T, sep = '\t', stringsAsFactors = F) 
  
    #correct the new purity files
    new_rho_file$rho = original_rho_file$rho
    new_purity_file$cellularity = original_purity_file$cellularity
  
    #update files in situ
    write.table(new_rho_file, paste0('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/', manifest$Sample[i], '_rho_and_psi.txt'), col.names = T, row.names = T, sep = '\t', quote = F)
    write.table(new_purity_file, paste0('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/', manifest$Sample[i], '_cellularity_ploidy.txt'), col.names = T, row.names = F, sep = '\t', quote = F)
  
  }
}

```

Several samples required further refitting or minor modification to final subclone finals:
PD42036a_lo0008 - change original Battenberg file to get chr8 configuration from merged
PD45543a_lo0002 - change merged Battenberg file to get chr12 from original
PD45544a_lo0014 - change merged Battenberg file to get chr12 configuration from original
PD42036b_lo0047 - refit 3+1 segment to 2+1 on original bam
PD45543a_lo0005 - 12p cannot break as BAF is very similar despite gain with no related samples possessing the exact same configuration, low purity so has little resultant impact
PD50656a - modify 5+0 chr9 segment to 4+0

```{r}

#gunzip all subclones files
#PD42036a_lo0008 - change original Battenberg file to get chr8 configuration from merged
#PD45543a_lo0002 - change merged Battenberg file to get chr12 from original
#PD45544a_lo0014 - change merged Battenberg file to get chr12 configuration from original

##PD42036a_lo0008
PD42036a_lo0008_new = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/PD42036a_lo0008_subclones.txt', header = T, sep = '\t', stringsAsFactors = F)
PD42036a_lo0008_old = read.table(gzfile('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/merged_file_bb/PD42036a_lo0008/PD42036a_lo0008_subclones.txt.gz'), header = T, sep = '\t', stringsAsFactors = F)

write.table(rbind(PD42036a_lo0008_new[PD42036a_lo0008_new$chr %in% c(1:7),], PD42036a_lo0008_old[PD42036a_lo0008_old$chr == 8,], PD42036a_lo0008_new[PD42036a_lo0008_new$chr %in% c(9:22, 'X', 'Y'),]), '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/PD42036a_lo0008_subclones.txt', col.names = T, row.names = F, sep = '\t', quote = F)

##PD45543a_lo0002
PD45543a_lo0002_new = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/PD45543a_lo0002_subclones.txt', header = T, sep = '\t', stringsAsFactors = F)
PD45543a_lo0002_old = read.table(gzfile('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/original_file_bb/PD45543a_lo0002/PD45543a_lo0002_subclones.txt.gz'), header = T, sep = '\t', stringsAsFactors = F)

write.table(rbind(PD45543a_lo0002_new[PD45543a_lo0002_new$chr %in% c(1:11),], PD45543a_lo0002_old[PD45543a_lo0002_old$chr == 12,], PD45543a_lo0002_new[PD45543a_lo0002_new$chr %in% c(13:22, 'X', 'Y'),]), '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/PD45543a_lo0002_subclones.txt', col.names = T, row.names = F, sep = '\t', quote = F)

##PD45544a_lo0014
PD45544a_lo0014_new = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/PD45544a_lo0014_subclones.txt', header = T, sep = '\t', stringsAsFactors = F)
PD45544a_lo0014_old = read.table(gzfile('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/original_file_bb/PD45544a_lo0014/PD45544a_lo0014_subclones.txt.gz'), header = T, sep = '\t', stringsAsFactors = F)

write.table(rbind(PD45544a_lo0014_new[PD45544a_lo0014_new$chr %in% c(1:11),], PD45544a_lo0014_old[PD45544a_lo0014_old$chr == 12,], PD45544a_lo0014_new[PD45544a_lo0014_new$chr %in% c(13:22, 'X', 'Y'),]), '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/PD45544a_lo0014_subclones.txt', col.names = T, row.names = F, sep = '\t', quote = F)

#PD42036b_lo0047 - refit 3+1 segment to 2+1 on original bam
system('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/original_file_bb_refit/PD42036b_lo0047/PD42036b_lo0047_rho_and_psi.txt /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input')
system('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/original_file_bb_refit/PD42036b_lo0047/PD42036b_lo0047_cellularity_ploidy.txt /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input')
system('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/original_file_bb_refit/PD42036b_lo0047/PD42036b_lo0047_subclones.txt.gz /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input')

```

####Concatenate high confidence copy number calls

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/00_final_calls')

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', header = T, sep = '\t', stringsAsFactors = F)

all_cns = c()
for(sample in manifest[manifest$Average.reads.per.chromosome.copy. >= 5,]$Sample){
  data = read.table(paste0(sample, '_subclones.txt'), header = T, sep = '\t', stringsAsFactors = F)
  data$sample = sample
  all_cns = rbind(all_cns, data)
}

all_cns$Case.ID = substr(all_cns$sample, 0, 7)

write.table(all_cns, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_cn_calls.txt', col.names = T, row.names = F, sep = '\t', quote = F)

```

##SINGLE SAMPLE DPCLUST QC

Single sample run, autosomal genome only to assess fit of subs and copy number calls. CN calling on X and CCF calculations not reliable, hence why excluded when assessing quality of calls.

###Collect input files
Copy over SNV VCF and copy number files
```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input

cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/00_final_calls/* /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input

cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/subs/*vcf /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input

```

Create a sample list
```{r}

#Functions
sample_list_generator_lcm <- function(filepath){
  sample.list <- unlist(strsplit(list.files(paste0(filepath), pattern = '_post_swater_final_snvs.annot.vcf'), '_post_swater_final_snvs.annot.vcf'))
  write.table(sample.list, 'sample_names.txt', sep = '\t', quote = F, row.names = F, col.names = F)
}

sample_list_generator_bulk <- function(filepath){
  sample.list <- unlist(strsplit(list.files(paste0(filepath), pattern = '.caveman_c.annot_filtered.vcf'), '.caveman_c.annot_filtered.vcf'))
  write.table(sample.list, 'sample_names_bulk_yst.txt', sep = '\t', quote = F, row.names = F, col.names = F)
}

#LCM
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input')
sample_list_generator_lcm(filepath = '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input')

#Bulk
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input')
sample_list_generator_bulk(filepath = '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input')

```

###Pre-processing

```{bash}

#LCM
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"

ROOT_DIR="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/"
SOURCE_DIR="/lustre/scratch119/casm/team294rr/to3/testes/tumour/dpclust/00_Scripts/CGP-DPC/"
REF_DIR=$SOURCE_DIR"data/"
DATA_DIR=$ROOT_DIR"01_Input/"
WORK_DIR=$ROOT_DIR"02_Intermediates/"
sample_ID_file=${DATA_DIR}"sample_names.txt"
OUTPUT_DIR=${WORK_DIR}

##female cases
for f in $(ls *_subclones.txt | grep PD455);
do
filename=$(basename -- "$f")
sample="${filename%_subclones.txt}"
Rscript --no-save --no-restore --verbose \
/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw/dpclust3p/example/preproc_pipeline_caveman.R -s ${sample} -v $DATA_DIR/${sample}_post_swater_final_snvs.annot.vcf -r $DATA_DIR/${sample}_rho_and_psi.txt -c $DATA_DIR/${sample}_subclones.txt --sex female -o $OUTPUT_DIR --fai $REF_DIR/"genome.fa.fai" --ign $REF_DIR/"ignore_chrs.txt" > $OUTPUT_DIR/preproc.female.outputFile.Rout 2>&1
done

##male cases
for f in $(ls *_subclones.txt | grep -v "PD455\|PD506"); 
do
filename=$(basename -- "$f")
sample="${filename%_subclones.txt}"
bsub -G team294-vwork -J${sample} -M10000 -R"select[mem>10000] rusage[mem=10000] span[hosts=1]" -n 1 -q normal -e %J.stderr -o %J.stdout "Rscript --no-save --no-restore --verbose \
/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw/dpclust3p/example/preproc_pipeline_caveman.R -s ${sample} -v $DATA_DIR/${sample}_post_swater_final_snvs.annot.vcf -r $DATA_DIR/${sample}_rho_and_psi.txt -c $DATA_DIR/${sample}_subclones.txt --sex male -o $OUTPUT_DIR --fai $REF_DIR/'genome.fa.fai' --ign $REF_DIR/'ignore_chrs.txt' > $OUTPUT_DIR/preproc.male.outputFile.Rout 2>&1"
done

#Bulk
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"

ROOT_DIR="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/"
SOURCE_DIR="/lustre/scratch119/casm/team294rr/to3/testes/tumour/dpclust/00_Scripts/CGP-DPC/"
REF_DIR=$SOURCE_DIR"data/"
DATA_DIR=$ROOT_DIR"01_Input/"
WORK_DIR=$ROOT_DIR"02_Intermediates/"
sample_ID_file=${DATA_DIR}"sample_names_bulk_yst.txt"
OUTPUT_DIR=${WORK_DIR}

##female cases
for f in PD50654a PD50657a PD50659a PD50660a PD50661a;
do
Rscript --no-save --no-restore --verbose \
/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw/dpclust3p/example/preproc_pipeline_caveman.R -s ${f} -v $DATA_DIR/${f}.caveman_c.annot_filtered.vcf -r $DATA_DIR/${f}_rho_and_psi.txt -c $DATA_DIR/${f}_subclones.txt --sex female -o $OUTPUT_DIR --fai $REF_DIR/"genome.fa.fai" --ign $REF_DIR/"ignore_chrs.txt" > $OUTPUT_DIR/preproc.female.outputFile.Rout 2>&1
done

##male cases
for f in PD50655a PD50656a;
do
Rscript --no-save --no-restore --verbose \
/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw/dpclust3p/example/preproc_pipeline_caveman.R -s ${f} -v $DATA_DIR/${f}.caveman_c.annot_filtered.vcf -r $DATA_DIR/${f}_rho_and_psi.txt -c $DATA_DIR/${f}_subclones.txt --sex male -o $OUTPUT_DIR --fai $REF_DIR/"genome.fa.fai" --ign $REF_DIR/"ignore_chrs.txt" > $OUTPUT_DIR/preproc.male.outputFile.Rout 2>&1
done

```

Remove X/Y subs from *_allDirichletProcessInfo.txt

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/02_Intermediates

for f in $(ls *_allDirichletProcessInfo.txt);
do
filename=$(basename -- "$f")
sample="${filename%_allDirichletProcessInfo.txt}"
echo $sample
grep -v 'X\|Y' $f > ${sample}_allDirichletProcessInfo_noX.txt
mv ${sample}_allDirichletProcessInfo_noX.txt $f
done

```


Create a master project file to put into the pipeline for each sample

```{bash}

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"

#LCM
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/02_Intermediates

##female cases
for f in $(ls *_allDirichletProcessInfo.txt | grep PD455);
do
filename=$(basename -- "$f")
sample="${filename%_allDirichletProcessInfo.txt}"
Rscript --no-save --no-restore --verbose ~/scripts/dpclust_master_project_file_creator_updated_20210513.R /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input ${sample} female
done

##male cases
for f in $(ls *_allDirichletProcessInfo.txt | grep -v "PD455\|PD506");
do
filename=$(basename -- "$f")
sample="${filename%_allDirichletProcessInfo.txt}"
Rscript --no-save --no-restore --verbose ~/scripts/dpclust_master_project_file_creator_updated_20210513.R /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input ${sample} male
done

#Bulk
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/02_Intermediates

##female cases
for f in PD50654a PD50657a PD50659a PD50660a PD50661a;
do
Rscript --no-save --no-restore --verbose ~/scripts/dpclust_master_project_file_creator_updated_20210513.R /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input ${f} female
done

##male cases
for f in PD50655a PD50656a;
do
Rscript --no-save --no-restore --verbose ~/scripts/dpclust_master_project_file_creator_updated_20210513.R /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input ${f} male
done

```

###Run DPClust

```{bash}

#LCM
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/03_Output

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"

ROOT_DIR="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/"
WORK_DIR=$ROOT_DIR"02_Intermediates/"
OUTPUT_DIR=${ROOT_DIR}"03_Output/"
MEM=20000
QUEUE=normal
CORES=10
n_its=10000
burnin=3000
sample_list="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input/sample_names.txt"

while read line;
do set $line
bsub -G team294-vwork -J"${1}" -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -q ${QUEUE} -e %J.dpclust.stderr -o %J.dpclust.stdout \
"xvfb-run -a Rscript --no-save --no-restore --verbose /nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw/DPClust/example/dpclust_pipeline.R -r 1 -d $WORK_DIR -o $OUTPUT_DIR -i ${WORK_DIR}${1}"_ProjectFile.txt" -k -a nd_dp --iterations ${n_its} --burnin ${burnin} --mut_assignment_type 1 --min_muts_cluster -1 --min_frac_muts_cluster 0.01 --num_muts_sample 4000 --bin_size NA --seed 42 > ${1}.dpclust.outputFile.Rout 2>&1"
done < $sample_list

#Bulk
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/03_Output

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"

ROOT_DIR="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/"
WORK_DIR=$ROOT_DIR"02_Intermediates/"
OUTPUT_DIR=${ROOT_DIR}"03_Output/"
MEM=20000
QUEUE=normal
CORES=10
n_its=10000
burnin=3000
sample_list="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/01_Input/sample_names_bulk_yst.txt"

while read line;
do set $line
bsub -G team294-vwork -J"${1}" -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -q ${QUEUE} -e %J.dpclust.stderr -o %J.dpclust.stdout \
"xvfb-run -a Rscript --no-save --no-restore --verbose /nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw/DPClust/example/dpclust_pipeline.R -r 1 -d $WORK_DIR -o $OUTPUT_DIR -i ${WORK_DIR}${1}"_ProjectFile.txt" -k -a nd_dp --iterations ${n_its} --burnin ${burnin} --mut_assignment_type 1 --min_muts_cluster -1 --min_frac_muts_cluster 0.01 --num_muts_sample 4000 --bin_size NA --seed 42 > ${1}.dpclust.outputFile.Rout 2>&1"
done < $sample_list

```

##VARIANT ANNOTATION

###-Subs (LCM and bulk)

####Functions

```{r}

library(VariantAnnotation)

annotated_subs_text_file = function(vcf_file, type){
  if(type == 'lcm'){
    
    vcf = readVcf(vcf_file)
    sample = unlist(strsplit(vcf_file, '_post_swater_final_snvs.annot.vcf'))
    case = substr(sample, 0, 7)
    reads = cbind((geno(vcf)$FAZ)[,2]+(geno(vcf)$RAZ)[,2],(geno(vcf)$FCZ)[,2]+(geno(vcf)$RCZ)[,2],
                (geno(vcf)$FGZ)[,2]+(geno(vcf)$RGZ)[,2],(geno(vcf)$FTZ)[,2]+(geno(vcf)$RTZ)[,2])
    var.df = data.frame(Chr = as.character(seqnames(rowRanges(vcf))),
                    Position = start(ranges(rowRanges(vcf))),
                    Wildtype = as.character(ref(vcf)),
                    Mutant = unlist(CharacterList(alt(vcf))))
    var.df$Total_read_depth=rowSums(reads)
    var.df$Variant_read_depth=NA
    var.df$Case_ID = case
    var.df$Sample = sample
    var.df = var.df[, c(7, 8, 1:6)]
    colnames(reads)=c("A","C","G","T")
    for (k in c("A","C","G","T")){
      var.df$Variant_read_depth[var.df$Mutant==k] = reads[var.df$Mutant==k,k]
    }
  
    var.df$VAF = var.df$Variant_read_depth/ var.df$Total_read_depth
    var.df$Mutation = info(vcf)$VT
    vagrent_annots = lapply(strsplit(info(vcf)$VW, '\\|'), `length<-`, max(lengths(strsplit(info(vcf)$VW, '\\|')))) #make all elements in list the same length for subsetting purposes
    var.df$Gene = sapply(vagrent_annots, '[[', 1)
    var.df$Transcript = sapply(vagrent_annots, '[[', 3)
    var.df$Codon = sapply(vagrent_annots, '[[', 4)
    var.df$Protein = sapply(vagrent_annots, '[[', 5)
    var.df$Consequence = unlist(lapply(info(vcf)$VC, function(x) if(identical(x, character(0))) NA_character_ else x))
    write.table(var.df, paste0(sample, '_post_swater_final_snvs.annot.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
    
  }
  
  if(type == 'bulk'){
    
    vcf = readVcf(vcf_file)
    sample = unlist(strsplit(vcf_file, '.caveman_c.annot_filtered.vcf'))
    case = substr(sample, 0, 7)
    reads = cbind((geno(vcf)$FAZ)[,2]+(geno(vcf)$RAZ)[,2],(geno(vcf)$FCZ)[,2]+(geno(vcf)$RCZ)[,2],
                (geno(vcf)$FGZ)[,2]+(geno(vcf)$RGZ)[,2],(geno(vcf)$FTZ)[,2]+(geno(vcf)$RTZ)[,2])
    var.df = data.frame(Chr = as.character(seqnames(rowRanges(vcf))),
                    Position = start(ranges(rowRanges(vcf))),
                    Wildtype = as.character(ref(vcf)),
                    Mutant = unlist(CharacterList(alt(vcf))))
    var.df$Total_read_depth=rowSums(reads)
    var.df$Variant_read_depth=NA
    var.df$Case_ID = case
    var.df$Sample = sample
    var.df = var.df[, c(7, 8, 1:6)]
    colnames(reads)=c("A","C","G","T")
    for (k in c("A","C","G","T")){
      var.df$Variant_read_depth[var.df$Mutant==k] = reads[var.df$Mutant==k,k]
    }
  
    var.df$VAF = var.df$Variant_read_depth/ var.df$Total_read_depth
    var.df$Mutation = info(vcf)$VT
    vagrent_annots = lapply(strsplit(info(vcf)$VW, '\\|'), `length<-`, max(lengths(strsplit(info(vcf)$VW, '\\|')))) #make all elements in list the same length for subsetting purposes
    var.df$Gene = sapply(vagrent_annots, '[[', 1)
    var.df$Transcript = sapply(vagrent_annots, '[[', 3)
    var.df$Codon = sapply(vagrent_annots, '[[', 4)
    var.df$Protein = sapply(vagrent_annots, '[[', 5)
    var.df$Consequence = unlist(lapply(info(vcf)$VC, function(x) if(identical(x, character(0))) NA_character_ else x))
    write.table(var.df, paste0(sample, '.caveman_c.annot_filtered.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
    
    }
}
  

```

####Run

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/subs')

subs_vcf = list.files('.', 'vcf')
sample_type = ifelse(grepl('PD506', subs_vcf), 'bulk', 'lcm')

for(i in 1:length(subs_vcf)){
  annotated_subs_text_file(vcf_file = subs_vcf[i], type = sample_type[i])
}

```

###-Indels (LCM and bulk)

####Functions

```{r}

#bulk_pindel_vcf_to_text_no_filter
#convert putative indels vcf into text file
#input file is *.pindel.pass.hp7_threshold.qual.depth.direction.filtered.nogermline.5bp_window.nosnp_1bp_window.annot.vcf

bulk_pindel_vcf_to_text_no_filter = function(vcf_file){
  vcf = readVcf(vcf_file)
  sample = unlist(strsplit(vcf_file, '.pindel.pass.hp7_threshold.qual.depth.direction.filtered.nogermline.5bp_window.nosnp_1bp_window.annot.vcf'))
  var.df = data.frame(Chr=as.character(seqnames(rowRanges(vcf))),
                  Position=start(ranges(rowRanges(vcf))),
                  Wildtype=as.character(ref(vcf)),
                  Mutant=as.character(unlist(CharacterList(alt(vcf)))),
                  width=width(ranges(rowRanges(vcf))),
                  Variant_read_depth=geno(vcf)$NU[,"TUMOUR"]+geno(vcf)$PU[,"TUMOUR"],
                  Total_read_depth=geno(vcf)$NR[,"TUMOUR"]+geno(vcf)$PR[,"TUMOUR"])
  var.df$VAF = var.df$Variant_read_depth / var.df$Total_read_depth
  var.df$Mutation = info(vcf)$VT
  vagrent_annots = lapply(strsplit(info(vcf)$VW, '\\|'), `length<-`, max(lengths(strsplit(info(vcf)$VW, '\\|')))) #make all elements in list the same length for subsetting purposes
  var.df$Gene = sapply(vagrent_annots, '[[', 1)
  var.df$Transcript = sapply(vagrent_annots, '[[', 3)
  var.df$Codon = sapply(vagrent_annots, '[[', 4)
  var.df$Protein = sapply(vagrent_annots, '[[', 5)
  var.df$Consequence = unlist(lapply(info(vcf)$VC, function(x) if(identical(x, character(0))) NA_character_ else x))
  var.df$Case_ID = substr(sample, 0, 7)
  var.df$Sample = sample
  
  
  write.table(var.df[, c("Case_ID", "Sample", "Chr", "Position", "Wildtype", "Mutant", "Total_read_depth", "Variant_read_depth", "VAF", "Mutation", "Gene",      "Transcript", "Codon", "Protein", "Consequence")], paste0(sample, '.pindel.pass.hp7_threshold.qual.depth.direction.filtered.nogermline.5bp_window.nosnp_1bp_window.annot.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
  
}

#lcm_pindel_vcf_to_text_no_filter
#convert putative indels vcf into text file
#input file is *_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.nosnp_1bp_window.annot.vagrent.annot.vcf
lcm_pindel_vcf_to_text_no_filter = function(vcf_file){
  vcf = readVcf(vcf_file)
  sample = unlist(strsplit(vcf_file, '_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.nosnp_1bp_window.annot.vagrent.annot.vcf'))
  var.df = data.frame(Chr=as.character(seqnames(rowRanges(vcf))),
                  Position=start(ranges(rowRanges(vcf))),
                  Wildtype=as.character(ref(vcf)),
                  Mutant=as.character(unlist(CharacterList(alt(vcf)))),
                  Variant_read_depth=geno(vcf)$MTR[,2],
                  Total_read_depth=geno(vcf)$MTR[,2]+geno(vcf)$WTR[,2])
  var.df$VAF = var.df$Variant_read_depth / var.df$Total_read_depth
  var.df$Mutation = info(vcf)$VT
  vagrent_annots = lapply(strsplit(info(vcf)$VW, '\\|'), `length<-`, max(lengths(strsplit(info(vcf)$VW, '\\|')))) #make all elements in list the same length for subsetting purposes
  var.df$Gene = sapply(vagrent_annots, '[[', 1)
  var.df$Transcript = sapply(vagrent_annots, '[[', 3)
  var.df$Codon = sapply(vagrent_annots, '[[', 4)
  var.df$Protein = sapply(vagrent_annots, '[[', 5)
  
  conseq_list = lapply(info(vcf)$VC, function(x) if(identical(x, character(0))) NA_character_ else x)
  conseq_vec = c()
  for(i in 1:length(conseq_list)){
    conseq_vec = c(conseq_vec, paste(unlist(conseq_list[i]), collapse = ';'))
  }
  
  var.df$Consequence = conseq_vec
  var.df$Case_ID = substr(sample, 0, 7)
  var.df$Sample = sample
  
  write.table(var.df[, c("Case_ID", "Sample", "Chr", "Position", "Wildtype", "Mutant", "Total_read_depth", "Variant_read_depth", "VAF", "Mutation", "Gene",      "Transcript", "Codon", "Protein", "Consequence")], paste0(sample, '_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.nosnp_1bp_window.annot.vagrent.annot.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
  
}

```

####Run

Move files to final repository

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/bulk_yst/variant_calls/indels

cp *.pindel.pass.hp7_threshold.qual.depth.direction.filtered.nogermline.5bp_window.nosnp_1bp_window.annot.vcf /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering
cp *_indels_geno.final.post_swater_final_indels.AMB_UNKfiltered.coverageFiltered.hp7_filtered.nogermline.5bp_window.annot.nosnp_1bp_window.annot.vagrent.annot.vcf ../

```

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/')

vcf_list = list.files('.', pattern = 'vcf')

for(i in vcf_list[grepl('PD506', vcf_list)]){
  bulk_pindel_vcf_to_text_no_filter(i)
}

for(i in vcf_list[!grepl('PD506', vcf_list)]){
  lcm_pindel_vcf_to_text_no_filter(i)
}

```

###-Unified list of annotated subs and indels

```{r}

sub_files = list.files('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/subs', pattern = 'txt', full.names = T)
indel_files = list.files('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels', pattern = 'txt', full.names = T)
all_files = c(sub_files, indel_files)

all_var = c()
for(i in all_files){
  data = read.table(i, header = T, sep = '\t', stringsAsFactors = F)
  all_var = rbind(all_var, data)
}

all_var$Chr = factor(all_var$Chr, levels = c(1:22, 'X', 'Y'))
all_var = all_var[order(all_var$Case_ID, all_var$Sample, all_var$Chr, all_var$Position),]

write.table(all_var, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/gct_vagrent_ssm_all_mutations_plus_depth20210915.txt', col.names = T, row.names = F, quote = F, sep = '\t')

uniq_all_var = all_var[!duplicated(all_var[,c(1, 3, 4, 5, 6)]),c(1, 3, 4, 5, 6)]

write.table(uniq_all_var, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/gct_uniq_ssm20210915.txt', col.names = T, row.names = F, quote = F, sep = '\t')

```

#02_RNA_PROCESSING

##Filtering
Raw count table, preQC with genes mapped to nuclear genome only.

```{r}

library(ggplot2)
library(ggpubr)

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #just normal testis and GCT raw count data, 500 microbiopsies

#remove transcriptomes that were:
#- identified as contaminated during microdissection
#- non-neoplastic tissues that are not healthy seminiferous tubules
#- uncertainty as to the histological categorisation of the microbiopsy

raw_data = raw_data[, !colnames(raw_data) %in% c("PD43298a_Z1_2_SL02_04_SEC02_04_F2", 'PR43299c_lo0009', 'PR43299c_lo0010', 'PR43299c_lo0011', 'PR43299c_lo0012', "PD43296a_SL02_SEC02_G1", "PD43296a_Z1_2_SL02_04_SEC02_04_E1A", "PD43298a_SL02_SEC02_B3", "PD43298a_SL02_SEC02_C3", "PD43298a_Z1_2_SL02_04_SEC02_04_G2", "PR43299c_lo0002", "PR43299c_lo0006", "PR42036a_SL0D_SEC02_A5", "PR42036a_SL0D_SEC02_G3", "PR42036a_SL0D_SEC02_H3", "PR43298a_SL04_SEC04_D5",             "PR43298a_SL04_SEC04_E5" ,"PR43298a_SL04_SEC04_G9", "PR45543a_SL02_SEC03_B4", "PR45543a_SL02_SEC03_D2", "PR45543a_SL02_SEC03_F2", "PR46269c_SL02_SEC03_E12", "PR46269c_SL02_SEC03_F12", "PR46270c_SL02_SEC03_B5", "PR46270c_SL02_SEC03_C5", "PR46270c_SL02_SEC03_D5", "PR46270c_SL02_SEC03_E3", "PR46270c_SL02_SEC03_F3",             "PR46270c_SL02_SEC03_G3", "PR46270c_SL02_SEC03_H3", "PR46967a_SL02_SEC03_D8", "PR46967a_SL02_SEC03_E4", "PR46967a_SL02_SEC03_G6", "PR46968d_SL02_SEC02_03_H5", "PR46969c_SL02_SEC03_C11", "PR46969c_SL02_SEC03_D11", "PR46969c_SL02_SEC03_E11", "PR46969c_SL02_SEC03_F11", "PR46969c_SL02_SEC03_G11", "PR46969c_SL02_SEC03_H11")] #460 samples left

#how many reads, across all 55502 mapped features do we have?
summary(colSums(raw_data[, 7:ncol(raw_data)])) #highly variable library sizes
#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#   682   86612  246980  502107  651605 3863511 

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/all_samples_read_count_QC.pdf', height = 6, width = 6, useDingbats = F)
ggplot() +
  geom_histogram(mapping = aes(colSums(raw_data[, 7:ncol(raw_data)])), bins = 100, fill = '#999999') + theme_pubr() + coord_cartesian(expand = F) + xlab('Number of reads across all mapped features') + ylab('Number of microbiopsies')
#spike in samples with under 1000 features expressed
dev.off()

#if we choose a threshold of 5 reads as the minimum to consider a gene truly expressed, how many are expressed per microbiopsy?
summary(colSums(raw_data[, 7:ncol(raw_data)] >= 5))
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#   0    4150    8916    9077   13654   22865 

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/all_samples_feature_expression_QC.pdf', height = 6, width = 6, useDingbats = F)
ggplot() +
  geom_histogram(mapping = aes(colSums(raw_data[, 7:ncol(raw_data)] >= 5)), bins = 100, fill = '#999999') + theme_pubr() + coord_cartesian(expand = F) + geom_vline(xintercept = 1000, lty = 2, col = 'red') + xlab('Features expressed per microbiopsy to a depth of at least 5 reads') + ylim(0, 15) + ylab('Number of microbiopsies')
#spike in samples with under 1000 features expressed
dev.off()

sum(colSums(raw_data[, 7:ncol(raw_data)] >= 5) >= 1000) #416

filtered_data = raw_data[, c(rep(TRUE, 6), colSums(raw_data[, 7:ncol(raw_data)] >= 5) >= 1000)]

write.table(filtered_data, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/germ_cell_lcm_raw_counts_filtered_samples_20210812.txt', col.names = T, row.names = F, quote = F, sep = '\t')

#generate tpm values
x = filtered_data[, 7:ncol(filtered_data)]
row.names(x) = filtered_data$Geneid
x = x / filtered_data$Length #normalise for transcript length
tpm <- t( t(x) * 1e6 / colSums(x)) #normalise to read depth
write.table(tpm, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/germ_cell_lcm_tpm_counts_filtered_samples_20210812.txt', col.names = T, row.names = T, quote = F, sep = '\t')

#add total counts to mrna data
mrna = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #descriptive column contains apostrophes
#raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/rna_counts/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #raw count data
raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/germ_cell_lcm_raw_counts_filtered_samples_20210812.txt', header = T, sep = '\t', stringsAsFactors = F) #raw count data
row.names(mrna) = mrna$Sample
mrna$raw_counts = colSums(raw_data[, 7:ncol(raw_data)])[row.names(mrna)]
summary(mrna$raw_counts)
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 13822  125931  298080  553863  716367 3863511 

write.table(mrna, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', col.names = T, row.names = F, sep = '\t', quote = F)

```

##Inspect clustering of final sample list

Use Seurat to establish if the final samples cluster by histology/tumour/patient/sequencing run to check if RNA of sufficient quality to proceed.

```{r}

library(Seurat)
library(ggplot2)
library(ggbeeswarm)
library(cowplot)
library(ggpubr)

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F)
mrna = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '')

row.names(mrna) = mrna$Sample
raw_counts = raw_data[, row.names(mrna)]
row.names(raw_counts) = raw_data$Geneid

#check that the tmm-normalised counts cluster by histology before performing DE analysis
gct_obj <- CreateSeuratObject(counts = raw_counts, meta.data = mrna, project = 'gct')
gct_obj <- NormalizeData(gct_obj, normalization.method = "LogNormalize", scale.factor = 10000)
gct_obj <- FindVariableFeatures(gct_obj, selection.method = "vst", nfeatures = 1000)
var.genes = VariableFeatures(gct_obj)
top10 <- head(VariableFeatures(gct_obj), 10)
plot1 <- VariableFeaturePlot(gct_obj)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
#all.genes <- rownames(gct_obj)
gct_obj  <- ScaleData(gct_obj , features = var.genes)
gct_obj <- RunPCA(gct_obj, features = VariableFeatures(object = gct_obj), npcs = 100)
DimPlot(gct_obj, reduction = "pca", group.by = 'Updated.Description')
ElbowPlot(gct_obj, ndims = 100)
gct_obj <- RunUMAP(gct_obj, dims = 1:30)

p1 = DimPlot(gct_obj, reduction = "umap", group.by = 'Updated.Description')
p2 = DimPlot(gct_obj, reduction = "umap", group.by = 'Diagnosis')
p3 = DimPlot(gct_obj, reduction = "umap", group.by = 'Case.ID')
p4 = DimPlot(gct_obj, reduction = "umap", group.by = 'Batch')

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/gct_seurat_30_dims_4_umaps_20210920.pdf', width = 13, height = 7, useDingbats = F)
plot_grid(p1, p2, p3, p4, ncol = 2, nrow = 2, align = 'v')
dev.off()

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/gct_seurat_30_dims_6_marker_genes_20210920.pdf', width = 8, height = 10.5, useDingbats = F)
FeaturePlot(gct_obj, features = c("KIT", "SOX17", "SOX2", "TNFRSF8", "AFP", "GPC3"), reduction = 'umap')
dev.off()

saveRDS(gct_obj, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/seurat/lcm_gct_seurat_object_20210920.rds')

#would using 20 components instead matter?
gct_obj = readRDS('/lustre/scratch126/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/seurat/lcm_gct_seurat_object_20210920.rds')

ElbowPlot(gct_obj, ndims = 100)
gct_obj1 <- RunUMAP(gct_obj, dims = 1:20)

DimPlot(gct_obj1, reduction = "umap", group.by = 'Updated.Description')
DimPlot(gct_obj1, reduction = "umap", group.by = 'Diagnosis')
DimPlot(gct_obj1, reduction = "umap", group.by = 'Case.ID')
DimPlot(gct_obj1, reduction = "umap", group.by = 'Batch')

```

##Retrieve MT counts *TO DO*

We performed our QC checks on the counts minus MT genes. For completeness and transparency in the supplementary material, I will retrieve the MT gene counts for these samples too.

```{r}

library(dplyr)

filtered_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/germ_cell_lcm_raw_counts_filtered_samples_20210812.txt', sep = '\t', header = T, stringsAsFactors = F)
row.names(filtered_data) = filtered_data$Geneid
mrna = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '')


batch1 = read.table("/lustre/scratch119/casm/team294rr/rr11/LCM_RNA/LCM_RNA_testes_normal_tumour/featurecounts/GRCh37_ERCC92_FeatureCounts.txt", header = T, sep = '\t', stringsAsFactors = F)
colnames(batch1) = c(colnames(batch1)[1:6], unlist(strsplit(unlist(strsplit(colnames(batch1)[7:length(colnames(batch1))], "..star_bam_GRCh37_ERCC92."))[c(F,T)], "Aligned.sortedByCoord.out.bam")))
row.names(batch1) = batch1$Geneid



batch2 = read.table("/lustre/scratch119/casm/team294rr/rs30/LCM_RNA_testes_normal_tumour_34959_analysis/featurecounts-Jan2/GRCh37_ERCC92_FeatureCounts.txt", header = T, sep = '\t', stringsAsFactors = F)
colnames(batch2) = c(colnames(batch2)[1:6], unlist(strsplit(unlist(strsplit(colnames(batch2)[7:length(colnames(batch2))], "X.lustre.scratch119.casm.team294rr.shared.LCM_RNA_testes_normal_tumour_34959."))[c(F,T)], ".bam")))
row.names(batch2) = batch2$Geneid

#batch2 = read.table("/lustre/scratch117/casm/team294/rr11/LCM_RNA/featurecounts/featurecounts_GRCh37_ERCC92.txt", header = T, sep = '\t', stringsAsFactors = F)
#colnames(batch2) = c(colnames(batch2)[1:6], unlist(strsplit(unlist(strsplit(colnames(batch2)[7:length(colnames(batch2))], "..Recalibration_bams."))[c(F,T)], ".cDNA.AnalysisReady.bam")))
#row.names(batch2) = batch2$Geneid

table(mrna$Batch)
#1   2 
#96 320

ncol(batch1[,colnames(batch1) %in% colnames(filtered_data)]) - 6 #96
ncol(batch2[,colnames(batch2) %in% colnames(filtered_data)]) - 6 #320 found the featurecounts for the samples from each run

original_fc = cbind(batch1[union(row.names(batch1), row.names(batch2)),], batch2[union(row.names(batch1), row.names(batch2)), c(7: ncol(batch2))])

#nrow(filtered_data[!row.names(filtered_data) %in% row.names(original_fc),]) 0, all genes found in the count table used for analysis in these original files

original_fc_proj = original_fc[, colnames(original_fc) %in% colnames(filtered_data)]

all.equal(filtered_data, original_fc_proj[row.names(filtered_data),])

df = data.frame((colSums(filtered_data[, 7:ncol(filtered_data)]) / colSums(original_fc_proj[row.names(filtered_data),][, 7:ncol(original_fc_proj)]))[(colSums(filtered_data[, 7:ncol(filtered_data)]) / colSums(original_fc_proj[row.names(filtered_data),][, 7:ncol(original_fc_proj)])) != 1])

df$sample = row.names(df)
names(df) = c("to3_fc_rs30_fc", "sample")

ggplot(df) + geom_histogram(mapping = aes(x = to3_fc_rs30_fc)) + xlim(0,1) + xlab("to3 featurecounts / rs30 featurecounts")

```

#03_DNA_ANALYSES

##BASIC WGS STATS

###Substitution and indel burden + VAF

```{r}

#subs
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/subs/combined')

#prior = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)
manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)

sub_files = list.files('.', 'txt')

for(i in 1:nrow(manifest)){
  sub_file = sub_files[grepl(manifest$Sample[i], sub_files)]
  data = read.table(sub_file, header = T, sep = '\t', stringsAsFactors = F)
  manifest$SNV.burden[i] = nrow(data)
  manifest$SNV.median.VAF[i] = median(data[data$Chr %in% c(1:22),]$VAF)
  manifest$Autosomal.SNVs[i] = nrow(data[data$Chr %in% c(1:22),])
  manifest$Autosomal.SNVs.Mb[i] = nrow(data[data$Chr %in% c(1:22) & data$Total_read_depth >= 4,]) / manifest$Autosomal.callable.region.length[i] * 1e6 #callable region determined as a minimum read depth of 4 so need to limit variants in a given sample to those with a minimum total read depth of 4
}

#sum(prior$SNV.burden != manifest$SNV.burden) 0
#sum(prior$SNV.median.VAF != manifest$SNV.median.VAF) 9 but these must be rounding errors because:
#prior$SNV.median.VAF[prior$SNV.median.VAF != manifest$SNV.median.VAF]
#manifest$SNV.median.VAF[manifest$SNV.median.VAF != prior$SNV.median.VAF]
#look the same to 7 d.p.
#sum(prior$Autosomal.SNVs != manifest$Autosomal.SNVs) 0
#sum(prior$Autosomal.SNVs.Mb != manifest$Autosomal.SNVs.Mb) 127 but these must be rounding errors because:
#prior$Autosomal.SNVs.Mb[prior$Autosomal.SNVs.Mb != manifest$Autosomal.SNVs.Mb]
#manifest$Autosomal.SNVs.Mb[manifest$Autosomal.SNVs.Mb != prior$Autosomal.SNVs.Mb]
#look the same to 7 d.p.


#indels
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/indels/combined')

indel_files = list.files('.', 'txt')

for(i in 1:nrow(manifest)){
  indel_file = indel_files[grepl(manifest$Sample[i], indel_files)]
  data = read.table(indel_file, header = T, sep = '\t', stringsAsFactors = F)
  manifest$Indel.burden[i] = nrow(data)
  manifest$Indel.median.VAF[i] = median(data[data$Chr %in% c(1:22),]$VAF)
  manifest$Autosomal.indels[i] = nrow(data[data$Chr %in% c(1:22),])
  manifest$Autosomal.indels.Mb[i] = nrow(data[data$Chr %in% c(1:22) & data$Total_read_depth >= 4,]) / manifest$Autosomal.callable.region.length[i] * 1e6
}

#sum(prior$Indel.burden != manifest$Indel.burden) 0
#sum(prior$Indel.median.VAF != manifest$Indel.median.VAF) 29 but these must be rounding errors because:
#prior$Indel.median.VAF[prior$Indel.median.VAF != manifest$Indel.median.VAF]
#manifest$Indel.median.VAF[manifest$Indel.median.VAF != prior$Indel.median.VAF]
#look the same to 7 d.p.
#sum(prior$Autosomal.indels != manifest$Autosomal.indels) 0
#sum(prior$Autosomal.indels.Mb != manifest$Autosomal.indels.Mb) 130 but these must be rounding errors because:
#prior$Autosomal.indels.Mb[prior$Autosomal.indels.Mb != manifest$Autosomal.indels.Mb]
#manifest$Autosomal.indels.Mb[manifest$Autosomal.indels.Mb != prior$Autosomal.indels.Mb]
#look the same to 7 d.p.

write.table(manifest, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', col.names = T, row.names = F, sep = '\t', quote = F)

#muts per Mb, invasive LCM data only
manifest[manifest$Histology != 'GCNIS' & manifest$Experimental_arm == 'LCM',]

median(aggregate(Autosomal.SNVs.Mb ~ Case.ID, manifest[manifest$Histology != 'GCNIS' & manifest$Experimental_arm == 'LCM',], median)[,2]) #0.4920662
median(aggregate(Autosomal.indels.Mb ~ Case.ID, manifest[manifest$Histology != 'GCNIS' & manifest$Experimental_arm == 'LCM',], median)[,2]) #0.02766861

#in situ versus invasive, matched per patient
median(manifest[manifest$Case.ID == 'PD42036' & manifest$Histology == "GCNIS",]$SNV.burden) #511
median(manifest[manifest$Case.ID == 'PD42036' & manifest$Histology == "Seminoma",]$SNV.burden) #1324
wilcox.test(manifest[manifest$Case.ID == 'PD42036' & manifest$Histology == "GCNIS",]$SNV.burden, manifest[manifest$Case.ID == 'PD42036' & manifest$Histology == "Seminoma",]$SNV.burden, alternative = "two.sided") #0.0003996

#compare purity too
median(manifest[manifest$Case.ID == 'PD42036' & manifest$Histology == "GCNIS",]$Purity) #0.336
median(manifest[manifest$Case.ID == 'PD42036' & manifest$Histology == "Seminoma",]$Purity) #0.839655
median(manifest[manifest$Case.ID == 'PD46966' & manifest$Histology == "GCNIS",]$Purity) #0.54648
median(manifest[manifest$Case.ID == 'PD46966' & manifest$Histology == "Seminoma",]$Purity) #0.60575

median(manifest[manifest$Case.ID == 'PD46966' & manifest$Histology == "GCNIS",]$SNV.burden) #893
median(manifest[manifest$Case.ID == 'PD46966' & manifest$Histology == "Seminoma",]$SNV.burden) #1205.5
wilcox.test(manifest[manifest$Case.ID == 'PD46966' & manifest$Histology == "GCNIS",]$SNV.burden, manifest[manifest$Case.ID == 'PD46966' & manifest$Histology == "Seminoma",]$SNV.burden, alternative = "two.sided") #0.02381

```

###Purity, ploidy & reads per chromosome copy

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/cnas/01_battenberg')

#prior = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)
manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)

cna_files = list.files('.', '_cellularity_ploidy.txt')

for(i in 1:nrow(manifest)){
  cna_file = cna_files[grepl(manifest$Sample[i], cna_files)]
  data = read.table(cna_file, header = T, sep = '\t', stringsAsFactors = F)
  manifest$Purity[i] = data$cellularity
  manifest$Ploidy[i] = data$ploidy
  manifest$Average.reads.per.chromosome.copy.[i] = manifest$Purity[i] / (manifest$Purity[i] * manifest$Ploidy[i] + (1 - manifest$Purity[i]) * 2) * manifest$Coverage[i]
}

#sum(prior$Purity != manifest$Purity) 0
#sum(prior$Ploidy != manifest$Ploidy) 0
#sum(prior$Average.reads.per.chromosome.copy. != manifest$Average.reads.per.chromosome.copy.) 130 but these are probably rounding errors, checked using:
#manifest$Average.reads.per.chromosome.copy. - prior$Average.reads.per.chromosome.copy.

write.table(manifest, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', col.names = T, row.names = F, sep = '\t', quote = F)

```


###Cancer cell fraction (Extended data figure 2)

Only plotted for those where we have sufficient tumour depth to confidently call copy number (5>= reads per chromosome copy). For consistency across samples/tumours, we will use the single sample DPClust output.

```{r}

library(ggplot2)

#prior = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)
manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)
#min_dep_prior = prior[prior$Average.reads.per.chromosome.copy. >= 5,]
min_dep_manifest = manifest[manifest$Average.reads.per.chromosome.copy. >= 5,]

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/ssDPClust')

input_files = list.files('.', '_allDirichletProcessInfo.txt')

manifest$Median.CCF = NA
all_input = c()
for(file in input_files){
  if(file %in% paste0(min_dep_manifest$Sample, '_allDirichletProcessInfo.txt')){
    data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
    data$sample = unlist(strsplit(file, '_allDirichletProcessInfo.txt'))
    all_input = rbind(all_input, data)
    manifest[manifest$Sample == unlist(strsplit(file, '_allDirichletProcessInfo.txt')),]$Median.CCF = median(data$subclonal.fraction, na.rm = T)
  }
}

write.table(all_input, "/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/00_SUPPLEMENTARY_DATA/supplementary_figure_2_input_20220321.txt", col.names = T, row.names = F, sep = '\t', quote = F)

#sum(prior$Median.CCF != manifest$Median.CCF, na.rm = T) 20, rounding errors as determined by
#prior$Median.CCF - manifest$Median.CCF

#median CCF
median(manifest$Median.CCF, na.rm = T) #0.9643007
range(manifest$Median.CCF, na.rm = T) #0.7299857 1.0804695

#median CCF for microdissections only
median(manifest[manifest$Experimental_arm == 'LCM',]$Median.CCF, na.rm = T) #0.9642857
range(manifest[manifest$Experimental_arm == 'LCM',]$Median.CCF, na.rm = T) #0.8054448 1.0804695

#pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/per_sample_ccf_plots_20210918.pdf', height = 9,  width = 10, useDingbats = F)
#ggplot(all_input) + geom_histogram(mapping = aes(x = subclonal.fraction)) + facet_wrap(.~sample, ncol = 8, nrow = 16) + ylab('Autosomal substitution count') + xlab('Cancer cell fraction') + geom_vline(xintercept = 1, lty = 2 , col = 'red') + theme_linedraw() + theme(axis.text = element_text(size = 4), strip.text = element_text(size = 6))
#dev.off()

g = ggplot(all_input) + geom_histogram(mapping = aes(x = subclonal.fraction)) + ylab('Autosomal substitution count') + xlab('Cancer cell fraction') + geom_vline(xintercept = 1, lty = 2 , col = 'red') + theme_linedraw() + theme(axis.text = element_text(size = 4), strip.text = element_text(size = 7))

library(tidyverse)
#Create similar data for labels
Labels <- all_input %>% group_by(sample) %>% summarise(n=paste0('n = ',n()))
Labels$ccf <- 2.5
Labels$count_n = 200 

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/06_FIGURES/per_sample_ccf_plots_20220321.pdf', height = 9,  width = 10, useDingbats = F)
g + geom_text(data = Labels, aes(x = ccf, y = count_n, label = n)) + facet_wrap(.~sample, ncol = 8, nrow = 16)
dev.off()

write.table(manifest, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', col.names = T, row.names = F, sep = '\t', quote = F)

```

###Non-RT structural variants

Merge the calls from bulk and lcm data into final call set.

```{r}

bulk = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/bulk/04_rt/final_bulk_gct_brass_calls_20220118.txt", header = T, sep = '\t', stringsAsFactors = F)
bulk.2 = bulk[,c("chr1", "start1", "end1", "chr2", "start2", "end2", "strand1", "strand2", "sample", "svclass", "bkdist", "non.template", "micro.homology", "gene1" , "gene2", "fusion_flag", "Genes.in.dup.del", "Cancer.gene", "Dom.Rec.COSMICv94")]

lcm = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/03_rt/final_lcm_gct_brass_calls_20220118.txt", header = T, sep = '\t', stringsAsFactors = F)
lcm.2 = lcm[,c("chr1", "start1", "end1", "chr2", "start2", "end2", "strand1", "strand2", "sample", "svclass", "bkdist", "non.template", "micro.homology", "gene1" , "gene2", "fusion_flag", "Genes.in.dup.del", "Cancer.gene", "Dom.Rec.COSMICv94")]
lcm.2$sample = unlist(lapply(strsplit(lcm.2$sample, ","), "[[", 1)) #a few LCM samples that pass filters have matched sample calls still

combined_svs = rbind(lcm.2, bulk.2)

write.table(combined_svs, "/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/final_combined_gct_brass_calls_20220118.txt", sep = '\t', quote = F, col.names = T, row.names = F)

#just for adding to supplementary tables
#bulk = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/bulk/04_rt/final_bulk_gct_brass_calls_20220118.txt", header = T, sep = '\t', stringsAsFactors = F)
#bulk$case.id = substr(bulk$sample, 0, 7)
#bulk.2 = bulk[,c("sample", "case.id", "chr1", "start1", "end1", "strand1", "gene1", "region1", "chr2", "start2", "end2", "strand2", "gene2", "region2", "svclass", "bkdist", "non.template", "micro.homology")]

#lcm = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/svs/lcm/03_rt/final_lcm_gct_brass_calls_20220118.txt", header = T, sep = '\t', stringsAsFactors = F)
#lcm$case.id = substr(lcm$sample, 0, 7)
#lcm.2 = lcm[,c("sample", "case.id", "chr1", "start1", "end1", "strand1", "gene1", "region1", "chr2", "start2", "end2", "strand2", "gene2", "region2", "svclass", "bkdist", "non.template", "micro.homology")]
#lcm.2$sample = unlist(lapply(strsplit(lcm.2$sample, ","), "[[", 1)) #a few LCM samples that pass filters have matched sample calls still

#combined_svs = rbind(lcm.2, bulk.2)

#write.table(combined_svs, "/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/final_combined_gct_brass_calls_20220122.txt", sep = '\t', quote = F, col.names = T, row.names = F)

```

Add counts to manifest

```{r}

#setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/structural_variants')

prior = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)
manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)

#rt_list = read.table('all_gct_rt_count.txt', header = F, sep = '\t', stringsAsFactors = F)[,1]
#lcm_brass_calls = read.table('final_lcm_gct_brass_calls_20210709.txt', header = T, sep = '\t', stringsAsFactors = F)
#bulk_brass_calls = read.table('final_bulk_gct_brass_calls_20210709.txt', header = T, sep = '\t', stringsAsFactors = F)
brass_calls = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/final_combined_gct_brass_calls_20220118.txt', header = T, sep = '\t', stringsAsFactors = F)

for(i in 1:nrow(manifest)){
  #if(manifest$Experimental_arm[i] == 'LCM'){
  manifest$dup_sv_count[i] = nrow(brass_calls[brass_calls$sample == manifest$Sample[i] & brass_calls$svclass == 'tandem-duplication',])
  manifest$del_sv_count[i] = nrow(brass_calls[brass_calls$sample == manifest$Sample[i] & brass_calls$svclass == 'deletion',])
  manifest$trans_sv_count[i] = nrow(brass_calls[brass_calls$sample == manifest$Sample[i] & brass_calls$svclass == 'translocation',])
  manifest$inv_sv_count[i] = nrow(brass_calls[brass_calls$sample == manifest$Sample[i] & brass_calls$svclass == 'inversion',])
  #}
  #if(manifest$Experimental_arm[i] == 'Bulk'){
  #  manifest$dup_sv_count[i] = nrow(bulk_brass_calls[bulk_brass_calls$sample == manifest$Sample[i] & bulk_brass_calls$svclass == 'tandem-duplication',])
  #  manifest$del_sv_count[i] = nrow(bulk_brass_calls[bulk_brass_calls$sample == manifest$Sample[i] & bulk_brass_calls$svclass == 'deletion',])
  #  manifest$trans_sv_count[i] = nrow(bulk_brass_calls[bulk_brass_calls$sample == manifest$Sample[i] & bulk_brass_calls$svclass == 'translocation',])
  #  manifest$inv_sv_count[i] = nrow(bulk_brass_calls[bulk_brass_calls$sample == manifest$Sample[i] & bulk_brass_calls$svclass == 'inversion',])
  #}
  #manifest$rt_sv_count[i] = length(rt_list[rt_list == manifest$Sample[i]])
}

manifest[manifest$dup_sv_count != prior$dup_sv_count,]$Sample #2 bulk samples since removal of all called in matched normal too
manifest[manifest$del_sv_count != prior$del_sv_count,]$Sample #2 additional calls in newly run samples plus removal from bulk sample of all called in matched normal too
manifest[manifest$trans_sv_count != prior$trans_sv_count,]$Sample #2 bulk samples since removal of all called in matched normal too
manifest[manifest$inv_sv_count != prior$inv_sv_count,]$Sample #2 bulk samples since removal of all called in matched normal too

nrow(brass_calls) #2732
sum(colSums(manifest[, 42:45])) #2732, all mapped back and accounted for


write.table(manifest, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', col.names = T, row.names = F, sep = '\t', quote = F)

```

###Retrotransposition events

```{r}

rt_ids = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/all_mgct_sample_IDs_only.txt", header = T, sep = '\t', stringsAsFactors = F) #298 but actually includes one call made in a sample that didn't pass QC (PD45545a_lo0006)

prior = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)
manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)

manifest$rt_sv_count = 0

for(j in 1:nrow(manifest)){
  manifest$rt_sv_count[j] = length(rt_ids[rt_ids$sample == manifest$Sample[j],])
}

#sum(manifest$rt_sv_count != prior$rt_sv_count) 0 same as before
sum(manifest$rt_sv_count) #297 consistent with input

write.table(manifest, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', col.names = T, row.names = F, sep = '\t', quote = F)

```

###Plot genomic overview, including signatures (Fig. 1b)

SBS signature extraction detailed below in the next section.

```{r}

library(reshape2)
library(ggplot2)
library(ggpubr)
library(cowplot)

#Read in burden, ploidy and purity data
manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', header = T, sep = '\t', stringsAsFactors = F)

#Read in SBS signatures
hdp_sigprofiler_mapped = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/sigprofiler_deconvolution_remapping/Decompose_Solution_Activities.txt', header = T, sep = '\t', stringsAsFactors = F)

summary_data = manifest[,c('Sample', 'Case.ID', 'Organ', 'Diagnosis', 'Histology', 'Age', 'Purity', 'Ploidy', 'SNV.burden', 'Indel.burden')]

summary_data$is.gcnis = ifelse(summary_data$Histology == 'GCNIS', 'yes', 'no')
summary_data$is.gcnis = factor(summary_data$is.gcnis, levels = c('yes', 'no'))
summary_data = summary_data[order(summary_data$Age, summary_data$Case.ID, summary_data$is.gcnis, summary_data$SNV.burden),]
summary_data$Case.ID = factor(summary_data$Case.ID, levels = unique(summary_data$Case.ID))
summary_data$Sample = factor(summary_data$Sample, levels = as.vector(summary_data$Sample))

p1 = ggplot(summary_data, mapping = aes(x = Sample, y = SNV.burden, fill = is.gcnis)) + 
  geom_bar(stat = 'identity') + 
  theme_pubr() + 
  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_rect(fill = 'grey97')) + 
  scale_y_continuous(expand = c(0,0), limits = c(0, 4000)) + 
  ylab('SNV burden') + 
  xlab('') + 
  scale_x_discrete(expand = c(0,0)) +
  facet_grid(. ~ Case.ID, scales = 'free_x') + 
  scale_fill_manual(breaks = c("yes", "no"), values=c("black", "#999999"))
p2 = ggplot(summary_data, mapping = aes(x = Sample, y = Indel.burden, fill = is.gcnis)) + 
  geom_bar(stat = 'identity') + 
  theme_pubr() + 
  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_rect(fill = 'grey97')) + 
  scale_y_continuous(expand = c(0,0), limits = c(0, 200)) + 
  ylab('Indel burden') + 
  xlab('') + 
  scale_x_discrete(expand = c(0,0)) +
  facet_grid(. ~ Case.ID, scales = 'free_x') + 
  scale_fill_manual(breaks = c("yes", "no"), values=c("black", "#999999"))
p3 = ggplot(summary_data, mapping = aes(x = Sample, y = Purity * 100)) + 
  geom_crossbar(mapping = aes(ymin = Purity * 100, ymax = Purity * 100, col = is.gcnis)) + 
  theme_pubr() + 
  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_rect(fill = 'grey97')) + 
  scale_y_continuous(expand = c(0,0), limits = c(0, 100)) + 
  ylab('Purity (%)') + 
  xlab('') + 
  scale_x_discrete(expand = c(0,0)) +
  facet_grid(. ~ Case.ID, scales = 'free_x') + 
  scale_color_manual(breaks = c("yes", "no"), values=c("black", "#999999"))
p4 = ggplot(summary_data, mapping = aes(x = Sample, y = Ploidy)) + 
 geom_crossbar(mapping = aes(ymin = Purity * 100, ymax = Purity * 100, col = is.gcnis)) + 
  theme_pubr() +
  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_rect(fill = 'grey97')) + 
  ylab('Ploidy') + 
  xlab('') + 
  scale_x_discrete(expand = c(0,0)) +
  facet_grid(. ~ Case.ID, scales = 'free_x') + 
  scale_color_manual(breaks = c("yes", "no"), values=c("black", "#999999")) +
  scale_y_continuous(expand = c(0.05,0.05), limits = c(0, 4.2)) 
p4 = p4 + coord_cartesian(ylim=c(1,4.2))

hdp_sigprofiler_mapped$SBS5_40 = hdp_sigprofiler_mapped$SBS5 + hdp_sigprofiler_mapped$SBS40 
hdp_sigprofiler_mapped = hdp_sigprofiler_mapped[, c("Samples", "SBS1", "SBS5_40", "SBS17a", "SBS17b", "SBS18", "SBS35", "SBS91", "N2")]

hdp_sigprofiler_mapped_norm = hdp_sigprofiler_mapped[, c(2:9)] / rowSums(hdp_sigprofiler_mapped[, c(2:9)])
hdp_sigprofiler_mapped_norm$sample = hdp_sigprofiler_mapped$Samples
hdp_sigprofiler_mapped_norm$tumour = substr(hdp_sigprofiler_mapped_norm$sample, 0, 7)

hdp_sigprofiler_mapped_norm.m = reshape2::melt(hdp_sigprofiler_mapped_norm, id.vars = c('sample', 'tumour'))

hdp_sigprofiler_mapped_norm.m$age = NA
hdp_sigprofiler_mapped_norm.m$diagnosis = NA
hdp_sigprofiler_mapped_norm.m$site = NA
hdp_sigprofiler_mapped_norm.m$histology = NA

for(i in 1:nrow(hdp_sigprofiler_mapped_norm.m)){
  hdp_sigprofiler_mapped_norm.m$age[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Age
  hdp_sigprofiler_mapped_norm.m$diagnosis[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Diagnosis
  hdp_sigprofiler_mapped_norm.m$site[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Organ
  hdp_sigprofiler_mapped_norm.m$histology[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Histology
}

hdp_sigprofiler_mapped_norm.m$variable = factor(hdp_sigprofiler_mapped_norm.m$variable, levels = rev(unique(hdp_sigprofiler_mapped_norm.m$variable)))
hdp_sigprofiler_mapped_norm.m$tumour = factor(hdp_sigprofiler_mapped_norm.m$tumour, levels = levels(summary_data$Case.ID))
hdp_sigprofiler_mapped_norm.m$sample = factor(hdp_sigprofiler_mapped_norm.m$sample, levels = levels(summary_data$Sample))

median(aggregate(value ~ tumour, hdp_sigprofiler_mapped_norm.m[hdp_sigprofiler_mapped_norm.m$variable == "N2" & hdp_sigprofiler_mapped_norm.m$value != 0,], median)[,2]) #0.2349856
min(aggregate(value ~ tumour, hdp_sigprofiler_mapped_norm.m[hdp_sigprofiler_mapped_norm.m$variable == "N2" & hdp_sigprofiler_mapped_norm.m$value != 0,], median)[,2]) #0.160037
max(aggregate(value ~ tumour, hdp_sigprofiler_mapped_norm.m[hdp_sigprofiler_mapped_norm.m$variable == "N2" & hdp_sigprofiler_mapped_norm.m$value != 0,], median)[,2]) #0.3820225

cols = get_palette(palette = 'Set1', k = length(unique(hdp_sigprofiler_mapped_norm.m$variable)))
cols = c("#F781BF", "#A65628", "#4DAF4A", "#984EA3", "#FF7F00", "#FFFF33", "dodgerblue", "#E52621")

h1 = ggplot(hdp_sigprofiler_mapped_norm.m) + 
  geom_bar(mapping = aes(x = sample, y = value, fill = variable), stat = 'identity') + 
  theme_pubr() + coord_cartesian(expand = F) + 
  scale_fill_manual(values=cols) + 
  ylab('Proportion of substitutions') + 
  xlab('') + 
  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_blank()) + facet_grid(. ~ tumour, scales = 'free_x')

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/20210919_lcm_genome_overview.pdf', height = 6, width = 11, useDingbats = F)
plot_grid(p1, p2, p3, p4, h1, ncol = 1, align = 'v', rel_heights = c(1, 1, 0.5, 0.5, 1))
dev.off()

```

##SBS SIGNATURES

###HDP (0.1.5, de novo signature extraction followed by fitting)

####Generate input files

Generate per patient mutation list.

```{bash}

cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/subs/*txt /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/raw_input

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/raw_input

```

Function to generate the counts per trinucleotide context

```{r}

count_muts_96 = function(mutations, genomeFile = "/nfs/cancer_ref01/Homo_sapiens/37/genome.fa"){
  library("GenomicRanges")
  library("Rsamtools")
  library("MASS")
  colnames(mutations) = c("chr","pos","ref","mut")
  mutations = mutations[(mutations$ref %in% c("A","C","G","T")) & (mutations$mut %in% c("A","C","G","T")) & mutations$chr %in% c(1:22,"X","Y"),]
  mutations$trinuc_ref = as.vector(scanFa(genomeFile, GRanges(mutations$chr, IRanges(as.numeric(mutations$pos)-1,
                                                                                     as.numeric(mutations$pos)+1))))
 
  # 2. Annotating the mutation from the pyrimidine base
  ntcomp = c(T="A",G="C",C="G",A="T")
  mutations$sub = paste(mutations$ref,mutations$mut,sep=">")
  mutations$trinuc_ref_py = mutations$trinuc_ref
  for (j in 1:nrow(mutations)) {
    if (mutations$ref[j] %in% c("A","G")) { # Purine base
      mutations$sub[j] = paste(ntcomp[mutations$ref[j]],ntcomp[mutations$mut[j]],sep=">")
      mutations$trinuc_ref_py[j] = paste(ntcomp[rev(strsplit(mutations$trinuc_ref[j],split="")[[1]])],collapse="")
    }
  }
  freqs = table(paste(mutations$sub,paste(substr(mutations$trinuc_ref_py,1,1),substr(mutations$trinuc_ref_py,3,3),sep="-"),sep=","))
  sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
  ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
  full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")
  freqs_full = freqs[full_vec]; freqs_full[is.na(freqs_full)] = 0; names(freqs_full) = full_vec
  return(freqs_full)
}

```

Generate input text files - unique subs per patient to avoid double counting.

```{r}

#setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/raw_input')
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp/01_INPUT')

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', header = T, stringsAsFactors = F, sep = '\t')

mut_txt_files = list.files('.', pattern = 'annot')

All_var = c()

#list of unique subs per patient
for(i in unique(manifest$Case.ID)){
  
  patient_mut_txt_files = mut_txt_files[grepl(i, mut_txt_files)]
  patient_muts = c()
  for(j in patient_mut_txt_files){
    sample_muts = read.table(j, header = T, sep = '\t', stringsAsFactors = F)
    sample_muts = sample_muts[,3:6]
    names(sample_muts) = c("chr","pos","ref","mut")
    sample_muts$sample = i
    sample_muts = sample_muts[, c(5, 1:4)]
    row.names(sample_muts) = NULL
    patient_muts = rbind(patient_muts, sample_muts)
  }
  
  patient_muts = patient_muts[!duplicated(patient_muts),]
  All_var = rbind(All_var, patient_muts)
  
}

write.table(All_var, 'uniq_subs_gct_20200916.txt', col.names = T, row.names = F, sep = '\t', quote = F)

samples=unique(All_var$sample)
mut_count = matrix(0,nrow=length(samples),ncol=96)

#create dataframe of context counts per patient so as not to duplicate the counting of individual subs
for (k in 1:length(samples)){
  mut_count[k,]=count_muts_96(All_var[All_var$sample==samples[k],2:5])
  print(k)
}

mut_count = data.frame(mut_count)
row.names(mut_count) = samples

sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")

names(mut_count) = full_vec

samples_df = data.frame(Sample = samples, Patient = samples)

write.table(mut_count, "trinuc_mut_mat.txt", sep = ' ', quote = F, col.names = T, row.names = T)
write.table(samples_df, "key_table.txt", sep = ' ', quote = F, col.names = T, row.names = F)

```

```{bash}

#in directory with relevant files

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/raw_input

cp trinuc_mut_mat.txt ../wkdir
cp key_table.txt ../wkdir

cd ../wkdir

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_hdp"

for n in $(seq 1 20);
do
bsub -o $PWD/log.%J -e $PWD/err.%J -q normal -R'select[mem>10000] rusage[mem=10000]' -M10000 -J $n /software/R-3.6.1/bin/Rscript /nfs/users/nfs_t/to3/scripts/hdp_single_chain_tc16.R $n
done 

```

####Review diagnostic plots & combine

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/wkdir')

mutations=read.table("trinuc_mut_mat.txt", header = T)
key_table=read.table("key_table.txt", header = T)

lower_threshold = 100
sample_remove=rownames(mutations)[rowSums(mutations)<lower_threshold]
mutations=mutations[!rownames(mutations)%in%sample_remove,]
key_table=key_table[!key_table$Sample%in%sample_remove,]

options(stringsAsFactors = F)
library(hdp)

chlist <- vector("list", 20)
for (i in 1:20){
  if(file.exists(paste0("hdp_chain_",i,".Rdata"))){
    chlist[[i]] <- readRDS(paste0("hdp_chain_",i,".Rdata"))
  }
}
if(any(unlist(lapply(chlist,is.null)))) chlist=chlist[-which(unlist(lapply(chlist,is.null)))]

mut_example_multi <- hdp_multi_chain(chlist)

pdf("QC_plots_chain.pdf") 
par(mfrow=c(2,2), mar=c(4, 4, 2, 1))
p1 <- lapply(chains(mut_example_multi), plot_lik, bty="L", start=1000)
p2 <- lapply(chains(mut_example_multi), plot_numcluster, bty="L")
p3 <- lapply(chains(mut_example_multi), plot_data_assigned, bty="L")
dev.off()

mut_example_multi <- hdp_extract_components(mut_example_multi) #This step can take a while. If too long, submit R script as job
saveRDS(mut_example_multi,"HDP_multi_chain.Rdata")

#mut_example_multi <- readRDS("HDP_multi_chain.Rdata")

pdf("muts_attributed.pdf")
plot_comp_size(mut_example_multi, bty="L")
dev.off()

trinuc_context <- sapply(strsplit(colnames(mut_count), '\\.'), `[`, 4)
group_factor <- as.factor(rep(c("C>A", "C>G", "C>T", "T>A", "T>C", "T>G"),
                              each=16))
mut_colours=c("dodgerblue","black","red","grey70","olivedrab3","plum2")

for (i in 0:mut_example_multi@numcomp){
  pdf(paste0("hdp_component_",i,".pdf"),width=12,height=4)
  
  plot_comp_distn(mut_example_multi, cat_names=trinuc_context,
                  grouping=group_factor, col=mut_colours,comp=i,
                  col_nonsig="grey80", show_group_labels=TRUE)
  dev.off()
}

plot_dp_comp_exposure(mut_example_multi,
                      dpindices=2:4, incl_numdata_plot=FALSE,
                      col=c(RColorBrewer::brewer.pal(12, "Paired"),"magenta","firebrick"),
                      incl_nonsig=TRUE, cex.names=0.8,
                      ylab_exp = 'Signature exposure', leg.title = 'Signature')

pdf("signature_attribution.pdf",width=10,height=8)

plot_dp_comp_exposure(mut_example_multi, dpindices=(length(unique(key_table$Patient))+2):length(mut_example_multi@comp_dp_counts), incl_nonsig = T, ylab_exp = 'Signature exposure', leg.title = 'Signature',
                      col=c(RColorBrewer::brewer.pal(12, "Set3"),"magenta","firebrick"))
dev.off()

mean_assignment=as.data.frame(comp_dp_distn(mut_example_multi)$mean)
write.table(mean_assignment,"mean_assignment_hdp.txt")
mean_sigs=as.data.frame(t(comp_categ_distn(mut_example_multi)$mean))
write.table(mean_sigs,"hdp_sigs.txt")

```

####Deconvolve signatures

Component 0 is the known noise element.

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/wkdir')

options(stringsAsFactors = F)
library(hdp)
library(RColorBrewer)
library(lsa)
library(lattice)

mut.cols = rep(c("dodgerblue","black","red","grey70","olivedrab3","plum2"),each=16)

#Load HDP signatures
hdp_sigs=read.table("hdp_sigs.txt")

#Load reference signatures
#ref = read.csv("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/PCAWG_sigProfiler_SBS_signatures.csv", header=T, stringsAsFactors = F, row.names = 1)
ref = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/COSMIC_v3.2_SBS_GRCh37.txt', header = T, sep = '\t', stringsAsFactors = F, row.names = 1)

#Assess cosine similarities for all reference signatures
cosine_matrix=data.frame(matrix(nrow=ncol(hdp_sigs), ncol=ncol(ref)))
rownames(cosine_matrix)=colnames(hdp_sigs)
colnames(cosine_matrix)=colnames(ref)

for (n in 1:nrow(cosine_matrix)) {
  for (m in 1:ncol(cosine_matrix)) {
    cosine_matrix[n,m] <- cosine(x=hdp_sigs[,rownames(cosine_matrix)[n]],
                                 y=ref[,colnames(cosine_matrix)[m]])
  }
}

write.table(cosine_matrix, "Cosine_similarities.txt",sep="\t",quote=F)

#plot output
pdf("cosine_similarities.pdf", height=5, width=15)
color.palette = colorRampPalette(c("white", "orange", "purple"))
levelplot(t(cosine_matrix[dim(cosine_matrix)[1]:1,]),col.regions=color.palette, aspect="fill", scales=list(x=list(rot=90)))
dev.off()

#select which signatures to decompose into reference signatures
sigs_to_decompose=rowSums(cosine_matrix>0.85)

#X0 is just noise
#X4 closely matches 17b only (>0.85), no further extraction required
#X7 closely matches 91 only (>0.85), no further extraction required

excluded_sigs = c('X0', 'X4', 'X7') #either noise or already a close match
sigs_close_cosine = c()
for(n in 1:nrow(cosine_matrix)){
  if(!row.names(cosine_matrix)[n] %in% excluded_sigs){
    sigs_close_cosine = c(sigs_close_cosine, colnames(cosine_matrix)[cosine_matrix[n,]>0.7])
    print(paste0(rownames(cosine_matrix)[n],": ",paste(colnames(cosine_matrix)[cosine_matrix[n,]>0.7],collapse=",")))
  }
}

#"X1: SBS3,SBS5,SBS40,SBS89,SBS92"
#"X2: SBS4,SBS10c,SBS18,SBS29,SBS36,SBS56"
#"X3: "
#"X5: SBS1,SBS6,SBS87"
#"X6: SBS31,SBS35"

#X3 novel

#First iteration; decomposed hdp sigs into all suspected sigs 
gdsigs = unique(sigs_close_cosine)

signatures=t(ref[,gdsigs])
sample_list=paste0("X",c(1, 2, 5, 6)) #remove noisy, novel and already matched signatures
profiles=hdp_sigs[,sample_list]

signature_fraction = matrix(NA,nrow=nrow(signatures),ncol=length(sample_list))
rownames(signature_fraction) = rownames(signatures)
colnames(signature_fraction) = sample_list
maxiter <- 1000

for (j in 1:length(sample_list)) {
  freqs = profiles[,j]
  freqs[is.na(freqs)] = 0
  # EM algo with to estimate the signature contribution
  alpha = runif(nrow(signatures)); alpha=alpha/sum(alpha) # Random start (seems to give ~identical results)
  for (iter in 1:maxiter) {
    contr = t(array(alpha,dim=c(nrow(signatures),96))) * t(signatures)
    probs = contr/array(rowSums(contr),dim=dim(contr))
    probs = probs * freqs
    old_alpha = alpha
    alpha = colSums(probs)/sum(probs)
    if (sum(abs(alpha-old_alpha))<1e-5) {
      break
    }
  }
  # Saving the signature contributions for the sample
  print(j/length(sample_list))
  signature_fraction[,j] = alpha
}

#Plot initial deconvolution and save results
pdf("Deconvolution_hdp_sigs_R1.pdf", height=5, width=10)
#dev.new(height=5, width=10)
color.palette = colorRampPalette(c("white", "orange", "purple"))
levelplot((signature_fraction[nrow(signature_fraction):1,]),col.regions=color.palette, aspect="fill", scales=list(x=list(rot=90)))
dev.off()

write.table(signature_fraction, "hdp_known_sigs_broken_down_into_cosmic_gd_sigs.txt", sep="\t", col.names=T, row.names = T, quote=F)

sigs_deconv_R2=list()

for(n in 1:length(sample_list)){
  sigs_deconv_R2[[n]]=rownames(signature_fraction)[signature_fraction[,n] > 0.2] #SBS87 shares SBS1's C>T peaks resulting in some difficulty distinguishing it cleanly from SBS1, threshold for calling a known signature in the hdp extraction raised to 0.2 in order to prevent overfitting SBS87 in samples with no prior thiopurine exposure
}

names(sigs_deconv_R2)=colnames(signature_fraction)

#X2 is predominantly SBS18

sigs_to_deconv=names(sigs_deconv_R2)[unlist(lapply(sigs_deconv_R2,length))>1]

ref_sigs_R2=sort(unique(unlist(sigs_deconv_R2)))
signature_fractionR2=matrix(NA,ncol=length(sigs_to_deconv),nrow=length(ref_sigs_R2))
rownames(signature_fractionR2)=ref_sigs_R2
colnames(signature_fractionR2)=sigs_to_deconv

#repeat the deconvolution with the identified constitutive signatures
n = 1
for(s in sigs_to_deconv){
  gdsigs <- sigs_deconv_R2[[s]]
  signatures <- t(ref[,gdsigs])
  
  signature_fraction = matrix(NA,nrow=nrow(signatures),ncol=length(sample_list))
  rownames(signature_fraction) = rownames(signatures)
  colnames(signature_fraction) = sample_list
  maxiter <- 1000
  
  freqs = profiles[,s]
  freqs[is.na(freqs)] = 0
  
  alpha = runif(nrow(signatures)); alpha=alpha/sum(alpha) # Random start (seems to give ~identical results)
  for (iter in 1:maxiter) {
    contr = t(array(alpha,dim=c(nrow(signatures),96))) * t(signatures)
    probs = contr/array(rowSums(contr),dim=dim(contr))
    probs = probs * freqs
    old_alpha = alpha
    alpha = colSums(probs)/sum(probs)
    if (sum(abs(alpha-old_alpha))<1e-5) {
      break
    }
  }
  # Saving the signature contributions for the sample
  signature_fractionR2[gdsigs,n] = alpha
  n=n+1
  reconsbs <- rep(0,96)
  for (g in gdsigs) {
    reconsbs=reconsbs+(ref[,g]*alpha[g])
  }
  cosine_reconst=cosine(x=reconsbs, y=hdp_sigs[,s])
  print(paste0(s,": ",cosine_reconst))
  pdf(paste0("HDP_",s,"_reconstitution2.pdf"))
  par(mfrow=c(length(alpha)+2,1))
  par(mar=c(1,2,4,1))
  barplot(hdp_sigs[,s], col=mut.cols, main=paste0("HDP ",s),names.arg="")
  barplot(reconsbs, col=mut.cols, main=paste0("Reconstituted ",s," cosine similarity to original: ", round(cosine_reconst, digits=2)))
  for (g in gdsigs) {
    barplot(ref[,g], col=mut.cols, main=paste0("PCAWG ", g, " accounts for ", round(alpha[g], digits=2)))
  }
  dev.off()
}

#add previous signatures back in
sigs_deconv_R2$X0 = "noise"
sigs_deconv_R2$X3 = "N1"
sigs_deconv_R2$X4 = "SBS17b"
sigs_deconv_R2$X7 = "SBS91"

saveRDS(sigs_deconv_R2,"hdp2refsigs.Rdata")

ref_sigs_R2 = c(ref_sigs_R2, 'SBS17b', 'SBS91')

#Combine the hdp signature that did not get deconvolved and reference signatures into final table
final_sigs = cbind(hdp_sigs[,c("X3")],ref[,ref_sigs_R2])
#Rename the HDP components that didn't get deconvoluted
colnames(final_sigs)[1] = "N1"

#Order them - optional
final_sigs = final_sigs[, c('N1', 'SBS1', 'SBS5', 'SBS17b', 'SBS18', 'SBS31', 'SBS35', 'SBS40', 'SBS91')]

write.table(final_sigs, "final_sigs_2021_09_16.txt", sep = '\t', col.names = T, row.names = T, quote = F)

```

#### Clean up novel signature

N1 still evidently has a minor component of SBS1. This is likely because the key peaks seem to coincide with samples containing slightly higher burdens of SBS1. To address this, we will run HDP again but with priors this time and the finite reportoire of known signatures we identified first time. We can also check if there was a smaller component of any signatures we missed the first time round.

#####Parts A, B & C
PART A : Arrange the prior sigs
PART B : Set up HDP
PART C : Run posterior sampling chains

```{r}

#/nfs/users/nfs_t/to3/scripts/hdp_prior_single_chain_updated_20210916.R

options(stringsAsFactors = F)
library(hdp)
lower_threshold=100

n=as.numeric(commandArgs(T)[1])
mutations=read.table("trinuc_mut_mat.txt", header = T)
key_table=read.table("key_table.txt", header = T)

ref = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/wkdir/final_sigs_2021_09_16.txt', header = T, stringsAsFactors = F, sep = '\t')
ref = ref[, colnames(ref) != 'N1']
prior_sigs = as.matrix(ref)

# number of prior signatures to condition on (8)
nps <- ncol(prior_sigs)

# take a look
#prior_sigs[1:6,1:6]
# columns should sum to one
#colSums(prior_sigs)

# have a look at the first two prior sigs:
#barplot(prior_sigs[,1], las=2, cex.names=0.5)
#barplot(prior_sigs[,'SBS18'], las=2, cex.names=0.5)
#barplot(prior_sigs[,'SBS91'], las=2, cex.names=0.5)

#If requiring a minimum number of mutations:
sample_remove=rownames(mutations)[rowSums(mutations)<lower_threshold]
mutations=mutations[!rownames(mutations)%in%sample_remove,]
key_table=key_table[!key_table$Sample%in%sample_remove,]

#Hierarchy is set per patient, can change if wanted
freq = table(key_table$Patient)

#nps <- ncol(prior_sigs)

hdp_prior <- hdp_prior_init(prior_distn = prior_sigs, # matrix of prior sigs
                             prior_pseudoc = rep(1000, nps), # pseudocount weights
                             hh=rep(1, 96), # uniform prior over 96 categories
                             alphaa=c(1, 1), # shape hyperparams for 2 CPs
                             alphab=c(1, 1)) # rate hyperparams for 2 CPs
#hdp_prior
#numdp(hdp_prior)
#pseudoDP(hdp_prior)
#conparam(hdp_prior)
#ppindex(hdp_prior)
#cpindex(hdp_prior)
#dpstate(hdp_prior) # 2 for active node, 1 for frozen, 0 for heldout

# make two more CPs available for the data we will add
hdp_prior <- hdp_addconparam(hdp_prior,
                              alphaa = rep(1,length(freq)+2), # shape hyperparams for x new CPs
                              alphab = rep(1,length(freq)+2)) # rate hyperparams for x new CPs

hdp_prior <- hdp_adddp(hdp_prior,
                        numdp = nrow(mutations) + 1,
                        ppindex = c(1, rep(1+nps+1:(length(freq)), times=freq)),
                        cpindex = c(3, rep(4:(length(freq)+3), times=freq)))

# assign the data to the relevant DP nodes
hdp_prior <- hdp_setdata(hdp_prior,
                          dpindex = (1 + nps + 1) + 1:nrow(mutations), 
                          mutations) # mutation counts in all GCTs


hdp_activated <- dp_activate(hdp_prior, 
                             dpindex = (1+nps+1)+0:nrow(mutations), 
                             initcc = nps+5,
                             seed = n * 1000)

chain=hdp_posterior(hdp_activated,
                    burnin=80000,
                    n=100,
                    seed=n*1000,
                    space=200,
                    cpiter=3)

saveRDS(chain,paste0("hdp_prior_chain_",n,".Rdata"))

```

```{bash}

#in directory with relevant files

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/raw_input

cp trinuc_mut_mat.txt ../novel_sig_cleanup
cp key_table.txt ../novel_sig_cleanup

cd ../novel_sig_cleanup

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_hdp"

for n in $(seq 1 20);
do
bsub -o $PWD/log.%J -e $PWD/err.%J -q normal -R'select[mem>10000] rusage[mem=10000]' -M10000 -J $n /software/R-3.6.1/bin/Rscript /nfs/users/nfs_t/to3/scripts/hdp_prior_single_chain_updated_20210916.R $n
done

```

#####Parts D & E

PART D : Review diagnostic plots
PART E : Extract consensus components / signatures

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/novel_sig_cleanup')

options(stringsAsFactors = F)
library(hdp)
library(ggpubr)

chlist <- vector("list", 20)
for (i in 1:20){
  if(file.exists(paste0("hdp_prior_chain_",i,".Rdata"))){
    chlist[[i]] <- readRDS(paste0("hdp_prior_chain_",i,".Rdata"))
  }
}
if(any(unlist(lapply(chlist,is.null)))) chlist=chlist[-which(unlist(lapply(chlist,is.null)))]

hdp_multi <- hdp_multi_chain(chlist)

pdf("QC_plots_chain.pdf") 
par(mfrow=c(2,2), mar=c(4, 4, 2, 1))
p1 <- lapply(chains(hdp_multi), plot_lik, bty="L", start=1000)
p2 <- lapply(chains(hdp_multi), plot_numcluster, bty="L")
p3 <- lapply(chains(hdp_multi), plot_data_assigned, bty="L")
dev.off()

#QC looks good 16/9/21

hdp_multi <- hdp_extract_components(hdp_multi)
saveRDS(hdp_multi,"HDP_multi_chain.Rdata")

numcomp(hdp_multi) #10
prop.ex(hdp_multi) #0.963

pdf("muts_attributed.pdf")
plot_comp_size(hdp_multi, bty="L")
dev.off()

# plot components / signatures
# pick your colours
mut_colours=c("dodgerblue","black","red","grey70","olivedrab3","plum2")
# labels along bottom x-axis
trinuc_context <- sapply(strsplit(colnames(mut_count), '\\.'), `[`, 4)
# group labels along the top (and controls colour grouping)
group_factor <- as.factor(rep(c("C>A", "C>G", "C>T", "T>A", "T>C", "T>G"),
                              each=16))

ref = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/wkdir/final_sigs_2021_09_16.txt', header = T, stringsAsFactors = F, sep = '\t')
ref = ref[, colnames(ref) != 'N1']
prior_sigs = as.matrix(ref)

# number of prior signatures to condition on (8)
nps <- ncol(prior_sigs)

pdf(paste0("hdp_components_with_priors.pdf"),width=12,height=4)
plot_comp_distn(hdp_multi, cat_names=trinuc_context,
                grouping=group_factor, col=mut_colours,
                col_nonsig="grey97", show_group_labels=TRUE)
dev.off()

lower_threshold = 100

mutations=read.table("trinuc_mut_mat.txt", header = T)
sample_remove=rownames(mutations)[rowSums(mutations)<lower_threshold]
mutations=mutations[!rownames(mutations)%in%sample_remove,]


pdf("signature_attribution.pdf",width=10,height=8)
plot_dp_comp_exposure(hdp_multi, dpindices = (1+nps+1)+1:nrow(mutations), incl_nonsig = F,
                      col=c('black', get_palette(palette = "Set1", 15)), cex.names=0.8,
                      ylab_exp = 'Signature exposure', leg.title = 'Signature')
dev.off()

mean_assignment=as.data.frame(comp_dp_distn(hdp_multi)$mean)
write.table(mean_assignment,"mean_assignment_prior_hdp.txt")
mean_sigs=as.data.frame(t(comp_categ_distn(hdp_multi)$mean))
write.table(mean_sigs,"hdp_prior_sigs.txt")

colnames(prior_sigs)[c(1, 2, 3, 4, 7, 8)]
#"SBS1"   "SBS5"   "SBS17b" "SBS18"  "SBS40"  "SBS91" 

colnames(prior_sigs)[c(7)] #SBS40 looks unstable on repeat extraction, leaving "SBS1"   "SBS5"   "SBS17b" "SBS18"  "SBS91"
```

#####Deconvolute 'novel' signatures

```{r}

#Deconvolute HDP signatures into reference signatures and arrive at final set

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/novel_sig_cleanup')

options(stringsAsFactors = F)
library(hdp)
library(RColorBrewer)
library(lsa)
library(lattice)

mut.cols = rep(c("dodgerblue","black","red","grey70","olivedrab3","plum2"),each=16)

#Load HDP signatures
mean_sigs = read.table("hdp_prior_sigs.txt", row.names = 1, header = T)
hdp_sigs = mean_sigs
#Load reference signatures, need formatting
#ref = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/MutationalPatterns/reformatted_cosmic_sig_v3.txt', header = T, stringsAsFactors = F, row.names = 1, check.names = F)
#ref = t(ref)
#cosmic.sigs = read.csv('/lustre/scratch119/casm/team294rr/to3/testes/tumour/MutationalPatterns/PCAWG_sigProfiler_SBS_signatures.csv', header = T, stringsAsFactors = F, row.names = 1, check.names = F)
#row.names(cosmic.sigs) = row.names(ref)
#prior_sigs = as.matrix(cosmic.sigs)
ref = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/COSMIC_v3.2_SBS_GRCh37.txt', header = T, sep = '\t', stringsAsFactors = F, row.names = 1)

#Assess cosine similarities for all unassigned signatures (X0, N1, N2)
cosine_matrix=data.frame(matrix(nrow=ncol(hdp_sigs[, c('N1', 'N2', 'N3', 'N4')]), ncol=ncol(ref)))
rownames(cosine_matrix)=colnames(hdp_sigs[, c('N1', 'N2', 'N3', 'N4')])
colnames(cosine_matrix)=colnames(ref)

for (n in 1:nrow(cosine_matrix)) {
  for (m in 1:ncol(cosine_matrix)) {
    cosine_matrix[n,m] <- cosine(x=hdp_sigs[,rownames(cosine_matrix)[n]],
                                 y=ref[,colnames(cosine_matrix)[m]])
  }
}

write.table(cosine_matrix, "Cosine_similarities_unassigned_to_priors.txt",sep="\t",quote=F)

hdp_sigs = hdp_sigs[, c('N1', 'N2', 'N3', 'N4')]

#plot output
pdf("cosine_similarities_unassigned_to_priors.pdf", height=5, width=15)
color.palette = colorRampPalette(c("white", "orange", "purple"))
levelplot(t(cosine_matrix[dim(cosine_matrix)[1]:1,]),col.regions=color.palette, aspect="fill", scales=list(x=list(rot=90)))
dev.off()

#select which signatures to decompose into reference signatures
sigs_to_decompose=rowSums(cosine_matrix>0.85)
#N3 is SBS35
#N4 is SBS17a

excluded_sigs = c('N3', 'N4') #already a close match
sigs_close_cosine = c()
for(n in 1:nrow(cosine_matrix)){
  if(!row.names(cosine_matrix)[n] %in% excluded_sigs){
    sigs_close_cosine = c(sigs_close_cosine, colnames(cosine_matrix)[cosine_matrix[n,]>0.7])
    print(paste0(rownames(cosine_matrix)[n],": ",paste(colnames(cosine_matrix)[cosine_matrix[n,]>0.7],collapse=",")))
  }
}

#[1] "N1: SBS3,SBS5,SBS40"
#"N2: "

#First iteration; decomposed hdp sigs into all suspected sigs 
gdsigs = sigs_close_cosine

signatures=t(ref[,gdsigs])
sample_list= c('N1', 'N2') #only keep N2 in to prevent errors from vector <-> dataframe conversions later
profiles=hdp_sigs[,sample_list]

signature_fraction = matrix(NA,nrow=nrow(signatures),ncol=length(sample_list))
rownames(signature_fraction) = rownames(signatures)
colnames(signature_fraction) = sample_list
maxiter <- 1000

for (j in 1:length(sample_list)) {
  freqs = profiles[,j]
  freqs[is.na(freqs)] = 0
  # EM algorithm with to estimate the signature contribution
  alpha = runif(nrow(signatures)); alpha=alpha/sum(alpha) # Random start (seems to give ~identical results)
  for (iter in 1:maxiter) {
    contr = t(array(alpha,dim=c(nrow(signatures),96))) * t(signatures)
    probs = contr/array(rowSums(contr),dim=dim(contr))
    probs = probs * freqs
    old_alpha = alpha
    alpha = colSums(probs)/sum(probs)
    if (sum(abs(alpha-old_alpha))<1e-5) {
      break
    }
  }
  # Saving the signature contributions for the sample
  print(j/length(sample_list))
  signature_fraction[,j] = alpha
}

#Plot initial deconvolution and save results
pdf("Deconvolution_hdp_sigs_after_prior_R1.pdf", height=5, width=10)
color.palette = colorRampPalette(c("white", "orange", "purple"))
levelplot((signature_fraction[nrow(signature_fraction):1,]),col.regions=color.palette, aspect="fill", scales=list(x=list(rot=90)))
dev.off()

write.table(signature_fraction[,'N1'], "hdp_unassigned_sig_broken_down_into_cosmic_sigs.txt", sep="\t", col.names=T, row.names = T, quote=F)

sigs_deconv_R2=list()

for(n in 1:length(sample_list)){
  sigs_deconv_R2[[n]]=rownames(signature_fraction)[signature_fraction[,n]>0.2]
}

names(sigs_deconv_R2)=colnames(signature_fraction)

#No signature has one major contributor only

sigs_to_deconv=names(sigs_deconv_R2)[unlist(lapply(sigs_deconv_R2,length))>1]

ref_sigs_R2=sort(unique(unlist(sigs_deconv_R2)))
signature_fractionR2=matrix(NA,ncol=length(sigs_to_deconv),nrow=length(ref_sigs_R2))
rownames(signature_fractionR2)=ref_sigs_R2
colnames(signature_fractionR2)=sigs_to_deconv

#repeat the deconvolution with the identified constitutive signatures
n = 1
for(s in sigs_to_deconv){
  gdsigs <- sigs_deconv_R2[[s]]
  signatures <- t(ref[,gdsigs])
  
  signature_fraction = matrix(NA,nrow=nrow(signatures),ncol=length(sample_list))
  rownames(signature_fraction) = rownames(signatures)
  colnames(signature_fraction) = sample_list
  maxiter <- 1000
  
  freqs = profiles[,s]
  freqs[is.na(freqs)] = 0
  
  alpha = runif(nrow(signatures)); alpha=alpha/sum(alpha) # Random start (seems to give ~identical results)
  for (iter in 1:maxiter) {
    contr = t(array(alpha,dim=c(nrow(signatures),96))) * t(signatures)
    probs = contr/array(rowSums(contr),dim=dim(contr))
    probs = probs * freqs
    old_alpha = alpha
    alpha = colSums(probs)/sum(probs)
    if (sum(abs(alpha-old_alpha))<1e-5) {
      break
    }
  }
  # Saving the signature contributions for the sample
  signature_fractionR2[gdsigs,n] = alpha
  n=n+1
  reconsbs <- rep(0,96)
  for (g in gdsigs) {
    reconsbs=reconsbs+(ref[,g]*alpha[g])
  }
  cosine_reconst=cosine(x=reconsbs, y=hdp_sigs[,s])
  print(paste0(s,": ",cosine_reconst))
  pdf(paste0("HDP_",s,"_reconstitution2.pdf"))
  par(mfrow=c(length(alpha)+2,1))
  par(mar=c(1,2,4,1))
  barplot(hdp_sigs[,s], col=mut.cols, main=paste0("HDP ",s),names.arg="")
  barplot(reconsbs, col=mut.cols, main=paste0("Reconstituted ",s," cosine similarity to original: ", round(cosine_reconst, digits=2)))
  for (g in gdsigs) {
    barplot(ref[,g], col=mut.cols, main=paste0("PCAWG ", g, " accounts for ", round(alpha[g], digits=2)))
  }
  dev.off()
}

#complete sig list
sigs_deconv_R2$N2 = 'N2'
sigs_deconv_R2$N3 = 'SBS35'
sigs_deconv_R2$N4 = 'SBS17a'

saveRDS(sigs_deconv_R2,"novel_hdp2refsigs.Rdata")

ref_sigs_R2 = c(ref_sigs_R2, "SBS18", "SBS1", "SBS17a", "SBS17b", 'SBS35', "SBS91") #lost SBS31, gained SBS17a

#Combine the hdp signature that did not get deconvolved and reference signatures into final table
final_sigs = cbind(hdp_sigs[,c("N2")],ref[,ref_sigs_R2])
#Rename the HDP components that didn't get deconvoluted
colnames(final_sigs)[1] = "N2"

#Order them - optional
final_sigs = final_sigs[, c('N2', 'SBS1', 'SBS5', 'SBS17a', 'SBS17b', 'SBS18', 'SBS35', 'SBS40', 'SBS91')]

write.table(final_sigs, "final_sigs_CLEANED_UP_2021_09_16.txt", col.names = T, row.names = T, sep = '\t', quote = F)

```

#####Compare N2 to updated COSMIC signatures

```{r}

library(lsa)

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/novel_sig_cleanup")

final_sigs = read.table('final_sigs_CLEANED_UP_2021_09_16.txt', header = T, sep = '\t', stringsAsFactors = F)
cosmic_v3_2 = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/COSMIC_v3.2_SBS_GRCh37.txt', header = T, sep = '\t', stringsAsFactors = F, row.names = 1)

#cosmic_v3_2 = cosmic_v3_2[row.names(final_sigs),] #put cosmic signatures in the correct order
#write.table(cosmic_v3_2, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/COSMIC_v3.2_SBS_GRCh37.txt', col.names = T, row.names = T, sep = '\t', quote = F)

cos_res = c()
for(i in 1:ncol(cosmic_v3_2)){
  cos_res = c(cos_res, cosine(final_sigs[,'N2'], cosmic_v3_2[,i])) 
}
names(cos_res) = colnames(cosmic_v3_2)
cos_res[cos_res == max(cos_res)]

#    SBS39 
#0.4206593 very dissimilar to published signatures

cos_res[cos_res >0.4]
#    SBS39     SBS87 
#0.4206593 0.4198666

```

#####Map back signatures with Sigprofiler (per sample)

Need substitutions per sample

```{r}

library(reshape2)
library(ggplot2)
library(ggpubr)
library(cowplot)

#setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/sigprofiler_deconvolution_remapping')
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp/04_SIGPROFILER_DECONVOLUTION')

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', header = T, stringsAsFactors = F, sep = '\t')

hdp_sigprofiler_mapped = read.table('Decompose_Solution_Activities.txt', header = T, sep = '\t', stringsAsFactors = F)
#hdp_sigprofiler_mapped$Samples = unlist(strsplit(hdp_sigprofiler_mapped$Samples, '_post_swater_final_snvs')) the sample names needed editing from the raw sigprofiler output
#write.table(hdp_sigprofiler_mapped, 'Decompose_Solution_Activities.txt', col.names = T, row.names = F, sep = '\t', quote = F)

hdp_sigprofiler_mapped$SBS5_40 = hdp_sigprofiler_mapped$SBS5 + hdp_sigprofiler_mapped$SBS40 
hdp_sigprofiler_mapped = hdp_sigprofiler_mapped[, c("Samples", "SBS1", "SBS5_40", "SBS17a", "SBS17b", "SBS18", "SBS35", "SBS91", "N2")]

hdp_sigprofiler_mapped_norm = hdp_sigprofiler_mapped[, c(2:9)] / rowSums(hdp_sigprofiler_mapped[, c(2:9)])
hdp_sigprofiler_mapped_norm$sample = hdp_sigprofiler_mapped$Samples
hdp_sigprofiler_mapped_norm$tumour = substr(hdp_sigprofiler_mapped_norm$sample, 0, 7)

hdp_sigprofiler_mapped_norm.m = reshape2::melt(hdp_sigprofiler_mapped_norm, id.vars = c('sample', 'tumour'))

hdp_sigprofiler_mapped_norm.m$age = NA
hdp_sigprofiler_mapped_norm.m$diagnosis = NA
hdp_sigprofiler_mapped_norm.m$site = NA
hdp_sigprofiler_mapped_norm.m$histology = NA

for(i in 1:nrow(hdp_sigprofiler_mapped_norm.m)){
  hdp_sigprofiler_mapped_norm.m$age[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Age
  hdp_sigprofiler_mapped_norm.m$diagnosis[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Diagnosis
  hdp_sigprofiler_mapped_norm.m$site[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Organ
  hdp_sigprofiler_mapped_norm.m$histology[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Histology
}

hdp_sigprofiler_mapped_norm.m$variable = factor(hdp_sigprofiler_mapped_norm.m$variable, levels = rev(unique(hdp_sigprofiler_mapped_norm.m$variable)))
hdp_sigprofiler_mapped_norm.m$tumour = factor(hdp_sigprofiler_mapped_norm.m$tumour, levels = unique(hdp_sigprofiler_mapped_norm.m[order(hdp_sigprofiler_mapped_norm.m$age),]$tumour))

cols = get_palette(palette = 'Set1', k = length(unique(hdp_sigprofiler_mapped_norm.m$variable)))

h1 = ggplot(hdp_sigprofiler_mapped_norm.m) + 
  geom_bar(mapping = aes(x = sample, y = value, fill = variable), stat = 'identity') + 
  theme_pubr() + coord_cartesian(expand = F) + 
  scale_fill_manual(values=cols) + 
  ylab('Proportion of substitutions') + 
  xlab('') + 
  theme(axis.text.x = element_blank(), legend.position = 'right', axis.ticks.x = element_blank()) + facet_grid(. ~ tumour, scales = 'free_x', space = 'free_x')
h2 = ggplot(hdp_sigprofiler_mapped_norm.m) + 
  geom_bar(mapping = aes(x = sample, y = 1, fill = histology),  stat = "identity", position = "dodge") + 
  theme_pubr() + coord_cartesian(expand = F) + 
  scale_fill_manual(values = get_palette('Dark2', length(unique(hdp_sigprofiler_mapped_norm.m$histology)))) + 
  ylab('') + 
  xlab('') + 
  theme(axis.text.x = element_blank(), legend.position = 'right', axis.ticks.x = element_blank(), strip.text = element_blank(), axis.ticks = element_blank(), axis.text = element_blank()) + facet_grid(. ~ tumour, scales = 'free_x', space = 'free_x')
h3 = ggplot(hdp_sigprofiler_mapped_norm.m) + 
  geom_bar(mapping = aes(x = sample, y = 1, fill = diagnosis),  stat = "identity", position = "dodge") + 
  theme_pubr() + coord_cartesian(expand = F) + 
  scale_fill_manual(values = get_palette('Dark2', length(unique(hdp_sigprofiler_mapped_norm.m$diagnosis)))) + 
  ylab('') + 
  xlab('') + 
  theme(axis.text.x = element_blank(), legend.position = 'right', axis.ticks.x = element_blank(), strip.text = element_blank(), axis.ticks = element_blank(), axis.text = element_blank()) + facet_grid(. ~ tumour, scales = 'free_x', space = 'free_x')
h4 = ggplot(hdp_sigprofiler_mapped_norm.m) + 
  geom_bar(mapping = aes(x = sample, y = 1, fill = site),  stat = "identity", position = "dodge") + 
  theme_pubr() + coord_cartesian(expand = F) + 
  scale_fill_manual(values = get_palette('default', length(unique(hdp_sigprofiler_mapped_norm.m$site)))) + 
  ylab('') + 
  xlab('') + 
  theme(axis.text.x = element_blank(), legend.position = 'right', axis.ticks.x = element_blank(), strip.text = element_blank(), axis.ticks = element_blank(), axis.text = element_blank()) + facet_grid(. ~ tumour, scales = 'free_x', space = 'free_x')

plot_grid(h1, h2, h3, h4,  align = 'v', ncol = 1, rel_heights = c(1, 0.2, 0.2, 0.2))

```

###Plot the trinucleotide context per tumour (Extended data figure 4)

Plot trinucleotide context per tumour

```{r}

library(GenomicRanges)
library(Rsamtools)
library(MASS)

#genomeFile = "/nfs/cancer_ref01/Homo_sapiens/37/genome.fa"

plot_trinuc_mat <- function(SNVsample, input_df, out_dir){
  # 1. Annotating the trinucleotide context
  genomeFile <- "/nfs/cancer_ref01/Homo_sapiens/37/genome.fa"
  mutations <- input_df[,c(2:5)]
  mutations <- as.data.frame(mutations)
  colnames(mutations) <- c("chr","pos","ref","mut")
  mutations$pos <- as.numeric(mutations$pos)
  mutations <- mutations[(mutations$ref %in% c("A","C","G","T")) & (mutations$mut %in% c("A","C","G","T")) & mutations$chr %in% c(1:22,"X","Y"),]
  mutations$trinuc_ref <- as.vector(scanFa(genomeFile, GRanges(mutations$chr, IRanges(as.numeric(mutations$pos)-1, 
                                                                                      as.numeric(mutations$pos)+1))))
  
  # 2. Annotating the mutation from the pyrimidine base
  ntcomp <- c(T="A",G="C",C="G",A="T")
  mutations$sub <- paste(mutations$ref, mutations$mut,sep=">")
  mutations$trinuc_ref_py <- mutations$trinuc_ref
  for (j in 1:nrow(mutations)) {
    if (mutations$ref[j] %in% c("A","G")) { # Purine base
      mutations$sub[j] = paste(ntcomp[mutations$ref[j]],ntcomp[mutations$mut[j]],sep=">")
      mutations$trinuc_ref_py[j] = paste(ntcomp[rev(strsplit(mutations$trinuc_ref[j],split="")[[1]])],collapse="")
    }
  }
  
  # 3. Counting subs
  freqs = table(paste(mutations$sub,paste(substr(mutations$trinuc_ref_py,1,1),substr(mutations$trinuc_ref_py,3,3),sep="-"),sep=","))
  sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
  ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
  full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")
  freqs_full = freqs[full_vec]; freqs_full[is.na(freqs_full)] = 0; names(freqs_full) = full_vec
  
  xstr = paste(substr(full_vec,5,5), substr(full_vec,1,1), substr(full_vec,7,7), sep="")
  
  pdf(file = paste0(out_dir, "/", SNVsample, "_trinucleotide.pdf"), width=12,height=4)
  colvec = rep(c("dodgerblue","black","red","grey70","olivedrab3","plum2"),each=16)
  y = freqs_full; maxy = max(y)
  h = barplot(y, las=2, col=colvec, border=NA, ylim=c(0,maxy*1.5), space=1, cex.names=0.6, names.arg=xstr, ylab="Number of mutations")
  for (j in 1:length(sub_vec)) {
    xpos = h[c((j-1)*16+1,j*16)]
    rect(xpos[1]-0.5, maxy*1.2, xpos[2]+0.5, maxy*1.3, border=NA, col=colvec[j*16])
    text(x=mean(xpos), y=maxy*1.3, pos=3, label=sub_vec[j])
  }    
  dev.off()
}

mutations = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/hdp_new_swater/raw_input/uniq_subs_gct_20200916.txt', header = T, sep = '\t', stringsAsFactors = F)
table(mutations$sample)

for(i in unique(mutations$sample)){
  plot_trinuc_mat(i, mutations[mutations$sample == i,], '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/per_tum_trinuc_plots')
}

```

###Plot novel signature (Fig. 1c)

```{r}

#without strand bias
library(dplyr)
library(ggplot2)
library(ggpubr)

final_sigs = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp/03_NOVEL_SIG_CLEANUP/final_sigs_CLEANED_UP_2021_09_16.txt', header = T, sep = '\t', stringsAsFactors = F)
final_sigs$trinuc = row.names(final_sigs)
final_sigs$trinuc_context = paste0(substr(final_sigs$trinuc, 1, 1), substr(final_sigs$trinuc, 3, 3), substr(final_sigs$trinuc, 7, 7))
final_sigs$trinuc = factor(final_sigs$trinuc, levels = final_sigs$trinuc)
final_sigs$trinuc_context = factor(final_sigs$trinuc_context, levels = unique(final_sigs$trinuc_context))
final_sigs$sub = substr(final_sigs$trinuc, 3, 5)
final_sigs$sub = factor(final_sigs$sub, levels = unique(final_sigs$sub))
  
pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/sbs_n2_plot_20210919.pdf', height = 2.5,  width = 7.5, useDingbats = F)
ggplot(final_sigs) + 
  geom_bar(mapping = aes(x = trinuc_context, y = N2, fill = sub), position="dodge", stat = 'identity', col = NA, size = 0.3) + 
  facet_grid(. ~ sub, scales = 'free_x') +
  scale_fill_manual(values = c("dodgerblue","black","red","grey70","olivedrab3","plum2")) +
  ylab("Proportion of substitutions assigned") + 
  xlab("Trinucleotide context") +
  guides(fill = FALSE) + 
  theme_pubr() +
  theme(strip.background = element_blank(), strip.text = element_text(face = 'bold'), axis.text.x = element_text(size = 5, angle = 90, vjust = 0.5), panel.spacing = unit(1, 'mm'), panel.border = element_rect(size = 0.3, fill = NA), legend.position = 'right') +
  coord_cartesian(ylim = c(0, 0.15), expand = F)
dev.off()

```

###Examine whether SBSN2 is unique amongst SBS signatures for its dominance by a single peak.

```{r}

#novel signature

final_sigs = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp/03_NOVEL_SIG_CLEANUP/final_sigs_CLEANED_UP_2021_09_16.txt', header = T, sep = '\t', stringsAsFactors = F)
ref = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/99_REFERENCE_DATA/COSMIC_v3.2_SBS_GRCh37.txt', header = T, sep = '\t', stringsAsFactors = F, row.names = 1)

row.names(final_sigs) == row.names(ref) #in same order

all_sigs = cbind(final_sigs[row.names(ref), "N2"], ref)
colnames(all_sigs)[1] = "SBSN2"

max(all_sigs$SBSN2) #0.1471272 highest single peak

library(ggplot2)

max_sig = data.frame(apply(all_sigs, 2, max))
max_sig$SBS = row.names(max_sig)
max_sig = max_sig[, c(2,1)]
colnames(max_sig) = c("SBS", "max_prop")
max_sig = max_sig[order(max_sig$max_prop),]
max_sig$SBS = factor(max_sig$SBS, levels = max_sig$SBS)
max_sig$highlight = ifelse(max_sig$SBS == "SBSN2", "Y", "N")

ggplot(max_sig) + geom_bar(mapping = aes(x = SBS, y = max_prop, fill = highlight), stat = "identity") + theme_linedraw() + coord_cartesian(expand = T, ylim = c(0, 0.75)) + theme(axis.text.x = element_text(angle = 90, size = 8, vjust = 0.5, hjust = 1)) + scale_y_continuous(expand = c(0.001,0.001)) + scale_fill_manual( values = c( "Y"="tomato", "N"="darkgray"), guide = FALSE ) + ylab("Maximum proportion from a single trinucleotide context") + xlab("SBS signature")

```

###HDP inc. Addies samples

####LiftOver Addies samples to 37 build

```{bash}

#Convert to bed files for liftover

grep -v "#" LP5000388-DNA_B01_pass.vcf | cut -f1,2,4,5 | awk '{print $1, $2 - 1, $2, $3, $4}' > LP5000388-DNA_B01_pass.bed
grep -v "#" LP5000405-DNA_D01_pass.vcf | cut -f1,2,4,5 | awk '{print $1, $2 - 1, $2, $3, $4}' > LP5000405-DNA_D01_pass.bed

#liftOver suing CrossMap

CrossMap.py bed /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/99_REFERENCE_DATA/grch38_to_grch37.over.chain.gz LP5000388-DNA_B01_pass.bed LP5000388-DNA_B01_pass.grch37.bed
CrossMap.py bed /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/99_REFERENCE_DATA/grch38_to_grch37.over.chain.gz LP5000405-DNA_D01_pass.bed LP5000405-DNA_D01_pass.grch37.bed

#only keep subs
awk -F'\t' '{if(length($4) == 1 && length($5) == 1) print }'  LP5000388-DNA_B01_pass.grch37.bed > LP5000388-DNA_B01_pass.grch37.subs.bed
awk -F'\t' '{if(length($4) == 1 && length($5) == 1) print }'  LP5000405-DNA_D01_pass.grch37.bed > LP5000405-DNA_D01_pass.grch37.subs.bed

```

####Generate input files

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp_inc_addies_samples/01_INPUT

cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp/01_INPUT/uniq_subs_gct_20200916.txt .
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp_inc_addies_samples/00_liftOver/*.grch37.subs.bed .

```

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp_inc_addies_samples/01_INPUT")

sanger_subs = read.table("uniq_subs_gct_20200916.txt", header = T, sep = '\t', stringsAsFactors = F)
addies1 = read.table("LP5000388-DNA_B01_pass.grch37.subs.bed", header = F, sep = '\t', stringsAsFactors = F)
addies2 = read.table("LP5000405-DNA_D01_pass.grch37.subs.bed", header = F, sep = '\t', stringsAsFactors = F)

addies1$V1 = unlist(strsplit(addies1$V1, "chr"))[c(F,T)]
addies2$V1 = unlist(strsplit(addies2$V1, "chr"))[c(F,T)]

addies1$sample = "LP5000388"
addies2$sample = "LP5000405"

addies1 = addies1[, c(6, 1, 3, 4, 5)]
addies2 = addies2[, c(6, 1, 3, 4, 5)]

names(addies1) = names(sanger_subs)
names(addies2) = names(sanger_subs)

all_subs = rbind(sanger_subs, addies1, addies2)

all_subs = all_subs[all_subs$chr %in% c(1:22, "X", "Y"),]

write.table(all_subs, "uniq_subs_gct_inc_addies_20220102.txt", col.names = T, row.names = T, sep = '\t', quote = F)

#Function to generate the counts per trinucleotide context
count_muts_96 = function(mutations, genomeFile = "/nfs/cancer_ref01/Homo_sapiens/37/genome.fa"){
  library("GenomicRanges")
  library("Rsamtools")
  library("MASS")
  colnames(mutations) = c("chr","pos","ref","mut")
  mutations = mutations[(mutations$ref %in% c("A","C","G","T")) & (mutations$mut %in% c("A","C","G","T")) & mutations$chr %in% c(1:22,"X","Y"),]
  mutations$trinuc_ref = as.vector(scanFa(genomeFile, GRanges(mutations$chr, IRanges(as.numeric(mutations$pos)-1,
                                                                                     as.numeric(mutations$pos)+1))))
 
  # 2. Annotating the mutation from the pyrimidine base
  ntcomp = c(T="A",G="C",C="G",A="T")
  mutations$sub = paste(mutations$ref,mutations$mut,sep=">")
  mutations$trinuc_ref_py = mutations$trinuc_ref
  for (j in 1:nrow(mutations)) {
    if (mutations$ref[j] %in% c("A","G")) { # Purine base
      mutations$sub[j] = paste(ntcomp[mutations$ref[j]],ntcomp[mutations$mut[j]],sep=">")
      mutations$trinuc_ref_py[j] = paste(ntcomp[rev(strsplit(mutations$trinuc_ref[j],split="")[[1]])],collapse="")
    }
  }
  freqs = table(paste(mutations$sub,paste(substr(mutations$trinuc_ref_py,1,1),substr(mutations$trinuc_ref_py,3,3),sep="-"),sep=","))
  sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
  ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
  full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")
  freqs_full = freqs[full_vec]; freqs_full[is.na(freqs_full)] = 0; names(freqs_full) = full_vec
  return(freqs_full)
}

samples=unique(all_subs$sample)
mut_count = matrix(0,nrow=length(samples),ncol=96)

#create dataframe of context counts per patient so as not to duplicate the counting of individual subs
for (k in 1:length(samples)){
  mut_count[k,]=count_muts_96(all_subs[all_subs$sample==samples[k],2:5])
  print(k)
}

mut_count = data.frame(mut_count)
row.names(mut_count) = samples

sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")

names(mut_count) = full_vec

samples_df = data.frame(Sample = samples, Patient = samples)

write.table(mut_count, "trinuc_mut_mat.txt", sep = ' ', quote = F, col.names = T, row.names = T)
write.table(samples_df, "key_table.txt", sep = ' ', quote = F, col.names = T, row.names = F)

```

```{bash}

#in directory with relevant files

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp_inc_addies_samples/01_INPUT

cp trinuc_mut_mat.txt ../02_INITIAL_OUTPUT
cp key_table.txt ../02_INITIAL_OUTPUT

cd ../02_INITIAL_OUTPUT

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_hdp"

for n in $(seq 1 20);
do
bsub -o $PWD/log.%J -e $PWD/err.%J -q normal -R'select[mem>10000] rusage[mem=10000]' -M10000 -J $n /software/R-3.6.1/bin/Rscript /nfs/users/nfs_t/to3/scripts/hdp_single_chain_tc16.R $n
done 

```

####Review diagnostic plots & combine

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp_inc_addies_samples/02_INITIAL_OUTPUT')

mutations=read.table("trinuc_mut_mat.txt", header = T)
key_table=read.table("key_table.txt", header = T)

lower_threshold = 100
sample_remove=rownames(mutations)[rowSums(mutations)<lower_threshold]
mutations=mutations[!rownames(mutations)%in%sample_remove,]
key_table=key_table[!key_table$Sample%in%sample_remove,]

options(stringsAsFactors = F)
library(hdp)

chlist <- vector("list", 20)
for (i in 1:20){
  if(file.exists(paste0("hdp_chain_",i,".Rdata"))){
    chlist[[i]] <- readRDS(paste0("hdp_chain_",i,".Rdata"))
  }
}
if(any(unlist(lapply(chlist,is.null)))) chlist=chlist[-which(unlist(lapply(chlist,is.null)))]

mut_example_multi <- hdp_multi_chain(chlist)

pdf("QC_plots_chain.pdf") 
par(mfrow=c(2,2), mar=c(4, 4, 2, 1))
p1 <- lapply(chains(mut_example_multi), plot_lik, bty="L", start=1000)
p2 <- lapply(chains(mut_example_multi), plot_numcluster, bty="L")
p3 <- lapply(chains(mut_example_multi), plot_data_assigned, bty="L")
dev.off()

mut_example_multi <- hdp_extract_components(mut_example_multi) #This step can take a while. If too long, submit R script as job
saveRDS(mut_example_multi,"HDP_multi_chain.Rdata")

#mut_example_multi <- readRDS("HDP_multi_chain.Rdata")

pdf("muts_attributed.pdf")
plot_comp_size(mut_example_multi, bty="L")
dev.off()

trinuc_context <- sapply(strsplit(colnames(mut_count), '\\.'), `[`, 4)
group_factor <- as.factor(rep(c("C>A", "C>G", "C>T", "T>A", "T>C", "T>G"),
                              each=16))
mut_colours=c("dodgerblue","black","red","grey70","olivedrab3","plum2")

for (i in 0:mut_example_multi@numcomp){
  pdf(paste0("hdp_component_",i,".pdf"),width=12,height=4)
  
  plot_comp_distn(mut_example_multi, cat_names=trinuc_context,
                  grouping=group_factor, col=mut_colours,comp=i,
                  col_nonsig="grey80", show_group_labels=TRUE)
  dev.off()
}

plot_dp_comp_exposure(mut_example_multi,
                      dpindices=2:4, incl_numdata_plot=FALSE,
                      col=c(RColorBrewer::brewer.pal(12, "Paired"),"magenta","firebrick"),
                      incl_nonsig=TRUE, cex.names=0.8,
                      ylab_exp = 'Signature exposure', leg.title = 'Signature')

pdf("signature_attribution.pdf",width=10,height=8)

plot_dp_comp_exposure(mut_example_multi, dpindices=(length(unique(key_table$Patient))+2):length(mut_example_multi@comp_dp_counts), incl_nonsig = T, ylab_exp = 'Signature exposure', leg.title = 'Signature',
                      col=c(RColorBrewer::brewer.pal(12, "Set3"),"magenta","firebrick"))
dev.off()

mean_assignment=as.data.frame(comp_dp_distn(mut_example_multi)$mean)
write.table(mean_assignment,"mean_assignment_hdp.txt")
mean_sigs=as.data.frame(t(comp_categ_distn(mut_example_multi)$mean))
write.table(mean_sigs,"hdp_sigs.txt")

```

####Deconvolve signatures

Component 0 is the known noise element.

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp_inc_addies_samples/02_INITIAL_OUTPUT')

options(stringsAsFactors = F)
library(hdp)
library(RColorBrewer)
library(lsa)
library(lattice)

mut.cols = rep(c("dodgerblue","black","red","grey70","olivedrab3","plum2"),each=16)

#Load HDP signatures
hdp_sigs=read.table("hdp_sigs.txt")

#Load reference signatures
#ref = read.csv("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/PCAWG_sigProfiler_SBS_signatures.csv", header=T, stringsAsFactors = F, row.names = 1)
ref = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/99_REFERENCE_DATA/COSMIC_v3.2_SBS_GRCh37.txt', header = T, sep = '\t', stringsAsFactors = F, row.names = 1)

#Assess cosine similarities for all reference signatures
cosine_matrix=data.frame(matrix(nrow=ncol(hdp_sigs), ncol=ncol(ref)))
rownames(cosine_matrix)=colnames(hdp_sigs)
colnames(cosine_matrix)=colnames(ref)

for (n in 1:nrow(cosine_matrix)) {
  for (m in 1:ncol(cosine_matrix)) {
    cosine_matrix[n,m] <- cosine(x=hdp_sigs[,rownames(cosine_matrix)[n]],
                                 y=ref[,colnames(cosine_matrix)[m]])
  }
}

write.table(cosine_matrix, "Cosine_similarities.txt",sep="\t",quote=F)

#plot output
pdf("cosine_similarities.pdf", height=5, width=15)
color.palette = colorRampPalette(c("white", "orange", "purple"))
levelplot(t(cosine_matrix[dim(cosine_matrix)[1]:1,]),col.regions=color.palette, aspect="fill", scales=list(x=list(rot=90)))
dev.off()

#select which signatures to decompose into reference signatures
sigs_to_decompose=rowSums(cosine_matrix>0.85)

#X7 is just noise
#X2 closely matches 31 only (>0.85), no further extraction required
#X5 closely matches 17b only (>0.85), no further extraction required
#X8 closely matches 91 only (>0.85), no further extraction required

excluded_sigs = c('X2', 'X5', 'X7', "X8") #either noise or already a close match
sigs_close_cosine = c()
for(n in 1:nrow(cosine_matrix)){
  if(!row.names(cosine_matrix)[n] %in% excluded_sigs){
    sigs_close_cosine = c(sigs_close_cosine, colnames(cosine_matrix)[cosine_matrix[n,]>0.7])
    print(paste0(rownames(cosine_matrix)[n],": ",paste(colnames(cosine_matrix)[cosine_matrix[n,]>0.7],collapse=",")))
  }
}

#"X0: SBS3,SBS5,SBS25,SBS39,SBS40,SBS89,SBS92"
#"X1: SBS3,SBS5,SBS40,SBS89,SBS92"
#"X3: SBS10c,SBS18,SBS29,SBS36,SBS56"
#"X4: "
#"X6: SBS1,SBS6,SBS87"
#"X9: "

#?X4 and 9 novel

#First iteration; decomposed hdp sigs into all suspected sigs 
gdsigs = unique(sigs_close_cosine)

signatures=t(ref[,gdsigs])
sample_list=paste0("X",c(0, 1, 3, 6)) #remove noisy, novel and already matched signatures
profiles=hdp_sigs[,sample_list]

signature_fraction = matrix(NA,nrow=nrow(signatures),ncol=length(sample_list))
rownames(signature_fraction) = rownames(signatures)
colnames(signature_fraction) = sample_list
maxiter <- 1000

for (j in 1:length(sample_list)) {
  freqs = profiles[,j]
  freqs[is.na(freqs)] = 0
  # EM algo with to estimate the signature contribution
  alpha = runif(nrow(signatures)); alpha=alpha/sum(alpha) # Random start (seems to give ~identical results)
  for (iter in 1:maxiter) {
    contr = t(array(alpha,dim=c(nrow(signatures),96))) * t(signatures)
    probs = contr/array(rowSums(contr),dim=dim(contr))
    probs = probs * freqs
    old_alpha = alpha
    alpha = colSums(probs)/sum(probs)
    if (sum(abs(alpha-old_alpha))<1e-5) {
      break
    }
  }
  # Saving the signature contributions for the sample
  print(j/length(sample_list))
  signature_fraction[,j] = alpha
}

#Plot initial deconvolution and save results
pdf("Deconvolution_hdp_sigs_R1.pdf", height=5, width=10)
#dev.new(height=5, width=10)
color.palette = colorRampPalette(c("white", "orange", "purple"))
levelplot((signature_fraction[nrow(signature_fraction):1,]),col.regions=color.palette, aspect="fill", scales=list(x=list(rot=90)))
dev.off()

write.table(signature_fraction, "hdp_known_sigs_broken_down_into_cosmic_gd_sigs.txt", sep="\t", col.names=T, row.names = T, quote=F)

sigs_deconv_R2=list()

for(n in 1:length(sample_list)){
  sigs_deconv_R2[[n]]=rownames(signature_fraction)[signature_fraction[,n] > 0.2] #SBS87 shares SBS1's C>T peaks resulting in some difficulty distinguishing it cleanly from SBS1, threshold for calling a known signature in the hdp extraction raised to 0.2 in order to prevent overfitting SBS87 in samples with no prior thiopurine exposure
}

names(sigs_deconv_R2)=colnames(signature_fraction)

#X3 is predominantly SBS18

sigs_to_deconv=names(sigs_deconv_R2)[unlist(lapply(sigs_deconv_R2,length))>1]

ref_sigs_R2=sort(unique(unlist(sigs_deconv_R2)))
signature_fractionR2=matrix(NA,ncol=length(sigs_to_deconv),nrow=length(ref_sigs_R2))
rownames(signature_fractionR2)=ref_sigs_R2
colnames(signature_fractionR2)=sigs_to_deconv

#repeat the deconvolution with the identified constitutive signatures
n = 1
for(s in sigs_to_deconv){
  gdsigs <- sigs_deconv_R2[[s]]
  signatures <- t(ref[,gdsigs])
  
  signature_fraction = matrix(NA,nrow=nrow(signatures),ncol=length(sample_list))
  rownames(signature_fraction) = rownames(signatures)
  colnames(signature_fraction) = sample_list
  maxiter <- 1000
  
  freqs = profiles[,s]
  freqs[is.na(freqs)] = 0
  
  alpha = runif(nrow(signatures)); alpha=alpha/sum(alpha) # Random start (seems to give ~identical results)
  for (iter in 1:maxiter) {
    contr = t(array(alpha,dim=c(nrow(signatures),96))) * t(signatures)
    probs = contr/array(rowSums(contr),dim=dim(contr))
    probs = probs * freqs
    old_alpha = alpha
    alpha = colSums(probs)/sum(probs)
    if (sum(abs(alpha-old_alpha))<1e-5) {
      break
    }
  }
  # Saving the signature contributions for the sample
  signature_fractionR2[gdsigs,n] = alpha
  n=n+1
  reconsbs <- rep(0,96)
  for (g in gdsigs) {
    reconsbs=reconsbs+(ref[,g]*alpha[g])
  }
  cosine_reconst=cosine(x=reconsbs, y=hdp_sigs[,s])
  print(paste0(s,": ",cosine_reconst))
  pdf(paste0("HDP_",s,"_reconstitution2.pdf"))
  par(mfrow=c(length(alpha)+2,1))
  par(mar=c(1,2,4,1))
  barplot(hdp_sigs[,s], col=mut.cols, main=paste0("HDP ",s),names.arg="")
  barplot(reconsbs, col=mut.cols, main=paste0("Reconstituted ",s," cosine similarity to original: ", round(cosine_reconst, digits=2)))
  for (g in gdsigs) {
    barplot(ref[,g], col=mut.cols, main=paste0("PCAWG ", g, " accounts for ", round(alpha[g], digits=2)))
  }
  dev.off()
}

#add previous signatures back in
sigs_deconv_R2$X2 = "SBS31"
sigs_deconv_R2$X3 = "SBS18"
sigs_deconv_R2$X4 = "N1"
sigs_deconv_R2$X5 = "SBS17b"
sigs_deconv_R2$X7 = "noise"
sigs_deconv_R2$X8 = "SBS91"
sigs_deconv_R2$X9 = "N2"

saveRDS(sigs_deconv_R2,"hdp2refsigs.Rdata")

ref_sigs_R2 = c(ref_sigs_R2, 'SBS17b', 'SBS31', 'SBS91')
#"SBS1"   "SBS18"  "SBS40"  "SBS5"   "SBS17b" "SBS31"  "SBS91"

#Combine the hdp signature that did not get deconvolved and reference signatures into final table
final_sigs = cbind(hdp_sigs[,c("X4", "X9")],ref[,ref_sigs_R2])
#Rename the HDP components that didn't get deconvoluted
colnames(final_sigs)[1:2] = c("N1", "N2")

#Order them - optional
final_sigs = final_sigs[, c('N1', "N2", 'SBS1', 'SBS5', 'SBS17b', 'SBS18', 'SBS31', 'SBS40', 'SBS91')]

write.table(final_sigs, "final_sigs_2022_01_02.txt", sep = '\t', col.names = T, row.names = T, quote = F)

```

#### Clean up novel signature

N1 still evidently has a minor component of SBS1. This is likely because the key peaks seem to coincide with samples containing slightly higher burdens of SBS1. To address this, we will run HDP again but with priors this time and the finite reportoire of known signatures we identified first time. We can also check if there was a smaller component of any signatures we missed the first time round.

#####Parts A, B & C
PART A : Arrange the prior sigs
PART B : Set up HDP
PART C : Run posterior sampling chains

```{r}

#/nfs/users/nfs_t/to3/scripts/hdp_prior_single_chain_updated_20220102.R

options(stringsAsFactors = F)
library(hdp)
lower_threshold=100

n=as.numeric(commandArgs(T)[1])
mutations=read.table("trinuc_mut_mat.txt", header = T)
key_table=read.table("key_table.txt", header = T)

ref = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp_inc_addies_samples/02_INITIAL_OUTPUT/final_sigs_2022_01_02.txt', header = T, stringsAsFactors = F, sep = '\t')
ref = ref[, !colnames(ref) %in% c('N1', 'N2')]
prior_sigs = as.matrix(ref)

# number of prior signatures to condition on (8)
nps <- ncol(prior_sigs)

# take a look
#prior_sigs[1:6,1:6]
# columns should sum to one
#colSums(prior_sigs)

# have a look at the first two prior sigs:
#barplot(prior_sigs[,1], las=2, cex.names=0.5)
#barplot(prior_sigs[,'SBS18'], las=2, cex.names=0.5)
#barplot(prior_sigs[,'SBS91'], las=2, cex.names=0.5)

#If requiring a minimum number of mutations:
sample_remove=rownames(mutations)[rowSums(mutations)<lower_threshold]
mutations=mutations[!rownames(mutations)%in%sample_remove,]
key_table=key_table[!key_table$Sample%in%sample_remove,]

#Hierarchy is set per patient, can change if wanted
freq = table(key_table$Patient)

#nps <- ncol(prior_sigs)

hdp_prior <- hdp_prior_init(prior_distn = prior_sigs, # matrix of prior sigs
                             prior_pseudoc = rep(1000, nps), # pseudocount weights
                             hh=rep(1, 96), # uniform prior over 96 categories
                             alphaa=c(1, 1), # shape hyperparams for 2 CPs
                             alphab=c(1, 1)) # rate hyperparams for 2 CPs
#hdp_prior
#numdp(hdp_prior)
#pseudoDP(hdp_prior)
#conparam(hdp_prior)
#ppindex(hdp_prior)
#cpindex(hdp_prior)
#dpstate(hdp_prior) # 2 for active node, 1 for frozen, 0 for heldout

# make two more CPs available for the data we will add
hdp_prior <- hdp_addconparam(hdp_prior,
                              alphaa = rep(1,length(freq)+2), # shape hyperparams for x new CPs
                              alphab = rep(1,length(freq)+2)) # rate hyperparams for x new CPs

hdp_prior <- hdp_adddp(hdp_prior,
                        numdp = nrow(mutations) + 1,
                        ppindex = c(1, rep(1+nps+1:(length(freq)), times=freq)),
                        cpindex = c(3, rep(4:(length(freq)+3), times=freq)))

# assign the data to the relevant DP nodes
hdp_prior <- hdp_setdata(hdp_prior,
                          dpindex = (1 + nps + 1) + 1:nrow(mutations), 
                          mutations) # mutation counts in all GCTs


hdp_activated <- dp_activate(hdp_prior, 
                             dpindex = (1+nps+1)+0:nrow(mutations), 
                             initcc = nps+5,
                             seed = n * 1000)

chain=hdp_posterior(hdp_activated,
                    burnin=80000,
                    n=100,
                    seed=n*1000,
                    space=200,
                    cpiter=3)

saveRDS(chain,paste0("hdp_prior_chain_",n,".Rdata"))

```

```{bash}

#in directory with relevant files

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp_inc_addies_samples/01_INPUT

cp trinuc_mut_mat.txt ../03_NOVEL_SIG_CLEANUP
cp key_table.txt ../03_NOVEL_SIG_CLEANUP

cd ../03_NOVEL_SIG_CLEANUP

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_hdp"

for n in $(seq 1 20);
do
bsub -o $PWD/log.%J -e $PWD/err.%J -q normal -R'select[mem>10000] rusage[mem=10000]' -M10000 -J $n /software/R-3.6.1/bin/Rscript /nfs/users/nfs_t/to3/scripts/hdp_prior_single_chain_updated_20220102.R $n
done

```

#####Parts D & E

PART D : Review diagnostic plots
PART E : Extract consensus components / signatures

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp_inc_addies_samples/03_NOVEL_SIG_CLEANUP')

options(stringsAsFactors = F)
library(hdp)
library(ggpubr)

chlist <- vector("list", 20)
for (i in 1:20){
  if(file.exists(paste0("hdp_prior_chain_",i,".Rdata"))){
    chlist[[i]] <- readRDS(paste0("hdp_prior_chain_",i,".Rdata"))
  }
}
if(any(unlist(lapply(chlist,is.null)))) chlist=chlist[-which(unlist(lapply(chlist,is.null)))]

hdp_multi <- hdp_multi_chain(chlist)

pdf("QC_plots_chain.pdf") 
par(mfrow=c(2,2), mar=c(4, 4, 2, 1))
p1 <- lapply(chains(hdp_multi), plot_lik, bty="L", start=1000)
p2 <- lapply(chains(hdp_multi), plot_numcluster, bty="L")
p3 <- lapply(chains(hdp_multi), plot_data_assigned, bty="L")
dev.off()

#QC looks good 16/9/21

hdp_multi <- hdp_extract_components(hdp_multi)
saveRDS(hdp_multi,"HDP_multi_chain.Rdata")

numcomp(hdp_multi) #11
prop.ex(hdp_multi) #0.999

pdf("muts_attributed.pdf")
plot_comp_size(hdp_multi, bty="L")
dev.off()

# plot components / signatures
# pick your colours
mut_colours=c("dodgerblue","black","red","grey70","olivedrab3","plum2")
# labels along bottom x-axis
trinuc_context <- sapply(strsplit(colnames(mut_count), '\\.'), `[`, 4)
# group labels along the top (and controls colour grouping)
group_factor <- as.factor(rep(c("C>A", "C>G", "C>T", "T>A", "T>C", "T>G"),
                              each=16))

ref = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp_inc_addies_samples/02_INITIAL_OUTPUT/final_sigs_2022_01_02.txt', header = T, stringsAsFactors = F, sep = '\t')
ref = ref[, !colnames(ref) %in% c('N1', 'N2')]
prior_sigs = as.matrix(ref)

# number of prior signatures to condition on (8)
nps <- ncol(prior_sigs)

pdf(paste0("hdp_components_with_priors.pdf"),width=12,height=4)
plot_comp_distn(hdp_multi, cat_names=trinuc_context,
                grouping=group_factor, col=mut_colours,
                col_nonsig="grey97", show_group_labels=TRUE)
dev.off()

lower_threshold = 100

mutations=read.table("trinuc_mut_mat.txt", header = T)
sample_remove=rownames(mutations)[rowSums(mutations)<lower_threshold]
mutations=mutations[!rownames(mutations)%in%sample_remove,]


pdf("signature_attribution.pdf",width=10,height=8)
plot_dp_comp_exposure(hdp_multi, dpindices = (1+nps+1)+1:nrow(mutations), incl_nonsig = F,
                      col=c('black', get_palette(palette = "Set1", 15)), cex.names=0.8,
                      ylab_exp = 'Signature exposure', leg.title = 'Signature')
dev.off()

mean_assignment=as.data.frame(comp_dp_distn(hdp_multi)$mean)
write.table(mean_assignment,"mean_assignment_prior_hdp.txt")
mean_sigs=as.data.frame(t(comp_categ_distn(hdp_multi)$mean))
write.table(mean_sigs,"hdp_prior_sigs.txt")

colnames(prior_sigs)[c(1, 2, 3, 4, 5, 7)]
#"SBS1"   "SBS5"   "SBS17b" "SBS18"  "SBS31"  "SBS91" 

colnames(prior_sigs)[c(7)] #SBS40 not found on repeat extraction, leaving "SBS1"   "SBS5"   "SBS17b" "SBS18"  "SBS31"  "SBS91"
```

#####Deconvolute 'novel' signatures

```{r}

#Deconvolute HDP signatures into reference signatures and arrive at final set

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp_inc_addies_samples/03_NOVEL_SIG_CLEANUP')

options(stringsAsFactors = F)
library(hdp)
library(RColorBrewer)
library(lsa)
library(lattice)

mut.cols = rep(c("dodgerblue","black","red","grey70","olivedrab3","plum2"),each=16)

#Load HDP signatures
mean_sigs = read.table("hdp_prior_sigs.txt", row.names = 1, header = T)
hdp_sigs = mean_sigs
#Load reference signatures, need formatting
#ref = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/MutationalPatterns/reformatted_cosmic_sig_v3.txt', header = T, stringsAsFactors = F, row.names = 1, check.names = F)
#ref = t(ref)
#cosmic.sigs = read.csv('/lustre/scratch119/casm/team294rr/to3/testes/tumour/MutationalPatterns/PCAWG_sigProfiler_SBS_signatures.csv', header = T, stringsAsFactors = F, row.names = 1, check.names = F)
#row.names(cosmic.sigs) = row.names(ref)
#prior_sigs = as.matrix(cosmic.sigs)
ref = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/99_REFERENCE_DATA/COSMIC_v3.2_SBS_GRCh37.txt', header = T, sep = '\t', stringsAsFactors = F, row.names = 1)

#Assess cosine similarities for all unassigned signatures (X0, N1, N2)
cosine_matrix=data.frame(matrix(nrow=ncol(hdp_sigs[, c('N1', 'N2', 'N3', 'N4', 'N5')]), ncol=ncol(ref)))
rownames(cosine_matrix)=colnames(hdp_sigs[, c('N1', 'N2', 'N3', 'N4', 'N5')])
colnames(cosine_matrix)=colnames(ref)

for (n in 1:nrow(cosine_matrix)) {
  for (m in 1:ncol(cosine_matrix)) {
    cosine_matrix[n,m] <- cosine(x=hdp_sigs[,rownames(cosine_matrix)[n]],
                                 y=ref[,colnames(cosine_matrix)[m]])
  }
}

write.table(cosine_matrix, "Cosine_similarities_unassigned_to_priors.txt",sep="\t",quote=F)

hdp_sigs = hdp_sigs[, c('N1', 'N2', 'N3', 'N4', 'N5')]

#plot output
pdf("cosine_similarities_unassigned_to_priors.pdf", height=5, width=15)
color.palette = colorRampPalette(c("white", "orange", "purple"))
levelplot(t(cosine_matrix[dim(cosine_matrix)[1]:1,]),col.regions=color.palette, aspect="fill", scales=list(x=list(rot=90)))
dev.off()

#select which signatures to decompose into reference signatures
sigs_to_decompose=rowSums(cosine_matrix>0.85)
#N3 is SBS17a
#N4 is noise

excluded_sigs = c('N3') #already a close match
sigs_close_cosine = c()
for(n in 1:nrow(cosine_matrix)){
  if(!row.names(cosine_matrix)[n] %in% excluded_sigs){
    sigs_close_cosine = c(sigs_close_cosine, colnames(cosine_matrix)[cosine_matrix[n,]>0.7])
    print(paste0(rownames(cosine_matrix)[n],": ",paste(colnames(cosine_matrix)[cosine_matrix[n,]>0.7],collapse=",")))
  }
}

#[1] "N1: SBS3,SBS5,SBS39,SBS40"
#"N2: "
#"N4: SBS3,SBS5,SBS40,SBS89"
#"N5: "

#First iteration; decomposed hdp sigs into all suspected sigs 
gdsigs = unique(sigs_close_cosine)

signatures=t(ref[,gdsigs])
sample_list= c('N1', 'N4')
profiles=hdp_sigs[,sample_list]

signature_fraction = matrix(NA,nrow=nrow(signatures),ncol=length(sample_list))
rownames(signature_fraction) = rownames(signatures)
colnames(signature_fraction) = sample_list
maxiter <- 1000

for (j in 1:length(sample_list)) {
  freqs = profiles[,j]
  freqs[is.na(freqs)] = 0
  # EM algorithm with to estimate the signature contribution
  alpha = runif(nrow(signatures)); alpha=alpha/sum(alpha) # Random start (seems to give ~identical results)
  for (iter in 1:maxiter) {
    contr = t(array(alpha,dim=c(nrow(signatures),96))) * t(signatures)
    probs = contr/array(rowSums(contr),dim=dim(contr))
    probs = probs * freqs
    old_alpha = alpha
    alpha = colSums(probs)/sum(probs)
    if (sum(abs(alpha-old_alpha))<1e-5) {
      break
    }
  }
  # Saving the signature contributions for the sample
  print(j/length(sample_list))
  signature_fraction[,j] = alpha
}

#Plot initial deconvolution and save results
pdf("Deconvolution_hdp_sigs_after_prior_R1.pdf", height=5, width=10)
color.palette = colorRampPalette(c("white", "orange", "purple"))
levelplot((signature_fraction[nrow(signature_fraction):1,]),col.regions=color.palette, aspect="fill", scales=list(x=list(rot=90)))
dev.off()

#N4 is noise

write.table(signature_fraction[,'N1'], "hdp_unassigned_sig_broken_down_into_cosmic_sigs.txt", sep="\t", col.names=T, row.names = T, quote=F)

sigs_deconv_R2=list()

for(n in 1:length(sample_list)){
  sigs_deconv_R2[[n]]=rownames(signature_fraction)[signature_fraction[,n]>0.2]
}

names(sigs_deconv_R2)=colnames(signature_fraction)

#No signature has one major contributor only

sigs_to_deconv=names(sigs_deconv_R2)[unlist(lapply(sigs_deconv_R2,length))>1]

ref_sigs_R2=sort(unique(unlist(sigs_deconv_R2)))
signature_fractionR2=matrix(NA,ncol=length(sigs_to_deconv),nrow=length(ref_sigs_R2))
rownames(signature_fractionR2)=ref_sigs_R2
colnames(signature_fractionR2)=sigs_to_deconv

#repeat the deconvolution with the identified constitutive signatures
n = 1
for(s in sigs_to_deconv){
  gdsigs <- sigs_deconv_R2[[s]]
  signatures <- t(ref[,gdsigs])
  
  signature_fraction = matrix(NA,nrow=nrow(signatures),ncol=length(sample_list))
  rownames(signature_fraction) = rownames(signatures)
  colnames(signature_fraction) = sample_list
  maxiter <- 1000
  
  freqs = profiles[,s]
  freqs[is.na(freqs)] = 0
  
  alpha = runif(nrow(signatures)); alpha=alpha/sum(alpha) # Random start (seems to give ~identical results)
  for (iter in 1:maxiter) {
    contr = t(array(alpha,dim=c(nrow(signatures),96))) * t(signatures)
    probs = contr/array(rowSums(contr),dim=dim(contr))
    probs = probs * freqs
    old_alpha = alpha
    alpha = colSums(probs)/sum(probs)
    if (sum(abs(alpha-old_alpha))<1e-5) {
      break
    }
  }
  # Saving the signature contributions for the sample
  signature_fractionR2[gdsigs,n] = alpha
  n=n+1
  reconsbs <- rep(0,96)
  for (g in gdsigs) {
    reconsbs=reconsbs+(ref[,g]*alpha[g])
  }
  cosine_reconst=cosine(x=reconsbs, y=hdp_sigs[,s])
  print(paste0(s,": ",cosine_reconst))
  pdf(paste0("HDP_",s,"_reconstitution2.pdf"))
  par(mfrow=c(length(alpha)+2,1))
  par(mar=c(1,2,4,1))
  barplot(hdp_sigs[,s], col=mut.cols, main=paste0("HDP ",s),names.arg="")
  barplot(reconsbs, col=mut.cols, main=paste0("Reconstituted ",s," cosine similarity to original: ", round(cosine_reconst, digits=2)))
  for (g in gdsigs) {
    barplot(ref[,g], col=mut.cols, main=paste0("PCAWG ", g, " accounts for ", round(alpha[g], digits=2)))
  }
  dev.off()
}

#complete sig list
sigs_deconv_R2$N2 = 'N2'
sigs_deconv_R2$N3 = 'SBS17a'
sigs_deconv_R2$N4 = 'Noise'
sigs_deconv_R2$N5 = 'SBS35 C>A component'

saveRDS(sigs_deconv_R2,"novel_hdp2refsigs.Rdata")

ref_sigs_R2 = c("SBS1", "SBS5", "SBS17a", "SBS17b", "SBS18", "SBS31", "SBS35", "SBS39", "SBS40", "SBS91") #gained SBS17a & 39, added SBS35 because likely different platinum signatures in different samples

#Combine the hdp signatures that did not get deconvolved and reference signatures into final table
final_sigs = cbind(hdp_sigs[,c("N2")],ref[,ref_sigs_R2])
#Rename the HDP components that didn't get deconvoluted
colnames(final_sigs)[1] = c("N2")

write.table(final_sigs, "final_sigs_CLEANED_UP_2022_01_02.txt", col.names = T, row.names = T, sep = '\t', quote = F)

```

#####Map back signatures with Sigprofiler (per sample) *TO DO*

Need substitutions per sample

```{r}

library(reshape2)
library(ggplot2)
library(ggpubr)

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp_inc_addies_samples/04_SIGPROFILER_DECONVOLUTION')

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', header = T, stringsAsFactors = F, sep = '\t')

hdp_sigprofiler_mapped = read.table('Decompose_Solution_Activities.txt', header = T, sep = '\t', stringsAsFactors = F)
#hdp_sigprofiler_mapped$Samples = unlist(strsplit(hdp_sigprofiler_mapped$Samples, '_post_swater_final_snvs')) the sample names needed editing from the raw sigprofiler output
#write.table(hdp_sigprofiler_mapped, 'Decompose_Solution_Activities.txt', col.names = T, row.names = F, sep = '\t', quote = F)

hdp_sigprofiler_mapped$SBS5_40 = hdp_sigprofiler_mapped$SBS5 + hdp_sigprofiler_mapped$SBS40 
hdp_sigprofiler_mapped = hdp_sigprofiler_mapped[, c("Samples", "SBS1", "SBS5_40", "SBS17a", "SBS17b", "SBS18", "SBS35", "SBS91", "N2")]

hdp_sigprofiler_mapped_norm = hdp_sigprofiler_mapped[, c(2:9)] / rowSums(hdp_sigprofiler_mapped[, c(2:9)])
hdp_sigprofiler_mapped_norm$sample = hdp_sigprofiler_mapped$Samples
hdp_sigprofiler_mapped_norm$tumour = substr(hdp_sigprofiler_mapped_norm$sample, 0, 7)

hdp_sigprofiler_mapped_norm.m = reshape2::melt(hdp_sigprofiler_mapped_norm, id.vars = c('sample', 'tumour'))

hdp_sigprofiler_mapped_norm.m$age = NA
hdp_sigprofiler_mapped_norm.m$diagnosis = NA
hdp_sigprofiler_mapped_norm.m$site = NA
hdp_sigprofiler_mapped_norm.m$histology = NA

for(i in 1:nrow(hdp_sigprofiler_mapped_norm.m)){
  hdp_sigprofiler_mapped_norm.m$age[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Age
  hdp_sigprofiler_mapped_norm.m$diagnosis[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Diagnosis
  hdp_sigprofiler_mapped_norm.m$site[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Organ
  hdp_sigprofiler_mapped_norm.m$histology[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Histology
}

hdp_sigprofiler_mapped_norm.m$variable = factor(hdp_sigprofiler_mapped_norm.m$variable, levels = rev(unique(hdp_sigprofiler_mapped_norm.m$variable)))
hdp_sigprofiler_mapped_norm.m$tumour = factor(hdp_sigprofiler_mapped_norm.m$tumour, levels = unique(hdp_sigprofiler_mapped_norm.m[order(hdp_sigprofiler_mapped_norm.m$age),]$tumour))

cols = get_palette(palette = 'Set1', k = length(unique(hdp_sigprofiler_mapped_norm.m$variable)))

h1 = ggplot(hdp_sigprofiler_mapped_norm.m) + 
  geom_bar(mapping = aes(x = sample, y = value, fill = variable), stat = 'identity') + 
  theme_pubr() + coord_cartesian(expand = F) + 
  scale_fill_manual(values=cols) + 
  ylab('Proportion of substitutions') + 
  xlab('') + 
  theme(axis.text.x = element_blank(), legend.position = 'right', axis.ticks.x = element_blank()) + facet_grid(. ~ tumour, scales = 'free_x', space = 'free_x')
h2 = ggplot(hdp_sigprofiler_mapped_norm.m) + 
  geom_bar(mapping = aes(x = sample, y = 1, fill = histology),  stat = "identity", position = "dodge") + 
  theme_pubr() + coord_cartesian(expand = F) + 
  scale_fill_manual(values = get_palette('Dark2', length(unique(hdp_sigprofiler_mapped_norm.m$histology)))) + 
  ylab('') + 
  xlab('') + 
  theme(axis.text.x = element_blank(), legend.position = 'right', axis.ticks.x = element_blank(), strip.text = element_blank(), axis.ticks = element_blank(), axis.text = element_blank()) + facet_grid(. ~ tumour, scales = 'free_x', space = 'free_x')
h3 = ggplot(hdp_sigprofiler_mapped_norm.m) + 
  geom_bar(mapping = aes(x = sample, y = 1, fill = diagnosis),  stat = "identity", position = "dodge") + 
  theme_pubr() + coord_cartesian(expand = F) + 
  scale_fill_manual(values = get_palette('Dark2', length(unique(hdp_sigprofiler_mapped_norm.m$diagnosis)))) + 
  ylab('') + 
  xlab('') + 
  theme(axis.text.x = element_blank(), legend.position = 'right', axis.ticks.x = element_blank(), strip.text = element_blank(), axis.ticks = element_blank(), axis.text = element_blank()) + facet_grid(. ~ tumour, scales = 'free_x', space = 'free_x')
h4 = ggplot(hdp_sigprofiler_mapped_norm.m) + 
  geom_bar(mapping = aes(x = sample, y = 1, fill = site),  stat = "identity", position = "dodge") + 
  theme_pubr() + coord_cartesian(expand = F) + 
  scale_fill_manual(values = get_palette('default', length(unique(hdp_sigprofiler_mapped_norm.m$site)))) + 
  ylab('') + 
  xlab('') + 
  theme(axis.text.x = element_blank(), legend.position = 'right', axis.ticks.x = element_blank(), strip.text = element_blank(), axis.ticks = element_blank(), axis.text = element_blank()) + facet_grid(. ~ tumour, scales = 'free_x', space = 'free_x')

plot_grid(h1, h2, h3, h4,  align = 'v', ncol = 1, rel_heights = c(1, 0.2, 0.2, 0.2))

```

###Are the A[C>G]G variants enriched to areas around retrotransposition events?

```{r}

library("GenomicRanges")
library("Rsamtools")
library("MASS")

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp/01_INPUT')

#all variants across each individual sample
all.var = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/gct_vagrent_ssm_all_mutations_plus_depth20210915.txt", header = T, sep = '\t', stringsAsFactors = F)
all.var = all.var[, c(2,3,4,5,6)]

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', header = T, stringsAsFactors = F, sep = '\t')

genomeFile = "/nfs/cancer_ref01/Homo_sapiens/37/genome.fa"

#annotate with trinucleotide context
mutations = all.var
colnames(mutations) = c("sample", "chr","pos","ref","mut")
mutations = mutations[(mutations$ref %in% c("A","C","G","T")) & (mutations$mut %in% c("A","C","G","T")) & mutations$chr %in% c(1:22,"X","Y"),]
mutations$trinuc_ref = as.vector(scanFa(genomeFile, GRanges(mutations$chr, IRanges(as.numeric(mutations$pos)-1,
                                                                                     as.numeric(mutations$pos)+1))))

ntcomp = c(T="A",G="C",C="G",A="T")
mutations$sub = paste(mutations$ref,mutations$mut,sep=">")
mutations$trinuc_ref_py = mutations$trinuc_ref

for (j in 1:nrow(mutations)) {
    if (mutations$ref[j] %in% c("A","G")) { # Purine base
      mutations$sub[j] = paste(ntcomp[mutations$ref[j]],ntcomp[mutations$mut[j]],sep=">")
      mutations$trinuc_ref_py[j] = paste(ntcomp[rev(strsplit(mutations$trinuc_ref[j],split="")[[1]])],collapse="")
    }
  }

#which samples have N2?
hdp_sigprofiler_mapped = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/sbs_signatures/hdp/04_SIGPROFILER_DECONVOLUTION/Decompose_Solution_Activities.txt', header = T, sep = '\t', stringsAsFactors = F)
n2_samples = hdp_sigprofiler_mapped[hdp_sigprofiler_mapped$N2 > 0,]$Samples

n2_sample_muts = mutations[mutations$sample %in% n2_samples,] #only look at samples with N2 signature

#rt coordinates per sample
rt_coord = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/retrotransposition_coordinates_20211231.txt", header = T, sep = '\t', stringsAsFactors = F)

#distance to nearest rt breakpoint
n2_sample_muts$nr_rt_breakpoint = NA
for(i in 1:nrow(n2_sample_muts)){
  if(nrow(rt_coord[rt_coord$sample == n2_sample_muts$sample[i] & rt_coord$chr == n2_sample_muts$chr[i], ] > 0)){
    put_rt_coord = rt_coord[rt_coord$sample == n2_sample_muts$sample[i] & rt_coord$chr == n2_sample_muts$chr[i], ]
    n2_sample_muts$nr_rt_breakpoint[i] = min(abs(n2_sample_muts$pos[i] - put_rt_coord$breakpoint))
  }
}

n2_sample_muts$a.c.g.g = "No"
n2_sample_muts[n2_sample_muts$sub == "C>G" & n2_sample_muts$trinuc_ref_py == "ACG", ]$a.c.g.g = "Yes"

n2_sample_muts$nr_rt_breakpoint_mb = n2_sample_muts$nr_rt_breakpoint / 1e6

ggplot(n2_sample_muts) + geom_density(mapping = aes(x= nr_rt_breakpoint_mb, col = a.c.g.g)) + theme_linedraw() + xlab("Nearest retrotransposon breakpoint (Mb, same chromsome)")

nrow(n2_sample_muts[n2_sample_muts$a.c.g.g == "Yes" & !is.na(n2_sample_muts$nr_rt_breakpoint_mb) & n2_sample_muts$nr_rt_breakpoint_mb < 1,]) / nrow(n2_sample_muts[n2_sample_muts$a.c.g.g == "Yes" ,]) #12/614 = 0.01954397
nrow(n2_sample_muts[n2_sample_muts$a.c.g.g == "No" & !is.na(n2_sample_muts$nr_rt_breakpoint_mb) & n2_sample_muts$nr_rt_breakpoint_mb < 1,]) / nrow(n2_sample_muts[n2_sample_muts$a.c.g.g == "No" ,]) #580 / 19012 = 0.03050705

```

##DRIVER ANALYSIS (SUBS + INDELS)

###Uncurated

####dnds (subs and indels)

```{r}

ssms.merged = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/gct_vagrent_ssm_all_mutations_plus_depth20210915.txt', header = T, sep = '\t', stringsAsFactors = F)

ssms.simplified = ssms.merged[, c(1, 3, 4, 5, 6, 10)]
ssms.simplified = ssms.simplified[!duplicated(ssms.simplified),]

max_NV = c()
for(i in 1:nrow(ssms.simplified)){
  data = ssms.merged[ssms.merged$Case_ID == ssms.simplified$Case_ID[i] & ssms.merged$Chr == ssms.simplified$Chr[i] & ssms.merged$Position == ssms.simplified$Position[i] & ssms.merged$Wildtype == ssms.simplified$Wildtype[i] & ssms.merged$Mutant == ssms.simplified$Mutant[i],]
  max_NV = c(max_NV, max(data$Variant_read_depth))
}

ssms.simplified$max_NV = max_NV

min(ssms.simplified[ssms.simplified$Mutation == 'Sub',]$max_NV) #2
min(ssms.simplified[ssms.simplified$Mutation != 'Sub',]$max_NV) #5

ssms.simplified[ssms.simplified$Mutation == 'Sub' & ssms.simplified$max_NV < 4,] #64233, #145997, #148428, #156324, #158729

merge(ssms.simplified[ssms.simplified$Mutation == 'Sub' & ssms.simplified$max_NV < 4,], ssms.merged, by = c("Case_ID",  "Chr","Position", "Wildtype", "Mutant", "Mutation"))

#library(devtools); install_github("im3sanger/dndscv")
library("dndscv")
library("seqinr")
library("Biostrings")
library("MASS")
library("GenomicRanges")
library(dplyr)

ssms.merged.conc = ssms.merged[, c('Case_ID', 'Chr', 'Position', 'Wildtype', 'Mutant')]
ssms.merged.conc.uniq = unique(ssms.merged.conc) #unique mutations per patient
names(ssms.merged.conc.uniq) = c("sampleID", "chr", "pos", "ref", "mut")

dnds_input <- ssms.merged.conc.uniq

dndsout = dndscv(dnds_input)

dndsout$globaldnds
#     name       mle     cilow   cihigh
#wmis wmis 0.9193256 0.7470043 1.131399
#wnon wnon 0.9015984 0.5625637 1.444956
#wspl wspl 1.2182946 0.6889338 2.154404
#wtru wtru 1.0083834 0.6892352 1.475312
#wall wall 0.9256158 0.7545217 1.135507

print(head(dndsout$sel_cv), digits = 3) #none significant after correction for multiple hypothesis testing
print(head(dndsout$annotmuts), digits = 3)

saveRDS(dndsout, "/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/driver_analysis/dndsoutput_20210915.rds")

```

####CHASMplus

```{r}

ssms.merged = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/gct_vagrent_ssm_all_mutations_plus_depth20210915.txt', header = T, sep = '\t', stringsAsFactors = F)

ssms.merged.conc = ssms.merged[, c('Case_ID', 'Chr', 'Position', 'Wildtype', 'Mutant', 'Gene')]
ssms.merged.conc.uniq = unique(ssms.merged.conc) #unique mutations per patient
ssms.merged.conc.uniq = ssms.merged.conc.uniq[!is.na(ssms.merged.conc.uniq$Gene), 1:5] #only those associated with a gene
names(ssms.merged.conc.uniq) = c("sampleID", "chr", "pos", "ref", "mut")
ssms.merged.conc.uniq$strand = "+" #CHASMplus seems to work best when this is set to + for all variants

ssms.merged.conc.uniq = ssms.merged.conc.uniq[, c("chr", "pos", "strand", "ref", "mut", "sampleID")]
ssms.merged.conc.uniq = ssms.merged.conc.uniq[(nchar(ssms.merged.conc.uniq$ref) == 1) & (nchar(ssms.merged.conc.uniq$mut) == 1),] #remove indels
ssms.merged.conc.uniq$chr = paste0('chr', ssms.merged.conc.uniq$chr)

write.table(ssms.merged.conc.uniq, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/driver_analysis/chasmplus_input_final_20210916.txt', col.names = F, row.names = F, sep = '\t', quote = F)

```


```{bash}

cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/driver_analysis/chasmplus_input_final_20210916.txt /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/driver_analysis/chasmplus

cravat -n chasmplus_output_final_20210916 -t excel -a chasmplus -d . -l hg19 chasmplus_input_final_20210916.txt

```

###Curated final drivers after filtering (SNV and indels)

####Read in total list of calls

```{r}

ssms = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/gct_vagrent_ssm_all_mutations_plus_depth20210915.txt', header = T,  sep = '\t', stringsAsFactors = F)
cosmic_v94 <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/99_REFERENCE_DATA/cosmic_v94_cancer_gene_census.csv', sep = ',', header = T, stringsAsFactors = F)

table(ssms$Consequence)

func_ssms = ssms[!ssms$Consequence %in% c('3prime_UTR_variant', '5prime_UTR_variant', 'downstream', 'inframe', 'intronic', 'nc_variant', 'silent', 'splice_region', 'upstream') & !is.na(ssms$Consequence),] #remove variants that are not of interest

uniq_func_ssms = func_ssms[, c(1, 3:6, 10:15)]
uniq_func_ssms = uniq_func_ssms[!duplicated(uniq_func_ssms),]

```

####Oncogenes
```{r}

oncogenes = unlist(strsplit(cosmic_v94[grepl('oncogene', cosmic_v94$Role.in.Cancer) | grepl('Dom', cosmic_v94$Molecular.Genetics),]$Synonyms, ",")) #including all synonyms
poss_onco_ssms = uniq_func_ssms[uniq_func_ssms$Gene %in% oncogenes,]
poss_onco_ssms$put_annot = "oncogene"
poss_onco_ssms[,6:11]

#Mutation    Gene          Transcript               Codon      Protein           Consequence
#Sub   RBM15           r.1969c>g           c.1069C>G      p.P357A              missense
#Sub    TBX3           r.1363u>c            c.973T>C      p.F325L              missense
#Sub   FGFR2           r.1417g>c            c.825G>C      p.E275D              missense
#Sub    KRAS             r.99g>u             c.35G>T       p.G12V              missense DRIVER
#Sub    AKT1           r.1716c>a            c.235C>A       p.Q79K              missense DRIVER
#Sub    KRAS             r.99g>a             c.35G>A       p.G12D              missense DRIVER
#Sub   PRRX1            r.932g>c            c.619G>C      p.A207P              missense
#Sub    SND1            r.791c>g            c.597C>G      p.I199M              missense
#Sub  NUP214            r.586g>a            c.442G>A      p.A148T              missense
#Sub    MTOR           r.7324c>a           c.7247C>A     p.A2416D              missense
#Sub    NRG1           r.2151a>c           c.1634A>C      p.K545T              missense
#Sub    TCF4           r.1781c>u           c.1724C>T      p.P575L              missense
#Sub   NTRK1           r.1005c>a            c.964C>A      p.L322I              missense
#Sub WHSC1L1           r.4484u>g           c.3966T>G     p.D1322E              missense
#Sub    KLF4            r.523c>u             c.49C>T       p.L17F              missense
#Sub    PAX8            r.962a>c            c.767A>C      p.K256T              missense
#Del     CIC    r.2385_2428del44    c.2345_2388del44 p.A782fs*134            frameshift
#Del   MED12 r.6416_6440+11del36 c.6384_6408+11del36          p.?            ess_splice
#Del    XPO1   r.351_401+37del88                 c.?          p.? 5prime_UTR_ess_splice

```


####Recessive cancer gene

```{r}

rcgenes = cosmic_v94[grepl('TSG', cosmic_v94$Role.in.Cancer) | grepl('Rec', cosmic_v94$Molecular.Genetics),]$Gene.Symbol
rcgenes = unlist(strsplit(cosmic_v94[grepl('TSG', cosmic_v94$Role.in.Cancer) | grepl('Rec', cosmic_v94$Molecular.Genetics),]$Synonyms, ",")) #including all synonyms

poss_rcg_ssms = uniq_func_ssms[uniq_func_ssms$Gene %in% rcgenes,]
poss_rcg_ssms$put_annot = "TSG"
poss_rcg_ssms[,6:11]
#Mutation   Gene          Transcript               Codon      Protein Consequence
#Sub   FAT4           r.1688g>a           c.1675G>A      p.V559I    missense
#Sub   MSH6           r.3381c>u           c.3229C>T     p.P1077S    missense
#Sub   TBX3           r.1363u>c            c.973T>C      p.F325L    missense
#Sub   CHD2           r.3770c>g           c.2702C>G      p.A901G    missense
#Sub  SETD2           r.7626a>c           c.7583A>C     p.E2528A    missense
#Sub  POLD1           r.2481g>a           c.2428G>A      p.A810T    missense
#Sub  CSMD3           r.6082g>a           c.5837G>A     p.R1946H    missense
#Sub   CBLB            r.837g>c            c.515G>C      p.R172P    missense
#Sub   NRG1           r.2151a>c           c.1634A>C      p.K545T    missense
#Sub FANCD2           r.1316g>c           c.1223G>C      p.R408P    missense
#Sub  PTPRD           r.3348u>a           c.2804T>A      p.V935D    missense
#Sub  NTRK1           r.1005c>a            c.964C>A      p.L322I    missense
#Sub   KLF4            r.523c>u             c.49C>T       p.L17F    missense
#Sub  ZMYM3            r.493c>g            c.305C>G      p.T102S    missense
#Del    CIC    r.2385_2428del44    c.2345_2388del44 p.A782fs*134  frameshift
#Del  MED12 r.6416_6440+11del36 c.6384_6408+11del36          p.?  ess_splice
#Sub  CDK12           r.1588g>u           c.1555G>T      p.E519*    nonsense

#PD50661 PD50661a  17 37627640        G      T               49                 10 0.2040816      Sub CDK12  r.1588g>u c.1555G>T p.E519*    nonsense but only 10/49 0.2 VAF in 2+2 location 71%
#PD50655  19  42795264                                             GCCCAGGGGGCCCCTGGTGGTGGGACCACTGCGGGCTCAGGAGCA      G      Del    CIC    r.2385_2428del44    c.2345_2388del44 p.A782fs*134            frameshift 20/59 0.34 VAF 2+2 62%

```

####Combined

```{r}

write.table(rbind(poss_onco_ssms, poss_rcg_ssms), "/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/driver_analysis/cosmic_v94/putative_sub_indel_drivers_20210102.txt", sep = '\t', row.names = F, col.names = T, quote = F)

```

###Curated final driver without downstream filtering

Only the following filters applied to LCM data.
- Subs - ms44 filters
- Indels - PASS & HP

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/driver_analysis_unfiltered/cosmic_v94/01_INPUT

cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/subs/lcm/02_lcm/*txt .
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/indels/lcm_filtering/*.pindel.HP.7.vcf .

```

Convert pindel files to text files.

#####Convert pindel vcf to txt file

######Functions

```{r}

#convert putative indels vcf into text file
#input file is *.pindel.pass.hp7_threshold.qual.depth.direction.filtered.nogermline.5bp_window.nosnp_1bp_window.annot.vcf

prefilter_pindel_vcf_to_text_no_filter = function(vcf_file){
  vcf = readVcf(vcf_file)
  sample = unlist(strsplit(vcf_file, '.pindel.HP.7.vcf'))
  var.df = data.frame(Chr=as.character(seqnames(rowRanges(vcf))),
                  Position=start(ranges(rowRanges(vcf))),
                  Wildtype=as.character(ref(vcf)),
                  Mutant=as.character(unlist(CharacterList(alt(vcf)))),
                  width=width(ranges(rowRanges(vcf))),
                  Variant_read_depth=geno(vcf)$NU[,"TUMOUR"]+geno(vcf)$PU[,"TUMOUR"],
                  Total_read_depth=geno(vcf)$NR[,"TUMOUR"]+geno(vcf)$PR[,"TUMOUR"])
  var.df$VAF = var.df$Variant_read_depth / var.df$Total_read_depth
  var.df$Mutation = info(vcf)$VT
  vagrent_annots = lapply(strsplit(info(vcf)$VW, '\\|'), `length<-`, max(lengths(strsplit(info(vcf)$VW, '\\|')))) #make all elements in list the same length for subsetting purposes
  var.df$Gene = sapply(vagrent_annots, '[[', 1)
  var.df$Transcript = sapply(vagrent_annots, '[[', 3)
  var.df$Codon = sapply(vagrent_annots, '[[', 4)
  var.df$Protein = sapply(vagrent_annots, '[[', 5)
  
  conseq_list = lapply(info(vcf)$VC, function(x) if(identical(x, character(0))) NA_character_ else x)
  conseq_vec = c()
  for(i in 1:length(conseq_list)){
    conseq_vec = c(conseq_vec, paste(unlist(conseq_list[i]), collapse = ';'))
  }
  
  var.df$Consequence = conseq_vec
  
  var.df$Case_ID = substr(sample, 0, 7)
  var.df$Sample = sample
  
  
  write.table(var.df[, c("Case_ID", "Sample", "Chr", "Position", "Wildtype", "Mutant", "Total_read_depth", "Variant_read_depth", "VAF", "Mutation", "Gene",      "Transcript", "Codon", "Protein", "Consequence")], paste0(sample, '.pindel.HP.7.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
  
}

```

######Run

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/driver_analysis_unfiltered/cosmic_v94/01_INPUT/')

vcf_list = list.files('.', pattern = 'vcf')

for(i in vcf_list){
  prefilter_pindel_vcf_to_text_no_filter(i)
}

```

#####Combine subs and indels

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/driver_analysis_unfiltered/cosmic_v94/01_INPUT")

sub_files = list.files(".", "_complete_final_allinfo_4.txt")

all_subs = data.frame()
for(file in sub_files){
  sample = unlist(strsplit(file, "_complete_final_allinfo_4.txt"))
  subs = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  subs = subs[, c(1, 3:7, 9, 10)]
  #subs$sample = sample
  subs$patient = substr(sample, 0, 7)
  all_subs = rbind(all_subs, subs)
}

all_uniq_subs = all_subs[!duplicated(all_subs),]
all_uniq_func_subs = all_uniq_subs[all_uniq_subs$ExonicFunc.refGene %in% c("nonsynonymous SNV", "stopgain", "unknown"), ]

indel_files = list.files(".", ".pindel.HP.7.txt")

all_indels = data.frame()
for(file in indel_files){
  #sample = unlist(strsplit(file, ".pindel.HP.7.txt"))
  indels = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  indels = indels[, c(3, 4, 5, 6, 15, 11, 13, 14, 1)]
  all_indels = rbind(all_indels, indels)
}

all_uniq_indels = all_indels[!duplicated(all_indels),]
all_uniq_func_indels = all_uniq_indels[!is.na(all_uniq_indels$Consequence) & !all_uniq_indels$Consequence %in% c("3prime_UTR_variant", "5prime_UTR_variant", "downstream", "inframe", "intronic", "nc_variant", "splice_region", "upstream"),]

all_uniq_func_subs_2 = all_uniq_func_subs[, c(1:4, 7, 6, 8, 8, 9)]

colnames(all_uniq_func_subs_2) = colnames(all_uniq_func_indels)

all_uniq_func_muts = rbind(all_uniq_func_subs_2, all_uniq_func_indels)

write.table(all_uniq_func_muts, "all_min_filtered_uniq_fun_lcm_gct_muts_20220102.txt", col.names = T, row.names = F, sep = '\t', quote = F)

```

####Read in total list of calls

```{r}

uniq_func_ssms = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/driver_analysis_unfiltered/cosmic_v94/01_INPUT/all_min_filtered_uniq_fun_lcm_gct_muts_20220102.txt', header = T,  sep = '\t', stringsAsFactors = F)
cosmic_v94 <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/99_REFERENCE_DATA/cosmic_v94_cancer_gene_census.csv', sep = ',', header = T, stringsAsFactors = F)

table(uniq_func_ssms$Consequence)

```

####Oncogenes
```{r}

oncogenes = unlist(strsplit(cosmic_v94[grepl('oncogene', cosmic_v94$Role.in.Cancer) | grepl('Dom', cosmic_v94$Molecular.Genetics),]$Synonyms, ",")) #including all synonyms
poss_onco_ssms = uniq_func_ssms[uniq_func_ssms$Gene %in% oncogenes,]
poss_onco_ssms$put_annot = "oncogene"
poss_onco_ssms[, c(1, 2, 3, 4, 5, 6)] #no extra over the filtered calls

```

####Recessive cancer gene

```{r}

rcgenes = cosmic_v94[grepl('TSG', cosmic_v94$Role.in.Cancer) | grepl('Rec', cosmic_v94$Molecular.Genetics),]$Gene.Symbol
rcgenes = unlist(strsplit(cosmic_v94[grepl('TSG', cosmic_v94$Role.in.Cancer) | grepl('Rec', cosmic_v94$Molecular.Genetics),]$Synonyms, ",")) #including all synonyms

poss_rcg_ssms = uniq_func_ssms[uniq_func_ssms$Gene %in% rcgenes,]
poss_rcg_ssms$put_annot = "TSG"
poss_rcg_ssms[, c(1, 2, 3, 4, 5, 6)] #no extra over the filtered calls


```

##DRIVER ANALYSIS (CN and SVs)

###Copy number

```{r}

#read in manifest
manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)

#COSMIC consensus list of genes
cosmic_v94 <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/cosmic_v94_cancer_gene_census.csv', sep = ',', header = T, stringsAsFactors = F)
cosmic_v94 = cosmic_v94[!grepl(":-", cosmic_v94$Genome.Location),]

#get GRCh37 build genomic coordinates
library(biomaRt)
ensembl37 <- useEnsembl(biomart = 'genes', 
                       dataset = 'hsapiens_gene_ensembl',
                       version = 'GRCh37')
gene_coord = getBM(attributes=c('hgnc_symbol','chromosome_name','start_position','end_position'), mart = ensembl37)
gene_coord = gene_coord[gene_coord$hgnc_symbol != "",]
gene_coord = gene_coord[!grepl("^MIR", gene_coord$hgnc_symbol),]
gene_coord = gene_coord[gene_coord$chromosome_name %in% c(1:22, "X", "Y"),]
gene_coord = gene_coord[gene_coord$hgnc_symbol != "UGT2A1",] #last duplicated entry
row.names(gene_coord) = gene_coord$hgnc_symbol

#add coords to cosmic data
cosmic_v94 = cosmic_v94[cosmic_v94$Gene.Symbol %in% gene_coord$hgnc_symbol,]
nrow(cosmic_v94) #707
cosmic_v94$chr_h19 = gene_coord[cosmic_v94$Gene.Symbol,]$chromosome_name
cosmic_v94$start_h19 = gene_coord[cosmic_v94$Gene.Symbol,]$start_position
cosmic_v94$end_h19 = gene_coord[cosmic_v94$Gene.Symbol,]$end_position

oncogenes = cosmic_v94[grepl('oncogene', cosmic_v94$Role.in.Cancer) | grepl('Dom', cosmic_v94$Molecular.Genetics),]
rcgenes = cosmic_v94[grepl('TSG', cosmic_v94$Role.in.Cancer) | grepl('Rec', cosmic_v94$Molecular.Genetics),]

#read in high quality copy number calls
all_cns = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/cnas/02_high_confidence/WGS_cn_calls.txt', header = T, sep = '\t', stringsAsFactors = F)

all_cns$seg_width = all_cns$endpos - all_cns$startpos

#which oncogenes or recessive cancer genes lie across each segment
all_cns$oncogenes = NA
all_cns$rcgenes = NA

for(i in 1:nrow(all_cns)){
  contained_onco = oncogenes[oncogenes$chr_h19 == all_cns$chr[i] & oncogenes$start_h19 >= all_cns$startpos[i] & oncogenes$end_h19 <= all_cns$endpos[i], ] #the oncogene must be contained entirely within amplified segment
  if(nrow(contained_onco) > 0) all_cns$oncogenes[i] = paste(contained_onco$Gene.Symbol, collapse = ',')
  contained_rcgenes = rcgenes[rcgenes$chr == all_cns$chr[i] & rcgenes$start_h19 <= all_cns$endpos[i] & rcgenes$end_h19 >= all_cns$startpos[i], ] #the recessive cancer gene must lie within or across a breakpoint
  if(nrow(contained_rcgenes) > 0) all_cns$rcgenes[i] = paste(contained_rcgenes$Gene.Symbol, collapse = ',')
}

row.names(manifest) = manifest$Sample

all_cns$ploidy = manifest[all_cns$sample,]$Ploidy

#focal oncogene amplifications
all_cns[!is.na(all_cns$oncogenes) & all_cns$ploidy <= 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A >= 5) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A >= 5)) & all_cns$seg_width < 1000000,] #none
all_cns[!is.na(all_cns$oncogenes) & all_cns$ploidy > 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A >= 9) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A >= 9)) & all_cns$seg_width < 1000000,] #none

#are there any amplifications on larger segments containing genes of interest in GCTs?
unique(c(all_cns[!is.na(all_cns$oncogenes) & all_cns$ploidy <= 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A >= 5) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A >= 5)) & all_cns$seg_width < 10000000,]$oncogenes, all_cns[!is.na(all_cns$oncogenes) & all_cns$ploidy > 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A >= 9) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A >= 9)) & all_cns$seg_width < 10000000,]$oncogenes))
#"U2AF1", "ERC1,KDM5A", "KDR,KIT,PDGFRA", "WWTR1", "ETNK1,KRAS", "GMPS,MLF1"  
#KRAS and KIT gains on larger segments

#non-focal KRAS & KIT amplification
all_cns[!is.na(all_cns$oncogenes) & all_cns$ploidy <= 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A >= 5) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A >= 5)) & grepl("KRAS", all_cns$oncogenes) & all_cns$seg_width < 10000000,] #none
all_cns[!is.na(all_cns$oncogenes) & all_cns$ploidy > 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A >= 9) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A >= 9)) & grepl("KRAS", all_cns$oncogenes) & all_cns$seg_width < 10000000,] #>30 copies in all samples from PD46966 with high quality copy number calls
all_cns[!is.na(all_cns$oncogenes) & all_cns$ploidy <= 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A >= 5) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A >= 5)) & grepl("KIT", all_cns$oncogenes) & all_cns$seg_width < 10000000,] #none
all_cns[!is.na(all_cns$oncogenes) & all_cns$ploidy > 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A >= 9) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A >= 9)) & grepl("KIT", all_cns$oncogenes) & all_cns$seg_width < 10000000,] #all samples from PD46271

#focal rcgenes losses
all_cns[!is.na(all_cns$rcgenes) & all_cns$ploidy <= 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A == 0) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A == 0)) & all_cns$seg_width < 1000000,] #none
all_cns[!is.na(all_cns$rcgenes) & all_cns$ploidy > 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A < all_cns$ploidy - 2.7) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A < all_cns$ploidy - 2.7)) & all_cns$seg_width < 1000000,] #homozygous PTPRD deletion in all PD42036 samples
#SDHB deletion in PD50660a only fulfils criteria in minor subclone with an single intact allele preserved

#are there any deletions on larger segments containing genes of interest in GCTs?
unique(c(all_cns[!is.na(all_cns$rcgenes) & all_cns$ploidy <= 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A == 0) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A == 0)) & all_cns$seg_width < 10000000,]$rcgenes, all_cns[!is.na(all_cns$rcgenes) & all_cns$ploidy > 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A < all_cns$ploidy - 2.7) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A < all_cns$ploidy - 2.7)) & all_cns$seg_width < 10000000,]$rcgenes))
#"PTPRD" "SDHB"

all_cns[!is.na(all_cns$rcgenes) & all_cns$ploidy <= 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A == 0) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A == 0)) & all_cns$seg_width < 10000000 & grepl("PTPRD", all_cns$rcgenes),] #none
all_cns[!is.na(all_cns$rcgenes) & all_cns$ploidy > 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A < all_cns$ploidy - 2.7) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A < all_cns$ploidy - 2.7)) & all_cns$seg_width < 10000000 & grepl("PTPRD", all_cns$rcgenes),] #original case, homozygous deletion PD42036
all_cns[!is.na(all_cns$rcgenes) & all_cns$ploidy <= 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A == 0) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A == 0)) & all_cns$seg_width < 10000000 & grepl("SDHB", all_cns$rcgenes),] #none
all_cns[!is.na(all_cns$rcgenes) & all_cns$ploidy > 2.7 & ((all_cns$nMaj1_A + all_cns$nMin1_A < all_cns$ploidy - 2.7) | (!is.na(all_cns$frac2_A)) & (all_cns$nMaj2_A + all_cns$nMin2_A < all_cns$ploidy - 2.7)) & all_cns$seg_width < 10000000 & grepl("SDHB", all_cns$rcgenes),] #original case, minor clone solution A, PD50660a

```

###Structural variants

Breakpoints that interrupt recessive cancer genes or involve cancer genes activated by fusion events

```{r}

#lcm_brass_calls = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/structural_variants/final_lcm_gct_brass_calls_20210709.txt', header = T, sep = '\t', stringsAsFactors = F)
#bulk_brass_calls = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/structural_variants/final_bulk_gct_brass_calls_20210709.txt', header = T, sep = '\t', stringsAsFactors = F)

brass_calls = combined_sv = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/final_combined_gct_brass_calls_20220118.txt', header = T, sep = '\t', stringsAsFactors = F)

cosmic_v94 <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/cosmic_v94_cancer_gene_census.csv', sep = ',', header = T, stringsAsFactors = F)

rcgenes = unlist(strsplit(cosmic_v94[grepl('TSG', cosmic_v94$Role.in.Cancer) | grepl('Rec', cosmic_v94$Molecular.Genetics),]$Synonyms, ","))

write.table(brass_calls[(brass_calls$gene1 %in% rcgenes | brass_calls$gene2 %in% rcgenes),], '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/gct_sample_putative_rc_gene_breaks_20220121.txt', col.names = T, row.names = F, sep = '\t')

fusion_genes = unlist(strsplit(cosmic_v94[grepl('fusion', cosmic_v94$Role.in.Cancer),]$Synonyms, ","))

write.table(brass_calls[(brass_calls$gene1 %in% fusion_genes | brass_calls$gene2 %in% fusion_genes),], '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/gct_sample_putative_fusion_gene_involved_20220121.txt', col.names = T, row.names = F, sep = '\t')

```

Check if re-write after notebook crash generates the same putative calls.

```{r}

old_rc = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/gct_sample_putative_rc_gene_breaks_20220119.txt", header = T, sep = '\t', stringsAsFactors = F)
new_rc = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/gct_sample_putative_rc_gene_breaks_20220121.txt", header = T, sep = '\t', stringsAsFactors = F)

sum(!old_rc$Dom.Rec.COSMICv94 %in% new_rc$Dom.Rec.COSMICv94) #0

old_fus = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/gct_sample_putative_fusion_gene_involved_20220119.txt", header = T, sep = '\t', stringsAsFactors = F)
new_fus = read.table("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/gct_sample_putative_fusion_gene_involved_20220121.txt", header = T, sep = '\t', stringsAsFactors = F)

sum(!old_fus$Dom.Rec.COSMICv94 %in% new_fus$Dom.Rec.COSMICv94) #0

```


###Calculate multiplicity of driver substitutions (supplementary table 7)

Used to time their occurence relative to WGD

```{r}

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)
row.names(manifest) = manifest$Sample

ssms.merged = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/gct_vagrent_ssm_all_mutations_plus_depth20210915.txt', header = T, sep = '\t', stringsAsFactors = F)
cn_calls = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/cnas/02_high_confidence/WGS_cn_calls.txt', sep = '\t', header = T, stringsAsFactors = F)

driver_samples = manifest[manifest$Case.ID %in% c('PD45543', 'PD45544', 'PD45545') & manifest$Average.reads.per.chromosome.copy. >= 5,]$Sample

#12 25398284 #both KRAS mutations PD45543, PD45545
#14	105243048 #AKT1 mutation PD45544

driver_subs = ssms.merged[((ssms.merged$Chr == 12 & ssms.merged$Position == 25398284) | (ssms.merged$Chr == 14 & ssms.merged$Position == 105243048)) & ssms.merged$Sample %in% driver_samples,]
driver_subs$purity = manifest[driver_subs$Sample, ]$Purity

driver_subs$total_cn = NA #taken from major clone in Battenberg output
driver_subs$maj_cn = NA #taken from major clone in Battenberg output
driver_subs$min_cn = NA
for(i in 1:nrow(driver_subs)){
  sub_cn_state = cn_calls[cn_calls$sample == driver_subs$Sample[i] & cn_calls$chr == driver_subs$Chr[i] & cn_calls$startpos < driver_subs$Position[i] & cn_calls$endpos > driver_subs$Position[i], ]
  if(is.na(sub_cn_state$frac2_A)){
    driver_subs$maj_cn[i] = sub_cn_state$nMaj1_A
    driver_subs$min_cn[i] = sub_cn_state$nMin1_A
    driver_subs$total_cn[i] = driver_subs$maj_cn[i] + driver_subs$min_cn[i]
  }
  if(!is.na(sub_cn_state$frac2_A)){
    dom_subclone = names(sub_cn_state[, c('frac1_A', 'frac2_A')])[sub_cn_state[, c('frac1_A', 'frac2_A')] == max(sub_cn_state[, c('frac1_A', 'frac2_A')])]
    if(dom_subclone == 'frac1_A'){
      driver_subs$maj_cn[i] = sub_cn_state$nMaj1_A
      driver_subs$min_cn[i] = sub_cn_state$nMin1_A
      driver_subs$total_cn[i] = driver_subs$maj_cn[i] + driver_subs$min_cn[i]
    }
    if(dom_subclone == 'frac2_A'){
      driver_subs$maj_cn[i] = sub_cn_state$nMaj2_A
      driver_subs$min_cn[i] = sub_cn_state$nMin2_A
      driver_subs$total_cn[i] = driver_subs$maj_cn[i] + driver_subs$min_cn[i]
    }
  }
}

driver_subs$mut_multiplicity = driver_subs$VAF * ((driver_subs$total_cn * driver_subs$purity) + (2 * (1 - driver_subs$purity)))/ driver_subs$purity

write.table(driver_subs, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/driver_multiplicity_calc_20210920.txt', col.names = T, row.names = F, quote = F, sep = '\t')

```

##COPY NUMBER PLOTTING

###Aggregated CN plot - study cohort vs TCGA (Extended data figure 3)

Turn our battenberg files into a long segment format

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/00_final_calls/

for f in $(ls *.battenberg.subclones.txt);
do
filename=$(basename -- "$f")
sampleID="${filename%.battenberg.subclones.txt}"
python /nfs/users/nfs_t/to3/scripts/copynumber.py -i $sampleID.battenberg_subclones.txt -c $sampleID.battenberg_rho_and_psi.txt -o long_format_bb -r
done

```

- In-house data binned into 10kb chunks and the copy number across that chunk used according to the Battenberg subclones file.
- The median CN for each bin per tumour is taken across all invasive samples with a purity >40%.
- The average of this composite per tumour CN is then taken as the snapshot for the cohort as a whole.
- Autosomal chromosomes only as we have some ovarian tumours too.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/combined_CN_RNA_exp

bsub -G team294-vwork -o $PWD/log.%J -e $PWD/err.%J -q yesterday -R'select[mem>30000] rusage[mem=30000]' -M30000 -J 'gct_cn' /software/R-3.6.1/bin/Rscript lcm_gct_WG_CN.R

```

```{r}

#lcm_gct_WG_CN.R
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/combined_CN_RNA_exp/')

library(plyr)

#locate copy number files and the eligible samples
project_dir <- "/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/00_final_calls/long_format_bb"
dna_cuts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', header = T, sep = '\t', stringsAsFactors = F)
pure_dna_cuts = dna_cuts[dna_cuts$Purity > 0.4 & dna_cuts$Histology != 'GCNIS' & dna_cuts$Experimental_arm == 'LCM',] #pure, invasive only
samples <- pure_dna_cuts$Sample #all samples from all donors
donors <- unique(substr(samples,1,7)) #all donors

#set bin size
bin_size <- 10000
filter_size = 10000 #remove CN changes smaller than this
adjust = bin_size - 1 #for arranging segments

#chromosome lengths, each will be rounded up to the nearest multiple of 10,000
#chrom_sizes <- read.table('/lustre/scratch119/casm/team294rr/rs30/Testicular_project/Final_Analyses/metadata/male.hg19.chrom.sizes.tsv', header = F, sep = '\t', stringsAsFactors = F) moved location, changed 3/1/22
chrom_sizes <- read.table('/lustre/scratch119/casm/team294rr/rs30/References/GRCh37d5/chrom.sizes.txt', header = F, sep = '\t', stringsAsFactors = F)
chrom_sizes$V1 = factor(chrom_sizes$V1, levels = c(1:22, "X", "Y"))
chrom_sizes = chrom_sizes[order(chrom_sizes$V1),]
chrom_sizes[,2] = round_any(chrom_sizes[,2], bin_size, ceiling) #chunk each chromosome into 10kb bins in order to accurate characterise the CN at the beginning and ends of chromosomes, has to be ceiling to capture the ends
chrom_sizes[,3] <- cumsum(as.numeric(chrom_sizes[,2]))
#chrom_sizes = chrom_sizes[!chrom_sizes$V1 %in% c('chrM', 'chrX', 'chrY'),]
#chrom_sizes$V1 = unlist(lapply(strsplit(chrom_sizes$V1, 'chr'), '[[', 2))
chrom_sizes = chrom_sizes[!chrom_sizes$V1 %in% c("X", "Y"),]
chrom_sizes$V4 = c(0, chrom_sizes$V3[1:(length(chrom_sizes$V3) - 1)])
names(chrom_sizes) = c('chr', 'chr_length', 'cumsum_length', 'prior_wg_length')

#create array with per donor and per 10kb bin cells
cn_arr <- array(data = NA, dim = c((chrom_sizes[22, 'cumsum_length'] / bin_size), length(donors)))
colnames(cn_arr) = donors
cn_df = data.frame(cn_arr)

for(i in 1:length(donors)) {
  print(donors[i])
  donor_samples = samples[grepl(donors[i], samples)]
  donor_sample_arr = array(data = NA, dim = c((chrom_sizes[22, 'cumsum_length'] / bin_size), length(donor_samples))) #need copy number per site per sample first before averaging across donor and ultimately across all tumours
  colnames(donor_sample_arr) = donor_samples
  donor_sample_df = data.frame(donor_sample_arr)
  for(j in 1:length(donor_samples)){
    if(file.exists(paste0(project_dir,"/",donor_samples[j],"_subclones_segments.txt"))){
    print(donor_samples[j])
    segments <- read.table(paste0(project_dir,"/",donor_samples[j],"_subclones_segments.txt"), sep = "\t", stringsAsFactors = F, header = T)
    segments = segments[segments$chromosome %in% c(1:22),] #avoid sex chromosomes due to sampling of men and women
    segments = segments[order(segments$clonal_frequency, decreasing = T),] #order by cn state with greatest clonal frequency first
    segments = segments[!duplicated(segments[,1:3]),] #only keep the integer value for the copy number state with the highest clonal frequency
    segments = segments[, c('chromosome', 'start', 'end', 'copy_number')]
    names(segments) = c('chr', 'startpos', 'endpos', 'ntot')
    
  
    segments = segments[(segments$endpos - segments$startpos) > filter_size,] #minimum copy number segment size
    segments$startpos = round_any(segments$startpos, bin_size, round) #round start and end positions to nearest 10kb
    segments$endpos = round_any(segments$endpos, bin_size, round)
    segments$startcumpos <- segments$startpos #treating genome as one continuous length, generate cumsum postions
    segments$endcumpos <- segments$endpos
   
    
   for(k in 1:nrow(segments)) { #annotate segments as cumulative positions
        segments$startcumpos[k] = segments$startpos[k] + chrom_sizes$prior_wg_length[which(chrom_sizes$chr == segments$chr[k])]
        segments$endcumpos[k] <- segments$endpos[k] + chrom_sizes$prior_wg_length[which(chrom_sizes$chr == segments$chr[k])]
    }
    
    sample_vec = c()
    for(m in 1:nrow(donor_sample_df)){ #get copy number estimates per 10000bp bin
      if(nrow(segments[segments$startcumpos <= (bin_size * (m - 1)) & segments$endcumpos >= (bin_size * m),]) > 0) sample_vec = c(sample_vec, segments[segments$startcumpos <= (bin_size * (m - 1)) & segments$endcumpos >= (bin_size * m),]$ntot)
      else {
        sample_vec = c(sample_vec, NA)
      }
    }
    donor_sample_df[, donor_samples[j]] = sample_vec
    }
  }
  cn_df[, donors[i]] = apply(donor_sample_df, 1, median, na.rm = T) #median per tumour for that bin
}
  
write.table(cn_df, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/combined_CN_RNA_exp/lcm_gct_cn_10000_bp_bin_per_GCT_WG_median.txt', col.names = T, row.names = F, sep = '\t', quote = F) 

```

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/combined_CN_RNA_exp/')

cn_df = read.table('lcm_gct_cn_10000_bp_bin_per_GCT_WG_median.txt', header = T, sep = '\t', stringsAsFactors = F)
cn_df$start = c(1:nrow(cn_df))*10000 - 9999
cn_df$end = c(1:nrow(cn_df))*10000

#adjust for ploidy, to express copy number relative to the overall ploidy 
dna_cuts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', header = T, sep = '\t', stringsAsFactors = F)
pure_dna_cuts = dna_cuts[dna_cuts$Purity > 0.4 & dna_cuts$Histology != 'GCNIS' & dna_cuts$Experimental_arm == 'LCM',] #pure, invasive only
lcm_ploidy_table = aggregate(Ploidy ~ Case.ID, pure_dna_cuts, median)
row.names(lcm_ploidy_table) = lcm_ploidy_table$Case.ID
cn_df[,1:12] = sweep(cn_df[,1:12], 2, lcm_ploidy_table[colnames(cn_df[,1:12]),]$Ploidy, FUN = '-', check.margin = T) #adjust the copy number calls for each tumour by their median ploidy

#add further statistics
#cn_df$all_samples = apply(cn_df[,1:12], MARGIN = 1, FUN = median, na.rm = T) 
cn_df$all_samples = rowMeans(cn_df[,1:12], na.rm = T)
cn_df$postp_samples = rowMeans(cn_df[,c(1:3, 5:12)], na.rm = T) 
cn_df$uq_all_samples = apply(cn_df[,c(1:12)], MARGIN = 1, FUN = function(z) {quantile(z, 0.75, na.rm = T)})
cn_df$lq_all_samples = apply(cn_df[,c(1:12)], MARGIN = 1, FUN = function(z) {quantile(z, 0.25, na.rm = T)})
cn_df$uq_postp_samples = apply(cn_df[,c(1:3, 5:12)], MARGIN = 1, FUN = function(z) {quantile(z, 0.75, na.rm = T)})
cn_df$lq_postp_samples = apply(cn_df[,c(1:3, 5:12)], MARGIN = 1, FUN = function(z) {quantile(z, 0.25, na.rm = T)})

write.table(cn_df, 'lcm_gct_cn_10000_bp_bin_per_GCT_WG_median_extra_cols.txt', col.names = T, row.names = F, sep = '\t', quote = F)

```

- To derive the IQR from the TCGA data, we need to do something similar with the ASCAT segments, also from the hg19 build using the same purity cut-off
- Using latest ASCAT segmentation and ploidy estimates provided by Peter van Loo and his group for the TCGA data.
- Limited to tumours which were over 40% pure.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/analyses/combined_CN_RNA_exp

bsub -G team294-vwork -o $PWD/log.%J -e $PWD/err.%J -q yesterday -R'select[mem>30000] rusage[mem=30000]' -M30000 -J 'tcga_cn' /software/R-3.6.1/bin/Rscript tcga_gct_WG_CN.R

```

```{r}

#tcga_gct_WG_CN.R
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/combined_CN_RNA_exp/')

library(plyr)

#get metadata, including ploidy values for all eligible tumours
tcga_ploidy = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/tcga_tgct/summary.ascatTCGA.penalty70.hg19.tsv', header = T, sep = '\t', stringsAsFactors = F) #changed 4/1/21
tcga_ploidy = tcga_ploidy[(tcga_ploidy$cancer_type == 'TGCT') & (tcga_ploidy$pass == TRUE) & (tcga_ploidy$rep == TRUE) & (tcga_ploidy$purity > 0.4), ] #103 samples

#get copy segments for all tumours in a single dataframe
tcga_segment_files = list.files(path = '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/reference_datasets/tcga_tgct', pattern = 'segments.txt')
paste0(tcga_ploidy$name, '.segments.txt') %in% tcga_segment_files #all present in segment files

tcga_ascat_segments = c()
for(i in tcga_ploidy$name){
  tumour_file = read.table(paste0('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/reference_datasets/tcga_tgct/', i, '.segments.txt'), header = T, sep = '\t', stringsAsFactors = F)
  tcga_ascat_segments = rbind(tcga_ascat_segments, tumour_file)
}

names(tcga_ascat_segments) = c('tumour', 'chr', 'start', 'end', 'majA', 'minA')
tcga_ascat_segments$ntot = tcga_ascat_segments$majA + tcga_ascat_segments$minA

samples <- unique(tcga_ascat_segments$tumour)

#set bin size
bin_size <- 10000
filter_size = 10000 #remove CN changes smaller than this
adjust = bin_size - 1 #for arranging segments

#chromosome lengths, each will be rounded up to the nearest multiple of 10,000
#chrom_sizes <- read.table('/lustre/scratch119/casm/team294rr/rs30/Testicular_project/Final_Analyses/metadata/male.hg19.chrom.sizes.tsv', header = F, sep = '\t', stringsAsFactors = F) moved location, changed 4/1/22
chrom_sizes <- read.table('/lustre/scratch119/casm/team294rr/rs30/References/GRCh37d5/chrom.sizes.txt', header = F, sep = '\t', stringsAsFactors = F)
chrom_sizes$V1 = factor(chrom_sizes$V1, levels = c(1:22, "X", "Y"))
chrom_sizes = chrom_sizes[order(chrom_sizes$V1),]
chrom_sizes[,2] = round_any(chrom_sizes[,2], bin_size, ceiling) #chunk each chromosome into 10kb bins in order to accurate characterise the CN at the beginning and ends of chromosomes, has to be ceiling to capture the ends
chrom_sizes[,3] <- cumsum(as.numeric(chrom_sizes[,2]))
#chrom_sizes = chrom_sizes[!chrom_sizes$V1 %in% c('chrM', 'chrX', 'chrY'),]
#chrom_sizes$V1 = unlist(lapply(strsplit(chrom_sizes$V1, 'chr'), '[[', 2))
chrom_sizes = chrom_sizes[!chrom_sizes$V1 %in% c("X", "Y"),]
chrom_sizes$V4 = c(0, chrom_sizes$V3[1:(length(chrom_sizes$V3) - 1)])
names(chrom_sizes) = c('chr', 'chr_length', 'cumsum_length', 'prior_wg_length')

tcga_ascat_segments = tcga_ascat_segments[(tcga_ascat_segments$end - tcga_ascat_segments$start) > filter_size,]
tcga_ascat_segments$mod_start = round_any(tcga_ascat_segments$start, bin_size, round) #fit genome to bins of 10,000bp
tcga_ascat_segments$mod_end = round_any(tcga_ascat_segments$end, bin_size, round)
tcga_ascat_segments = tcga_ascat_segments[tcga_ascat_segments$chr %in% c(1:22),]

tcga_ascat_segments$startcumpos = tcga_ascat_segments$mod_start
tcga_ascat_segments$endcumpos = tcga_ascat_segments$mod_end

for(j in 1:nrow(tcga_ascat_segments)) { #annotate segments as cumulative positions
        tcga_ascat_segments$startcumpos[j] = tcga_ascat_segments$mod_start[j] + chrom_sizes$prior_wg_length[which(chrom_sizes$chr == tcga_ascat_segments$chr[j])]
        tcga_ascat_segments$endcumpos[j] <- tcga_ascat_segments$mod_end[j] + chrom_sizes$prior_wg_length[which(chrom_sizes$chr == tcga_ascat_segments$chr[j])]
}

#create array per sample and per 10kb bin cells
tcga_cn_arr <- array(data = NA, dim = c((chrom_sizes[22, 'cumsum_length'] / bin_size), length(samples)))
colnames(tcga_cn_arr) = samples
tcga_cn_df = data.frame(tcga_cn_arr)
colnames(tcga_cn_df) = gsub('\\.', '-', colnames(tcga_cn_df))

for(k in 1:length(samples)){
  print(samples[k])
  for(m in 1:nrow(tcga_cn_df)){ #get copy number estimates per 10000bp bin
      if(nrow(tcga_ascat_segments[tcga_ascat_segments$startcumpos <= (10000 * (m - 1)) & tcga_ascat_segments$endcumpos >= (10000 * m) & tcga_ascat_segments$tumour == samples[k],]) > 0){
        tcga_cn_df[m, samples[k]] = tcga_ascat_segments[tcga_ascat_segments$startcumpos <= (10000 * (m - 1)) & tcga_ascat_segments$endcumpos >= (10000 * m) & tcga_ascat_segments$tumour == samples[k],]$ntot
    }
  }
}

write.table(tcga_cn_df, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/combined_CN_RNA_exp/tcga_cn_10000_bp_GCT_WG_median.txt', col.names = T, row.names = F, sep = '\t', quote = F) 

```


```{r}

tcga_cn_df = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/combined_CN_RNA_exp/tcga_cn_10000_bp_GCT_WG_median.txt', header = T, sep = '\t', stringsAsFactors = F, check.names = F)

#read in ploidy values
#identify samples with RNA seq data
tcga_ploidy = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/reference_datasets/tcga_tgct/summary.ascatTCGA.penalty70.hg19.tsv', header = T, sep = '\t', stringsAsFactors = F)
tcga_ploidy = tcga_ploidy[(tcga_ploidy$cancer_type == 'TGCT') & (tcga_ploidy$pass == TRUE) & (tcga_ploidy$rep == TRUE) & (tcga_ploidy$purity > 0.4), ] #103 samples
row.names(tcga_ploidy) = tcga_ploidy$name

#substract the ploidy from the counts for each tumour
tcga_cn_df = sweep(tcga_cn_df, 2, tcga_ploidy[colnames(tcga_cn_df),]$ploidy, FUN = '-', check.margin = T)
#tcga_cn_df$all_samples = apply(tcga_cn_df[, 1:103], 1, median, na.rm = T)
tcga_cn_df$all_samples = rowMeans(tcga_cn_df[, 1:103], na.rm = T)
tcga_cn_df$start = c(1:nrow(tcga_cn_df))*10000 - 9999
tcga_cn_df$end = c(1:nrow(tcga_cn_df))*10000
tcga_cn_df$uq = apply(tcga_cn_df[,1:103], MARGIN = 1, FUN = function(z) {quantile(z, 0.75, na.rm = T)})
tcga_cn_df$lq = apply(tcga_cn_df[,1:103], MARGIN = 1, FUN = function(z) {quantile(z, 0.25, na.rm = T)})

write.table(tcga_cn_df, 'tcga_cn_10000_bp_GCT_WG_median_extra_cols.txt', col.names = T, row.names = F, sep = '\t', quote = F)

```

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/combined_CN_RNA_exp/')

library(ggplot2)
library(ggpubr)
library(plyr)

cn_df = read.table('lcm_gct_cn_10000_bp_bin_per_GCT_WG_median_extra_cols.txt', header = T, sep = '\t', stringsAsFactors = F, check.names = F)
tcga_cn_df = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/combined_CN_RNA_exp/tcga_cn_10000_bp_GCT_WG_median_extra_cols.txt', header = T, sep = '\t', stringsAsFactors = F, check.names = F)

#chromosome lengths, each will be rounded up to the nearest multiple of 10,000
#chrom_sizes <- read.table('/lustre/scratch119/casm/team294rr/rs30/Testicular_project/Final_Analyses/metadata/male.hg19.chrom.sizes.tsv', header = F, sep = '\t', stringsAsFactors = F) moved location, changed 4/1/22
chrom_sizes <- read.table('/lustre/scratch119/casm/team294rr/rs30/References/GRCh37d5/chrom.sizes.txt', header = F, sep = '\t', stringsAsFactors = F)
chrom_sizes$V1 = factor(chrom_sizes$V1, levels = c(1:22, "X", "Y"))
chrom_sizes = chrom_sizes[order(chrom_sizes$V1),]
chrom_sizes[,2] = round_any(chrom_sizes[,2], bin_size, ceiling) #chunk each chromosome into 10kb bins in order to accurate characterise the CN at the beginning and ends of chromosomes, has to be ceiling to capture the ends
chrom_sizes[,3] <- cumsum(as.numeric(chrom_sizes[,2]))
#chrom_sizes = chrom_sizes[!chrom_sizes$V1 %in% c('chrM', 'chrX', 'chrY'),]
#chrom_sizes$V1 = unlist(lapply(strsplit(chrom_sizes$V1, 'chr'), '[[', 2))
chrom_sizes = chrom_sizes[!chrom_sizes$V1 %in% c("X", "Y"),]
chrom_sizes$V4 = c(0, chrom_sizes$V3[1:(length(chrom_sizes$V3) - 1)])
names(chrom_sizes) = c('chr', 'chr_length', 'cumsum_length', 'prior_wg_length')

#get genes loci, examine overall hotspots
data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F)
row.names(data) = data$Geneid
data$Start = as.numeric(unlist(lapply(strsplit(data$Start, ';'), '[[', 1)))
data$End = as.numeric(unlist(lapply(strsplit(data$End, ';'), '[[', 1)))
data$startcumpos = data$Start
data$endcumpos = data$End
data = data[data$Chr %in% c(1:22),] #only autosomal chromosomes

for(k in 1:nrow(data)) { #annotate segments as cumulative positions
    if(data$Chr[k] != "1"){
        data$startcumpos[k] = data$Start[k] + chrom_sizes$prior_wg_length[which(chrom_sizes$chr == data$Chr[k])]
        data$endcumpos[k] = data$End[k] + chrom_sizes$prior_wg_length[which(chrom_sizes$chr == data$Chr[k])]
      }
}
data = data[, c('Geneid', 'Chr', 'Start', 'End', 'startcumpos', 'endcumpos')]
genes_to_plot = data[c('KRAS', 'KIT', 'CCND2'),]
genes_to_plot$plot_pos = NA
for(i in 1:nrow(genes_to_plot)){
  genes_to_plot$plot_pos[i] = cn_df[cn_df$start < genes_to_plot$startcumpos[i] & cn_df$end > genes_to_plot$startcumpos[i], ]$postp_samples
}

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/lcm_gct_v_tcga_cn_20210910.pdf', height = 3, width = 12, useDingbats = F)
ggplot() + 
  scale_x_continuous(expand = c(0,0), breaks = c(chrom_sizes$cumsum_length - (chrom_sizes$chr_length / 2)), labels = c(1:22), limits = c(0, 2881130000)) +
  geom_ribbon(tcga_cn_df, mapping = aes(x = start, ymin = lq, ymax = uq), alpha = 0.4, colour = 'transparent', fill = '#999999') +
  geom_point(cn_df, mapping = aes(x = start, y = postp_samples), size = 0.5) + 
  geom_vline(xintercept = chrom_sizes$prior_wg_length) + labs(y = 'copy number - ploidy', x = 'chr') + theme_pubr() + theme(panel.border = element_rect(fill = NA)) + coord_cartesian(expand = F) + geom_hline(yintercept = 0, col = '#999999', lty = 2) + ylim(-2, 8) + geom_text(genes_to_plot, mapping = aes(x = startcumpos, y = plot_pos + 0.5), label = c('KRAS', 'KIT', 'CCND2'))
dev.off()
                                                                                                                                               
```

###Genes in highly amplified 12 segment

```{r}

#chromosome lengths, each will be rounded up to the nearest multiple of 10,000
#chrom_sizes <- read.table('/lustre/scratch119/casm/team294rr/rs30/Testicular_project/Final_Analyses/metadata/male.hg19.chrom.sizes.tsv', header = F, sep = '\t', stringsAsFactors = F) moved location, changed 4/1/22
chrom_sizes <- read.table('/lustre/scratch119/casm/team294rr/rs30/References/GRCh37d5/chrom.sizes.txt', header = F, sep = '\t', stringsAsFactors = F)
chrom_sizes$V1 = factor(chrom_sizes$V1, levels = c(1:22, "X", "Y"))
chrom_sizes = chrom_sizes[order(chrom_sizes$V1),]
chrom_sizes[,2] = round_any(chrom_sizes[,2], bin_size, ceiling) #chunk each chromosome into 10kb bins in order to accurate characterise the CN at the beginning and ends of chromosomes, has to be ceiling to capture the ends
chrom_sizes[,3] <- cumsum(as.numeric(chrom_sizes[,2]))
#chrom_sizes = chrom_sizes[!chrom_sizes$V1 %in% c('chrM', 'chrX', 'chrY'),]
#chrom_sizes$V1 = unlist(lapply(strsplit(chrom_sizes$V1, 'chr'), '[[', 2))
chrom_sizes = chrom_sizes[!chrom_sizes$V1 %in% c("X", "Y"),]
chrom_sizes$V4 = c(0, chrom_sizes$V3[1:(length(chrom_sizes$V3) - 1)])
names(chrom_sizes) = c('chr', 'chr_length', 'cumsum_length', 'prior_wg_length')

#chr12 is 1950970000 - 2084830000

cn_df = read.table('lcm_gct_cn_10000_bp_bin_per_GCT_WG_median_extra_cols.txt', header = T, sep = '\t', stringsAsFactors = F, check.names = F)

gain12 = cn_df[cn_df$start >= 1950970000 & cn_df$end <= 2084830000 & !is.na(cn_df$postp_samples) & cn_df$postp_samples > 4,]
gain12$start12 = gain12$start - 1950970000
gain12$end12 = gain12$end - 1950970000

#get genes loci, examine overall hotspots
data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F)
row.names(data) = data$Geneid
data$Start = as.numeric(unlist(lapply(strsplit(data$Start, ';'), '[[', 1)))
data$End = as.numeric(unlist(lapply(strsplit(data$End, ';'), '[[', 1)))
data$startcumpos = data$Start
data$endcumpos = data$End
data = data[data$Chr %in% c(1:22),] #only autosomal chromosomes

for(k in 1:nrow(data)) { #annotate segments as cumulative positions
    if(data$Chr[k] != "1"){
        data$startcumpos[k] = data$Start[k] + chrom_sizes$prior_wg_length[which(chrom_sizes$chr == data$Chr[k])]
        data$endcumpos[k] = data$End[k] + chrom_sizes$prior_wg_length[which(chrom_sizes$chr == data$Chr[k])]
      }
}
data = data[, c('Geneid', 'Chr', 'Start', 'End', 'startcumpos', 'endcumpos')]

paste(data[data$startcumpos >= min(gain12$start) & data$endcumpos <= max(gain12$end), ]$Geneid, collapse = ", ")

```


##STRUCTURAL VARIANT PLOTTING

###Chromothripsis detection

Generate long format CN calls

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/cnas/01_battenberg

for f in $(ls *_subclones.txt);
do
filename=$(basename -- "$f")
sample="${filename%_subclones.txt}"
python /nfs/users/nfs_t/to3/scripts/copynumber.py -i ${sample}_subclones.txt -c ${sample}_rho_and_psi.txt -o . -r
done

mkdir long_format_bb

mv *_subclones_segments.txt long_format_bb
rm *_subclones_baflogr.txt

```

We need the largest subclonal population and to merge contiguous segments that share the same copy number

```{r}

library(GenomicRanges)

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/cnas/01_battenberg/long_format_bb')

#per sample bb calls
bb_files = list.files('.', '_subclones_segments.txt')


for(file in bb_files){
  sample = unlist(strsplit(file, '_subclones_segments.txt')) #sample name
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F) # read in data
  
  if(nrow(data[duplicated(data[,c(1:3)]),]) > 0){ #if multiple copy number subclones are present
  remove_df = c() #empty vector that will ultimately contain the all the smaller subclonal copy number configurations of any segment
  
  for(i in 1:nrow(data[duplicated(data[,c(1:3)]),])){
    dup_rows = data[data$chromosome == data[duplicated(data[,c(1:3)]),][i,]$chromosome & data$start == data[duplicated(data[,c(1:3)]),][i,]$start & data$end == data[duplicated(data[,c(1:3)]),][i,]$end,] #identify all configurations of a given segment reported
    remove_df = rbind(remove_df, dup_rows[dup_rows$clonal_frequency == min(dup_rows$clonal_frequency),]) #flag the versions with smaller subclonal contributions
  }
  
  sample_majA_calls = data[!row.names(data) %in% row.names(remove_df),] #leave only the clonal or largest subclone copy number calls
  
  }
  
  sample_majA_calls = sample_majA_calls[, c(1:4)]
  colnames(sample_majA_calls) = c("chr", "start", "end", "total_cn")
  
  #merge contiguous segments with the same copy number, using the code provided with shatterseek
  
  d = sample_majA_calls
  dd <- d
  if(length(dd$total_cn[dd$total_cn == 0]) > 0) dd$total_cn[dd$total_cn == 0] <- 150000
  dd$total_cn[is.na(dd$total_cn)] <- 0
  dd <- as(dd,"GRanges")
  cov <- coverage(dd, weight = dd$total_cn)
  dd1 <- as(cov,"GRanges")
  dd1 <- as.data.frame(dd1)
  dd1 <- dd1[dd1$score !=0,]
  dd1 = dd1[,c(1,2,3,6)]
  names(dd1) <- names(d)[1:4]
  dd1$total_cn[dd1$total_cn == 150000] <- 0
  d = dd1; rm(dd)
  d$chr = as.character(d$chr)
  
  chr_df = c() #merge adjacent copy number segments with the same total cn state
  for(i in c(1:22, 'X')){
    cumsum_vec = cumsum(rle(d[d$chr == i,]$total_cn)$lengths)
    cumsum_vec = c(0, cumsum_vec[1:length(cumsum_vec)])
    for(j in 2:length(cumsum_vec)){
      chr_df = rbind(chr_df, c(d[d$chr == i,][cumsum_vec[j],]$chr, d[d$chr == i,][cumsum_vec[j-1] + 1,]$start, d[d$chr == i,][cumsum_vec[j],]$end, d[d$chr == i,][cumsum_vec[j],]$total_cn))  
      }
  }
  
  chr_df = data.frame(chr_df)
  colnames(chr_df) = colnames(d)
  
  write.table(chr_df, paste0('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/shatterseek/', sample, '_cn_input.txt'), col.names = T, row.names = F, sep = '\t', quote = F)
  
}

```

Generate per sample SV calls

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA')

combined_sv = read.table('final_combined_gct_brass_calls_20220118.txt', header = T, sep = '\t', stringsAsFactors = F)

for(sample in unique(combined_sv$sample)){
  
  sample_data = combined_sv[combined_sv$sample == sample,]
  sample_data = sample_data[, c('chr1', 'start1', 'chr2', 'end2', 'svclass', 'strand1', 'strand2')]
  sample_data$brass_strand = paste0(sample_data$strand1, '/', sample_data$strand2)
  
  if(nrow(sample_data[sample_data$svclass == 'translocation',]) > 0) sample_data[sample_data$svclass == 'translocation',]$svclass = 'TRA'
  if(nrow(sample_data[sample_data$svclass == 'tandem-duplication',]) > 0) sample_data[sample_data$svclass == 'tandem-duplication',]$svclass = 'DUP'
  if(nrow(sample_data[sample_data$svclass == 'deletion',]) > 0) sample_data[sample_data$svclass == 'deletion',]$svclass = 'DEL'
  if(nrow(sample_data[sample_data$svclass == 'inversion' & sample_data$strand1 == '+' & sample_data$strand2 == '-',]) > 0) sample_data[sample_data$svclass == 'inversion' & sample_data$strand1 == '+' & sample_data$strand2 == '-',]$svclass = 'h2hINV'
  if(nrow(sample_data[sample_data$svclass == 'inversion' & sample_data$strand1 == '-' & sample_data$strand2 == '+',]) > 0) sample_data[sample_data$svclass == 'inversion' & sample_data$strand1 == '-' & sample_data$strand2 == '+',]$svclass = 't2tINV'
  
  sample_data$pcawg_strand = NA
  
  for(i in 1:nrow(sample_data)){
    if(nrow(sample_data[sample_data$brass_strand == '+/-',]) > 0) sample_data[sample_data$brass_strand == '+/-',]$pcawg_strand = '+/+'
    if(nrow(sample_data[sample_data$brass_strand == '-/+',]) > 0) sample_data[sample_data$brass_strand == '-/+',]$pcawg_strand = '-/-'
    if(nrow(sample_data[sample_data$brass_strand == '+/+',]) > 0) sample_data[sample_data$brass_strand == '+/+',]$pcawg_strand = '+/-'
    if(nrow(sample_data[sample_data$brass_strand == '-/-',]) > 0) sample_data[sample_data$brass_strand == '-/-',]$pcawg_strand = '-/+'
  }
  
  sample_data$pcawg_strand1 = unlist(strsplit(sample_data$pcawg_strand, '/'))[c(T, F)]
  sample_data$pcawg_strand2 = unlist(strsplit(sample_data$pcawg_strand, '/'))[c(F, T)]
  
  sample_data = sample_data[,c('chr1', 'start1', 'chr2', 'end2', 'svclass', 'pcawg_strand1', 'pcawg_strand2')]
  names(sample_data) = c('chrom1', 'pos1', 'chrom2', 'pos2', 'SVtype', 'strand1', 'strand2')
  
  write.table(sample_data, paste0('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/shatterseek/', sample, '_sv_input.txt'), col.names = T, row.names = F, sep = '\t', quote = F)
  
}

```

First we need to detect chromothripsis

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/shatterseek')

sv_files = list.files('.', '_sv_input.txt') #only samples that have structural variants called in them, 134/138 samples

library(ShatterSeek)
library(gridExtra)
library(cowplot)

chr_df = c() #empty vector, will become dataframe containing all putative chromothripsis calls

for(i in sv_files){
  
  sample = unlist(strsplit(i, '_sv_input.txt')) #extract sample name
  print(sample)
  
  cn_input = read.table(paste0(sample, '_cn_input.txt'), header = T, sep = '\t', stringsAsFactors = F) # read in bb data
  
  sv_input = read.table(paste0(sample, '_sv_input.txt'), header = T, sep = '\t', stringsAsFactors = F) #read in structural variant data
  sv_input = sv_input[sv_input$chrom1 != 'Y' & sv_input$chrom2 != 'Y',] #remove any SVs involving Y chromosome
  
  CN_data <- CNVsegs(chrom = as.character(cn_input$chr), 
                     start = cn_input$start,
                     end = cn_input$end, 
                     total_cn = cn_input$total_cn)
  
  SV_data = SVs(chrom1 = as.character(sv_input$chrom1), 
                pos1 = as.numeric(sv_input$pos1),
                chrom2 = as.character(sv_input$chrom2),
                pos2 = as.numeric(sv_input$pos2),
                SVtype = as.character(sv_input$SVtype),
                strand1 = as.character(sv_input$strand1),
                strand2 = as.character(sv_input$strand2))
  
  chromothripsis <- shatterseek(SV.sample=SV_data, seg.sample=CN_data) #run shatterseek
  
  #run BH correction for p vals as per authors advice
  chromothripsis@chromSummary$qval_fragment_joins = p.adjust(chromothripsis@chromSummary$pval_fragment_joins, method = 'BH') #fragment joins test
  chromothripsis@chromSummary$chr_breakpoint_enrichment_qval = p.adjust(chromothripsis@chromSummary$chr_breakpoint_enrichment, method = 'BH') #chromosomal breakpoint enrichment
  chromothripsis@chromSummary$qval_exp_cluster = p.adjust(chromothripsis@chromSummary$pval_exp_cluster, method = 'BH') #exponential distribution of breakpoints 
  
  #add sample name and column for confidence of chromothripsis call
  chromothripsis@chromSummary$sample = sample
  chromothripsis@chromSummary$chromothripsis_call = NA
  
  #high confidence, limited to a single chromosome
  if(nrow(chromothripsis@chromSummary[(chromothripsis@chromSummary$clusterSize_including_TRA - chromothripsis@chromSummary$number_TRA) >= 6 & (!is.na(chromothripsis@chromSummary$max_number_oscillating_CN_segments_2_states) & chromothripsis@chromSummary$max_number_oscillating_CN_segments_2_states >= 7) & chromothripsis@chromSummary$qval_fragment_joins > 0.2 & (chromothripsis@chromSummary$chr_breakpoint_enrichment_qval < 0.2 | chromothripsis@chromSummary$qval_exp_cluster < 0.2),]) > 0) chromothripsis@chromSummary[(chromothripsis@chromSummary$clusterSize_including_TRA - chromothripsis@chromSummary$number_TRA) >= 6 & (!is.na(chromothripsis@chromSummary$max_number_oscillating_CN_segments_2_states) & chromothripsis@chromSummary$max_number_oscillating_CN_segments_2_states >= 7) & chromothripsis@chromSummary$qval_fragment_joins > 0.2 & (chromothripsis@chromSummary$chr_breakpoint_enrichment_qval < 0.2 | chromothripsis@chromSummary$qval_exp_cluster < 0.2),]$chromothripsis_call = "Hi_conf_single" #6 or more interleaved intrachromosomal SVs, 7 or more contiguous segments oscillating between 2 CN states, passes the fragments join test (fits the polynomial distribution) and either the chromosomal enrichment or exponential distribution of breakpoints tests
  if(nrow(chromothripsis@chromSummary[(chromothripsis@chromSummary$clusterSize_including_TRA - chromothripsis@chromSummary$number_TRA) >= 3 & chromothripsis@chromSummary$number_TRA >= 4 & (!is.na(chromothripsis@chromSummary$max_number_oscillating_CN_segments_2_states) & chromothripsis@chromSummary$max_number_oscillating_CN_segments_2_states >= 7) & chromothripsis@chromSummary$qval_fragment_joins > 0.2,]) > 0) chromothripsis@chromSummary[(chromothripsis@chromSummary$clusterSize_including_TRA - chromothripsis@chromSummary$number_TRA) >= 3 & chromothripsis@chromSummary$number_TRA >= 4 & (!is.na(chromothripsis@chromSummary$max_number_oscillating_CN_segments_2_states) & chromothripsis@chromSummary$max_number_oscillating_CN_segments_2_states >= 7) & chromothripsis@chromSummary$qval_fragment_joins > 0.2,]$chromothripsis_call = "Hi_conf_multi" #3 or more interleaved intrachromosomal SVs, 4 or more interchromosomal SVs, 7 contiguous segments oscillating between 2 CN states and passes the fragments join test (fits the polynomial distribution)
  if(nrow(chromothripsis@chromSummary[(chromothripsis@chromSummary$clusterSize_including_TRA - chromothripsis@chromSummary$number_TRA) >= 6 & (!is.na(chromothripsis@chromSummary$max_number_oscillating_CN_segments_2_states) & chromothripsis@chromSummary$max_number_oscillating_CN_segments_2_states >= 4 & chromothripsis@chromSummary$max_number_oscillating_CN_segments_2_states < 7) & chromothripsis@chromSummary$qval_fragment_joins > 0.2 & (chromothripsis@chromSummary$chr_breakpoint_enrichment_qval < 0.2 | chromothripsis@chromSummary$qval_exp_cluster < 0.2),]) > 0) chromothripsis@chromSummary[(chromothripsis@chromSummary$clusterSize_including_TRA - chromothripsis@chromSummary$number_TRA) >= 6 & (!is.na(chromothripsis@chromSummary$max_number_oscillating_CN_segments_2_states) & chromothripsis@chromSummary$max_number_oscillating_CN_segments_2_states >= 4 & chromothripsis@chromSummary$max_number_oscillating_CN_segments_2_states < 7) & chromothripsis@chromSummary$qval_fragment_joins > 0.2 & (chromothripsis@chromSummary$chr_breakpoint_enrichment_qval < 0.2 | chromothripsis@chromSummary$qval_exp_cluster < 0.2),]$chromothripsis_call = "Lo_conf" #6 or more interleaved intrachromosomal SVs, 3-6 contiguous segments oscillating between 2 CN states, passes the fragments join test (fits the polynomial distribution) and either the chromosomal enrichment or exponential distribution of breakpoints tests
  
  chr_df = rbind(chr_df, chromothripsis@chromSummary[!is.na(chromothripsis@chromSummary$chromothripsis_call),])
  
}

write.table(chr_df, 'gct_putative_chromothripsis_calls.txt', col.names = T, row.names = F, sep = '\t', quote = F)

source('/nfs/users/nfs_t/to3/scripts/plot_chromothripsis_to3_edit.R') #edited plotting function for chromothripsis

for(sample in unique(chr_df$sample)){

  print(sample)
  
  cn_input = read.table(paste0(sample, '_cn_input.txt'), header = T, sep = '\t', stringsAsFactors = F) # read in bb data
  
  sv_input = read.table(paste0(sample, '_sv_input.txt'), header = T, sep = '\t', stringsAsFactors = F) #read in structural variant data
  sv_input = sv_input[sv_input$chrom1 != 'Y' & sv_input$chrom2 != 'Y',] #remove any SVs involving Y chromosome
  
  CN_data <- CNVsegs(chrom = as.character(cn_input$chr), 
                     start = cn_input$start,
                     end = cn_input$end, 
                     total_cn = cn_input$total_cn)
  
  SV_data = SVs(chrom1 = as.character(sv_input$chrom1), 
                pos1 = as.numeric(sv_input$pos1),
                chrom2 = as.character(sv_input$chrom2),
                pos2 = as.numeric(sv_input$pos2),
                SVtype = as.character(sv_input$SVtype),
                strand1 = as.character(sv_input$strand1),
                strand2 = as.character(sv_input$strand2))
  
  chromothripsis <- shatterseek(SV.sample=SV_data, seg.sample=CN_data) #run shatterseek
  
  for(j in chr_df[chr_df$sample == sample,]$chrom){
    
    plots_chr = plot_chromothripsis_to3_edit(ShatterSeek_output = chromothripsis, chr = paste0('chr',j), sample_name = sample) 
    plot_chr = arrangeGrob(plots_chr[[1]],
plots_chr[[2]], plots_chr[[3]], plots_chr[[4]],
nrow=4,ncol=1,heights=c(0.2,.4,.4,.4))
    
    title <- ggdraw() + 
  draw_label(
    paste0(sample),
    fontface = 'bold',
    x = 0,
    hjust = 0
  ) +
  theme(
    # add margin on the left of the drawing canvas,
    # so title is aligned with left edge of first plot
    plot.margin = margin(0, 0, 0, 7)
  )
  
 final_plot = plot_grid(title, plot_chr, ncol = 1, rel_heights = c(0.1, 1))
 
 save_plot(paste0(sample, '_chr', j, '_shatterseek_plot_to3_edit.pdf'), final_plot, base_height = 7, base_width = 5)
  }

}

```

Having reviewed the calls and the other samples from the same tumour, the following follow a chromothripsis-like pattern:

the following look good:
- PD43299a_lo0005 - 17    445003  81048871
- PD45545a_lo0004 - 3   2946853 195566011
- PD45545a_lo0005 - 3   2946853 195566011 (missed by Shattersheek as low purity means copy number changes not detected although matching rearrangements still found)
- PD45545a_lo0007 - 3   2946853 195566011
- PD50657a - 9  40898959 140813552 (possibly two cth events in one region)
- PD50657a - 4    165407 190600589 (cth-like)
- PD50659a - 10   5286480  72671352 (cth-like, rather than true cth perhaps)

####Add to manifest

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/shatterseek')

chr_df = read.table('gct_putative_chromothripsis_calls.txt', header = T, sep = '\t', stringsAsFactors = F)
filtered_chr_df = chr_df[c(1:3, 6:8), c('chrom', 'start', 'end', 'sample')] #after review with isidro
filtered_chr_df = rbind(filtered_chr_df, c(3, 2946853, 195566011, 'PD45545a_lo0005')) #missed in this sample

prior = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)
manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)

manifest$chromothripsis_call = 0
manifest$chromothripsis_coord = NA
for(i in 1:nrow(manifest)){
  if(manifest$Sample[i] %in% unique(filtered_chr_df$sample)) manifest$chromothripsis_call[i] = nrow(filtered_chr_df[filtered_chr_df$sample == manifest$Sample[i],])
  if(manifest$Sample[i] %in% unique(filtered_chr_df$sample)) manifest$chromothripsis_coord[i] = paste(paste0(filtered_chr_df[filtered_chr_df$sample == manifest$Sample[i],]$chrom, ':', filtered_chr_df[filtered_chr_df$sample == manifest$Sample[i],]$start, '-', filtered_chr_df[filtered_chr_df$sample == manifest$Sample[i],]$end),collapse = ', ')
}

#manifest[manifest$chromothripsis_call != prior$chromothripsis_call,] 0
#manifest[manifest$chromothripsis_coord != prior$chromothripsis_coord & !is.na(manifest$chromothripsis_coord),] 0

write.table(manifest, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', col.names = T, row.names = F, sep = '\t', quote = F)

```

###Plot SV burden (Extended data figure 5)
 
```{r}

library(reshape2)
library(ggplot2)
library(ggpubr)
library(ggbeeswarm)

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)

sv_df = manifest[, c('Sample', 'Case.ID', 'Organ', 'Diagnosis', 'Histology', 'Age', 'dup_sv_count', 'del_sv_count', 'trans_sv_count', 'inv_sv_count', 'rt_sv_count', 'chromothripsis_call')]
sv_df$not_rt_svs = sv_df$dup_sv_count + sv_df$del_sv_count + sv_df$trans_sv_count + sv_df$inv_sv_count

not_rt_svs_per_tumour = aggregate(not_rt_svs ~ Case.ID, sv_df[sv_df$chromothripsis_call == 0 & sv_df$Histology != "GCNIS",], mean) #remove cases with chromothripsis which will skew SV calls despite likely being a single event, plus in situ disease for burden comparisons

rt_svs_per_tumour = aggregate(rt_sv_count ~ Case.ID, sv_df[sv_df$Histology != "GCNIS",], mean) #remove in situ disease

merged_per_tumour = merge(not_rt_svs_per_tumour, rt_svs_per_tumour, by = 'Case.ID', all = T)

non_dup_rows = sv_df[!duplicated(sv_df$Case.ID),]
row.names(non_dup_rows) = non_dup_rows$Case.ID

#annotate calls
merged_per_tumour$Organ = non_dup_rows[merged_per_tumour$Case.ID,]$Organ
merged_per_tumour$Age = non_dup_rows[merged_per_tumour$Case.ID,]$Age
merged_per_tumour$Diagnosis = non_dup_rows[merged_per_tumour$Case.ID,]$Diagnosis

merged_per_tumour$age_group = NA

merged_per_tumour[merged_per_tumour$Age <= 12,]$age_group = 'Paediatric'
merged_per_tumour[merged_per_tumour$Age > 12 & merged_per_tumour$Age < 18,]$age_group = 'Adolescent'
merged_per_tumour[merged_per_tumour$Age >= 18,]$age_group = 'Adult'

merged_per_tumour$age_group = factor(merged_per_tumour$age_group, levels = c('Paediatric', 'Adolescent', 'Adult'))

aggregate(not_rt_svs ~ age_group, merged_per_tumour[merged_per_tumour$age_group %in% c("Paediatric", "Adult"),], median)
#age_group not_rt_svs
#Paediatric   36.50000
#Adult   23.33333

aggregate(rt_sv_count ~ age_group, merged_per_tumour[merged_per_tumour$age_group %in% c("Paediatric", "Adult"),], median)
#age_group rt_sv_count
#Paediatric        16.5
#Adult         0.0

ggplot(merged_per_tumour) + 
  geom_boxplot(mapping = aes(x = age_group, y = not_rt_svs)) + 
  geom_point(mapping = aes(x = age_group, y = not_rt_svs, col = Diagnosis)) + theme_pubr() + xlab('Age group') + ylab('Non-RT rearrangements') + scale_color_manual(values = c('grey80', 'grey50', 'grey20'))

table(merged_per_tumour[!is.na(merged_per_tumour$not_rt_svs),]$age_group)
#Paediatric Adolescent      Adult 
#6          3          7 
table(merged_per_tumour[!is.na(merged_per_tumour$rt_sv_count),]$age_group)
#Paediatric Adolescent      Adult 
#        8          4          7

#plot
pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/06_FIGURES/not_rt_svs_plot_20220504.pdf', height = 3,  width = 3, useDingbats = F)
ggplot(merged_per_tumour) +
  geom_quasirandom(mapping = aes(x = age_group, y = not_rt_svs, col = Diagnosis, shape = Organ), size = 3) +
  geom_pointrange(mapping = aes(x = age_group, y = not_rt_svs),
                  stat = "summary", shape=19, 
                  fun.min = function(z) {quantile(z,0.25)},
                  fun.max = function(z) {quantile(z,0.75)},
                  fun = median, fill="black") +
  theme_pubr() + 
  xlab('') + 
  ylab('Non-RT rearrangements') + 
  scale_color_manual(values = c('grey80', 'grey50', 'grey20')) +
  scale_x_discrete(expand = c(0,0.5)) +
  scale_y_continuous(limits = c(0, 120), expand = c(0.01, 0.01)) +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.8, hjust = 0.9), legend.position = 'none')
dev.off()

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/06_FIGURES/rt_svs_plot_20220504.pdf', height = 3,  width = 3, useDingbats = F)
ggplot(merged_per_tumour) +
  geom_quasirandom(mapping = aes(x = age_group, y = rt_sv_count, col = Diagnosis, shape = Organ), size = 3) +
  geom_pointrange(mapping = aes(x = age_group, y = rt_sv_count),
                  stat = "summary", shape=19, 
                  fun.min = function(z) {quantile(z,0.25)},
                  fun.max = function(z) {quantile(z,0.75)},
                  fun = median, fill="black") +
  theme_pubr() + 
  xlab('') + 
  ylab('Retrotransposition events') + 
  scale_color_manual(values = c('grey80', 'grey50', 'grey20')) +
  scale_x_discrete(expand = c(0,0.5)) +
  scale_y_continuous(limits = c(0, 80), expand = c(0.01, 0.01)) +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.8, hjust = 0.9), legend.position = 'none')
dev.off()

wilcox.test(merged_per_tumour[merged_per_tumour$age_group == 'Paediatric',]$not_rt_svs, merged_per_tumour[merged_per_tumour$age_group == 'Adolescent',]$not_rt_svs, alternative = 'two.sided', paired = F) #W = 18, p-value = 0.02381
wilcox.test(merged_per_tumour[merged_per_tumour$age_group == 'Paediatric',]$not_rt_svs, merged_per_tumour[merged_per_tumour$age_group == 'Adult',]$not_rt_svs, alternative = 'two.sided', paired = F) #W = 42, p-value = 0.001166
wilcox.test(merged_per_tumour[merged_per_tumour$age_group == 'Paediatric',]$rt_sv_count, merged_per_tumour[merged_per_tumour$age_group == 'Adolescent',]$rt_sv_count, alternative = 'two.sided', paired = F) #W = 25.5, p-value = 0.125
wilcox.test(merged_per_tumour[merged_per_tumour$age_group == 'Paediatric',]$rt_sv_count, merged_per_tumour[merged_per_tumour$age_group == 'Adult',]$rt_sv_count, alternative = 'two.sided', paired = F) #W = 55.5, p-value = 0.001448

```

##COPY GAIN TIMING

###mutationtimeR

Although mutationtimeR times gains on all chromosomes, it is informed by the clusters found using single sample dpclust using the autosomal subs only.

Move all the necessary input files into the working directory.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir

infile="/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt"

grep -v 'Sample' $infile | awk -F "\t" '{ if ($23 >= 7) {print}}' | while IFS=$'\t' read -r -a tmp;
do
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/00_final_calls/${tmp[0]}* /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/subs/${tmp[0]}*vcf /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/single_sample_dpclust_noX_new_swater/03_Output/${tmp[0]}*/${tmp[0]}*bestClusterInfo.txt /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir
done

```

####Functions

```{r}

sample_list_generator_lcm <- function(filepath){
  sample.list <- unlist(strsplit(list.files(paste0(filepath), pattern = '_post_swater_final_snvs.annot.vcf'), '_post_swater_final_snvs.annot.vcf'))
  write.table(sample.list, 'sample_names.txt', sep = '\t', quote = F, row.names = F, col.names = F)
}

sample_list_generator_bulk <- function(filepath){
  sample.list <- unlist(strsplit(list.files(paste0(filepath), pattern = '.caveman_c.annot_filtered.vcf'), '.caveman_c.annot_filtered.vcf'))
  write.table(sample.list, 'sample_names_bulk_yst.txt', sep = '\t', quote = F, row.names = F, col.names = F)
}

```

Helper functions on github that I found useful to add.
```{r}

loadBB <- function(file){
  tab <- read.table(file, header=TRUE, sep='\t')
  GRanges(tab$chromosome, IRanges(tab$start, tab$end), strand="*", tab[-3:-1])
}

mtHeader <- function() {
  DataFrame(Number=c(1,1,1,1,1,1,1,1,".",1,1,1,1,".",1),Type=c("Float","Float","Integer","Integer","Integer","Integer","Float","Float","Integer","Float","Float","Float","Float","String","String"), 
            Description=c("Mutation copy number","Change in MutCN between ancestral and derived state","Major copy number (ancestral)","Minor copy number (ancestral)","Major copy number (derived)","Minor copy number (derived)","Copy number frequency (relative to all cancer cells)", "MutCN probability","BB segment ids","Posterior prob: Early clonal","Posterior prob: Late clonal","Posterior prob: Subclonal", "Tail prob of mixture model", "Assignment probability of mutation to subclone","Mutation Time {clonal [early], clonal [late], clonal [NA], subclonal} - MAP assignments"),
            row.names=c("MutCN","MutDeltaCN","MajCN","MinCN","MajDerCN","MinDerCN","CNF","pMutCN","CNID","pGain","pSingle","pSub","pMutCNTail","pAllSubclones","CLS"))
}

mcnHeader <- function() {
  DataFrame(Number=c(1,1,1,1,1,1,1,1,".",1,1,1,1,"."),Type=c("Float","Float","Integer","Integer","Integer","Integer","Float","Float","Integer","Float","Float","Float","Float","String"), 
            Description=c("Mutation copy number","Change in MutCN between ancestral and derived state","Major copy number (ancestral)","Minor copy number (ancestral)","Major copy number (derived)","Minor copy number (derived)","Copy number frequency (relative to all cancer cells)", "MutCN probability","BB segment ids","Posterior prob: Early clonal","Posterior prob: Late clonal","Posterior prob: Subclonal", "Tail prob of mixture model", "Assignment probability of mutation to subclone"),
            row.names=c("MutCN","MutDeltaCN","MajCN","MinCN","MajDerCN","MinDerCN","CNF","pMutCN","CNID","pGain","pSingle","pSub","pMutCNTail","pAllSubclones"))
}

addMutCn <- function(vcf, bb=allBB[[meta(header(vcf))["ID",]]], clusters=allClusters[[meta(header(vcf))["ID",]]]){
  MCN = computeMutCn(vcf, bb, clusters)
  i = info(header(vcf))
  info(header(vcf)) <- rbind(i, mtHeader())
  info(vcf) <- cbind(info(vcf), MCN$D)
  return(vcf)
}

```

####Preprocessing

Create a sample list
```{r}

#LCM
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir')
sample_list_generator_lcm(filepath = '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir')


#Bulk
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir')
sample_list_generator_bulk(filepath = '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir')

```

Coerce Battenberg files into expected input.
```{bash}

#LCM
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir

while read line;
do
set $line
python /nfs/users/nfs_t/to3/scripts/copynumber.py -i ${1}_subclones.txt -c ${1}_rho_and_psi.txt -o . -r
done < sample_names.txt #list of samples to generate new bb txt files for

#Bulk
while read line;
do
set $line
python /nfs/users/nfs_t/to3/scripts/copynumber.py -i ${1}_subclones.txt -c ${1}_rho_and_psi.txt -o . -r
done < sample_names_bulk_yst.txt #list of samples to generate new bb txt files for

```

Change vcf header to the expected one and remove problematic header line that prevents us writing out the vcf output

```{bash}

#LCM
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir
while read line;
do
set $line
sed 's/#CHROM.*/#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tNORMAL\tTUMOUR/' ${1}_post_swater_final_snvs.annot.vcf | grep "##vcfProcessLog=" -v > ${1}.edited.vcf
done < sample_names.txt

#Bulk
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir
while read line;
do
set $line
sed 's/#CHROM.*/#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tNORMAL\tTUMOUR/' ${1}.caveman_c.annot_filtered.vcf | grep "##vcfProcessLog=" -v > ${1}.edited.vcf
done < sample_names_bulk_yst.txt

```

####Run mutationtimeR

```{r}

#Bulk
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir')

library(VariantAnnotation)
library(MutationTimeR)

manifest <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)
filepath = '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir'
sample.list <- unlist(strsplit(list.files(paste0(filepath), pattern = '_10000iters_3000burnin_bestClusterInfo.txt'), '_10000iters_3000burnin_bestClusterInfo.txt'))
reference_metadata <- manifest[manifest$Sample %in% c(sample.list), c('Sample', 'Sex')]
row.names(reference_metadata) = c(1:length(reference_metadata$Sample))

for(k in 1:nrow(reference_metadata)){
#for(k in 97:nrow(reference_metadata)){
  vcf <- readVcf(paste0(reference_metadata$Sample[k], '.edited.vcf'))
  vcf
  bb <- loadBB(paste0(reference_metadata$Sample[k],'_subclones_segments.txt'))
  clusters <- read.table(paste0(reference_metadata$Sample[k],'_10000iters_3000burnin_bestClusterInfo.txt'), sep = '\t', header = T, stringsAsFactors = F)
  clusters$proportion = clusters$location * max(bb$clonal_frequency)
  clusters <- clusters[, c('cluster.no', 'proportion', 'no.of.mutations')]
  names(clusters) <- c('cluster', 'proportion', 'n_ssms')
  
  mut_timed <- mutationTime(vcf, bb, clusters=clusters,
                            purity=max(bb$clonal_frequency,na.rm=TRUE), gender = paste0(reference_metadata$Sex[k]), isWgd = TRUE)
  
  i = info(header(vcf))
  info(header(vcf)) <- rbind(i, mtHeader())
  info(vcf) <- cbind(info(vcf), mut_timed$V)
  writeVcf(vcf, paste0(reference_metadata$Sample[k], '_ssm_post_mutationtime.vcf')) #not working, need to investigate
  mcols(bb) <- cbind(mcols(bb),mut_timed$T)
  saveRDS(bb, paste0(reference_metadata$Sample[k], '_post_mutationtime_segments.rds'))
  bb_out = data.frame(chr = seqnames(bb), 
                      start = start(bb), 
                      end = end(bb), 
                      strand = strand(bb), 
                      copy_number = elementMetadata(bb)$copy_number, 
                      major_cn = elementMetadata(bb)$major_cn, 
                      minor_cn = elementMetadata(bb)$minor_cn, 
                      clonal_frequency = elementMetadata(bb)$clonal_frequency,
                      type = elementMetadata(bb)$type,
                      time = elementMetadata(bb)$time,
                      time.lo = elementMetadata(bb)$time.lo,
                      time.up = elementMetadata(bb)$time.up,
                      time.2nd = elementMetadata(bb)$time.2nd,
                      time.2nd.lo = elementMetadata(bb)$time.2nd.lo,
                      time.2nd.up = elementMetadata(bb)$time.2nd.up,
                      time.star = elementMetadata(bb)$time.star,
                      n.snv_mnv = elementMetadata(bb)$n.snv_mnv)
  write.table(bb_out, paste0(reference_metadata$Sample[k], '_post_mutationtime_segments.txt'), quote = F, row.names = F, sep = '\t')

  pdf(paste0(reference_metadata$Sample[k], '_mutationtime.pdf'), height = 12, width = 8)
  plotSample(vcf = vcf, cn = bb, sv = NULL, title = paste0(reference_metadata$Sample[k]))
  dev.off()

  print(reference_metadata$Sample[k])
}

```

####Plot copy number states and their timing (Extended data figure 7)

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir')

subclone.files = list.files(pattern = '_post_mutationtime_segments.txt')

male.genome = read.table('/lustre/scratch119/casm/team294rr/rs30/References/GRCh37d5/chrom.sizes.txt', header = F, sep = '\t', stringsAsFactors = F)
row.names(male.genome) = male.genome$V1
male.genome = male.genome[c(1:22, 'X', 'Y'),]
auto.genome = male.genome[1:22,]
names(auto.genome) = c('chr', 'length')
l = as.numeric(auto.genome$length)

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/timed_cn_segments_gct_cohort_20210924.pdf', height = 9,  width = 8)
#with mutation timing
par(mar = c(3, 8, 2, 1), xpd = F)
plot(NA,NA, ylab="",xlab="", xlim= c(0, sum(l)), ylim=c(0,length(subclone.files)), xaxt="n", yaxt="n", xaxs="i", yaxs="i")
c = cumsum(l)
c_ref = c(0, c[-length(c)])
chr_offset = data.frame(chr =c(1:22), offset = c_ref)
g <- colorRampPalette(RColorBrewer::brewer.pal(4,"Set1")[c(3,2,4)])(101)
names(g) = seq(0.0, 1, 0.01)
d = cumsum(table(substr(subclone.files, 0, 7)))
axis(side=1, at=c(0,c), labels=rep('', length(l)+1))
axis(side=2, at=c(0,d), labels=rep('', length(d)+1))

for(k in 1:length(subclone.files)){
  data = read.table(subclone.files[k], header = T, stringsAsFactors = F, sep = '\t')
  data$chr = factor(data$chr, levels = c(1:22, 'X', 'Y'))
  data = data[data$chr %in% c(1:22),] #only autosomal
  data = data[order(data$chr, data$start, data$end, data$clonal_frequency, decreasing = T),]
  data = data[!duplicated(data[,1:3]),]
  data = data[order(data$chr, data$start, data$end),] #only major clone for any given segment, occasional major clone <0.5, hence doing it this way
  data$chr = as.numeric(data$chr)
  data = data[!is.na(data$copy_number),] #remove row with missing copy number state
  if(length(data[data$copy_number %in% c(0, 1),]$type)) data[data$copy_number %in% c(0, 1),]$type = "Deletion"
  if(length(data[is.na(data$type),]$type)) data[is.na(data$type),]$type = "NA"
  data[data$copy_number > 2 | data$type == 'CN-LOH',]$type = "Gain"
  if(nrow(data[data$major_cn == 1 & data$minor_cn == 1,]) > 0) data[data$major_cn == 1 & data$minor_cn == 1,]$type = "Normal"
  for(i in 1:nrow(data)){
  if(data$type[i] == 'Gain' & !is.na(data$time[i])){
    rect(xleft = chr_offset[chr_offset == data$chr[i],]$offset + data$start[i], ybottom = k - 1, xright = chr_offset[chr_offset == data$chr[i],]$offset + data$end[i], ytop = k, col = g[paste0(round(data$time[i], digits = 2))], border = NA)
  }
  if(data$type[i] == 'Gain' & is.na(data$time[i])){
    rect(xleft = chr_offset[chr_offset == data$chr[i],]$offset + data$start[i], ybottom = k - 1, xright = chr_offset[chr_offset == data$chr[i],]$offset + data$end[i], ytop = k, col = '#999999', border = NA)
  }
  if(data$type[i] == 'Normal'){
    rect(xleft = chr_offset[chr_offset == data$chr[i],]$offset + data$start[i], ybottom = k - 1, xright = chr_offset[chr_offset == data$chr[i],]$offset + data$end[i], ytop = k, col = 'white', border = NA)
  }
  if(data$type[i] == 'Deletion'){
    rect(xleft = chr_offset[chr_offset == data$chr[i],]$offset + data$start[i], ybottom = k - 1, xright = chr_offset[chr_offset == data$chr[i],]$offset + data$end[i], ytop = k, col = 'red', border = NA)
  }
}
}

abline(v = c, h = cumsum(table(substr(subclone.files, 0, 7))))
axis(side = 1, at = chr_offset$offset + auto.genome$length/2, labels = paste0(1:22), lwd.tick=0)
axis(side = 2, at = as.vector(c(0, d[-length(d)])) + as.vector(table(substr(subclone.files, 0, 7)) / 2), labels = names(d), lwd.tick=0, las = 2)
legend(x = 2800000000, y = 72,
       legend = c(0, rep('', 99), 1),
       fill = colorRampPalette(RColorBrewer::brewer.pal(4,"Set1")[c(3,2,4)])(101),
       border = NA,
       y.intersp = 0.5,
       cex = 0.4)
dev.off()


#without mutation timing, include all samples with >=5 reads per chromosome copy

all_cns = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_cn_calls.txt', header = T, sep = '\t', stringsAsFactors = F)
manifest <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)

all_cns$sample = factor(all_cns$sample, levels = c(unique(manifest[order(manifest$Age),]$Sample)))
all_cns = all_cns[order(all_cns$sample),]

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/untimed_cn_segments_gct_cohort_20210924.pdf', height = 9,  width = 8)
par(mar = c(3, 8, 2, 1), xpd = F)
plot(NA,NA, ylab="",xlab="", xlim= c(0, sum(l)), ylim=c(0,length(unique(all_cns$sample))), xaxt="n", yaxt="n", xaxs="i", yaxs="i")
c = cumsum(l)
c_ref = c(0, c[-length(c)])
chr_offset = data.frame(chr =c(1:22), offset = c_ref)
g <- c('blue', 'dodgerblue', 'white', '#e96868', '#FF2400', '#B22222')
names(g) = c(0, 1, 2, 3, 4, 5)

d = cumsum(table(substr(unique(all_cns$sample), 0, 7))[unique(substr(unique(all_cns$sample), 0, 7))])
axis(side=1, at=c(0,c), labels=rep('', length(l)+1))
axis(side=2, at=c(0,d), labels=rep('', length(d)+1))

for(k in unique(all_cns$sample)){
  data = all_cns[all_cns$sample == k,]
  data$chr = factor(data$chr, levels = c(1:22, 'X', 'Y'))
  data = data[data$chr %in% c(1:22),] #only autosomal
  data$copy_number = NA
  
  for(j in 1:nrow(data)){
    if(is.na(data$frac2_A[j])) data$copy_number[j] = data$nMaj1_A[j] + data$nMin1_A[j]
    if(!is.na(data$frac2_A[j])){
      if(data$frac1_A[j] > data$frac2_A[j]){
        data$copy_number[j] = data$nMaj1_A[j] + data$nMin1_A[j]
      }
      if(data$frac2_A[j] > data$frac1_A[j]){
        data$copy_number[j] = data$nMaj2_A[j] + data$nMin2_A[j]
      }
    }
  }
  
  data$chr = as.numeric(data$chr)
  data = data[!is.na(data$copy_number),] #remove row with missing copy number state
  if(length(data[data$copy_number > 4,]$copy_number)) data[data$copy_number > 4,]$copy_number = 5
  for(i in 1:nrow(data)){
    rect(xleft = chr_offset[chr_offset == data$chr[i],]$offset + data$startpos[i], ybottom = which(unique(all_cns$sample) == k) - 1, xright = chr_offset[chr_offset == data$chr[i],]$offset + data$endpos[i], ytop = which(unique(all_cns$sample) == k), col = g[paste0(data$copy_number[i])], border = NA)
  }
}

abline(v = c, h = cumsum(table(substr(unique(all_cns$sample), 0, 7))[unique(substr(unique(all_cns$sample), 0, 7))]))
axis(side = 1, at = chr_offset$offset + auto.genome$length/2, labels = paste0(1:22), lwd.tick=0)
axis(side = 2, at = as.vector(c(0, d[-length(d)])) + as.vector(table(substr(unique(all_cns$sample), 0, 7))[unique(substr(unique(all_cns$sample), 0, 7))] / 2), labels = names(d), lwd.tick=0, las = 2)
dev.off()
#legend(x = 2000000000, y = 72,
#       legend = c(0, 1, 2, 3, 4, '5+'),
#       fill = c('blue', 'dodgerblue', 'white', '#e96868', '#FF2400', '#B22222'),
#       title = 'Total CN',
#       border = 'black',
#       cex = 1)

```

###WGD estimation analysis

####Functions

Functions taken from PCAWG paper

```{r}

source('/nfs/users/nfs_t/to3/scripts/additional_mutationtimeR_functions.R') #source extra, unchanged functions

#edited functions from mutationtimeR
findMainCluster <- function(bb, min.dist = 0.05){ #at least 20 SNVs
    w <- which(bb$n.snv_mnv >= 20 & !is.na(bb$time) & ((bb$time.up - bb$time.lo) < 0.5)) #edited to exclude regions with very wide confidence intervals when estimating WGD time, to3 03/03/21
    s <- seq(0,1,0.001) #increased granularity in view of how early WGD is, to3 03/03/21
    l2 <- pmin(bb$time.lo, bb$time - min.dist)[w]
    u2 <- pmax(bb$time.up, bb$time + min.dist)[w]
    l1 <- (l2 +  bb$time[w])/2
    u1 <- (u2 +  bb$time[w])/2
    wd <- as.numeric(width(bb)[w])
    o <- sapply(s, function(i) sum(wd * ( (l2 <= i & u2 >=i) + (l1 <= i & u1 >= i))))
    s[which.max(o)]
}

#remove the deamination component as we have too few mutations, especially associated with SBS1 to do this and filter low confidence timing segments from analysis
computeWgdParam <- function(vcf, bb, clusters, purity, sex){ #added sex in
    # 1. Find segments compatible with WGD
    min.dist <- 0.05
    m <- findMainCluster(bb) #modifed as above to higher confidence regions on which to time WGD, to3 03/03/21
    l <- pmin(bb$time.lo, bb$time - min.dist)
    u <- pmax(bb$time.up, bb$time + min.dist)
    o <- which(l <= m & u >= m & ((bb$time.up - bb$time.lo) < 0.5)) #only going to count and time over segments with higher confidence, to3 03/03/21
    #note that this line has been left as is in fractionGenomeWgdCompatible as there we only want to know the regions compatable with WGD, not necessarily depend on them all for downstream analyses, to3 03/03/21
    # 2. Find substitutions in compatible segments
    intersect_seg = findOverlaps(vcf, bb[o], ignore.strand = T) #%over% method not working so this is a workaround, to3 03/03/21
    vcf <- vcf[queryHits(intersect_seg),] 
    w <- which(info(vcf)$MajCN==2 & sapply(info(vcf)$CNID, length)==1)
    v <- vcf[w]
    if(nrow(v)< 100) return(NULL) # At least 100 SNVs total over which to estimate the time of acquisition
    seqnames(rowRanges(v)) <- factor(3-info(v)$MinCN, levels=seqlevels(v))
    v <- sort(v)
    
    # 3. Merged CN segments
    b <- GRanges(1:3, IRanges(rep(1,3),rep(max(end(v)),3)), copy_number=4:2, major_cn=2, minor_cn=2:0, clonal_frequency=as.numeric(purity))
    
    # 4. Calculate times
    l <- computeMutCn(v, b, clusters, purity, isWgd=TRUE, n.boot=200, rho=0.01, xmin=3, gender = sex) #recalculates times calculated before on single segments
    b$n.snv_mnv <- l$n <- table(factor(info(v)$MinCN, levels=2:0))
    l$time <- bbToTime(b, l$P)

    row.names(l$D) = row.names(info(v)) #get sub annotations back, to3 03/03/21
    l$cn_state_adj_burden = sum(l$D$MutCN == 2, na.rm = T) + (sum(l$D$MutCN == 2, na.rm = T) * ((unname(l$n['0']) +  unname(l$n['1'])) /  (unname(l$n['0']) +  unname(l$n['1']) + unname(l$n['2'])))) #correct for the 2+1 and 2+0 regions where pre-duplication mutations on the minor allele cannot be counted, to3 04/03/21
    l$hq_seg_width = sum(width(bb[o][bb[o]$major_cn == 2])) #over what length of the genome were these substitutions identified? to3 03/03/21
    l$adj_genome_sub_burden = l$cn_state_adj_burden * (genome_length / l$hq_seg_width) #number of mutations estimated to be on both copies, corrected for genome coverage by good confidence WGD segments, to3 04/03/21
         
    return(l)
}

#fractionGenomeWgdCompatible <- function(bb, min.dist=0.05){
#    m <- findMainCluster(bb)
#    l <- pmin(bb$time.lo, bb$time - min.dist)
#    u <- pmax(bb$time.up, bb$time + min.dist)
#    w <- which(l <= m & u >= m)
#    avgCi <- weighted.mean(bb$time.up- bb$time.lo, width(bb), na.rm=TRUE)
#    sd.wgd <- sqrt(weighted.mean((bb$time[w] - m)^2, width(bb)[w], na.rm=TRUE))
#    sd.all <- sqrt(weighted.mean((bb$time - m)^2, width(bb), na.rm=TRUE))
#    c(nt.wgd=sum(as.numeric(width(bb))[w]), nt.total=sum(as.numeric(width(bb))[!is.na(bb$time)]), time.wgd=m, n.wgd=length(w), n.all = sum(!is.na(bb$time)), chr.wgd = length(unique(seqnames(bb)[w])), chr.all = length(unique(seqnames(bb)[!is.na(bb$time)])), sd.wgd=sd.wgd, avg.ci=avgCi, sd.all=sd.all) 
#}

```

####Generate estimates

Create lists of pertinent input data, including output files from mutationtimeR

```{r}

library(VariantAnnotation)

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir')

chrom_sizes <- read.table('/lustre/scratch119/casm/team294rr/rs30/References/GRCh37d5/chrom.sizes.txt', header = F, sep = '\t', stringsAsFactors = F)
genome_length = sum(chrom_sizes$V2) #3095677412

bb_files = list.files(pattern = '_post_mutationtime_segments.rds')

finalBB <- list()
for( f in bb_files){
    g = readRDS(f)
    finalBB[[unlist(strsplit(f, '_post_mutationtime_segments.rds'))]] <- g
}

snv_files = list.files(pattern = '_ssm_post_mutationtime.vcf')

finalSnv = list()
for(f in snv_files){
  vcf = readVcf(f)
  finalSnv[[unlist(strsplit(f, '_ssm_post_mutationtime.vcf'))]] <- vcf
}

cluster_files = list.files(pattern = '_10000iters_3000burnin_bestClusterInfo.txt')

finalClusters <- list()
for(f in cluster_files){
  clusters <- read.table(f, sep = '\t', header = T, stringsAsFactors = F)
  sample = unlist(strsplit(f, '_10000iters_3000burnin_bestClusterInfo.txt'))
  bb = read.table(paste0(sample, '_post_mutationtime_segments.txt'), sep = '\t', header = T, stringsAsFactors = F)
  clusters$proportion = clusters$location * max(bb$clonal_frequency)
  clusters <- clusters[, c('cluster.no', 'proportion', 'no.of.mutations')]
  names(clusters) <- c('cluster', 'proportion', 'n_ssms')
  finalClusters[[sample]] <- clusters
}

purity_files = list.files(pattern = '_rho_and_psi.txt')

finalPurity <- numeric()
for(f in purity_files){
  pur_df = read.table(f, sep = '\t', header = T, stringsAsFactors = F)
  purity = pur_df['FRAC_GENOME', 'rho']
  finalPurity[unlist(strsplit(f, '_rho_and_psi.txt'))] = purity
}

manifest <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F)
filepath = '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir'
sample.list <- unlist(strsplit(list.files(paste0(filepath), pattern = '_10000iters_3000burnin_bestClusterInfo.txt'), '_10000iters_3000burnin_bestClusterInfo.txt'))
reference_metadata <- manifest[manifest$Sample %in% c(sample.list), c('Sample', 'Sex')]
row.names(reference_metadata) = c(1:length(reference_metadata$Sample))


sex_vec = reference_metadata$Sex
names(sex_vec) = reference_metadata$Sample

```

Sanity check for final, eligible samples - are they consistent with WGD?

```{r}

finalPloidy <- sapply(finalBB, averagePloidy)
names(finalPloidy) <- names(finalBB)

finalHom <- sapply(finalBB, averageHom)
names(finalHom) <- names(finalBB)

isWgd <- .classWgd(finalPloidy, finalHom) #2.9 -2*hom <= ploidy
table(isWgd) #all samples classed as WGD based on ploidy and homozygosity

fracGenomeWgdComp <- t(sapply(finalBB, function(bb) {
                    fgw <- try(fractionGenomeWgdCompatible(bb)); 
                    if(class(fgw)!='try-error') fgw
                    else rep(NA,10)}))
rownames(fracGenomeWgdComp) <- names(finalBB)
fracGenomeWgdComp = data.frame(fracGenomeWgdComp)
fracGenomeWgdComp$patient = substr(row.names(fracGenomeWgdComp), 0, 7)
aggregate(time.wgd ~ patient, fracGenomeWgdComp, median)

#   patient time.wgd
#  PD42036   0.0200
#  PD43296   0.0170
#  PD43298   0.0485
#  PD43299   0.3950
#  PD45543   0.0430
#  PD45544   0.0010
#  PD45545   0.0150
#  PD46269   0.0010
#  PD46271   0.0175
# PD46966   0.0380
# PD46969   0.0010
# PD50654   0.0160
# PD50655   0.8140
# PD50656   0.5680
# PD50657   0.0520
# PD50660   0.4890
# PD50661   0.2290

wgdStar <- factor(rep(1,nrow(fracGenomeWgdComp)), levels=0:3, labels=c("unlikely","uninformative","likely","very likely"))
wgdStar[fracGenomeWgdComp[,"avg.ci"]<=0.75 & fracGenomeWgdComp[,"nt.total"]/genome_length >= 0.33 ] <- "likely"
wgdStar[fracGenomeWgdComp[,"nt.wgd"]/fracGenomeWgdComp[,"nt.total"] < 0.66] <- "unlikely"
wgdStar[wgdStar=="likely" & fracGenomeWgdComp[,"nt.wgd"]/fracGenomeWgdComp[,"nt.total"] > 0.8 & fracGenomeWgdComp[,"sd.wgd"] < 0.1 &  fracGenomeWgdComp[,"nt.total"]/genome_length > 0.5] <- "very likely"
names(wgdStar) <-  names(finalBB)
prop.table(table(wgdStar[isWgd])) #only in a subset of microbiopsies thought unlikely using this metric

wgdPoss <- !isWgd & 2.5 - 1.5 * finalHom <= finalPloidy #none

#wgdStat <- factor(wgdPoss + 2*isWgd - wgdPoss*isWgd, labels=c("absent","possible","present"))
#table(wgdStat, wgdStar)

```

Generate parameters for estimating the pre-duplication mutation burden.

```{r}

wgdParam <- mclapply(names(finalSnv)[isWgd], function(ID){
            try(computeWgdParam(finalSnv[[ID]], finalBB[[ID]], clusters=finalClusters[[ID]], purity=finalPurity[ID], sex=sex_vec[ID]))
        })
names(wgdParam) <- names(finalSnv)[isWgd]
void <- sapply(wgdParam, function(x) is.null(x) | class(x)=="try-error") #none

saveRDS(wgdParam, 'mutationtimeR_WGD_segments_20210918.rds')

predup_burden_df = data.frame(sample = names(wgdParam))
row.names(predup_burden_df) = predup_burden_df$sample

predup_burden_df$adj_genome_sub_burden = NA
predup_burden_df$hq_seg_width = NA

for(i in 1:nrow(predup_burden_df)){
  predup_burden_df$adj_genome_sub_burden[i] = wgdParam[[predup_burden_df$sample[i]]]$adj_genome_sub_burden
  predup_burden_df$hq_seg_width[i] = wgdParam[[predup_burden_df$sample[i]]]$hq_seg_width
}

predup_burden_df$patient = substr(predup_burden_df$sample, 0, 7)

write.table(fracGenomeWgdComp, 'gct_wgd_estimate_mut_time_20210918.txt', col.names = T, row.names = T, sep = '\t', quote = F)
write.table(predup_burden_df, 'gct_predup_sub_burden_estimate_20210918.txt', col.names = T, row.names = F, sep = '\t', quote = F)

```

####Age vs pre-duplication sub burden (Fig. 2a)

```{r}

library(reshape2)
library(ggplot2)
library(ggpubr)
library(VariantAnnotation)
library(ggbeeswarm)
library(investr)

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir')
setwd('/lustre/scratch126/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir')

fracGenomeWgdComp = read.table('gct_wgd_estimate_mut_time_20210918.txt', header = T, sep = '\t', stringsAsFactors = F)
predup_burden_df = read.table('gct_predup_sub_burden_estimate_20210918.txt', header = T, sep = '\t', stringsAsFactors = F)

predup_burden_df_per_patient = aggregate(adj_genome_sub_burden ~ patient, predup_burden_df, mean)

#  patient adj_genome_sub_burden
#  PD42036              2.880271
#  PD43296              8.520985
#  PD43298             17.974153
#  PD43299            399.311492
#  PD45543              0.000000
#  PD45544              0.000000
#  PD45545              8.143693
#  PD46269              2.056881
#  PD46271              5.366702
#  PD46966              6.169503
#  PD46969             11.098779
#  PD50654             29.871324
#  PD50655            788.782149
#  PD50656            312.862374
#  PD50657             10.601263
#  PD50660            399.891822
#  PD50661            359.694023

pt_ages = manifest[!duplicated(manifest$Case.ID), c('Case.ID', 'Age', 'Diagnosis', 'Organ')]

predup_burden_df_per_patient$age = NA
predup_burden_df_per_patient$dx = NA
predup_burden_df_per_patient$site = NA
for(i in 1:nrow(predup_burden_df_per_patient)){
  predup_burden_df_per_patient$age[i] = pt_ages[pt_ages$Case.ID == predup_burden_df_per_patient$patient[i],]$Age
  predup_burden_df_per_patient$dx[i] = pt_ages[pt_ages$Case.ID == predup_burden_df_per_patient$patient[i],]$Diagnosis
  predup_burden_df_per_patient$site[i] = pt_ages[pt_ages$Case.ID == predup_burden_df_per_patient$patient[i],]$Organ
}

median(predup_burden_df_per_patient[predup_burden_df_per_patient$age > 12,]$adj_genome_sub_burden) #5.768102
range(predup_burden_df_per_patient[predup_burden_df_per_patient$age > 12,]$adj_genome_sub_burden) #0.00000 17.97415
median(predup_burden_df_per_patient[predup_burden_df_per_patient$age <= 12,]$adj_genome_sub_burden) #359.694
range(predup_burden_df_per_patient[predup_burden_df_per_patient$age <= 12,]$adj_genome_sub_burden) #10.60126 788.78215

ggplot(predup_burden_df_per_patient) + geom_point(mapping = aes(x = age, y = adj_genome_sub_burden, col = dx, shape = site)) +
  theme_pubr() + ylab('Estimated pre-duplication substitution burden') + xlab('Age (years)') +
  theme(legend.position = 'right')

#all models
asympt.model = nls(adj_genome_sub_burden ~ SSasymp(age, Asym, R0, lrc), predup_burden_df_per_patient) #asymptotic regression
null.model = lm(formula = adj_genome_sub_burden ~ 1, data = predup_burden_df_per_patient) # null model
#predup_burden_df_per_patient$log.adj_genome_sub_burden = log(predup_burden_df_per_patient$adj_genome_sub_burden + 1)
#log.model = lm(formula = log.adj_genome_sub_burden ~ age, data = predup_burden_df_per_patient) # log model
simple.linear.model = lm(formula = adj_genome_sub_burden ~ age, data = predup_burden_df_per_patient) #simple linear

#identify best fit
AIC(asympt.model, null.model, simple.linear.model) #asymptotic regression fits best

summary(asympt.model)

#generate confidence intervals for asymptotic regression model
age_range = seq(0.75, 58, 0.25) #new age range to generate interval over
cimat <- as.data.frame(investr::predFit(asympt.model, data.frame(age = age_range), interval = "confidence", level = 0.95))
cimat$age = age_range

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/age_vs_pre_dup_burden_plot_20210919.pdf', height = 3,  width = 4, useDingbats = F)
ggplot() +
  theme_pubr() + ylab('Estimated pre-duplication substitution burden') + xlab('Age (years)') +
  theme(legend.position = 'none') +
  geom_ribbon(data = cimat, mapping = aes(x = age, ymin = lwr, ymax = upr), color = NA, alpha = 0.1, fill = "grey50") +
  #geom_point(predup_burden_df_per_patient, mapping = aes(x = age, y = adj_genome_sub_burden, fill = dx, shape = site), col = 'black', size = 5) + 
  geom_point(predup_burden_df_per_patient, mapping = aes(x = age, y = adj_genome_sub_burden, fill = dx), shape = 21, col = 'black', size = 5) + 
  stat_smooth(data = predup_burden_df_per_patient, mapping = aes(x = age, y = adj_genome_sub_burden), method = "nls", formula = y ~ SSasymp(x, Asym, R0, lrc), se = F, lty = 2, col = 'red') +
  scale_fill_manual(values = c('#9d9d9c', '#ffffff', '#1d1d1b'))
  #+ scale_shape_manual(values = c(21, 24, 22))
dev.off()

```

####Postpubertal WGD timing vs PCAWG data (Fig. 2b)

PCAWG data kindly provided by Peter van Loo

```{r}

library(RColorBrewer)
library(ggplot2)

summary_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', header = T, sep = '\t', stringsAsFactors = F)
row.names(summary_data) = summary_data$Sample

#read in PCAWG data
pcawg_wgd = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/reference_datasets/20190824_timing_gains_adj.txt', header = T, sep = '\t', stringsAsFactors = F)

pcawg_wgd[pcawg_wgd$samplename %in% pcawg_wgd[pcawg_wgd$type == 'WGD',]$samplename,] #samples with wgd


sample_hist_key = pcawg_wgd[!duplicated(pcawg_wgd[, c('samplename', 'histology_abbreviation')]), c('samplename', 'histology_abbreviation')] #unique entries per sample
pcawg_wgd = pcawg_wgd[pcawg_wgd$histology_abbreviation %in% sort(unique(sample_hist_key$histology_abbreviation))[table(sample_hist_key$histology_abbreviation) >= 10], ] #only keep tumour types with at least 10 samples to make proportion values sensible

sample_hist_key = sample_hist_key[sample_hist_key$histology_abbreviation %in% sort(unique(sample_hist_key$histology_abbreviation))[table(sample_hist_key$histology_abbreviation) >= 10], ] #only do it for tumour types with at least 10 tumours represented

sample_hist_key$WGD = NA
sample_hist_key$WGD_timing = NA

for(i in 1:nrow(sample_hist_key)){
  sample_hist_key$WGD[i] = 'No'
  if(sample_hist_key$samplename[i] %in% pcawg_wgd[pcawg_wgd$chr == 'WGD',]$samplename){
    sample_hist_key$WGD[i] = 'Yes'
    sample_hist_key$WGD_timing[i] = pcawg_wgd[pcawg_wgd$samplename == sample_hist_key$samplename[i] & pcawg_wgd$chr == 'WGD',]$time
  } 
}

nrow(sample_hist_key) #2096
length(unique(sample_hist_key$histology_abbreviation)) #31

#read in GCT data
timing_df = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir/gct_wgd_estimate_mut_time_20210918.txt', sep = '\t', header = T, stringsAsFactors = F)

median_wgd = aggregate(time.wgd ~ patient, timing_df, median)
median_wgd$age = NA
for(i in 1:nrow(median_wgd)){
    median_wgd$age[i] = unique(summary_data[summary_data$Case.ID == median_wgd$patient[i],]$Age)
}
  
median_wgd_postpubertal = median_wgd[median_wgd$age > 12,]
median_wgd_postpubertal$histology_abbreviation = 'Postpubertal_GCT'
median_wgd_postpubertal = median_wgd_postpubertal[, c(1,2,4)]


gct_samples = data.frame(samplename = median_wgd_postpubertal$patient, histology_abbreviation = rep('Postpubertal_GCT', nrow(median_wgd_postpubertal)), WGD = rep('Yes', nrow(median_wgd_postpubertal)), WGD_timing = median_wgd_postpubertal$time.wgd)

median(median_wgd_postpubertal[median_wgd_postpubertal$patient %in% c('PD45543', 'PD45544', 'PD45545'),]$time.wgd) #post-pubertal ovary 0.015
median(median_wgd_postpubertal[!median_wgd_postpubertal$patient %in% c('PD45543', 'PD45544', 'PD45545'),]$time.wgd) #post-pubertal testis 0.0175

#combined data gct and pcawg data
sample_hist_key = rbind(sample_hist_key, gct_samples)

sample_hist_key$WGD_timing_plot = round(sample_hist_key$WGD_timing, digits = 2)
sample_hist_key[is.na(sample_hist_key$WGD_timing_plot),]$WGD_timing_plot = 'None'
sample_hist_key$WGD_timing_plot = factor(sample_hist_key$WGD_timing_plot, levels = rev(sort(unique(sample_hist_key$WGD_timing_plot))))

#rank by proportion of samples that undergo WGD and then arrange by the estimated time at which it occurs
ranked_histo = as.data.frame(unique(sample_hist_key$histology_abbreviation))
names(ranked_histo) = 'Dx'

ranked_histo$noWGD_count = NA
ranked_histo$total = NA

for(i in 1:nrow(ranked_histo)){
  ranked_histo$noWGD_count[i] = nrow(sample_hist_key[sample_hist_key$histology_abbreviation == ranked_histo$Dx[i] & sample_hist_key$WGD == 'No',])
  ranked_histo$total[i] = nrow(sample_hist_key[sample_hist_key$histology_abbreviation == ranked_histo$Dx[i],])
}

ranked_histo$prop_noWGD = ranked_histo$noWGD_count / ranked_histo$total
sample_hist_key$histology_abbreviation = factor(sample_hist_key$histology_abbreviation, levels = ranked_histo[order(ranked_histo$prop_noWGD, decreasing = F),]$Dx)

#g <- colorRampPalette(RColorBrewer::brewer.pal(4,"Set1")[c(3,2,4)])(101)
g <- colorRampPalette(c('black', 'grey95'))(101)
names(g) = seq(0.0, 1, 0.01)
h = c(g, 'white')
names(h)[102] = 'None'

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/wgd_gct_v_pcawg_plot_20210919.pdf', height = 8, width = 6)
ggplot(sample_hist_key, aes(y = histology_abbreviation, fill = WGD_timing_plot)) +
  geom_bar(position="fill") + 
  scale_fill_manual(values = h) + 
  coord_cartesian(expand = F) +
  theme_classic() +
  theme(legend.position = 'none') +
  xlab('Proportion undergone WGD')  + ylab('')
dev.off()

```

####Number of cell division prior to WGD

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/mutationtimeR_new_swater/wkdir')

predup_burden_df = read.table('gct_predup_sub_burden_estimate_20210918.txt', header = T, sep = '\t', stringsAsFactors = F)
manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', header = T, sep = '\t', stringsAsFactors = F)

predup_burden_df_per_patient = aggregate(adj_genome_sub_burden ~ patient, predup_burden_df, mean)
pt_ages = manifest[!duplicated(manifest$Case.ID), c('Case.ID', 'Age', 'Diagnosis', 'Organ')]

predup_burden_df_per_patient$age = NA
predup_burden_df_per_patient$dx = NA
predup_burden_df_per_patient$site = NA
for(i in 1:nrow(predup_burden_df_per_patient)){
  predup_burden_df_per_patient$age[i] = pt_ages[pt_ages$Case.ID == predup_burden_df_per_patient$patient[i],]$Age
  predup_burden_df_per_patient$dx[i] = pt_ages[pt_ages$Case.ID == predup_burden_df_per_patient$patient[i],]$Diagnosis
  predup_burden_df_per_patient$site[i] = pt_ages[pt_ages$Case.ID == predup_burden_df_per_patient$patient[i],]$Organ
}

#Rahbari et al 2016 report 0.5-0.7 mutations per haploid genome per cell division. We can, therefore, extrapolate to 1-1.4 mutations per diploid genome per cell division (the assumed diploid PGC precursor to postpubertal GCTs)
#this is interpreted as 1.2 +/- 0.2
#assuming no uncertainty for the pre-duplication substitution burden
#Number of cell division prior to WGD (f) = Pre-WGD sub burden (A) / Estimated subs per diploid genome per cell division post-PGC specification (B)
#sig_f = f * (sig_B / B) where B = 1.2 and sig_B = 0.2

predup_burden_df_per_patient$estimated_postPGC_cell_divisions = predup_burden_df_per_patient$adj_genome_sub_burden / 1.2 #estimated subs per diploid genome per cell division post PGC specification

#take estimate of postPGC mutation rate per cell division in testis and ovary from Rahbari et al. 2016
predup_burden_df_per_patient$estimated_postPGC_cell_divisions_err = predup_burden_df_per_patient$estimated_postPGC_cell_divisions * (0.2 / 1.2)


median(predup_burden_df_per_patient[predup_burden_df_per_patient$age > 12,]$estimated_postPGC_cell_divisions) #4.806752
range(predup_burden_df_per_patient[predup_burden_df_per_patient$age > 12,]$estimated_postPGC_cell_divisions) #0.00000 14.97846
median(predup_burden_df_per_patient[predup_burden_df_per_patient$age > 12,]$estimated_postPGC_cell_divisions_err) #0.8011253

```

##PHYLOGENY

###multidimDPClust

Not using sex chromosomes.

Check that the input files were the same.

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/')

cp_files = list.files(".", "_cellularity_ploidy.txt")

all_cp = data.frame()
for(file in cp_files){
  sample = unlist(strsplit(file, "_cellularity_ploidy.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  data$sample = sample
  all_cp = rbind(all_cp, data)
}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/01_DNA_PROCESSING/variant_filtering/cnas/01_battenberg')

new_cp_files = list.files(".", "_cellularity_ploidy.txt")

new_all_cp = data.frame()
for(file in cp_files){
  sample = unlist(strsplit(file, "_cellularity_ploidy.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  data$sample = sample
  new_all_cp = rbind(new_all_cp, data)
}

sum(all_cp$cellularity != new_all_cp$cellularity) #1
sum(all_cp$ploidy != new_all_cp$ploidy) #1
sum(all_cp$psi != new_all_cp$psi) #1
sum(all_cp$sample != new_all_cp$sample) #0

all_cp[all_cp$cellularity != new_all_cp$cellularity,] #PD50657a only had a change so we are good

```


####Prepare data files 

With the basic canpipe output files copied into the directory, generate files 1), 2), and 3).

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input')

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', header = T, stringsAsFactors = F, sep = '\t')
multidim = c("PD42036", "PD43296", "PD43298", "PD43299", "PD45544", "PD46269", "PD46966", "PD46969")

for(i in 1:nrow(manifest)){
  if(manifest$Case.ID[i] %in% multidim & manifest$Average.reads.per.chromosome.copy.[i] >= 5){
    system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/', manifest$Sample[i], '_rho_and_psi.txt .'))
    system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/', manifest$Sample[i], '_subclones.txt .'))
    system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/', manifest$Sample[i], '_cellularity_ploidy.txt .'))
    system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/subs/', manifest$Sample[i], '_post_swater_final_snvs.annot.vcf .'))
  }
}

```

Change column names in vcf to expected one
```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input

for f in $(ls *_post_swater_final_snvs.annot.vcf);
do
sample=${f:0:15}
echo $sample
sed 's/#CHROM.*/#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tNORMAL\tTUMOUR/' ${sample}_post_swater_final_snvs.annot.vcf > ${sample}.edited.vcf
done

```

Create separate directories per patient.

```{bash}

for f in $(ls *_post_swater_final_snvs.annot.vcf | cut -c1-7 | uniq);
do
mkdir -p $f
mv $f*.* $f
done 

```

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input')

sample_list_generator <- function(filepath){
  sample.list <- unlist(strsplit(list.files(paste0(filepath), pattern = '_rho_and_psi.txt'), '_rho_and_psi.txt'))
  write.table(sample.list, paste0(filepath, '/sample_names.txt'), sep = '\t', quote = F, row.names = F, col.names = F)
}

cellularity_list_generator <- function(filepath){
  cellularity <- c()
  for(f in unlist(strsplit(list.files(paste0(filepath), pattern = '_rho_and_psi.txt'), '_rho_and_psi.txt'))){
    file <- read.table(paste0(filepath, '/', f, '_rho_and_psi.txt'), stringsAsFactors = F, header = T, row.names = 1, sep = '\t')
    cellularity <- rbind(cellularity, file['FRAC_GENOME', 'rho']) #corrected to right way around 7/3/22
    write.table(cellularity, paste0(filepath, '/cellularity.txt'), sep = '\t', quote = F, row.names = F, col.names = F)
  }
}

subclone_list_generator <- function(filepath){
  subclone.list <- unlist(list.files(paste0(filepath), pattern = '_subclones.txt'))
  write.table(subclone.list, paste0(filepath, '/subclone.files.txt'), sep = '\t', quote = F, row.names = F, col.names = F)
}

multidimdp_preproc <- function(filepath){
  sample_list_generator(filepath)
  cellularity_list_generator(filepath)
  subclone_list_generator(filepath)
}

```

```{r}

for ( i in multidim){
  multidimdp_preproc(paste0('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/', i))
}

```

Create input files for allelecount
```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input

for f in PD42036 PD43296 PD43298 PD43299 PD45544 PD46269 PD46966 PD46969;
do
cut -f 1,2,4,5 $f/*.edited.vcf | sort -k1,1 -k2,2n -k3,4 | uniq > $f/${f}_loci.txt
grep -v '#' $f/${f}_loci.txt > $f/${f}_loci_SNVs_nohash.txt
mv $f/${f}_loci_SNVs_nohash.txt $f/${f}_loci.txt
done

#remove sex chromosome mutations

for f in PD42036 PD43296 PD43298 PD43299 PD45544 PD46269 PD46966 PD46969;
do
grep -v 'X\|Y' $f/${f}_loci.txt > $f/${f}_loci_noX.txt
mv $f/${f}_loci_noX.txt $f/${f}_loci.txt
done

```

Create a txt file containing a list of the locations of the BAM files. 

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input

for f in $(ls PD*/*loci.txt);
do
patient=$f
dir=${patient%/*}
echo $dir
done >> patients.txt # edit to include the project id as a second column and sex as third column

while read line;
do
set $line
echo $1
echo $2
for f in $(ls $1/*.edited.vcf);
do
#echo $f
filename=$(basename -- "$f")
SAMPLE="${filename%.edited.vcf}"
echo "/lustre/scratch117/casm/team274/to3/gct_bams/${2}/${SAMPLE}/mapped_sample/${SAMPLE}.sample.dupmarked.bam" >> $1/${1}_bam_list.txt
done
done < patients.txt

```

Run allelecounter on cgpfarm.
```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input
module load alleleCount

for f in PD42036 PD43296 PD43298 PD43299 PD45544 PD46269 PD46966 PD46969;
do
while read i;
do
sampleID=$(basename $i | sed "s/[.].*//")
echo $sampleID
echo $i
bsub -G team294-vwork -J"${sampleID}" -o ${f}/${sampleID}.o -e ${f}/${sampleID}.e -q normal -n 5 -R"select[mem>10000] rusage[mem=10000] span[hosts=1]" -M10000 "alleleCounter -l ${f}/${f}_loci.txt -b $i -o /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/${f}/${sampleID}_alleleFrequencies.txt -m 25 -q 30"
done < ${f}/${f}_bam_list.txt 
done

```

Now we need to split the WT and MT counts across all samples into two files. We can do this using the loci and allelefrequency output files.
```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input')

wt_mt_count_file_generator <- function(filepath){
  file = list.files(filepath, '_loci.txt')
  loci_file = read.table(paste0(filepath, '/', file), header = F, sep = '\t', stringsAsFactors = F)
  loci_file = loci_file[!(loci_file$V1 %in% loci_file[duplicated(loci_file[,1:2]),]$V1 & loci_file$V2 %in% loci_file[duplicated(loci_file[,1:2]),]$V2),] #remove loci that violate infinite sites assumption
  wt_mat = as.data.frame(matrix(NA, 
                                ncol =  length(unlist(strsplit(list.files(path = filepath, '_alleleFrequencies.txt'), '_alleleFrequencies.txt'))) + 1, 
                                nrow = nrow(loci_file),
                                dimnames = list(NULL, c('ID', unlist(strsplit(list.files(path = filepath, '_alleleFrequencies.txt'), '_alleleFrequencies.txt'), use.names = F)))))
  wt_mat$ID = paste(loci_file$V1, loci_file$V2, sep = '_')
  
  mt_mat = as.data.frame(matrix(NA, 
                                ncol =  length(unlist(strsplit(list.files(path = filepath, '_alleleFrequencies.txt'), '_alleleFrequencies.txt'))) + 1, 
                                nrow = nrow(loci_file),
                                dimnames = list(NULL, c('ID', unlist(strsplit(list.files(path = filepath, '_alleleFrequencies.txt'), '_alleleFrequencies.txt'), use.names = F)))))
  mt_mat$ID = paste(loci_file$V1, loci_file$V2, sep = '_')
  
  for ( i in unlist(strsplit(list.files(path = filepath, '_alleleFrequencies.txt'), '_alleleFrequencies.txt')) ){
    alleleFreq_file = read.table(paste0(filepath, '/', i, '_alleleFrequencies.txt'), header = T, sep = '\t', comment.char = '')
    for ( j in 1:nrow( wt_mat ) ){
      chr = unlist(strsplit(wt_mat$ID[j], '_'))[1]
      pos = unlist(strsplit(wt_mat$ID[j], '_'))[2]
      ref = loci_file[loci_file$V1 == chr & loci_file$V2 == pos, ]$V3
      alt = loci_file[loci_file$V1 == chr & loci_file$V2 == pos, ]$V4
      wt_mat[j, i] = alleleFreq_file[alleleFreq_file$X.CHR == chr & alleleFreq_file$POS == pos, paste0('Count_', ref)]
      mt_mat[j, i] = alleleFreq_file[alleleFreq_file$X.CHR == chr & alleleFreq_file$POS == pos, paste0('Count_', alt)]
    }
  }
  wt_mat <- wt_mat[rowSums(mt_mat[2:ncol(mt_mat)]) > 0,] #remove any rows where no SNVs are called, this shouldn't happen with the loci list given but is a precautionary extra filter
  mt_mat <- mt_mat[rowSums(mt_mat[2:ncol(mt_mat)]) > 0,]
  write.table(wt_mat, paste0(filepath, '/wt_count_file.txt'), sep = '\t', quote = F, row.names = F, col.names = T)
  write.table(mt_mat, paste0(filepath, '/mt_count_file.txt'), sep = '\t', quote = F, row.names = F, col.names = T)
}

```

```{r}

directory <- '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/'

for (k in multidim){
  wt_mt_count_file_generator(paste0(directory, k))
}

```

####Create the multidimensional input files for DPClust
- Run in the directory with all the prerequisite files in one place
```{bash}

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input

HOME_DIR='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/'

while read line;
do
set $line
cd ${HOME_DIR}$1
/software/R-3.6.1/bin/Rscript /nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/GetDirichletProcessInfo_multipleSamples_updated_NAP.R \
sample_names.txt \
cellularity.txt \
subclone.files.txt \
mt_count_file.txt \
wt_count_file.txt \
$3 #is.male
done < ${HOME_DIR}patients.txt


```

####Run DPClust on samples

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input

HOME_DIR='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/'

for f in PD42036 PD43299 PD45544 PD46269 PD46966 PD46969;
do
JOB_NAME=${f}_dpclust
SAMPLE_NAME=$f
MEM=60000
CORES=10
QUEUE=long

cd ${HOME_DIR}$f
#-J jobname -M mem -R select host -n cores -q queue name -e stderror putput -o stdout

bsub -G team294-vwork -J"${JOB_NAME}" -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -q ${QUEUE} -e %J.stderr -o %J.stdout \
"xvfb-run -a Rscript --no-save --no-restore --verbose /nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/runDP_new_nonzero_NAP.R sample_names.txt cellularity.txt ${SAMPLE_NAME} > output_multidim_dpclust.Rout 2>&1"

done

#the two most heavily sampled tumours will take considerably longer to finish..
for f in PD43296 PD43298;
do
JOB_NAME=${f}_dpclust
SAMPLE_NAME=$f
MEM=60000
CORES=10
QUEUE=basement

cd ${HOME_DIR}$f
#-J jobname -M mem -R select host -n cores -q queue name -e stderror putput -o stdout

bsub -G team294-vwork -J"${JOB_NAME}" -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -q ${QUEUE} -e %J.stderr -o %J.stdout \
"xvfb-run -a Rscript --no-save --no-restore --verbose /nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/runDP_new_nonzero_NAP.R sample_names.txt cellularity.txt ${SAMPLE_NAME} > output_multidim_dpclust.Rout 2>&1"

done

```

####Run assignment of mutations to clusters per triplet combination of samples (where n > 3) 

Calculate the number of jobs in the job array required.
PD42034 PD42036 PD43296 PD43298 PD43299 PD45543 PD45544 PD45545 PD46269 PD46270 PD46271 PD46966 PD46969 
      1       8      34      26       6       4       9       3       9       1       2       8       6 

PD42036 - 8 samples
PD43296 - 34 samples
PD43298 - 26 samples - 2589 samples with null output
PD43299 - 6 samples
PD45544 - 9 samples
PD46269 - 9 samples
PD46966 - 8 samples
PD46969 - 6 samples

1st run - PD42036, PD45544, PD46966
2nd run - PD46269
3rd run - PD43299, PD46969

```{r}

n = 6 #number of samples you have
choose(n,3) #20 jobs in the array

n = 8 #number of samples you have
choose(n,3) #56 jobs in the array

n = 9 #number of samples you have
choose(n,3) #84 jobs in the array

n = 26 #number of samples you have
choose(n,3) #2600 jobs in the array

n = 34 #number of samples you have
choose(n,3) #5984 jobs in the array

```

Create a text file (triplet_file) with:
patient     no_of_samples     binom_coef

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/

#Run #1

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

HOME_DIR='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/'

infile='triplet_file_run1.txt'
less $infile | while IFS=$'\t' read -r -a tmp;
do

echo ${tmp[0]} #case id
echo ${tmp[1]} #number of samples
echo ${tmp[2]} #binomial coefficient

cd ${HOME_DIR}${tmp[0]}
MEM=30000
QUEUE=normal #may need long queue for 1000s
CORES=10
ARRAY=${tmp[2]} #binomial coefficient
RUN=1 #number of patients
CHOOSE_FROM=${tmp[1]} #number of samples
CHOOSE_NUMBER=3 #number of samples to run in each batch
JOB_NAME="multidim"
PATIENT=${tmp[0]} #case id
SAMPLE_NAMES='sample_names.txt'
CELLULARITY='cellularity.txt'

bsub -G team294-vwork -J"${JOB_NAME}[1-${ARRAY}]%100" -q ${QUEUE} -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -e %I.stderr -o %I.stdout \
"/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/BASH/multidimclustparallel_NAP.sh ${RUN} ${CHOOSE_FROM} ${CHOOSE_NUMBER} ${PATIENT} ${SAMPLE_NAMES} ${CELLULARITY}"

done

#Run 2
export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

HOME_DIR='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/'

infile='triplet_file_run2.txt'
less $infile | while IFS=$'\t' read -r -a tmp;
do

echo ${tmp[0]} #case id
echo ${tmp[1]} #number of samples
echo ${tmp[2]} #binomial coefficient

cd ${HOME_DIR}${tmp[0]}
MEM=30000
QUEUE=normal #may need long queue for 1000s
CORES=10
ARRAY=${tmp[2]} #binomial coefficient
RUN=1 #number of patients
CHOOSE_FROM=${tmp[1]} #number of samples
CHOOSE_NUMBER=3 #number of samples to run in each batch
JOB_NAME="multidim"
PATIENT=${tmp[0]} #case id
SAMPLE_NAMES='sample_names.txt'
CELLULARITY='cellularity.txt'

bsub -G team294-vwork -J"${JOB_NAME}[1-${ARRAY}]%100" -q ${QUEUE} -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -e %I.stderr -o %I.stdout \
"/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/BASH/multidimclustparallel_NAP.sh ${RUN} ${CHOOSE_FROM} ${CHOOSE_NUMBER} ${PATIENT} ${SAMPLE_NAMES} ${CELLULARITY}"

done

#Run 3
export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

HOME_DIR='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/'

infile='triplet_file_run3.txt'
less $infile | while IFS=$'\t' read -r -a tmp;
do

echo ${tmp[0]} #case id
echo ${tmp[1]} #number of samples
echo ${tmp[2]} #binomial coefficient

cd ${HOME_DIR}${tmp[0]}
MEM=30000
QUEUE=normal #may need long queue for 1000s
CORES=10
ARRAY=${tmp[2]} #binomial coefficient
RUN=1 #number of patients
CHOOSE_FROM=${tmp[1]} #number of samples
CHOOSE_NUMBER=3 #number of samples to run in each batch
JOB_NAME="multidim"
PATIENT=${tmp[0]} #case id
SAMPLE_NAMES='sample_names.txt'
CELLULARITY='cellularity.txt'

bsub -G team294-vwork -J"${JOB_NAME}[1-${ARRAY}]%100" -q ${QUEUE} -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -e %I.stderr -o %I.stdout \
"/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/BASH/multidimclustparallel_NAP.sh ${RUN} ${CHOOSE_FROM} ${CHOOSE_NUMBER} ${PATIENT} ${SAMPLE_NAMES} ${CELLULARITY}"

done

#a few job failed on initial run, timed out in normal queue
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

HOME_DIR='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/'

infile='triplet_file_pd43298.txt'
less $infile | while IFS=$'\t' read -r -a tmp;
do

echo ${tmp[0]} #case id
echo ${tmp[1]} #number of samples
echo ${tmp[2]} #binomial coefficient

cd ${HOME_DIR}${tmp[0]}
MEM=30000
QUEUE=long #may need long queue for 1000s
CORES=10
ARRAY=${tmp[2]} #binomial coefficient
RUN=1 #number of patients
CHOOSE_FROM=${tmp[1]} #number of samples
CHOOSE_NUMBER=3 #number of samples to run in each batch
JOB_NAME="multidim"
PATIENT=${tmp[0]} #case id
SAMPLE_NAMES='sample_names.txt'
CELLULARITY='cellularity.txt'

bsub -G team294-vwork -J"${JOB_NAME}[1-${ARRAY}]%200" -q ${QUEUE} -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -e %I.stderr -o %I.stdout \
"/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/BASH/multidimclustparallel_NAP.sh ${RUN} ${CHOOSE_FROM} ${CHOOSE_NUMBER} ${PATIENT} ${SAMPLE_NAMES} ${CELLULARITY}"

done

#PD43296
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

HOME_DIR='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/'

infile='triplet_file_pd43296.txt'
less $infile | while IFS=$'\t' read -r -a tmp;
do

echo ${tmp[0]} #case id
echo ${tmp[1]} #number of samples
echo ${tmp[2]} #binomial coefficient

cd ${HOME_DIR}${tmp[0]}
MEM=30000
QUEUE=normal #may need long queue for 1000s
CORES=10
ARRAY=${tmp[2]} #binomial coefficient
RUN=1 #number of patients
CHOOSE_FROM=${tmp[1]} #number of samples
CHOOSE_NUMBER=3 #number of samples to run in each batch
JOB_NAME="multidim"
PATIENT=${tmp[0]} #case id
SAMPLE_NAMES='sample_names.txt'
CELLULARITY='cellularity.txt'

bsub -G team294-vwork -J"${JOB_NAME}[1-${ARRAY}]%300" -q ${QUEUE} -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -e %I.stderr -o %I.stdout \
"/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/BASH/multidimclustparallel_NAP.sh ${RUN} ${CHOOSE_FROM} ${CHOOSE_NUMBER} ${PATIENT} ${SAMPLE_NAMES} ${CELLULARITY}"

done

#ran out of time in normal queue


#grep "Killed" output_multidim_parallel.*out
#output_multidim_parallel.1051.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 1051 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.1197.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 1197 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.1307.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 1307 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.1831.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 1831 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.238.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 238 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.2412.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 2412 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.3220.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 3220 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.3302.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 3302 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.3432.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 3432 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.3806.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 3806 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.476.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 476 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.565.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 565 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.656.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 656 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.680.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 680 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.763.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 763 PD43296 sample_names.txt cellularity.txt"

#output_multidim_parallel.820.Rout:Killed
bsub -q normal -e %I.stderr -o %I.stdout -n 4 -R"select[mem>20000] span[hosts=1] order[-slots] rusage[mem=20000]" -M20000 "/software/R-3.6.1/lib/R/bin/R --slave --no-restore --no-save --no-restore --file=/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/multidimclustparallel_NAP.R --args 1 34 3 820 PD43296 sample_names.txt cellularity.txt"

```

####Combine clustering assignments across triplets

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/

#Run #1
export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

HOME_DIR='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/'

infile='triplet_file_run1.txt'
less $infile | while IFS=$'\t' read -r -a tmp;
do

echo ${tmp[0]} #case id
echo ${tmp[1]} #number of samples
echo ${tmp[2]} #binomial coefficient

cd ${HOME_DIR}${tmp[0]}
MEM=30000
QUEUE=long #may need long queue for 1000s
CORES=10
ARRAY=${tmp[2]} #binomial coefficient
RUN=1 #number of patients
CHOOSE_FROM=${tmp[1]} #number of samples
CHOOSE_NUMBER=3 #number of samples to run in each batch
JOB_NAME="paraclust"
PATIENT=${tmp[0]} #case id
SAMPLE_NAMES='sample_names.txt'
CELLULARITY='cellularity.txt'

bsub -G team294-vwork -J"${JOB_NAME}" -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -q ${QUEUE} -e %J.stderr -o %J.stdout \
"xvfb-run -a Rscript --no-save --no-restore --verbose /nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/combineClusteringParallelByAssignment_NAP.R \
${RUN} ${CHOOSE_FROM} ${CHOOSE_NUMBER} ${PATIENT} ${SAMPLE_NAMES} ${CELLULARITY} > output_combineClusteringParallelByAssignment.Rout 2>&1"

done

#Run #2
export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

HOME_DIR='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/'

infile='triplet_file_run2.txt'
less $infile | while IFS=$'\t' read -r -a tmp;
do

echo ${tmp[0]} #case id
echo ${tmp[1]} #number of samples
echo ${tmp[2]} #binomial coefficient

cd ${HOME_DIR}${tmp[0]}
MEM=30000
QUEUE=long #may need long queue for 1000s
CORES=10
ARRAY=${tmp[2]} #binomial coefficient
RUN=1 #number of patients
CHOOSE_FROM=${tmp[1]} #number of samples
CHOOSE_NUMBER=3 #number of samples to run in each batch
JOB_NAME="paraclust"
PATIENT=${tmp[0]} #case id
SAMPLE_NAMES='sample_names.txt'
CELLULARITY='cellularity.txt'

bsub -G team294-vwork -J"${JOB_NAME}" -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -q ${QUEUE} -e %J.stderr -o %J.stdout \
"xvfb-run -a Rscript --no-save --no-restore --verbose /nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/combineClusteringParallelByAssignment_NAP.R \
${RUN} ${CHOOSE_FROM} ${CHOOSE_NUMBER} ${PATIENT} ${SAMPLE_NAMES} ${CELLULARITY} > output_combineClusteringParallelByAssignment.Rout 2>&1"

done

#Run #3
export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

HOME_DIR='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/'

infile='triplet_file_run3.txt'
less $infile | while IFS=$'\t' read -r -a tmp;
do

echo ${tmp[0]} #case id
echo ${tmp[1]} #number of samples
echo ${tmp[2]} #binomial coefficient

cd ${HOME_DIR}${tmp[0]}
MEM=30000
QUEUE=long #may need long queue for 1000s
CORES=10
ARRAY=${tmp[2]} #binomial coefficient
RUN=1 #number of patients
CHOOSE_FROM=${tmp[1]} #number of samples
CHOOSE_NUMBER=3 #number of samples to run in each batch
JOB_NAME="paraclust"
PATIENT=${tmp[0]} #case id
SAMPLE_NAMES='sample_names.txt'
CELLULARITY='cellularity.txt'

bsub -G team294-vwork -J"${JOB_NAME}" -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -q ${QUEUE} -e %J.stderr -o %J.stdout \
"xvfb-run -a Rscript --no-save --no-restore --verbose /nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/combineClusteringParallelByAssignment_NAP.R \
${RUN} ${CHOOSE_FROM} ${CHOOSE_NUMBER} ${PATIENT} ${SAMPLE_NAMES} ${CELLULARITY} > output_combineClusteringParallelByAssignment.Rout 2>&1"

done

#re-run for PD43298
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

HOME_DIR='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/'

infile='triplet_file_pd43298.txt'
less $infile | while IFS=$'\t' read -r -a tmp;
do

echo ${tmp[0]} #case id
echo ${tmp[1]} #number of samples
echo ${tmp[2]} #binomial coefficient

cd ${HOME_DIR}${tmp[0]}
MEM=30000
QUEUE=long #may need long queue for 1000s
CORES=10
ARRAY=${tmp[2]} #binomial coefficient
RUN=1 #number of patients
CHOOSE_FROM=${tmp[1]} #number of samples
CHOOSE_NUMBER=3 #number of samples to run in each batch
JOB_NAME="paraclust"
PATIENT=${tmp[0]} #case id
SAMPLE_NAMES='sample_names.txt'
CELLULARITY='cellularity.txt'

bsub -G team294-vwork -J"${JOB_NAME}" -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -q ${QUEUE} -e %J.stderr -o %J.stdout \
"xvfb-run -a Rscript --no-save --no-restore --verbose /nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/combineClusteringParallelByAssignment_NAP.R \
${RUN} ${CHOOSE_FROM} ${CHOOSE_NUMBER} ${PATIENT} ${SAMPLE_NAMES} ${CELLULARITY} > output_combineClusteringParallelByAssignment.Rout 2>&1"

done

#run for PD43296
cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

HOME_DIR='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/'

infile='triplet_file_pd43296.txt'
less $infile | while IFS=$'\t' read -r -a tmp;
do

echo ${tmp[0]} #case id
echo ${tmp[1]} #number of samples
echo ${tmp[2]} #binomial coefficient

cd ${HOME_DIR}${tmp[0]}
MEM=30000
QUEUE=long #may need long queue for 1000s
CORES=10
ARRAY=${tmp[2]} #binomial coefficient
RUN=1 #number of patients
CHOOSE_FROM=${tmp[1]} #number of samples
CHOOSE_NUMBER=3 #number of samples to run in each batch
JOB_NAME="paraclust"
PATIENT=${tmp[0]} #case id
SAMPLE_NAMES='sample_names.txt'
CELLULARITY='cellularity.txt'

bsub -G team294-vwork -J"${JOB_NAME}" -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -q ${QUEUE} -e %J.stderr -o %J.stdout \
"xvfb-run -a Rscript --no-save --no-restore --verbose /nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/combineClusteringParallelByAssignment_NAP.R \
${RUN} ${CHOOSE_FROM} ${CHOOSE_NUMBER} ${PATIENT} ${SAMPLE_NAMES} ${CELLULARITY} > output_combineClusteringParallelByAssignment.Rout 2>&1"

done

```

####Review the output

```{r}

library(ggplot2)
library(ggpubr)
library(reshape2)

patient = c("PD42036", "PD45544", "PD46966", "PD46269", "PD43299", "PD46969", "PD43296", "PD43298")

for(i in patient){
  setwd(paste0('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/', i, '/md_out'))
  mut_clusters = read.table(paste0(i, '_consensusClustersByParallelNodeAssignment.txt'), header = T, sep = '\t', stringsAsFactors = F)
  mut_clusters = mut_clusters[((mut_clusters$no.muts / sum(mut_clusters$no.muts)) >= 0.01) & mut_clusters$no.muts >= 20, ]
  write.table(mut_clusters, paste0(i, '_filtered_clusters.txt'), sep ='\t', row.names = F, col.names = T, quote = F)
  
  clusters = read.table(paste0(i, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)

  clusters = clusters[order(clusters$cluster.no, clusters$chr, clusters$pos),]
  clusters = clusters[clusters$cluster.no %in% names(table(clusters$cluster.no))[table(clusters$cluster.no) >= 20 & table(clusters$cluster.no) > (0.01 * nrow(clusters))],]

  clusters.m = reshape2::melt(clusters, id.vars = c('cluster.no', 'chr', 'pos'))
  
  clusters.l = reshape(aggregate(value ~ cluster.no + variable, clusters.m, median), idvar = "cluster.no", timevar = "variable", direction = "wide")
  clusters.l$mut.no = as.vector(table(clusters$cluster.no))
  colnames(clusters.l) = gsub('value.', '', colnames(clusters.l))
  write.table(clusters.l, paste0(i, '_filtered_clusters_median_CCF.txt'), sep ='\t', row.names = F, col.names = T, quote = F)
  
  clusters.m$cluster.no = as.factor(clusters.m$cluster.no)
  clusters.m$variant = as.factor(paste(clusters.m$cluster.no, clusters.m$chr, clusters.m$pos, sep = '_'))

ggplot(clusters.m) + geom_bar(mapping = aes(x = variant, y = value, fill = cluster.no), stat = 'identity') + facet_grid(variable ~ .) + coord_cartesian(expand = F, ylim = c(0,1)) + theme_pubr() + theme(strip.text.y = element_text(angle = 0), strip.background = element_blank(), axis.text.x = element_blank()) + ggsave(paste0(i, '_cluster_plot.png'), device = 'png')
  
}

```

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/03_DNA_ANALYSES/phylogeny/ndDPClust

cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/PD*/md_out/*_allClusterassignmentsFromParallelRuns.txt .
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/PD*/md_out/*_consensusClustersByParallelNodeAssignment.txt .
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/PD*/md_out/*_filtered_clusters_median_CCF.txt .
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/PD*/md_out/*_cluster_plot.png .
cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/PD*/md_out/*_filtered_clusters.txt .

```


Also see if calculating mean instead is better as there are quite a few clusters that have a median CCF of 0.

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/03_DNA_ANALYSES/phylogeny/ndDPClust")

library(reshape2)

patients = c("PD42036", "PD45544", "PD46966", "PD46269", "PD43299", "PD46969", "PD43296", "PD43298")

for(patient in patients){
  
  clusters = read.table(paste0(patient, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)
  clusters = clusters[order(clusters$cluster.no, clusters$chr, clusters$pos),]
  clusters = clusters[clusters$cluster.no %in% names(table(clusters$cluster.no))[table(clusters$cluster.no) >= 20 & table(clusters$cluster.no) >= (0.01 * nrow(clusters))],]

  clusters.m = reshape2::melt(clusters, id.vars = c('cluster.no', 'chr', 'pos'))
  
  clusters.l = reshape(aggregate(value ~ cluster.no + variable, clusters.m, mean), idvar = "cluster.no", timevar = "variable", direction = "wide")
  clusters.l$mut.no = as.vector(table(clusters$cluster.no))
  colnames(clusters.l) = gsub('value.', '', colnames(clusters.l))
  write.table(clusters.l, paste0(patient, '_filtered_clusters_mean_CCF.txt'), sep ='\t', row.names = F, col.names = T, quote = F)
  
}

```

#####PD45544

Large mutation cluster 3 at lower CCF in three samples. Check mutations.

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/03_DNA_ANALYSES/phylogeny/ndDPClust")

i = "PD45544"
clusters = read.table(paste0(i, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)

#check loci
table(clusters[clusters$cluster.no == 3,]$chr) #not restricted

#now check context - do they have a particular pattern?
poss_artefact_clusters = clusters[clusters$cluster.no %in% c(3),]

#read in subs
all_uniq_var = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/gct_uniq_ssm20210915.txt', header = T, sep = '\t', stringsAsFactors = F)
subs = all_uniq_var[all_uniq_var$Case_ID == i & nchar(all_uniq_var$Wildtype) == 1 & nchar(all_uniq_var$Mutant) == 1,]
#row.names(subs) = paste(subs$Chr, subs$Position, subs$sep = '_')

merged_subs = merge(subs, poss_artefact_clusters[, c(1:2, ncol(poss_artefact_clusters))], by.x = c('Chr', 'Position'), by.y = c('chr', 'pos'))
dim(merged_subs) #228, matches cluster size

library("GenomicRanges")
library("Rsamtools")
library("MASS")

genomeFile <- "/nfs/cancer_ref01/Homo_sapiens/37/genome.fa"

mutations = merged_subs
names(mutations) = c("chr", "pos", "Case_ID", "ref", "mut", "cluster.no")

mutations = mutations[(mutations$ref %in% c("A","C","G","T")) & (mutations$mut %in% c("A","C","G","T")) & mutations$chr %in% c(1:22,"X","Y"),]
mutations$trinuc_ref = as.vector(scanFa(genomeFile, GRanges(mutations$chr, IRanges(as.numeric(mutations$pos)-1,
                                                                                     as.numeric(mutations$pos)+1))))

# 2. Annotating the mutation from the pyrimidine base
ntcomp = c(T="A",G="C",C="G",A="T")
mutations$sub = paste(mutations$ref,mutations$mut,sep=">")
mutations$trinuc_ref_py = mutations$trinuc_ref
for (j in 1:nrow(mutations)) {
  if (mutations$ref[j] %in% c("A","G")) { # Purine base
    mutations$sub[j] = paste(ntcomp[mutations$ref[j]],ntcomp[mutations$mut[j]],sep=">")
    mutations$trinuc_ref_py[j] = paste(ntcomp[rev(strsplit(mutations$trinuc_ref[j],split="")[[1]])],collapse="")
  }
}
freqs = table(paste(mutations$sub,paste(substr(mutations$trinuc_ref_py,1,1),substr(mutations$trinuc_ref_py,3,3),sep="-"),sep=","))
sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")
freqs_full = freqs[full_vec]; freqs_full[is.na(freqs_full)] = 0; names(freqs_full) = full_vec
  
#3. plot
xstr = paste(substr(full_vec,5,5), substr(full_vec,1,1), substr(full_vec,7,7), sep="")

colvec = rep(c("dodgerblue","black","red","grey70","olivedrab3","plum2"),each=16)
y = freqs_full; maxy = max(y)
h = barplot(y, las=2, col=colvec, border=NA, ylim=c(0,maxy*1.5), space=1, cex.names=0.6, names.arg=xstr, ylab="Number of mutations")
for (j in 1:length(sub_vec)) {
  xpos = h[c((j-1)*16+1,j*16)]
  rect(xpos[1]-0.5, maxy*1.2, xpos[2]+0.5, maxy*1.3, border=NA, col=colvec[j*16])
  text(x=mean(xpos), y=maxy*1.3, pos=3, label=sub_vec[j])
}    

#SBS91

table(mutations$cluster.no, mutations$sub)

```

#####PD46269

Large clusters 2 and 3 seen as low CCF across all samples - are they restricted to specific chromosomes or trinucleotide contexts?

 
```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/03_DNA_ANALYSES/phylogeny/ndDPClust")

i = "PD46269"
clusters = read.table(paste0(i, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)

#check loci
table(clusters[clusters$cluster.no == 2,]$chr) #not restricted
table(clusters[clusters$cluster.no == 3,]$chr) #not restricted

#now check context - do they have a particular pattern?
poss_artefact_clusters = clusters[clusters$cluster.no %in% c(2, 3),]

#read in subs
all_uniq_var = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/gct_uniq_ssm20210915.txt', header = T, sep = '\t', stringsAsFactors = F)
subs = all_uniq_var[all_uniq_var$Case_ID == i & nchar(all_uniq_var$Wildtype) == 1 & nchar(all_uniq_var$Mutant) == 1,]
#row.names(subs) = paste(subs$Chr, subs$Position, subs$sep = '_')

merged_subs = merge(subs, poss_artefact_clusters[, c(1:2, ncol(poss_artefact_clusters))], by.x = c('Chr', 'Position'), by.y = c('chr', 'pos'))
dim(merged_subs) #863, matches total size of clusters

library("GenomicRanges")
library("Rsamtools")
library("MASS")

genomeFile <- "/nfs/cancer_ref01/Homo_sapiens/37/genome.fa"

mutations = merged_subs
names(mutations) = c("chr", "pos", "Case_ID", "ref", "mut", "cluster.no")

mutations = mutations[(mutations$ref %in% c("A","C","G","T")) & (mutations$mut %in% c("A","C","G","T")) & mutations$chr %in% c(1:22,"X","Y"),]
mutations$trinuc_ref = as.vector(scanFa(genomeFile, GRanges(mutations$chr, IRanges(as.numeric(mutations$pos)-1,
                                                                                     as.numeric(mutations$pos)+1))))

# 2. Annotating the mutation from the pyrimidine base
ntcomp = c(T="A",G="C",C="G",A="T")
mutations$sub = paste(mutations$ref,mutations$mut,sep=">")
mutations$trinuc_ref_py = mutations$trinuc_ref
for (j in 1:nrow(mutations)) {
  if (mutations$ref[j] %in% c("A","G")) { # Purine base
    mutations$sub[j] = paste(ntcomp[mutations$ref[j]],ntcomp[mutations$mut[j]],sep=">")
    mutations$trinuc_ref_py[j] = paste(ntcomp[rev(strsplit(mutations$trinuc_ref[j],split="")[[1]])],collapse="")
  }
}
freqs = table(paste(mutations$sub,paste(substr(mutations$trinuc_ref_py,1,1),substr(mutations$trinuc_ref_py,3,3),sep="-"),sep=","))
sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")
freqs_full = freqs[full_vec]; freqs_full[is.na(freqs_full)] = 0; names(freqs_full) = full_vec
  
#3. plot
xstr = paste(substr(full_vec,5,5), substr(full_vec,1,1), substr(full_vec,7,7), sep="")

colvec = rep(c("dodgerblue","black","red","grey70","olivedrab3","plum2"),each=16)
y = freqs_full; maxy = max(y)
h = barplot(y, las=2, col=colvec, border=NA, ylim=c(0,maxy*1.5), space=1, cex.names=0.6, names.arg=xstr, ylab="Number of mutations")
for (j in 1:length(sub_vec)) {
  xpos = h[c((j-1)*16+1,j*16)]
  rect(xpos[1]-0.5, maxy*1.2, xpos[2]+0.5, maxy*1.3, border=NA, col=colvec[j*16])
  text(x=mean(xpos), y=maxy*1.3, pos=3, label=sub_vec[j])
}    

#SBS91

table(mutations$cluster.no, mutations$sub)

```

#####PD46966

Cluster 13 is a small cluster with variable mean CCF values per sample - infraclonal in lo0003 and lo0010 but supraclonal in lo0005, lo0006, lo0012 and lo0013.

```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/03_DNA_ANALYSES/phylogeny/ndDPClust")

i = "PD46966"
clusters = read.table(paste0(i, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)

#check loci
table(clusters[clusters$cluster.no == 13,]$chr) #not restricted

probs = read.table(paste0('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/multidimdpclust_noX_new_swater_20220307/01_Input/', i, '/md_out/', i, '_allClusterProbabilitiesFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)
#check loci

hist(merge(probs, clusters[clusters$cluster.no == 13, c(1:2)], by = c('chr', 'pos'))[,'prob.cluster13'])

median(merge(probs, clusters[clusters$cluster.no == 13, c(1:2)], by = c('chr', 'pos'))[,'prob.cluster13'])


```

#####PD46969

Cluster 19 inconsistent with rest of tree - can this be explained by a gain then subsequent loss?

 
```{r}

setwd("/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/03_DNA_ANALYSES/phylogeny/ndDPClust")

i = "PD46969"
clusters = read.table(paste0(i, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)

#check loci
table(clusters[clusters$cluster.no == 19,]$chr) #restricted to chr6 and chr14, consistent with a gain followed by subsequent loss

#d_lo0001 6p 1+1 6q 1+1 14p 1+1 (poss subclonal 2+1) 14q 1+1
#d_lo0003 6p 2+1 6q 1+1 14p 2+1 14q 2+1
#d_lo0005 6p 2+1 6q 1+1 14p 2+1 14q 2+1
#e_lo0001 6p 2+1 6q 2+1 14p 2+1 14q 2+1
#e_lo0002 6p 2+1 6q 1+1 14p 2+1 14q 2+1
#e_lo0005 6p 2+1 6q 2+1 14p 2+1 14q 2+1

```

####DPClust up to 5 samples

Not using sex chromosomes.

#####Prepare data files

Copy across necessary files for samples that meet minimum reads per chromosome copy threshold

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/upto5_dpclust_noX_new_swater/01_Input')

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', header = T, stringsAsFactors = F, sep = '\t')
standard = c("PD42034", "PD45543", "PD45545", "PD46270", "PD46271")

for(i in 1:nrow(manifest)){
  if(manifest$Case.ID[i] %in% standard & manifest$Average.reads.per.chromosome.copy.[i] >= 5){
    system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/', manifest$Sample[i], '_rho_and_psi.txt .'))
    system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/', manifest$Sample[i], '_subclones.txt .'))
    system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/dpclust_qc_input/', manifest$Sample[i], '_cellularity_ploidy.txt .'))
    system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/subs/', manifest$Sample[i], '_post_swater_final_snvs.annot.vcf .'))
  }
}

```

Change column names in vcf to expected one
```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/upto5_dpclust_noX_new_swater/01_Input

for f in $(ls *_post_swater_final_snvs.annot.vcf);
do
sample=${f:0:15}
#echo $sample
sed 's/#CHROM.*/#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tNORMAL\tTUMOUR/' ${sample}_post_swater_final_snvs.annot.vcf > ${sample}.edited.vcf
done 

```

Create separate directories per patient.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/upto5_dpclust_noX_new_swater/01_Input

for f in $(ls *_post_swater_final_snvs.annot.vcf | cut -c1-7 | uniq);
do
mkdir -p $f
mv $f*.* $f
done 

```

Create input files for allelecount
```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/upto5_dpclust_noX_new_swater/01_Input

for f in PD42034 PD45543 PD45545 PD46270 PD46271;
do
cut -f 1,2,4,5 $f/*.edited.vcf | sort -k1,1 -k2,2n -k3,4 | uniq > $f/${f}_loci.txt
grep -v '#' $f/${f}_loci.txt > $f/${f}_loci_SNVs_nohash.txt
mv $f/${f}_loci_SNVs_nohash.txt $f/${f}_loci.txt
done

#remove sex chromosome mutations

for f in PD42034 PD45543 PD45545 PD46270 PD46271;
do
grep -v 'X\|Y' $f/${f}_loci.txt > $f/${f}_loci_noX.txt
mv $f/${f}_loci_noX.txt $f/${f}_loci.txt
done

```

Create a txt file containing a list of the locations of the BAM files. 

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/upto5_dpclust_noX_new_swater/01_Input

for f in $(ls PD*/*loci.txt);
do
patient=$f
dir=${patient%/*}
echo $dir
done >> patients.txt # edit to include the project id as a second column and sex as third column

while read line;
do
set $line
echo $1
echo $2
for f in $(ls $1/*.edited.vcf);
do
#echo $f
filename=$(basename -- "$f")
SAMPLE="${filename%.edited.vcf}"
echo "/lustre/scratch117/casm/team294/rs30/GCT_analyses/Staged_BAM/${2}/${SAMPLE}/mapped_sample/${SAMPLE}.sample.dupmarked.bam" >> $1/${1}_bam_list.txt
done
done < patients.txt

```

Run allelecounter on cgpfarm.
```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/upto5_dpclust_noX_new_swater/01_Input

module load alleleCount

for f in PD42034 PD45543 PD45545 PD46270 PD46271;
do
while read i;
do
sampleID=$(basename $i | sed "s/[.].*//")
echo $sampleID
echo $i
bsub -G team294-vwork -J"${sampleID}" -o ${f}/${sampleID}.o -e ${f}/${sampleID}.e -q normal -n 5 -R"select[mem>10000] rusage[mem=10000] span[hosts=1]" -M10000 "alleleCounter -l ${f}/${f}_loci.txt -b $i -o /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/upto5_dpclust_noX_new_swater/01_Input/${f}/${sampleID}_alleleFrequencies.txt -m 25 -q 30"
done < ${f}/${f}_bam_list.txt 
done

```

Final step to generate the _allDirichletProcessInfo.txt files.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/upto5_dpclust_noX_new_swater/01_Input

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"

/software/R-3.6.1/bin/Rscript /nfs/users/nfs_t/to3/scripts/allDirichletProcessInfo_generator_20210604.R /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/upto5_dpclust_noX_new_swater/01_Input

```


```{r}

#allDirichletProcessInfo_generator_20210604.R

library(dpclust3p)
library(optparse)

args <- commandArgs(TRUE)
options(stringsAsFactors = F)

INPUT_PATH <- args[1] #parent directory

data = read.table(paste0(INPUT_PATH, '/patients.txt'), header = F, sep = '\t')
names(data) = c('Patient', 'ProjectID', 'Sex')

for( k in 1:nrow(data) ){
  samples = unlist(strsplit(list.files(path = paste0(data$Patient[k]), pattern = '_cellularity_ploidy.txt'), '_cellularity_ploidy.txt'))
  for(i in 1:length(samples)){
    runGetDirichletProcessInfo(loci_file = paste0(INPUT_PATH, '/', data$Patient[k], '/', data$Patient[k], '_loci.txt'), 
                               allele_frequencies_file = paste0(INPUT_PATH, '/', data$Patient[k], '/', samples[i], '_alleleFrequencies.txt'), 
                               cellularity_file = paste0(INPUT_PATH, '/', data$Patient[k], '/', samples[i], '_rho_and_psi.txt'), 
                               subclone_file = paste0(INPUT_PATH, '/', data$Patient[k], '/', samples[i], '_subclones.txt'), 
                               gender = data$Sex[k], 
                               SNP.phase.file="NA", 
                               mut.phase.file="NA", 
                               output_file = paste0(INPUT_PATH, '/', data$Patient[k], '/', samples[i], '_allDirichletProcessInfo.txt'))
  }
}

```

- Create a master project file to put into the pipeline for each sample.

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/upto5_dpclust_noX_new_swater/01_Input

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"

/software/R-3.6.1/bin/Rscript /nfs/users/nfs_t/to3/scripts/createProjectFile_generator_20210604.R /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/upto5_dpclust_noX_new_swater/01_Input

```


- This is more convoluted than the creating a file on a per sample basis and can even be used to submit samples from multiple donors simultaneously - you need to repeat each value to match up the lengths of your comma separated lists

```{r}

library(dpclust3p)
library(optparse)

args <- commandArgs(TRUE)
options(stringsAsFactors = F)

INPUT_PATH <- args[1] #parent directory

data = read.table(paste0(INPUT_PATH, '/patients.txt'), header = F, sep = '\t')
names(data) = c('Patient', 'ProjectID', 'Sex')

for( k in 1:nrow(data) ){
  samples = unlist(strsplit(list.files(path = paste0(data$Patient[k]), pattern = '_cellularity_ploidy.txt'), '_cellularity_ploidy.txt'))
  createProjectFile(outputfile = paste0(data$Patient[k], '_ProjectFile.txt'), 
                    donornames = rep(data$Patient[k], length(samples)), 
                    samplenames = sapply(strsplit(samples, '_'), "[[", 2), 
                    rho_and_psi_files = paste0(data$Patient[k], '/', samples, '_rho_and_psi.txt'),
                    sex = rep(data$Sex[k],  length(samples)), 
                    datafiles = paste0(data$Patient[k], '/', samples, '_allDirichletProcessInfo.txt'))
}

```

#####Run multi-sample DPClust

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/upto5_dpclust_noX_new_swater/01_Input

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"
INPUT_DIR='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/upto5_dpclust_noX_new_swater/01_Input/'
OUTPUT_DIR='/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/upto5_dpclust_noX_new_swater/03_Output/'
MEM=30000
QUEUE=normal
n_its=3000
burnin=1000
CORES=10

for f in PD42034 PD45543 PD45545 PD46270 PD46271;
do
bsub -G team294-vwork -J${f} -q ${QUEUE} -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -e %I.stderr -o %I.stdout \
"xvfb-run -a Rscript --no-save --no-restore --verbose /nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw/DPClust/example/dpclust_pipeline.R -r 1 -d ${INPUT_DIR} -o $OUTPUT_DIR -i ${INPUT_DIR}/${f}_ProjectFile.txt -k -a nd_dp --iterations $n_its --burnin $burnin --mut_assignment_type 1 --min_muts_cluster 20 --min_frac_muts_cluster 0.01 --num_muts_sample 4000 --bin_size NA --seed 123 --assign_sampled_muts T > ${OUTPUT_DIR}${f}_outputFile2.Rout 2>&1"
done

```

#04_RNA_ANALYSES

##Limma-voom DE analysis

###Per histology (vs EC & seminiferous tubules)

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom/')

library(limma)
library(edgeR)

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #just normal testis and GCT raw count data, 500 microbiopsies
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #read in metadata for filtered, final cuts
row.names(metadata) = metadata$Sample
gct_mat = raw_data[, 7:ncol(raw_data)]
row.names(gct_mat) = raw_data$Geneid

#remove cuts that failed QC
gct_mat = gct_mat[, colnames(gct_mat) %in% metadata$Sample] #416 samples

#define factors that will be input into the model
histology = factor(metadata[colnames(gct_mat), ]$Updated.Description)
patient = factor(metadata[colnames(gct_mat), ]$Case.ID)
metadata = metadata[colnames(gct_mat),]

#preprocessing
y <- DGEList(counts = gct_mat, group = histology)
#keep <- filterByExpr.default(y, group = histology)
keep <- filterByExpr(y, group = histology)
y <- y[keep, , keep.lib.sizes=FALSE]
y <- calcNormFactors(y) #33185 genes kept

plotMDS(y, col=as.numeric(patient), labels = histology) #no obvious outlier samples

#library sizes are variable for voom is preferred over limma-trend
#specify model to be fitted, no intercept to be fitted, each histology will have its own group mean
design = model.matrix(~ 0 + histology)
colnames(design) = levels(histology)

contr.matrix <- makeContrasts(
  fat_v_ec = Adipose_teratoma - Embryonal_carcinoma,
  dys_v_ec = Dysgerminoma - Embryonal_carcinoma,
  gcnis_v_ec = GCNIS - Embryonal_carcinoma,
  cartilage_v_ec = Cartilage_teratoma - Embryonal_carcinoma, 
  glandular_v_ec = Mature_glandular_teratoma - Embryonal_carcinoma, 
  ne_v_ec = Neuroepithelium - Embryonal_carcinoma, 
  other_ter_v_ec = Other_epithelial_teratoma - Embryonal_carcinoma, 
  muscle_v_ec = Smooth_muscle_teratoma - Embryonal_carcinoma, 
  sct_v_ec = Syncytiotrophoblasts - Embryonal_carcinoma, 
  yst_v_ec = Yolk_sac_tumour - Embryonal_carcinoma, 
  sem_v_ec = Seminoma - Embryonal_carcinoma,
  str_v_ec = Stroma - Embryonal_carcinoma,
  fat_v_st = Adipose_teratoma - Seminiferous_tubule,
  dys_v_st = Dysgerminoma - Seminiferous_tubule,
  ec_v_st = Embryonal_carcinoma - Seminiferous_tubule,
  gcnis_v_st = GCNIS - Seminiferous_tubule,
  cartilage_v_st = Cartilage_teratoma - Seminiferous_tubule, 
  glandular_v_st = Mature_glandular_teratoma - Seminiferous_tubule, 
  ne_v_st = Neuroepithelium - Seminiferous_tubule, 
  other_ter_v_st = Other_epithelial_teratoma - Seminiferous_tubule, 
  muscle_v_st = Smooth_muscle_teratoma - Seminiferous_tubule, 
  sct_v_st = Syncytiotrophoblasts - Seminiferous_tubule, 
  yst_v_st = Yolk_sac_tumour - Seminiferous_tubule, 
  sem_v_st = Seminoma - Seminiferous_tubule,
  str_v_st = Stroma - Seminiferous_tubule,
  levels = colnames(design))

v <- voom(y, design, plot=TRUE)
vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)
efit <- eBayes(vfit)
plotSA(efit, main="Final model: Mean-variance trend")

saveRDS(v, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom/limma_voom_de_LCM_GCT_DE_voom_20210925.rds')
saveRDS(efit, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom/limma_voom_de_LCM_GCT_DE_model_fit_20210925.rds')

for(i in colnames(contr.matrix)){
  data = topTable(efit, coef = i, n = Inf, sort = "p")
  write.table(data, paste0(i, "_voom_de_20210925.txt"), col.names = T, row.names = T, sep = '\t', quote = F)
}

```

####Benchmark against EMDomics

```{bash}

cd /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics

Rscript ec_vs_ct_EMDomics_20210921.R

bsub -G team294-vwork -M60000 -R"select[mem>60000] rusage[mem=60000] span[hosts=1]" -n 4 -q long -e %J.stderr -o %J.stdout "/software/R-4.1.0/bin/Rscript ec_vs_ct_EMDomics_20210921.R > ec_vs_ct_EMDomics_20210921.outputFile.Rout 2>&1"
bsub -G team294-vwork -M60000 -R"select[mem>60000] rusage[mem=60000] span[hosts=1]" -n 4 -q long -e %J.stderr -o %J.stdout "/software/R-4.1.0/bin/Rscript ec_vs_ne_EMDomics_20210921.R > ec_vs_ne_EMDomics_20210921.outputFile.Rout 2>&1"
bsub -G team294-vwork -M60000 -R"select[mem>60000] rusage[mem=60000] span[hosts=1]" -n 4 -q long -e %J.stderr -o %J.stdout "/software/R-4.1.0/bin/Rscript ec_vs_sm_EMDomics_20210921.R > ec_vs_sm_EMDomics_20210921.outputFile.Rout 2>&1"
bsub -G team294-vwork -M60000 -R"select[mem>60000] rusage[mem=60000] span[hosts=1]" -n 4 -q long -e %J.stderr -o %J.stdout "/software/R-4.1.0/bin/Rscript ec_vs_sem_EMDomics_20210921.R > ec_vs_sem_EMDomics_20210921.outputFile.Rout 2>&1"
bsub -G team294-vwork -M60000 -R"select[mem>60000] rusage[mem=60000] span[hosts=1]" -n 4 -q long -e %J.stderr -o %J.stdout "/software/R-4.1.0/bin/Rscript ec_vs_str_EMDomics_20210921.R > ec_vs_str_EMDomics_20210921.outputFile.Rout 2>&1"
bsub -G team294-vwork -M60000 -R"select[mem>60000] rusage[mem=60000] span[hosts=1]" -n 4 -q long -e %J.stderr -o %J.stdout "/software/R-4.1.0/bin/Rscript ec_vs_yst_EMDomics_20210921.R > ec_vs_yst_EMDomics_20210921.outputFile.Rout 2>&1"
bsub -G team294-vwork -M60000 -R"select[mem>60000] rusage[mem=60000] span[hosts=1]" -n 4 -q long -e %J.stderr -o %J.stdout "/software/R-4.1.0/bin/Rscript ec_vs_gc_EMDomics_20210921.R > ec_vs_gc_EMDomics_20210921.outputFile.Rout 2>&1"

```

#####EMDomics scripts

```{r}

#ec_vs_ct_EMDomics_20210921.R
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics')

library(EMDomics)

edger_pseudo_counts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/lcm_rna_tmm_norm_counts_20210812.txt', sep = '\t', header = T, stringsAsFactors = F)

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #just normal testis and GCT raw count data, 500 microbiopsies
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #read in metadata for filtered, final cuts
row.names(metadata) = metadata$Sample
gct_mat = raw_data[, 7:ncol(raw_data)]
row.names(gct_mat) = raw_data$Geneid

#remove cuts that failed QC
gct_mat = gct_mat[, colnames(gct_mat) %in% metadata$Sample] #416 samples
gct_mat = gct_mat[row.names(edger_pseudo_counts),]

#EC vs CT
ec_vs_ct = metadata[metadata$Updated.Description %in% c('Embryonal_carcinoma', 'Cartilage_teratoma'),]
ec_vs_ct_counts = gct_mat[,ec_vs_ct$Sample]
ec_vs_ct_counts = ec_vs_ct_counts[rowSums(ec_vs_ct_counts == 0) != ncol(ec_vs_ct_counts),]
histology = ec_vs_ct[colnames(ec_vs_ct_counts), ]$Updated.Description
names(histology) = colnames(ec_vs_ct_counts)
results = calculate_emd(ec_vs_ct_counts, histology, nperm = 100, parallel = T, seq = T, verbose = T, binSize = 1)

saveRDS(results, 'ec_vs_ct_EMDomics_20210921.rds')

```

```{r}

#ec_vs_ne_EMDomics_20210921.R
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics')

library(EMDomics)

edger_pseudo_counts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/lcm_rna_tmm_norm_counts_20210812.txt', sep = '\t', header = T, stringsAsFactors = F)

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #just normal testis and GCT raw count data, 500 microbiopsies
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #read in metadata for filtered, final cuts
row.names(metadata) = metadata$Sample
gct_mat = raw_data[, 7:ncol(raw_data)]
row.names(gct_mat) = raw_data$Geneid

#remove cuts that failed QC
gct_mat = gct_mat[, colnames(gct_mat) %in% metadata$Sample] #416 samples
gct_mat = gct_mat[row.names(edger_pseudo_counts),]

#EC vs NE
ec_vs_ne = metadata[metadata$Updated.Description %in% c('Embryonal_carcinoma', 'Neuroepithelium'),]
ec_vs_ne_counts = gct_mat[,ec_vs_ne$Sample]
ec_vs_ne_counts = ec_vs_ne_counts[rowSums(ec_vs_ne_counts == 0) != ncol(ec_vs_ne_counts),]
histology = ec_vs_ne[colnames(ec_vs_ne_counts), ]$Updated.Description
names(histology) = colnames(ec_vs_ne_counts)
results = calculate_emd(ec_vs_ne_counts, histology, nperm = 100, parallel = T, seq = T, verbose = T, binSize = 1)

saveRDS(results, 'ec_vs_ne_EMDomics_20210921.rds')

```

```{r}

#ec_vs_sm_EMDomics_20210921.R
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics')

library(EMDomics)

edger_pseudo_counts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/lcm_rna_tmm_norm_counts_20210812.txt', sep = '\t', header = T, stringsAsFactors = F)

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #just normal testis and GCT raw count data, 500 microbiopsies
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #read in metadata for filtered, final cuts
row.names(metadata) = metadata$Sample
gct_mat = raw_data[, 7:ncol(raw_data)]
row.names(gct_mat) = raw_data$Geneid

#remove cuts that failed QC
gct_mat = gct_mat[, colnames(gct_mat) %in% metadata$Sample] #416 samples
gct_mat = gct_mat[row.names(edger_pseudo_counts),]

#EC vs SM
ec_vs_sm = metadata[metadata$Updated.Description %in% c('Embryonal_carcinoma', 'Smooth_muscle_teratoma'),]
ec_vs_sm_counts = gct_mat[,ec_vs_sm$Sample]
ec_vs_sm_counts = ec_vs_sm_counts[rowSums(ec_vs_sm_counts == 0) != ncol(ec_vs_sm_counts),]
histology = ec_vs_sm[colnames(ec_vs_sm_counts), ]$Updated.Description
names(histology) = colnames(ec_vs_sm_counts)
results = calculate_emd(ec_vs_sm_counts, histology, nperm = 100, parallel = T, seq = T, verbose = T, binSize = 1)

saveRDS(results, 'ec_vs_sm_EMDomics_20210921.rds')

```

```{r}

#ec_vs_at_EMDomics_20210921.R
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics')

library(EMDomics)

edger_pseudo_counts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/lcm_rna_tmm_norm_counts_20210812.txt', sep = '\t', header = T, stringsAsFactors = F)

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #just normal testis and GCT raw count data, 500 microbiopsies
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #read in metadata for filtered, final cuts
row.names(metadata) = metadata$Sample
gct_mat = raw_data[, 7:ncol(raw_data)]
row.names(gct_mat) = raw_data$Geneid

#remove cuts that failed QC
gct_mat = gct_mat[, colnames(gct_mat) %in% metadata$Sample] #416 samples
gct_mat = gct_mat[row.names(edger_pseudo_counts),]

#EC vs SM
ec_vs_at = metadata[metadata$Updated.Description %in% c('Embryonal_carcinoma', 'Adipose_teratoma'),]
ec_vs_at_counts = gct_mat[,ec_vs_at$Sample]
ec_vs_at_counts = ec_vs_at_counts[rowSums(ec_vs_at_counts == 0) != ncol(ec_vs_at_counts),]
histology = ec_vs_at[colnames(ec_vs_at_counts), ]$Updated.Description
names(histology) = colnames(ec_vs_at_counts)
results = calculate_emd(ec_vs_at_counts, histology, nperm = 100, parallel = T, seq = T, verbose = T, binSize = 1)

saveRDS(results, 'ec_vs_at_EMDomics_20210921.rds')

```

```{r}

#ec_vs_ds_EMDomics_20210921.R
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics')

library(EMDomics)

edger_pseudo_counts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/lcm_rna_tmm_norm_counts_20210812.txt', sep = '\t', header = T, stringsAsFactors = F)

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #just normal testis and GCT raw count data, 500 microbiopsies
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #read in metadata for filtered, final cuts
row.names(metadata) = metadata$Sample
gct_mat = raw_data[, 7:ncol(raw_data)]
row.names(gct_mat) = raw_data$Geneid

#remove cuts that failed QC
gct_mat = gct_mat[, colnames(gct_mat) %in% metadata$Sample] #416 samples
gct_mat = gct_mat[row.names(edger_pseudo_counts),]

#EC vs DS
ec_vs_ds = metadata[metadata$Updated.Description %in% c('Embryonal_carcinoma', 'Dysgerminoma'),]
ec_vs_ds_counts = gct_mat[,ec_vs_ds$Sample]
ec_vs_ds_counts = ec_vs_ds_counts[rowSums(ec_vs_ds_counts == 0) != ncol(ec_vs_ds_counts),]
histology = ec_vs_ds[colnames(ec_vs_ds_counts), ]$Updated.Description
names(histology) = colnames(ec_vs_ds_counts)
results = calculate_emd(ec_vs_ds_counts, histology, nperm = 100, parallel = T, seq = T, verbose = T, binSize = 1)

saveRDS(results, 'ec_vs_ds_EMDomics_20210921.rds')

```

```{r}

#ec_vs_gc_EMDomics_20210921.R
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics')

library(EMDomics)

edger_pseudo_counts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/lcm_rna_tmm_norm_counts_20210812.txt', sep = '\t', header = T, stringsAsFactors = F)

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #just normal testis and GCT raw count data, 500 microbiopsies
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #read in metadata for filtered, final cuts
row.names(metadata) = metadata$Sample
gct_mat = raw_data[, 7:ncol(raw_data)]
row.names(gct_mat) = raw_data$Geneid

#remove cuts that failed QC
gct_mat = gct_mat[, colnames(gct_mat) %in% metadata$Sample] #416 samples
gct_mat = gct_mat[row.names(edger_pseudo_counts),]

#EC vs GCNIS
ec_vs_gc = metadata[metadata$Updated.Description %in% c('Embryonal_carcinoma', 'GCNIS'),]
ec_vs_gc_counts = gct_mat[,ec_vs_gc$Sample]
ec_vs_gc_counts = ec_vs_gc_counts[rowSums(ec_vs_gc_counts == 0) != ncol(ec_vs_gc_counts),]
histology = ec_vs_gc[colnames(ec_vs_gc_counts), ]$Updated.Description
names(histology) = colnames(ec_vs_gc_counts)
results = calculate_emd(ec_vs_gc_counts, histology, nperm = 100, parallel = T, seq = T, verbose = T, binSize = 1)

saveRDS(results, 'ec_vs_gc_EMDomics_20210921.rds')

```

```{r}

#ec_vs_mg_EMDomics_20210921.R
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics')

library(EMDomics)

edger_pseudo_counts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/lcm_rna_tmm_norm_counts_20210812.txt', sep = '\t', header = T, stringsAsFactors = F)

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #just normal testis and GCT raw count data, 500 microbiopsies
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #read in metadata for filtered, final cuts
row.names(metadata) = metadata$Sample
gct_mat = raw_data[, 7:ncol(raw_data)]
row.names(gct_mat) = raw_data$Geneid

#remove cuts that failed QC
gct_mat = gct_mat[, colnames(gct_mat) %in% metadata$Sample] #416 samples
gct_mat = gct_mat[row.names(edger_pseudo_counts),]

#EC vs GCNIS
ec_vs_mg = metadata[metadata$Updated.Description %in% c('Embryonal_carcinoma', 'Mature_glandular_teratoma'),]
ec_vs_mg_counts = gct_mat[,ec_vs_mg$Sample]
ec_vs_mg_counts = ec_vs_mg_counts[rowSums(ec_vs_mg_counts == 0) != ncol(ec_vs_mg_counts),]
histology = ec_vs_mg[colnames(ec_vs_mg_counts), ]$Updated.Description
names(histology) = colnames(ec_vs_mg_counts)
results = calculate_emd(ec_vs_mg_counts, histology, nperm = 100, parallel = T, seq = T, verbose = T, binSize = 1)

saveRDS(results, 'ec_vs_mg_EMDomics_20210921.rds')

```

```{r}

#ec_vs_sem_EMDomics_20210921.R
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics')

library(EMDomics)

edger_pseudo_counts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/lcm_rna_tmm_norm_counts_20210812.txt', sep = '\t', header = T, stringsAsFactors = F)

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #just normal testis and GCT raw count data, 500 microbiopsies
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #read in metadata for filtered, final cuts
row.names(metadata) = metadata$Sample
gct_mat = raw_data[, 7:ncol(raw_data)]
row.names(gct_mat) = raw_data$Geneid

#remove cuts that failed QC
gct_mat = gct_mat[, colnames(gct_mat) %in% metadata$Sample] #416 samples
gct_mat = gct_mat[row.names(edger_pseudo_counts),]

#EC vs Seminoma
ec_vs_sem = metadata[metadata$Updated.Description %in% c('Embryonal_carcinoma', 'Seminoma'),]
ec_vs_sem_counts = gct_mat[,ec_vs_sem$Sample]
ec_vs_sem_counts = ec_vs_sem_counts[rowSums(ec_vs_sem_counts == 0) != ncol(ec_vs_sem_counts),]
histology = ec_vs_sem[colnames(ec_vs_sem_counts), ]$Updated.Description
names(histology) = colnames(ec_vs_sem_counts)
results = calculate_emd(ec_vs_sem_counts, histology, nperm = 100, parallel = T, seq = T, verbose = T, binSize = 1)

saveRDS(results, 'ec_vs_sem_EMDomics_20210921.rds')

```

```{r}

#ec_vs_str_EMDomics_20210921.R
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics')

library(EMDomics)

edger_pseudo_counts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/lcm_rna_tmm_norm_counts_20210812.txt', sep = '\t', header = T, stringsAsFactors = F)

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #just normal testis and GCT raw count data, 500 microbiopsies
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #read in metadata for filtered, final cuts
row.names(metadata) = metadata$Sample
gct_mat = raw_data[, 7:ncol(raw_data)]
row.names(gct_mat) = raw_data$Geneid

#remove cuts that failed QC
gct_mat = gct_mat[, colnames(gct_mat) %in% metadata$Sample] #416 samples
gct_mat = gct_mat[row.names(edger_pseudo_counts),]

#EC vs Stroma
ec_vs_str = metadata[metadata$Updated.Description %in% c('Embryonal_carcinoma', 'Stroma'),]
ec_vs_str_counts = gct_mat[,ec_vs_str$Sample]
ec_vs_str_counts = ec_vs_str_counts[rowSums(ec_vs_str_counts == 0) != ncol(ec_vs_str_counts),]
histology = ec_vs_str[colnames(ec_vs_str_counts), ]$Updated.Description
names(histology) = colnames(ec_vs_str_counts)
results = calculate_emd(ec_vs_str_counts, histology, nperm = 100, parallel = T, seq = T, verbose = T, binSize = 1)

saveRDS(results, 'ec_vs_str_EMDomics_20210921.rds')

```

```{r}

#ec_vs_yst_EMDomics_20210921.R
setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics')

library(EMDomics)

edger_pseudo_counts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/lcm_rna_tmm_norm_counts_20210812.txt', sep = '\t', header = T, stringsAsFactors = F)

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #just normal testis and GCT raw count data, 500 microbiopsies
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #read in metadata for filtered, final cuts
row.names(metadata) = metadata$Sample
gct_mat = raw_data[, 7:ncol(raw_data)]
row.names(gct_mat) = raw_data$Geneid

#remove cuts that failed QC
gct_mat = gct_mat[, colnames(gct_mat) %in% metadata$Sample] #416 samples
gct_mat = gct_mat[row.names(edger_pseudo_counts),]

#EC vs YST
ec_vs_yst = metadata[metadata$Updated.Description %in% c('Embryonal_carcinoma', 'Yolk_sac_tumour'),]
ec_vs_yst_counts = gct_mat[,ec_vs_yst$Sample]
ec_vs_yst_counts = ec_vs_yst_counts[rowSums(ec_vs_yst_counts == 0) != ncol(ec_vs_yst_counts),]
histology = ec_vs_yst[colnames(ec_vs_yst_counts), ]$Updated.Description
names(histology) = colnames(ec_vs_yst_counts)
results = calculate_emd(ec_vs_yst_counts, histology, nperm = 100, parallel = T, seq = T, verbose = T, binSize = 1)

saveRDS(results, 'ec_vs_yst_EMDomics_20210921.rds')

```


#####Compare significant genes between both methods

```{r}

library(ggvenn)
library(cowplot)

edger_pseudo_counts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/lcm_rna_tmm_norm_counts_20210812.txt', sep = '\t', header = T, stringsAsFactors = F)
all_genes = row.names(edger_pseudo_counts)

#Cartilage teratoma
limma_results = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom/cartilage_v_ec_voom_de_20210925.txt', header = T, sep = '\t', stringsAsFactors = F)
emdomics_rds = readRDS('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics/ec_vs_ct_EMDomics_20210921.rds')
emd <- emdomics_rds$emd
emdomics_results <- data.frame(emd[(order(emd[,"q-value"])),])

limma_sig_genes = row.names(limma_results[limma_results$adj.P.Val <0.01 & (limma_results$logFC > 2 | limma_results$logFC < -2),])
emdomics_sig_genes = row.names(emdomics_results[emdomics_results$q.value <0.05,])

comparison = list(first.vector = limma_sig_genes, second.vector = emdomics_sig_genes, third.vector = all_genes)
names(comparison) = c("Limma (q <0.01 &\n > |2| log2FC)", "EMDomics (q <0.05)", "All minimally expressed genes")

ct_plot = ggvenn(comparison, fill_color = c("#0073C2FF", "#EFC000FF", "#CD534CFF"), stroke_size = 0.5, set_name_size = 4) + ggtitle("Cartilage teratoma vs Embryonal carcinoma") + theme(plot.title = element_text(hjust = 0.5, face = "bold"))

#Neuroepithelium
limma_results = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom/ne_v_ec_voom_de_20210925.txt', header = T, sep = '\t', stringsAsFactors = F)
emdomics_rds = readRDS('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics/ec_vs_ne_EMDomics_20210921.rds')
emd <- emdomics_rds$emd
emdomics_results <- data.frame(emd[(order(emd[,"q-value"])),])

limma_sig_genes = row.names(limma_results[limma_results$adj.P.Val <0.01 & (limma_results$logFC > 2 | limma_results$logFC < -2),])
emdomics_sig_genes = row.names(emdomics_results[emdomics_results$q.value <0.05,])

comparison = list(first.vector = limma_sig_genes, second.vector = emdomics_sig_genes, third.vector = all_genes)
names(comparison) = c("Limma (q <0.01 &\n > |2| log2FC)", "EMDomics (q <0.05)", "All minimally expressed genes")

ne_plot = ggvenn(comparison, fill_color = c("#0073C2FF", "#EFC000FF", "#CD534CFF"), stroke_size = 0.5, set_name_size = 4) + ggtitle("Neuroepithelium vs Embryonal carcinoma") + theme(plot.title = element_text(hjust = 0.5, face = "bold"))

#Smooth muscle
limma_results = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom/muscle_v_ec_voom_de_20210925.txt', header = T, sep = '\t', stringsAsFactors = F)
emdomics_rds = readRDS('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics/ec_vs_sm_EMDomics_20210921.rds')
emd <- emdomics_rds$emd
emdomics_results <- data.frame(emd[(order(emd[,"q-value"])),])

limma_sig_genes = row.names(limma_results[limma_results$adj.P.Val <0.01 & (limma_results$logFC > 2 | limma_results$logFC < -2),])
emdomics_sig_genes = row.names(emdomics_results[emdomics_results$q.value <0.05,])

comparison = list(first.vector = limma_sig_genes, second.vector = emdomics_sig_genes, third.vector = all_genes)
names(comparison) = c("Limma (q <0.01 &\n > |2| log2FC)", "EMDomics (q <0.05)", "All minimally expressed genes")

sm_plot = ggvenn(comparison, fill_color = c("#0073C2FF", "#EFC000FF", "#CD534CFF"), stroke_size = 0.5, set_name_size = 4) + ggtitle("Smooth muscle teratoma vs Embryonal carcinoma") + theme(plot.title = element_text(hjust = 0.5, face = "bold"))

#Seminoma
limma_results = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom/sem_v_ec_voom_de_20210925.txt', header = T, sep = '\t', stringsAsFactors = F)
emdomics_rds = readRDS('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics/ec_vs_sem_EMDomics_20210921.rds')
emd <- emdomics_rds$emd
emdomics_results <- data.frame(emd[(order(emd[,"q-value"])),])

limma_sig_genes = row.names(limma_results[limma_results$adj.P.Val <0.01 & (limma_results$logFC > 2 | limma_results$logFC < -2),])
emdomics_sig_genes = row.names(emdomics_results[emdomics_results$q.value <0.05,])

comparison = list(first.vector = limma_sig_genes, second.vector = emdomics_sig_genes, third.vector = all_genes)
names(comparison) = c("Limma (q <0.01 &\n > |2| log2FC)", "EMDomics (q <0.05)", "All minimally expressed genes")

sem_plot = ggvenn(comparison, fill_color = c("#0073C2FF", "#EFC000FF", "#CD534CFF"), stroke_size = 0.5, set_name_size = 4) + ggtitle("Seminoma vs Embryonal carcinoma") + theme(plot.title = element_text(hjust = 0.5, face = "bold"))

#Yolk sac tumour
limma_results = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom/yst_v_ec_voom_de_20210925.txt', header = T, sep = '\t', stringsAsFactors = F)
emdomics_rds = readRDS('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics/ec_vs_yst_EMDomics_20210921.rds')
emd <- emdomics_rds$emd
emdomics_results <- data.frame(emd[(order(emd[,"q-value"])),])

limma_sig_genes = row.names(limma_results[limma_results$adj.P.Val <0.05,])
limma_sig_genes = row.names(limma_results[limma_results$adj.P.Val <0.01 & (limma_results$logFC > 2 | limma_results$logFC < -2),])
emdomics_sig_genes = row.names(emdomics_results[emdomics_results$q.value <0.05,])

comparison = list(first.vector = limma_sig_genes, second.vector = emdomics_sig_genes, third.vector = all_genes)
names(comparison) = c("Limma (q <0.01 &\n > |2| log2FC)", "EMDomics (q <0.05)", "All minimally expressed genes")

yst_plot = ggvenn(comparison, fill_color = c("#0073C2FF", "#EFC000FF", "#CD534CFF"), stroke_size = 0.5, set_name_size = 4) + ggtitle("Yolk sac tumour vs Embryonal carcinoma") + theme(plot.title = element_text(hjust = 0.5, face = "bold"))

#Stroma
limma_results = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom/str_v_ec_voom_de_20210925.txt', header = T, sep = '\t', stringsAsFactors = F)
emdomics_rds = readRDS('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/EMDomics/ec_vs_str_EMDomics_20210921.rds')
emd <- emdomics_rds$emd
emdomics_results <- data.frame(emd[(order(emd[,"q-value"])),])

limma_sig_genes = row.names(limma_results[limma_results$adj.P.Val <0.05,])
limma_sig_genes = row.names(limma_results[limma_results$adj.P.Val <0.01 & (limma_results$logFC > 2 | limma_results$logFC < -2),])
emdomics_sig_genes = row.names(emdomics_results[emdomics_results$q.value <0.05,])

comparison = list(first.vector = limma_sig_genes, second.vector = emdomics_sig_genes, third.vector = all_genes)
names(comparison) = c("Limma (q <0.01 &\n > |2| log2FC)", "EMDomics (q <0.05)", "All minimally expressed genes")

str_plot = ggvenn(comparison, fill_color = c("#0073C2FF", "#EFC000FF", "#CD534CFF"), stroke_size = 0.5, set_name_size = 4) + ggtitle("Malignant stroma vs Embryonal carcinoma") + theme(plot.title = element_text(hjust = 0.5, face = "bold"))

#plot
pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/GCT_benchmarking_limma_vs_EMDomics_Venn_diagrams.20210925.pdf', height = 12, width = 6)
plot_grid(ct_plot, ne_plot, sm_plot, sem_plot, yst_plot, str_plot, ncol = 2, labels="auto")
dev.off()



```

####Example volcano plots (Fig. 3d, e and f)

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom')

library(EnhancedVolcano)
library(cowplot)

#read in example DE analyses
ne_v_ec = read.table('ne_v_ec_voom_de_20210925.txt', header = T, sep = '\t', stringsAsFactors = F)
sm_v_ec = read.table('muscle_v_ec_voom_de_20210925.txt', header = T, sep = '\t', stringsAsFactors = F)
ct_v_ec = read.table('cartilage_v_ec_voom_de_20210925.txt', header = T, sep = '\t', stringsAsFactors = F)

ne_v_ec$gene = row.names(ne_v_ec)
ggplot(ne_v_ec[(ne_v_ec$logFC > -2 & ne_v_ec$logFC < 2) | ne_v_ec$adj.P.Val > 0.01, ]) + geom_point(mapping = aes(x = logFC, y = -log10(adj.P.Val)), col = 'grey20') + theme_bw() + geom_point(ne_v_ec[(ne_v_ec$logFC < -2 | ne_v_ec$logFC > 2) & ne_v_ec$adj.P.Val < 0.01, ], mapping = aes(x = logFC, y = -log10(adj.P.Val)), col = 'red') + ylab('-Log10P') + xlab('Log2 fold change') + geom_vline(xintercept = c(-2, 2), lty = 2) + geom_hline(yintercept = -log10(0.01), lty = 2) + geom_text(ne_v_ec[c('SOX1', 'SOX3', 'PAX6', 'CDH2', 'SOX2', 'POU5F1', 'NANOG'),], mapping = aes(x = logFC, y = -log10(adj.P.Val), label = gene), hjust=0,vjust=0, size = 10)

ggplot(ne_v_ec[(ne_v_ec$logFC > -2 & ne_v_ec$logFC < 2) | ne_v_ec$adj.P.Val > 0.01, ]) + geom_point(mapping = aes(x = logFC, y = -log10(adj.P.Val)), col = 'grey20') + theme_bw() + geom_point(ne_v_ec[(ne_v_ec$logFC < -2 | ne_v_ec$logFC > 2) & ne_v_ec$adj.P.Val < 0.01, ], mapping = aes(x = logFC, y = -log10(adj.P.Val)), col = 'red') + geom_vline(xintercept = c(-2, 2), lty = 2) + geom_hline(yintercept = -log10(0.01), lty = 2) + theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.title = element_blank()) + ggsave('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/ne_v_ec_volcano_plot_20210927.png', device = 'png', width = 4, height = 4, units = 'in')

sm_v_ec$gene = row.names(sm_v_ec)
ggplot(sm_v_ec[(sm_v_ec$logFC > -2 & sm_v_ec$logFC < 2) | sm_v_ec$adj.P.Val > 0.01, ]) + geom_point(mapping = aes(x = logFC, y = -log10(adj.P.Val)), col = 'grey20') + theme_bw() + geom_point(sm_v_ec[(sm_v_ec$logFC < -2 | sm_v_ec$logFC > 2) & sm_v_ec$adj.P.Val < 0.01, ], mapping = aes(x = logFC, y = -log10(adj.P.Val)), col = 'red') + ylab('-Log10P') + xlab('Log2 fold change') + geom_vline(xintercept = c(-2, 2), lty = 2) + geom_hline(yintercept = -log10(0.01), lty = 2) + geom_text(sm_v_ec[c('ACTA2', 'MIR145', 'MIR143HG', 'MYOCD', 'SOX2', 'POU5F1', 'NANOG'),], mapping = aes(x = logFC, y = -log10(adj.P.Val), label = gene), hjust=0,vjust=0, size = 6)

ggplot(sm_v_ec[(sm_v_ec$logFC > -2 & sm_v_ec$logFC < 2) | sm_v_ec$adj.P.Val > 0.01, ]) + geom_point(mapping = aes(x = logFC, y = -log10(adj.P.Val)), col = 'grey20') + theme_bw() + geom_point(sm_v_ec[(sm_v_ec$logFC < -2 | sm_v_ec$logFC > 2) & sm_v_ec$adj.P.Val < 0.01, ], mapping = aes(x = logFC, y = -log10(adj.P.Val)), col = 'red') + geom_vline(xintercept = c(-2, 2), lty = 2) + geom_hline(yintercept = -log10(0.01), lty = 2) + theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.title = element_blank()) + ggsave('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/sm_v_ec_volcano_plot_20210927.png', device = 'png', width = 4, height = 4, units = 'in')

ct_v_ec$gene = row.names(ct_v_ec)
ggplot(ct_v_ec[(ct_v_ec$logFC > -2 & ct_v_ec$logFC < 2) | ct_v_ec$adj.P.Val > 0.01, ]) + geom_point(mapping = aes(x = logFC, y = -log10(adj.P.Val)), col = 'grey20') + theme_bw() + geom_point(ct_v_ec[(ct_v_ec$logFC < -2 | ct_v_ec$logFC > 2) & ct_v_ec$adj.P.Val < 0.01, ], mapping = aes(x = logFC, y = -log10(adj.P.Val)), col = 'red') + ylab('-Log10P') + xlab('Log2 fold change') + geom_vline(xintercept = c(-2, 2), lty = 2) + geom_hline(yintercept = -log10(0.01), lty = 2) + geom_text(ct_v_ec[c('COL2A1', 'COMP', 'ACAN', 'SOX2', 'POU5F1', 'NANOG'),], mapping = aes(x = logFC, y = -log10(adj.P.Val), label = gene), hjust=0,vjust=0, size = 6)

ggplot(ct_v_ec[(ct_v_ec$logFC > -2 & ct_v_ec$logFC < 2) | ct_v_ec$adj.P.Val > 0.01, ]) + geom_point(mapping = aes(x = logFC, y = -log10(adj.P.Val)), col = 'grey20') + theme_bw() + geom_point(ct_v_ec[(ct_v_ec$logFC < -2 | ct_v_ec$logFC > 2) & ct_v_ec$adj.P.Val < 0.01, ], mapping = aes(x = logFC, y = -log10(adj.P.Val)), col = 'red') + geom_vline(xintercept = c(-2, 2), lty = 2) + geom_hline(yintercept = -log10(0.01), lty = 2) + theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.title = element_blank()) + ggsave('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/ct_v_ec_volcano_plot_20210927.png', device = 'png', width = 4, height = 4, units = 'in')

p1 = EnhancedVolcano(ne_v_ec,
    lab = rownames(ne_v_ec),
    x = 'logFC',
    y = 'adj.P.Val', selectLab = c('SOX1', 'SOX3', 'PAX6', 'CDH2', 'SOX2', 'POU5F1', 'NANOG'), 
    title = 'Neuroepithelium',
    subtitle = '',
    labFace = 'bold', 
    boxedLabels = T, 
    drawConnectors = TRUE,
    widthConnectors = 1.0,
    typeConnectors = 'open',
    colConnectors = 'black', 
    legendPosition = 'none', 
    pCutoff = 0.01,
    FCcutoff = 2,
    labSize = 5, 
    gridlines.major = FALSE,
    gridlines.minor = FALSE,
    caption = '')
p2 = EnhancedVolcano(sm_v_ec,
    lab = rownames(sm_v_ec),
    x = 'logFC',
    y = 'adj.P.Val', selectLab = c('MYOCD', 'ACTA2', 'MIR145', 'MIR143HG', 'SOX2', 'POU5F1', 'NANOG'), 
    title = 'Smooth muscle teratoma',
    subtitle = '',
    labFace = 'bold', 
    boxedLabels = T, 
    drawConnectors = TRUE,
    widthConnectors = 1.0,
    typeConnectors = 'closed',
    colConnectors = 'black', 
    legendPosition = 'none', 
    pCutoff = 0.01,
    FCcutoff = 2,
    labSize = 5, 
    gridlines.major = FALSE,
    gridlines.minor = FALSE,
    caption = '')
p3 = EnhancedVolcano(ct_v_ec,
    lab = rownames(ct_v_ec),
    x = 'logFC',
    y = 'adj.P.Val', selectLab = c('COL2A1', 'ACAN', 'COMP', 'SOX2', 'POU5F1', 'NANOG'), 
    title = 'Cartilage teratoma',
    subtitle = '',
    labFace = 'bold', 
    boxedLabels = T, 
    drawConnectors = TRUE,
    widthConnectors = 1.0,
    typeConnectors = 'closed',
    colConnectors = 'black', 
    legendPosition = 'none', 
    pCutoff = 0.01,
    FCcutoff = 2,
    labSize = 5,
    gridlines.major = FALSE,
    gridlines.minor = FALSE,
    caption = '')

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/gct_tissues_v_ec_de_plot_20210826.pdf', height = 6, width = 18, useDingbats = F)
plot_grid(p3, p1, p2, ncol = 3, align = 'h')
dev.off()

```

####Check expression of these transcripts in the TPM data

```{r}

tpm = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/germ_cell_lcm_tpm_counts_filtered_samples_20210812.txt', header = T, sep = '\t', stringsAsFactors = F) 
log2tpm = log2(as.matrix(tpm) + 1)
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #read in metadata for filtered, final cuts
row.names(metadata) = metadata$Sample

log2tpm_subset = log2tpm[c('POU5F1', 'SOX2', 'NANOG', 'COL2A1', 'COMP', 'ACAN', 'SOX9', 'SOX1', 'CDH2', 'SOX3', 'PAX6', 'MYOCD', 'MIR143HG', 'MIR145', 'ACTA2'),]

log2tpm_subset.t = data.frame(t(log2tpm_subset))
log2tpm_subset.t$histology = metadata[row.names(log2tpm_subset.t),]$Updated.Description
log2tpm_subset.t$Case.ID = metadata[row.names(log2tpm_subset.t),]$Case.ID

log2tpm_subset.t = log2tpm_subset.t[log2tpm_subset.t$histology %in% c('Embryonal_carcinoma', 'Cartilage_teratoma', 'Neuroepithelium', 'Smooth_muscle_teratoma'),]

library(reshape2)
library(ggplot2)
library(ggbeeswarm)

log2tpm_subset.t.m = reshape2::melt(log2tpm_subset.t, id.vars = c('histology', 'Case.ID'))

#cartilage teratoma
ggplot(log2tpm_subset.t.m[log2tpm_subset.t.m$histology %in% c('Embryonal_carcinoma', 'Cartilage_teratoma') & log2tpm_subset.t.m$variable %in% c('POU5F1', 'SOX2', 'NANOG', 'COL2A1', 'COMP', 'ACAN'), ]) + 
  geom_quasirandom(mapping = aes(x = histology, y = value, col = Case.ID)) +
  geom_pointrange(mapping = aes(x = histology, y = value), stat = "summary", shape=19, fun.min = function(z) {quantile(z,0.25)}, 
                  fun.max = function(z) {quantile(z,0.75)}, 
                  fun = median, size = 0.8, fatten = 5, linetype = 2, col = "black") +
  facet_grid(. ~ variable) + 
  theme_linedraw() + ylab('Log2(TPM)') + xlab('') +
  scale_x_discrete(labels= c('CT', 'EC'))

#Neuroepithelium
ggplot(log2tpm_subset.t.m[log2tpm_subset.t.m$histology %in% c('Embryonal_carcinoma', 'Neuroepithelium') & log2tpm_subset.t.m$variable %in% c('POU5F1', 'NANOG', 'SOX1', 'CDH2', 'SOX3', 'PAX6'), ]) + 
  geom_quasirandom(mapping = aes(x = histology, y = value, col = Case.ID)) +
  geom_pointrange(mapping = aes(x = histology, y = value), stat = "summary", shape=19, fun.min = function(z) {quantile(z,0.25)}, 
                  fun.max = function(z) {quantile(z,0.75)}, 
                  fun = median, size = 0.8, fatten = 5, linetype = 2, col = "black") +
  facet_grid(. ~ variable) + 
  theme_linedraw() + ylab('Log2(TPM)') + xlab('') +
  scale_x_discrete(labels= c('EC', 'NE'))

#Smooth muscle teratoma
ggplot(log2tpm_subset.t.m[log2tpm_subset.t.m$histology %in% c('Embryonal_carcinoma', 'Smooth_muscle_teratoma') & log2tpm_subset.t.m$variable %in% c('POU5F1', 'SOX2', 'NANOG', 'MYOCD', 'MIR143HG', 'MIR145', 'ACTA2'), ]) + 
  geom_quasirandom(mapping = aes(x = histology, y = value, col = Case.ID)) +
  geom_pointrange(mapping = aes(x = histology, y = value), stat = "summary", shape=19, fun.min = function(z) {quantile(z,0.25)}, 
                  fun.max = function(z) {quantile(z,0.75)}, 
                  fun = median, size = 0.8, fatten = 5, linetype = 2, col = "black") +
  facet_grid(. ~ variable) + 
  theme_linedraw() + ylab('Log2(TPM)') + xlab('') +
  scale_x_discrete(labels= c('EC', 'SM'))

geom_beeswarm(data = gc_exp.t.m[gc_exp.t.m$cell_types == 'Dysgerminoma' & gc_exp.t.m$variable %in% c('DDX4', 'TNP1'), ], mapping = aes(x = gene_use, y = value), dodge.width=1, cex=1, shape = 21, fill = 'black', col = 'black') + 
  facet_wrap(. ~ cell_types, strip.position = 'bottom', ncol = 8) +
  theme_pubr() + 
  theme(axis.text.x = element_blank(), 
        panel.spacing = unit(0, "mm"), 
        strip.background = element_blank(), 
        strip.placement = "outside", 
        legend.position = 'none', 
        axis.ticks.x = element_blank()) +
  labs(y = 'Log2(TPM) expression', x = '') + 
  geom_pointrange(mapping = aes(col = gene_use), stat = "summary", shape=19, fun.min = function(z) {quantile(z,0.25)}, 
                  fun.max = function(z) {quantile(z,0.75)}, 
                  fun = median, size = 0.8, fatten = 5, linetype = 2) + 
  scale_color_manual(values = c('#FF6C5C', '#58CCED', '#072F5F'))

```


###All invasive against seminiferous tubules

```{r}

library(limma)
library(edgeR)

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #just normal testis and GCT raw count data, 500 microbiopsies
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #read in metadata for filtered, final cuts
row.names(metadata) = metadata$Sample
gct_mat = raw_data[, 7:ncol(raw_data)]
row.names(gct_mat) = raw_data$Geneid

#remove cuts that failed QC
gct_mat = gct_mat[, colnames(gct_mat) %in% metadata[metadata$Updated.Description != 'GCNIS',]$Sample] #380 samples after removing GCNIS

#define factors that will be input into the model
histology = factor(metadata[colnames(gct_mat), ]$Updated.Description)
tumour = factor(metadata[colnames(gct_mat), ]$Normal.Tumour)
metadata = metadata[colnames(gct_mat),]

#preprocessing
y <- DGEList(counts = gct_mat, group = tumour)
keep <- filterByExpr(y, group = tumour)
y <- y[keep, , keep.lib.sizes=FALSE]
y <- calcNormFactors(y) #12394 genes kept
y <- estimateCommonDisp(y) #to generate pseudocounts, adjusted for library size

#visualise if microbiopsies from the same histology cluster and if there is patient effect
plotMDS(y, labels = tumour, col = as.numeric(histology))

#specify model to be fitted, no intercept to be fitted, each normal-tumour status will have its own group mean
design = model.matrix(~ 0 + tumour)
colnames(design) = levels(tumour)

#run double voom to optimise regression weights, adjust for histology as 'block' to find underlying generic tumour signal
v <- voom(y, design)
corfit <- duplicateCorrelation(v, design, block = histology)
v <- voom(y, design, block = histology, correlation = corfit$consensus)
corfit <- duplicateCorrelation(v, design, block = histology)
fit <- lmFit(v, design, block = histology, correlation = corfit$consensus)
corfit$consensus #0.1987738

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom')

saveRDS(v, 'LCM_invasive_GCT_v_testis_DE_voom_object_20210812.rds')
saveRDS(fit, 'LCM_invasive_GCT_v_testis_DE_model_fit_20210812.rds')

v = readRDS('LCM_invasive_GCT_v_testis_DE_voom_object_20210812.rds')
fit = readRDS('LCM_invasive_GCT_v_testis_DE_model_fit_20210812.rds')

cont_matrix = makeContrasts(Tumour - Normal, levels = colnames(coef(fit)))
fit2 <- contrasts.fit(fit, cont_matrix)
fit2 <- eBayes(fit2)
top.table <- topTable(fit2, number = nrow(fit$coefficients), sort.by = "P")

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F)
row.names(raw_data) = raw_data$Geneid
raw_data$Start = unlist(lapply(strsplit(raw_data$Start, ';'), '[[', 1))

top.table$chr = raw_data[row.names(top.table),]$Chr
top.table$start = raw_data[row.names(top.table),]$Start
top.table$start = as.numeric(top.table$start)

write.table(top.table, 'invasive_gct_v_normtestis_voom_de_20210812.txt', col.names = T, sep = '\t', row.names = T, quote = F)

```

###Cytoband-level gene expression (Fig. 4a)

```{r}

library(msigdbr)
library(limma)
library(scales)
library(ggpubr)
library(reshape2)
library(ComplexHeatmap)
library(circlize)

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom')

gct_v = readRDS('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom/limma_voom_de_LCM_GCT_DE_voom_20210925.rds')
gct_fit = readRDS('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom/limma_voom_de_LCM_GCT_DE_model_fit_20210925.rds')

contr.matrix <- makeContrasts(
  fat_v_st = Adipose_teratoma - Seminiferous_tubule,
  dys_v_st = Dysgerminoma - Seminiferous_tubule,
  ec_v_st = Embryonal_carcinoma - Seminiferous_tubule,
  gcnis_v_st = GCNIS - Seminiferous_tubule,
  cartilage_v_st = Cartilage_teratoma - Seminiferous_tubule, 
  glandular_v_st = Mature_glandular_teratoma - Seminiferous_tubule, 
  ne_v_st = Neuroepithelium - Seminiferous_tubule, 
  other_ter_v_st = Other_epithelial_teratoma - Seminiferous_tubule, 
  muscle_v_st = Smooth_muscle_teratoma - Seminiferous_tubule, 
  sct_v_st = Syncytiotrophoblasts - Seminiferous_tubule, 
  yst_v_st = Yolk_sac_tumour - Seminiferous_tubule, 
  sem_v_st = Seminoma - Seminiferous_tubule,
  str_v_st = Stroma - Seminiferous_tubule,
  levels = colnames(gct_v$design))

all_hum_gene_sets = msigdbr(species = "Homo sapiens")

c1_gene_set = list()

for(i in unique(all_hum_gene_sets[all_hum_gene_sets$gs_cat == 'C1', ]$gs_name)){
  c1_gene_set[[i]] = all_hum_gene_sets[all_hum_gene_sets$gs_cat == 'C1' & all_hum_gene_sets$gs_name == i, ]$human_gene_symbol
}
c1_idx <- ids2indices(c1_gene_set,id=rownames(gct_v))

all_histo_regions_enrich = c()
for(i in 1:ncol(contr.matrix)){
  temp = camera(gct_v, c1_idx, gct_fit$design, contrast = contr.matrix[,i])
  tissue = names(contr.matrix[,i][contr.matrix[,i] == 1])
  temp$tissue = tissue
  temp$cytoband = row.names(temp)
  row.names(temp) = NULL
  all_histo_regions_enrich = rbind(all_histo_regions_enrich, temp)
}

all_histo_regions_enrich$chr = substr(unlist(lapply(strsplit(all_histo_regions_enrich$cytoband, 'p|q'), '[[', 1)), 4, 5)
all_histo_regions_enrich$chr = factor(all_histo_regions_enrich$chr, levels = c(1:22, 'X', 'Y'))
all_histo_regions_enrich$arm = substr(unlist(lapply(strsplit(all_histo_regions_enrich$cytoband, paste(unique(paste0('chr', all_histo_regions_enrich$chr)), collapse = '|')), '[[', 2)), 0, 1)
all_histo_regions_enrich$cytoband_pos = substr(unlist(lapply(strsplit(all_histo_regions_enrich$cytoband, 'p|q'), '[[', 2)), 0, 2)
  
all_histo_regions_enrich = all_histo_regions_enrich[order(all_histo_regions_enrich$tissue, all_histo_regions_enrich$chr, all_histo_regions_enrich$arm, all_histo_regions_enrich$cytoband_pos),]
all_histo_regions_enrich$cytoband = factor(all_histo_regions_enrich$cytoband, levels = unique(all_histo_regions_enrich$cytoband))
all_histo_regions_enrich$sig_variable = 1 / all_histo_regions_enrich$FDR #invert for plotting purposes

for(i in 1:nrow(all_histo_regions_enrich)){
  if(all_histo_regions_enrich$Direction[i] == 'Down') all_histo_regions_enrich$sig_variable[i] = -1 * all_histo_regions_enrich$sig_variable[i]
}

write.table(all_histo_regions_enrich, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/gct_cytoband_region_enrichment20210927.txt', col.names = T, row.names = F, quote = F, sep = '\t')

all_histo_regions_enrich_wide <- dcast(all_histo_regions_enrich, tissue ~ cytoband, value.var="sig_variable")
row.names(all_histo_regions_enrich_wide) = all_histo_regions_enrich_wide$tissue

all_histo_regions_enrich_wide = all_histo_regions_enrich_wide[,2:276]
all_histo_regions_enrich_wide_auto = all_histo_regions_enrich_wide[, !grepl('chrX|chrY', colnames(all_histo_regions_enrich_wide)),]
 
col_fun = colorRamp2(c(-100, -10, 0, 10, 100), c('purple', '#520160', 'black', '#435E1C', '#95D840'))
chrom_order = factor(unlist(lapply(strsplit(colnames(all_histo_regions_enrich_wide_auto), 'p|q'), '[[', 1)), levels = paste0('chr', 1:22))
levels(chrom_order) = 1:22

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/enriched_regions_gct_expression20210927.pdf', width = 15, height = 8)
Heatmap(all_histo_regions_enrich_wide_auto, cluster_columns = F, cluster_rows = F, col = col_fun, column_split = chrom_order, heatmap_legend_param = list(title = "FDR", at = c(-100, -10, 10, 100), labels = c(0.01, 0.1, 0.1, 0.01), legend_height = unit(4, "cm")), show_column_names = F, row_names_side = 'left', row_order = c('GCNIS', 'Dysgerminoma', 'Seminoma', 'Embryonal_carcinoma', 'Yolk_sac_tumour', 'Syncytiotrophoblasts', 'Neuroepithelium', 'Mature_glandular_teratoma', 'Other_epithelial_teratoma', 'Adipose_teratoma', 'Smooth_muscle_teratoma', 'Cartilage_teratoma', 'Stroma'))
dev.off()

```

#####Which 12p genes are upregulated in which GCT tissues?

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom')

invasive_comparison = list.files(pattern = 'st_voom_de_20210925.txt')[!grepl('gcnis', list.files(pattern = 'st_voom_de_20210925.txt'))]

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #raw count data
row.names(raw_data) = raw_data$Geneid
raw_data$Start = unlist(lapply(strsplit(raw_data$Start, ';'), '[[', 1))
raw_data$Start = as.numeric(raw_data$Start)
genes_list = row.names(raw_data[raw_data$Chr == 12 & raw_data$Start < 35800000,]) #759


all_tissue_df = c()
for(i in invasive_comparison){
  tissue_file = read.table(i, header = T, sep = '\t', stringsAsFactors = F)
  sig_up = tissue_file[tissue_file$logFC >= 2 & tissue_file$adj.P.Val < 0.01,]
  sig_up = sig_up[row.names(sig_up) %in% genes_list,]
  tissue_df = data.frame(row.names(sig_up), i)
  all_tissue_df = rbind(all_tissue_df, tissue_df)
}

all_tissue_df$i = as.character(all_tissue_df$i)
all_tissue_df$i = unlist(strsplit(all_tissue_df$i, '_v_st_voom_de_20210925.txt'))
all_tissue_tab = as.data.frame(table(all_tissue_df$row.names.sig_up.))
all_tissue_tab = all_tissue_tab[order(all_tissue_tab$Freq, decreasing = T),]

tf_list = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/reference_datasets/church_tfs.txt', header = F, sep = '\t', stringsAsFactors = F)[,1]

all_tissue_tab$Var1 = as.character(all_tissue_tab$Var1)
all_tissue_tab$is_tf = 'no'
for(i in 1:nrow(all_tissue_tab)){
  if(all_tissue_tab$Var1[i] %in% tf_list) all_tissue_tab$is_tf[i] = 'yes'
}

names(all_tissue_tab) = c('12p gene', 'No. of tissues upregulated in (/12)', 'Transcription factor')

all_tissue_tab$tissues_involved = NA
for(j in 1:nrow(all_tissue_tab)){
  all_tissue_tab$tissues_involved[j] = paste(all_tissue_df[all_tissue_df$row.names.sig_up. == all_tissue_tab$`12p gene`[j],]$i, collapse = ', ')
}

write.table(all_tissue_tab, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/gct_tissues_enriched_for_12p_genes_20210928.txt', col.names = T, row.names = F, sep = '\t', quote = F)

tpm = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/germ_cell_lcm_tpm_counts_filtered_samples_20210812.txt', header = T, sep = '\t', stringsAsFactors = F)
log2tpm <- log2(as.matrix(tpm) + 1)

metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #read in metadata for filtered, final cuts
row.names(metadata) = metadata$Sample
log2tpm = log2tpm[, metadata$Sample]
histology = metadata$Updated.Description
names(histology) = colnames(log2tpm)

genes_explore_df = data.frame(t(log2tpm[c('CCND2', 'ATN1', 'PTMS', 'KRAS'),]))
genes_explore_df$sampleID = row.names(genes_explore_df)
genes_explore_df$histology = metadata[row.names(genes_explore_df),]$Updated.Description
genes_explore_df$patient = metadata[row.names(genes_explore_df),]$Case.ID

library(reshape2)

genes_explore_df.m = reshape2::melt(genes_explore_df, id.vars = c('sampleID', 'histology', 'patient'))

ggplot(genes_explore_df.m) + 
  geom_quasirandom(mapping = aes(x = histology, y = value, col = patient)) + 
  geom_pointrange(mapping = aes(x = histology, y = value), stat = "summary", shape=19, fun.min = function(z) {quantile(z,0.25)}, 
              fun.max = function(z) {quantile(z,0.75)}, 
              fun = median) +
  facet_wrap(.~variable, nrow = 4) + 
  theme_linedraw() + 
  ylab('Log2tpm counts') + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) + xlab('')
  

```

######KRAS check

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom')

invasive_comparison = list.files(pattern = 'st_voom_de_20210925.txt')[!grepl('gcnis', list.files(pattern = 'st_voom_de_20210925.txt'))]

raw_data = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/analyses/RNA/filtering/germ_cell_lcm_rna_preQC.txt', header = T, sep = '\t', stringsAsFactors = F) #raw count data
row.names(raw_data) = raw_data$Geneid
raw_data$Start = unlist(lapply(strsplit(raw_data$Start, ';'), '[[', 1))
raw_data$Start = as.numeric(raw_data$Start)

all_tissue_df = c()
for(i in invasive_comparison){
  tissue = unlist(strsplit(i, '_v_st_voom_de_20210925.txt'))
  tissue_file = read.table(i, header = T, sep = '\t', stringsAsFactors = F)
  tissue_file$tissue = tissue
  tissue_file$gene = row.names(tissue_file)
  row.names(tissue_file) = NULL
  all_tissue_df = rbind(all_tissue_df, tissue_file)
}

all_tissue_df[all_tissue_df$gene == 'KRAS', c('gene', 'tissue', 'logFC', 'adj.P.Val')]
all_tissue_df[all_tissue_df$gene == 'ATN1', c('gene', 'tissue', 'logFC', 'adj.P.Val')]
```


####12p CN vs expression (Fig. 4b)

```{r}

#fetch cn data from Fig. 1

cn_df = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/combined_CN_RNA_exp/lcm_gct_cn_10000_bp_bin_per_GCT_WG_median_extra_cols.txt', header = T, sep = '\t', stringsAsFactors = F)

#limit to chromosome 12 (cumulative coordinates from 10kb bins in file)

lcm_chr12_df = cn_df[(cn_df$start > 1950970000) & (cn_df$end <= 2084830000), ]
lcm_chr12_df$chr = 'chr12'
lcm_chr12_df$start = lcm_chr12_df$start - 1950970000
lcm_chr12_df$end = lcm_chr12_df$end - 1950970000
lcm_chr12_df = lcm_chr12_df[!is.na(lcm_chr12_df$all_samples),]

lcm_chr12_df.copy = lcm_chr12_df
lcm_chr12_df.copy = lcm_chr12_df.copy[, c("chr", "start", "end", "all_samples")]

write.table(lcm_chr12_df.copy, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/combined_CN_RNA_exp/lcm_gct_cn_10000_bp_bin_per_GCT_chr12_source_data.txt', col.names = T, row.names = F, sep = '\t', quote = F)# source data for paper

#fetch logFC data from invasive v normal comparison

top.table.lcm = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom/invasive_gct_v_normtestis_voom_de_20210812.txt', header = T, sep = '\t', stringsAsFactors = F)

#limit to chromsome 12 data for plotting
lcm_chr12_exp_data = top.table.lcm[top.table.lcm$chr == 12,]
lcm_chr12_exp_data$chr = paste0('chr', lcm_chr12_exp_data$chr)
lcm_chr12_exp_data = lcm_chr12_exp_data[order(lcm_chr12_exp_data$start),]

library(karyoploteR)
library(zoo)

#generic plotting settings
chromosomes = 'chr12'
dna_r0 <- 0.9
dna_r1 <- 0
rna_r0 <- 0
rna_r1 <- 0.9
karyoplotr_plottype = 1

#in-house rna.data
window_size = 50 #must be an even number
y_data = rollmean(lcm_chr12_exp_data$logFC, k = window_size, align = 'center')
x_data = lcm_chr12_exp_data$start[(window_size / 2):(length(lcm_chr12_exp_data$start) - (window_size / 2))]
pos_y_data = ifelse(y_data > 0, y_data, 0) 
neg_y_data = ifelse(y_data < 0, y_data, 0)

##plot in-house data
title = 'In-house'

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/gct_cn_v_exp_karyoplotter_20210812.pdf', height = 6, width = 6, useDingbats = F)
kp <- karyoploteR::plotKaryotype(genome = "hg19", chromosomes = 'chr12', plot.type = karyoplotr_plottype, cex = 0.85, main = gsub(".*__", "", title))
karyoploteR::kpDataBackground(kp, r0 = dna_r0, r1 = dna_r1, col = 'grey90')
karyoploteR::kpAxis(karyoplot = kp, col = "black", ymin = -1.5, ymax = 1.5, r0 = rna_r0, r1 = rna_r1, cex = 1, tick.pos = c(-1.5, -1, -0.5, 0, 0.5, 1, 1.5), side = 2)
karyoploteR::kpBars(kp, chr = 'chr12', x0 = x_data - 100000, x1 = x_data + 100000, y0 = 0, y1 = pos_y_data, col = '#95D840', data.panel = 1, ymin = -1.5, ymax = 1.5, r0 = rna_r0, r1 = rna_r1, clipping = T, border = 'NA')
karyoploteR::kpBars(kp, chr = 'chr12', x0 = x_data - 100000, x1 = x_data + 100000, y0 = 0, y1 = neg_y_data, col = '#A020F0', data.panel = 1, ymin = -1.5, ymax = 1.5, r0 = rna_r0, r1 = rna_r1, clipping = T, border = 'NA')
karyoploteR::kpAxis(karyoplot = kp, col = "black", ymin = 6, ymax = -3, r1 = dna_r1, r0 = dna_r0, cex = 1, tick.pos = c(-3, 0, 3, 6), side = 1)
karyoploteR::kpPoints(kp, chr = lcm_chr12_df$chr, x = lcm_chr12_df$start, y = lcm_chr12_df$all_samples, col = "black", ymin = 6, ymax = -3, r0 = dna_r0, r1 = dna_r1, clipping = T) #reintroduced prepubertal samples
dev.off()

```

####12p gene enrichment compared to regions of near baseline ploidy (Supplementary Figure 9)

```{r}

library(plyr)

chrom_sizes <- read.table('/lustre/scratch119/casm/team294rr/rs30/References/GRCh37d5/chrom.sizes.txt', header = F, sep = '\t', stringsAsFactors = F)
chrom_sizes$V1 = factor(chrom_sizes$V1, levels = c(1:22, "X", "Y"))
chrom_sizes = chrom_sizes[order(chrom_sizes$V1),]
chrom_sizes[,2] = round_any(chrom_sizes[,2], bin_size, ceiling) #chunk each chromosome into 10kb bins in order to accurate characterise the CN at the beginning and ends of chromosomes, has to be ceiling to capture the ends
chrom_sizes[,3] <- cumsum(as.numeric(chrom_sizes[,2]))
#chrom_sizes = chrom_sizes[!chrom_sizes$V1 %in% c('chrM', 'chrX', 'chrY'),]
#chrom_sizes$V1 = unlist(lapply(strsplit(chrom_sizes$V1, 'chr'), '[[', 2))
chrom_sizes = chrom_sizes[!chrom_sizes$V1 %in% c("X", "Y"),]
chrom_sizes$V4 = c(0, chrom_sizes$V3[1:(length(chrom_sizes$V3) - 1)])
names(chrom_sizes) = c('chr', 'chr_length', 'cumsum_length', 'prior_wg_length')

#fetch cn data from Fig. 1

cn_df = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/combined_CN_RNA_exp/lcm_gct_cn_10000_bp_bin_per_GCT_WG_median_extra_cols.txt', header = T, sep = '\t', stringsAsFactors = F)

#limit to chromosome 12 (cumulative coordinates from 10kb bins in file)

lcm_chr12_df = cn_df[(cn_df$start > 1950970000) & (cn_df$end <= 2084830000), ]
lcm_chr12_df$chr = 'chr12'
lcm_chr12_df$start = lcm_chr12_df$start - 1950970000
lcm_chr12_df$end = lcm_chr12_df$end - 1950970000
lcm_chr12_df = lcm_chr12_df[!is.na(lcm_chr12_df$all_samples),]

#fetch logFC data from invasive v normal comparison

top.table.lcm = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/limma_voom/invasive_gct_v_normtestis_voom_de_20210812.txt', header = T, sep = '\t', stringsAsFactors = F)
top.table.lcm = top.table.lcm[top.table.lcm$chr %in% c(1:22),] #only performed copy number plotting across autosomal genome for comparison

#limit to chromsome 12 data for plotting
lcm_chr12_exp_data = top.table.lcm[top.table.lcm$chr == 12,]
lcm_chr12_exp_data$chr = paste0('chr', lcm_chr12_exp_data$chr)
lcm_chr12_exp_data = lcm_chr12_exp_data[order(lcm_chr12_exp_data$start),]


#get regions with a CN-ploidy within 0.5 of the overall ploidy located outside of chr12
av_cn = cn_df[!(is.na(cn_df$all_samples)) & (cn_df$all_samples > -0.5) & (cn_df$all_samples < 0.5) & ((cn_df$start < 1950970000) | (cn_df$start > 2084830000)),]

#provided cumulative loci for genes
top.table.lcm$mod_start = top.table.lcm$start
for(k in 1:nrow(top.table.lcm)) {
    if(top.table.lcm$chr[k] != "1"){
        top.table.lcm$mod_start[k] = top.table.lcm$start[k] + chrom_sizes$prior_wg_length[which(chrom_sizes$chr == top.table.lcm$chr[k])]
      }
}

top.table.lcm$av_cn_region = NA

for(i in 1:nrow(top.table.lcm)){
  if(nrow(av_cn[(av_cn$start < top.table.lcm$mod_start[i]) & (av_cn$end > top.table.lcm$mod_start[i]), ]) > 0){
    top.table.lcm$av_cn_region[i] = 'Yes'
  }
  if(nrow(av_cn[(av_cn$start < top.table.lcm$mod_start[i]) & (av_cn$end > top.table.lcm$mod_start[i]), ]) == 0){
    top.table.lcm$av_cn_region[i] = 'No'
  }
}

n_12p_genes = nrow(top.table.lcm[top.table.lcm$chr == 12 & top.table.lcm$start < 35800000,]) #226 used
mean_12p = mean(top.table.lcm[top.table.lcm$chr == 12 & top.table.lcm$start < 35800000,]$logFC) #  0.7004505

nRand = 100000
randScores=c()
set.seed(42)
av_cn_exp_regions = top.table.lcm[top.table.lcm$av_cn_region == 'Yes',]

for(i in 1:nRand){
  randScores=c(randScores, mean(sample(av_cn_exp_regions$logFC, n_12p_genes, replace = F)))
}

pVal = sum(randScores >= mean_12p)/nRand #0...

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/rand_sampling_gene_exp_v_12p_20210913.pdf', height = 3, width = 4)
ggplot() + geom_density(mapping = aes(x = randScores), fill = '#999999') + xlim(-2, 1) + theme_classic() + geom_vline(xintercept = mean_12p) + labs(x = 'log2FC of gene expression in regions with baseline ploidy') + coord_cartesian(expand = T)
dev.off()

```

##Expression of fetal transcripts in GCT tissues (Fig. 3a, b and c)

###Generate log2tpm values
Updated to standardise tpm calculation across all samples included 23/5/22.

```{r}

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/reference_datasets')

#read in gct raw counts
gct_raw_counts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/germ_cell_lcm_raw_counts_filtered_samples_20210812.txt', header = T, sep = '\t', stringsAsFactors = F)

#read in R object with counts from reference datasets
ref_scrna_final = readRDS('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/reference_scrna.rds')

#re-label cell annotations
ref_scrna_final@meta.data$cell_types = as.factor(ref_scrna_final@meta.data$cell_types)
ref_scrna_final@meta.data$cell_types = factor(ref_scrna_final@meta.data$cell_types, levels = rev(c("PGC-like", "Spermatogonia", "Spermatocytes", "Spermatid", "Fetal neuronal stem cells", "Fetal oligodendrocyte progenitor cells", "Adult oligodendrocyte progenitor cells", "Adult oligodendrocytes", "Fetal astrocytes", "Adult astrocytes", "Fetal interneurons", "Adult interneurons", "Fetal excitatory neurons", "Adult excitatory neurons", "Fetal cardiac smooth muscle cells", "Fetal other smooth muscle cells", "Adult other smooth muscle cells")))

#subset to genes that are shared between reference data and GCT data
intersect_genes = intersect(row.names(ref_scrna_final), gct_raw_counts$Geneid)
length(intersect_genes) #34866

gct_raw_counts_subset = gct_raw_counts[gct_raw_counts$Geneid %in% intersect_genes,]
nrow(gct_raw_counts_subset) #34866

sum(!row.names(ref_scrna_final) %in% intersect_genes) #0, no genes not in the intersect list

#generate log2(tpm) values for our modified smartseq2 data (needs adjustment for gene length)
x = gct_raw_counts_subset[, 7:ncol(gct_raw_counts_subset)]
row.names(x) = gct_raw_counts_subset$Geneid
x = x / gct_raw_counts_subset$Length #normalise for transcript length
gct_tpm <- t( t(x) * 1e6 / colSums(x)) #normalise to read depth

colSums(gct_tpm) #check it has worked
gct_log2tpm = log2(as.matrix(gct_tpm) + 1)

write.table(gct_log2tpm, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/00_SUPPLEMENTARY_DATA/gct_lcm_data_renormalised_log2tpm_counts_20220523.txt', col.names = T, row.names = T, sep = '\t', quote = F)

#generate log2(tpm) values for reference data - all used UMI-based RNAseq approaches, no adjustment for gene length required
ref_scrna_final_df = as.data.frame(as.matrix(ref_scrna_final@assays$RNA@counts))
write.table(ref_scrna_final_df, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/00_SUPPLEMENTARY_DATA/reference_scrna_counts_20220523.txt', col.names = T, row.names = T, sep = '\t', quote = F)

#a lot of very large datasets in memory, wipe them and read in the relevant files again
ref_scrna_final_df = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/00_SUPPLEMENTARY_DATA/reference_scrna_counts_20220523.txt', header = T, sep = '\t', stringsAsFactors = F)

#check that no NA values are in the reference data, e.g. from a gene being in one dataset and not another before merging
colSums(is.na(ref_scrna_final_df))[colSums(is.na(ref_scrna_final_df)) > 0] #none

#generate log2tpm
ref_scrna_final_tpm <- t( t(ref_scrna_final_df) * 1e6 / colSums(ref_scrna_final_df)) #tpm conversion
table(colSums(ref_scrna_final_tpm)) #looks good
ref_scrna_final_log2tpm <- log2(as.matrix(ref_scrna_final_tpm) + 1)

write.table(ref_scrna_final_log2tpm, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/00_SUPPLEMENTARY_DATA/reference_scrna_log2tpm.txt', col.names = T, row.names = T, sep = '\t', quote = F)

#clear again and read in necessary files only
gct_log2tpm = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/00_SUPPLEMENTARY_DATA/gct_lcm_data_renormalised_log2tpm_counts_20220523.txt', header = T, sep = '\t', stringsAsFactors = F)
ref_scrna_final_log2tpm = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/00_SUPPLEMENTARY_DATA/reference_scrna_log2tpm.txt', header = T, sep = '\t', stringsAsFactors = F)

#get metadata
ref_scrna_final = readRDS('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/reference_scrna.rds')
ref_scrna_final@meta.data$sample = row.names(ref_scrna_final@meta.data)
ref_scrna_final@meta.data$source = 'scrna'
ref_scrna_final_meta = ref_scrna_final@meta.data[, c('sample', 'cell_types', 'source')]

rna_cuts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #descriptive column contains apostrophes
row.names(rna_cuts) = rna_cuts$Sample
rna_cuts$cell_types = rna_cuts$Updated.Description
rna_cuts$sample = row.names(rna_cuts)
rna_cuts$source = 'lcm'
gct_final_meta = rna_cuts[, c('sample', 'cell_types', 'source')]

#combine dataframes
all_data_final_log2tpm = cbind(ref_scrna_final_log2tpm, gct_log2tpm[row.names(ref_scrna_final_log2tpm),])
all_data_final_meta = rbind(ref_scrna_final_meta, gct_final_meta)

#standardise naming
colnames(all_data_final_log2tpm) = gsub('\\.', '_', colnames(all_data_final_log2tpm))
row.names(all_data_final_meta) = gsub('-', '_', row.names(all_data_final_meta))
row.names(all_data_final_meta) = gsub('\\.', '_', row.names(all_data_final_meta))

#write out
write.table(all_data_final_log2tpm, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/00_SUPPLEMENTARY_DATA/combined_gct_scrna_ref_log2tpm.txt', col.names = T, row.names = T, sep = '\t', quote = F)
write.table(all_data_final_meta, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/00_SUPPLEMENTARY_DATA/combined_gct_scrna_ref_meta.txt', col.names = T, row.names = T, sep = '\t', quote = F)

```

###Plot

```{r}

library(ggplot2)
library(reshape2)
library(cowplot)
library(Seurat)
library(ggbeeswarm)
library(dplyr)
library(ggpubr)
library(gridExtra)
library(scales)

all_data_final_log2tpm = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/00_SUPPLEMENTARY_DATA/combined_gct_scrna_ref_log2tpm.txt', header = T, sep = '\t', stringsAsFactors = F)
all_data_final_log2tpm.sub = all_data_final_log2tpm[ , !colnames(all_data_final_log2tpm) %in% row.names(all_data_final_meta[all_data_final_meta$cell_types == "Fetal cardiac smooth muscle cells",])]

all_data_final_meta = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/00_SUPPLEMENTARY_DATA/combined_gct_scrna_ref_meta.txt', header = T, sep = '\t', stringsAsFactors = F)
all_data_final_meta.sub = all_data_final_meta[all_data_final_meta$cell_types != "Fetal cardiac smooth muscle cells", ]

write.table(all_data_final_log2tpm.sub[c("NANOG", "POU5F1", "SOX17", "DDX4", "TNP1", "NES", "PAX6", "SOX2", "MBP", "SLC1A2", "RBFOX3", "GAD1", "IGF2", "ACTA2", "MYH11", "TAGLN"), ], '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/00_SUPPLEMENTARY_DATA/combined_gct_scrna_ref_log2tpm_plotting.txt', col.names = T, row.names = T, sep = '\t', quote = F)
write.table(all_data_final_meta.sub, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/00_SUPPLEMENTARY_DATA/combined_gct_scrna_ref_meta_plotting.txt', col.names = T, row.names = T, sep = '\t', quote = F)

#germ cell comparison
gc_exp = all_data_final_log2tpm[,row.names(all_data_final_meta[all_data_final_meta$cell_types %in% c('PGC-like', 'Spermatogonia', 'Spermatocytes', 'Spermatid', 'GCNIS', 'Seminoma', 'Dysgerminoma', 'Seminiferous_tubule'), ])]
gc_exp.t = as.data.frame(t(as.data.frame(gc_exp)))
gc_exp.t$cell_types = all_data_final_meta[row.names(gc_exp.t),]$cell_types
gc_exp.t$sample = row.names(gc_exp.t)
gc_exp.t.m = reshape2::melt(gc_exp.t[, c('cell_types', 'sample', 'POU5F1', 'NANOG', 'SOX17', 'DDX4', 'TNP1')], id.vars = c('cell_types', 'sample'))

gc_exp.t.m$gene_use = NA
for(i in 1:nrow(gc_exp.t.m)){
  if(gc_exp.t.m$variable[i] %in% c('POU5F1', 'NANOG', 'SOX17')) gc_exp.t.m$gene_use[i] = 'fetal'
  if(gc_exp.t.m$variable[i] == 'DDX4') gc_exp.t.m$gene_use[i] = 'gen_gc'
  if(gc_exp.t.m$variable[i] == 'TNP1') gc_exp.t.m$gene_use[i] = 'spermatid'
}

gc_exp.t.m$gene_use = factor(gc_exp.t.m$gene_use, levels = c('fetal', 'gen_gc', 'spermatid'))
gc_exp.t.m$cell_types = factor(gc_exp.t.m$cell_types, levels = c('PGC-like', 'Spermatogonia', 'Spermatocytes', 'Spermatid', 'Seminiferous_tubule', 'GCNIS', 'Seminoma', 'Dysgerminoma'))

aggregate(value ~ cell_types + variable, gc_exp.t.m, length) #Dysgerminoma will need individual data points for DDX4 and TNP1 plotting

#check that effect consistent across each gene
ggplot(gc_exp.t.m) + 
  geom_quasirandom(mapping = aes(x = variable, y = value), dodge.width=.8, cex=1, shape = 21) +
  geom_pointrange(mapping = aes(x = variable, y = value, col = gene_use), stat = "summary", shape=19, fun.min = function(z) {quantile(z,0.25)}, 
                  fun.max = function(z) {quantile(z,0.75)}, 
                  fun = median, size = 0.8, fatten = 5, linetype = 2) +
  facet_wrap(. ~ cell_types, strip.position = 'bottom', ncol = 8)

p1 = ggplot(gc_exp.t.m, mapping = aes(x = gene_use, y = value)) + 
  geom_quasirandom(dodge.width=.8, cex=1, aes(x = gene_use, y = value), shape = 21, fill = 'transparent', col = 'transparent') + 
  geom_beeswarm(data = gc_exp.t.m[gc_exp.t.m$cell_types == 'Dysgerminoma' & gc_exp.t.m$variable %in% c('DDX4', 'TNP1'), ], mapping = aes(x = gene_use, y = value), dodge.width=1, cex=1, shape = 21, fill = 'black', col = 'black') + 
  facet_wrap(. ~ cell_types, strip.position = 'bottom', ncol = 8) +
  theme_pubr() + 
  theme(axis.text.x = element_blank(), 
        panel.spacing = unit(0, "mm"), 
        strip.background = element_blank(),
        panel.background = element_rect(fill = alpha("#b3b2b2", 0.5)),
        strip.placement = "outside", 
        legend.position = 'none', 
        axis.ticks.x = element_blank()) +
  labs(y = 'Log2(TPM) expression', x = '') + 
  geom_pointrange(mapping = aes(col = gene_use), stat = "summary", shape=19, fun.min = function(z) {quantile(z,0.25)}, 
                  fun.max = function(z) {quantile(z,0.75)}, 
                  fun = median, size = 0.8, fatten = 5, linetype = 2) + 
  scale_color_manual(values = c('#078e38', '#61c5e6', '#20335a'))

#neural comparison
ne_exp = all_data_final_log2tpm[,row.names(all_data_final_meta[all_data_final_meta$cell_types %in% c('Neuroepithelium', 'Fetal interneurons', 'Fetal neuronal stem cells', 'Fetal oligodendrocyte progenitor cells', 'Fetal astrocytes', 'Fetal excitatory neurons', 'Adult interneurons', 'Adult oligodendrocyte progenitor cells', 'Adult astrocytes', 'Adult excitatory neurons', 'Adult oligodendrocytes'), ])]
ne_exp.t = as.data.frame(t(as.data.frame(ne_exp)))
ne_exp.t$cell_types = all_data_final_meta[row.names(ne_exp.t),]$cell_types
ne_exp.t$sample = row.names(ne_exp.t)
ne_exp.t.m = reshape2::melt(ne_exp.t[, c('cell_types', 'sample', 'PAX6', 'SOX2', 'NES', 'MBP', 'SLC1A2', 'RBFOX3', 'GAD1')], id.vars = c('cell_types', 'sample'))

ne_exp.t.m$gene_use = NA
for(i in 1:nrow(ne_exp.t.m)){
  if(ne_exp.t.m$variable[i] %in% c('PAX6', 'SOX2', 'NES')) ne_exp.t.m$gene_use[i] = 'fetal'
  if(ne_exp.t.m$variable[i] == 'MBP') ne_exp.t.m$gene_use[i] = 'oligo'
  if(ne_exp.t.m$variable[i] == 'SLC1A2') ne_exp.t.m$gene_use[i] = 'astro'
  if(ne_exp.t.m$variable[i] == 'RBFOX3') ne_exp.t.m$gene_use[i] = 'neun'
  if(ne_exp.t.m$variable[i] == 'GAD1') ne_exp.t.m$gene_use[i] = 'inter'
}

ne_exp.t.m$gene_use = factor(ne_exp.t.m$gene_use, levels = c('fetal', 'oligo', 'astro', 'neun', 'inter'))
ne_exp.t.m$cell_types = factor(ne_exp.t.m$cell_types, levels = c('Fetal neuronal stem cells', 'Fetal oligodendrocyte progenitor cells', 'Fetal astrocytes', 'Fetal excitatory neurons', 'Fetal interneurons', 'Adult oligodendrocyte progenitor cells', 'Adult oligodendrocytes', 'Adult astrocytes', 'Adult excitatory neurons', 'Adult interneurons', 'Neuroepithelium'))

aggregate(value ~ cell_types + variable, ne_exp.t.m, length) #no individual points need to be plotted

#check that effect consistent across each gene in our data
ggplot(ne_exp.t.m) + 
  geom_quasirandom(mapping = aes(x = variable, y = value), dodge.width=.8, cex=1, shape = 21) +
  geom_pointrange(mapping = aes(x = variable, y = value, col = gene_use), stat = "summary", shape=19, fun.min = function(z) {quantile(z,0.25)}, 
                  fun.max = function(z) {quantile(z,0.75)}, 
                  fun = median, size = 0.8, fatten = 5, linetype = 2) +
  facet_wrap(. ~ cell_types, strip.position = 'bottom', ncol = 11)

p2 = ggplot(ne_exp.t.m, mapping = aes(x = gene_use, y = value)) + 
  geom_quasirandom(dodge.width=.8, cex=1, aes(x = gene_use, y = value), shape = 21, fill = 'transparent', col = 'transparent') + 
  facet_wrap(. ~ cell_types, strip.position = 'bottom', ncol = 11) +
  theme_pubr() + 
  theme(axis.text.x = element_blank(), 
        panel.spacing = unit(0, "mm"), 
        strip.background = element_blank(),
        panel.background = element_rect(fill = alpha("#b3b2b2", 0.5)), 
        strip.placement = "outside", 
        legend.position = 'none', 
        axis.ticks.x = element_blank()) +
  labs(y = 'Log2(TPM) expression', x = '') + 
  geom_pointrange(mapping = aes(col = gene_use), stat = "summary", shape=19, fun.min = function(z) {quantile(z,0.25)}, 
                  fun.max = function(z) {quantile(z,0.75)}, 
                  fun = median, size = 0.8, fatten = 5, linetype = 2) + 
  scale_color_manual(values = c('#078e38', '#61c5e6', '#3a93d0', '#1063a1', '#20335a'))

#smooth muscle comparison
#sm_exp = all_data_final_log2tpm[,row.names(all_data_final_meta[all_data_final_meta$cell_types %in% c("Fetal cardiac smooth muscle cells", "Fetal other smooth muscle cells", "Adult other smooth muscle cells", "Smooth_muscle_teratoma"), ])]
sm_exp = all_data_final_log2tpm[,row.names(all_data_final_meta[all_data_final_meta$cell_types %in% c("Fetal other smooth muscle cells", "Adult other smooth muscle cells", "Smooth_muscle_teratoma"), ])]
sm_exp.t = as.data.frame(t(as.data.frame(sm_exp)))
sm_exp.t$cell_types = all_data_final_meta[row.names(sm_exp.t),]$cell_types
sm_exp.t$sample = row.names(sm_exp.t)
sm_exp.t.m = reshape2::melt(sm_exp.t[, c('cell_types', 'sample', 'IGF2', 'TAGLN', 'ACTA2', 'MYH11')], id.vars = c('cell_types', 'sample'))

sm_exp.t.m$gene_use = NA
for(i in 1:nrow(sm_exp.t.m)){
  if(sm_exp.t.m$variable[i] %in% c('TAGLN', 'ACTA2', 'MYH11')) sm_exp.t.m$gene_use[i] = 'adult'
  if(sm_exp.t.m$variable[i] == 'IGF2') sm_exp.t.m$gene_use[i] = 'fetal'
}

sm_exp.t.m$gene_use = factor(sm_exp.t.m$gene_use, levels = c('fetal', 'adult'))
#sm_exp.t.m$cell_types = factor(sm_exp.t.m$cell_types, levels = c("Fetal cardiac smooth muscle cells", "Fetal other smooth muscle cells", "Adult other smooth muscle cells", "Smooth_muscle_teratoma"))
sm_exp.t.m$cell_types = factor(sm_exp.t.m$cell_types, levels = c("Fetal other smooth muscle cells", "Adult other smooth muscle cells", "Smooth_muscle_teratoma"))

aggregate(value ~ cell_types + variable, sm_exp.t.m, length) #add data points for GCT data for IGF2

#check that effect consistent across each gene in our data
ggplot(sm_exp.t.m) + 
  geom_quasirandom(mapping = aes(x = variable, y = value), dodge.width=.8, cex=1, shape = 21) +
  geom_pointrange(mapping = aes(x = variable, y = value, col = gene_use), stat = "summary", shape=19, fun.min = function(z) {quantile(z,0.25)}, 
                  fun.max = function(z) {quantile(z,0.75)}, 
                  fun = median, size = 0.8, fatten = 5, linetype = 2) +
  facet_wrap(. ~ cell_types, strip.position = 'bottom', ncol = 3)

p3 = ggplot(sm_exp.t.m, mapping = aes(x = gene_use, y = value)) + 
  geom_quasirandom(dodge.width=.8, cex=1, aes(x = gene_use, y = value), shape = 21, fill = 'transparent', col = 'transparent') + 
  geom_beeswarm(data = sm_exp.t.m[sm_exp.t.m$cell_types == 'Smooth_muscle_teratoma' & sm_exp.t.m$variable == 'IGF2', ], mapping = aes(x = gene_use, y = value), dodge.width=1, cex=1, shape = 21, fill = 'black', col = 'black') + 
  facet_wrap(. ~ cell_types, strip.position = 'bottom', ncol = 8) +
  theme_pubr() + 
  theme(axis.text.x = element_blank(), 
        panel.spacing = unit(0, "mm"), 
        strip.background = element_blank(),
        panel.background = element_rect(fill = alpha("#b3b2b2", 0.5)), 
        strip.placement = "outside", 
        legend.position = 'none', 
        axis.ticks.x = element_blank()) +
  labs(y = 'Log2(TPM) expression', x = '') + 
  geom_pointrange(mapping = aes(col = gene_use), stat = "summary", shape=19, fun.min = function(z) {quantile(z,0.25)}, 
                  fun.max = function(z) {quantile(z,0.75)}, 
                  fun = median, size = 0.8, fatten = 5, linetype = 2) + 
  scale_color_manual(values = c('#078e38', '#61c5e6'))

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/06_FIGURES/log2tpm_expression_gct_vs_ref_tissue20220524.pdf', width = 14, height = 9, useDingbats = F)
ggdraw() +
  draw_plot(p1, 0, 2/3, 8/11, 1/3) +
  draw_plot(p2, 0, 1/3, 1, 1/3) +
  draw_plot(p3, 0, 0, 4.25/11, 1/3) +
  draw_plot_label(c("A", "B", "C"), c(0, 0, 0), c(1, 2/3, 1/3), size = 15)
dev.off()


```

###Old


```{r}

library(ggplot2)
library(reshape2)
library(cowplot)
library(Seurat)
library(ggbeeswarm)
library(dplyr)
library(ggpubr)
library(gridExtra)

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/original_nature_submission/reference_datasets')

#read in scRNA seurat objects containing the count data
testis_ref = readRDS('postnatal_testis_final.rds') #neonatal and adult testis - Sohni et al., no downsampling, 10x, scRNA
fetal_brain = readRDS('pfc_final.rds') #fetal brain tissues - Zhong et al., no downsampling, smartseq2, scRNA
adult_brain = readRDS('adult_brain_li.rds') #adult brain - Li et al., no downsampling, 10x for snRNA from adult brains
fetal_muscle = readRDS('fetal_muscle_cao.rds') #fetal cardiac, smooth and skeletal muscle - Cao et al., each cell type downsampled to 500 cells, snRNA, sci-RNA-seq3
adult_smooth_musc = readRDS('adult_smooth_muscle.rds') #adult cardiac, smooth and skeletal muscle - Litviňuková et al., downsampled to 1000 cells, sc and snRNA, 10x

#read in our TPM data
gct_tpm = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/germ_cell_lcm_tpm_counts_filtered_samples_20210812.txt', header = T, sep = '\t', stringsAsFactors = F)

#merge reference scRNA data and restrict to genes also found in our data
ref_scrna = merge(testis_ref, y = c(fetal_brain, adult_brain, fetal_muscle, adult_smooth_musc), add.cell.ids = c("testis", "fetal_brain", "adult_brain","fetal_smooth_muscle", "adult_smooth_muscle"), project = 'ref_scrna_data') #49128 features across 32036 samples within 1 assay 

genes_to_keep = row.names(ref_scrna)[row.names(ref_scrna) %in% row.names(gct_tpm)]

ref_scrna_final = CreateSeuratObject(counts = ref_scrna@assays$RNA@counts[genes_to_keep,], project = 'final_ref_scrna', assay = 'RNA', meta.data = ref_scrna@meta.data) #34866 features across 32036 samples within 1 assay

ref_scrna_final = subset(ref_scrna_final, subset = cell_types != 'Microglia' & cell_types != 'Pre-Spermatogonia') #34866 features across 31692 samples within 1 assay

ref_scrna_final = NormalizeData(object = ref_scrna_final, normalization.method = "LogNormalize")
all.genes <- rownames(ref_scrna_final)
ref_scrna_final <- ScaleData(ref_scrna_final, features = all.genes)
ref_scrna_final@meta.data$cell_types = factor(ref_scrna_final@meta.data$cell_types)

levels(ref_scrna_final@meta.data$cell_types) = c("Adult astrocytes", "Fetal astrocytes", "Fetal excitatory neurons", "Adult excitatory neurons", "Fetal cardiac smooth muscle cells", "Adult interneurons", "Fetal interneurons", "Fetal other smooth muscle cells", "Fetal neuronal stem cells", "Adult oligodendrocytes", "Adult oligodendrocyte progenitor cells", "Fetal oligodendrocyte progenitor cells", "PGC-like", "Adult other smooth muscle cells", "Spermatid", "Spermatocytes", "Spermatogonia")

saveRDS(ref_scrna_final, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/reference_scrna.rds')

#ref_scrna_final = readRDS('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/reference_scrna.rds')

ref_scrna_final@meta.data$cell_types = as.factor(ref_scrna_final@meta.data$cell_types)

ref_scrna_final@meta.data$cell_types = factor(ref_scrna_final@meta.data$cell_types, levels = rev(c("PGC-like", "Spermatogonia", "Spermatocytes", "Spermatid", "Fetal neuronal stem cells", "Fetal oligodendrocyte progenitor cells", "Adult oligodendrocyte progenitor cells", "Adult oligodendrocytes", "Fetal astrocytes", "Adult astrocytes", "Fetal interneurons", "Adult interneurons", "Fetal excitatory neurons", "Adult excitatory neurons", "Fetal cardiac smooth muscle cells", "Fetal other smooth muscle cells", "Adult other smooth muscle cells")))

#limit gct tpm counts to genes shared between datasets
gct_tpm_filtered = gct_tpm[row.names(ref_scrna_final),]
gct_log2tpm <- log2(as.matrix(gct_tpm_filtered) + 1)
rna_cuts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '') #descriptive column contains apostrophes
row.names(rna_cuts) = rna_cuts$Sample

#extract ref data counts and convert to tpm
ref_scrna_final_df = as.data.frame(ref_scrna_final@assays$RNA@counts)
write.table(ref_scrna_final_df, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/reference_scrna_counts.txt', col.names = T, row.names = T, sep = '\t', quote = F)

ref_scrna_final_tpm <- t( t(ref_scrna_final_df) * 1e6 / colSums(ref_scrna_final_df)) #tpm conversion, 
ref_scrna_final_log2tpm <- log2(as.matrix(ref_scrna_final_tpm) + 1)

write.table(ref_scrna_final_log2tpm, '/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/reference_scrna_log2tpm.txt', col.names = T, row.names = T, sep = '\t', quote = F)

#ref_scrna_final_log2tpm = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/reference_scrna_log2tpm.txt', header = T, sep = '\t', stringsAsFactors = F)

ref_scrna_final@meta.data$sample = row.names(ref_scrna_final@meta.data)
ref_scrna_final@meta.data$source = 'scrna'
ref_scrna_final_meta = ref_scrna_final@meta.data[, c('sample', 'cell_types', 'source')]

rna_cuts$cell_types = rna_cuts$Updated.Description
rna_cuts$sample = row.names(rna_cuts)
rna_cuts$source = 'lcm'
gct_final_meta = rna_cuts[, c('sample', 'cell_types', 'source')]

all_data_final_log2tpm = cbind(ref_scrna_final_log2tpm, gct_log2tpm)
all_data_final_meta = rbind(ref_scrna_final_meta, gct_final_meta)

colnames(all_data_final_log2tpm) = gsub('\\.', '_', colnames(all_data_final_log2tpm))
row.names(all_data_final_meta) = gsub('-', '_', row.names(all_data_final_meta))
row.names(all_data_final_meta) = gsub('\\.', '_', row.names(all_data_final_meta))

#germ cell comparison
gc_exp = all_data_final_log2tpm[,row.names(all_data_final_meta[all_data_final_meta$cell_types %in% c('PGC-like', 'Spermatogonia', 'Spermatocytes', 'Spermatid', 'GCNIS', 'Seminoma', 'Dysgerminoma', 'Seminiferous_tubule'), ])]
gc_exp.t = as.data.frame(t(as.data.frame(gc_exp)))
gc_exp.t$cell_types = all_data_final_meta[row.names(gc_exp.t),]$cell_types
gc_exp.t$sample = row.names(gc_exp.t)
gc_exp.t.m = reshape2::melt(gc_exp.t[, c('cell_types', 'sample', 'POU5F1', 'NANOG', 'SOX17', 'DDX4', 'TNP1')], id.vars = c('cell_types', 'sample'))

gc_exp.t.m$gene_use = NA
for(i in 1:nrow(gc_exp.t.m)){
  if(gc_exp.t.m$variable[i] %in% c('POU5F1', 'NANOG', 'SOX17')) gc_exp.t.m$gene_use[i] = 'fetal'
  if(gc_exp.t.m$variable[i] == 'DDX4') gc_exp.t.m$gene_use[i] = 'gen_gc'
  if(gc_exp.t.m$variable[i] == 'TNP1') gc_exp.t.m$gene_use[i] = 'spermatid'
}

gc_exp.t.m$gene_use = factor(gc_exp.t.m$gene_use, levels = c('fetal', 'gen_gc', 'spermatid'))
gc_exp.t.m$cell_types = factor(gc_exp.t.m$cell_types, levels = c('PGC-like', 'Spermatogonia', 'Spermatocytes', 'Spermatid', 'Seminiferous_tubule', 'GCNIS', 'Seminoma', 'Dysgerminoma'))

aggregate(value ~ cell_types + variable, gc_exp.t.m, length) #Dysgerminoma will need individual data points for DDX4 and TNP1 plotting

p1 = ggplot(gc_exp.t.m, mapping = aes(x = gene_use, y = value)) + 
  geom_quasirandom(dodge.width=.8, cex=1, aes(x = gene_use, y = value), shape = 21, fill = 'transparent', col = 'transparent') + 
  geom_beeswarm(data = gc_exp.t.m[gc_exp.t.m$cell_types == 'Dysgerminoma' & gc_exp.t.m$variable %in% c('DDX4', 'TNP1'), ], mapping = aes(x = gene_use, y = value), dodge.width=1, cex=1, shape = 21, fill = 'black', col = 'black') + 
  facet_wrap(. ~ cell_types, strip.position = 'bottom', ncol = 8) +
  theme_pubr() + 
  theme(axis.text.x = element_blank(), 
        panel.spacing = unit(0, "mm"), 
        strip.background = element_blank(), 
        strip.placement = "outside", 
        legend.position = 'none', 
        axis.ticks.x = element_blank()) +
  labs(y = 'Log2(TPM) expression', x = '') + 
  geom_pointrange(mapping = aes(col = gene_use), stat = "summary", shape=19, fun.min = function(z) {quantile(z,0.25)}, 
                  fun.max = function(z) {quantile(z,0.75)}, 
                  fun = median, size = 0.8, fatten = 5, linetype = 2) + 
  scale_color_manual(values = c('#FF6C5C', '#58CCED', '#072F5F'))

#neural comparison
ne_exp = all_data_final_log2tpm[,row.names(all_data_final_meta[all_data_final_meta$cell_types %in% c('Neuroepithelium', 'Fetal interneurons', 'Fetal neuronal stem cells', 'Fetal oligodendrocyte progenitor cells', 'Fetal astrocytes', 'Fetal excitatory neurons', 'Adult interneurons', 'Adult oligodendrocyte progenitor cells', 'Adult astrocytes', 'Adult excitatory neurons', 'Adult oligodendrocytes'), ])]
ne_exp.t = as.data.frame(t(as.data.frame(ne_exp)))
ne_exp.t$cell_types = all_data_final_meta[row.names(ne_exp.t),]$cell_types
ne_exp.t$sample = row.names(ne_exp.t)
ne_exp.t.m = reshape2::melt(ne_exp.t[, c('cell_types', 'sample', 'PAX6', 'SOX2', 'NES', 'MBP', 'SLC1A2', 'RBFOX3', 'GAD1')], id.vars = c('cell_types', 'sample'))

ne_exp.t.m$gene_use = NA
for(i in 1:nrow(ne_exp.t.m)){
  if(ne_exp.t.m$variable[i] %in% c('PAX6', 'SOX2', 'NES')) ne_exp.t.m$gene_use[i] = 'fetal'
  if(ne_exp.t.m$variable[i] == 'MBP') ne_exp.t.m$gene_use[i] = 'oligo'
  if(ne_exp.t.m$variable[i] == 'SLC1A2') ne_exp.t.m$gene_use[i] = 'astro'
  if(ne_exp.t.m$variable[i] == 'RBFOX3') ne_exp.t.m$gene_use[i] = 'neun'
  if(ne_exp.t.m$variable[i] == 'GAD1') ne_exp.t.m$gene_use[i] = 'inter'
}

ne_exp.t.m$gene_use = factor(ne_exp.t.m$gene_use, levels = c('fetal', 'oligo', 'astro', 'neun', 'inter'))
ne_exp.t.m$cell_types = factor(ne_exp.t.m$cell_types, levels = c('Fetal neuronal stem cells', 'Fetal oligodendrocyte progenitor cells', 'Fetal astrocytes', 'Fetal excitatory neurons', 'Fetal interneurons', 'Adult oligodendrocyte progenitor cells', 'Adult oligodendrocytes', 'Adult astrocytes', 'Adult excitatory neurons', 'Adult interneurons', 'Neuroepithelium'))

aggregate(value ~ cell_types + variable, ne_exp.t.m, length)

p2 = ggplot(ne_exp.t.m, mapping = aes(x = gene_use, y = value)) + 
  geom_quasirandom(dodge.width=.8, cex=1, aes(x = gene_use, y = value), shape = 21, fill = 'transparent', col = 'transparent') + 
  facet_wrap(. ~ cell_types, strip.position = 'bottom', ncol = 11) +
  theme_pubr() + 
  theme(axis.text.x = element_blank(), 
        panel.spacing = unit(0, "mm"), 
        strip.background = element_blank(), 
        strip.placement = "outside", 
        legend.position = 'none', 
        axis.ticks.x = element_blank()) +
  labs(y = 'Log2(TPM) expression', x = '') + 
  geom_pointrange(mapping = aes(col = gene_use), stat = "summary", shape=19, fun.min = function(z) {quantile(z,0.25)}, 
                  fun.max = function(z) {quantile(z,0.75)}, 
                  fun = median, size = 0.8, fatten = 5, linetype = 2) + 
  scale_color_manual(values = c('#FF6C5C', '#58CCED', '#3895D3', '#1261A0', '#072F5F'))

#smooth muscle comparison
sm_exp = all_data_final_log2tpm[,row.names(all_data_final_meta[all_data_final_meta$cell_types %in% c("Fetal cardiac smooth muscle cells", "Fetal other smooth muscle cells", "Adult other smooth muscle cells", "Smooth_muscle_teratoma"), ])]
sm_exp.t = as.data.frame(t(as.data.frame(sm_exp)))
sm_exp.t$cell_types = all_data_final_meta[row.names(sm_exp.t),]$cell_types
sm_exp.t$sample = row.names(sm_exp.t)
sm_exp.t.m = reshape2::melt(sm_exp.t[, c('cell_types', 'sample', 'IGF2', 'TAGLN', 'ACTA2', 'MYH11')], id.vars = c('cell_types', 'sample'))

sm_exp.t.m$gene_use = NA
for(i in 1:nrow(sm_exp.t.m)){
  if(sm_exp.t.m$variable[i] %in% c('TAGLN', 'ACTA2', 'MYH11')) sm_exp.t.m$gene_use[i] = 'adult'
  if(sm_exp.t.m$variable[i] == 'IGF2') sm_exp.t.m$gene_use[i] = 'fetal'
}

sm_exp.t.m$gene_use = factor(sm_exp.t.m$gene_use, levels = c('fetal', 'adult'))
sm_exp.t.m$cell_types = factor(sm_exp.t.m$cell_types, levels = c("Fetal cardiac smooth muscle cells", "Fetal other smooth muscle cells", "Adult other smooth muscle cells", "Smooth_muscle_teratoma"))

aggregate(value ~ cell_types + variable, sm_exp.t.m, length) #add data points for GCT data for IGF2

p3 = ggplot(sm_exp.t.m, mapping = aes(x = gene_use, y = value)) + 
  geom_quasirandom(dodge.width=.8, cex=1, aes(x = gene_use, y = value), shape = 21, fill = 'transparent', col = 'transparent') + 
  geom_beeswarm(data = sm_exp.t.m[sm_exp.t.m$cell_types == 'Smooth_muscle_teratoma' & sm_exp.t.m$variable == 'IGF2', ], mapping = aes(x = gene_use, y = value), dodge.width=1, cex=1, shape = 21, fill = 'black', col = 'black') + 
  facet_wrap(. ~ cell_types, strip.position = 'bottom', ncol = 8) +
  theme_pubr() + 
  theme(axis.text.x = element_blank(), 
        panel.spacing = unit(0, "mm"), 
        strip.background = element_blank(), 
        strip.placement = "outside", 
        legend.position = 'none', 
        axis.ticks.x = element_blank()) +
  labs(y = 'Log2(TPM) expression', x = '') + 
  geom_pointrange(mapping = aes(col = gene_use), stat = "summary", shape=19, fun.min = function(z) {quantile(z,0.25)}, 
                  fun.max = function(z) {quantile(z,0.75)}, 
                  fun = median, size = 0.8, fatten = 5, linetype = 2) + 
  scale_color_manual(values = c('#FF6C5C', '#58CCED'))

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/log2tpm_expression_gct_vs_ref_tissue20210812.pdf', width = 14, height = 9, useDingbats = F)
ggdraw() +
  draw_plot(p1, 0, 2/3, 8/11, 1/3) +
  draw_plot(p2, 0, 1/3, 1, 1/3) +
  draw_plot(p3, 0, 0, 4.25/11, 1/3) +
  draw_plot_label(c("A", "B", "C"), c(0, 0, 0), c(1, 2/3, 1/3), size = 15)
dev.off()

```

##Measuring transcriptional impact of chromothripsis in PD43299 (Extended data figure 8)

```{r}

library(Seurat)

#genes to remove to leave ones of more biological interest to compare
source('/lustre/scratch119/casm/team294rr/to3/testes/tumour/RNA_analysis/geneSets.R')
prot_cod = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/RNA_analysis/protein_coding_genes.txt', header = F, sep='\t', stringsAsFactors = F)[,1]

tpm = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/germ_cell_lcm_tpm_counts_filtered_samples_20210812.txt', header = T, sep = '\t', stringsAsFactors = F)
log2tpm <- log2(as.matrix(tpm) + 1)
log2tpm = log2tpm[(row.names(log2tpm) %in% prot_cod) & !(row.names(log2tpm) %in% unique(c(hgGenes, hkGenes, igGenes, cc.genes.updated.2019$s.genes, cc.genes.updated.2019$g2m.genes))),] #remove genes whose expression will be related to normal contamination or cellular functions not related to phenotype
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '')
yst_samples = metadata[metadata$Updated.Description == "Yolk_sac_tumour",]$Sample
PD43299_samples = metadata[metadata$Case.ID == 'PD43299',]$Sample

cor_mat = cor(log2tpm[,yst_samples], method = "pearson")
m = cor_mat

m.long = data.frame(row=rownames(m)[row(m)[upper.tri(m)]], 
           col=colnames(m)[col(m)[upper.tri(m)]], 
           corr=m[upper.tri(m)])

m.long$row.ID = substr(m.long$row, 3, 7)
m.long$col.ID = substr(m.long$col, 3, 7)

m.long.intra = m.long[m.long$row.ID == m.long$col.ID,]

nRand = 1000
set.seed(42)
var_corr = c()

choose(nrow(m.long.intra), 6) #170230452

for(i in 1:nRand){
  rand_sample = sample(m.long.intra$corr, 6, replace = F)
  
  var_corr = c(var_corr, (var(rand_sample)))
}

hist(var_corr, breaks = 100)

pd43299.long = m.long.intra[m.long.intra$row.ID == "43299",]

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/rand_sampling_intra_tumour_yst_transcriptional_heterogeneity_20210920.pdf', height = 3, width = 4)
ggplot() + geom_density(mapping = aes(x = var_corr), fill = '#999999') + xlim(0, 0.1) + theme_classic() + geom_vline(xintercept = var(pd43299.long$corr)) + labs(x = 'Variance in transcriptional similarity within yolk sac tumour elements from the same tumour') + coord_cartesian(expand = F)
dev.off()

pVal = length(var_corr[var_corr > var(pd43299.long$corr)]) / length(var_corr) #0.89, 89% of samples of transcriptional similarity within a single tumour had a larger variance in transcriptional values, PD43299 actually remains remarkably homogeneous, despite chromothripsis

```

#05_COMBINED_DNA_RNA_ANALYSES

##Comparing genomic and transcriptome similarity (Fig. 2c)

###Genomic

```{r}

library(ggplot2)
library(ggpubr)
library(ggbeeswarm)

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/05_COMBINED_DNA_RNA_ANALYSES/genomic/')

#patient = c("PD43296", "PD43298", "PD45544", "PD46269") #copy over relevant files
#for(i in patient){
  system(paste0('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/03_DNA_ANALYSES/phylogeny/ndDPClust/', i, '_filtered_clusters_mean_CCF.txt .'))
#}

system('cp /lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/03_DNA_ANALYSES/phylogeny/DPClust/PD45543_3000iters_1000burnin_bestClusterInfo.txt .') #changed headers and name to match other files 16.3.22

patient = c("PD43296", "PD43298", "PD45543", "PD45544", "PD46269") #only mixed tumours with multiple histologies represented within each biopsy taken

ignore_clusters = list(PD43296 = c(12), PD43298 = c(NULL), PD45543 = c(NULL), PD45544 = c(3), PD46269 = c(2, 3)) #low VAF/artefactual clusters that shouldn't be included in similarity metric

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/WGS_summary_metadata.txt', header = T, sep = '\t', stringsAsFactors = F)

#between histologies
interhisto_g_list = list()
for(i in patient){ #per patient
  mut_clusters = read.table(paste0(i, '_filtered_clusters_mean_CCF.txt'), header = T, sep = '\t', stringsAsFactors = F) #dataframe of mutations assigned per cluster by multiDPClust
  mut_clusters = mut_clusters[!mut_clusters$cluster.no %in% ignore_clusters[[i]],] # remove bad clusters
  assessed_histo = c() #record of histologies already analysed
  g_prox_score_vec = c() #vector of per patient comparison scores
  
  patient_meta = manifest[manifest$Case.ID == i & !is.na(manifest$Median.CCF),] #patient samples that are included in the phylogeny
  
  for(j in unique(patient_meta$Histology)){ #per histology
    
    assessed_histo = c(assessed_histo, j) #add histology to list not to compare against
    
    if(j != unique(patient_meta$Histology)[length(unique(patient_meta$Histology))]){ #no groups left for the final histology to be compared against
        
        for(k in patient_meta[patient_meta$Histology == j,]$Sample){ #each patient sample per histology
          choice_sample_clusters_vec = mut_clusters[mut_clusters[,paste0(k, "_subclonalFraction")] > 0.1,]$cluster.no #clusters found at a CCF of at least 10% in sample of interest
          
          for(l in patient_meta[patient_meta$Histology != j & !(patient_meta$Histology %in% assessed_histo),]$Sample){ #each sample from another histology
          comparator_sample_clusters_vec = mut_clusters[mut_clusters[,paste0(l, "_subclonalFraction")] > 0.1,]$cluster.no #clusters found at a CCF of at least 10% in sample of interest
          
          g_prox = sum(mut_clusters[mut_clusters$cluster.no %in% intersect(choice_sample_clusters_vec, comparator_sample_clusters_vec), ]$mut.no) / ((sum(mut_clusters[mut_clusters$cluster.no %in% choice_sample_clusters_vec, ]$mut.no) + sum(mut_clusters[mut_clusters$cluster.no %in% comparator_sample_clusters_vec, ]$mut.no))  / 2) #average sharedness of mutations between all clones in each sample
        g_prox_score_vec = c(g_prox_score_vec, g_prox) #add pairwise comparison to patient vector
      }
    }
    }
  }
  interhisto_g_list = c(interhisto_g_list, list(g_prox_score_vec))
}

names(interhisto_g_list) = patient
sapply(interhisto_g_list, median)

#within histologies
intrahisto_g_list = list()

for(i in patient){ #per patient
  
  mut_clusters = read.table(paste0(i, '_filtered_clusters_mean_CCF.txt'), header = T, sep = '\t', stringsAsFactors = F) #dataframe of mutations assigned per cluster by multiDPClust
  mut_clusters = mut_clusters[!mut_clusters$cluster.no %in% ignore_clusters[[i]],] # remove bad clusters
  
  patient_meta = manifest[manifest$Case.ID == i & !is.na(manifest$Median.CCF),]

  g_prox_score_vec = c() #vector of per patient comparison scores
  
  for(j in unique(patient_meta$Histology)){ #per histology
    if(nrow(patient_meta[patient_meta$Histology == j,]) > 1){ #if histology has more than one microbiopsy
      
      assessed_samples = c() #record of samples already analysed
      
      for(k in patient_meta[patient_meta$Histology == j,]$Sample){ #per sample
        
        assessed_samples = c(assessed_samples, k) #add to vector of analysed samples
        choice_sample_clusters_vec = mut_clusters[mut_clusters[,paste0(k, "_subclonalFraction")] > 0.1,]$cluster.no

        if(k != patient_meta[patient_meta$Histology == j,]$Sample[length(patient_meta[patient_meta$Histology == j,]$Sample)]){ #as the last sample cannot be compared against itself
          for(l in patient_meta[patient_meta$Histology == j & !(patient_meta$Sample %in% assessed_samples),]$Sample){
            comparator_sample_clusters_vec = mut_clusters[mut_clusters[,paste0(l, "_subclonalFraction")] > 0.1,]$cluster.no
            
           g_prox = sum(mut_clusters[mut_clusters$cluster.no %in% intersect(choice_sample_clusters_vec, comparator_sample_clusters_vec), ]$mut.no) / ((sum(mut_clusters[mut_clusters$cluster.no %in% choice_sample_clusters_vec, ]$mut.no) + sum(mut_clusters[mut_clusters$cluster.no %in% comparator_sample_clusters_vec, ]$mut.no))  / 2) #average sharedness of mutations between all clones in each sample
            g_prox_score_vec = c(g_prox_score_vec, g_prox) #add pairwise comparison to patient vector
        }
        }
        
      }
    }
  }
  intrahisto_g_list = c(intrahisto_g_list, list(g_prox_score_vec))
}

names(intrahisto_g_list) = patient
sapply(intrahisto_g_list, median)

#combine results
interhisto_g_df = data.frame(unlist(interhisto_g_list))
interhisto_g_df$patient = substr(row.names(interhisto_g_df), 0, 7)
interhisto_g_df$comparison = 'Inter-histology'
row.names(interhisto_g_df) = NULL
colnames(interhisto_g_df) = c('gen_prox', 'patient', 'comparison')

intrahisto_g_df = data.frame(unlist(intrahisto_g_list))
intrahisto_g_df$patient = substr(row.names(intrahisto_g_df), 0, 7)
intrahisto_g_df$comparison = 'Intra-histology'
row.names(intrahisto_g_df) = NULL
colnames(intrahisto_g_df) = c('gen_prox', 'patient', 'comparison')

all_histo_g_df = rbind(interhisto_g_df, intrahisto_g_df)
all_histo_g_df$comparison = factor(all_histo_g_df$comparison, levels = c('Intra-histology', 'Inter-histology'))
all_histo_g_df$patient = as.factor(all_histo_g_df$patient)

#are these differences significant? initially using Wilcoxon rank sum test
wilcox.test(gen_prox ~ comparison, all_histo_g_df[all_histo_g_df$patient == 'PD43296',]) #W = 26462, p-value = 0.3298
wilcox.test(gen_prox ~ comparison, all_histo_g_df[all_histo_g_df$patient == 'PD43298',]) #W = 12376, p-value = 0.1501
wilcox.test(gen_prox ~ comparison, all_histo_g_df[all_histo_g_df$patient == 'PD45543',]) #W = 4, p-value = 1
wilcox.test(gen_prox ~ comparison, all_histo_g_df[all_histo_g_df$patient == 'PD45544',]) #W = 97.5, p-value = 0.2571
wilcox.test(gen_prox ~ comparison, all_histo_g_df[all_histo_g_df$patient == 'PD46269',]) #W = 193, p-value = 0.00195

#retry with permutation test for reviewer
permutation_list = list()
set.seed(42)

#check how many permutations for label switching can be achieved per tumour
patient = c("PD43296", "PD43298", "PD45543", "PD45544", "PD46269") 

for(i in patient){
  df = all_histo_g_df[all_histo_g_df$patient == i,]
  max_perm = choose(nrow(df), nrow(df[df$comparison == 'Inter-histology',]))
  print(paste0(i, ' ', max_perm))
}

#Only 15 possible combinations for PD45543 - here will we compare the other 14 combinations with the one we observe as 1000 permutations would simply resample the same sample space many times over

#PD45543 assessment
i = "PD45543"
df = all_histo_g_df[all_histo_g_df$patient == i,]
all_comb = data.frame(combn(nrow(df), nrow(df[df$comparison == 'Inter-histology',]), simplify = T)) #all possible combinations of label switching
which(df$comparison == 'Inter-histology') #X1 matches our observations
remaining_comb = all_comb[, c(2:15)]
median_diff_vec = c()

for(j in 1:ncol(remaining_comb)){
  df$label.switch = 'Intra-histology'
  df$label.switch[remaining_comb[,j]] = 'Inter-histology'
  median_diff_vec = c(median_diff_vec, abs(median(df[df$label.switch == 'Inter-histology',]$gen_prox) - median(df[df$label.switch == 'Intra-histology',]$gen_prox)))
}

permutation_list = c(permutation_list, list(median_diff_vec))

#remaining tumours
patient = c("PD43296", "PD43298", "PD45544", "PD46269")
nRand = 1000


for(i in patient){
  median_diff_vec = c()
  df = all_histo_g_df[all_histo_g_df$patient == i,]
  for(j in 1:nRand){
    df$label.switch = sample(df$comparison, length(df$comparison), replace = F)
    median_diff_vec = c(median_diff_vec, abs(median(df[df$label.switch == 'Inter-histology',]$gen_prox) - median(df[df$label.switch == 'Intra-histology',]$gen_prox)))
  }
  permutation_list = c(permutation_list, list(median_diff_vec))
}

names(permutation_list) = c('PD45543', patient)

#what are the difference between the inter and intra histology values that we observe?
obs_data = aggregate(gen_prox ~ patient + comparison, all_histo_g_df, median)
patient = c("PD45543", "PD43296", "PD43298", "PD45544", "PD46269")

obs_med_diff = c()
for(i in patient){
  obs_med_diff = rbind(obs_med_diff, c(i, abs(obs_data[obs_data$patient == i & obs_data$comparison == "Intra-histology",]$gen_prox - obs_data[obs_data$patient == i & obs_data$comparison == "Inter-histology",]$gen_prox)))
}

obs_med_diff = data.frame(obs_med_diff)
names(obs_med_diff) = c('patient', 'observed_median_diff')
obs_med_diff$observed_median_diff = as.numeric(obs_med_diff$observed_median_diff)

#what is the probability this was observed by chance?
obs_med_diff$pval = NA
for(i in patient){
  obs_med_diff[obs_med_diff$patient == i,]$pval = sum(permutation_list[[i]] >= obs_med_diff[obs_med_diff$patient == i,]$observed_median_diff) / length(permutation_list[[i]]) #what proportion of label switched combinations generates a larger difference between the inter and intra-histology genomic proximity?
}

#patient observed_median_diff      pval
#PD45543          0.036548715 0.4285714
#PD43296          0.010493501 0.3550000
#PD43298          0.004696293 0.7200000
#PD45544          0.010150518 0.2570000
#PD46269          0.075844487 0.0000000

library(ggbeeswarm)
library(ggplot2)

ggplot(all_histo_g_df) + 
  geom_quasirandom(mapping = aes(x = gen_prox, y = comparison, col = comparison), width = 0.3, groupOnX = F) +
  geom_pointrange(mapping = aes(x = gen_prox, y = comparison), stat = "summary", 
                  shape=19, 
                  fun.min = function(z) {quantile(z,0.25)},
                  fun.max = function(z) {quantile(z,0.75)},
                  fun = median, fill="black") +
  facet_grid(patient ~ ., scales = 'free_y', space = 'free_y') + theme_pubr() + theme(axis.text.y = element_blank(), panel.spacing = unit(0, 'mm'), strip.background = element_blank(), strip.placement = "outside", strip.text = element_text(size = 11)) + xlab('Genetic similarity') + ylab ('') + scale_y_discrete(position = "right")  + scale_x_continuous(expand = c(0,0), limits = c(0,1))

ggplot(all_histo_g_df) + 
  geom_boxplot(mapping = aes(x = gen_prox, y = comparison, fill = comparison), width = 0.5) +
  facet_grid(patient ~ ., scales = 'free_y', space = 'free_y') + theme_pubr() + theme(axis.text.y = element_blank(), panel.spacing = unit(0, 'mm'), strip.background = element_blank(), strip.placement = "outside", strip.text = element_text(size = 11)) + xlab('Genetic similarity') + ylab ('') + scale_y_discrete(position = "right")  + scale_x_continuous(expand = c(0,0), limits = c(0,1)) + scale_fill_manual(values = c('grey25', 'grey75'), name = 'Comparison')

```

###Transcriptomic

Updated to recalculate TPM values for restricted set of genes used in the analysis. 25/5/22.

```{r}

#libraries to use
library(Seurat)
library(ggplot2)
library(ggpubr)
library(ggbeeswarm)

#genes to remove to leave ones of more biological interest to compare
source('/lustre/scratch119/casm/team294rr/to3/testes/tumour/RNA_analysis/geneSets.R')
prot_cod = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/RNA_analysis/protein_coding_genes.txt', header = F, sep='\t', stringsAsFactors = F)[,1]

#read in data
gct_raw_counts = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/germ_cell_lcm_raw_counts_filtered_samples_20210812.txt', header = T, sep = '\t', stringsAsFactors = F)
row.names(gct_raw_counts) = gct_raw_counts$Geneid

gct_raw_counts_subset = gct_raw_counts[(row.names(gct_raw_counts) %in% prot_cod) & !(row.names(gct_raw_counts) %in% unique(c(hgGenes, hkGenes, igGenes, cc.genes.updated.2019$s.genes, cc.genes.updated.2019$g2m.genes))),]

#generate log2(tpm) values for our modified smartseq2 data (needs adjustment for gene length)
x = gct_raw_counts_subset[, 7:ncol(gct_raw_counts_subset)]
row.names(x) = gct_raw_counts_subset$Geneid
x = x / gct_raw_counts_subset$Length #normalise for transcript length
gct_tpm <- t( t(x) * 1e6 / colSums(x)) #normalise to read depth
log2tpm <- log2(as.matrix(gct_tpm) + 1)

write.table(log2tpm, "/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/00_SUPPLEMENTARY_DATA/figure2c_log2tpm_mat_additional_source_data.txt", col.names = T, row.names = T, sep = '\t', quote = F)

#read in metadata and 
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '')
row.names(metadata) = metadata$Sample
inv_mixed_rna = metadata[metadata$Diagnosis == 'Post-pubertal mixed germ cell tumour' & metadata$Normal.Tumour == 'Tumour' & metadata$Updated.Description != 'GCNIS',]
inv_mixed_rna = inv_mixed_rna[inv_mixed_rna$Case.ID %in% c('PD43296', 'PD43298', 'PD45543', 'PD45544', 'PD46269'),] #make direct contrast in each tumour to their transcriptome

#between histologies
interhisto_t_list = list()
for(i in unique(inv_mixed_rna$Case.ID)){ #per patient
  assessed_histo = c() #record of histologies already analysed
  t_prox_score_vec = c() #vector per patient comparison scores
  for(j in unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i,]$Updated.Description)){
    assessed_histo = c(assessed_histo, j) #add histology to list to not compare against
    if(length(assessed_histo) != length(unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i,]$Updated.Description))){ #stop when you are at the last histology as you have already compared it to all of the others
       for(k in unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i & inv_mixed_rna$Updated.Description == j,]$Sample)){ #samples per histology per tumour
      choice_sample = log2tpm[, k] #vector from sample in histology of interest
      for(l in unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i & !(inv_mixed_rna$Updated.Description %in% assessed_histo),]$Sample)){
        comparator_sample = log2tpm[, l] #vector from a sample in another histology
        t_prox = cor(choice_sample, comparator_sample, method = 'pearson')
        t_prox_score_vec = c(t_prox_score_vec, t_prox)
    }
    }
    }
   
  }
  interhisto_t_list = c(interhisto_t_list, list(t_prox_score_vec))
}

names(interhisto_t_list) = unique(inv_mixed_rna$Case.ID)

sapply(interhisto_t_list, median)

#within histologies
intrahisto_t_list = list()
for(i in unique(inv_mixed_rna$Case.ID)){ #per patient
  t_prox_score_vec = c() #vector per patient comparison scores
  for(j in unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i,]$Updated.Description)){ #per histology
      if(nrow(inv_mixed_rna[inv_mixed_rna$Case.ID == i & inv_mixed_rna$Updated.Description == j,]) > 1){ #if histology has more than one microbiopsy
      assessed_samples = c() #record of samples already analysed
      for(k in unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i & inv_mixed_rna$Updated.Description == j,]$Sample)){ #per sample
        assessed_samples = c(assessed_samples, k) #add to vector analysed
        choice_sample = log2tpm[, k]
        
        if(k != unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i & inv_mixed_rna$Updated.Description == j,]$Sample)[length(unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i & inv_mixed_rna$Updated.Description == j,]$Sample))]){ #so the last sample isn't compared against itself
          for(l in unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i & inv_mixed_rna$Updated.Description == j & !(inv_mixed_rna$Sample %in% assessed_samples),]$Sample)){
          comparator_sample = log2tpm[, l]
          t_prox = cor(choice_sample, comparator_sample, method = 'pearson')
          t_prox_score_vec = c(t_prox_score_vec, t_prox)
        }
        }
      }
    }
  }
  intrahisto_t_list = c(intrahisto_t_list, list(t_prox_score_vec))
}

names(intrahisto_t_list) = unique(inv_mixed_rna$Case.ID)
sapply(intrahisto_t_list, median)

#control - against normal testis
norm_control = metadata[metadata$Normal.Tumour == 'Normal',]

norm_tumour_t_list = list()
for(i in unique(inv_mixed_rna$Case.ID)){ #per patient
  t_prox_score_vec = c() #vector per patient comparison scores
  for(j in inv_mixed_rna[inv_mixed_rna$Case.ID == i,]$Sample){
    choice_sample = log2tpm[, j]
    for(k in norm_control$Sample){
      comparator_sample = log2tpm[, k]
      t_prox = cor(choice_sample, comparator_sample, method = 'pearson')
      #t_prox = cosine(choice_sample, comparator_sample)[1,1]
      #t_prox = cor(choice_sample, comparator_sample, method = 'spearman')
      t_prox_score_vec = c(t_prox_score_vec, t_prox)
    }
  }
  norm_tumour_t_list = c(norm_tumour_t_list, list(t_prox_score_vec))
}

names(norm_tumour_t_list) = unique(inv_mixed_rna$Case.ID)

sapply(norm_tumour_t_list, median)

#add together
interhisto_t_df = data.frame(unlist(interhisto_t_list))
interhisto_t_df$patient = substr(row.names(interhisto_t_df), 0, 7)
interhisto_t_df$comparison = 'Inter-histology'
row.names(interhisto_t_df) = NULL
colnames(interhisto_t_df) = c('t_prox', 'patient', 'comparison')

intrahisto_t_df = data.frame(unlist(intrahisto_t_list))
intrahisto_t_df$patient = substr(row.names(intrahisto_t_df), 0, 7)
intrahisto_t_df$comparison = 'Intra-histology'
row.names(intrahisto_t_df) = NULL
colnames(intrahisto_t_df) = c('t_prox', 'patient', 'comparison')

norm_tumour_t_df = data.frame(unlist(norm_tumour_t_list))
norm_tumour_t_df$patient = substr(row.names(norm_tumour_t_df), 0, 7)
norm_tumour_t_df$comparison = 'Normal v Tumour'
row.names(norm_tumour_t_df) = NULL
colnames(norm_tumour_t_df) = c('t_prox', 'patient', 'comparison')

all_histo_t_df = rbind(interhisto_t_df, intrahisto_t_df, norm_tumour_t_df)

ggplot(all_histo_t_df) + 
  geom_quasirandom(mapping = aes(x = comparison, y = t_prox, col = comparison), width = 0.3) +
  geom_pointrange(mapping = aes(x = comparison, y = t_prox), stat = "summary", 
                  shape=19, 
                  fun.min = function(z) {quantile(z,0.25)},
                  fun.max = function(z) {quantile(z,0.75)},
                  fun = median, fill="black") +
  facet_grid(. ~ patient, scales = 'free_x', space = 'free_x', switch = 'x') + theme_pubr() + theme(axis.text.x = element_blank(), panel.spacing = unit(0, 'mm'), strip.background = element_blank(), strip.placement = "outside", strip.text = element_text(size = 11)) + ylab('Transcriptomic similarity') + xlab ('') + ylim(0, 1)

#are these differences significant? initially using Wilcoxon rank sum test
wilcox.test(t_prox ~ comparison, all_histo_t_df[all_histo_t_df$patient == 'PD43296' & all_histo_t_df$comparison %in% c('Inter-histology', 'Intra-histology'),]) #W = 163720, p-value < 2.2e-16
wilcox.test(t_prox ~ comparison, all_histo_t_df[all_histo_t_df$patient == 'PD43298' & all_histo_t_df$comparison %in% c('Inter-histology', 'Intra-histology'),]) #W = 71255, p-value < 2.2e-16
wilcox.test(t_prox ~ comparison, all_histo_t_df[all_histo_t_df$patient == 'PD45543' & all_histo_t_df$comparison %in% c('Inter-histology', 'Intra-histology'),]) #W = 10, p-value = 0.04009
wilcox.test(t_prox ~ comparison, all_histo_t_df[all_histo_t_df$patient == 'PD45544' & all_histo_t_df$comparison %in% c('Inter-histology', 'Intra-histology'),]) #W = 3759, p-value = 5.267e-16
wilcox.test(t_prox ~ comparison, all_histo_t_df[all_histo_t_df$patient == 'PD46269' & all_histo_t_df$comparison %in% c('Inter-histology', 'Intra-histology'),]) #W = 20651, p-value = 1.298e-09

#retry with permutation test for reviewer
permutation_list = list()
set.seed(42)

#check how many permutations for label switching can be achieved per tumour
patient = c("PD43296", "PD43298", "PD45543", "PD45544", "PD46269") 

for(i in patient){
  df = all_histo_t_df[all_histo_t_df$patient == i & all_histo_t_df$comparison != "Normal v Tumour",] #label swapping between intra and inter only
  max_perm = choose(nrow(df), nrow(df[df$comparison == 'Inter-histology',]))
  print(paste0(i, ' ', max_perm))
}

#well over 1000 in all cases
#all tumours
patient = c("PD45543", "PD43296", "PD43298", "PD45544", "PD46269")
nRand = 1000

for(i in patient){
  median_diff_vec = c()
  df = all_histo_t_df[all_histo_t_df$patient == i & all_histo_t_df$comparison != "Normal v Tumour",] #label swapping between intra and inter only
  for(j in 1:nRand){
    df$label.switch = sample(df$comparison, length(df$comparison), replace = F)
    median_diff_vec = c(median_diff_vec, abs(median(df[df$label.switch == 'Inter-histology',]$t_prox) - median(df[df$label.switch == 'Intra-histology',]$t_prox)))
  }
  permutation_list = c(permutation_list, list(median_diff_vec))
}

names(permutation_list) = patient

#what are the difference between the inter and intra histology values that we observe?
obs_t_data = aggregate(t_prox ~ patient + comparison, all_histo_t_df, median)
patient = c("PD45543", "PD43296", "PD43298", "PD45544", "PD46269")

obs_t_med_diff = data.frame()
for(i in patient){
  obs_t_med_diff = rbind(obs_t_med_diff, c(i, abs(obs_t_data[obs_t_data$patient == i & obs_t_data$comparison == "Intra-histology",]$t_prox - obs_t_data[obs_t_data$patient == i & obs_t_data$comparison == "Inter-histology",]$t_prox)))
}

#obs_t_med_diff = data.frame(obs_t_med_diff)
names(obs_t_med_diff) = c('patient', 'observed_median_diff')
obs_t_med_diff$observed_median_diff = as.numeric(obs_t_med_diff$observed_median_diff)

#what is the probability this was observed by chance?
obs_t_med_diff$pval = NA
for(i in patient){
  obs_t_med_diff[obs_t_med_diff$patient == i,]$pval = sum(permutation_list[[i]] >= obs_t_med_diff[obs_t_med_diff$patient == i,]$observed_median_diff) / length(permutation_list[[i]]) #what proportion of label switched combinations generates a larger difference between the inter and intra-histology transcriptomic proximity?
}

#patient observed_median_diff  pval
#PD45543           0.08199448 0.015
#PD43296           0.04012713 0.000
#PD43298           0.10661239 0.000
#PD45544           0.17628309 0.000
#PD46269           0.11245925 0.000

```

Old, without normalising for restricted gene set.

```{r}

#libraries to use
#library(lsa)
library(Seurat)
library(ggplot2)
library(ggpubr)
library(ggbeeswarm)

#genes to remove to leave ones of more biological interest to compare
source('/lustre/scratch119/casm/team294rr/to3/testes/tumour/RNA_analysis/geneSets.R')
prot_cod = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/RNA_analysis/protein_coding_genes.txt', header = F, sep='\t', stringsAsFactors = F)[,1]

#read in data
tpm = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/germ_cell_lcm_tpm_counts_filtered_samples_20210812.txt', header = T, sep = '\t', stringsAsFactors = F)
log2tpm <- log2(as.matrix(tpm) + 1)
log2tpm = log2tpm[(row.names(log2tpm) %in% prot_cod) & !(row.names(log2tpm) %in% unique(c(hgGenes, hkGenes, igGenes, cc.genes.updated.2019$s.genes, cc.genes.updated.2019$g2m.genes))),] #remove genes whose expression will be related to normal contamination or cellular functions not related to phenotype
metadata = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R0/00_SUPPLEMENTARY_DATA/mRNA_summary_metadata.txt', sep = '\t', header = T, stringsAsFactors = F, quote = '')
row.names(metadata) = metadata$Sample
inv_mixed_rna = metadata[metadata$Diagnosis == 'Post-pubertal mixed germ cell tumour' & metadata$Normal.Tumour == 'Tumour' & metadata$Updated.Description != 'GCNIS',]
inv_mixed_rna = inv_mixed_rna[inv_mixed_rna$Case.ID %in% c('PD43296', 'PD43298', 'PD45543', 'PD45544', 'PD46269'),] #make direct contrast in each tumour to their transcriptome

#between histologies
interhisto_t_list = list()
for(i in unique(inv_mixed_rna$Case.ID)){ #per patient
  assessed_histo = c() #record of histologies already analysed
  t_prox_score_vec = c() #vector per patient comparison scores
  for(j in unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i,]$Updated.Description)){
    assessed_histo = c(assessed_histo, j) #add histology to list to not compare against
    if(length(assessed_histo) != length(unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i,]$Updated.Description))){ #stop when you are at the last histology as you have already compared it to all of the others
       for(k in unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i & inv_mixed_rna$Updated.Description == j,]$Sample)){ #samples per histology per tumour
      choice_sample = log2tpm[, k] #vector from sample in histology of interest
      for(l in unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i & !(inv_mixed_rna$Updated.Description %in% assessed_histo),]$Sample)){
        comparator_sample = log2tpm[, l] #vector from a sample in another histology
        t_prox = cor(choice_sample, comparator_sample, method = 'pearson')
        t_prox_score_vec = c(t_prox_score_vec, t_prox)
    }
    }
    }
   
  }
  interhisto_t_list = c(interhisto_t_list, list(t_prox_score_vec))
}

names(interhisto_t_list) = unique(inv_mixed_rna$Case.ID)

sapply(interhisto_t_list, median)

#within histologies
intrahisto_t_list = list()
for(i in unique(inv_mixed_rna$Case.ID)){ #per patient
  t_prox_score_vec = c() #vector per patient comparison scores
  for(j in unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i,]$Updated.Description)){ #per histology
      if(nrow(inv_mixed_rna[inv_mixed_rna$Case.ID == i & inv_mixed_rna$Updated.Description == j,]) > 1){ #if histology has more than one microbiopsy
      assessed_samples = c() #record of samples already analysed
      for(k in unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i & inv_mixed_rna$Updated.Description == j,]$Sample)){ #per sample
        assessed_samples = c(assessed_samples, k) #add to vector analysed
        choice_sample = log2tpm[, k]
        
        if(k != unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i & inv_mixed_rna$Updated.Description == j,]$Sample)[length(unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i & inv_mixed_rna$Updated.Description == j,]$Sample))]){ #so the last sample isn't compared against itself
          for(l in unique(inv_mixed_rna[inv_mixed_rna$Case.ID == i & inv_mixed_rna$Updated.Description == j & !(inv_mixed_rna$Sample %in% assessed_samples),]$Sample)){
          comparator_sample = log2tpm[, l]
          t_prox = cor(choice_sample, comparator_sample, method = 'pearson')
          t_prox_score_vec = c(t_prox_score_vec, t_prox)
        }
        }
      }
    }
  }
  intrahisto_t_list = c(intrahisto_t_list, list(t_prox_score_vec))
}

names(intrahisto_t_list) = unique(inv_mixed_rna$Case.ID)
sapply(intrahisto_t_list, median)

#control - against normal testis
norm_control = metadata[metadata$Normal.Tumour == 'Normal',]

norm_tumour_t_list = list()
for(i in unique(inv_mixed_rna$Case.ID)){ #per patient
  t_prox_score_vec = c() #vector per patient comparison scores
  for(j in inv_mixed_rna[inv_mixed_rna$Case.ID == i,]$Sample){
    choice_sample = log2tpm[, j]
    for(k in norm_control$Sample){
      comparator_sample = log2tpm[, k]
      t_prox = cor(choice_sample, comparator_sample, method = 'pearson')
      #t_prox = cosine(choice_sample, comparator_sample)[1,1]
      #t_prox = cor(choice_sample, comparator_sample, method = 'spearman')
      t_prox_score_vec = c(t_prox_score_vec, t_prox)
    }
  }
  norm_tumour_t_list = c(norm_tumour_t_list, list(t_prox_score_vec))
}

names(norm_tumour_t_list) = unique(inv_mixed_rna$Case.ID)

sapply(norm_tumour_t_list, median)

#add together
interhisto_t_df = data.frame(unlist(interhisto_t_list))
interhisto_t_df$patient = substr(row.names(interhisto_t_df), 0, 7)
interhisto_t_df$comparison = 'Inter-histology'
row.names(interhisto_t_df) = NULL
colnames(interhisto_t_df) = c('t_prox', 'patient', 'comparison')

intrahisto_t_df = data.frame(unlist(intrahisto_t_list))
intrahisto_t_df$patient = substr(row.names(intrahisto_t_df), 0, 7)
intrahisto_t_df$comparison = 'Intra-histology'
row.names(intrahisto_t_df) = NULL
colnames(intrahisto_t_df) = c('t_prox', 'patient', 'comparison')

norm_tumour_t_df = data.frame(unlist(norm_tumour_t_list))
norm_tumour_t_df$patient = substr(row.names(norm_tumour_t_df), 0, 7)
norm_tumour_t_df$comparison = 'Normal v Tumour'
row.names(norm_tumour_t_df) = NULL
colnames(norm_tumour_t_df) = c('t_prox', 'patient', 'comparison')

all_histo_t_df = rbind(interhisto_t_df, intrahisto_t_df, norm_tumour_t_df)

ggplot(all_histo_t_df) + 
  geom_quasirandom(mapping = aes(x = comparison, y = t_prox, col = comparison), width = 0.3) +
  geom_pointrange(mapping = aes(x = comparison, y = t_prox), stat = "summary", 
                  shape=19, 
                  fun.min = function(z) {quantile(z,0.25)},
                  fun.max = function(z) {quantile(z,0.75)},
                  fun = median, fill="black") +
  facet_grid(. ~ patient, scales = 'free_x', space = 'free_x', switch = 'x') + theme_pubr() + theme(axis.text.x = element_blank(), panel.spacing = unit(0, 'mm'), strip.background = element_blank(), strip.placement = "outside", strip.text = element_text(size = 11)) + ylab('Transcriptomic similarity') + xlab ('') + ylim(0, 1)

#are these differences significant? initially using Wilcoxon rank sum test
wilcox.test(t_prox ~ comparison, all_histo_t_df[all_histo_t_df$patient == 'PD43296' & all_histo_t_df$comparison %in% c('Inter-histology', 'Intra-histology'),]) #W = 154448, p-value < 2.2e-16
wilcox.test(t_prox ~ comparison, all_histo_t_df[all_histo_t_df$patient == 'PD43298' & all_histo_t_df$comparison %in% c('Inter-histology', 'Intra-histology'),]) #W = 68099, p-value < 2.2e-16
wilcox.test(t_prox ~ comparison, all_histo_t_df[all_histo_t_df$patient == 'PD45543' & all_histo_t_df$comparison %in% c('Inter-histology', 'Intra-histology'),]) #W = 10, p-value = 0.04009
wilcox.test(t_prox ~ comparison, all_histo_t_df[all_histo_t_df$patient == 'PD45544' & all_histo_t_df$comparison %in% c('Inter-histology', 'Intra-histology'),]) #W = 3784, p-value = 7.244e-16
wilcox.test(t_prox ~ comparison, all_histo_t_df[all_histo_t_df$patient == 'PD46269' & all_histo_t_df$comparison %in% c('Inter-histology', 'Intra-histology'),]) #W = 19852, p-value = 8.553e-11

#retry with permutation test for reviewer
permutation_list = list()
set.seed(42)

#check how many permutations for label switching can be achieved per tumour
patient = c("PD43296", "PD43298", "PD45543", "PD45544", "PD46269") 

for(i in patient){
  df = all_histo_t_df[all_histo_t_df$patient == i & all_histo_t_df$comparison != "Normal v Tumour",] #label swapping between intra and inter only
  max_perm = choose(nrow(df), nrow(df[df$comparison == 'Inter-histology',]))
  print(paste0(i, ' ', max_perm))
}

#well over 1000 in all cases
#all tumours
patient = c("PD45543", "PD43296", "PD43298", "PD45544", "PD46269")
nRand = 1000

for(i in patient){
  median_diff_vec = c()
  df = all_histo_t_df[all_histo_t_df$patient == i & all_histo_t_df$comparison != "Normal v Tumour",] #label swapping between intra and inter only
  for(j in 1:nRand){
    df$label.switch = sample(df$comparison, length(df$comparison), replace = F)
    median_diff_vec = c(median_diff_vec, abs(median(df[df$label.switch == 'Inter-histology',]$t_prox) - median(df[df$label.switch == 'Intra-histology',]$t_prox)))
  }
  permutation_list = c(permutation_list, list(median_diff_vec))
}

names(permutation_list) = patient

#what are the difference between the inter and intra histology values that we observe?
obs_t_data = aggregate(t_prox ~ patient + comparison, all_histo_t_df, median)
patient = c("PD45543", "PD43296", "PD43298", "PD45544", "PD46269")

obs_t_med_diff = data.frame()
for(i in patient){
  obs_t_med_diff = rbind(obs_t_med_diff, c(i, abs(obs_t_data[obs_t_data$patient == i & obs_t_data$comparison == "Intra-histology",]$t_prox - obs_t_data[obs_t_data$patient == i & obs_t_data$comparison == "Inter-histology",]$t_prox)))
}

#obs_t_med_diff = data.frame(obs_t_med_diff)
names(obs_t_med_diff) = c('patient', 'observed_median_diff')
obs_t_med_diff$observed_median_diff = as.numeric(obs_t_med_diff$observed_median_diff)

#what is the probability this was observed by chance?
obs_t_med_diff$pval = NA
for(i in patient){
  obs_t_med_diff[obs_t_med_diff$patient == i,]$pval = sum(permutation_list[[i]] >= obs_t_med_diff[obs_t_med_diff$patient == i,]$observed_median_diff) / length(permutation_list[[i]]) #what proportion of label switched combinations generates a larger difference between the inter and intra-histology transcriptomic proximity?
}

# patient observed_median_diff  pval
# PD45543           0.08906052 0.015
# PD43296           0.05536096 0.000
# PD43298           0.12263423 0.000
# PD45544           0.17752213 0.000
# PD46269           0.12499585 0.000

```

###Plot

```{r}

library(cowplot)

table(all_histo_g_df$patient, all_histo_g_df$comparison) #=<10 for PD45543 intra and inter and PD46269 intra
table(all_histo_t_df$patient, all_histo_t_df$comparison) #=<10 for PD45543 intra and inter

all_histo_g_df$comparison = factor(all_histo_g_df$comparison, levels = c('Intra-histology', 'Inter-histology'))
all_histo_t_df$comparison = factor(all_histo_t_df$comparison, levels = c('Intra-histology', 'Inter-histology', 'Normal v Tumour'))

p1 = ggplot(all_histo_g_df) + 
  geom_quasirandom(dodge.width=.8, cex=1, aes(x = gen_prox, y = comparison), shape = 21, fill = 'transparent', col = 'transparent', groupOnX = F) + 
  geom_quasirandom(all_histo_g_df[(all_histo_g_df$patient == 'PD45543') | (all_histo_g_df$patient == 'PD46269' & all_histo_g_df$comparison == 'Intra-histology'),], mapping = aes(x = gen_prox, y = comparison), width = 0.3, groupOnX = F, col = "black") +
  geom_pointrange(mapping = aes(x = gen_prox, y = comparison, fill = comparison),
                  stat = "summary", shape=21, 
                  fun.min = function(z) {quantile(z,0.25)},
                  fun.max = function(z) {quantile(z,0.75)},
                  fun = median, size = 1.3) +
  facet_grid(patient ~ ., scales = 'free_y', space = 'free_y') + theme_pubr() + theme(axis.text.y = element_blank(), panel.spacing = unit(0, 'mm'), strip.background = element_blank(), strip.placement = "none", strip.text = element_text(size = 11), panel.background = element_blank(), legend.position = "none") + xlab('Genetic similarity') + ylab ('') + scale_y_discrete(position = "right")  + scale_x_continuous(expand = c(0,0.01), limits = c(0,1)) + scale_fill_manual(values = c('grey25', 'grey75'), name = 'Comparison')

p2 = ggplot(all_histo_t_df) + 
  geom_quasirandom(cex=1, aes(x = t_prox, y = comparison), shape = 21, fill = 'transparent', col = 'transparent', groupOnX = F) + 
  geom_quasirandom(all_histo_t_df[(all_histo_t_df$patient == 'PD45543' & all_histo_t_df$comparison %in% c('Intra-histology', 'Inter-histology')),], mapping = aes(x = t_prox, y = comparison), width = 0.4, groupOnX = F, col = "black") +
  geom_pointrange(mapping = aes(x = t_prox, y = comparison, fill = comparison),
                  stat = "summary", shape=21, 
                  fun.min = function(z) {quantile(z,0.25)},
                  fun.max = function(z) {quantile(z,0.75)},
                  fun = median, size = 1.3) +
  facet_grid(patient ~ ., scales = 'free_y', space = 'free_y', switch = 'y') + theme_pubr() + theme(axis.text.y = element_blank(), panel.spacing = unit(0, 'mm'), strip.background = element_blank(), strip.placement = "outside", strip.text = element_text(size = 11), panel.background = element_blank(), legend.position = "none") + xlab('Transcriptomic similarity') + ylab ('') + scale_y_discrete(position = "left")  + scale_x_continuous(expand = c(0,0.01), limits = c(0,1)) + scale_fill_manual(values = c('grey25', 'grey75', 'white'), name = 'Comparison')

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nat_comms_2022_R1/06_FIGURES/gct_genetic_proximity_pearson_rna_20220525.pdf', height = 4, width = 4, useDingbats = F)
plot_grid(p1, p2, align = 'h', ncol = 2)
dev.off()

```

#06_figures

##Fig1 - Sample overview & genomic features

###Fig1b - Burden, ploidy and purity
```{r}

library(ggpubr)
library(cowplot)

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', header = T, sep = '\t', stringsAsFactors = F)
#summary_data = manifest[manifest$Experimental_arm == 'LCM',]
summary_data = manifest[,c('Sample', 'Case.ID', 'Organ', 'Diagnosis', 'Histology', 'Age', 'Purity', 'Ploidy', 'SNV.burden', 'Indel.burden')]

summary_data$is.gcnis = ifelse(summary_data$Histology == 'GCNIS', 'yes', 'no')
summary_data$is.gcnis = factor(summary_data$is.gcnis, levels = c('yes', 'no'))
summary_data = summary_data[order(summary_data$Age, summary_data$Case.ID, summary_data$is.gcnis, summary_data$SNV.burden),]
summary_data$Case.ID = factor(summary_data$Case.ID, levels = unique(summary_data$Case.ID))
summary_data$Sample = factor(summary_data$Sample, levels = as.vector(summary_data$Sample))

p1 = ggplot(summary_data, mapping = aes(x = Sample, y = SNV.burden, fill = is.gcnis)) + 
  geom_bar(stat = 'identity') + 
  theme_pubr() + 
  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_rect(fill = 'grey97')) + 
  scale_y_continuous(expand = c(0,0), limits = c(0, 4000)) + 
  ylab('SNV burden') + 
  xlab('') + 
  scale_x_discrete(expand = c(0,0)) +
  facet_grid(. ~ Case.ID, scales = 'free_x') + 
  scale_fill_manual(breaks = c("yes", "no"), values=c("black", "#999999"))
p2 = ggplot(summary_data, mapping = aes(x = Sample, y = Indel.burden, fill = is.gcnis)) + 
  geom_bar(stat = 'identity') + 
  theme_pubr() + 
  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_rect(fill = 'grey97')) + 
  scale_y_continuous(expand = c(0,0), limits = c(0, 200)) + 
  ylab('Indel burden') + 
  xlab('') + 
  scale_x_discrete(expand = c(0,0)) +
  facet_grid(. ~ Case.ID, scales = 'free_x') + 
  scale_fill_manual(breaks = c("yes", "no"), values=c("black", "#999999"))
#p3 = ggplot(summary_data, mapping = aes(x = Sample, y = Purity * 100, fill = is.gcnis)) + 
#  geom_bar(stat = 'identity') + 
#  theme_pubr() + 
#  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_blank()) + 
#  scale_y_continuous(expand = c(0,0), limits = c(0, 100)) + 
#  ylab('Purity (%)') + 
#  xlab('') + 
#  scale_x_discrete(expand = c(0,0)) +
#  facet_grid(. ~ Case.ID, scales = 'free_x') + 
#  scale_fill_manual(breaks = c("yes", "no"), values=c("black", "#999999"))
  
p3 = ggplot(summary_data, mapping = aes(x = Sample, y = Purity * 100)) + 
  geom_crossbar(mapping = aes(ymin = Purity * 100, ymax = Purity * 100, col = is.gcnis)) + 
  theme_pubr() + 
  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_rect(fill = 'grey97')) + 
  scale_y_continuous(expand = c(0,0), limits = c(0, 100)) + 
  ylab('Purity (%)') + 
  xlab('') + 
  scale_x_discrete(expand = c(0,0)) +
  facet_grid(. ~ Case.ID, scales = 'free_x') + 
  scale_color_manual(breaks = c("yes", "no"), values=c("black", "#999999"))

#p4 = ggplot(summary_data, mapping = aes(x = Sample, y = Ploidy, fill = is.gcnis)) + 
#  geom_bar(stat = 'identity') + 
#  theme_pubr() +
#  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_blank()) + 
#  ylab('Ploidy') + 
#  xlab('') + 
#  scale_x_discrete(expand = c(0,0)) +
#  facet_grid(. ~ Case.ID, scales = 'free_x') + 
#  scale_fill_manual(breaks = c("yes", "no"), values=c("black", "#999999"))
p4 = ggplot(summary_data, mapping = aes(x = Sample, y = Ploidy)) + 
 geom_crossbar(mapping = aes(ymin = Purity * 100, ymax = Purity * 100, col = is.gcnis)) + 
  theme_pubr() +
  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_rect(fill = 'grey97')) + 
  ylab('Ploidy') + 
  xlab('') + 
  scale_x_discrete(expand = c(0,0)) +
  facet_grid(. ~ Case.ID, scales = 'free_x') + 
  scale_color_manual(breaks = c("yes", "no"), values=c("black", "#999999")) +
  scale_y_continuous(expand = c(0.05,0.05), limits = c(0, 4.2)) 




p4 = p4 + coord_cartesian(ylim=c(1,4.2))


pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/20210823_lcm_genome_overview.pdf', height = 8, width = 11, useDingbats = F)
plot_grid(p1, p2, p3, p4, ncol = 1, align = 'v', rel_heights = c(1, 1, 0.5, 0.5))
dev.off()

p3 = ggplot(summary_data, mapping = aes(x = Sample, y = Purity * 100, fill = is.gcnis)) + 
  geom_point(shape = '|', size = 3) + 
  theme_pubr() + 
  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_blank()) + 
  scale_y_continuous(expand = c(0.05,0.05), limits = c(0, 100)) + 
  ylab('Purity (%)') + 
  xlab('') + 
  #scale_x_discrete(expand = c(0,0)) +
  facet_grid(. ~ Case.ID, scales = 'free_x') + 
  scale_fill_manual(breaks = c("yes", "no"), values=c("black", "#999999"))

p4 = ggplot(summary_data, mapping = aes(x = Sample, y = Ploidy, fill = is.gcnis)) + 
 geom_point(shape = "|", size = 3) + 
  theme_pubr() +
  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_blank()) + 
  ylab('Ploidy') + 
  xlab('') + 
  #scale_x_discrete(expand = c(0.5,0.5)) +
  facet_grid(. ~ Case.ID, scales = 'free_x') + 
  scale_fill_manual(breaks = c("yes", "no"), values=c("black", "#999999")) +
  scale_y_continuous(expand = c(0.05,0.05), limits = c(0, 4.2)) 

```

###Fig1c - SBS signatures

```{r}

library(reshape2)
library(ggplot2)
library(ggpubr)
library(cowplot)

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/sigprofiler/hdp_sig_prior/per_sample')

manifest = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/supplementary_data/WGS_summary_metadata.txt', header = T, sep = '\t', stringsAsFactors = F)
#summary_data = manifest[manifest$Experimental_arm == 'LCM',]
summary_data = manifest[,c('Sample', 'Case.ID', 'Organ', 'Diagnosis', 'Histology', 'Age', 'Purity', 'Ploidy', 'SNV.burden', 'Indel.burden')]

summary_data$is.gcnis = ifelse(summary_data$Histology == 'GCNIS', 'yes', 'no')
summary_data$is.gcnis = factor(summary_data$is.gcnis, levels = c('yes', 'no'))
summary_data = summary_data[order(summary_data$Age, summary_data$Case.ID, summary_data$is.gcnis, summary_data$SNV.burden),]
summary_data$Case.ID = factor(summary_data$Case.ID, levels = unique(summary_data$Case.ID))
summary_data$Sample = factor(summary_data$Sample, levels = as.vector(summary_data$Sample))

hdp_sigprofiler_mapped = read.table('Decompose_Solution_Activities.txt', header = T, sep = '\t', stringsAsFactors = F)
#hdp_sigprofiler_mapped$Samples = unlist(strsplit(hdp_sigprofiler_mapped$Samples, '_post_swater_final_snvs')) the sample names needed editing from the raw sigprofiler output
#write.table(hdp_sigprofiler_mapped, 'Decompose_Solution_Activities.txt', col.names = T, row.names = F, sep = '\t', quote = F)

hdp_sigprofiler_mapped$SBS5_40 = hdp_sigprofiler_mapped$SBS5 + hdp_sigprofiler_mapped$SBS40 
hdp_sigprofiler_mapped = hdp_sigprofiler_mapped[, c("Samples", "SBS1", "SBS5_40", "SBS17a", "SBS17b", "SBS18", "SBS35", "SBS91", "N2")]

hdp_sigprofiler_mapped_norm = hdp_sigprofiler_mapped[, c(2:9)] / rowSums(hdp_sigprofiler_mapped[, c(2:9)])
hdp_sigprofiler_mapped_norm$sample = hdp_sigprofiler_mapped$Samples
hdp_sigprofiler_mapped_norm$tumour = substr(hdp_sigprofiler_mapped_norm$sample, 0, 7)

hdp_sigprofiler_mapped_norm.m = reshape2::melt(hdp_sigprofiler_mapped_norm, id.vars = c('sample', 'tumour'))

hdp_sigprofiler_mapped_norm.m$age = NA
hdp_sigprofiler_mapped_norm.m$diagnosis = NA
hdp_sigprofiler_mapped_norm.m$site = NA
hdp_sigprofiler_mapped_norm.m$histology = NA

for(i in 1:nrow(hdp_sigprofiler_mapped_norm.m)){
  hdp_sigprofiler_mapped_norm.m$age[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Age
  hdp_sigprofiler_mapped_norm.m$diagnosis[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Diagnosis
  hdp_sigprofiler_mapped_norm.m$site[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Organ
  hdp_sigprofiler_mapped_norm.m$histology[i] = manifest[manifest$Sample == hdp_sigprofiler_mapped_norm.m$sample[i],]$Histology
}

hdp_sigprofiler_mapped_norm.m$variable = factor(hdp_sigprofiler_mapped_norm.m$variable, levels = rev(unique(hdp_sigprofiler_mapped_norm.m$variable)))
hdp_sigprofiler_mapped_norm.m$tumour = factor(hdp_sigprofiler_mapped_norm.m$tumour, levels = levels(summary_data$Case.ID))
hdp_sigprofiler_mapped_norm.m$sample = factor(hdp_sigprofiler_mapped_norm.m$sample, levels = levels(summary_data$Sample))

cols = get_palette(palette = 'Set1', k = length(unique(hdp_sigprofiler_mapped_norm.m$variable)))
cols = c("#F781BF", "#A65628", "#4DAF4A", "#984EA3", "#FF7F00", "#FFFF33", "dodgerblue", "#E52621")

ggplot(summary_data, mapping = aes(x = Sample, y = Ploidy, fill = is.gcnis)) + 
  geom_bar(stat = 'identity') + 
  theme_pubr() +
  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_blank()) + 
  scale_y_continuous(expand = c(0,0), limits = c(0, 4.2)) + 
  ylab('Ploidy') + 
  xlab('') + 
  scale_x_discrete(expand = c(0,0)) +
  facet_grid(. ~ Case.ID, scales = 'free_x') + 
  scale_fill_manual(breaks = c("yes", "no"), values=c("dodgerblue", "grey40"))


h1 = ggplot(hdp_sigprofiler_mapped_norm.m) + 
  geom_bar(mapping = aes(x = sample, y = value, fill = variable), stat = 'identity') + 
  theme_pubr() + coord_cartesian(expand = F) + 
  scale_fill_manual(values=cols) + 
  ylab('Proportion of substitutions') + 
  xlab('') + 
  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_blank()) + facet_grid(. ~ tumour, scales = 'free_x')
h2 = ggplot(hdp_sigprofiler_mapped_norm.m) + 
  geom_bar(mapping = aes(x = sample, y = 1, fill = histology),  stat = "identity", position = "dodge") + 
  theme_pubr() + coord_cartesian(expand = F) + 
  scale_fill_manual(values = get_palette('Dark2', length(unique(hdp_sigprofiler_mapped_norm.m$histology)))) + 
  ylab('') + 
  xlab('') + 
  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_blank()) + facet_grid(. ~ tumour, scales = 'free_x')
#h3 = ggplot(hdp_sigprofiler_mapped_norm.m) + 
#  geom_bar(mapping = aes(x = sample, y = 1, fill = diagnosis),  stat = "identity", position = "dodge") + 
#  theme_pubr() + coord_cartesian(expand = F) + 
#  scale_fill_manual(values = c('grey80', 'grey50', 'grey20')) + 
#  ylab('') + 
#  xlab('') + 
#  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_blank()) + facet_grid(. ~ tumour, scales = 'free_x')
#h4 = ggplot(hdp_sigprofiler_mapped_norm.m) + 
#  geom_bar(mapping = aes(x = sample, y = 1, fill = site),  stat = "identity", position = "dodge") + 
#  theme_pubr() + coord_cartesian(expand = F) + 
#  scale_fill_manual(values = get_palette('default', length(unique(hdp_sigprofiler_mapped_norm.m$site)))) + 
#  ylab('') + 
#  xlab('') + 
#  theme(panel.border = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin = unit(c(0.1,0.1,0,0.1), "cm"), strip.text.x = element_blank(), panel.spacing = unit(1, "mm"), legend.position = 'none', strip.background = element_blank(), panel.background = element_blank()) + facet_grid(. ~ tumour, scales = 'free_x')

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/hdp_extracted_sigprofiler_mapped_sigs20210823.pdf', height = 4,  width = 10, useDingbats = F)
plot_grid(h1, h2, align = 'v', ncol = 1, rel_heights = c(1, 0.2))
dev.off()

```

###Fig1d - N2 trinucleotide context

```{r}

library(dplyr)
library(ggplot2)
library(ggpubr)

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/signature_extraction/tsb')

tsb = read.table('sigTSBdf.tsv', header = T, sep = '\t', stringsAsFactors = F)

n2_tsb = tsb[tsb$sig == 'SBS_N2',]
n2_tsb$prop = n2_tsb$count / sum(n2_tsb$count)

n2_tsb$sub = factor(n2_tsb$sub, levels = unique(n2_tsb$sub))
n2_tsb$trinuc = factor(n2_tsb$trinuc, levels = unique(n2_tsb$trinuc))
n2_tsb$txStrand = factor(n2_tsb$txStrand, levels = rev(unique(n2_tsb$txStrand)))

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/tsb_sbs_n2_20210813.pdf', height = 2.5,  width = 7.5, useDingbats = F)
ggplot(n2_tsb) + 
  geom_bar(mapping = aes(x = trinuc, y = prop, fill = sub, alpha = txStrand), position="dodge", stat = 'identity', col = NA, size = 0.3) + 
  facet_grid(. ~ sub, scales = 'free_x') + scale_alpha_discrete(range = c(1, 0.4)) + 
  scale_fill_manual(values = c("dodgerblue","black","red","grey70","olivedrab3","plum2")) +
  ylab("Proportion of substitutions assigned") + 
  xlab("Trinucleotide context") +
  labs(alpha = "Strand") + 
  guides(fill = FALSE) + 
  theme_pubr() +
  theme(strip.background = element_blank(), strip.text = element_text(face = 'bold'), axis.text.x = element_text(size = 5, angle = 90, vjust = 0.5), panel.spacing = unit(1, 'mm'), panel.border = element_rect(size = 0.3, fill = NA), legend.position = 'right') +
  coord_cartesian(ylim = c(0, 0.08), expand = F)
dev.off()



```

###Fig1e/f - rearrangements (including retrotransposition activity) by age *TO DO

##Fig2

###Fig2f - plot copy number for PD43299 chr17

####Normalised coverage

```{r}

library(ggplot2)
library(ggpubr)

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/offpipe_run/original_file_bb')

lo0005 = read.table('PD43299a_lo0005/PD43299a_lo0005_allelecounts/PD43299a_lo0005_alleleFrequencies_chr17.txt', header = T, sep = '\t', stringsAsFactors = F, comment.char = '')
lo0011 = read.table('PD43299a_lo0011/PD43299a_lo0011_allelecounts/PD43299a_lo0011_alleleFrequencies_chr17.txt', header = T, sep = '\t', stringsAsFactors = F, comment.char = '')
b = read.table('PD43299a_lo0011/PD43299b_allelecounts/PD43299b_alleleFrequencies_chr17.txt', header = T, sep = '\t', stringsAsFactors = F, comment.char = '')

#normalise coverage by matched normal
lo0005$norm_good_depth = lo0005$Good_depth / b$Good_depth
lo0011$norm_good_depth = lo0011$Good_depth / b$Good_depth

#log the normalisation
lo0005$log_norm_good_depth = log(lo0005$norm_good_depth)
lo0011$log_norm_good_depth = log(lo0011$norm_good_depth)

#combine dataframes
lo0005$sample = 'lo0005'
lo0011$sample = 'lo0011'

combined = rbind(lo0005, lo0011)

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/PD43299a_chr17_overview.pdf', height = 3, width = 5, useDingbats = F)
ggplot(combined) + geom_point(mapping = aes(x = POS, y = log_norm_good_depth), pch = '.', alpha = 0.1) + coord_cartesian(ylim = c(-1.5, 1.5), expand = F, xlim = c(0, max(combined$POS))) + facet_grid(sample ~ ., space = 'free', scales = 'free') + theme_pubr() + theme(panel.spacing = unit(5, 'mm'), panel.border = element_rect(size = 0.3, fill = NA), strip.background = element_blank()) + ylab('Log normalised coverage') + xlab('Chromosome 17 position (Mb)')
dev.off()

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/PD43299a_chr17_overview_empty_plot.pdf', height = 3, width = 5, useDingbats = F)
ggplot(combined, mapping = aes(x = POS, y = log_norm_good_depth), pch = '.', alpha = 0.1) + geom_blank() + coord_cartesian(ylim = c(-1.5, 1.5), expand = F, xlim = c(0, max(combined$POS))) + facet_grid(sample ~ ., space = 'free', scales = 'free') + theme_pubr() + theme(panel.spacing = unit(5, 'mm'), panel.border = element_rect(size = 0.3, fill = NA), strip.background = element_blank(), ) + ylab('Log normalised coverage') + xlab('Chromosome 17 position (Mb)') + scale_x_continuous(breaks = c(0,20000000,40000000,60000000,80000000))
dev.off()

ggplot(combined[combined$sample == 'lo0005',]) + geom_point(mapping = aes(x = POS, y = log_norm_good_depth), pch = '.', alpha = 0.1) + coord_cartesian(ylim = c(-1.5, 1.5), expand = F, xlim = c(0, max(combined$POS))) + theme_pubr() + theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.line = element_blank(), axis.title = element_blank()) + ggsave('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/PD43299a_lo0005_chr17_overview.png', height = 3, width = 5, units = 'in')

ggplot(combined[combined$sample == 'lo0011',]) + geom_point(mapping = aes(x = POS, y = log_norm_good_depth), pch = '.', alpha = 0.1) + coord_cartesian(ylim = c(-1.5, 1.5), expand = F, xlim = c(0, max(combined$POS))) + theme_pubr() + theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.line = element_blank(), axis.title = element_blank()) + ggsave('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/PD43299a_lo0011_chr17_overview.png', height = 3, width = 5, units = 'in')


```

####Battenberg plot

```{r}

library(ggplot2)
library(ggpubr)

#0, 81194907 chr 17 coord

setwd('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/variant_calls/battenberg/00_final_calls')

lo0005_bb = read.table('PD43299a_lo0005_subclones.txt', header = T, sep = '\t', stringsAsFactors = F)
lo0011_bb = read.table('PD43299a_lo0011_subclones.txt', header = T, sep = '\t', stringsAsFactors = F)

lo0005_bb_chr17 = lo0005_bb[lo0005_bb$chr == 17,]
lo0011_bb_chr17 = lo0011_bb[lo0011_bb$chr == 17,]

ggplot(lo0005_bb_chr17) + geom_segment(mapping = aes(x = startpos, xend = endpos, y = ntot, yend = ntot)) + theme_pubr() + coord_cartesian(ylim = c(3, 5.5), expand = F, xlim = c(0, 81194907)) + scale_y_continuous(sec.axis = dup_axis()) + theme(panel.border = element_rect(fill = NA, color = 'black'), axis.title.y.right = element_blank(), axis.text.y.right = element_blank()) + xlab('') + ylab('') + geom_vline(xintercept = c(7565097, 7590856, 29421945, 29709134), col = 'red')


pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/PD43299a_lo0005_chr17_bb_overview_plot.pdf', height = 3, width = 5, useDingbats = F)
ggplot(lo0005_bb_chr17) + geom_segment(mapping = aes(x = startpos, xend = endpos, y = ntot, yend = ntot)) + theme_pubr() + coord_cartesian(ylim = c(3, 5.5), expand = F, xlim = c(0, 81194907)) + scale_y_continuous(sec.axis = dup_axis()) + theme(panel.border = element_rect(fill = NA, color = 'black'), axis.title.y.right = element_blank(), axis.text.y.right = element_blank()) + xlab('') + ylab('')
dev.off()

pdf('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/figures/PD43299a_lo0011_chr17_bb_overview_plot.pdf', height = 3, width = 5, useDingbats = F)
ggplot(lo0011_bb_chr17) + geom_segment(mapping = aes(x = startpos, xend = endpos, y = ntot, yend = ntot)) + theme_pubr() + coord_cartesian(ylim = c(3, 5.5), expand = F, xlim = c(0, 81194907)) + scale_y_continuous(sec.axis = dup_axis()) + theme(panel.border = element_rect(fill = NA, color = 'black'), axis.title.y.right = element_blank(), axis.text.y.right = element_blank()) + xlab('') + ylab('')
dev.off()

```




