---
  title: "Paediatric autopsy project - WGS data"
author: "Thomas R W Oliver"
output: 
  html_document:
  toc: true
toc_depth: 2
number_sections: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Review of code/clean up complete - TO3 22/12/22

#00_supplementary_data

```{r}

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

```

#03_DNA_analyses

##Check coverage and sample overview

```{r}

manifest_wgs = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
manifest_wes = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wes_data/00_supplementary_data/wes_paeds_pm_sample_metadata_20220722.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

manifest_wgs_flt = manifest_wgs[manifest_wgs$keep == "Y", ]

table(manifest_wgs_flt$experimental_arm)
#WGS_bulk  WGS_LCM 
#     71      839
nrow(manifest_wes) #180

median(manifest_wgs_flt[manifest_wgs_flt$experimental_arm == "WGS_bulk",]$picard_median_coverage) #121
summary(manifest_wgs_flt[manifest_wgs_flt$experimental_arm == "WGS_bulk",]$picard_median_coverage)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#  59.0    80.5   121.0   121.1   156.0   193.0
median(manifest_wgs_flt[manifest_wgs_flt$experimental_arm == "WGS_LCM",]$picard_median_coverage) #28
summary(manifest_wgs_flt[manifest_wgs_flt$experimental_arm == "WGS_LCM",]$picard_median_coverage)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#11.00   22.50   28.00   29.58   34.00  141.00 
median(manifest_wes$pct_target_30x) #0.843673

summary(manifest_wgs[manifest_wgs$keep == "Y" & manifest_wgs$purity_trunk < 0.01 & manifest_wgs$CNS == "Y",]$sub_burden_tier1 / manifest_wgs[manifest_wgs$keep == "Y" & manifest_wgs$purity_trunk < 0.01 & manifest_wgs$CNS == "Y",]$sub_burden) 
#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#0.4722  1.0000  1.0000  0.9751  1.0000  1.0000 
sum(manifest_wgs[manifest_wgs$keep == "Y" & manifest_wgs$purity_trunk < 0.01 & manifest_wgs$CNS == "Y",]$sub_burden_tier1 / manifest_wgs[manifest_wgs$keep == "Y" & manifest_wgs$purity_trunk < 0.01 & manifest_wgs$CNS == "Y",]$sub_burden < 0.6) #2
manifest_wgs[manifest_wgs$keep == "Y" & manifest_wgs$purity_trunk < 0.01 & manifest_wgs$CNS == "Y",][manifest_wgs[manifest_wgs$keep == "Y" & manifest_wgs$purity_trunk < 0.01 & manifest_wgs$CNS == "Y",]$sub_burden_tier1 / manifest_wgs[manifest_wgs$keep == "Y" & manifest_wgs$purity_trunk < 0.01 & manifest_wgs$CNS == "Y",]$sub_burden < 0.6,]$sample #"PD51122af_lo0018" "PD51122af_lo0024"

all_subs = read.table("/lustre/scratch119/realdata/mdt1/team274/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/subs/07_indel/all_sample_subs_distribution_annotated_20221201.txt", header = T, sep = "\t", stringsAsFactors = F)

temp = aggregate(Variant_read_depth ~ mutID, all_subs[all_subs$mutID %in% unique(all_subs[all_subs$Variant_read_depth < 4,]$mutID),], max)

```

##Fig. 1b - number of samples per tissue compartment per case

```{r}

manifest_wgs = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
manifest_wgs_flt = manifest_wgs[manifest_wgs$keep == "Y",]
manifest_wes = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wes_data/00_supplementary_data/wes_paeds_pm_sample_metadata_20220722.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

manifest_flt = rbind(manifest_wgs_flt[, c("sample", "case.id", "bulk_sample", "bulk_tissue", "experimental_arm")], manifest_wes[, c("sample", "case.id", "bulk_sample", "bulk_tissue", "experimental_arm")])

sort(unique(manifest_flt$bulk_tissue))

manifest_flt$bulk_tissue_annot = NA

manifest_flt[manifest_flt$bulk_tissue %in% c("Frontal cortex", "Parietal cortex", "Temporal cortex", "Occipital cortex", "Hippocampus", "Subventricular zone", "Thalamus", "Hypothalamus", "Septum pellucidum", "Pineal gland", "Pituitary gland", "Optic nerve", "Pia-arachnoid meninges"),]$bulk_tissue_annot = "Supratentorial CNS"
manifest_flt[manifest_flt$bulk_tissue %in% c("Midbrain", "Pons", "Medulla"),]$bulk_tissue_annot = "Brainstem CNS"
manifest_flt[manifest_flt$bulk_tissue == "Cerebellum",]$bulk_tissue_annot = "Cerebellum CNS"
manifest_flt[manifest_flt$bulk_tissue == "Spinal cord",]$bulk_tissue_annot = "Spinal cord CNS"
manifest_flt[manifest_flt$bulk_tissue %in% c("Trigeminal nerve", "Cauda equina", "Dorsal root ganglion", "Sciatic nerve"),]$bulk_tissue_annot = "PNS"
manifest_flt[manifest_flt$bulk_tissue == "Skin",]$bulk_tissue_annot = "Skin"
manifest_flt[manifest_flt$bulk_tissue %in% c("Adrenal gland", "Blood", "Heart", "Skeletal muscle", "Soft tissue", "Spleen", "Kidney"),]$bulk_tissue_annot = "Mesodermal tissues"
manifest_flt[manifest_flt$bulk_tissue %in% c("Appendix", "Large intestine"),]$bulk_tissue_annot = "Large intestine"
manifest_flt[manifest_flt$bulk_tissue %in% c("Small intestine"),]$bulk_tissue_annot = "Small intestine"
manifest_flt[manifest_flt$bulk_tissue %in% c("Bladder", "Trachea", "Bronchus", "Thyroid gland", "Stomach", "Liver"),]$bulk_tissue_annot = "Other endodermal tissues"
manifest_flt[manifest_flt$bulk_sample == "PD51122ai",]$bulk_tissue_annot = "Antemortem diagnostic biopsy"

manifest_flt[is.na(manifest_flt$bulk_tissue_annot),]$bulk_tissue #none
table(manifest_flt$bulk_tissue_annot)

manifest_flt$bulk_tissue_annot = factor(manifest_flt$bulk_tissue_annot, levels = c("Supratentorial CNS", "Brainstem CNS", "Cerebellum CNS", "Spinal cord CNS", "PNS", "Skin", "Large intestine", "Small intestine", "Other endodermal tissues", "Mesodermal tissues", "Antemortem diagnostic biopsy"))

lcm_wgs_summary = table(manifest_flt[manifest_flt$experimental_arm == "WGS_LCM",]$bulk_tissue_annot, manifest_flt[manifest_flt$experimental_arm == "WGS_LCM",]$case.id)
colnames(lcm_wgs_summary) = paste0(colnames(lcm_wgs_summary), "_lcm_wgs")
lcm_wes_summary = table(manifest_flt[manifest_flt$experimental_arm == "WES_LCM",]$bulk_tissue_annot, manifest_flt[manifest_flt$experimental_arm == "WES_LCM",]$case.id)
colnames(lcm_wes_summary) = paste0(colnames(lcm_wes_summary), "_lcm_wes")
bulk_summary = table(manifest_flt[manifest_flt$experimental_arm == "WGS_bulk",]$bulk_tissue_annot, manifest_flt[manifest_flt$experimental_arm == "WGS_bulk",]$case.id)
colnames(bulk_summary) = paste0(colnames(bulk_summary), "_bulk")
all_summary = cbind(lcm_wgs_summary, lcm_wes_summary, bulk_summary)
all_summary = all_summary[, sort(colnames(all_summary))]

```

##Get list of all subs, annotated by distribution

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/subs/07_indel")

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

manifest_flt = manifest[manifest$keep == "Y",] #4 samples failed QC

all_subs = data.frame()
for(sample in manifest_flt$sample){
  data = read.table(paste0(sample, ".standard.gnomAD.AF.0.001.PON.read.direction.low_cov.bq.mq.binom.bbinom.5bp.indel.filtered.tier.annot.txt"), header = T, sep = '\t', stringsAsFactors = F)
    print(unique(data$Sample))
    all_subs = rbind(all_subs, data)
}

all_subs$bulk_sample = unlist(lapply(strsplit(all_subs$Sample, "_"), "[[", 1))
all_subs[all_subs$bulk_sample == "PD51123x3",]$bulk_sample = "PD51123x"
all_subs[all_subs$bulk_sample == "PD51123aa3",]$bulk_sample = "PD51123aa"
all_subs[all_subs$bulk_sample == "PD51123r3",]$bulk_sample = "PD51123r"
all_subs[all_subs$bulk_sample == "PD51122ae2",]$bulk_sample = "PD51122ae"
all_subs$mut_status = NA
all_subs$pure_tumour_mut = F

#single sample only called 
pd50297_private_muts = names(table(all_subs[all_subs$Case_ID == "PD50297",]$mutID))[table(all_subs[all_subs$Case_ID == "PD50297",]$mutID) == 1]
pd51122_private_muts = names(table(all_subs[all_subs$Case_ID == "PD51122",]$mutID))[table(all_subs[all_subs$Case_ID == "PD51122",]$mutID) == 1]
pd51123_private_muts = names(table(all_subs[all_subs$Case_ID == "PD51123",]$mutID))[table(all_subs[all_subs$Case_ID == "PD51123",]$mutID) == 1]

#bulk only called, anything taken from the original bulk sample
bulk_sample.list = list()
for(bulk in unique(all_subs$bulk_sample)){
  
  bulk_sample.list[[bulk]] = unique(all_subs[all_subs$bulk_sample == bulk,]$mutID)
  
}

bulk_sample.df = data.frame(bulk_sample = rep(names(bulk_sample.list), as.numeric(unlist(lapply(bulk_sample.list, length)))),
                            mutID = as.character(unlist(bulk_sample.list)))
bulk_sample.df$case.id = substr(bulk_sample.df$bulk_sample, 0, 7)

pd50297_single_bulk_muts = names(table(bulk_sample.df[bulk_sample.df$case.id == "PD50297",]$mutID))[table(bulk_sample.df[bulk_sample.df$case.id == "PD50297",]$mutID) == 1]
pd50297_multi_bulk_muts = names(table(bulk_sample.df[bulk_sample.df$case.id == "PD50297",]$mutID))[table(bulk_sample.df[bulk_sample.df$case.id == "PD50297",]$mutID) > 1]
pd51122_single_bulk_muts = names(table(bulk_sample.df[bulk_sample.df$case.id == "PD51122",]$mutID))[table(bulk_sample.df[bulk_sample.df$case.id == "PD51122",]$mutID) == 1]
pd51122_multi_bulk_muts = names(table(bulk_sample.df[bulk_sample.df$case.id == "PD51122",]$mutID))[table(bulk_sample.df[bulk_sample.df$case.id == "PD51122",]$mutID) > 1]
pd51123_single_bulk_muts = names(table(bulk_sample.df[bulk_sample.df$case.id == "PD51123",]$mutID))[table(bulk_sample.df[bulk_sample.df$case.id == "PD51123",]$mutID) == 1]
pd51123_multi_bulk_muts = names(table(bulk_sample.df[bulk_sample.df$case.id == "PD51123",]$mutID))[table(bulk_sample.df[bulk_sample.df$case.id == "PD51123",]$mutID) > 1]

#pure tumour variants
pd50297_tumour_samples = manifest[manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$case.id == "PD50297",]$sample
pd51122_tumour_samples = manifest[manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$case.id == "PD51122",]$sample
pd51123_tumour_samples = manifest[manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$case.id == "PD51123",]$sample

pd50297_tumour_muts = unique(all_subs[all_subs$Sample %in% pd50297_tumour_samples,]$mutID)
pd51122_tumour_muts = unique(all_subs[all_subs$Sample %in% pd51122_tumour_samples,]$mutID)
pd51123_tumour_muts = unique(all_subs[all_subs$Sample %in% pd51123_tumour_samples,]$mutID)

#add annotations
all_subs[all_subs$mutID %in% pd50297_private_muts & all_subs$Case_ID == "PD50297",]$mut_status = "sample"
all_subs[all_subs$mutID %in% pd51122_private_muts & all_subs$Case_ID == "PD51122",]$mut_status = "sample"
all_subs[all_subs$mutID %in% pd51123_private_muts & all_subs$Case_ID == "PD51123",]$mut_status = "sample"

all_subs[all_subs$mutID %in% pd50297_single_bulk_muts & all_subs$Case_ID == "PD50297" & !all_subs$mutID %in% pd50297_private_muts,]$mut_status = "bulk"
all_subs[all_subs$mutID %in% pd51122_single_bulk_muts & all_subs$Case_ID == "PD51122" & !all_subs$mutID %in% pd51122_private_muts,]$mut_status = "bulk"
all_subs[all_subs$mutID %in% pd51123_single_bulk_muts & all_subs$Case_ID == "PD51123" & !all_subs$mutID %in% pd51123_private_muts,]$mut_status = "bulk"

all_subs[all_subs$mutID %in% pd50297_multi_bulk_muts & all_subs$Case_ID == "PD50297",]$mut_status = "multi"
all_subs[all_subs$mutID %in% pd51122_multi_bulk_muts & all_subs$Case_ID == "PD51122",]$mut_status = "multi"
all_subs[all_subs$mutID %in% pd51123_multi_bulk_muts & all_subs$Case_ID == "PD51123",]$mut_status = "multi"

#all_subs[is.na(all_subs$mut_status),] none
#table(all_subs$mut_status)

all_subs[all_subs$mutID %in% pd50297_tumour_muts & all_subs$Case_ID == "PD50297",]$pure_tumour_mut = T
all_subs[all_subs$mutID %in% pd51122_tumour_muts & all_subs$Case_ID == "PD51122",]$pure_tumour_mut = T
all_subs[all_subs$mutID %in% pd51123_tumour_muts & all_subs$Case_ID == "PD51123",]$pure_tumour_mut = T
#table(all_subs$pure_tumour_mut)

write.table(all_subs, "all_sample_subs_distribution_annotated_20221201.txt", col.names = T, row.names = F, sep = "\t", quote = F)

#no T/TRUE issues
table(all_subs$Wildtype)
table(all_subs$Mutant)

```

##Get list of all indels, annotated by distribution

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/indels/04_bbinom")

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

manifest_flt = manifest[manifest$keep == "Y",] #4 samples failed QC

indel_files = list.files(".", ".pindel.HP.9.read.direction.germline.beta.binom.unk.amb.min.depth.filtered.txt")

all_indels = data.frame()
for(file in indel_files){
  
  sample = unlist(strsplit(file, ".pindel.HP.9.read.direction.germline.beta.binom.unk.amb.min.depth.filtered.txt"))
  
  if(file.exists(file)){
    
    data = read.table(file, header = T, sep = '\t', stringsAsFactors = F, colClasses = c("character", "character", "character", "numeric", "character", "character", "numeric", "numeric", "numeric", "character", "character", "character", "character", "character", "character"))
    print(sample)
    all_indels = rbind(all_indels, data)
    
  } 
}

all_indels$bulk_sample = unlist(lapply(strsplit(all_indels$Sample, "_"), "[[", 1))
all_indels = all_indels[all_indels$Sample %in% manifest_flt$sample,]

table(all_indels$bulk_sample)
#all_indels[all_indels$bulk_sample == "PD51123x3",]$bulk_sample = "PD51123x"
#all_indels[all_indels$bulk_sample == "PD51123aa3",]$bulk_sample = "PD51123aa"
#all_indels[all_indels$bulk_sample == "PD51123r3",]$bulk_sample = "PD51123r"
#all_indels[all_indels$bulk_sample == "PD51122ae2",]$bulk_sample = "PD51122ae"
all_indels$mut_status = NA
all_indels$pure_tumour_mut = F
all_indels$mutID = paste(all_indels$Chr, all_indels$Position, all_indels$Wildtype, all_indels$Mutant, sep = "_")

#single sample only called 
pd50297_private_muts = names(table(all_indels[all_indels$Case_ID == "PD50297",]$mutID))[table(all_indels[all_indels$Case_ID == "PD50297",]$mutID) == 1]
pd51122_private_muts = names(table(all_indels[all_indels$Case_ID == "PD51122",]$mutID))[table(all_indels[all_indels$Case_ID == "PD51122",]$mutID) == 1]
pd51123_private_muts = names(table(all_indels[all_indels$Case_ID == "PD51123",]$mutID))[table(all_indels[all_indels$Case_ID == "PD51123",]$mutID) == 1]

#bulk only called, anything taken from the original bulk sample
bulk_sample.list = list()
for(bulk in unique(all_indels$bulk_sample)){
  
  bulk_sample.list[[bulk]] = unique(all_indels[all_indels$bulk_sample == bulk,]$mutID)
  
}

bulk_sample.df = data.frame(bulk_sample = rep(names(bulk_sample.list), as.numeric(unlist(lapply(bulk_sample.list, length)))),
                            mutID = as.character(unlist(bulk_sample.list)))
bulk_sample.df$case.id = substr(bulk_sample.df$bulk_sample, 0, 7)

pd50297_single_bulk_muts = names(table(bulk_sample.df[bulk_sample.df$case.id == "PD50297",]$mutID))[table(bulk_sample.df[bulk_sample.df$case.id == "PD50297",]$mutID) == 1]
pd50297_multi_bulk_muts = names(table(bulk_sample.df[bulk_sample.df$case.id == "PD50297",]$mutID))[table(bulk_sample.df[bulk_sample.df$case.id == "PD50297",]$mutID) > 1]
pd51122_single_bulk_muts = names(table(bulk_sample.df[bulk_sample.df$case.id == "PD51122",]$mutID))[table(bulk_sample.df[bulk_sample.df$case.id == "PD51122",]$mutID) == 1]
pd51122_multi_bulk_muts = names(table(bulk_sample.df[bulk_sample.df$case.id == "PD51122",]$mutID))[table(bulk_sample.df[bulk_sample.df$case.id == "PD51122",]$mutID) > 1]
pd51123_single_bulk_muts = names(table(bulk_sample.df[bulk_sample.df$case.id == "PD51123",]$mutID))[table(bulk_sample.df[bulk_sample.df$case.id == "PD51123",]$mutID) == 1]
pd51123_multi_bulk_muts = names(table(bulk_sample.df[bulk_sample.df$case.id == "PD51123",]$mutID))[table(bulk_sample.df[bulk_sample.df$case.id == "PD51123",]$mutID) > 1]

#pure tumour variants
pd50297_tumour_samples = manifest[manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$case.id == "PD50297",]$sample
pd51122_tumour_samples = manifest[manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$case.id == "PD51122",]$sample
pd51123_tumour_samples = manifest[manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$case.id == "PD51123",]$sample

pd50297_tumour_muts = unique(all_indels[all_indels$Sample %in% pd50297_tumour_samples,]$mutID)
pd51122_tumour_muts = unique(all_indels[all_indels$Sample %in% pd51122_tumour_samples,]$mutID)
pd51123_tumour_muts = unique(all_indels[all_indels$Sample %in% pd51123_tumour_samples,]$mutID)

#add annotations
all_indels[all_indels$mutID %in% pd50297_private_muts & all_indels$Case_ID == "PD50297",]$mut_status = "sample"
all_indels[all_indels$mutID %in% pd51122_private_muts & all_indels$Case_ID == "PD51122",]$mut_status = "sample"
all_indels[all_indels$mutID %in% pd51123_private_muts & all_indels$Case_ID == "PD51123",]$mut_status = "sample"

all_indels[all_indels$mutID %in% pd50297_single_bulk_muts & all_indels$Case_ID == "PD50297" & !all_indels$mutID %in% pd50297_private_muts,]$mut_status = "bulk"
all_indels[all_indels$mutID %in% pd51122_single_bulk_muts & all_indels$Case_ID == "PD51122" & !all_indels$mutID %in% pd51122_private_muts,]$mut_status = "bulk"
all_indels[all_indels$mutID %in% pd51123_single_bulk_muts & all_indels$Case_ID == "PD51123" & !all_indels$mutID %in% pd51123_private_muts,]$mut_status = "bulk"

all_indels[all_indels$mutID %in% pd50297_multi_bulk_muts & all_indels$Case_ID == "PD50297",]$mut_status = "multi"
all_indels[all_indels$mutID %in% pd51122_multi_bulk_muts & all_indels$Case_ID == "PD51122",]$mut_status = "multi"
all_indels[all_indels$mutID %in% pd51123_multi_bulk_muts & all_indels$Case_ID == "PD51123",]$mut_status = "multi"

#all_indels[is.na(all_indels$mut_status),] none
#table(all_indels$mut_status)

all_indels[all_indels$mutID %in% pd50297_tumour_muts & all_indels$Case_ID == "PD50297",]$pure_tumour_mut = T
all_indels[all_indels$mutID %in% pd51122_tumour_muts & all_indels$Case_ID == "PD51122",]$pure_tumour_mut = T
all_indels[all_indels$mutID %in% pd51123_tumour_muts & all_indels$Case_ID == "PD51123",]$pure_tumour_mut = T
#table(all_indels$pure_tumour_mut)

write.table(all_indels, "all_sample_indels_distribution_annotated_20221206.txt", col.names = T, row.names = F, sep = "\t", quote = F)

```

##ndDPClust   

Alter the PD51122n chr1 file to correct the subclonal chr1 status - should be clonal 3+2.

###Prepare data files

Copy across necessary files for samples that meet minimum purity threshold (40%). This threshold is passed by the same samples, whether the purity is taken from Battenberg or our method.

Copy over SNV VCF and copy number files

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input

#copy number
infile="/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt"
cat $infile | awk -F "\t" '{ if($28 >= 0.4 && $28 != "NA" && $1 != "sample") { print $1} }' | while IFS=$'\t' read -r -a tmp; 
do
#echo ${tmp[0]} #sample ID
cp /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/copy_number/battenberg/bulk/tumour/${tmp[0]}* /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input
done

#subs
infile="/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt"
cat $infile | awk -F "\t" '{ if($28 >= 0.4 && $28 != "NA" && $1 != "sample") { print $1} }' | while IFS=$'\t' read -r -a tmp; 
do
#echo ${tmp[0]} #sample ID
cp /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/subs/07_indel/${tmp[0]}.standard.gnomAD.AF.0.001.PON.read.direction.low_cov.bq.mq.binom.bbinom.5bp.indel.filtered.vcf /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input
done

```

Change column names in vcf to expected one
```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input

for f in $(ls *.standard.gnomAD.AF.0.001.PON.read.direction.low_cov.bq.mq.binom.bbinom.5bp.indel.filtered.vcf);
do
sample=${f%.standard.gnomAD.AF.0.001.PON.read.direction.low_cov.bq.mq.binom.bbinom.5bp.indel.filtered.vcf}
echo $sample
sed 's/#CHROM.*/#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tNORMAL\tTUMOUR/' ${sample}.standard.gnomAD.AF.0.001.PON.read.direction.low_cov.bq.mq.binom.bbinom.5bp.indel.filtered.vcf > ${sample}.edited.vcf
done

```

Create separate directories per patient.

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input

for f in $(ls *.battenberg.subclones.txt | cut -c1-7 | uniq);
do
mkdir -p $f
mv $f*.* $f
done 

```

Create input files for allelecount. Keep sex chromosomes here.

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input

for f in $(ls -d PD*);
do
cut -f 1,2,4,5 $f/*.edited.vcf | sort -k1,1 -k2,2n -k3,4 | uniq > $f/${f}_loci.txt
grep -v '#' $f/${f}_loci.txt > $f/${f}_loci_SNVs_nohash.txt
mv $f/${f}_loci_SNVs_nohash.txt $f/${f}_loci.txt
done

```

Create a txt file containing a list of the locations of the BAM files. 

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input

infile="/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt"
cat $infile | awk -F "\t" '{ if($28 >= 0.4 && $28 != "NA" && $1 != "sample") { print $3, $6, $17} }' | uniq > patients.txt

while read line;
do
set $line
echo $1
echo $2
for f in $(ls $1/*.edited.vcf);
do
#echo $f
filename=$(basename -- "$f")
SAMPLE="${filename%.edited.vcf}"
echo "/nfs/cancer_ref01/nst_links/live/${2}/${SAMPLE}/${SAMPLE}.sample.dupmarked.bam" >> $1/${1}_bam_list.txt
done
done < patients.txt

```

Run allelecounter on cgpfarm.

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/

for f in $(ls -d PD*);
do

while read i;
do

sampleID=$(basename $i | sed "s/[.].*//")
echo $sampleID
#echo $i

bsub -J"${sampleID}" -o ${f}/${sampleID}.o -e ${f}/${sampleID}.e -q normal -n 5 -R"select[mem>10000] rusage[mem=10000] span[hosts=1]" -M10000 "/nfs/users/nfs_m/my4/bin/alleleCounter -l ${f}/${f}_loci.txt -b /nfs/cancer_ref01/nst_links/live/2571/${sampleID}/${sampleID}.sample.dupmarked.bam -o /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/${f}/${sampleID}_alleleFrequencies.txt -r /lustre/scratch119/casm/team78pipelines/reference/human/GRCh38_full_analysis_set_plus_decoy_hla/genome.fa -m 25 -q 30"

done < ${f}/${f}_bam_list.txt 

done


wc -l PD*/PD*_loci.txt
wc -l PD*/PD*_alleleFrequencies.txt
#none truncated

```

Convert chromosomes back to hg19 style
```{r}

setwd('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/')

INPUT_PATH = "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/"

data = read.table(paste0(INPUT_PATH, '/patients.txt'), header = F, sep = ' ')
names(data) = c('Patient', 'ProjectID', 'Sex')
data$Sex = tolower(data$Sex)

for( k in 1:nrow(data) ){
  samples = unlist(strsplit(list.files(path = paste0(data$Patient[k]), pattern = '.battenberg.cellularity_ploidy.txt'), '.battenberg.cellularity_ploidy.txt'))
  
  #loci file
    loci_file = read.table(paste0(INPUT_PATH, '/', data$Patient[k], '/', data$Patient[k], '_loci.txt'), header = F, sep = '\t', stringsAsFactors = F)
    if(sum(grepl("chr", loci_file$V1)) != 0){
      loci_file$V1 = unlist(strsplit(loci_file$V1, 'chr'))[c(F, T)]
      write.table(loci_file, paste0(INPUT_PATH, '/', data$Patient[k], '/', data$Patient[k], '_loci.txt'), col.names = F, row.names = F, sep = '\t', quote = F)
    }
  for(i in 1:length(samples)){
    
    #allele frequencies
    allele_frequencies_file = read.table(paste0(INPUT_PATH, '/', data$Patient[k], '/', samples[i], '_alleleFrequencies.txt'), header = T, sep = '\t', stringsAsFactors = F, comment.char = '', check.names = F)
    if(sum(grepl("chr", allele_frequencies_file$`#CHR`)) != 0){
      allele_frequencies_file$`#CHR` = unlist(strsplit(allele_frequencies_file$`#CHR`, 'chr'))[c(F, T)]
      write.table(allele_frequencies_file, paste0(INPUT_PATH, '/', data$Patient[k], '/', samples[i], '_alleleFrequencies.txt'), col.names = T, row.names = F, sep = '\t', quote = F)
    }
    
    #subclone files
    subclone_file = read.table(paste0(INPUT_PATH, '/', data$Patient[k], '/', samples[i], '.battenberg.subclones.txt'), header = T, sep = '\t', stringsAsFactors = F, comment.char = '', check.names = F)
    if(sum(grepl("chr", subclone_file$chr)) != 0){
      subclone_file$chr = unlist(strsplit(subclone_file$chr, 'chr'))[c(F, T)]
      write.table(subclone_file, paste0(INPUT_PATH, '/', data$Patient[k], '/', samples[i], '.battenberg.subclones.txt'), col.names = T, row.names = F, sep = '\t', quote = F)
    }
  }
}
    
```

Now we need to split the WT and MT counts across all samples into two files. We can do this using the loci and allelefrequency output files.
```{r}

setwd('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/')

wt_mt_count_file_generator <- function(filepath){
  file = list.files(filepath, '_loci.txt')
  loci_file = read.table(paste0(filepath, '/', file), header = F, sep = '\t', stringsAsFactors = F)
  loci_file = loci_file[!(loci_file$V1 %in% loci_file[duplicated(loci_file[,1:2]),]$V1 & loci_file$V2 %in% loci_file[duplicated(loci_file[,1:2]),]$V2),] #remove loci that violate infinite sites assumption
  wt_mat = as.data.frame(matrix(NA, 
                                ncol =  length(unlist(strsplit(list.files(path = filepath, '_alleleFrequencies.txt'), '_alleleFrequencies.txt'))) + 1, 
                                nrow = nrow(loci_file),
                                dimnames = list(NULL, c('ID', unlist(strsplit(list.files(path = filepath, '_alleleFrequencies.txt'), '_alleleFrequencies.txt'), use.names = F)))))
  wt_mat$ID = paste(loci_file$V1, loci_file$V2, sep = '_')
  
  mt_mat = as.data.frame(matrix(NA, 
                                ncol =  length(unlist(strsplit(list.files(path = filepath, '_alleleFrequencies.txt'), '_alleleFrequencies.txt'))) + 1, 
                                nrow = nrow(loci_file),
                                dimnames = list(NULL, c('ID', unlist(strsplit(list.files(path = filepath, '_alleleFrequencies.txt'), '_alleleFrequencies.txt'), use.names = F)))))
  mt_mat$ID = paste(loci_file$V1, loci_file$V2, sep = '_')
  
  for ( i in unlist(strsplit(list.files(path = filepath, '_alleleFrequencies.txt'), '_alleleFrequencies.txt')) ){
    alleleFreq_file = read.table(paste0(filepath, '/', i, '_alleleFrequencies.txt'), header = T, sep = '\t', comment.char = '')
    for ( j in 1:nrow( wt_mat ) ){
      chr = unlist(strsplit(wt_mat$ID[j], '_'))[1]
      pos = unlist(strsplit(wt_mat$ID[j], '_'))[2]
      ref = loci_file[loci_file$V1 == chr & loci_file$V2 == pos, ]$V3
      alt = loci_file[loci_file$V1 == chr & loci_file$V2 == pos, ]$V4
      wt_mat[j, i] = alleleFreq_file[alleleFreq_file$X.CHR == chr & alleleFreq_file$POS == pos, paste0('Count_', ref)]
      mt_mat[j, i] = alleleFreq_file[alleleFreq_file$X.CHR == chr & alleleFreq_file$POS == pos, paste0('Count_', alt)]
    }
  }
  wt_mat <- wt_mat[rowSums(mt_mat[2:ncol(mt_mat)]) > 0,] #remove any rows where no SNVs are called, this shouldn't happen with the loci list given but is a precautionary extra filter
  mt_mat <- mt_mat[rowSums(mt_mat[2:ncol(mt_mat)]) > 0,]
  write.table(wt_mat, paste0(filepath, '/wt_count_file.txt'), sep = '\t', quote = F, row.names = F, col.names = T)
  write.table(mt_mat, paste0(filepath, '/mt_count_file.txt'), sep = '\t', quote = F, row.names = F, col.names = T)
}

```

```{r}

directory <- '/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/'

for (k in c("PD50297", "PD51122", "PD51123")){
  wt_mt_count_file_generator(paste0(directory, k))
}

```

```{r}

setwd('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/')

sample_list_generator <- function(filepath){
  sample.list <- unlist(strsplit(list.files(paste0(filepath), pattern = '.battenberg.rho_and_psi.txt'), '.battenberg.rho_and_psi.txt'))
  write.table(sample.list, paste0(filepath, '/sample_names.txt'), sep = '\t', quote = F, row.names = F, col.names = F)
}

cellularity_list_generator <- function(filepath){
  cellularity <- c()
  for(f in unlist(strsplit(list.files(paste0(filepath), pattern = '.battenberg.rho_and_psi.txt'), '.battenberg.rho_and_psi.txt'))){
    file <- read.table(paste0(filepath, '/', f, '.battenberg.rho_and_psi.txt'), stringsAsFactors = F, header = T, row.names = 1, sep = '\t')
    cellularity <- rbind(cellularity, file['FRAC_GENOME', 'rho']) #corrected to right way around 7/3/22
    write.table(cellularity, paste0(filepath, '/cellularity.txt'), sep = '\t', quote = F, row.names = F, col.names = F)
  }
}

subclone_list_generator <- function(filepath){
  subclone.list <- unlist(list.files(paste0(filepath), pattern = '.battenberg.subclones.txt'))
  write.table(subclone.list, paste0(filepath, '/subclone.files.txt'), sep = '\t', quote = F, row.names = F, col.names = F)
}

multidimdp_preproc <- function(filepath){
  sample_list_generator(filepath)
  cellularity_list_generator(filepath)
  subclone_list_generator(filepath)
}

```

```{r}

for ( i in c("PD50297", "PD51122", "PD51123")){
  multidimdp_preproc(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', i))
}

```

Remove sex chromosomes here.

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/")

patients = c("PD50297", "PD51122", "PD51123")

for(patient in patients){
  wt_df = read.table(paste0(patient, "/wt_count_file.txt"), header = T, sep = '\t', stringsAsFactors = F)
  wt_df_auto = wt_df[!grepl("X|Y", wt_df$ID),]
  write.table(wt_df_auto, paste0(patient, "/wt_count_file_auto_only.txt"), col.names = T, sep = '\t', quote = F, row.names = F)
  mt_df = read.table(paste0(patient, "/mt_count_file.txt"), header = T, sep = '\t', stringsAsFactors = F)
  mt_df_auto = mt_df[!grepl("X|Y", mt_df$ID),]
  write.table(mt_df_auto, paste0(patient, "/mt_count_file_auto_only.txt"), col.names = T, sep = '\t', quote = F, row.names = F)
}

```

#####Create the multidimensional input files for DPClust

- Run in the directory with all the prerequisite files in one place

```{bash}

#change sex line

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/

HOME_DIR='/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/'

#edit patients file to have is.male T/F instead

while read line;
do
set $line
cd ${HOME_DIR}$1
/software/R-3.6.1/bin/Rscript /nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/GetDirichletProcessInfo_multipleSamples_updated_NAP.R \
sample_names.txt \
cellularity.txt \
subclone.files.txt \
mt_count_file_auto_only.txt \
wt_count_file_auto_only.txt \
$3 #is.male
done < ${HOME_DIR}patients.txt

```

###Run DPClust on samples

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

HOME_DIR='/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/'

for f in $(ls -d PD*);
do
JOB_NAME=${f}_dpclust
SAMPLE_NAME=$f
MEM=60000
CORES=10
QUEUE=long

cd ${HOME_DIR}$f
#-J jobname -M mem -R select host -n cores -q queue name -e stderror putput -o stdout

bsub -G team294-vwork -J"${JOB_NAME}" -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -q ${QUEUE} -e %J.stderr -o %J.stdout \
"xvfb-run -a Rscript --no-save --no-restore --verbose /nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/runDP_new_nonzero_NAP.R sample_names.txt cellularity.txt ${SAMPLE_NAME} > output_multidim_dpclust.Rout 2>&1"

done

```

###Run assignment of mutations to clusters per triplet combination of samples (where n > 3)

```{r}

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

table(manifest[manifest$battenberg_purity >= 0.4 & !is.na(manifest$battenberg_purity),]$case.id)
#PD50297 PD51122 PD51123 
#     7       7       5 


```

Calculate jobs in array

```{r}

n = 5 #number of samples you have
choose(n,3) #10 jobs in the array

n = 7 #number of samples you have
choose(n,3) #35 jobs in the array

```

Create a text file (triplet_file) with:
patient     no_of_samples     binom_coef

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/")

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

write.table(cbind(data.frame(table(manifest[manifest$battenberg_purity >= 0.4 & !is.na(manifest$battenberg_purity),]$case.id)), c(35, 35, 10)), "triplet_file.txt", col.names = F, row.names = F, sep = '\t', quote = F)

```


```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

HOME_DIR='/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/'

infile='triplet_file.txt'
less $infile | while IFS=$'\t' read -r -a tmp;
do

echo ${tmp[0]} #case id
echo ${tmp[1]} #number of samples
echo ${tmp[2]} #binomial coefficient

cd ${HOME_DIR}${tmp[0]}
MEM=30000
QUEUE=normal #may need long queue for 1000s
CORES=10
ARRAY=${tmp[2]} #binomial coefficient
RUN=1 #number of patients
CHOOSE_FROM=${tmp[1]} #number of samples
CHOOSE_NUMBER=3 #number of samples to run in each batch
JOB_NAME="multidim"
PATIENT=${tmp[0]} #case id
SAMPLE_NAMES='sample_names.txt'
CELLULARITY='cellularity.txt'

bsub -G team294-vwork -J"${JOB_NAME}[1-${ARRAY}]%100" -q ${QUEUE} -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -e %I.stderr -o %I.stdout \
"/nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/BASH/multidimclustparallel_NAP.sh ${RUN} ${CHOOSE_FROM} ${CHOOSE_NUMBER} ${PATIENT} ${SAMPLE_NAMES} ${CELLULARITY}"

done

```

###Combine clustering assignments across triplets

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim"

HOME_DIR='/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/'

infile='triplet_file.txt'
less $infile | while IFS=$'\t' read -r -a tmp;
do

echo ${tmp[0]} #case id
echo ${tmp[1]} #number of samples
echo ${tmp[2]} #binomial coefficient

cd ${HOME_DIR}${tmp[0]}
MEM=30000
QUEUE=long #may need long queue for 1000s
CORES=10
ARRAY=${tmp[2]} #binomial coefficient
RUN=1 #number of patients
CHOOSE_FROM=${tmp[1]} #number of samples
CHOOSE_NUMBER=3 #number of samples to run in each batch
JOB_NAME="paraclust"
PATIENT=${tmp[0]} #case id
SAMPLE_NAMES='sample_names.txt'
CELLULARITY='cellularity.txt'

bsub -G team294-vwork -J"${JOB_NAME}" -M${MEM} -R"select[mem>${MEM}] rusage[mem=${MEM}] span[hosts=1]" -n ${CORES} -q ${QUEUE} -e %J.stderr -o %J.stdout \
"xvfb-run -a Rscript --no-save --no-restore --verbose /nfs/users/nfs_t/to3/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_multidim/multidimDPClust/R/combineClusteringParallelByAssignment_NAP.R \
${RUN} ${CHOOSE_FROM} ${CHOOSE_NUMBER} ${PATIENT} ${SAMPLE_NAMES} ${CELLULARITY} > output_combineClusteringParallelByAssignment.Rout 2>&1"

done

```

###Review the output

```{r}

library(ggplot2)
library(ggpubr)
library(reshape2)

patient = c("PD50297", "PD51122", "PD51123")

for(i in patient){
  setwd(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', i, '/md_out'))
  mut_clusters = read.table(paste0(i, '_consensusClustersByParallelNodeAssignment.txt'), header = T, sep = '\t', stringsAsFactors = F)
  mut_clusters = mut_clusters[((mut_clusters$no.muts / sum(mut_clusters$no.muts)) >= 0.01), ]
  write.table(mut_clusters, paste0(i, '_filtered_clusters.txt'), sep ='\t', row.names = F, col.names = T, quote = F)
  
  clusters = read.table(paste0(i, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)

  clusters = clusters[order(clusters$cluster.no, clusters$chr, clusters$pos),]
  clusters = clusters[clusters$cluster.no %in% names(table(clusters$cluster.no)[table(clusters$cluster.no) >= (0.01 * nrow(clusters))]),]

  clusters.m = reshape2::melt(clusters, id.vars = c('cluster.no', 'chr', 'pos'))
  
  clusters.l = reshape(aggregate(value ~ cluster.no + variable, clusters.m, mean), idvar = "cluster.no", timevar = "variable", direction = "wide")
  clusters.l$mut.no = as.vector(table(clusters$cluster.no))
  colnames(clusters.l) = gsub('value.', '', colnames(clusters.l))
  write.table(clusters.l, paste0(i, '_filtered_clusters_mean_CCF.txt'), sep ='\t', row.names = F, col.names = T, quote = F)
  
  clusters.m$cluster.no = as.factor(clusters.m$cluster.no)
  clusters.m$variant = as.character(paste(clusters.m$cluster.no, clusters.m$chr, clusters.m$pos, sep = '_'))
  clusters.m = clusters.m[order(clusters.m$cluster.no),]
  #clusters.m$variant.no = c(1:nrow(clusters.m))
  clusters.m$variant = factor(clusters.m$variant, levels = unique(as.character(clusters.m$variant)))

  p = ggplot(clusters.m) + geom_bar(mapping = aes(x = variant, y = value, fill = cluster.no), stat = 'identity') + facet_grid(variable ~ .) + coord_cartesian(expand = F, ylim = c(0,1.5)) + theme_pubr() + theme(strip.text.y = element_text(angle = 0), strip.background = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank()) + geom_hline(yintercept = 1, lty = 2) + geom_hline(yintercept = 0, lty = 1)
  ggsave(paste0(i, '_cluster_plot.png'), plot = p, device = 'png', units = "in", width = 12, height = 6)
  
}

```

####PD50297

Cluster 9 is the trunk.

```{r}

patient = "PD50297"

setwd(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient, '/md_out'))
clusters = read.table(paste0(patient, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)

keep_clusters = names(table(clusters$cluster.no)[(table(clusters$cluster.no)/ nrow(clusters)) >= 0.01])

clusters = clusters[clusters$cluster.no %in% keep_clusters,]

table(clusters[clusters$cluster.no == 1,]$chr) #spread across genome

library(ComplexHeatmap)

Heatmap(clusters[clusters$cluster.no == 1, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 2, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 3, c(3:9)]) #combination of au, ax and ad clusters
Heatmap(clusters[clusters$cluster.no == 4, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 5, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 6, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 7, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 9, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 10, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 11, c(3:9)]) #OK

#clearly only a proportion found in a given sample, cluster needs splitting
hist(clusters[clusters$cluster.no == 3,]$PD50297au_subclonalFraction, breaks = 100)
hist(clusters[clusters$cluster.no == 3,]$PD50297ax_subclonalFraction, breaks = 100)
hist(clusters[clusters$cluster.no == 3,]$PD50297ad_subclonalFraction, breaks = 100)

clusters$cluster.no.new = clusters$cluster.no

#assign to new clusters
clusters[clusters$cluster.no == 3 &
           clusters$PD50297au_subclonalFraction > 0,]$cluster.no.new = "3a"
clusters[clusters$cluster.no == 3 &
           clusters$PD50297ax_subclonalFraction > 0 &
           ! clusters$cluster.no.new %in% c("3a"),]$cluster.no.new = "3b"
clusters[clusters$cluster.no == 3 &
           clusters$PD50297ad_subclonalFraction > 0 &
           ! clusters$cluster.no.new %in% c("3a", "3b"),]$cluster.no.new = "3c"

#keep all final clusters don't contain further merged clusters
table(clusters$cluster.no.new)
#   1   10   11    2   3a   3b   3c    4    5    6    7    9 
#2594 1184  733 1251  733  465   54  397  223  538 1054  929 

Heatmap(clusters[clusters$cluster.no.new == "3a", c(3:9)])
Heatmap(clusters[clusters$cluster.no.new == "3b", c(3:9)])
Heatmap(clusters[clusters$cluster.no.new == "3c", c(3:9)]) # a few of these are more present in av but too few to form their own cluster
#all look OK - small numbers of mutations probably belong to other clusters but these are a very definite minority

clusters$cluster.no = clusters$cluster.no.new

write.table(clusters[, c(1:10)], paste0(patient, '_allClusterassignmentsFromParallelRuns_clusters_reviewed.txt'), col.names = T, row.names = F, sep = "\t", quote = F)

clusters = clusters[, c(1:10)]
clusters$mut_ID = paste0(clusters$chr, "_", clusters$pos)

#read in loci file
loci = read.table(paste0("../", patient, "_loci.txt"), header = F, sep = "\t", stringsAsFactors = F)
loci$mut_ID = paste0(loci$V1, "_", loci$V2)
loci$mut_ID2 = paste0(loci$V1, "_", loci$V2, "_", loci$V3, "_", loci$V4)
row.names(loci) = loci$mut_ID

clusters$mut_ID2 = loci[clusters$mut_ID,]$mut_ID2

library("GenomicRanges")
library("Rsamtools")
library("MASS")

genomeFile <- "/lustre/scratch119/casm/team78pipelines/reference/human/GRCh38_full_analysis_set_plus_decoy_hla/genome.fa"

cluster_vec = unique(clusters$cluster.no)

plot_trinuc_context <- function(input){
  mutations = data.frame(matrix(unlist(strsplit(input, "_")), ncol = 4, byrow= T))
  names(mutations) = c("chr", "pos", "ref", "mut")
  mutations = mutations[(mutations$ref %in% c("A","C","G","T")) & (mutations$mut %in% c("A","C","G","T")) & mutations$chr %in% paste0("chr", c(1:22,"X","Y")),]
  mutations$trinuc_ref = as.vector(scanFa(genomeFile, GRanges(mutations$chr, IRanges(as.numeric(mutations$pos)-1,
                                                                                     as.numeric(mutations$pos)+1))))

  # 2. Annotating the mutation from the pyrimidine base
  ntcomp = c(T="A",G="C",C="G",A="T")
  mutations$sub = paste(mutations$ref,mutations$mut,sep=">")
  mutations$trinuc_ref_py = mutations$trinuc_ref
  for (j in 1:nrow(mutations)) {
    if (mutations$ref[j] %in% c("A","G")) { # Purine base
      mutations$sub[j] = paste(ntcomp[mutations$ref[j]],ntcomp[mutations$mut[j]],sep=">")
      mutations$trinuc_ref_py[j] = paste(ntcomp[rev(strsplit(mutations$trinuc_ref[j],split="")[[1]])],collapse="")
    }
  }
  freqs = table(paste(mutations$sub,paste(substr(mutations$trinuc_ref_py,1,1),substr(mutations$trinuc_ref_py,3,3),sep="-"),sep=","))
  sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
  ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
  full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")
  freqs_full = freqs[full_vec]; freqs_full[is.na(freqs_full)] = 0; names(freqs_full) = full_vec
  
  #3. plot
  xstr = paste(substr(full_vec,5,5), substr(full_vec,1,1), substr(full_vec,7,7), sep="")

  colvec = rep(c("dodgerblue","black","red","grey70","olivedrab3","plum2"),each=16)
  y = freqs_full; maxy = max(y)
  h = barplot(y, las=2, col=colvec, border=NA, ylim=c(0,maxy*1.5), space=1, cex.names=0.6, names.arg=xstr, ylab="Number of mutations")
  for (j in 1:length(sub_vec)) {
    xpos = h[c((j-1)*16+1,j*16)]
    rect(xpos[1]-0.5, maxy*1.2, xpos[2]+0.5, maxy*1.3, border=NA, col=colvec[j*16])
    text(x=mean(xpos), y=maxy*1.3, pos=3, label=sub_vec[j])
  } 
  title(main = paste0(patient), sub = paste0("filtered subs = ", length(input)))
}

for(cluster in cluster_vec){
  cluster_vars = paste0("chr", clusters[clusters$cluster.no == cluster,]$mut_ID2)
  pdf(paste0(patient, "_cluster_", cluster, "_mut_spec.pdf"), height = 5, width = 12)
  plot_trinuc_context(cluster_vars)
  dev.off()
}

```

####PD51122

6 is the root. Looks like 9 needs merging with it.

```{r}

patient = "PD51122"

setwd(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient, '/md_out'))
clusters = read.table(paste0(patient, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)

keep_clusters = names(table(clusters$cluster.no)[(table(clusters$cluster.no)/ nrow(clusters)) >= 0.01])

clusters = clusters[clusters$cluster.no %in% keep_clusters,]

table(clusters$cluster.no)
#  1   2   3   4   5   6   7   8   9  10  11  12  13  14 
#999 120  80 360 386 874 203  86  98 127 172 203  83 248 

Heatmap(clusters[clusters$cluster.no == 1, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 2, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 3, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 4, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 5, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 6, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 7, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 8, c(3:9)]) #22n has a few with very high CCF
clusters[clusters$cluster.no == 8 & clusters$PD51122n_subclonalFraction > 0.8, ]$chr #mix of chromosomes

Heatmap(clusters[clusters$cluster.no == 9, c(3:9)]) #OK
table(clusters[clusters$cluster.no == 9, ]$chr) #many chromosomes

Heatmap(clusters[clusters$cluster.no == 10, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 11, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 12, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 13, c(3:9)]) #OK
Heatmap(clusters[clusters$cluster.no == 14, c(3:9)]) #OK

Heatmap(clusters[clusters$cluster.no %in% c(6, 9), c(3:9)]) 

#merge clusters
clusters[clusters$cluster.no %in% c(6, 9),]$cluster.no = "6_9"

write.table(clusters[, c(1:10)], paste0(patient, '_allClusterassignmentsFromParallelRuns_clusters_reviewed.txt'), col.names = T, row.names = F, sep = "\t", quote = F)

clusters = clusters[, c(1:10)]
clusters$mut_ID = paste0(clusters$chr, "_", clusters$pos)

#read in loci file
loci = read.table(paste0("../", patient, "_loci.txt"), header = F, sep = "\t", stringsAsFactors = F)
loci$mut_ID = paste0(loci$V1, "_", loci$V2)
loci$mut_ID2 = paste0(loci$V1, "_", loci$V2, "_", loci$V3, "_", loci$V4)
loci = loci[loci$mut_ID != loci[duplicated(loci$mut_ID),]$mut_ID,] #multi-allelic site remove for now
row.names(loci) = loci$mut_ID

clusters$mut_ID2 = loci[clusters$mut_ID,]$mut_ID2

library("GenomicRanges")
library("Rsamtools")
library("MASS")

genomeFile <- "/lustre/scratch119/casm/team78pipelines/reference/human/GRCh38_full_analysis_set_plus_decoy_hla/genome.fa"

cluster_vec = unique(clusters$cluster.no)

plot_trinuc_context <- function(input){
  mutations = data.frame(matrix(unlist(strsplit(input, "_")), ncol = 4, byrow= T))
  names(mutations) = c("chr", "pos", "ref", "mut")
  mutations = mutations[(mutations$ref %in% c("A","C","G","T")) & (mutations$mut %in% c("A","C","G","T")) & mutations$chr %in% paste0("chr", c(1:22,"X","Y")),]
  mutations$trinuc_ref = as.vector(scanFa(genomeFile, GRanges(mutations$chr, IRanges(as.numeric(mutations$pos)-1,
                                                                                     as.numeric(mutations$pos)+1))))

  # 2. Annotating the mutation from the pyrimidine base
  ntcomp = c(T="A",G="C",C="G",A="T")
  mutations$sub = paste(mutations$ref,mutations$mut,sep=">")
  mutations$trinuc_ref_py = mutations$trinuc_ref
  for (j in 1:nrow(mutations)) {
    if (mutations$ref[j] %in% c("A","G")) { # Purine base
      mutations$sub[j] = paste(ntcomp[mutations$ref[j]],ntcomp[mutations$mut[j]],sep=">")
      mutations$trinuc_ref_py[j] = paste(ntcomp[rev(strsplit(mutations$trinuc_ref[j],split="")[[1]])],collapse="")
    }
  }
  freqs = table(paste(mutations$sub,paste(substr(mutations$trinuc_ref_py,1,1),substr(mutations$trinuc_ref_py,3,3),sep="-"),sep=","))
  sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
  ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
  full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")
  freqs_full = freqs[full_vec]; freqs_full[is.na(freqs_full)] = 0; names(freqs_full) = full_vec
  
  #3. plot
  xstr = paste(substr(full_vec,5,5), substr(full_vec,1,1), substr(full_vec,7,7), sep="")

  colvec = rep(c("dodgerblue","black","red","grey70","olivedrab3","plum2"),each=16)
  y = freqs_full; maxy = max(y)
  h = barplot(y, las=2, col=colvec, border=NA, ylim=c(0,maxy*1.5), space=1, cex.names=0.6, names.arg=xstr, ylab="Number of mutations")
  for (j in 1:length(sub_vec)) {
    xpos = h[c((j-1)*16+1,j*16)]
    rect(xpos[1]-0.5, maxy*1.2, xpos[2]+0.5, maxy*1.3, border=NA, col=colvec[j*16])
    text(x=mean(xpos), y=maxy*1.3, pos=3, label=sub_vec[j])
  } 
  title(main = paste0(patient), sub = paste0("filtered subs = ", length(input)))
}

for(cluster in cluster_vec){
  cluster_vars = paste0("chr", clusters[clusters$cluster.no == cluster,]$mut_ID2)
  pdf(paste0(patient, "_cluster_", cluster, "_mut_spec.pdf"), height = 5, width = 12)
  plot_trinuc_context(cluster_vars)
  dev.off()
}


```

####PD51123

Cluster 2 is root.

```{r}

patient = "PD51123"

setwd(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient, '/md_out'))
clusters = read.table(paste0(patient, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)

keep_clusters = names(table(clusters$cluster.no)[(table(clusters$cluster.no)/ nrow(clusters)) >= 0.01])

clusters = clusters[clusters$cluster.no %in% keep_clusters,]

table(clusters$cluster.no)
#   1    2    3    4    5    7    8    9   12 
#1739  699  167 1085 1392  232 1363  244 1644 

#check no clusters are merged
Heatmap(clusters[clusters$cluster.no == 1, c(3:7)]) #OK
Heatmap(clusters[clusters$cluster.no == 2, c(3:7)]) #OK
Heatmap(clusters[clusters$cluster.no == 3, c(3:7)]) #OK
Heatmap(clusters[clusters$cluster.no == 4, c(3:7)]) #OK
Heatmap(clusters[clusters$cluster.no == 5, c(3:7)]) #OK
Heatmap(clusters[clusters$cluster.no == 7, c(3:7)]) #OK
Heatmap(clusters[clusters$cluster.no == 8, c(3:7)]) #two merged clusters of mostly private mutations to 23o and 23q
Heatmap(clusters[clusters$cluster.no == 9, c(3:7)]) #OK
Heatmap(clusters[clusters$cluster.no == 12, c(3:7)]) #OK
#cluster 8 needs splitting into 2

clusters$cluster.no.new = clusters$cluster.no

#split cluster 8
clusters[clusters$cluster.no == 8 & clusters$PD51123o_subclonalFraction == 0,]$cluster.no.new = "8b"
clusters[clusters$cluster.no.new == 8,]$cluster.no.new = "8a"

#check new split
Heatmap(clusters[clusters$cluster.no.new == "8a", c(3:7)])
Heatmap(clusters[clusters$cluster.no.new == "8b", c(3:7)])
#look OK

clusters$cluster.no = clusters$cluster.no.new
clusters = clusters[, c(1:8)]

write.table(clusters, paste0(patient, '_allClusterassignmentsFromParallelRuns_clusters_reviewed.txt'), col.names = T, row.names = F, sep = "\t", quote = F)

clusters$mut_ID = paste0(clusters$chr, "_", clusters$pos)

#read in loci file
loci = read.table(paste0("../", patient, "_loci.txt"), header = F, sep = "\t", stringsAsFactors = F)
loci$mut_ID = paste0(loci$V1, "_", loci$V2)
loci$mut_ID2 = paste0(loci$V1, "_", loci$V2, "_", loci$V3, "_", loci$V4)
#multi_allele_loci = loci[duplicated(loci$mut_ID),]$mut_ID #remove loci with more than one mutant allele
#loci = loci[loci$mut_ID != multi_allele_loci,]
row.names(loci) = loci$mut_ID

clusters$mut_ID2 = loci[clusters$mut_ID,]$mut_ID2

library("GenomicRanges")
library("Rsamtools")
library("MASS")

genomeFile <- "/lustre/scratch119/casm/team78pipelines/reference/human/GRCh38_full_analysis_set_plus_decoy_hla/genome.fa"

cluster_vec = unique(clusters$cluster.no)

plot_trinuc_context <- function(input){
  mutations = data.frame(matrix(unlist(strsplit(input, "_")), ncol = 4, byrow= T))
  names(mutations) = c("chr", "pos", "ref", "mut")
  mutations = mutations[(mutations$ref %in% c("A","C","G","T")) & (mutations$mut %in% c("A","C","G","T")) & mutations$chr %in% paste0("chr", c(1:22,"X","Y")),]
  mutations$trinuc_ref = as.vector(scanFa(genomeFile, GRanges(mutations$chr, IRanges(as.numeric(mutations$pos)-1,
                                                                                     as.numeric(mutations$pos)+1))))

  # 2. Annotating the mutation from the pyrimidine base
  ntcomp = c(T="A",G="C",C="G",A="T")
  mutations$sub = paste(mutations$ref,mutations$mut,sep=">")
  mutations$trinuc_ref_py = mutations$trinuc_ref
  for (j in 1:nrow(mutations)) {
    if (mutations$ref[j] %in% c("A","G")) { # Purine base
      mutations$sub[j] = paste(ntcomp[mutations$ref[j]],ntcomp[mutations$mut[j]],sep=">")
      mutations$trinuc_ref_py[j] = paste(ntcomp[rev(strsplit(mutations$trinuc_ref[j],split="")[[1]])],collapse="")
    }
  }
  freqs = table(paste(mutations$sub,paste(substr(mutations$trinuc_ref_py,1,1),substr(mutations$trinuc_ref_py,3,3),sep="-"),sep=","))
  sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
  ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
  full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")
  freqs_full = freqs[full_vec]; freqs_full[is.na(freqs_full)] = 0; names(freqs_full) = full_vec
  
  #3. plot
  xstr = paste(substr(full_vec,5,5), substr(full_vec,1,1), substr(full_vec,7,7), sep="")

  colvec = rep(c("dodgerblue","black","red","grey70","olivedrab3","plum2"),each=16)
  y = freqs_full; maxy = max(y)
  h = barplot(y, las=2, col=colvec, border=NA, ylim=c(0,maxy*1.5), space=1, cex.names=0.6, names.arg=xstr, ylab="Number of mutations")
  for (j in 1:length(sub_vec)) {
    xpos = h[c((j-1)*16+1,j*16)]
    rect(xpos[1]-0.5, maxy*1.2, xpos[2]+0.5, maxy*1.3, border=NA, col=colvec[j*16])
    text(x=mean(xpos), y=maxy*1.3, pos=3, label=sub_vec[j])
  } 
  title(main = paste0(patient), sub = paste0("filtered subs = ", length(input)))
}


for(cluster in cluster_vec){
  cluster_vars = paste0("chr", clusters[clusters$cluster.no == cluster,]$mut_ID2)
  pdf(paste0(patient, "_cluster_", cluster, "_mut_spec.pdf"), height = 5, width = 12)
  plot_trinuc_context(cluster_vars)
  dev.off()
}

```

###Regenerate additional plots

```{r}

library(ggplot2)
library(ggpubr)
library(reshape2)

patient = c("PD50297", "PD51122", "PD51123")

for(i in patient){
  setwd(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', i, '/md_out'))
  
  clusters = read.table(paste0(i, '_allClusterassignmentsFromParallelRuns_clusters_reviewed.txt'), header = T, sep = '\t', stringsAsFactors = F)

  clusters = clusters[order(clusters$cluster.no, clusters$chr, clusters$pos),]
  clusters = clusters[clusters$cluster.no %in% names(table(clusters$cluster.no)[table(clusters$cluster.no) >= (0.01 * nrow(clusters))]),]

  clusters.m = reshape2::melt(clusters, id.vars = c('cluster.no', 'chr', 'pos'))
  
  clusters.l = reshape(aggregate(value ~ cluster.no + variable, clusters.m, mean), idvar = "cluster.no", timevar = "variable", direction = "wide")
  clusters.l$mut.no = as.vector(table(clusters$cluster.no))
  colnames(clusters.l) = gsub('value.', '', colnames(clusters.l))
  write.table(clusters.l, paste0(i, '_filtered_clusters_mean_CCF_clusters_reviewed.txt'), sep ='\t', row.names = F, col.names = T, quote = F)
  
  clusters.m$cluster.no = as.factor(clusters.m$cluster.no)
  clusters.m$variant = as.character(paste(clusters.m$cluster.no, clusters.m$chr, clusters.m$pos, sep = '_'))
  clusters.m = clusters.m[order(clusters.m$cluster.no),]
  #clusters.m$variant.no = c(1:nrow(clusters.m))
  clusters.m$variant = factor(clusters.m$variant, levels = unique(as.character(clusters.m$variant)))

  p = ggplot(clusters.m) + geom_bar(mapping = aes(x = variant, y = value, fill = cluster.no), stat = 'identity') + facet_grid(variable ~ .) + coord_cartesian(expand = F, ylim = c(0,1.5)) + theme_pubr() + theme(strip.text.y = element_text(angle = 0), strip.background = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank()) + geom_hline(yintercept = 1, lty = 2) + geom_hline(yintercept = 0, lty = 1)
  ggsave(paste0(i, '_cluster_plot_clusters_reviewed.png'), plot = p, device = 'png', units = "in", width = 12, height = 6)
  
}

```

##Estimate tumour involvement per sample

This will be less accurate in the lower coverage samples.

###PD51122

```{r}

patient = "PD51122"
truncal_cluster = "6_9"

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

samples = manifest[manifest$case.id == patient & manifest$battenberg_purity >= 0.4 & !is.na(manifest$battenberg_purity), ]$sample

#get variants in trunk
vars = read.table(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient, '/md_out/', patient, '_allClusterassignmentsFromParallelRuns_clusters_reviewed.txt'), header = T, sep = '\t', stringsAsFactors = F) #reviewed and merged clusters
vars[vars$chr == 3 & vars$pos == 179218294,] #not assigned
vars[vars$chr == 17 & vars$pos == 7675140,] #root

cluster_vars = vars[vars$cluster.no == truncal_cluster,]
cluster_vars$ID = paste0(cluster_vars$chr, "_", cluster_vars$pos)

#identify which truncal variants cover two chromomes copies in 2+0 sites
dp_files = list.files(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient), "allDirichletProcessInfo_fromMultipleSamplesWithoutChromosomeLosses.txt", full.names = T)

trunc_pre_dup = list()
for(file in dp_files){
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  sample = unlist(strsplit(basename(file), "_allDirichletProcessInfo_fromMultipleSamplesWithoutChromosomeLosses.txt"))
  data$ID = paste0(data$chr, "_", data$pos)
  trunc_pre_dup[[sample]] = data[data$ID %in% cluster_vars$ID & data$subclonal.CN == 2 & is.na(data$nMaj2) & data$nMin1 == 0 & data$no.chrs.bearing.mut == 2,]$ID
}

hi_conf_loci = Reduce(intersect, trunc_pre_dup)
hi_conf_loci = paste0("chr", hi_conf_loci)
length(hi_conf_loci) #60

#read in VAF dataframe
VAF = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/", patient, "_VAF_final_merged_snps.txt"), header = T, sep = "\t", stringsAsFactors = F)

VAF_filter = VAF
VAF_filter$ID = paste0(unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 1)), "_", unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 2)))

mean_no_outliers <- function(x, remove_na = TRUE) { #early embryonic variants might skew overall calculation
  qnt <- quantile(x, probs = c(.25, .75), na.rm = remove_na)
  val <- 1.5 * IQR(x, na.rm = remove_na)
  y <- x
  y[x < (qnt[1] - val)] <- NA
  y[x > (qnt[2] + val)] <- NA
  z = mean(y, na.rm = remove_na)
  return(z)
}


tum_infiltration_est = data.frame(apply(VAF[VAF_filter$ID %in% hi_conf_loci, ], 2, function(x) mean_no_outliers(x)))
#tum_infiltration_est =  data.frame(apply(NV[row.names(VAF_filter)[VAF_filter$ID %in% hi_conf_loci], ], 2, sum) / apply(NR[row.names(VAF_filter)[VAF_filter$ID %in% hi_conf_loci], ], 2, sum))
#tum_infiltration_est_median = data.frame(apply(VAF[VAF_filter$ID %in% hi_conf_loci, ], 2, median)) #2 copies at 2+0 sites
tum_infiltration_est$PD = row.names(tum_infiltration_est)
colnames(tum_infiltration_est)[1] = c("purity_trunc")
tum_infiltration_est = tum_infiltration_est[, c(2,1)]

#old = manifest
manifest$purity_trunk = NA

for(i in 1:nrow(manifest)){
  if(manifest$case.id[i] == patient & manifest$keep[i] == "Y"){
    manifest$purity_trunk[i] = tum_infiltration_est[tum_infiltration_est$PD == manifest$sample[i], ]$purity_trunc
  }
}

write.table(manifest, '/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', sep = '\t', quote = F, col.names = T, row.names = F)

```

###PD50297 & PD51123

```{r}

#PD50297
patient = "PD50297"
truncal_cluster = 9

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

samples = manifest[manifest$case.id == patient & manifest$battenberg_purity >= 0.4 & !is.na(manifest$battenberg_purity), ]$sample

#get variants in trunk
vars = read.table(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient, '/md_out/', patient, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F) #from original as root unchanged
vars[vars$chr == 1 & vars$pos == 226064434,] #root

cluster_vars = vars[vars$cluster.no == truncal_cluster,]
cluster_vars$ID = paste0(cluster_vars$chr, "_", cluster_vars$pos)

#identify which truncal variants occur at 1+1 sites
dp_files = list.files(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient), "allDirichletProcessInfo_fromMultipleSamplesWithoutChromosomeLosses.txt", full.names = T)

trunc_pre_dup = list()
for(file in dp_files){
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  sample = unlist(strsplit(basename(file), "_allDirichletProcessInfo_fromMultipleSamplesWithoutChromosomeLosses.txt"))
  data$ID = paste0(data$chr, "_", data$pos)
  trunc_pre_dup[[sample]] = data[data$ID %in% cluster_vars$ID & data$subclonal.CN == 2 & is.na(data$nMaj2) & data$nMin1 == 1 & data$nMaj1 == 1,]$ID
}

hi_conf_loci = Reduce(intersect, trunc_pre_dup)
hi_conf_loci = paste0("chr", hi_conf_loci)
length(hi_conf_loci) #385

#read in VAF dataframe
VAF = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/", patient, "_VAF_final_merged_snps.txt"), header = T, sep = "\t", stringsAsFactors = F)

VAF_filter = VAF
VAF_filter$ID = paste0(unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 1)), "_", unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 2)))

mean_no_outliers <- function(x, remove_na = TRUE) {
  qnt <- quantile(x, probs = c(.25, .75), na.rm = remove_na)
  val <- 1.5 * IQR(x, na.rm = remove_na)
  y <- x
  y[x < (qnt[1] - val)] <- NA
  y[x > (qnt[2] + val)] <- NA
  z = mean(y, na.rm = remove_na)
  return(z)
}

tum_infiltration_est = data.frame(apply(VAF[VAF_filter$ID %in% hi_conf_loci, ], 2, function(x) 2 * mean_no_outliers(x))) #x2 at 1+1 sites for cellularity
tum_infiltration_est$PD = row.names(tum_infiltration_est)
colnames(tum_infiltration_est)[1] = c("purity_trunc")
tum_infiltration_est = tum_infiltration_est[, c(2,1)]

for(i in 1:nrow(manifest)){
  if(manifest$case.id[i] == patient & manifest$keep[i] == "Y"){
    manifest$purity_trunk[i] = tum_infiltration_est[tum_infiltration_est$PD == manifest$sample[i], ]$purity_trunc
  }
}

write.table(manifest, '/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', sep = '\t', quote = F, col.names = T, row.names = F)

#PD51123
patient = "PD51123"
truncal_cluster = 2

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

samples = manifest[manifest$case.id == patient & manifest$battenberg_purity >= 0.4 & !is.na(manifest$battenberg_purity), ]$sample

#get variants in trunk
vars = read.table(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient, '/md_out/', patient, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)
cluster_vars = vars[vars$cluster.no == truncal_cluster,]
cluster_vars$ID = paste0(cluster_vars$chr, "_", cluster_vars$pos)

#identify which truncal variants occur at 1+1 sites
dp_files = list.files(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient), "allDirichletProcessInfo_fromMultipleSamplesWithoutChromosomeLosses.txt", full.names = T)

trunc_pre_dup = list()
for(file in dp_files){
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  sample = unlist(strsplit(basename(file), "_allDirichletProcessInfo_fromMultipleSamplesWithoutChromosomeLosses.txt"))
  data$ID = paste0(data$chr, "_", data$pos)
  trunc_pre_dup[[sample]] = data[data$ID %in% cluster_vars$ID & data$subclonal.CN == 2 & is.na(data$nMaj2) & data$nMin1 == 1 & data$nMaj1 == 1,]$ID
}

hi_conf_loci = Reduce(intersect, trunc_pre_dup)
hi_conf_loci = paste0("chr", hi_conf_loci)
length(hi_conf_loci) #260

#read in VAF dataframe
VAF = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/", patient, "_VAF_final_merged_snps.txt"), header = T, sep = "\t", stringsAsFactors = F)

VAF_filter = VAF
VAF_filter$ID = paste0(unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 1)), "_", unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 2)))

mean_no_outliers <- function(x, remove_na = TRUE) {
  qnt <- quantile(x, probs = c(.25, .75), na.rm = remove_na)
  val <- 1.5 * IQR(x, na.rm = remove_na)
  y <- x
  y[x < (qnt[1] - val)] <- NA
  y[x > (qnt[2] + val)] <- NA
  z = mean(y, na.rm = remove_na)
  return(z)
}

tum_infiltration_est = data.frame(apply(VAF[VAF_filter$ID %in% hi_conf_loci, ], 2, function(x) 2 * mean_no_outliers(x))) #x2 at 1+1 sites for cellularity
tum_infiltration_est$PD = row.names(tum_infiltration_est)
colnames(tum_infiltration_est)[1] = c("purity_trunc")
tum_infiltration_est = tum_infiltration_est[, c(2,1)]

for(i in 1:nrow(manifest)){
  if(manifest$case.id[i] == patient & manifest$keep[i] == "Y"){
    manifest$purity_trunk[i] = tum_infiltration_est[tum_infiltration_est$PD == manifest$sample[i], ]$purity_trunc
  }
}

write.table(manifest, '/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', sep = '\t', quote = F, col.names = T, row.names = F)

```

###EDF12 - Comparison of Battenberg sensitivity with truncal VAF estimates

Sanity check against the Battenberg estimates for samples with >=10% tumour involvement

```{r}

library(ggplot2)
library(ggpubr)

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

bulk = manifest[!is.na(manifest$battenberg_purity), ]

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/pure_tumour_involvement_method_comparison_20221214.pdf", height = 5, width = 5, useDingbats = F)
ggplot(bulk, mapping = aes(x = battenberg_purity, y = purity_trunk)) + 
  geom_point() + 
  geom_smooth(method='lm', col = "red") + 
  stat_cor(method="pearson") + 
  theme_bw() + 
  coord_cartesian(xlim = c(0,1), ylim = c(0,1)) + 
  xlab("Purity (Battenberg)") + 
  ylab("Purity (curated truncal sub VAF method)") + 
  ggtitle("Comparison of method for estimating tumour infiltration") +
  geom_abline(slope=1, intercept=0, lty = 2)
dev.off()

```

###EDF13 - Estimate sensitivity of method using simulations

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tumour_detection_sensitivity_analysis

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"
bsub -q yesterday -Is -n 4 -R 'select[mem>=80000] rusage[mem=80000] span[hosts=1]' -M80000 R

```

Purity estimate uses somewhere between 50-400 subs from the trunk

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tumour_detection_sensitivity_analysis")

mean_no_outliers <- function(x, remove_na = TRUE) {
  qnt <- quantile(x, probs = c(.25, .75), na.rm = remove_na)
  val <- 1.5 * IQR(x, na.rm = remove_na)
  y <- x
  y[x < (qnt[1] - val)] <- NA
  y[x > (qnt[2] + val)] <- NA
  z = mean(y, na.rm = remove_na)
  return(z)
}

subs_range <- seq(50, 400, 50)
depth_range <- seq(10, 200, 10)
tumour_inf_range <- seq(0, 0.1, 0.01)
nRand = 1000

res_arr = array(dim = c(length(subs_range), length(depth_range), length(tumour_inf_range), nRand), 
                dimnames = list(n_subs = subs_range, depth = depth_range, purity = tumour_inf_range, sim = 1:nRand))

set.seed(42)

for(n_subs in subs_range){
  
  print(n_subs)
  
  for(depth in depth_range){
    
    for(purity in tumour_inf_range){
      
      for(sim in 1:nRand){
        
        #generate simulated data
        NR = rpois(n_subs, depth)
        NV = sapply(NR, function(x) rbinom(1, x, purity / 2)) #create situation like PD50297 and PD51123
  
        NR[NR == 0] = 1 #should 0 be generated for the distribution
  
        VAF = NV / NR
        
        res_arr[paste0(n_subs), paste0(depth), paste0(purity), paste0(sim)] = mean_no_outliers(VAF) * 2
        
      }
      
    }
    
  }
  
}

saveRDS(res_arr, "tumour_infiltration_sensitivity_sim.rds")

#res_arr = readRDS("tumour_infiltration_sensitivity_sim.rds")
#three patients use different numbers of mutations from trunk to derive tumour infiltration estimate - ~50, 250 and 400

sensitivity_calc = data.frame(matrix(nrow = 660, ncol = 3, dimnames = list(1:660, c("n_subs", "depth", "purity"))))
sensitivity_calc$n_subs = c(rep(50, 220), rep(250, 220), rep(400, 220))
sensitivity_calc$depth = rep(depth_range, 33)
sensitivity_calc$purity = rep(sort(rep(tumour_inf_range, 20)), 3)
sensitivity_calc$sens = as.vector(apply(sensitivity_calc[,1:3], 1, function(x) sum(as.vector(res_arr[as.character(x[1]), as.character(x[2]), as.character(x[3]),]) > 0) / length(as.vector(res_arr[as.character(x[1]), as.character(x[2]), as.character(x[3]),]))))

sensitivity_calc[sensitivity_calc$n_subs == 50 & sensitivity_calc$depth == 30 & sensitivity_calc$purity == 0.05,]
sensitivity_calc[sensitivity_calc$n_subs == 50 & sensitivity_calc$depth == 30,]
sensitivity_calc[sensitivity_calc$n_subs == 50 & sensitivity_calc$depth == 60,]

library(ggplot2)
library(cowplot)

p1 = ggplot(sensitivity_calc[sensitivity_calc$n_subs == 50 & sensitivity_calc$depth <= 150,]) + 
  geom_point(mapping = aes(x = purity, y = sens)) +
  geom_line(mapping = aes(x = purity, y = sens)) +
  facet_wrap(. ~ depth, ncol = 5) +
  theme_bw() +
  ylab("Fraction of simulations where tumour infiltration was detected") +
  xlab("Purity") +
  ggtitle("Using 50 truncal mutations from tumour") +
  scale_x_continuous(breaks = seq(0, 0.1, by = 0.05))
p2 = ggplot(sensitivity_calc[sensitivity_calc$n_subs == 250 & sensitivity_calc$depth <= 150,]) + 
  geom_point(mapping = aes(x = purity, y = sens)) +
  geom_line(mapping = aes(x = purity, y = sens)) +
  facet_wrap(. ~ depth, ncol = 5) +
  theme_bw() +
  ylab("Fraction of simulations where tumour infiltration was detected") +
  xlab("Purity") +
  ggtitle("Using 250 truncal mutations from tumour") +
  scale_x_continuous(breaks = seq(0, 0.1, by = 0.05))
p3 = ggplot(sensitivity_calc[sensitivity_calc$n_subs == 400 & sensitivity_calc$depth <= 150,]) + 
  geom_point(mapping = aes(x = purity, y = sens)) +
  geom_line(mapping = aes(x = purity, y = sens)) +
  facet_wrap(. ~ depth, ncol = 5) +
  theme_bw() +
  ylab("Fraction of simulations where tumour infiltration was detected") +
  xlab("Purity") +
  ggtitle("Using 400 truncal mutations from tumour") +
  scale_x_continuous(breaks = seq(0, 0.1, by = 0.05))

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/truncal_mutation_sensitivity_simulation_20221221.pdf", height = 10, width = 6)
plot_grid(p1, p2, p3, align = "v", ncol = 1)
dev.off()

```

####Low-level tumour infiltration impact

Assuming a mutation lies within a 1+1 region, was is the probability low level tumour infiltration will gather sufficient variant support in the normal sample?

```{r}

min_nv = 4
mean_cov = 30
purity = 0.02

ppois(min_nv - 1, mean_cov * (purity /2), lower.tail = F)

```

###EDF14 - Estimate false positive rate for calling tumour root variants using simulations

```{r}

vaf_cutoff = 0.1
mean_cov = 30
purity = 0.03

ppois(ceiling(mean_cov * vaf_cutoff) - 1, (mean_cov * (purity / 2)), lower.tail = F) #greater than 2
ppois(floor(mean_cov * vaf_cutoff), (mean_cov * (purity / 2)), lower.tail = F) #greater than 3


#fp_calc[fp_calc$vaf_cutoff == 0.1 & fp_calc$mean_cov == 30 & fp_calc$purity == 0.01,]

```

####Method 1

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tumour_detection_sensitivity_analysis")

depth_range <- seq(10, 200, 10)
tumour_inf_range <- seq(0, 0.1, 0.01)
vaf_cutoff_range = seq(0, 0.3, 0.05)
#mean_cov = 100
#purity = 0.05
#vaf_cutoff = 0.1

fp_arr = array(dim = c(length(depth_range), length(tumour_inf_range), length(vaf_cutoff_range)), 
                dimnames = list(mean_cov = depth_range, purity = tumour_inf_range, vaf_cutoff = vaf_cutoff_range))

set.seed(42)
for(mean_cov in depth_range){
    
  print(mean_cov)
    
  for(purity in tumour_inf_range){
      
    for(vaf_cutoff in vaf_cutoff_range){
    
    fp_arr[paste0(mean_cov), paste0(purity), paste0(vaf_cutoff)] = mean(unlist(lapply(rpois(n = 100000, lambda = mean_cov), function(x) pbinom(q = ceiling(x * vaf_cutoff) - 1, size = x, p = (purity / 2), lower.tail = F))))
      
    }
      
  }
    
}
  
saveRDS(fp_arr, "fp_tum_root_calling_sim.rds")
#fp_arr = readRDS("fp_tum_root_calling_sim.rds")

#almost all LCM samples have coverage 10-50
#6 * 11 * 7 #462
fp_calc = data.frame(matrix(nrow = 462, ncol = 3, dimnames = list(1:462, c("mean_cov", "purity", "vaf_cutoff"))))
fp_calc$mean_cov = c(rep(10, 77), rep(20, 77), rep(30, 77), rep(40, 77), rep(50, 77), rep(60, 77))
fp_calc$purity = rep(sort(rep(tumour_inf_range, 7)), 6)
fp_calc$vaf_cutoff = rep(vaf_cutoff_range, 66)
fp_calc$fp = apply(fp_calc[, 1:3], 1, function(x) fp_arr[as.character(x[1]), as.character(x[2]), as.character(x[3])])


p1 = ggplot(fp_calc[fp_calc$mean_cov == 30 & fp_calc$vaf_cutoff > 0,]) + 
  geom_point(mapping = aes(x = purity, y = fp)) +
  geom_line(mapping = aes(x = purity, y = fp)) +
  facet_wrap(. ~ vaf_cutoff) +
  theme_bw() +
  ylab("Sensitivity to detect variants from tumour infiltration") +
  xlab("Purity") +
  ggtitle("Average 30x depth of sequencing") +
  scale_x_continuous(breaks = seq(0, 0.1, by = 0.05)) +
  ylim(0, 1)
p2 = ggplot(fp_calc[fp_calc$mean_cov == 50 & fp_calc$vaf_cutoff > 0,]) + 
  geom_point(mapping = aes(x = purity, y = fp)) +
  geom_line(mapping = aes(x = purity, y = fp)) +
  facet_wrap(. ~ vaf_cutoff) +
  theme_bw() +
  ylab("Sensitivity to detect variants from tumour infiltration") +
  xlab("Purity") +
  ggtitle("Average 50x depth of sequencing") +
  scale_x_continuous(breaks = seq(0, 0.1, by = 0.05)) +
  ylim(0, 1)

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/fp_shared_mutation_rate_simulation_20221214.pdf", height = 6.6, width = 6)
plot_grid(p1, p2, align = "v", ncol = 1)
dev.off()

fp_calc[fp_calc$mean_cov == 30 & fp_calc$vaf_cutoff == 0.1,]
fp_calc[fp_calc$mean_cov == 50 & fp_calc$vaf_cutoff == 0.1,]
fp_calc[fp_calc$mean_cov == 60 & fp_calc$vaf_cutoff == 0.1,]
```

####Method 2

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tumour_detection_sensitivity_analysis")

depth_range <- seq(10, 200, 10)
tumour_inf_range <- seq(0, 0.1, 0.01)
vaf_cutoff_range = seq(0, 0.3, 0.05)
#mean_cov = 100
#purity = 0.05
#vaf_cutoff = 0.1

fp_arr = array(dim = c(length(depth_range), length(tumour_inf_range), length(vaf_cutoff_range)), 
                dimnames = list(mean_cov = depth_range, purity = tumour_inf_range, vaf_cutoff = vaf_cutoff_range))

#set.seed(42)
for(mean_cov in depth_range){
    
  print(mean_cov)
    
  for(purity in tumour_inf_range){
      
    for(vaf_cutoff in vaf_cutoff_range){
    
    fp_arr[paste0(mean_cov), paste0(purity), paste0(vaf_cutoff)] = ppois(floor(mean_cov * vaf_cutoff), (mean_cov * (purity / 2)), lower.tail = F)
      
    }
      
  }
    
}
  
saveRDS(fp_arr, "fp_tum_root_calling_sim_2_20230109.rds")
#fp_arr = readRDS("fp_tum_root_calling_sim_2_20230109.rds")

#almost all LCM samples have coverage 10-50
#6 * 11 * 7 #462
fp_calc = data.frame(matrix(nrow = 462, ncol = 3, dimnames = list(1:462, c("mean_cov", "purity", "vaf_cutoff"))))
fp_calc$mean_cov = c(rep(10, 77), rep(20, 77), rep(30, 77), rep(40, 77), rep(50, 77), rep(60, 77))
fp_calc$purity = rep(sort(rep(tumour_inf_range, 7)), 6)
fp_calc$vaf_cutoff = rep(vaf_cutoff_range, 66)
fp_calc$fp = apply(fp_calc[, 1:3], 1, function(x) fp_arr[as.character(x[1]), as.character(x[2]), as.character(x[3])])

library(ggplot2)

p1 = ggplot(fp_calc[fp_calc$mean_cov == 30 & fp_calc$vaf_cutoff > 0,]) + 
  geom_point(mapping = aes(x = purity, y = fp)) +
  geom_line(mapping = aes(x = purity, y = fp)) +
  facet_wrap(. ~ vaf_cutoff) +
  theme_bw() +
  ylab("Sensitivity to detect variants from tumour infiltration") +
  xlab("Purity") +
  ggtitle("Average 30x depth of sequencing") +
  scale_x_continuous(breaks = seq(0, 0.1, by = 0.05)) +
  ylim(0, 1)
p2 = ggplot(fp_calc[fp_calc$mean_cov == 60 & fp_calc$vaf_cutoff > 0,]) + 
  geom_point(mapping = aes(x = purity, y = fp)) +
  geom_line(mapping = aes(x = purity, y = fp)) +
  facet_wrap(. ~ vaf_cutoff) +
  theme_bw() +
  ylab("Sensitivity to detect variants from tumour infiltration") +
  xlab("Purity") +
  ggtitle("Average 60x depth of sequencing") +
  scale_x_continuous(breaks = seq(0, 0.1, by = 0.05)) +
  ylim(0, 1)

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/fp_shared_mutation_rate_simulation_20230109.pdf", height = 6.6, width = 6)
plot_grid(p1, p2, align = "v", ncol = 1)
dev.off()

fp_calc[fp_calc$mean_cov == 30 & fp_calc$vaf_cutoff == 0.1,]
fp_calc[fp_calc$mean_cov == 60 & fp_calc$vaf_cutoff == 0.1,]
fp_calc[fp_calc$mean_cov == 10 & fp_calc$vaf_cutoff == 0.05,]
fp_calc[fp_calc$mean_cov == 30 & fp_calc$vaf_cutoff == 0.05,]
fp_calc[fp_calc$mean_cov == 60 & fp_calc$vaf_cutoff == 0.05,]

```

##Burden

###Subs - all

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/subs/07_indel")

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

manifest$sub_burden = 0
manifest$autosomal_sub_burden = 0
manifest$median_sub_VAF = 0
manifest$sub_burden_tier1 = 0
manifest$median_sub_VAF_tier1 = 0

sub_files = list.files(".", ".standard.gnomAD.AF.0.001.PON.read.direction.low_cov.bq.mq.binom.bbinom.5bp.indel.filtered.tier.annot.txt")

for(file in sub_files){
  
  sample = unlist(strsplit(file, ".standard.gnomAD.AF.0.001.PON.read.direction.low_cov.bq.mq.binom.bbinom.5bp.indel.filtered.tier.annot.txt"))
  data = read.table(file, header = T, sep = "\t", stringsAsFactors = F)
  data_tier1 = data[data$tier == 1,]
  
  manifest[manifest$sample == sample, ]$sub_burden = nrow(data)
  manifest[manifest$sample == sample, ]$autosomal_sub_burden = nrow(data[data$Chr %in% paste0("chr", 1:22),])
  manifest[manifest$sample == sample, ]$median_sub_VAF = median(data[data$Chr %in% paste0("chr", 1:22),]$VAF)
  manifest[manifest$sample == sample, ]$sub_burden_tier1 = nrow(data_tier1)
  manifest[manifest$sample == sample, ]$median_sub_VAF_tier1 = median(data_tier1[data_tier1$Chr %in% paste0("chr", 1:22),]$VAF)
  
}

manifest[manifest$keep == "N",]$sub_burden = NA
manifest[manifest$keep == "N",]$autosomal_sub_burden = NA
manifest[manifest$keep == "N",]$median_sub_VAF = NA
manifest[manifest$keep == "N",]$sub_burden_tier1 = NA
manifest[manifest$keep == "N",]$median_sub_VAF_tier1 = NA

write.table(manifest, '/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', sep = '\t', quote = F, col.names = T, row.names = F)

```


###Subs - normal brain (EDF1)

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/subs/07_indel")

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

normal_brain = manifest[manifest$keep == "Y" & 
                          manifest$purity_trunk < 0.01 &
                          manifest$CNS == "Y",]
normal_brain_lcm = normal_brain[normal_brain$experimental_arm == "WGS_LCM",]

median(normal_brain_lcm$sub_burden_tier1) #7
range(normal_brain_lcm$sub_burden_tier1) #2 91
median(normal_brain_lcm$median_sub_VAF_tier1) #0.2083333
range(normal_brain_lcm$median_sub_VAF_tier1) # 0.09210526 0.51851852

all_subs = read.table("/lustre/scratch119/realdata/mdt1/team274/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/subs/07_indel/all_sample_subs_distribution_annotated_20221201.txt", header = T, sep = "\t", stringsAsFactors = F)

private_norm_brain_subs_flt = all_subs[all_subs$Sample %in% normal_brain_lcm$sample &
                              all_subs$mut_status %in% c("sample", "bulk") &
                              all_subs$pure_tumour_mut == F &
                              all_subs$tier == 1,]

normal_brain_lcm$private_sub_burden = NA
normal_brain_lcm$private_sub_vaf = NA
for(i in 1:nrow(normal_brain_lcm)){
  
  normal_brain_lcm$private_sub_burden[i] = nrow(private_norm_brain_subs_flt[private_norm_brain_subs_flt$Sample == normal_brain_lcm$sample[i],])
  normal_brain_lcm$private_sub_vaf[i] = median(private_norm_brain_subs_flt[private_norm_brain_subs_flt$Sample == normal_brain_lcm$sample[i] & private_norm_brain_subs_flt$Chr %in% paste0("chr", c(1:22)),]$VAF)
  
}

nrow(normal_brain_lcm[normal_brain_lcm$private_sub_burden <= 5,]) #294
nrow(normal_brain_lcm) #321
nrow(normal_brain_lcm[normal_brain_lcm$private_sub_burden <= 5,]) / nrow(normal_brain_lcm) #0.9158879

median_private_burden = aggregate(private_sub_burden ~ bulk_tissue, normal_brain_lcm, median)
median_private_burden = median_private_burden[order(median_private_burden$private_sub_burden, decreasing = F),]
table(normal_brain_lcm$bulk_tissue)

normal_brain_lcm$bulk_tissue = factor(normal_brain_lcm$bulk_tissue, levels = median_private_burden$bulk_tissue)

normal_brain_lcm[normal_brain_lcm$private_sub_burden >= 30,]

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/private_subs_per_biopsy_20221206.pdf", height = 5, width = 5, useDingbats = F)
ggplot(normal_brain_lcm) +
  geom_quasirandom(mapping = aes(x = bulk_tissue, y = private_sub_burden, col = case.id)) +
  geom_boxplot(mapping = aes(x = bulk_tissue, y = private_sub_burden), fill = NA, outlier.shape = NA) +
  theme_bw() +
  xlab("") +
  ylab("Private substitution count (tier 1)") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_color_manual(values = c("dodgerblue", "grey20", "#F9B233"))
dev.off()

```

###Indels

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/indels/04_bbinom")

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

manifest$indel_burden = 0
manifest$autosomal_indel_burden = 0
manifest$median_indel_VAF = 0

indel_files = list.files(".", ".pindel.HP.9.read.direction.germline.beta.binom.unk.amb.min.depth.filtered.txt")

for(file in indel_files){
  
  sample = unlist(strsplit(file, ".pindel.HP.9.read.direction.germline.beta.binom.unk.amb.min.depth.filtered.txt"))
  
  if(file.exists(file)){
    
    data = read.table(file, header = T, sep = "\t", stringsAsFactors = F)
  
    manifest[manifest$sample == sample, ]$indel_burden = nrow(data)
    manifest[manifest$sample == sample, ]$autosomal_indel_burden = nrow(data[data$Chr %in% paste0("chr", 1:22),])
    manifest[manifest$sample == sample, ]$median_indel_VAF = median(data[data$Chr %in% paste0("chr", 1:22),]$VAF)
    
  } else {
    
    manifest[manifest$sample == sample, ]$indel_burden = 0
    manifest[manifest$sample == sample, ]$autosomal_indel_burden = 0
    manifest[manifest$sample == sample, ]$median_indel_VAF = 0
    
  }
  
}

manifest[manifest$keep == "N",]$indel_burden = NA
manifest[manifest$keep == "N",]$autosomal_indel_burden = NA
manifest[manifest$keep == "N",]$median_indel_VAF = NA

ggplot(manifest) +
  geom_histogram(mapping = aes(x = indel_burden / sub_burden, fill = case.id), bins = 100)

ggplot(manifest) +
 geom_point(mapping = aes(x = sub_burden, y = indel_burden, col = case.id))

write.table(manifest, '/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', sep = '\t', quote = F, col.names = T, row.names = F)

```

##Fig. 1c - normal LCM WGS substitution count vs median VAF

```{r}

library(ggplot2)

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

norm_lcm_manifest = manifest[manifest$keep == "Y" &
                              manifest$experimental_arm == "WGS_LCM" &
                              manifest$purity_trunk < 0.01,]

#only use tier 1 mutations for CNS
norm_lcm_manifest$sub_burden_plot = norm_lcm_manifest$sub_burden
norm_lcm_manifest$sub_burden_plot[norm_lcm_manifest$CNS == "Y"] = norm_lcm_manifest$sub_burden_tier1[norm_lcm_manifest$CNS == "Y"]
norm_lcm_manifest$sub_median_vaf_plot = norm_lcm_manifest$median_sub_VAF
norm_lcm_manifest$sub_median_vaf_plot[norm_lcm_manifest$CNS == "Y"] = norm_lcm_manifest$median_sub_VAF_tier1[norm_lcm_manifest$CNS == "Y"]

norm_lcm_manifest$col_annot = NA
norm_lcm_manifest[norm_lcm_manifest$CNS == "Y", ]$col_annot = "CNS"
norm_lcm_manifest[norm_lcm_manifest$bulk_tissue %in% c("Large intestine", "Appendix") & norm_lcm_manifest$histo == "Crypt",]$col_annot = "Crypt(s), large intestine"
norm_lcm_manifest[norm_lcm_manifest$bulk_tissue == "Small intestine" & norm_lcm_manifest$histo %in% c("Crypt", "Epithelium"),]$col_annot = "Crypt(s), small intestine"
norm_lcm_manifest[norm_lcm_manifest$histo == "Epidermis",]$col_annot = "Epidermis, skin"
norm_lcm_manifest[is.na(norm_lcm_manifest$col_annot),]$col_annot = "Other"

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/non_tumour_lcm_sample_mut_vs_vaf.pdf", height = 4, width = 6, useDingbats = F)
ggplot(norm_lcm_manifest) +
  geom_point(mapping = aes(x = sub_median_vaf_plot, y = sub_burden_plot, col = col_annot), alpha = 0.8, size = 3, stroke = 0) +
  scale_y_log10() +
  theme_bw() +
  ylab("Substitution count") +
  xlab("Median VAF") +
  scale_color_manual(values = c("black", "royalblue", "skyblue", "brown2", "#999999"))
dev.off()

```

##Fig. 1d - SBS mutational signature extraction

###Functions

```{r}

library(lsa)
library(VariantAnnotation)
library("GenomicRanges")
library("Rsamtools")
library("MASS")

genomeFile_38 <- "/lustre/scratch119/casm/team78pipelines/reference/human/GRCh38_full_analysis_set_plus_decoy_hla/genome.fa"

count_trinuc_context_38 <- function(input){
  mutations = data.frame(matrix(unlist(strsplit(input, "_")), ncol = 4, byrow= T))
  names(mutations) = c("chr", "pos", "ref", "mut")
  mutations = mutations[(mutations$ref %in% c("A","C","G","T")) & (mutations$mut %in% c("A","C","G","T")) & mutations$chr %in% paste0("chr", c(1:22,"X","Y")),]
  mutations$trinuc_ref = as.vector(scanFa(genomeFile_38, GRanges(mutations$chr, IRanges(as.numeric(mutations$pos)-1,
                                                                                     as.numeric(mutations$pos)+1))))

  # 2. Annotating the mutation from the pyrimidine base
  ntcomp = c(T="A",G="C",C="G",A="T")
  mutations$sub = paste(mutations$ref,mutations$mut,sep=">")
  mutations$trinuc_ref_py = mutations$trinuc_ref
  for (j in 1:nrow(mutations)) {
    if (mutations$ref[j] %in% c("A","G")) { # Purine base
      mutations$sub[j] = paste(ntcomp[mutations$ref[j]],ntcomp[mutations$mut[j]],sep=">")
      mutations$trinuc_ref_py[j] = paste(ntcomp[rev(strsplit(mutations$trinuc_ref[j],split="")[[1]])],collapse="")
    }
  }
  freqs = table(paste(mutations$sub,paste(substr(mutations$trinuc_ref_py,1,1),substr(mutations$trinuc_ref_py,3,3),sep="-"),sep=","))
  sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
  ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
  full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")
  freqs_full = freqs[full_vec]; freqs_full[is.na(freqs_full)] = 0; names(freqs_full) = full_vec
  
  #3. plot
  xstr = paste(substr(full_vec,5,5), substr(full_vec,1,1), substr(full_vec,7,7), sep="")

  colvec = rep(c("dodgerblue","black","red","grey70","olivedrab3","plum2"),each=16)
  y = freqs_full; maxy = max(y)
  h = barplot(y, las=2, col=colvec, border=NA, ylim=c(0,maxy*1.5), space=1, cex.names=0.6, names.arg=xstr, ylab="Number of mutations")
  for (j in 1:length(sub_vec)) {
    xpos = h[c((j-1)*16+1,j*16)]
    rect(xpos[1]-0.5, maxy*1.2, xpos[2]+0.5, maxy*1.3, border=NA, col=colvec[j*16])
    text(x=mean(xpos), y=maxy*1.3, pos=3, label=sub_vec[j])
  } 
  title(sub = paste0("filtered subs = ", nrow(mutations)))
  
  mut.props = freqs_full / sum(freqs_full)
  
  return(mut.props)
}

```

```{r}

count_muts_96 = function(mutations, genomeFile = "/lustre/scratch119/casm/team78pipelines/reference/human/GRCh38_full_analysis_set_plus_decoy_hla/genome.fa"){
  library("GenomicRanges")
  library("Rsamtools")
  library("MASS")
  colnames(mutations) = c("chr","pos","ref","mut")
  mutations = mutations[(mutations$ref %in% c("A","C","G","T")) & (mutations$mut %in% c("A","C","G","T")) & mutations$chr %in% paste0("chr", c(1:22,"X","Y")),]
  mutations$trinuc_ref = as.vector(scanFa(genomeFile, GRanges(mutations$chr, IRanges(as.numeric(mutations$pos)-1,
                                                                                     as.numeric(mutations$pos)+1))))
 
  # 2. Annotating the mutation from the pyrimidine base
  ntcomp = c(T="A",G="C",C="G",A="T")
  mutations$sub = paste(mutations$ref,mutations$mut,sep=">")
  mutations$trinuc_ref_py = mutations$trinuc_ref
  for (j in 1:nrow(mutations)) {
    if (mutations$ref[j] %in% c("A","G")) { # Purine base
      mutations$sub[j] = paste(ntcomp[mutations$ref[j]],ntcomp[mutations$mut[j]],sep=">")
      mutations$trinuc_ref_py[j] = paste(ntcomp[rev(strsplit(mutations$trinuc_ref[j],split="")[[1]])],collapse="")
    }
  }
  freqs = table(paste(mutations$sub,paste(substr(mutations$trinuc_ref_py,1,1),substr(mutations$trinuc_ref_py,3,3),sep="-"),sep=","))
  sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
  ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
  full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")
  freqs_full = freqs[full_vec]; freqs_full[is.na(freqs_full)] = 0; names(freqs_full) = full_vec
  return(freqs_full)
}

```

###Input

Per biopsy:
- Epidermis, skin
- Crypt(s), large intestine
- Crypt(s), small intestine

Otherwise aggregate normal tissues together. Run tumour per mutation cluster.

```{r}

#get all subs
setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/subs/07_indel")

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
manifest_flt = manifest[manifest$keep == "Y",]

all_subs = data.frame()
for(sample in manifest_flt$sample){
  print(sample)
  data = read.table(paste0(sample, ".standard.gnomAD.AF.0.001.PON.read.direction.low_cov.bq.mq.binom.bbinom.5bp.indel.filtered.tier.annot.txt"), header = T, sep = '\t', stringsAsFactors = F)
  all_subs = rbind(all_subs, data)
}

#get normal tissue re-annotated
norm_lcm_manifest = manifest[manifest$keep == "Y" &
                              manifest$experimental_arm == "WGS_LCM" &
                              manifest$purity_trunk < 0.01,]

#only use tier 1 mutations for brain
norm_lcm_manifest$sub_burden_plot = norm_lcm_manifest$sub_burden
norm_lcm_manifest$sub_burden_plot[norm_lcm_manifest$CNS == "Y"] = norm_lcm_manifest$sub_burden_tier1[norm_lcm_manifest$CNS == "Y"]
norm_lcm_manifest$sub_median_vaf_plot = norm_lcm_manifest$median_sub_VAF
norm_lcm_manifest$sub_median_vaf_plot[norm_lcm_manifest$CNS == "Y"] = norm_lcm_manifest$median_sub_VAF_tier1[norm_lcm_manifest$CNS == "Y"]

table(norm_lcm_manifest[norm_lcm_manifest$CNS == "Y",]$histo)

#Relabel metadata
norm_lcm_manifest[norm_lcm_manifest$histo == "Tumour",]$histo = "Thalamus"
norm_lcm_manifest[norm_lcm_manifest$histo == "Anterior horn",]$histo = "Grey matter, spinal cord"
norm_lcm_manifest[norm_lcm_manifest$histo == "Posterior horn",]$histo = "Grey matter, spinal cord" #fixed 5/1/23 from WM
norm_lcm_manifest[norm_lcm_manifest$histo %in% c("Convoluted tubules", "Medulla tubule", "Medullary ray tubule", "Tubule"),]$histo = "Renal tubules"
norm_lcm_manifest[norm_lcm_manifest$histo == "Dentate gyrus",]$histo = "Grey matter, cerebrum"
norm_lcm_manifest[norm_lcm_manifest$histo == "Pyramidal layer",]$histo = "Grey matter, cerebrum"
norm_lcm_manifest[norm_lcm_manifest$histo == "Epithelium",]$histo = "Crypt(s), small intestine"
norm_lcm_manifest[norm_lcm_manifest$histo %in% c("Zona fasciculata", "Zona glomerulosa", "Zona reticularis"),]$histo = "Cortex, adrenal gland"
norm_lcm_manifest[norm_lcm_manifest$histo == "Grey matter" & norm_lcm_manifest$bulk_tissue != "Spinal cord",]$histo = "Grey matter, cerebrum"
norm_lcm_manifest[norm_lcm_manifest$histo == "Grey matter" & norm_lcm_manifest$bulk_tissue == "Spinal cord",]$histo = "Grey matter, spinal cord"
norm_lcm_manifest[norm_lcm_manifest$histo == "White matter" & norm_lcm_manifest$bulk_tissue != "Spinal cord",]$histo = "White matter, cerebrum"
norm_lcm_manifest[norm_lcm_manifest$histo == "White matter" & norm_lcm_manifest$bulk_tissue == "Spinal cord",]$histo = "White matter, spinal cord"
norm_lcm_manifest[norm_lcm_manifest$histo == "Crypt" & norm_lcm_manifest$bulk_tissue == "Small intestine",]$histo = "Crypt(s), large intestine"
norm_lcm_manifest[norm_lcm_manifest$histo == "Crypt" & norm_lcm_manifest$bulk_tissue %in% c("Large intestine", "Appendix"),]$histo = "Crypt(s), large intestine"
norm_lcm_manifest[norm_lcm_manifest$histo == "Medulla" & norm_lcm_manifest$bulk_tissue == "Adrenal gland",]$histo = "Medulla, adrenal gland"
norm_lcm_manifest[norm_lcm_manifest$histo == "Epidermis",]$histo = "Epidermis, skin"
norm_lcm_manifest[norm_lcm_manifest$histo == "Anterior pituitary",]$histo = "Adenohypophysis, pituitary gland"
norm_lcm_manifest[norm_lcm_manifest$histo == "Posterior pituitary",]$histo = "Neurohypophysis, pituitary gland"
norm_lcm_manifest[norm_lcm_manifest$histo == "Ganglion",]$histo = "Ganglion, PNS"
norm_lcm_manifest[norm_lcm_manifest$histo == "Nerve",]$histo = "Nerve, PNS"
norm_lcm_manifest[norm_lcm_manifest$histo == "Glomerulus",]$histo = "Glomerulus, kidney"
norm_lcm_manifest[norm_lcm_manifest$histo == "Hair follicle",]$histo = "Hair follicle, skin"
norm_lcm_manifest[norm_lcm_manifest$histo == "Hepatocytes",]$histo = "Hepatocytes, liver"
norm_lcm_manifest[norm_lcm_manifest$histo == "Lobular parenchyma",]$histo = "Parenchyma, pineal gland"
norm_lcm_manifest[norm_lcm_manifest$histo == "Granular layer",]$histo = "Granular layer, cerebellum"
norm_lcm_manifest[norm_lcm_manifest$histo == "Molecular layer",]$histo = "Molecular layer, cerebellum"
norm_lcm_manifest[norm_lcm_manifest$histo == "Muscularis propria",]$histo = "Smooth muscle, multiple viscera"
norm_lcm_manifest[norm_lcm_manifest$histo == "Myocardium",]$histo = "Myocardium, heart"
norm_lcm_manifest[norm_lcm_manifest$histo == "Purkinje layer",]$histo = "Purkinje layer, cerebellum"
norm_lcm_manifest[norm_lcm_manifest$histo == "Red pulp",]$histo = "Red pulp, spleen"
norm_lcm_manifest[norm_lcm_manifest$histo == "White pulp",]$histo = "White pulp, spleen"
norm_lcm_manifest[norm_lcm_manifest$histo == "Renal tubules",]$histo = "Tubule(s), kidney"
norm_lcm_manifest[grepl("Submucosal gland", norm_lcm_manifest$histo),]$histo = "Submucosal gland(s), Trachea/Bronchi"
norm_lcm_manifest[norm_lcm_manifest$histo == "Thyroid follicle",]$histo = "Follicle(s), thyroid"
norm_lcm_manifest[norm_lcm_manifest$histo == "White matter tract",]$histo = "White matter, cerebellum"

table(norm_lcm_manifest$histo)

norm_lcm_manifest_min = norm_lcm_manifest[norm_lcm_manifest$histo %in% names(table(norm_lcm_manifest$histo))[table(norm_lcm_manifest$histo) >= 5],]

table(norm_lcm_manifest_min$histo)

row.names(norm_lcm_manifest_min) = norm_lcm_manifest_min$sample

subs_flt = all_subs[all_subs$Sample %in% norm_lcm_manifest_min$sample,]
subs_flt$histo = norm_lcm_manifest_min[subs_flt$Sample,]$histo

aggregate(mutID ~ histo, subs_flt, function(x) length(unique(x)))

average_bx_burden = aggregate(sub_burden_plot ~ histo, norm_lcm_manifest_min, mean)
average_bx_burden[order(average_bx_burden$sub_burden_plot, decreasing = T),]

count_trinuc_context_38(unique(subs_flt[subs_flt$histo == "Crypt(s), large intestine",]$mutID))
count_trinuc_context_38(unique(subs_flt[subs_flt$histo == "Crypt(s), small intestine",]$mutID))
count_trinuc_context_38(unique(subs_flt[subs_flt$histo == "Follicle(s), thyroid",]$mutID))
count_trinuc_context_38(unique(subs_flt[subs_flt$histo == "Hepatocytes, liver",]$mutID))
count_trinuc_context_38(unique(subs_flt[subs_flt$histo == "Epidermis, skin",]$mutID))
count_trinuc_context_38(unique(subs_flt[subs_flt$histo == "Hair follicle, skin",]$mutID))
count_trinuc_context_38(unique(subs_flt[subs_flt$histo == "White matter, cerebrum",]$mutID))
count_trinuc_context_38(unique(subs_flt[subs_flt$histo == "Grey matter, cerebrum",]$mutID))
count_trinuc_context_38(unique(subs_flt[subs_flt$histo == "Adenohypophysis, pituitary gland",]$mutID))
count_trinuc_context_38(unique(subs_flt[subs_flt$histo == "Cortex, adrenal gland",]$mutID))
count_trinuc_context_38(unique(subs_flt[subs_flt$histo == "Tubule(s), kidney",]$mutID))
count_trinuc_context_38(unique(subs_flt[subs_flt$histo == "Submucosal gland(s), Trachea/Bronchi",]$mutID))
count_trinuc_context_38(unique(subs_flt[subs_flt$histo == "Nerve, PNS",]$mutID))

```

```{r}

#get tumour clone mutations
##PD50297
patient = "PD50297"
clusters = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/", patient, "/md_out/", patient, "_allClusterassignmentsFromParallelRuns_clusters_reviewed.txt"), header = T, sep = "\t")
clusters$mut_ID = paste(clusters$chr, clusters$pos, sep = "_")
loci = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/", patient, "/", patient, "_loci.txt"), header = F, sep = "\t")
loci$mut_ID = paste0(loci$V1, "_", loci$V2)
loci$mut_ID2 = paste0(loci$V1, "_", loci$V2, "_", loci$V3, "_", loci$V4)
row.names(loci) = loci$mut_ID

clusters$mut_ID2 = loci[clusters$mut_ID,]$mut_ID2
#clusters = clusters[clusters$cluster.no %in% names(table(clusters$cluster.no))[table(clusters$cluster.no) >= 100],]

pd50297_clusters = clusters

##PD51122
patient = "PD51122"
clusters = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/", patient, "/md_out/", patient, "_allClusterassignmentsFromParallelRuns_clusters_reviewed.txt"), header = T, sep = "\t")
clusters$mut_ID = paste(clusters$chr, clusters$pos, sep = "_")
clusters = clusters[!clusters$mut_ID %in% clusters$mut_ID[duplicated(clusters$mut_ID)],]
loci = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/", patient, "/", patient, "_loci.txt"), header = F, sep = "\t")
loci$mut_ID = paste0(loci$V1, "_", loci$V2)
loci = loci[!loci$mut_ID %in% loci$mut_ID[duplicated(loci$mut_ID)],]
loci$mut_ID2 = paste0(loci$V1, "_", loci$V2, "_", loci$V3, "_", loci$V4)
row.names(loci) = loci$mut_ID

clusters$mut_ID2 = loci[clusters$mut_ID,]$mut_ID2
#clusters = clusters[clusters$cluster.no %in% names(table(clusters$cluster.no))[table(clusters$cluster.no) >= 100],]

pd51122_clusters = clusters

##PD51123
patient = "PD51123"
clusters = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/", patient, "/md_out/", patient, "_allClusterassignmentsFromParallelRuns_clusters_reviewed.txt"), header = T, sep = "\t")
clusters$mut_ID = paste(clusters$chr, clusters$pos, sep = "_")
loci = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/", patient, "/", patient, "_loci.txt"), header = F, sep = "\t")
loci$mut_ID = paste0(loci$V1, "_", loci$V2)
loci$mut_ID2 = paste0(loci$V1, "_", loci$V2, "_", loci$V3, "_", loci$V4)
row.names(loci) = loci$mut_ID

clusters$mut_ID2 = loci[clusters$mut_ID,]$mut_ID2
#clusters = clusters[clusters$cluster.no %in% names(table(clusters$cluster.no))[table(clusters$cluster.no) >= 100],]

pd51123_clusters = clusters

#get biopsy level mutations from most clonal tissues
clonal_bx_subs_flt = subs_flt[subs_flt$histo %in% c("Epidermis, skin", "Crypt(s), large intestine", "Crypt(s), small intestine"),]

#get aggregate calls for other LCM tissues
non_clonal_bx_subs_flt = subs_flt[!subs_flt$histo %in% c("Epidermis, skin", "Crypt(s), large intestine", "Crypt(s), small intestine"),]
non_clonal_bx_subs_flt_uniq = non_clonal_bx_subs_flt[!duplicated(non_clonal_bx_subs_flt[, c("histo", "mutID")]),]

#combined into a single dataframe
input = clonal_bx_subs_flt[, c("Sample", "mutID")]
input = rbind(input, data.frame(Sample = non_clonal_bx_subs_flt_uniq$histo, mutID = non_clonal_bx_subs_flt_uniq$mutID))
input = rbind(input, data.frame(Sample = paste0("PD50297_", pd50297_clusters$cluster.no), mutID = pd50297_clusters$mut_ID2))
input = rbind(input, data.frame(Sample = paste0("PD51122_", pd51122_clusters$cluster.no), mutID = pd51122_clusters$mut_ID2))
input = rbind(input, data.frame(Sample = paste0("PD51123_", pd51123_clusters$cluster.no), mutID = pd51123_clusters$mut_ID2))
input[!grepl("chr", input$mutID),]$mutID = paste0("chr", input[!grepl("chr", input$mutID),]$mutID)

#relabel
temp = input$Sample
temp = gsub(",| ", "_", temp)
temp = gsub("\\(", "", temp)
temp = gsub("\\)", "", temp)
temp = gsub("__", "_", temp)

input$Sample = temp

All_var = cbind(input$Sample, data.frame(matrix(unlist(strsplit(input$mutID, "_")), ncol = 4, byrow = T)))

library(cosmicsig)
colnames(All_var) = c("sample", "chr","pos","ref","mut")

#write.table(All_var, '/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/sig_analysis/mSigHDP/01_Input/brain_autopsy_mSigHDP_input_20221106.txt', col.names = T, row.names = F, sep = '\t', quote = F)
write.table(All_var, '/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/sig_analysis/mSigHDP/01_Input/brain_autopsy_mSigHDP_input_20230105.txt', col.names = T, row.names = F, sep = '\t', quote = F)

```

```{r}

#All_var = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/sig_analysis/mSigHDP/01_Input/brain_autopsy_mSigHDP_input_20221106.txt', header = T, sep = '\t', stringsAsFactors = F)
All_var = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/sig_analysis/mSigHDP/01_Input/brain_autopsy_mSigHDP_input_20230105.txt', header = T, sep = '\t', stringsAsFactors = F)
example_input = read.table("/lustre/scratch117/casm/team294/rs30/Macaque_Signatures-Conrad/cgpvaf_mpboot/18607/Signatures/Input1/output/SBS/Input1.SBS96.all", header = T, sep = "\t", stringsAsFactors = F)

samples=unique(All_var$sample)
mut_count = matrix(0,nrow=length(samples),ncol=96)

#create dataframe of context counts per patient so as not to duplicate the counting of individual subs
for (k in 1:length(samples)){
  mut_count[k,]=count_muts_96(All_var[All_var$sample==samples[k],2:5])
  print(k)
}

mut_count = data.frame(mut_count)
row.names(mut_count) = samples

sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")

names(mut_count) = full_vec

colnames(mut_count) = paste0(substr(colnames(mut_count), 5, 5), "[", substr(colnames(mut_count), 1, 1), ">", substr(colnames(mut_count), 3, 3), "]", substr(colnames(mut_count), 7, 7))
mut_count2 = mut_count[, example_input$MutationType] #reorder to match order in example file
mut_count.t = as.data.frame(t(mut_count2)) 
mut_count.t$MutationType = row.names(mut_count.t)
mut_count.t = mut_count.t[, c(ncol(mut_count.t), 1:(ncol(mut_count.t) - 1))] #reorder columns

#write.table(mut_count.t, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/sig_analysis/sigProfiler/01_Input/sigprofiler_input_20221105.txt", col.names = T, sep = "\t", row.names = F, quote = F)
write.table(mut_count.t, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/sig_analysis/sigProfiler/01_Input/sigprofiler_input_20230105.txt", col.names = T, sep = "\t", row.names = F, quote = F)

```

####Run

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/sig_analysis/sigProfiler/01_Input/

source activate /lustre/scratch119/casm/team294rr/rs30/Anaconda/anaconda2/envs/sigprofiler_env

bsub -J ../02_Output_20230105/sigprofiler_sub -o ../02_Output_20230105/sigprofiler_sub.out -e ../02_Output_20230105/sigprofiler_sub.err -q gpu-basement -gpu - -R 'select[mem>=32000] rusage[mem=32000]' -M32000 -env "all" "python /lustre/scratch119/casm/team294rr/rs30/git_repositories/sigprofiler-scripts/run_SigProfiler_Extractor_matrix.v1.1.py \
/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/sig_analysis/sigProfiler/01_Input/sigprofiler_input_20230105.txt \
-o ../02_Output_20230105 \
-r GRCh38"

```

####Refit

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/sig_analysis/sigProfiler/01_Input/

cp /lustre/scratch119/casm/team294rr/rs30/git_repositories/sigprofiler-scripts/run_SigProfiler_Extractor_matrix.v1.1.py .
mv run_SigProfiler_Extractor_matrix.v1.1.py run_SigProfiler_Extractor_matrix.v1.1_refit_20221106.py #edit to have min and max of 4

source activate /lustre/scratch119/casm/team294rr/rs30/Anaconda/anaconda2/envs/sigprofiler_env

bsub -J ../02_Output_20230105_refit/sigprofiler_sub -o ../02_Output_20230105_refit/sigprofiler_sub.out -e ../02_Output_20230105_refit/sigprofiler_sub.err -q gpu-basement -gpu - -R 'select[mem>=32000] rusage[mem=32000]' -M32000 -env "all" "python run_SigProfiler_Extractor_matrix.v1.1_refit_20221106.py \
/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/sig_analysis/sigProfiler/01_Input/sigprofiler_input_20230105.txt \
-o ../02_Output_20230105_refit \
-r GRCh38"

```

###Plot

```{r}

#sig_activities_old = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/sig_analysis/sigProfiler/02_Output_20221106_refit/SBS96/Suggested_Solution/COSMIC_SBS96_Decomposed_Solution/Activities/COSMIC_SBS96_Activities.txt", header = T, sep = "\t", stringsAsFactors = F)
sig_activities = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/sig_analysis/sigProfiler/02_Output_20230105_refit/SBS96/Suggested_Solution/COSMIC_SBS96_Decomposed_Solution/Activities/COSMIC_SBS96_Activities.txt", header = T, sep = "\t", stringsAsFactors = F)

#sig_activities_old[, 2:10] - sig_activities[, 2:10] #only changes rows 95 and 96
#sig_activities[95:96,] #spinal cord unsurprisingly
#rowSums(sig_activities[95:96,2:10]) #both under 100 so wouldn't be included in figure anyway

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

sig_activities$tissue = NA

for(i in 1:nrow(sig_activities)){
  if(grepl("lo", sig_activities$Samples[i]) & sig_activities$Samples[i] != "Glomerulus_kidney"){
    
    sig_activities$tissue[i] = manifest[manifest$sample == sig_activities$Samples[i],]$bulk_tissue
    
  }
}

sig_activities$tissue[sig_activities$tissue == "Small intestine"] = "Crypt(s), small intestine"
sig_activities$tissue[sig_activities$tissue %in% c("Large intestine", "Appendix")] = "Crypt(s), large intestine"
sig_activities$tissue[sig_activities$tissue == "Skin"] = "Epidermis, skin"
sig_activities$tissue[!grepl("PD", sig_activities$Samples)] = sig_activities$Samples[!grepl("PD", sig_activities$Samples)]
sig_activities$tissue[is.na(sig_activities$tissue)] = "Tumour"
sig_activities$burden = rowSums(sig_activities[, 2:10])
sig_activities$case.id = ifelse(grepl("PD", sig_activities$Samples), substr(sig_activities$Samples, 0, 7), "merged")

sig_activities2 = sig_activities[sig_activities$tissue != "Tumour",]

sig_activities2[,2:10] = sig_activities2[,2:10] / rowSums(sig_activities2[, 2:10])

library(reshape2)
library(ggplot2)
library(ggpubr)

sig_activities.m = reshape2::melt(sig_activities2, id.vars= c("Samples", "tissue", "case.id", "burden"))
#sig_activities.m$patient = substr(sig_activities.m$Samples, 0, 7)
sig_activities.m$variable = factor(sig_activities.m$variable, levels = rev(levels(sig_activities.m$variable)))

names(table(sig_activities.m$tissue))
  
#cols = rev(c("black", "white", "#BCA7E8", "#A689E1", "#7851A9", "#FC4E07", "#00AFBB", "#E7B800", "#999999"))
cols = rev(c("#75529d", "white", "#F4A582", "#D6604D", "#B2182B", "#FC4E07", "#fbe45b", "#E7B800", "dodgerblue"))


sig_activities.m$case.id = factor(sig_activities.m$case.id, levels = c("PD50297", "PD51122", "PD51123", "merged"))
sig_activities.m.large = sig_activities.m[sig_activities.m$burden >= 100, ]

names(table(sig_activities.m.large$tissue))
#"Adenohypophysis_pituitary_gland"   "Cortex_adrenal_gland"              "Crypt(s), large intestine"         "Crypt(s), small intestine"         "Epidermis, skin" "Follicles_thyroid"                 "Glomerulus_kidney"                 "Grey_matter_cerebrum"              "Hair_follicle_skin"                "Hepatocytes_liver"  "Nerve_PNS"                         "Submucosal_glands_Trachea/Bronchi" "Thalamus"                          "Tubules_kidney"                    "White_matter_cerebrum"  

sig_activities.m.large$tissue = factor(sig_activities.m.large$tissue, levels = c("Epidermis, skin", "Crypt(s), small intestine", "Crypt(s), large intestine", "Adenohypophysis_pituitary_gland", "Grey_matter_cerebrum", "Thalamus", "White_matter_cerebrum", "Nerve_PNS", "Hair_follicle_skin", "Cortex_adrenal_gland", "Glomerulus_kidney", "Tubules_kidney", "Follicles_thyroid", "Hepatocytes_liver", "Submucosal_glands_Trachea/Bronchi"))

sig_activities2.large = sig_activities2[sig_activities2$Samples %in% unique(sig_activities.m.large$Samples),]
sig_activities2.large$case.id = factor(sig_activities2.large$case.id, levels = c("PD50297", "PD51122", "PD51123", "merged"))
sig_activities2.large$tissue = factor(sig_activities2.large$tissue, levels = c("Epidermis, skin", "Crypt(s), small intestine", "Crypt(s), large intestine", "Adenohypophysis_pituitary_gland", "Grey_matter_cerebrum", "Thalamus", "White_matter_cerebrum", "Nerve_PNS", "Hair_follicle_skin", "Cortex_adrenal_gland", "Glomerulus_kidney", "Tubules_kidney", "Follicles_thyroid", "Hepatocytes_liver", "Submucosal_glands_Trachea/Bronchi"))
sig_activities.m.large$Samples = factor(sig_activities.m.large$Samples, levels = sig_activities2.large[order(sig_activities2.large$case.id, sig_activities2.large$tissue, sig_activities2.large$SBS88, sig_activities2.large$SBS7a + sig_activities2.large$SBS7b + sig_activities2.large$SBS7d, sig_activities2.large$SBS18, sig_activities2.large$SBS1, sig_activities2.large$SBS5),]$Samples)

#sig_activities.m.large$Samples = factor(sig_activities.m.large$Samples, levels = unique(sig_activities.m.large[order(sig_activities.m.large$case.id, sig_activities.m.large$tissue),]$Samples))

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/non_tumour_lcm_signatures_20230106.pdf", height = 5, width = 11, useDingbats = F)
ggplot(sig_activities.m.large) +
  geom_bar(mapping = aes(x = Samples, y = value, fill = variable), stat = "identity", position = "stack", col = NA, width = 1) +
  theme_bw() +
  scale_fill_manual(values = cols) +
  theme(strip.text = element_text(angle = 90), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  coord_cartesian(expand = F) +
  xlab("") +
  ylab("SBS signature proportion")
dev.off()

```

##Plot PD51123 11p LOH (EDF2)

```{r}

library(VariantAnnotation)

vcf = readVcf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/localised_copy_number_calling/PD51123t_lo0028.battenberg.logR_Baf_segmented.vcf")
ascat = read.table(gzfile("/nfs/cancer_ref01/nst_links/live/2571/PD51123t_lo0028/PD51123t_lo0028.ascat.counts.gz"), header = T, sep = "\t", stringsAsFactors = F, comment.char = "")

var.df = data.frame(Chr = as.character(seqnames(rowRanges(vcf))),
                    Position = start(ranges(rowRanges(vcf))))

var.df$BAFP = info(vcf)[,1]
var.df$BAFE = info(vcf)[,2]
var.df$BAFS = info(vcf)[,3]
var.df$LOGR = info(vcf)[,4]

library(cowplot)

p1 <- ggplot(ascat[ascat$X.CHR == 11,])+
  geom_point(mapping = aes(x = POS, y = Good_depth), pch = ".", col = "#999999", alpha = 0.5) +
  ylim(0, 75) +
  theme_bw() +
  theme(axis.text = element_blank(), 
        axis.ticks.x = element_blank(),
        plot.margin = unit(c(0, 0, 0, 0), "cm")) +
  xlim(0, 100000000) +
  xlab("") +
  ylab("")

p2 <- ggplot(var.df[var.df$Chr == "chr11",]) +
  geom_point(mapping = aes(x = Position, y = BAFE), pch = ".", col = "#999999", alpha = 0.5) +
  geom_line(mapping = aes(x = Position, y = BAFS), col = "#e83b3d", size = 1) +
  theme_bw() +
  theme(axis.text = element_blank(),
        plot.margin = unit(c(0, 0, 0, 0), "cm")) +
  xlim(0, 100000000) +
  xlab("") +
  ylab("")

png("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/PD51123t_chr11p_LOH_plot.png", width = 3, height = 2, units = "in", res = 300)
plot_grid(p1, p2, ncol = 1)
dev.off()

```

##Driver analysis

###Subs/indels

These are reviewed prior to final filtering to avoid removing real drivers during filtering.

####Intersect with COSMIC

#####Subs

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/subs/05_read_direction") #use the unmatched version to catch germline  and very small clones that miss beta binomial filter

sub.files = list.files(".", ".standard.gnomAD.AF.0.001.PON.read.direction.filtered.txt")

cosmic_v94 <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/cosmic_v94_cancer_gene_census.csv', sep = ',', header = T, stringsAsFactors = F)
cancer_genes <- c(unique(unlist(strsplit(cosmic_v94$Synonyms, ","))), "H3-3A")
hgg_genes = c("H3-3A", "H3.3A", "H3F3A", "HIST1H3B", "HIST1H3C", "HIST2H3C", "TP53", "ATRX", "PDGFRA", "ACVR1", "PIK3CA", "CDKN2A", "CDKN2B", "NF1", "KIT", "KDR", "PIK3R1", "MYCN", "PPM1D", "EGFR", "PTEN", "BCOR", "TOP3A", "BRAF", "MET", "FGFR1", "ID2", "ATM", "CDK4", "SETD2", "CCND2", "MYC", "RB1", "TERT", "ASXL1", "CDK6", "KDM6B", "NTRK1", "NTRK2", "NTRK3") #mackay et al. 2017

prc2_genes = c("EZH1", "EZH2", "EED", "SUZ12", "RBBP7", "RBBP4")

all_subs = data.frame()
for(file in sub.files){
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  if(unique(data$Sample) != "PD50297o_lo0013"){
    print(unique(data$Sample))
    data = data[data$Gene %in% c(cancer_genes, hgg_genes, prc2_genes),]
    all_subs = rbind(all_subs, data)
  }
}

table(all_subs[!is.na(all_subs$Gene),]$Consequence)

poss_drivers = all_subs[all_subs$Consequence %in% c("ess_splice", "missense", "nonsense", "start_lost", "stop_lost"),] #no start lost or stop lost variants 20/10/22
poss_drivers = poss_drivers[order(poss_drivers$Chr, poss_drivers$Position),]
poss_drivers[poss_drivers$Gene == "KRAS",] #none

germline_vars = poss_drivers[poss_drivers$Sample %in% c("PD50297b", "PD51122q", "PD51123aa3"), 3:6]
germline_vars = unique(paste(germline_vars$Chr, germline_vars$Position, germline_vars$Wildtype, germline_vars$Mutant, sep = "_"))

uniq_poss_drivers = poss_drivers
#uniq_poss_drivers = poss_drivers[!duplicated(poss_drivers[, c(1,3, 4, 5, 6)]), c(1, 3:6, 10:ncol(poss_drivers))]
uniq_poss_drivers$mut_ID = paste(uniq_poss_drivers$Chr, uniq_poss_drivers$Position, uniq_poss_drivers$Wildtype, uniq_poss_drivers$Mutant, sep = "_")

uniq_poss_drivers$Molecular.Genetics = NA
uniq_poss_drivers$Mutation.Types = NA
uniq_poss_drivers$Tumour.Types.Somatic. = NA
uniq_poss_drivers$Role.in.Cancer = NA
uniq_poss_drivers$Germline.Somatic = NA

for(i in 1:nrow(uniq_poss_drivers)){
  if(nrow(cosmic_v94[cosmic_v94$Gene.Symbol == uniq_poss_drivers$Gene[i],]) == 1){
    uniq_poss_drivers$Molecular.Genetics[i] = cosmic_v94[cosmic_v94$Gene.Symbol == uniq_poss_drivers$Gene[i],]$Molecular.Genetics
    uniq_poss_drivers$Mutation.Types[i] = cosmic_v94[cosmic_v94$Gene.Symbol == uniq_poss_drivers$Gene[i],]$Mutation.Types
    uniq_poss_drivers$Tumour.Types.Somatic.[i] = cosmic_v94[cosmic_v94$Gene.Symbol == uniq_poss_drivers$Gene[i],]$Tumour.Types.Somatic.
    uniq_poss_drivers$Role.in.Cancer[i] = cosmic_v94[cosmic_v94$Gene.Symbol == uniq_poss_drivers$Gene[i],]$Role.in.Cancer
  }
  
  if(uniq_poss_drivers$mut_ID[i] %in% germline_vars){
    uniq_poss_drivers$Germline.Somatic[i] = "Germline"
  } else {
    uniq_poss_drivers$Germline.Somatic[i] = "Somatic"
  }
  
}

uniq_poss_drivers$hgg_gene = ifelse(uniq_poss_drivers$Gene %in% hgg_genes, T, F)
uniq_poss_drivers$prc2_gene = ifelse(uniq_poss_drivers$Gene %in% prc2_genes, T, F)

#write.table(uniq_poss_drivers, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/driver_analysis/putative_sub_drivers_20220814.txt", col.names = T, row.names = F, sep = "\t", quote = F)
write.table(uniq_poss_drivers, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/driver_analysis/putative_sub_drivers_wgs_20221023.txt", col.names = T, row.names = F, sep = "\t", quote = F) #additional columns

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
hq_tumour_samples = manifest[manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$average.reads.per.chromosome.copy.tumour >= 5 & (manifest$CNAqc_check == "PASS" | manifest$ascatPCA_gof >= 95), ]$sample #gold standard LCM samples

pure_tumour_poss_drivers = uniq_poss_drivers[uniq_poss_drivers$Sample %in% hq_tumour_samples,]

write.table(pure_tumour_poss_drivers, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/driver_analysis/putative_sub_drivers_wgs_pure_tumour_20221023.txt", col.names = T, row.names = F, sep = "\t", quote = F)

uniq_poss_drivers$purity = NA

for(i in 1:nrow(uniq_poss_drivers)){
  
  uniq_poss_drivers$purity[i] = manifest[manifest$sample == uniq_poss_drivers$Sample[i],]$purity_trunk
  
}

write.table(uniq_poss_drivers, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/driver_analysis/putative_sub_drivers_wgs_purity_added_20221206.txt", col.names = T, row.names = F, sep = "\t", quote = F)

```

Are any of the clonal driver subs in normal tissues?

```{r}

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

uniq_poss_drivers = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/driver_analysis/putative_sub_drivers_wgs_purity_added_20221206.txt", header =T ,sep = "\t", stringsAsFactors = F)

min(manifest[manifest$sample %in% uniq_poss_drivers[uniq_poss_drivers$mut_ID == "chr1_226064434_A_T",]$Sample, ]$purity_trunk) #0.08841944
uniq_poss_drivers[uniq_poss_drivers$mut_ID == "chr1_226064434_A_T" & uniq_poss_drivers$VAF == min(uniq_poss_drivers[uniq_poss_drivers$mut_ID == "chr1_226064434_A_T",]$VAF),]

min(manifest[manifest$sample %in% uniq_poss_drivers[uniq_poss_drivers$mut_ID == "chr17_7673802_C_T",]$Sample, ]$purity_trunk) #0.06322258

uniq_poss_drivers[uniq_poss_drivers$mut_ID == "chr17_7673802_C_T" & uniq_poss_drivers$VAF == min(uniq_poss_drivers[uniq_poss_drivers$mut_ID == "chr17_7673802_C_T",]$VAF),]

min(manifest[manifest$sample %in% uniq_poss_drivers[uniq_poss_drivers$mut_ID == "chr17_7675088_C_T",]$Sample, ]$purity_trunk) #0.1472562

uniq_poss_drivers[uniq_poss_drivers$mut_ID == "chr17_7675088_C_T" & uniq_poss_drivers$VAF == min(uniq_poss_drivers[uniq_poss_drivers$mut_ID == "chr17_7675088_C_T",]$VAF),]

min(manifest[manifest$sample %in% uniq_poss_drivers[uniq_poss_drivers$mut_ID == "chr17_7675140_G_C",]$Sample, ]$purity_trunk) #0.04408267

uniq_poss_drivers[uniq_poss_drivers$mut_ID == "chr17_7675140_G_C" & uniq_poss_drivers$VAF == min(uniq_poss_drivers[uniq_poss_drivers$mut_ID == "chr17_7675140_G_C",]$VAF),]

```

#####Indels

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/indels/03_read_direction") #use the unmatched version to catch germline  and very small clones that miss beta binomial filter

indel.files = list.files(".", ".pindel.HP.9.read.direction.filtered.txt")

cosmic_v94 <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/cosmic_v94_cancer_gene_census.csv', sep = ',', header = T, stringsAsFactors = F)
cancer_genes <- c(unique(unlist(strsplit(cosmic_v94$Synonyms, ","))), "H3-3A")
hgg_genes = c("H3-3A", "H3.3A", "H3F3A", "HIST1H3B", "HIST1H3C", "HIST2H3C", "TP53", "ATRX", "PDGFRA", "ACVR1", "PIK3CA", "CDKN2A", "CDKN2B", "NF1", "KIT", "KDR", "PIK3R1", "MYCN", "PPM1D", "EGFR", "PTEN", "BCOR", "TOP3A", "BRAF", "MET", "FGFR1", "ID2", "ATM", "CDK4", "SETD2", "CCND2", "MYC", "RB1", "TERT", "ASXL1", "CDK6", "KDM6B", "NTRK1", "NTRK2", "NTRK3") #mackay et al. 2017
prc2_genes = c("EZH1", "EZH2", "EED", "SUZ12", "RBBP7", "RBBP4")

all_indels = data.frame()
for(file in indel.files){
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  if(unique(data$Sample) != "PD50297o_lo0013"){
    print(unique(data$Sample))
    data = data[data$Gene %in% c(cancer_genes, hgg_genes, prc2_genes),]
    all_indels = rbind(all_indels, data)
  }
}

table(all_indels[!is.na(all_indels$Gene),]$Consequence)

poss_drivers = all_indels[all_indels$Consequence %in% c("ess_splice", "frameshift", "inframe", "SO:0000010:protein_coding;SO:0000147:exon;SO:0000204:five_prime_UTR;SO:0001993:extended_cis_splice_site;SO:0000159:deletion;SO:0001623:5_prime_UTR_variant;SO:0001577:complex_change_in_transcript;SO:0001576:transcript_variant"),]
poss_drivers = poss_drivers[order(poss_drivers$Chr, poss_drivers$Position),]
poss_drivers[poss_drivers$Gene == "KRAS",] #none

germline_vars = poss_drivers[poss_drivers$Sample %in% c("PD50297b", "PD51122q", "PD51123aa3"), 3:6]
germline_vars = unique(paste(germline_vars$Chr, germline_vars$Position, germline_vars$Wildtype, germline_vars$Mutant, sep = "_"))

uniq_poss_drivers = poss_drivers
#uniq_poss_drivers = poss_drivers[!duplicated(poss_drivers[, c(1,3, 4, 5, 6)]), c(1, 3:6, 10:ncol(poss_drivers))]
uniq_poss_drivers$mut_ID = paste(uniq_poss_drivers$Chr, uniq_poss_drivers$Position, uniq_poss_drivers$Wildtype, uniq_poss_drivers$Mutant, sep = "_")

uniq_poss_drivers$Molecular.Genetics = NA
uniq_poss_drivers$Mutation.Types = NA
uniq_poss_drivers$Tumour.Types.Somatic. = NA
uniq_poss_drivers$Role.in.Cancer = NA
uniq_poss_drivers$Germline.Somatic = NA

for(i in 1:nrow(uniq_poss_drivers)){
  if(nrow(cosmic_v94[cosmic_v94$Gene.Symbol == uniq_poss_drivers$Gene[i],]) == 1){
    uniq_poss_drivers$Molecular.Genetics[i] = cosmic_v94[cosmic_v94$Gene.Symbol == uniq_poss_drivers$Gene[i],]$Molecular.Genetics
    uniq_poss_drivers$Mutation.Types[i] = cosmic_v94[cosmic_v94$Gene.Symbol == uniq_poss_drivers$Gene[i],]$Mutation.Types
    uniq_poss_drivers$Tumour.Types.Somatic.[i] = cosmic_v94[cosmic_v94$Gene.Symbol == uniq_poss_drivers$Gene[i],]$Tumour.Types.Somatic.
    uniq_poss_drivers$Role.in.Cancer[i] = cosmic_v94[cosmic_v94$Gene.Symbol == uniq_poss_drivers$Gene[i],]$Role.in.Cancer
  }
  
  if(uniq_poss_drivers$mut_ID[i] %in% germline_vars){
    uniq_poss_drivers$Germline.Somatic[i] = "Germline"
  } else {
    uniq_poss_drivers$Germline.Somatic[i] = "Somatic"
  }
  
}

uniq_poss_drivers$hgg_gene = ifelse(uniq_poss_drivers$Gene %in% hgg_genes, T, F)
uniq_poss_drivers$prc2_gene = ifelse(uniq_poss_drivers$Gene %in% prc2_genes, T, F)

#write.table(uniq_poss_drivers, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/driver_analysis/putative_indel_drivers_20220814.txt", col.names = T, row.names = F, sep = "\t", quote = F)
write.table(uniq_poss_drivers, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/driver_analysis/putative_indel_drivers_wgs_20221023.txt", col.names = T, row.names = F, sep = "\t", quote = F) #additional columns

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
hq_tumour_samples = manifest[manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$average.reads.per.chromosome.copy.tumour >= 5 & (manifest$CNAqc_check == "PASS" | manifest$ascatPCA_gof >= 95), ]$sample #gold standard LCM samples

pure_tumour_poss_drivers = uniq_poss_drivers[uniq_poss_drivers$Sample %in% hq_tumour_samples,]

write.table(pure_tumour_poss_drivers, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/driver_analysis/putative_indel_drivers_wgs_pure_tumour_20221023.txt", col.names = T, row.names = F, sep = "\t", quote = F)

uniq_poss_drivers$purity = NA

for(i in 1:nrow(uniq_poss_drivers)){
  
  uniq_poss_drivers$purity[i] = manifest[manifest$sample == uniq_poss_drivers$Sample[i],]$purity_trunk
  
}

write.table(uniq_poss_drivers, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/driver_analysis/putative_indel_drivers_wgs_purity_added_20221206.txt", col.names = T, row.names = F, sep = "\t", quote = F)

```

###Copy number aberrations

####Bulk (Battenberg)

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/copy_number/battenberg/bulk/tumour/")

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
bulk_tumour = manifest[manifest$keep == "Y" & !is.na(manifest$battenberg_purity) & manifest$battenberg_purity >= 0.4,]
row.names(bulk_tumour) = bulk_tumour$sample
samples = bulk_tumour$sample

#all_cn = read.table("battenberg_all_segments_20221021.txt", header = T, sep = "\t", stringsAsFactors = F)
all_cn = read.table("battenberg_all_segments_20221219.txt", header = T, sep = "\t", stringsAsFactors = F)

cosmic_v94 <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/cosmic_v94_cancer_gene_census.csv', sep = ',', header = T, stringsAsFactors = F)
cosmic_v94 = cosmic_v94[!grepl(":-", cosmic_v94$Genome.Location),]
cosmic_v94$chr = paste0("chr", unlist(lapply(strsplit(cosmic_v94$Genome.Location, ":"), "[[", 1)))
cosmic_v94$start = as.numeric(unlist(lapply(strsplit(unlist(lapply(strsplit(cosmic_v94$Genome.Location, ":"), "[[", 2)), "-"), "[[", 1)))
cosmic_v94$end = as.numeric(unlist(lapply(strsplit(unlist(lapply(strsplit(cosmic_v94$Genome.Location, ":"), "[[", 2)), "-"), "[[", 2)))

oncogenes = cosmic_v94[grepl("oncogene", cosmic_v94$Role.in.Cancer) | grepl("Dom", cosmic_v94$Molecular.Genetics), ]
tsg = cosmic_v94[grepl("TSG", cosmic_v94$Role.in.Cancer) | grepl("Rec", cosmic_v94$Molecular.Genetics), ]

all_cn$segment_length = all_cn$endpos - all_cn$startpos

all_cn$oncogenes = NA
all_cn$tsg = NA

for(i in 1:nrow(all_cn)){
  if(nrow(oncogenes[oncogenes$chr == all_cn$chr[i] & oncogenes$start <= all_cn$endpos[i] & oncogenes$end >= all_cn$startpos[i], ]) > 0){
    all_cn$oncogenes[i] = paste(oncogenes[oncogenes$chr == all_cn$chr[i] & oncogenes$start <= all_cn$endpos[i] & oncogenes$end >= all_cn$startpos[i], ]$Gene.Symbol, collapse = ",")
    
  }
}

for(i in 1:nrow(all_cn)){
  if(nrow(tsg[tsg$chr == all_cn$chr[i] & tsg$start <= all_cn$endpos[i] & tsg$end >= all_cn$startpos[i], ]) > 0){
    all_cn$tsg[i] = paste(tsg[tsg$chr == all_cn$chr[i] & tsg$start <= all_cn$endpos[i] & tsg$end >= all_cn$startpos[i], ]$Gene.Symbol, collapse = ",")
    
  }
}

all_cn$tumour_ploidy = NA
all_cn$tumour_ploidy = bulk_tumour[all_cn$sample, ]$battenberg_ploidy

#write.table(all_cn, "battenberg_pure_tumour_segments_20221021.txt", col.names = T, row.names = F, sep = "\t", quote = F)
write.table(all_cn, "battenberg_pure_tumour_segments_20221219.txt", col.names = T, row.names = F, sep = "\t", quote = F) #same length as old one

#report all <1e6 and the previously reported <1e7

#oncogene
##1e7, wgd
wgd_1e7_no_cn_onc = all_cn[all_cn$tumour_ploidy > 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        !is.na(all_cn$oncogenes) & 
                        all_cn$segment_length < 1e7, ] #MYCN
wgd_1e7_hi_cn_onc = all_cn[all_cn$tumour_ploidy > 2.7 & 
                        !is.na(all_cn$nMaj1_A) & 
                        ((all_cn$nMaj1_A + all_cn$nMin1_A >= 9) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A >= 9))) &
                        !is.na(all_cn$oncogenes) & 
                        all_cn$segment_length < 1e7, ] #Estimated 31 copies of NUTM1 (amplification) in PD51122l, uncertain significance

##check for hgg, histone and prc2 genes beyond cosmic oncogenes - none
###hgg
all_cn[all_cn$tumour_ploidy > 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        is.na(all_cn$oncogenes) & 
                        !is.na(all_cn$hgg_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
all_cn[all_cn$tumour_ploidy > 2.7 & 
                        !is.na(all_cn$nMaj1_A) & 
                        ((all_cn$nMaj1_A + all_cn$nMin1_A >= 9) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A >= 9))) &
                        is.na(all_cn$oncogenes) & 
                        !is.na(all_cn$hgg_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
###histone genes
all_cn[all_cn$tumour_ploidy > 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        is.na(all_cn$oncogenes) & 
                        !is.na(all_cn$histone_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
all_cn[all_cn$tumour_ploidy > 2.7 & 
                        !is.na(all_cn$nMaj1_A) & 
                        ((all_cn$nMaj1_A + all_cn$nMin1_A >= 9) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A >= 9))) &
                        is.na(all_cn$oncogenes) & 
                        !is.na(all_cn$histone_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
###prc2 genes
all_cn[all_cn$tumour_ploidy > 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        is.na(all_cn$oncogenes) & 
                        !is.na(all_cn$prc2_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
all_cn[all_cn$tumour_ploidy > 2.7 & 
                        !is.na(all_cn$nMaj1_A) & 
                        ((all_cn$nMaj1_A + all_cn$nMin1_A >= 9) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A >= 9))) &
                        is.na(all_cn$oncogenes) & 
                        !is.na(all_cn$prc2_genes) &
                        all_cn$segment_length < 1e7, ] #no rows

#1e7, no wgd
no_wgd_1e7_no_cn_onc = all_cn[all_cn$tumour_ploidy <= 2.7 & 
                          is.na(all_cn$nMaj1_A) & 
                          !is.na(all_cn$oncogenes) & 
                          all_cn$segment_length < 1e7, ]
no_wgd_1e7_hi_cn_onc = all_cn[all_cn$tumour_ploidy <= 2.7 & 
                          !is.na(all_cn$nMaj1_A) & 
                          ((all_cn$nMaj1_A + all_cn$nMin1_A >= 5) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A >= 5))) &
                          !is.na(all_cn$oncogenes) & 
                          all_cn$segment_length < 1e7, ]

##check for hgg, histone and prc2 genes beyond cosmic oncogenes - none
###hgg
all_cn[all_cn$tumour_ploidy <= 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        is.na(all_cn$oncogenes) & 
                        !is.na(all_cn$hgg_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
all_cn[all_cn$tumour_ploidy <= 2.7 & 
                        !is.na(all_cn$nMaj1_A) & 
                        ((all_cn$nMaj1_A + all_cn$nMin1_A >= 5) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A >= 5))) &
                        is.na(all_cn$oncogenes) & 
                        !is.na(all_cn$hgg_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
###histone genes
all_cn[all_cn$tumour_ploidy <= 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        is.na(all_cn$oncogenes) & 
                        !is.na(all_cn$histone_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
all_cn[all_cn$tumour_ploidy <= 2.7 & 
                        !is.na(all_cn$nMaj1_A) & 
                        ((all_cn$nMaj1_A + all_cn$nMin1_A >= 5) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A >= 5))) &
                        is.na(all_cn$oncogenes) & 
                        !is.na(all_cn$histone_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
###prc2 genes
all_cn[all_cn$tumour_ploidy <= 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        is.na(all_cn$oncogenes) & 
                        !is.na(all_cn$prc2_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
all_cn[all_cn$tumour_ploidy <= 2.7 & 
                        !is.na(all_cn$nMaj1_A) & 
                        ((all_cn$nMaj1_A + all_cn$nMin1_A >= 5) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A >= 5))) &
                        is.na(all_cn$oncogenes) & 
                        !is.na(all_cn$prc2_genes) &
                        all_cn$segment_length < 1e7, ] #no rows

#tsg
##1e7, wgd
wgd_1e7_no_cn_tsg = all_cn[all_cn$tumour_ploidy > 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        !is.na(all_cn$tsg) & 
                        all_cn$segment_length < 1e7, ]

wgd_1e7_lo_cn_tsg = all_cn[all_cn$tumour_ploidy > 2.7 & 
         !is.na(all_cn$nMaj1_A) & 
         ((all_cn$nMaj1_A + all_cn$nMin1_A < all_cn$tumour_ploidy - 2.7) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A < all_cn$tumour_ploidy - 2.7))) &
         !is.na(all_cn$tsg) & 
         all_cn$segment_length < 1e7, ]

##check for hgg, histone and prc2 genes beyond cosmic oncogenes - none
###hgg
all_cn[all_cn$tumour_ploidy > 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        is.na(all_cn$tsg) & 
                        !is.na(all_cn$hgg_genes) &
                        all_cn$segment_length < 1e7, ] #MYCN  oncogene, already captured in oncogenes
all_cn[all_cn$tumour_ploidy > 2.7 & 
                        !is.na(all_cn$nMaj1_A) & 
                        ((all_cn$nMaj1_A + all_cn$nMin1_A < all_cn$tumour_ploidy - 2.7) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A < all_cn$tumour_ploidy - 2.7))) &
                        is.na(all_cn$tsg) & 
                        !is.na(all_cn$hgg_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
###histone genes
all_cn[all_cn$tumour_ploidy > 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        is.na(all_cn$tsg) & 
                        !is.na(all_cn$histone_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
all_cn[all_cn$tumour_ploidy > 2.7 & 
                        !is.na(all_cn$nMaj1_A) & 
                        ((all_cn$nMaj1_A + all_cn$nMin1_A < all_cn$tumour_ploidy - 2.7) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A < all_cn$tumour_ploidy - 2.7))) &
                        is.na(all_cn$tsg) & 
                        !is.na(all_cn$histone_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
###prc2 genes
all_cn[all_cn$tumour_ploidy > 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        is.na(all_cn$tsg) & 
                        !is.na(all_cn$prc2_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
all_cn[all_cn$tumour_ploidy > 2.7 & 
                        !is.na(all_cn$nMaj1_A) & 
                        ((all_cn$nMaj1_A + all_cn$nMin1_A < all_cn$tumour_ploidy - 2.7) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A < all_cn$tumour_ploidy - 2.7))) &
                        is.na(all_cn$tsg) & 
                        !is.na(all_cn$prc2_genes) &
                        all_cn$segment_length < 1e7, ] #no rows

##1e7, no wgd
no_wgd_1e7_no_cn_tsg = all_cn[all_cn$tumour_ploidy <= 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        !is.na(all_cn$tsg) & 
                        all_cn$segment_length < 1e7, ]
no_wgd_1e7_lo_cn_tsg = all_cn[all_cn$tumour_ploidy <= 2.7 & 
         !is.na(all_cn$nMaj1_A) & 
         ((all_cn$nMaj1_A + all_cn$nMin1_A == 0) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A == 0))) &
         !is.na(all_cn$tsg) & 
         all_cn$segment_length < 1e7, ]

##check for hgg, histone and prc2 genes beyond cosmic oncogenes - none
###hgg
all_cn[all_cn$tumour_ploidy <= 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        is.na(all_cn$tsg) & 
                        !is.na(all_cn$hgg_genes) &
                        all_cn$segment_length < 1e7, ] #MYCN  oncogene, already captured in oncogenes
all_cn[all_cn$tumour_ploidy <= 2.7 & 
                        !is.na(all_cn$nMaj1_A) & 
                        ((all_cn$nMaj1_A + all_cn$nMin1_A == 0) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A == 0))) &
                        is.na(all_cn$tsg) & 
                        !is.na(all_cn$hgg_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
###histone genes
all_cn[all_cn$tumour_ploidy <= 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        is.na(all_cn$tsg) & 
                        !is.na(all_cn$histone_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
all_cn[all_cn$tumour_ploidy <= 2.7 & 
                        !is.na(all_cn$nMaj1_A) & 
                        ((all_cn$nMaj1_A + all_cn$nMin1_A == 0) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A == 0))) &
                        is.na(all_cn$tsg) & 
                        !is.na(all_cn$histone_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
###prc2 genes
all_cn[all_cn$tumour_ploidy <= 2.7 & 
                        is.na(all_cn$nMaj1_A) &
                        is.na(all_cn$tsg) & 
                        !is.na(all_cn$prc2_genes) &
                        all_cn$segment_length < 1e7, ] #no rows
all_cn[all_cn$tumour_ploidy <= 2.7 & 
                        !is.na(all_cn$nMaj1_A) & 
                        ((all_cn$nMaj1_A + all_cn$nMin1_A == 0) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A == 0))) &
                        is.na(all_cn$tsg) & 
                        !is.na(all_cn$prc2_genes) &
                        all_cn$segment_length < 1e7, ] #no rows

poss_cn_drivers = rbind(wgd_1e7_no_cn_onc, wgd_1e7_hi_cn_onc, no_wgd_1e7_no_cn_onc, no_wgd_1e7_hi_cn_onc, wgd_1e7_no_cn_tsg, wgd_1e7_lo_cn_tsg, no_wgd_1e7_no_cn_tsg, no_wgd_1e7_lo_cn_tsg)
poss_cn_drivers_uniq = poss_cn_drivers[!duplicated(poss_cn_drivers),] #no change.

#write.table(poss_cn_drivers_uniq, "putative_bulk_tumour_driver_cnas_20220928.txt", col.names = T, row.names = F, sep = "\t", quote = F)
#write.table(poss_cn_drivers_uniq, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/driver_analysis/putative_bulk_tumour_driver_cnas_20221021.txt", col.names = T, row.names = F, sep = "\t", quote = F) #added columns
write.table(poss_cn_drivers_uniq, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/driver_analysis/putative_bulk_tumour_driver_cnas_20221219.txt", col.names = T, row.names = F, sep = "\t", quote = F) #added columns, same length

#check that no oncogenes have focal deletions which might be activating
setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/copy_number/battenberg/bulk/tumour/")

#all_cn = read.table("battenberg_pure_tumour_segments_20221021.txt", header = T, sep = "\t", stringsAsFactors = F)
all_cn = read.table("battenberg_pure_tumour_segments_20221219.txt", header = T, sep = "\t", stringsAsFactors = F)

all_cn[all_cn$tumour_ploidy > 2.7 & 
         !is.na(all_cn$nMaj1_A) & 
         ((all_cn$nMaj1_A + all_cn$nMin1_A < all_cn$tumour_ploidy - 2.7) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A < all_cn$tumour_ploidy - 2.7))) &
         !is.na(all_cn$oncogenes) & 
         all_cn$segment_length < 1e7, ] #CDKN2A mediates, not the MLLT3
all_cn[all_cn$tumour_ploidy <= 2.7 & 
         !is.na(all_cn$nMaj1_A) & 
         ((all_cn$nMaj1_A + all_cn$nMin1_A == 0) | (!is.na(all_cn$frac2_A) & (all_cn$nMaj2_A + all_cn$nMin2_A == 0))) &
         !is.na(all_cn$oncogenes) & 
         all_cn$segment_length < 1e7, ] #none

```

####LCM (ascatPCA)

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/copy_number/ascatPCA/seg_100/")

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
bulk_tumour = manifest[manifest$keep == "Y" & !is.na(manifest$battenberg_purity) & manifest$battenberg_purity >= 0.4,]
row.names(bulk_tumour) = bulk_tumour$sample
samples = bulk_tumour$sample

#all_cn = read.table("ascatPCA_all_segments_20221022.txt", header = T, sep = "\t", stringsAsFactors = F)
all_cn = read.table("ascatPCA_all_segments_20221219.txt", header = T, sep = "\t", stringsAsFactors = F) #same length as each other

cosmic_v94 <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/cosmic_v94_cancer_gene_census.csv', sep = ',', header = T, stringsAsFactors = F)
cosmic_v94 = cosmic_v94[!grepl(":-", cosmic_v94$Genome.Location),]
cosmic_v94$chr = paste0("chr", unlist(lapply(strsplit(cosmic_v94$Genome.Location, ":"), "[[", 1)))
cosmic_v94$start = as.numeric(unlist(lapply(strsplit(unlist(lapply(strsplit(cosmic_v94$Genome.Location, ":"), "[[", 2)), "-"), "[[", 1)))
cosmic_v94$end = as.numeric(unlist(lapply(strsplit(unlist(lapply(strsplit(cosmic_v94$Genome.Location, ":"), "[[", 2)), "-"), "[[", 2)))

oncogenes = cosmic_v94[grepl("oncogene", cosmic_v94$Role.in.Cancer) | grepl("Dom", cosmic_v94$Molecular.Genetics), ]
tsg = cosmic_v94[grepl("TSG", cosmic_v94$Role.in.Cancer) | grepl("Rec", cosmic_v94$Molecular.Genetics), ]

names(all_cn) = c("chr", "startpos", "endpos", "total_cn", "minor_cn", "sample", "est_ploidy", "cosmic_genes", "histone_genes", "hgg_genes", "prc2_genes")

all_cn$segment_length = all_cn$endpos - all_cn$startpos

all_cn$oncogenes = NA
all_cn$tsg = NA

for(i in 1:nrow(all_cn)){
  if(nrow(oncogenes[oncogenes$chr == all_cn$chr[i] & oncogenes$start <= all_cn$endpos[i] & oncogenes$end >= all_cn$startpos[i], ]) > 0){
    all_cn$oncogenes[i] = paste(oncogenes[oncogenes$chr == all_cn$chr[i] & oncogenes$start <= all_cn$endpos[i] & oncogenes$end >= all_cn$startpos[i], ]$Gene.Symbol, collapse = ",")
    
  }
}

for(i in 1:nrow(all_cn)){
  if(nrow(tsg[tsg$chr == all_cn$chr[i] & tsg$start <= all_cn$endpos[i] & tsg$end >= all_cn$startpos[i], ]) > 0){
    all_cn$tsg[i] = paste(tsg[tsg$chr == all_cn$chr[i] & tsg$start <= all_cn$endpos[i] & tsg$end >= all_cn$startpos[i], ]$Gene.Symbol, collapse = ",")
    
  }
}

hq_tumour_samples = manifest[manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$average.reads.per.chromosome.copy.tumour >= 5 & (manifest$CNAqc_check == "PASS" | manifest$ascatPCA_gof >= 95), ]$sample

all_cn$case.id = substr(all_cn$sample, 0, 7)

#write.table(all_cn[all_cn$sample %in% hq_tumour_samples,], "ascatPCA_pure_tumour_segments_20221025.txt", col.names = T, row.names = F, sep = "\t", quote = F)
write.table(all_cn[all_cn$sample %in% hq_tumour_samples,], "ascatPCA_pure_tumour_segments_20221219.txt", col.names = T, row.names = F, sep = "\t", quote = F) #same length as before

#cna_df = all_cn
cna_df = all_cn[all_cn$sample %in% hq_tumour_samples,]

#oncogene
##1e7, wgd
cna_df[cna_df$est_ploidy > 2.7 &
            is.na(cna_df$total_cn) &
            !is.na(cna_df$oncogenes) &
            cna_df$segment_length < 1e7,] #nil
wgd_1e7_hi_cn_onc = cna_df[cna_df$est_ploidy > 2.7 &
                            cna_df$total_cn >= 9 &
                            !is.na(cna_df$oncogenes) &
                            cna_df$segment_length < 1e7,]

##1e7, no wgd
cna_df[cna_df$est_ploidy <= 2.7 &
                            is.na(cna_df$total_cn) &
                            !is.na(cna_df$oncogenes) &
                            cna_df$segment_length < 1e7,] #nil
no_wgd_1e7_hi_cn_onc = cna_df[cna_df$est_ploidy <= 2.7 &
                            cna_df$total_cn >= 5 &
                            !is.na(cna_df$oncogenes) &
                            cna_df$segment_length < 1e7,]


#tsg
##1e7, wgd
cna_df[cna_df$est_ploidy > 2.7 &
                            is.na(cna_df$total_cn) &
                            !is.na(cna_df$tsg) &
                            cna_df$segment_length < 1e7,] #nil
wgd_1e7_lo_cn_tsg = cna_df[cna_df$est_ploidy > 2.7 &
                            cna_df$total_cn < (cna_df$est_ploidy - 2.7) &
                            !is.na(cna_df$tsg) &
                            cna_df$segment_length < 1e7,]

##1e7, no wgd
cna_df[cna_df$est_ploidy <= 2.7 &
                            is.na(cna_df$total_cn) &
                            !is.na(cna_df$tsg) &
                            cna_df$segment_length < 1e7,] #nil
no_wgd_1e7_lo_cn_tsg = cna_df[cna_df$est_ploidy <= 2.7 &
                            cna_df$total_cn == 0 &
                            !is.na(cna_df$tsg) &
                            cna_df$segment_length < 1e7,]

poss_cn_drivers = rbind(wgd_1e7_hi_cn_onc, no_wgd_1e7_hi_cn_onc, wgd_1e7_lo_cn_tsg, no_wgd_1e7_lo_cn_tsg)
poss_cn_drivers_uniq = poss_cn_drivers[!duplicated(poss_cn_drivers),] #no change.

row.names(manifest) = manifest$sample

poss_cn_drivers_uniq$tumour_infiltration = manifest[poss_cn_drivers_uniq$sample,]$purity_trunk
#poss_cn_drivers_uniq$bulk_tissue = manifest[poss_cn_drivers_uniq$sample,]$bulk_tissue

#write.table(poss_cn_drivers_uniq, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/driver_analysis/putative_ascatPCA_driver_cnas_20221025.txt", col.names = T, row.names = F, sep = "\t", quote = F)
write.table(poss_cn_drivers_uniq, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/driver_analysis/putative_ascatPCA_driver_cnas_20221219.txt", col.names = T, row.names = F, sep = "\t", quote = F) #same length as above

#identify additional events in normal tissues

normal_samples = manifest[manifest$purity_trunk < 0.01 & manifest$keep == "Y" & manifest$ascatPCA_gof >= 95,]$sample

#check that sex chromsomes are correct in normal tissue for each patient
sex_chrom_events1 = all_cn[all_cn$sample %in% normal_samples &
        all_cn$case.id %in% c("PD50297", "PD51122") &
        all_cn$chr %in% c("chrX", "chrY") & 
        all_cn$total_cn != 1,]
sex_chrom_events2 = all_cn[all_cn$sample %in% normal_samples &
        all_cn$case.id %in% c("PD51123") &
        ((all_cn$chr == "chrX" & all_cn$total_cn != 2) | (all_cn$chr == "chrY" & all_cn$total_cn != 0)),]
auto_chrom_events = all_cn[all_cn$sample %in% normal_samples &
         (all_cn$chr %in% paste0("chr", 1:22) & all_cn$total_cn != 2),]
poss_norm_events = rbind(sex_chrom_events1, sex_chrom_events2, auto_chrom_events)

poss_norm_events$tumour_infiltration = manifest[poss_norm_events$sample,]$purity_trunk
poss_norm_events$bulk_tissue = manifest[poss_norm_events$sample,]$bulk_tissue

#write.table(poss_norm_events, "putative_ascatPCA_normal_tissue_cnas_20221025.txt", col.names = T, row.names = F, sep = "\t", quote = F)
write.table(poss_norm_events, "putative_ascatPCA_normal_tissue_cnas_20221219.txt", col.names = T, row.names = F, sep = "\t", quote = F)

#check that no oncogenes have focal deletions which might be activating
setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/copy_number/ascatPCA/seg_100/")

cna_df = read.table("ascatPCA_pure_tumour_segments_20221219.txt", header = T, sep = "\t", stringsAsFactors = F)

cna_df[cna_df$est_ploidy > 2.7 &
                            cna_df$total_cn < (cna_df$est_ploidy - 2.7) &
                            !is.na(cna_df$oncogenes) &
                            cna_df$segment_length < 1e7,] #CDKN2A mediates, not the MLLT3
cna_df[cna_df$est_ploidy <= 2.7 &
                            cna_df$total_cn == 0 &
                            !is.na(cna_df$oncogenes) &
                            cna_df$segment_length < 1e7,] #PTEN annotated as oncogene too due to dominant effect in COSMIC, FLT3 is a single SNP, VTI1A deletion intragenic in gene that mediates oncogenesis via fusion events, KAT6B also thought to be a fusion partner driver

```

###Structural variants (GRIDSS)

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/structural_variants")

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

svs = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/structural_variants/paeds_autopsy_gridss_calls.txt", header = T, sep = "\t", stringsAsFactors = F)
svs$case.id = substr(svs$sample, 0, 7)

cosmic_v94 <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/cosmic_v94_cancer_gene_census.csv', sep = ',', header = T, stringsAsFactors = F)
cosmic_v94 = cosmic_v94[!grepl(":-", cosmic_v94$Genome.Location),]
cosmic_v94$chr = paste0("chr", unlist(lapply(strsplit(cosmic_v94$Genome.Location, ":"), "[[", 1)))
cosmic_v94$start = as.numeric(unlist(lapply(strsplit(unlist(lapply(strsplit(cosmic_v94$Genome.Location, ":"), "[[", 2)), "-"), "[[", 1)))
cosmic_v94$end = as.numeric(unlist(lapply(strsplit(unlist(lapply(strsplit(cosmic_v94$Genome.Location, ":"), "[[", 2)), "-"), "[[", 2)))

table(cosmic_v94$Role.in.Cancer)

bk_genes = unlist(strsplit(cosmic_v94[grepl("TSG|fusion", cosmic_v94$Role.in.Cancer),]$Synonyms, ",")) #cancer genes which are TSG or oncogenic by fusion
hgg_genes = c("H3-3A", "H3.3A", "H3F3A", "HIST1H3B", "HIST1H3C", "HIST2H3C", "TP53", "ATRX", "PDGFRA", "ACVR1", "PIK3CA", "CDKN2A", "CDKN2B", "NF1", "KIT", "KDR", "PIK3R1", "MYCN", "PPM1D", "EGFR", "PTEN", "BCOR", "TOP3A", "BRAF", "MET", "FGFR1", "ID2", "ATM", "CDK4", "SETD2", "CCND2", "MYC", "RB1", "TERT", "ASXL1", "CDK6", "KDM6B", "NTRK1", "NTRK2", "NTRK3") #mackay et al. 2017
prc2_genes = c("EZH1", "EZH2", "EED", "SUZ12", "RBBP7", "RBBP4")
oncogenes = cosmic_v94[grepl("oncogene", cosmic_v94$Role.in.Cancer) | grepl("Dom", cosmic_v94$Molecular.Genetics), ]
tsg = cosmic_v94[grepl("TSG", cosmic_v94$Role.in.Cancer) | grepl("Rec", cosmic_v94$Molecular.Genetics), ]

svs$bk_dist = ifelse(svs$chro1 == svs$chro2, svs$end2 - svs$start1, NA)
svs$oncogenes = NA
svs$tsg = NA
svs$chro1 = paste0("chr", svs$chro1)
svs$chro2 = paste0("chr", svs$chro2)

#annotate with any cancer drivers lying  within segments or breakpoints
for(i in 1:nrow(svs)){
  
  if(svs$chro1[i] == svs$chro2[i]){
    
    if(nrow(oncogenes[oncogenes$chr == svs$chro1[i] & oncogenes$start <= svs$end2[i] & oncogenes$end >= svs$start1[i], ]) > 0){
    svs$oncogenes[i] = paste(oncogenes[oncogenes$chr == svs$chro1[i] & oncogenes$start <= svs$end2[i] & oncogenes$end >= svs$start1[i], ]$Gene.Symbol, collapse = ",")
  }  
    
  }
  
}


for(i in 1:nrow(svs)){
  
  if(svs$chro1[i] == svs$chro2[i]){
    
    if(nrow(tsg[tsg$chr == svs$chro1[i] & tsg$start <= svs$end2[i] & tsg$end >= svs$start1[i], ]) > 0){
    svs$tsg[i] = paste(tsg[tsg$chr == svs$chro1[i] & tsg$start <= svs$end2[i] & tsg$end >= svs$start1[i], ]$Gene.Symbol, collapse = ",")
  }  
    
  }
  
}

svs$cancer_bk_genes = F
svs[svs$gene %in% bk_genes | svs$gene2 %in% bk_genes,]$cancer_bk_genes = T
svs$hgg_bk_genes = F
svs[svs$gene %in% hgg_genes | svs$gene2 %in% hgg_genes,]$hgg_bk_genes = T
svs$prc2_bk_genes = F
#svs[svs$gene %in% prc2_genes | svs$gene2 %in% prc2_genes,]$prc2_bk_genes = T there are none
svs$normal_sample = F
svs[svs$sample %in% manifest[manifest$purity_trunk < 0.01 & manifest$keep == "Y",]$sample,]$normal_sample = T

hq_tumour_samples = manifest[manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$average.reads.per.chromosome.copy.tumour >= 5 & (manifest$CNAqc_check == "PASS" | manifest$ascatPCA_gof >= 95), ]$sample

svs_tumour = svs[svs$sample %in% hq_tumour_samples,]

write.table(svs, "paeds_autopsy_annotated_svs_20221023.txt", col.names = T, row.names = F, sep = "\t", quote = F)
write.table(svs_tumour, "GRIDSS_pure_tumour_variants_20221023.txt", col.names = T, row.names = F, sep = "\t", quote = F)

#additional refinement of output done on workbook 02

```

##PD51122 chr17 phasing hetSNPs

Include WES samples in this analysis.

Using a pure tumour sample that has undergone widespread haploidisation to phase most of the genome.

PD51122m - tumour
PD51122q - blood

###Step 1: Get SNPs in matched normal (PD51122q)

```{bash, eval=F}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/PD51122_phasing/PD51122q_hetero_snps

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"
#bsub -Is -q yesterday -R 'select[mem>=30000] rusage[mem=30000]' -M30000 bash
#enter screen mode

for sample in PD51122q;
do
bsub -G team294-vwork -o $PWD/log.%J -e $PWD/err.%J -q normal -R'select[mem>30000] rusage[mem=30000]' -M30000 -J $sample /software/R-3.6.1/bin/Rscript /nfs/users/nfs_t/to3/scripts/allele_count_1000G_hg38.R $sample /nfs/cancer_ref01/nst_links/live/2571/$sample/$sample.sample.dupmarked.bam /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/PD51122_phasing/PD51122q_hetero_snps
done

#default mapping quality of 35 and base quality of 20

```

###Step 2: Identify heterozygous SNPs

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/PD51122_phasing/PD51122q_hetero_snps")

snp.files = list.files(".", "PD51122q_1000G_af_")

for(file in snp.files){
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F, comment.char = "")
  chr = unlist(strsplit(unlist(strsplit(file, ".txt")), "PD51122q_1000G_af_"))[c(F,T)]
  #if(chr == "chr23") data.good.depth = data[data$Good_depth >= 25 & data$Good_depth < 250,]
  if(chr != "chr23") data.good.depth = data[data$Good_depth >= 50 & data$Good_depth < 500,]
  data.good.depth[, c(3:6)] = data.good.depth[, c(3:6)] / data.good.depth$Good_depth
  data.good.depth = data.good.depth[rowSums(data.good.depth[, c(3:6)] > 0) == 2, ]
  data.good.depth = data.good.depth[rowSums(data.good.depth[, c(3:6)] > 0.2) == 2, ]
  write.table(data.good.depth[, c(1:2)], paste0("PD51122q_1000G_af_", chr, "_hetero_snps.txt"), col.names = F, row.names = F, sep = '\t', quote = F)
  }

```

###Step 3: Run allelecounter across samples for PD51122

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/PD51122_phasing/PD51122q_hetero_snps 
#wgs
infile="/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt" 

#SAMPLE="PD51122m"
#PROJECT=2571
grep PD51122 $infile | cut -f1,6 | while read SAMPLE PROJECT;
do

#echo $SAMPLE
#echo $PROJECT

bsub -J"${SAMPLE}" -o ../chr17_allelecount/${SAMPLE}.o -e ../chr17_allelecount/${SAMPLE}.e -q normal -n 5 -R"select[mem>10000] rusage[mem=10000] span[hosts=1]" -M10000 "/nfs/users/nfs_m/my4/bin/alleleCounter -l PD51122q_1000G_af_chr17_hetero_snps.txt -b /nfs/cancer_ref01/nst_links/live/${PROJECT}/${SAMPLE}/${SAMPLE}.sample.dupmarked.bam -o ../chr17_allelecount/${SAMPLE}_chr17_alleleFrequencies.txt -r /lustre/scratch119/casm/team78pipelines/reference/human/GRCh38_full_analysis_set_plus_decoy_hla/genome.fa -m 25 -q 30"

done

#wes
infile="/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wes_data/00_supplementary_data/wes_paeds_pm_sample_metadata_20220722.txt" 
grep PD51122 $infile | cut -f1,6 | while read SAMPLE PROJECT;
do

#echo $SAMPLE
#echo $PROJECT

bsub -J"${SAMPLE}" -o ../chr17_allelecount/${SAMPLE}.o -e ../chr17_allelecount/${SAMPLE}.e -q normal -n 5 -R"select[mem>10000] rusage[mem=10000] span[hosts=1]" -M10000 "/nfs/users/nfs_m/my4/bin/alleleCounter -l PD51122q_1000G_af_chr17_hetero_snps.txt -b /nfs/cancer_ref01/nst_links/live/${PROJECT}/${SAMPLE}/${SAMPLE}.sample.dupmarked.bam -o ../chr17_allelecount/${SAMPLE}_chr17_alleleFrequencies.txt -r /lustre/scratch119/casm/team78pipelines/reference/human/GRCh38_full_analysis_set_plus_decoy_hla/genome.fa -m 25 -q 30"

done

```

###Step 4: identify tumour and normal chr17 allele

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/PD51122_phasing/chr17_allelecount")

tum_counts = read.table("PD51122m_chr17_alleleFrequencies.txt", sep = "\t", comment.char = "", header = T)
tum_counts = tum_counts[rowSums(tum_counts[, 3:6] > 0) == 2,] #need two alleles present still to phase

tum_allele = substr(as.vector(unlist(apply(tum_counts[, 3:6], 1, function(x) names(sort(x)[4])))), 7, 7)
normal_allele = substr(as.vector(unlist(apply(tum_counts[, 3:6], 1, function(x) names(sort(x)[3])))), 7, 7)

tum_counts$tum_allele = tum_allele
tum_counts$normal_allele = normal_allele
tum_counts$coord = paste0(tum_counts$X.CHR, "_", tum_counts$POS)

write.table(tum_counts, "PD51122m_chr17_hetSNPs_phased.txt", col.names = T, row.names = F, sep = "\t", quote = F)

```

###Step 5: Plot phased SNPs per sample

```{r}

library(ggplot2)
library(cowplot)
library(plyr)
library(zoo)

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/PD51122_phasing/chr17_allelecount")
options(stringsAsFactors = F)

allelecount_files = list.files(".", "_chr17_alleleFrequencies.txt")
phased_chr17 = read.table("PD51122m_chr17_hetSNPs_phased.txt", header = T, comment.char = "", sep = "\t")
row.names(phased_chr17) = phased_chr17$coord

manifest_wgs = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
manifest_wgs$exp_tum_allele_baf = NA
manifest_wgs$obs_tum_allele_baf = NA
manifest_wgs$exp.tum.baf.pval = NA

#file = "PD51122m_chr17_alleleFrequencies.txt" initial test
#file = "PD51122v_lo0003_chr17_alleleFrequencies.txt" wes test
#file = "PD51122ad_lo0002_chr17_alleleFrequencies.txt" normal, non CNS test
#file = "PD51122n_lo0014_chr17_alleleFrequencies.txt"
for(file in allelecount_files){
  
  sample = unlist(strsplit(file, "_chr17_alleleFrequencies.txt"))
  print(sample)
  
  data = read.table(file, sep = "\t", comment.char = "", header = T)
  data$coord = paste0(data$X.CHR, "_", data$POS)
  data_flt = data[data$coord %in% phased_chr17$coord,]
  data_flt$tum_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$tum_allele)
  data_flt$normal_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$normal_allele)
  
  data_flt$tum_count = as.numeric(apply(data_flt, 1, function(x) x[x[9]]))
  data_flt$normal_count = as.numeric(apply(data_flt, 1, function(x) x[x[10]]))
  
  data_flt = data_flt[data_flt$Good_depth >= 10,] #missing regions for WES
  
  data_flt$tum_VAF = data_flt$tum_count / data_flt$Good_depth
  data_flt$normal_VAF = data_flt$normal_count / data_flt$Good_depth
  data_flt$tum_VAF_rollmean = rollmean(data_flt$tum_VAF, 50, fill = NA)
  data_flt$normal_VAF_rollmean = rollmean(data_flt$normal_VAF, 50, fill = NA)
  
  data_flt_nf1 = data_flt[data_flt$POS >= 31094977 - 1.5e5 & data_flt$POS <= 31377675 + 1.5e5, ]
  data_flt_nf1$tum_VAF_rollmean = rollmean(data_flt_nf1$tum_VAF, 10, fill = NA)
  data_flt_nf1$normal_VAF_rollmean = rollmean(data_flt_nf1$normal_VAF, 10, fill = NA)
  
  p1 = ggplot(data_flt) +
    geom_point(mapping = aes(x = POS, y = Good_depth), pch = ".") +
    theme_bw() +
    ylab("Good depth") +
    xlab("") +
    geom_vline(xintercept = c(31094977, 31377675)) +
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          plot.margin = unit(c(0.5, 0.5, 0, 0.5), "cm"),
          plot.title = element_text(hjust = 0.5)) +
    ylim(0, round_any(mean(data_flt$Good_depth) + 6 * sd(data_flt$Good_depth), 100, f = ceiling)) +
    ggtitle(paste0(sample))
  p2 = ggplot(data_flt) +
    geom_point(mapping = aes(x = POS, y = normal_VAF), col = "blue", pch = ".", alpha = 0.1) +
    geom_point(mapping = aes(x = POS, y = tum_VAF), col = "red", pch = ".", alpha = 0.1) +
    geom_vline(xintercept = c(31094977, 31377675)) +
    ylab("BAF") +
    xlab("Chromosome 17") +
    theme_bw() +
    theme(plot.margin = unit(c(0, 0.5, 0.5, 0.5), "cm")) +
    ylim(c(0, 1))
  p2.5 = ggplot(data_flt) +
    geom_line(mapping = aes(x = POS, y = tum_VAF_rollmean), col = "red", alpha = 0.5) +
    geom_line(mapping = aes(x = POS, y = normal_VAF_rollmean), col = "blue", alpha = 0.5) +
    geom_vline(xintercept = c(31094977, 31377675)) +
    ylab("BAF") +
    xlab("Chromosome 17") +
    theme_bw() +
    theme(plot.title = element_text(hjust = 0.5)) +
    ylim(c(0, 1)) +
    ggtitle(paste0(sample))
  
  png(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/PD51122_phasing/phased_plots/", sample, "_phased_chr17_hetSNPs.png"), width = 12, height = 8, units = "cm", res = 800)
  print(plot_grid(p1, p2, align = "v", ncol = 1))
  dev.off()
  
  png(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/PD51122_phasing/phased_plots/", sample, "_phased_chr17_smoothed_BAF_hetSNPs.png"), width = 12, height = 8, units = "cm", res = 800)
  print(p2.5)
  dev.off()
  
  #tumour has 2+0 at NF1
  if(sample %in% manifest_wgs[manifest_wgs$keep == "Y",]$sample){
    
    exp_tum_allele_baf = manifest_wgs[manifest_wgs$sample == sample,]$purity_trunk + (1 - manifest_wgs[manifest_wgs$sample == sample,]$purity_trunk) * 0.5
    
    nf1_tum_allele_count = sum(data_flt[data_flt$POS >= 31094977 & data_flt$POS <= 31377675, ]$tum_count)
    nf1_total_depth_count = sum(data_flt[data_flt$POS >= 31094977 & data_flt$POS <= 31377675, ]$Good_depth)
    
    obs_tum_allele_baf = nf1_tum_allele_count / nf1_total_depth_count
    
    manifest_wgs[manifest_wgs$sample == sample,]$exp_tum_allele_baf = exp_tum_allele_baf
    manifest_wgs[manifest_wgs$sample == sample,]$obs_tum_allele_baf = obs_tum_allele_baf
    manifest_wgs[manifest_wgs$sample == sample,]$exp.tum.baf.pval = binom.test(nf1_tum_allele_count, nf1_total_depth_count, p = exp_tum_allele_baf)$p.value
    
    p3 = ggplot(data_flt_nf1) +
    geom_point(mapping = aes(x = POS, y = Good_depth), pch = 19) +
    theme_bw() +
    ylab("Good depth") +
    xlab("") +
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          plot.margin = unit(c(0.5, 0.5, 0, 0.5), "cm"),
          plot.title = element_text(hjust = 0.5)) +
    ylim(0, round_any(mean(data_flt$Good_depth) + 6 * sd(data_flt$Good_depth), 100, f = ceiling)) +
    geom_vline(xintercept = c(31094977, 31377675)) +
    geom_vline(xintercept = 31230383, col = "orange") +
    ggtitle(paste0(sample))
    
    p4 = ggplot(data_flt_nf1) +
      geom_point(mapping = aes(x = POS, y = normal_VAF), col = "dodgerblue", pch = 20, alpha = 0.15) +
      geom_point(mapping = aes(x = POS, y = tum_VAF), col = "tomato", pch = 20, alpha = 0.15) +
      ylab("BAF") +
      xlab("NF1 genomic coordinate") +
      theme_bw() +
      theme(plot.margin = unit(c(0, 0.5, 0.5, 0.5), "cm")) +
      ylim(c(0, 1)) +
      geom_line(mapping = aes(x = POS, y = tum_VAF_rollmean), col = "red") +
      geom_line(mapping = aes(x = POS, y = normal_VAF_rollmean), col = "blue") +
      geom_vline(xintercept = c(31094977, 31377675)) +
      geom_vline(xintercept = 31230383, col = "orange") +
      geom_hline(yintercept = exp_tum_allele_baf, lty = 2)
    
  } else {
    
     p3 = ggplot(data_flt_nf1) +
        geom_point(mapping = aes(x = POS, y = Good_depth), pch = 19) +
        theme_bw() +
        ylab("Good depth") +
        xlab("") +
        theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          plot.margin = unit(c(0.5, 0.5, 0, 0.5), "cm"),
          plot.title = element_text(hjust = 0.5)) +
        ylim(0, round_any(mean(data_flt$Good_depth) + 6 * sd(data_flt$Good_depth), 100, f = ceiling)) +
        geom_vline(xintercept = c(31094977, 31377675)) +
        geom_vline(xintercept = 31230383, col = "orange") +
        ggtitle(paste0(sample))
    
     p4 = ggplot(data_flt_nf1) +
        geom_point(mapping = aes(x = POS, y = normal_VAF), col = "dodgerblue", pch = 20, alpha = 0.15) +
        geom_point(mapping = aes(x = POS, y = tum_VAF), col = "tomato", pch = 20, alpha = 0.15) +
        ylab("BAF") +
        xlab("NF1 genomic coordinate") +
        theme_bw() +
        theme(plot.margin = unit(c(0, 0.5, 0.5, 0.5), "cm")) +
        ylim(c(0, 1)) +
        geom_line(mapping = aes(x = POS, y = tum_VAF_rollmean), col = "red") +
        geom_line(mapping = aes(x = POS, y = normal_VAF_rollmean), col = "blue") +
        geom_vline(xintercept = c(31094977, 31377675)) +
        geom_vline(xintercept = 31230383, col = "orange")
    
  }
  
  png(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/PD51122_phasing/phased_plots/", sample, "_phased_nf1_hetSNPs.png"), width = 12, height = 8, units = "cm", res = 800)
  print(plot_grid(p3, p4, align = "v", ncol = 1))
  dev.off()
  
}

manifest_wgs$exp.tum.baf.qval = p.adjust(manifest_wgs$exp.tum.baf.pval, "fdr")

write.table(manifest_wgs, '/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', sep = '\t', quote = F, col.names = T, row.names = F)

```

```{r}

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

manifest$independent_nf1_dh = F

manifest[manifest$exp.tum.baf.qval <0.001 & !is.na(manifest$exp.tum.baf.qval) & manifest$picard_median_coverage >= 30 & manifest$purity_trunk < 0.01 & (manifest$obs_tum_allele_baf > manifest$exp_tum_allele_baf),]$independent_nf1_dh = T
manifest[manifest$keep == "N",]$independent_nf1_dh = NA
manifest[manifest$sample %in% c("PD51122d", "PD51122g_lo0003", "PD51122b_lo0018"),]$independent_nf1_dh = T #sub and indel hits

write.table(manifest, '/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', sep = '\t', quote = F, col.names = T, row.names = F)

```

###Estimate NF1 DH clone sizes per sample

```{r}

manifest_wgs = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
manifest_wes = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wes_data/00_supplementary_data/wes_paeds_pm_sample_metadata_20220722.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

manifest_wgs$nf1_dh_clone_size = NA
manifest_wes$nf1_dh_clone_size = NA

manifest_wgs[manifest_wgs$independent_nf1_dh == T & !is.na(manifest_wgs$independent_nf1_dh), ]$sample
#probable mechanisms
#deletion 1+0 PD51122ac_lo0013
#LOH PD51122ag_lo0007 PD51122b PD51122e_lo0045 PD51122f PD51122h_lo0012 PD51122h_lo0015 PD51122p_lo0032
#sub/indel PD51122g_lo0003 PD51122b_lo0018 PD51122d
loh_wgs = c("PD51122ag_lo0007", "PD51122b", "PD51122e_lo0045", "PD51122f", "PD51122h_lo0012", "PD51122h_lo0015", "PD51122p_lo0032")

manifest_wes[manifest_wes$independent_nf1_dh == T & !is.na(manifest_wes$independent_nf1_dh), ]$sample
#LOH PD51122w_lo0038
#indel = PD51122w_lo0040
loh_wes = c("PD51122w_lo0038")

#calculate for LOH samples
manifest_wgs[manifest_wgs$sample %in% loh_wgs,]$nf1_dh_clone_size = manifest_wgs[manifest_wgs$sample %in% loh_wgs,]$obs_tum_allele_baf * 2 - 1
manifest_wes[manifest_wes$sample %in% loh_wes,]$nf1_dh_clone_size = manifest_wes[manifest_wes$sample %in% loh_wes,]$obs_tum_allele_baf * 2 - 1

#calculate for deletion
manifest_wgs[manifest_wgs$sample == "PD51122ac_lo0013",]$nf1_dh_clone_size = 2 - (1 / manifest_wgs[manifest_wgs$sample == "PD51122ac_lo0013",]$obs_tum_allele_baf)

#add for subs/indels
manifest_wgs[manifest_wgs$sample == "PD51122g_lo0003",]$nf1_dh_clone_size = 2 * 0.105263158
manifest_wgs[manifest_wgs$sample == "PD51122b_lo0018",]$nf1_dh_clone_size = 2 * 0.28125
manifest_wgs[manifest_wgs$sample == "PD51122d",]$nf1_dh_clone_size = 2 * 0.096045198
manifest_wes[manifest_wes$sample == "PD51122w_lo0040",]$nf1_dh_clone_size = 2 * 0.172413793

write.table(manifest_wgs, '/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', sep = '\t', quote = F, col.names = T, row.names = F)
write.table(manifest_wes, '/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wes_data/00_supplementary_data/wes_paeds_pm_sample_metadata_20220722.txt', sep = '\t', quote = F, col.names = T, row.names = F)

library(ggplot2)
library(ggbeeswarm)

ggplot(manifest_wgs[!is.na(manifest_wgs$nf1_dh_clone_size) & !manifest_wgs$sample %in% c("PD51122ac_lo0013", "PD51122ag_lo0007"),]) +
  geom_point(mapping = aes(x = nf1_dh_clone_size, y = sub_burden_tier1, col = picard_median_coverage)) +
  ylim(0, 100) #none have a subsituton count over 50

VAF_subs = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/PD51122_VAF_final_merged_snps.txt"), header = T, sep = "\t", stringsAsFactors = F)
VAF_indels = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_indels/PD51122_VAF_final_merged_indels.txt"), header = T, sep = "\t", stringsAsFactors = F)

#what samples have the
unique(c(colnames(VAF_subs)[VAF_subs["chr17_31200443_C_T",] > 0], colnames(VAF_subs)[VAF_subs["chr17_31214524_A_G",] > 0], colnames(VAF_indels)[VAF_indels["chr17_31226459_A_AC",] > 0]))

table(manifest_wgs[manifest_wgs$sample %in% unique(c(colnames(VAF_subs)[VAF_subs["chr17_31200443_C_T",] > 0], colnames(VAF_subs)[VAF_subs["chr17_31214524_A_G",] > 0], colnames(VAF_indels)[VAF_indels["chr17_31226459_A_AC",] > 0])),]$CNS)
table(manifest_wgs[manifest_wgs$keep == "Y" & manifest_wgs$case.id == "PD51122",]$CNS) #very similar ratio

table(manifest_wgs[manifest_wgs$sample %in% unique(c(colnames(VAF_subs)[VAF_subs["chr17_31200443_C_T",] > 0], colnames(VAF_subs)[VAF_subs["chr17_31214524_A_G",] > 0])),]$histo)

table(manifest_wgs[manifest_wgs$sample %in% unique(c(colnames(VAF_subs)[VAF_subs["chr17_31200443_C_T",] > 0], colnames(VAF_subs)[VAF_subs["chr17_31214524_A_G",] > 0], colnames(VAF_indels)[VAF_indels["chr17_31226459_A_AC",] > 0])),]$histo)

manifest_wgs[manifest_wgs$sample %in% unique(c(colnames(VAF_subs)[VAF_subs["chr17_31200443_C_T",] > 0], colnames(VAF_subs)[VAF_subs["chr17_31214524_A_G",] > 0], colnames(VAF_indels)[VAF_indels["chr17_31226459_A_AC",] > 0])) & manifest_wgs$histo == "Tumour",]

manifest_wgs[manifest_wgs$sample %in% unique(c(colnames(VAF_subs)[VAF_subs["chr17_31200443_C_T",] > 0], colnames(VAF_subs)[VAF_subs["chr17_31214524_A_G",] > 0])) & manifest_wgs$histo == "Tumour",]

VAF_subs[c("chr17_31200443_C_T", "chr17_31214524_A_G"), "PD51122ai_lo0025"] #0.03448276 0.00000000 very low level in diagnostic biopsy

row.names(manifest_wgs) = manifest_wgs$sample

df = data.frame(samples = colnames(VAF_subs), chr17_31200443_C_T = as.numeric(VAF_subs["chr17_31200443_C_T",]), chr17_31214524_A_G = as.numeric(VAF_subs["chr17_31214524_A_G",]), chr17_31226459_A_AC = as.numeric(VAF_indels["chr17_31226459_A_AC",]), bulk_tissue = manifest_wgs[colnames(VAF_subs),]$bulk_tissue, is.cns = manifest_wgs[colnames(VAF_subs),]$CNS)

ggplot(df) +
  geom_quasirandom(mapping = aes(x = bulk_tissue, y = chr17_31200443_C_T, col = is.cns)) +
  theme_bw() +
  ylab("chr17_31200443_C_T VAF") +
  xlab("") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
ggplot(df) +
  geom_quasirandom(mapping = aes(x = bulk_tissue, y = chr17_31214524_A_G, col = is.cns)) +
  theme_bw() +
  ylab("chr17_31214524_A_G VAF") +
  xlab("") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
ggplot(df) +
  geom_quasirandom(mapping = aes(x = bulk_tissue, y = chr17_31226459_A_AC, col = is.cns)) +
  theme_bw() +
  ylab("chr17_31226459_A_AC VAF") +
  xlab("") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

df[df$chr17_31226459_A_AC > 0.1,]

```

##Copy number changes across NF-1 lesional tissues (EDF5)

```{r}

library(ggplot2)
library(cowplot)
library(plyr)
library(zoo)

#example tumour PD51122m
#CALM PD51122ag_lo0007
#SCL PD51122ac_lo0013

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/PD51122_phasing/chr17_allelecount")
options(stringsAsFactors = F)

allelecount_files = list.files(".", "_chr17_alleleFrequencies.txt")
phased_chr17 = read.table("PD51122m_chr17_hetSNPs_phased.txt", header = T, comment.char = "", sep = "\t")
row.names(phased_chr17) = phased_chr17$coord

#PD51122m
sample = "PD51122m"
file = paste0(sample, "_chr17_alleleFrequencies.txt")
cn_calls = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/copy_number/ascatPCA/seg_100/sort_ascat_", sample), header = F, sep = "\t", stringsAsFactors = F)

data = read.table(file, sep = "\t", comment.char = "", header = T)
data$coord = paste0(data$X.CHR, "_", data$POS)
data_flt = data[data$coord %in% phased_chr17$coord,]
data_flt$tum_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$tum_allele)
data_flt$normal_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$normal_allele)
  
data_flt$tum_count = as.numeric(apply(data_flt, 1, function(x) x[x[9]]))
data_flt$normal_count = as.numeric(apply(data_flt, 1, function(x) x[x[10]]))
  
data_flt = data_flt[data_flt$Good_depth >= 10,] #missing regions for WES
  
data_flt$tum_VAF = data_flt$tum_count / data_flt$Good_depth
data_flt$normal_VAF = data_flt$normal_count / data_flt$Good_depth
data_flt$tum_VAF_rollmean = rollmean(data_flt$tum_VAF, 50, fill = NA)
data_flt$normal_VAF_rollmean = rollmean(data_flt$normal_VAF, 50, fill = NA)

p1 = ggplot(cn_calls[cn_calls$V1 == "chr17",]) +
            geom_segment(mapping = aes(x = V2, xend = V3, y = V4, yend = V4), size = 2) +
            geom_vline(xintercept = 31094977, lty = 2) +
            ylim(0, 4) +
            theme_bw() +
            theme(axis.text.x = element_blank(),
                  axis.text.y = element_blank(),
                  axis.ticks.x = element_blank(),
                  axis.title.x = element_blank(),
                  axis.title.y = element_blank(),
                  plot.margin = unit(c(0, 0, 0, 0), "cm"))
p2 = ggplot(data_flt) +
            geom_line(mapping = aes(x = POS, y = tum_VAF_rollmean), col = "#e83b3e", alpha = 0.8) +
            geom_line(mapping = aes(x = POS, y = normal_VAF_rollmean), col = "#4764ad", alpha = 0.8) +
            geom_vline(xintercept = 31094977, lty = 2) +
            theme_bw() +
            ylim(c(0, 1)) +
            theme(axis.text.x = element_blank(),
                  axis.text.y = element_blank(),
                  axis.ticks.x = element_blank(),
                  axis.title.x = element_blank(),
                  axis.title.y = element_blank(),
                  plot.margin = unit(c(0, 0, 0, 0), "cm"))

#PD51122ag_lo0007
sample = "PD51122ag_lo0007"
file = paste0(sample, "_chr17_alleleFrequencies.txt")
cn_calls = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/copy_number/ascatPCA/seg_100/sort_ascat_", sample), header = F, sep = "\t", stringsAsFactors = F)

data = read.table(file, sep = "\t", comment.char = "", header = T)
data$coord = paste0(data$X.CHR, "_", data$POS)
data_flt = data[data$coord %in% phased_chr17$coord,]
data_flt$tum_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$tum_allele)
data_flt$normal_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$normal_allele)
  
data_flt$tum_count = as.numeric(apply(data_flt, 1, function(x) x[x[9]]))
data_flt$normal_count = as.numeric(apply(data_flt, 1, function(x) x[x[10]]))
  
data_flt = data_flt[data_flt$Good_depth >= 10,] #missing regions for WES
  
data_flt$tum_VAF = data_flt$tum_count / data_flt$Good_depth
data_flt$normal_VAF = data_flt$normal_count / data_flt$Good_depth
data_flt$tum_VAF_rollmean = rollmean(data_flt$tum_VAF, 50, fill = NA)
data_flt$normal_VAF_rollmean = rollmean(data_flt$normal_VAF, 50, fill = NA)

p3 = ggplot(cn_calls[cn_calls$V1 == "chr17",]) +
            geom_segment(mapping = aes(x = V2, xend = V3, y = V4, yend = V4), size = 2) +
            geom_vline(xintercept = 31094977, lty = 2) +
            ylim(0, 4) +
            theme_bw() +
            theme(axis.text.x = element_blank(),
                  axis.text.y = element_blank(),
                  axis.ticks.x = element_blank(),
                  axis.title.x = element_blank(),
                  axis.title.y = element_blank(),
                  plot.margin = unit(c(0, 0, 0, 0), "cm"))
p4 = ggplot(data_flt) +
            geom_line(mapping = aes(x = POS, y = tum_VAF_rollmean), col = "#e83b3e", alpha = 0.8) +
            geom_line(mapping = aes(x = POS, y = normal_VAF_rollmean), col = "#4764ad", alpha = 0.8) +
            geom_vline(xintercept = 31094977, lty = 2) +
            theme_bw() +
            ylim(c(0, 1)) +
            theme(axis.text.x = element_blank(),
                  axis.text.y = element_blank(),
                  axis.ticks.x = element_blank(),
                  axis.title.x = element_blank(),
                  axis.title.y = element_blank(),
                  plot.margin = unit(c(0, 0, 0, 0), "cm"))

#PD51122ac_lo0013
sample = "PD51122ac_lo0013"
file = paste0(sample, "_chr17_alleleFrequencies.txt")
cn_calls = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/copy_number/ascatPCA/seg_100/sort_ascat_", sample), header = F, sep = "\t", stringsAsFactors = F)

data = read.table(file, sep = "\t", comment.char = "", header = T)
data$coord = paste0(data$X.CHR, "_", data$POS)
data_flt = data[data$coord %in% phased_chr17$coord,]
data_flt$tum_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$tum_allele)
data_flt$normal_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$normal_allele)
  
data_flt$tum_count = as.numeric(apply(data_flt, 1, function(x) x[x[9]]))
data_flt$normal_count = as.numeric(apply(data_flt, 1, function(x) x[x[10]]))
  
data_flt = data_flt[data_flt$Good_depth >= 10,] #missing regions for WES
  
data_flt$tum_VAF = data_flt$tum_count / data_flt$Good_depth
data_flt$normal_VAF = data_flt$normal_count / data_flt$Good_depth
data_flt$tum_VAF_rollmean = rollmean(data_flt$tum_VAF, 50, fill = NA)
data_flt$normal_VAF_rollmean = rollmean(data_flt$normal_VAF, 50, fill = NA)

p5 = ggplot(cn_calls[cn_calls$V1 == "chr17",]) +
            geom_segment(mapping = aes(x = V2, xend = V3, y = V4, yend = V4), size = 2) +
            geom_vline(xintercept = 31094977, lty = 2) +
            ylim(0, 4) +
            theme_bw() +
            theme(axis.text.x = element_blank(),
                  axis.text.y = element_blank(),
                  axis.ticks.x = element_blank(),
                  axis.title.x = element_blank(),
                  axis.title.y = element_blank(),
                  plot.margin = unit(c(0, 0, 0, 0), "cm"))
p6 = ggplot(data_flt) +
            geom_line(mapping = aes(x = POS, y = tum_VAF_rollmean), col = "#e83b3e", alpha = 0.8) +
            geom_line(mapping = aes(x = POS, y = normal_VAF_rollmean), col = "#4764ad", alpha = 0.8) +
            geom_vline(xintercept = 31094977, lty = 2) +
            theme_bw() +
            ylim(c(0, 1)) +
            theme(axis.text.x = element_blank(),
                  axis.text.y = element_blank(),
                  axis.ticks.x = element_blank(),
                  axis.title.x = element_blank(),
                  axis.title.y = element_blank(),
                  plot.margin = unit(c(0, 0, 0, 0), "cm"))

png("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/PD51122_chr17_lesions_20221207.png", width = 4, height = 6, units = "in", res = 300)
plot_grid(p1, p2, p3, p4, p5, p6, ncol = 1, align = "v")
dev.off()

```


```{r}

library(ggplot2)
library(cowplot)

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/PD51122_phasing/chr17_allelecount")

manifest_wgs = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

#example tumour PD51122m
#CALM PD51122ag_lo0007
#SCL PD51122ac_lo0013

#allelecount_files = list.files(".", "_chr17_alleleFrequencies.txt")
phased_chr17 = read.table("PD51122m_chr17_hetSNPs_phased.txt", header = T, comment.char = "", sep = "\t")
row.names(phased_chr17) = phased_chr17$coord

#PD51122m
sample = "PD51122m"

data = read.table(paste0(sample, "_chr17_alleleFrequencies.txt"), sep = "\t", comment.char = "", header = T)
data$coord = paste0(data$X.CHR, "_", data$POS)
data_flt = data[data$coord %in% phased_chr17$coord,]
data_flt$tum_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$tum_allele)
data_flt$normal_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$normal_allele)
  
data_flt$tum_count = as.numeric(apply(data_flt, 1, function(x) x[x[9]]))
data_flt$normal_count = as.numeric(apply(data_flt, 1, function(x) x[x[10]]))
  
data_flt = data_flt[data_flt$Good_depth >= 10,] #missing regions for WES
  
data_flt$tum_VAF = data_flt$tum_count / data_flt$Good_depth
data_flt$normal_VAF = data_flt$normal_count / data_flt$Good_depth

p1 = ggplot(data_flt) +
  geom_point(mapping = aes(x = POS, y = normal_VAF), col = "#4663ac", alpha = 0.7, pch = ".") +
  geom_point(mapping = aes(x = POS, y = tum_VAF), col = "#e83b3d", alpha = 0.7, pch = ".") +
  geom_vline(mapping = aes(xintercept = 31094977), lty = 2) +
  theme_bw()+
  theme(axis.text = element_blank(), 
        axis.ticks.x = element_blank(),
        plot.margin = unit(c(0, 0, 0, 0), "cm")) +
  xlab("") +
  ylab("")

ggplot(data_flt) +
  geom_point(mapping = aes(x = POS, y = Good_depth), col = "#999999", alpha = 0.7, pch = ".") +
  geom_vline(mapping = aes(xintercept = 31094977), lty = 2) +
  theme_bw()+
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(),
        plot.margin = unit(c(0, 0, 0, 0), "cm")) +
  xlab("") +
  ylim(0, 150)

#PD51122ag_lo0007
sample = "PD51122ag_lo0007"

data = read.table(paste0(sample, "_chr17_alleleFrequencies.txt"), sep = "\t", comment.char = "", header = T)
data$coord = paste0(data$X.CHR, "_", data$POS)
data_flt = data[data$coord %in% phased_chr17$coord,]
data_flt$tum_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$tum_allele)
data_flt$normal_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$normal_allele)
  
data_flt$tum_count = as.numeric(apply(data_flt, 1, function(x) x[x[9]]))
data_flt$normal_count = as.numeric(apply(data_flt, 1, function(x) x[x[10]]))
  
data_flt = data_flt[data_flt$Good_depth >= 10,] #missing regions for WES
  
data_flt$tum_VAF = data_flt$tum_count / data_flt$Good_depth
data_flt$normal_VAF = data_flt$normal_count / data_flt$Good_depth

p2 = ggplot(data_flt) +
  geom_point(mapping = aes(x = POS, y = normal_VAF), col = "#4663ac", alpha = 0.7, pch = ".") +
  geom_point(mapping = aes(x = POS, y = tum_VAF), col = "#e83b3d", alpha = 0.7, pch = ".") +
  geom_vline(mapping = aes(xintercept = 31094977), lty = 2) +
  theme_bw()+
  theme(axis.text = element_blank(), 
        axis.ticks.x = element_blank(),
        plot.margin = unit(c(0, 0, 0, 0), "cm")) +
  xlab("") +
  ylab("") +
  ylim(0.1, 0.9)

#PD51122h_lo0012
sample = "PD51122h_lo0012"

data = read.table(paste0(sample, "_chr17_alleleFrequencies.txt"), sep = "\t", comment.char = "", header = T)
data$coord = paste0(data$X.CHR, "_", data$POS)
data_flt = data[data$coord %in% phased_chr17$coord,]
data_flt$tum_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$tum_allele)
data_flt$normal_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$normal_allele)
  
data_flt$tum_count = as.numeric(apply(data_flt, 1, function(x) x[x[9]]))
data_flt$normal_count = as.numeric(apply(data_flt, 1, function(x) x[x[10]]))
  
data_flt = data_flt[data_flt$Good_depth >= 10,] #missing regions for WES
  
data_flt$tum_VAF = data_flt$tum_count / data_flt$Good_depth
data_flt$normal_VAF = data_flt$normal_count / data_flt$Good_depth

p3 = ggplot(data_flt) +
  geom_point(mapping = aes(x = POS, y = normal_VAF), col = "#4663ac", alpha = 0.5, pch = ".") +
  geom_point(mapping = aes(x = POS, y = tum_VAF), col = "#e83b3d", alpha = 0.5, pch = ".") +
  geom_vline(mapping = aes(xintercept = 31094977), lty = 2) +
  theme_bw() +
  theme(axis.text = element_blank(), 
        plot.margin = unit(c(0, 0, 0, 0), "cm")) +
  xlab("") +
  ylab("")

png("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/PD51122_chr17_BAF_plot_20221108.png", width = 3, height = 3, units = "in", res = 300)
plot_grid(p1, p2, p3, ncol = 1)
dev.off()

```

##Fig. 2a & c - NF1 LOH hetSNP allele density plots

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/PD51122_phasing/chr17_allelecount")

manifest_wgs = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
manifest_wes = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wes_data/00_supplementary_data/wes_paeds_pm_sample_metadata_20220722.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

manifest_wgs[manifest_wgs$independent_nf1_dh == T & !is.na(manifest_wgs$independent_nf1_dh) & manifest_wgs$exp.tum.baf.qval <0.001,]$sample
#"PD51122ac_lo0013" "PD51122ag_lo0007" "PD51122b"         "PD51122e_lo0045"  "PD51122f"         "PD51122h_lo0012"  "PD51122h_lo0015"  "PD51122p_lo0032" ignore first two as they are separate lesions
manifest_wes[manifest_wes$independent_nf1_dh == T & !is.na(manifest_wes$independent_nf1_dh) & manifest_wes$exp.tum.baf.qval <0.01,]$sample #"PD51122w_lo0038" #

sig_samples = c("PD51122b", "PD51122e_lo0045", "PD51122f", "PD51122h_lo0012", "PD51122h_lo0015", "PD51122p_lo0032", "PD51122w_lo0038", "PD51122ai_lo0005", "PD51122q") #add in tumour and blood to comparison

phased_chr17 = read.table("PD51122m_chr17_hetSNPs_phased.txt", header = T, comment.char = "", sep = "\t")
row.names(phased_chr17) = phased_chr17$coord

all_nf1_dh = rbind(manifest_wgs[manifest_wgs$independent_nf1_dh == T & !is.na(manifest_wgs$independent_nf1_dh) & manifest_wgs$exp.tum.baf.qval <0.001, c("sample", "exp_tum_allele_baf", "obs_tum_allele_baf", "exp.tum.baf.qval")],
                   manifest_wes[manifest_wes$independent_nf1_dh == T & !is.na(manifest_wes$independent_nf1_dh) & manifest_wes$exp.tum.baf.qval <0.01, c("sample", "exp_tum_allele_baf", "obs_tum_allele_baf", "exp.tum.baf.qval")])
all_nf1_dh = rbind(all_nf1_dh, manifest_wgs[manifest_wgs$sample == "PD51122q", c("sample", "exp_tum_allele_baf", "obs_tum_allele_baf", "exp.tum.baf.qval")])
all_nf1_dh = rbind(all_nf1_dh, manifest_wgs[manifest_wgs$sample == "PD51122ai_lo0005", c("sample", "exp_tum_allele_baf", "obs_tum_allele_baf", "exp.tum.baf.qval")])

sig_sample_df = data.frame()
for(sample in sig_samples){
  
  data = read.table(paste0(sample, "_chr17_alleleFrequencies.txt"), sep = "\t", comment.char = "", header = T)
  data$coord = paste0(data$X.CHR, "_", data$POS)
  data_flt = data[data$coord %in% phased_chr17$coord,]
  data_flt$tum_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$tum_allele)
  data_flt$normal_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$normal_allele)
  
  data_flt$tum_count = as.numeric(apply(data_flt, 1, function(x) x[x[9]]))
  data_flt$normal_count = as.numeric(apply(data_flt, 1, function(x) x[x[10]]))
  
  data_flt = data_flt[data_flt$Good_depth >= 10,] #missing regions for WES
  
  data_flt$tum_VAF = data_flt$tum_count / data_flt$Good_depth
  data_flt$normal_VAF = data_flt$normal_count / data_flt$Good_depth
  nf1_data_flt = data_flt[data_flt$POS >= 31094977 & data_flt$POS <= 31377675, ]
  nf1_data_flt$exp_baf = all_nf1_dh[all_nf1_dh$sample == sample,]$exp_tum_allele_baf
  nf1_data_flt$sample = sample
  
  sig_sample_df = rbind(sig_sample_df, nf1_data_flt)
}

sig_sample_df$sample = factor(sig_sample_df$sample, levels = c("PD51122b", "PD51122e_lo0045", "PD51122f", "PD51122h_lo0012", "PD51122h_lo0015", "PD51122p_lo0032", "PD51122w_lo0038", "PD51122ai_lo0005", "PD51122q"))

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/non_tumour_independent_nf1_hits_density_plots_20221220.pdf", height = 4.5, width = 5, useDingbats = F)
ggplot(sig_sample_df) +
  geom_density(mapping = aes(x = normal_VAF), fill = "#4663ac", alpha = 0.7) +
  geom_density(mapping = aes(x = tum_VAF), fill = "#e83b3d", alpha = 0.7) +
  facet_wrap(. ~ sample, ncol = 4) +
  theme_bw() +
  geom_vline(mapping = aes(xintercept = exp_baf), lty = 2) +
  xlab("BAF") +
  ylab("Density") +
  theme(strip.background = element_rect(fill = "black"),
        strip.text = element_text(colour = "white", ),
        panel.grid = element_blank())
dev.off()

```

##Fig. 2a & d - chromosome 17 BAF plots

```{r}

library(ggplot2)
library(cowplot)

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/PD51122_phasing/chr17_allelecount")

#allelecount_files = list.files(".", "_chr17_alleleFrequencies.txt")
phased_chr17 = read.table("PD51122m_chr17_hetSNPs_phased.txt", header = T, comment.char = "", sep = "\t")
row.names(phased_chr17) = phased_chr17$coord

#PD51122m
sample = "PD51122m"

data = read.table(paste0(sample, "_chr17_alleleFrequencies.txt"), sep = "\t", comment.char = "", header = T)
data$coord = paste0(data$X.CHR, "_", data$POS)
data_flt = data[data$coord %in% phased_chr17$coord,]
data_flt$tum_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$tum_allele)
data_flt$normal_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$normal_allele)
  
data_flt$tum_count = as.numeric(apply(data_flt, 1, function(x) x[x[9]]))
data_flt$normal_count = as.numeric(apply(data_flt, 1, function(x) x[x[10]]))
  
data_flt = data_flt[data_flt$Good_depth >= 10,] #missing regions for WES
  
data_flt$tum_VAF = data_flt$tum_count / data_flt$Good_depth
data_flt$normal_VAF = data_flt$normal_count / data_flt$Good_depth

p1 = ggplot(data_flt) +
  geom_point(mapping = aes(x = POS, y = normal_VAF), col = "#4663ac", alpha = 0.7, pch = ".") +
  geom_point(mapping = aes(x = POS, y = tum_VAF), col = "#e83b3d", alpha = 0.7, pch = ".") +
  geom_vline(mapping = aes(xintercept = 31094977), lty = 2) +
  theme_bw()+
  theme(axis.text = element_blank(), 
        axis.ticks.x = element_blank(),
        plot.margin = unit(c(0, 0, 0, 0), "cm")) +
  xlab("") +
  ylab("")

#PD51122e_lo0045
sample = "PD51122e_lo0045"

data = read.table(paste0(sample, "_chr17_alleleFrequencies.txt"), sep = "\t", comment.char = "", header = T)
data$coord = paste0(data$X.CHR, "_", data$POS)
data_flt = data[data$coord %in% phased_chr17$coord,]
data_flt$tum_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$tum_allele)
data_flt$normal_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$normal_allele)
  
data_flt$tum_count = as.numeric(apply(data_flt, 1, function(x) x[x[9]]))
data_flt$normal_count = as.numeric(apply(data_flt, 1, function(x) x[x[10]]))
  
data_flt = data_flt[data_flt$Good_depth >= 10,] #missing regions for WES
  
data_flt$tum_VAF = data_flt$tum_count / data_flt$Good_depth
data_flt$normal_VAF = data_flt$normal_count / data_flt$Good_depth

p2 = ggplot(data_flt) +
  geom_point(mapping = aes(x = POS, y = normal_VAF), col = "#4663ac", alpha = 0.7, pch = ".") +
  geom_point(mapping = aes(x = POS, y = tum_VAF), col = "#e83b3d", alpha = 0.7, pch = ".") +
  geom_vline(mapping = aes(xintercept = 31094977), lty = 2) +
  theme_bw()+
  theme(axis.text = element_blank(), 
        axis.ticks.x = element_blank(),
        plot.margin = unit(c(0, 0, 0, 0), "cm")) +
  xlab("") +
  ylab("")

#PD51122h_lo0012
sample = "PD51122h_lo0012"

data = read.table(paste0(sample, "_chr17_alleleFrequencies.txt"), sep = "\t", comment.char = "", header = T)
data$coord = paste0(data$X.CHR, "_", data$POS)
data_flt = data[data$coord %in% phased_chr17$coord,]
data_flt$tum_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$tum_allele)
data_flt$normal_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$normal_allele)
  
data_flt$tum_count = as.numeric(apply(data_flt, 1, function(x) x[x[9]]))
data_flt$normal_count = as.numeric(apply(data_flt, 1, function(x) x[x[10]]))
  
data_flt = data_flt[data_flt$Good_depth >= 10,] #missing regions for WES
  
data_flt$tum_VAF = data_flt$tum_count / data_flt$Good_depth
data_flt$normal_VAF = data_flt$normal_count / data_flt$Good_depth

p3 = ggplot(data_flt) +
  geom_point(mapping = aes(x = POS, y = normal_VAF), col = "#4663ac", alpha = 0.5, pch = ".") +
  geom_point(mapping = aes(x = POS, y = tum_VAF), col = "#e83b3d", alpha = 0.5, pch = ".") +
  geom_vline(mapping = aes(xintercept = 31094977), lty = 2) +
  theme_bw() +
  theme(axis.text = element_blank(), 
        plot.margin = unit(c(0, 0, 0, 0), "cm")) +
  xlab("") +
  ylab("")

png("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/PD51122_chr17_BAF_plot_20221108.png", width = 3, height = 3, units = "in", res = 300)
plot_grid(p1, p2, p3, ncol = 1)
dev.off()

```

##Fig. 2e - Normal brain burden vs median VAF

```{r}

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

normal_brain = manifest[manifest$keep == "Y" & 
                          manifest$purity_trunk < 0.01 &
                          manifest$CNS == "Y",]

library(ggplot2)
library(cowplot)

p1 = ggplot() +
      geom_point(normal_brain[normal_brain$experimental_arm == "WGS_LCM",], mapping = aes(x = median_sub_VAF_tier1, y = sub_burden_tier1), pch = 21, cex = 100, stroke = NA, alpha = 0.8, fill = "#999999") +
      geom_point(normal_brain[normal_brain$experimental_arm == "WGS_LCM" & normal_brain$independent_nf1_dh == T,], mapping = aes(x = median_sub_VAF_tier1, y = sub_burden_tier1), col = "#e83b3d", size = 5) +
      theme_bw() +
      xlab("Median substitution VAF") +
      ylab("Substitution count") +
  scale_x_continuous(breaks = c(0.1, 0.3, 0.5), limits = c(0.05, 0.55)) +
  scale_y_continuous(breaks = c(0, 50, 100), limits = c(0, 100)) +
  xlab("") +
  ylab("Substitution count") + 
  theme(plot.margin = unit(c(0, 0, 0, 0), "cm"))

p2 = ggplot() +
      geom_point(normal_brain[normal_brain$experimental_arm == "WGS_bulk",], mapping = aes(x = median_sub_VAF_tier1, y = sub_burden_tier1), pch = 21, cex = 100, stroke = NA, alpha = 0.8, fill = "#999999") +
      geom_point(normal_brain[normal_brain$experimental_arm == "WGS_bulk" & normal_brain$independent_nf1_dh == T,], mapping = aes(x = median_sub_VAF_tier1, y = sub_burden_tier1), col = "#e83b3d", size = 5) +
      theme_bw() +
          xlab("Median substitution VAF") +
          ylab("") +
      scale_y_continuous(breaks = c(0, 50, 100), limits = c(0, 100)) +
      scale_x_continuous(breaks = c(0.1, 0.3, 0.5), limits = c(0.05, 0.55)) +
  theme(plot.margin = unit(c(0, 0, 0, 0), "cm"),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/normal_brain_burden_vs_vaf_20221122.pdf", height = 2.5, width = 5, useDingbats = F)
plot_grid(p1, p2, ncol = 2, align = "vh")
dev.off()

```

##Fig. 2f - somatic reversion evidence

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/PD51122_phasing/chr17_allelecount")

#allelecount_files = list.files(".", "_chr17_alleleFrequencies.txt")
phased_chr17 = read.table("PD51122m_chr17_hetSNPs_phased.txt", header = T, comment.char = "", sep = "\t")
row.names(phased_chr17) = phased_chr17$coord

sample = "PD51122s_lo0012"

data = read.table(paste0(sample, "_chr17_alleleFrequencies.txt"), sep = "\t", comment.char = "", header = T)
data$coord = paste0(data$X.CHR, "_", data$POS)
data_flt = data[data$coord %in% phased_chr17$coord,]
data_flt$tum_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$tum_allele)
data_flt$normal_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$normal_allele)
  
data_flt$tum_count = as.numeric(apply(data_flt, 1, function(x) x[x[9]]))
data_flt$normal_count = as.numeric(apply(data_flt, 1, function(x) x[x[10]]))
  
data_flt = data_flt[data_flt$Good_depth >= 10,] #missing regions for WES
  
data_flt$tum_VAF = data_flt$tum_count / data_flt$Good_depth
data_flt$normal_VAF = data_flt$normal_count / data_flt$Good_depth

res <- binom.test(sum(data_flt$tum_count), sum(data_flt$normal_count + data_flt$tum_count), p = 0.5)
res$p.value #4.446591e-323

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/PD51122s_lo0012_chr17_hetSNP_BAF_density_plot_20221109.pdf", height = 2, width = 2, useDingbats = F)
ggplot(data_flt) +
  geom_density(mapping = aes(x = tum_VAF), fill = "#e83b3d", alpha = 0.7) +
  geom_density(mapping = aes(x = normal_VAF), fill = "#4663ac", alpha = 0.7) +
  theme_bw() +
  geom_vline(mapping = aes(xintercept = 0.5), lty = 2) +
  xlab("BAF") +
  ylab("Density")
dev.off()

sample = "PD51122u_lo0009"

data = read.table(paste0(sample, "_chr17_alleleFrequencies.txt"), sep = "\t", comment.char = "", header = T)
data$coord = paste0(data$X.CHR, "_", data$POS)
data_flt = data[data$coord %in% phased_chr17$coord,]
data_flt$tum_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$tum_allele)
data_flt$normal_allele = paste0("Count_", phased_chr17[data_flt$coord, ]$normal_allele)
  
data_flt$tum_count = as.numeric(apply(data_flt, 1, function(x) x[x[9]]))
data_flt$normal_count = as.numeric(apply(data_flt, 1, function(x) x[x[10]]))
  
data_flt = data_flt[data_flt$Good_depth >= 10,] #missing regions for WES
  
data_flt$tum_VAF = data_flt$tum_count / data_flt$Good_depth
data_flt$normal_VAF = data_flt$normal_count / data_flt$Good_depth

res <- binom.test(sum(data_flt$tum_count), sum(data_flt$normal_count + data_flt$tum_count), p = 0.5)
res$p.value #0.00741927

ggplot(data_flt) +
  geom_density(mapping = aes(x = tum_VAF), fill = "#e83b3d", alpha = 0.7) +
  geom_density(mapping = aes(x = normal_VAF), fill = "#4663ac", alpha = 0.7) +
  theme_bw() +
  geom_vline(mapping = aes(xintercept = 0.5), lty = 2) +
  xlab("BAF") +
  ylab("Density")

```

##Identify mutations at the root of the tumour

1) Identify truncal mutations:
  i) Anything assigned to the root cluster in the phylogeny that is found in a very high confidence normal sample (0% tumour involvement, 30+ coverage)
  ii) Anything called in all bulk tumour samples that is also called in one of the aforementioned samples (mutations found via this route are seen in samples with coverage in excess of 60+)

###PD50297

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars")

patient = "PD50297"
truncal_cluster = 9

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

tumour_bulk_samples = manifest[manifest$case.id == patient & manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$experimental_arm == "WGS_bulk",]$sample
non_tumour_samples = manifest[manifest$case.id == patient & manifest$keep == "Y" & manifest$purity_trunk == 0 & manifest$picard_median_coverage >= 30,]$sample #high depth so that low depth with low level tumour purity not misidentified as shared

#get variants in trunk
vars = read.table(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient, '/md_out/', patient, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)
cluster_vars = vars[vars$cluster.no == truncal_cluster,]
cluster_vars$ID = paste0(cluster_vars$chr, "_", cluster_vars$pos)

truncal_vars = paste0("chr", cluster_vars$ID)
length(truncal_vars) #929

#read in VAF dataframe
VAF = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/", patient, "_VAF_final_merged_snps.txt"), header = T, sep = "\t", stringsAsFactors = F)

VAF_filter = VAF
VAF_filter$ID = paste0(unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 1)), "_", unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 2)))

trunc_normal_VAF = VAF[VAF_filter$ID %in% truncal_vars, non_tumour_samples]
nrow(trunc_normal_VAF) #929
#put_ee_vars = row.names(trunc_normal_VAF)[rowSums(trunc_normal_VAF >= 0.1) > 0] #greater or equal to
put_ee_vars = row.names(trunc_normal_VAF)[rowSums(trunc_normal_VAF > 0.1) > 0] #greater than

#intersection method
sub.files = list.files("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/subs/07_indel", ".standard.gnomAD.AF.0.001.PON.read.direction.low_cov.bq.mq.binom.bbinom.5bp.indel.filtered.txt", full.names = T)
patient.sub.files = sub.files[grepl(patient, sub.files)]

manifest_flt = manifest[manifest$keep == "Y" & manifest$case.id == patient,]

#read in patient subs
all_subs = data.frame()
for(file in patient.sub.files){

  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F, colClasses = c("character", "character", "character", "numeric", "character", "character", "numeric", "numeric", "numeric", "character", "character", "character", "character", "character", "character"))
  all_subs = rbind(all_subs, data)
}

all_subs$mutID = paste(all_subs$Chr, all_subs$Position, all_subs$Wildtype, all_subs$Mutant, sep = "_")
all_subs_flt = all_subs[all_subs$Sample %in% manifest_flt$sample,]

intersect_trunk_vars = names(table(all_subs_flt[all_subs_flt$Sample %in% tumour_bulk_samples,]$mutID)[table(all_subs_flt[all_subs_flt$Sample %in% tumour_bulk_samples,]$mutID) == length(tumour_bulk_samples)])
intersect_root_vars = unique(all_subs_flt[all_subs_flt$mutID %in% intersect_trunk_vars & all_subs_flt$Sample %in% non_tumour_samples,]$mutID)

aggregate(VAF ~ mutID, all_subs_flt[all_subs_flt$mutID %in% intersect_root_vars & all_subs_flt$Sample %in% non_tumour_samples,], max) #3 are seen at a VAF less than 0.1 - where are these found?

#                 mutID        VAF
#  chr12_112505263_G_A 0.09000000
#   chr17_67362034_C_A 0.69696970
#   chr17_72915405_C_A 0.13953488
#   chr19_37176351_C_T 0.12195122
#    chr4_68023594_C_T 0.56000000
#   chr6_120782973_G_A 0.18367347
#    chr6_78642457_C_T 0.08823529
#   chr7_143282253_C_T 0.08163265
#   chr7_146198061_C_T 0.20588235
#   chr8_87155015_G_A 0.47058824
#   chrX_52010308_C_T 1.00000000

all_subs_flt[all_subs_flt$mutID %in% c("chr12_112505263_G_A", "chr6_78642457_C_T", "chr7_143282253_C_T") & all_subs_flt$Sample %in% non_tumour_samples,] #all have read depths much higher than 30x
min(all_subs_flt[all_subs_flt$mutID %in% c("chr12_112505263_G_A", "chr6_78642457_C_T", "chr7_143282253_C_T") & all_subs_flt$Sample %in% non_tumour_samples,]$Total_read_depth) #68

all_put_root_vars = sort(unique(c(put_ee_vars, intersect_root_vars)))

write.table(VAF[all_put_root_vars, ], paste0(patient, "_put_tum_root_snp_combined_method.txt"), col.names = T, row.names = T, sep = "\t", quote = F)

```

###PD51122

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars")

patient = "PD51122"
truncal_cluster = "6_9"

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

tumour_bulk_samples = manifest[manifest$case.id == patient & manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$experimental_arm == "WGS_bulk",]$sample
non_tumour_samples = manifest[manifest$case.id == patient & manifest$keep == "Y" & manifest$purity_trunk == 0 & manifest$picard_median_coverage >= 30,]$sample #high depth so that low depth with low level tumour purity not misidentified as shared

#get variants in trunk
vars = read.table(paste0('/lustre/scratch119/realdata/mdt1/team274/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient, '/md_out/', patient, '_allClusterassignmentsFromParallelRuns_clusters_reviewed.txt'), header = T, sep = '\t', stringsAsFactors = F) # truncal cluster different from the original output
cluster_vars = vars[vars$cluster.no == truncal_cluster,]
cluster_vars$ID = paste0(cluster_vars$chr, "_", cluster_vars$pos)

truncal_vars = paste0("chr", cluster_vars$ID)
length(truncal_vars) #972

#read in VAF dataframe
VAF = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/", patient, "_VAF_final_merged_snps.txt"), header = T, sep = "\t", stringsAsFactors = F)

VAF_filter = VAF
VAF_filter$ID = paste0(unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 1)), "_", unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 2)))

trunc_normal_VAF = VAF[VAF_filter$ID %in% truncal_vars, non_tumour_samples]
nrow(trunc_normal_VAF) #972
#put_ee_vars = row.names(trunc_normal_VAF)[rowSums(trunc_normal_VAF >= 0.1) > 0] #greater or equal to
put_ee_vars = row.names(trunc_normal_VAF)[rowSums(trunc_normal_VAF > 0.1) > 0] #greater than

#intersection method
sub.files = list.files("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/subs/07_indel", ".standard.gnomAD.AF.0.001.PON.read.direction.low_cov.bq.mq.binom.bbinom.5bp.indel.filtered.txt", full.names = T)
patient.sub.files = sub.files[grepl(patient, sub.files)]

manifest_flt = manifest[manifest$keep == "Y" & manifest$case.id == patient,]

#read in patient subs
all_subs = data.frame()
for(file in patient.sub.files){

  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F, colClasses = c("character", "character", "character", "numeric", "character", "character", "numeric", "numeric", "numeric", "character", "character", "character", "character", "character", "character"))
  all_subs = rbind(all_subs, data)
}

all_subs$mutID = paste(all_subs$Chr, all_subs$Position, all_subs$Wildtype, all_subs$Mutant, sep = "_")
all_subs_flt = all_subs[all_subs$Sample %in% manifest_flt$sample,]

intersect_trunk_vars = names(table(all_subs_flt[all_subs_flt$Sample %in% tumour_bulk_samples,]$mutID)[table(all_subs_flt[all_subs_flt$Sample %in% tumour_bulk_samples,]$mutID) == length(tumour_bulk_samples)])
intersect_root_vars = unique(all_subs_flt[all_subs_flt$mutID %in% intersect_trunk_vars & all_subs_flt$Sample %in% non_tumour_samples,]$mutID)

aggregate(VAF ~ mutID, all_subs_flt[all_subs_flt$mutID %in% intersect_root_vars & all_subs_flt$Sample %in% non_tumour_samples,], max) #1 is seen at a VAF less than 0.1 - where is this found?

#               mutID       VAF
#chr1_169603900_G_T 0.5909091
#chr15_59361799_T_A 0.4117647
# chr17_4949465_C_A 0.0776699
#chr8_133329433_A_G 0.3703704

all_subs_flt[all_subs_flt$mutID %in% c("chr17_4949465_C_A") & all_subs_flt$Sample %in% non_tumour_samples,] #read depths much higher than 30x
min(all_subs_flt[all_subs_flt$mutID %in% c("chr17_4949465_C_A") & all_subs_flt$Sample %in% non_tumour_samples,]$Total_read_depth) #103

all_put_root_vars = sort(unique(c(put_ee_vars, intersect_root_vars)))

write.table(VAF[all_put_root_vars, ], paste0(patient, "_put_tum_root_snp_combined_method.txt"), col.names = T, row.names = T, sep = "\t", quote = F)

```

###PD51123

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars")

patient = "PD51123"
truncal_cluster = 2

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

tumour_bulk_samples = manifest[manifest$case.id == patient & manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$experimental_arm == "WGS_bulk",]$sample
non_tumour_samples = manifest[manifest$case.id == patient & manifest$keep == "Y" & manifest$purity_trunk == 0 & manifest$picard_median_coverage >= 30,]$sample #high depth so that low depth with low level tumour purity not misidentified as shared

#get variants in trunk
vars = read.table(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient, '/md_out/', patient, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)
cluster_vars = vars[vars$cluster.no == truncal_cluster,]
cluster_vars$ID = paste0(cluster_vars$chr, "_", cluster_vars$pos)

truncal_vars = paste0("chr", cluster_vars$ID)
length(truncal_vars) #699

#read in VAF dataframe
VAF = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/", patient, "_VAF_final_merged_snps.txt"), header = T, sep = "\t", stringsAsFactors = F)

VAF_filter = VAF
VAF_filter$ID = paste0(unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 1)), "_", unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 2)))

trunc_normal_VAF = VAF[VAF_filter$ID %in% truncal_vars, non_tumour_samples]
nrow(trunc_normal_VAF) #699
#put_ee_vars = row.names(trunc_normal_VAF)[rowSums(trunc_normal_VAF >= 0.1) > 0] #greater or equal to
put_ee_vars = row.names(trunc_normal_VAF)[rowSums(trunc_normal_VAF > 0.1) > 0] #greater than

#intersection method
sub.files = list.files("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/subs/07_indel", ".standard.gnomAD.AF.0.001.PON.read.direction.low_cov.bq.mq.binom.bbinom.5bp.indel.filtered.txt", full.names = T)
patient.sub.files = sub.files[grepl(patient, sub.files)]

manifest_flt = manifest[manifest$keep == "Y" & manifest$case.id == patient,]

#read in patient subs
all_subs = data.frame()
for(file in patient.sub.files){

  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F, colClasses = c("character", "character", "character", "numeric", "character", "character", "numeric", "numeric", "numeric", "character", "character", "character", "character", "character", "character"))
  all_subs = rbind(all_subs, data)
}

all_subs$mutID = paste(all_subs$Chr, all_subs$Position, all_subs$Wildtype, all_subs$Mutant, sep = "_")
all_subs_flt = all_subs[all_subs$Sample %in% manifest_flt$sample,]

intersect_trunk_vars = names(table(all_subs_flt[all_subs_flt$Sample %in% tumour_bulk_samples,]$mutID)[table(all_subs_flt[all_subs_flt$Sample %in% tumour_bulk_samples,]$mutID) == length(tumour_bulk_samples)])
intersect_root_vars = unique(all_subs_flt[all_subs_flt$mutID %in% intersect_trunk_vars & all_subs_flt$Sample %in% non_tumour_samples,]$mutID)

aggregate(VAF ~ mutID, all_subs_flt[all_subs_flt$mutID %in% intersect_root_vars & all_subs_flt$Sample %in% non_tumour_samples,], max) #none seen at a max VAF <0.1
#mutID       VAF
#chr1_162729402_A_G 0.6000000
#chr6_14609040_A_G 0.5142857
#chr7_157223062_G_A 0.6296296
#chr7_158935187_G_A 0.1315789
#chr7_159159464_G_A 0.1282051
#chr9_132072514_C_A 0.6923077

all_put_root_vars = sort(unique(c(put_ee_vars, intersect_root_vars)))

write.table(VAF[all_put_root_vars, ], paste0(patient, "_put_tum_root_snp_combined_method.txt"), col.names = T, row.names = T, sep = "\t", quote = F)

```

##Fig. 3b & EDF6 - distribution of shared mutations across normal tissues

All mutations called in non tumour samples, with their distribution assessed in LCM samples only.

```{r}

#get all mutations in non tumour samples
setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/subs/07_indel")

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

norm_manifest = manifest[manifest$keep == "Y" &
                          manifest$purity_trunk < 0.01,]

norm_subs = data.frame()
for(sample in norm_manifest$sample){
  print(sample)
  data = read.table(paste0(sample, ".standard.gnomAD.AF.0.001.PON.read.direction.low_cov.bq.mq.binom.bbinom.5bp.indel.filtered.tier.annot.txt"), header = T, sep = '\t', stringsAsFactors = F)
  norm_subs = rbind(norm_subs, data)
}

tumour_manifest = manifest[manifest$keep == "Y" &
                              manifest$purity_trunk >= 0.4,]

tum_subs = data.frame()
for(sample in tumour_manifest$sample){
  print(sample)
  data = read.table(paste0(sample, ".standard.gnomAD.AF.0.001.PON.read.direction.low_cov.bq.mq.binom.bbinom.5bp.indel.filtered.tier.annot.txt"), header = T, sep = '\t', stringsAsFactors = F)
  tum_subs = rbind(tum_subs, data)
}

#check how many variants in normal variants are found in pure tumour samples
length(unique(norm_subs[norm_subs$mutID %in% tum_subs$mutID,]$mutID)) #56
length(unique(norm_subs$mutID)) #29082

#read in muts tables
pd50297_df = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/shearwater_framework/PD50297/shearwater_combined_snv_qval_pass_mat.txt", header = T, sep = " ", stringsAsFactors = F)

pd51122_df = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/shearwater_framework/PD51122/shearwater_combined_snv_qval_pass_mat.txt", header = T, sep = " ", stringsAsFactors = F)

pd51123_df = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/shearwater_framework/PD51123/shearwater_combined_snv_qval_pass_mat.txt", header = T, sep = " ", stringsAsFactors = F)

ncol(pd50297_df) + ncol(pd51122_df) + ncol(pd51123_df) #910, correct

#PD50297
lcm_samples = norm_manifest[norm_manifest$case.id == "PD50297" & norm_manifest$experimental_arm == "WGS_LCM",]$sample #216
muts = sort(unique(norm_subs[norm_subs$Case_ID == "PD50297",]$mutID)) #22189
snps = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars/PD50297_put_tum_root_snp_combined_method_reviewed.txt"), header = T, sep = "\t", stringsAsFactors = F, row.names = 1)
snps_flt = row.names(snps[snps$Keep == "Y",])
case.norm.mut.df = pd50297_df[sort(unique(c(muts, snps_flt))), lcm_samples] #22192
lcm_crypts = norm_manifest[norm_manifest$case.id == "PD50297" & 
                             norm_manifest$experimental_arm == "WGS_LCM" &
                             norm_manifest$bulk_tissue %in% c("Appendix", "Large intestine", "Small intestine") & 
                             norm_manifest$histo %in% c("Crypt", "Epithelium"),]$sample
lcm_non_crypt = lcm_samples[!lcm_samples %in% lcm_crypts]

#either
#3+ crypts
#2+ non crypt cuts
#1+ crypt and 1+ non crypt
#in tumour root
case.norm.mut.df.shared = case.norm.mut.df[((rowSums(case.norm.mut.df[, lcm_crypts] > 0)) > 2 | 
                                             (rowSums(case.norm.mut.df[, lcm_non_crypt] > 0)) > 1 | 
                                             ((rowSums(case.norm.mut.df[, lcm_crypts] > 0)) > 0 & (rowSums(case.norm.mut.df[, lcm_non_crypt] > 0)) > 0) |
                                             row.names(case.norm.mut.df) %in% snps_flt),] #343

write.table(case.norm.mut.df.shared, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/PD50297_norm_sample_shared_mut_df.txt", col.names = T, row.names = T, sep = "\t", quote = F)

#PD51122
lcm_samples = norm_manifest[norm_manifest$case.id == "PD51122" & norm_manifest$experimental_arm == "WGS_LCM",]$sample #269
muts = unique(norm_subs[norm_subs$Case_ID == "PD51122",]$mutID) #3354
snps = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars/PD51122_put_tum_root_snp_combined_method_reviewed.txt"), header = T, sep = "\t", stringsAsFactors = F, row.names = 1)
snps_flt = row.names(snps[snps$Keep == "Y",])
case.norm.mut.df = pd51122_df[sort(unique(c(muts, snps_flt))), lcm_samples]
lcm_crypts = norm_manifest[norm_manifest$case.id == "PD51122" & 
                             norm_manifest$experimental_arm == "WGS_LCM" &
                             norm_manifest$bulk_tissue %in% c("Appendix", "Large intestine", "Small intestine") & 
                             norm_manifest$histo %in% c("Crypt", "Epithelium"),]$sample
lcm_non_crypt = lcm_samples[!lcm_samples %in% lcm_crypts]

#either
#3+ crypts
#2+ non crypt cuts
#1+ crypt and 1+ non crypt
#in tumour root
case.norm.mut.df.shared = case.norm.mut.df[((rowSums(case.norm.mut.df[, lcm_crypts] > 0)) > 2 | 
                                             (rowSums(case.norm.mut.df[, lcm_non_crypt] > 0)) > 1 | 
                                             ((rowSums(case.norm.mut.df[, lcm_crypts] > 0)) > 0 & (rowSums(case.norm.mut.df[, lcm_non_crypt] > 0)) > 0) |
                                             row.names(case.norm.mut.df) %in% snps_flt),] #153
write.table(case.norm.mut.df.shared, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/PD51122_norm_sample_shared_mut_df.txt", col.names = T, row.names = T, sep = "\t", quote = F)

#PD51123
lcm_samples = norm_manifest[norm_manifest$case.id == "PD51123" & norm_manifest$experimental_arm == "WGS_LCM",]$sample #117
muts = unique(norm_subs[norm_subs$Case_ID == "PD51123",]$mutID) #3539
snps = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars/PD51123_put_tum_root_snp_combined_method_reviewed.txt"), header = T, sep = "\t", stringsAsFactors = F, row.names = 1)
snps_flt = row.names(snps[snps$Keep == "Y",])
case.norm.mut.df = pd51123_df[sort(unique(c(muts, snps_flt))), lcm_samples]
lcm_crypts = norm_manifest[norm_manifest$case.id == "PD51123" & 
                             norm_manifest$experimental_arm == "WGS_LCM" &
                             norm_manifest$bulk_tissue %in% c("Appendix", "Large intestine", "Small intestine") & 
                             norm_manifest$histo %in% c("Crypt", "Epithelium"),]$sample #none
lcm_non_crypt = lcm_samples[!lcm_samples %in% lcm_crypts]

#either
#3+ crypts
#2+ non crypt cuts
#1+ crypt and 1+ non crypt
#in tumour root
case.norm.mut.df.shared = case.norm.mut.df[((rowSums(case.norm.mut.df[, lcm_crypts] > 0)) > 2 | 
                                             (rowSums(case.norm.mut.df[, lcm_non_crypt] > 0)) > 1 | 
                                             ((rowSums(case.norm.mut.df[, lcm_crypts] > 0)) > 0 & (rowSums(case.norm.mut.df[, lcm_non_crypt] > 0)) > 0) |
                                             row.names(case.norm.mut.df) %in% snps_flt),] #80

#as there are no crypts this should be the same as the filter below
nrow(case.norm.mut.df[((rowSums(case.norm.mut.df[, lcm_non_crypt] > 0)) > 1 | row.names(case.norm.mut.df) %in% snps_flt),]) #80, correct

write.table(case.norm.mut.df.shared, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/PD51123_norm_sample_shared_mut_df.txt", col.names = T, row.names = T, sep = "\t", quote = F)

```

```{r}

pd50297_shared = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/PD50297_norm_sample_shared_mut_df.txt", header = T, sep = "\t", stringsAsFactors = F)
pd51122_shared = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/PD51122_norm_sample_shared_mut_df.txt", header = T, sep = "\t", stringsAsFactors = F)
pd51123_shared = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/PD51123_norm_sample_shared_mut_df.txt", header = T, sep = "\t", stringsAsFactors = F)
nrow(pd50297_shared) #343
nrow(pd51122_shared) #153
nrow(pd51123_shared) #80

nrow(pd50297_shared) + nrow(pd51122_shared) + nrow(pd51123_shared) #576

```

###PD51122 (Fig. 3b)

```{r}

library(ComplexHeatmap)
library(circlize)

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/")

patient = "PD51122"

#read in dataframes
manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
patient_norm_manifest = manifest[manifest$keep == "Y" &
                            manifest$purity_trunk < 0.01 &
                            manifest$case.id == patient,]

table(patient_norm_manifest$bulk_tissue)
table(patient_norm_manifest$histo)

patient_df = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/", patient, "_norm_sample_shared_mut_df.txt"), header = T, sep = "\t", stringsAsFactors = F) #in a given lcm sample, does the variant occur at a sufficient depth when compared to error profiles created from the non-CNS LCM samples and bulk samples of the other two patients - shared variants only

vaf_input = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/", patient, "_VAF_merged_snps.txt"), header = T, sep = "\t", stringsAsFactors = F) #VAF of all called subs

vaf = vaf_input[row.names(patient_df),]

mut_vec = row.names(patient_df)

#germ layers
ecto = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Ectoderm",]$sample
endo = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Endoderm",]$sample
meso = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Mesoderm",]$sample

gl_vec = rep(NA, length(mut_vec))

ecto_muts = rowSums(patient_df[, ecto] > 0) > 0
endo_muts = rowSums(patient_df[, endo] > 0) > 0
meso_muts = rowSums(patient_df[, meso] > 0) > 0

gl_vec[ecto_muts] = "ectoderm"
gl_vec[endo_muts] = "endoderm"
gl_vec[meso_muts] = "mesoderm"
gl_vec[(ecto_muts & endo_muts & meso_muts) |
          (ecto_muts & endo_muts) |
          (endo_muts & meso_muts) |
          (ecto_muts & meso_muts)] = "mixed"
names(gl_vec) = mut_vec

#col functions
bin_vec_cols = c("white", "grey8")
names(bin_vec_cols) = c(0, 1)

germ_layer_col_fun = c("#75539e", "#1fd655", "#FFAE42", "grey85")
names(germ_layer_col_fun) = c("ectoderm", "endoderm", "mesoderm", "mixed")

max(cor(t(vaf))) #1
min(cor(t(vaf))) #-0.8768577

col_fun = colorRamp2(c(-1, -0.5, 0, 0.5, 1), c("#1167B1", "#2a9df4", "white", "#C00000", "#800000"))

#skin_non_cns_ecto annotation
skin_non_cns_ecto = patient_norm_manifest[patient_norm_manifest$CNS == "N" &
                                            patient_norm_manifest$predominant_germ_layer == "Ectoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM" &
                                            patient_norm_manifest$bulk_tissue == "Skin",]$sample

skin_non_cns_ecto_muts = mut_vec[rowSums(patient_df[mut_vec, skin_non_cns_ecto] > 0) > 0]

skin_non_cns_ecto_vec = rep(0, length(mut_vec))
names(skin_non_cns_ecto_vec) = mut_vec
skin_non_cns_ecto_vec[skin_non_cns_ecto_muts] = 1

#pns_ecto annotation
pns_ecto = patient_norm_manifest[patient_norm_manifest$CNS == "N" &
                                            patient_norm_manifest$predominant_germ_layer == "Ectoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM" &
                                            patient_norm_manifest$histo %in% c("Nerve", "Ganglion"),]$sample #not including NB-like cluster

pns_ecto_muts = mut_vec[rowSums(patient_df[mut_vec, pns_ecto] > 0) > 0]

pns_ecto_vec = rep(0, length(mut_vec))
names(pns_ecto_vec) = mut_vec
pns_ecto_vec[pns_ecto_muts] = 1

#non cns endo annotation
non_cns_endo = patient_norm_manifest[patient_norm_manifest$CNS == "N" &
                                            patient_norm_manifest$predominant_germ_layer == "Endoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM",]$sample
non_cns_endo_muts = mut_vec[rowSums(patient_df[mut_vec, non_cns_endo] > 0) > 0]

non_cns_endo_vec = rep(0, length(mut_vec))
names(non_cns_endo_vec) = mut_vec
non_cns_endo_vec[non_cns_endo_muts] = 1

#non_cns_meso
non_cns_meso = patient_norm_manifest[patient_norm_manifest$CNS == "N" &
                                            patient_norm_manifest$predominant_germ_layer == "Mesoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM",]$sample
non_cns_meso_muts = mut_vec[rowSums(patient_df[mut_vec, non_cns_meso] > 0) > 0]

non_cns_meso_vec = rep(0, length(mut_vec))
names(non_cns_meso_vec) = mut_vec
non_cns_meso_vec[non_cns_meso_muts] = 1

#tumour root variant
snps = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars/", patient, "_put_tum_root_snp_combined_method_reviewed.txt"), header = T, sep = "\t", stringsAsFactors = F, row.names = 1)
snps_flt = row.names(snps[snps$Keep == "Y",])

tumour_root_vec = rep(0, length(mut_vec))
tumour_root_vec[mut_vec %in% snps_flt] = 1
names(tumour_root_vec) = mut_vec

#cns ecto annotation
cns_ecto = patient_norm_manifest[patient_norm_manifest$CNS == "Y" &
                                            patient_norm_manifest$predominant_germ_layer == "Ectoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM",]$sample

cns_ecto_muts = mut_vec[rowSums(patient_df[mut_vec, cns_ecto] > 0) > 0]

cns_ecto_vec = rep(0, length(mut_vec))
names(cns_ecto_vec) = mut_vec
cns_ecto_vec[cns_ecto_muts] = 1

#CALM annotation
calm = "PD51122ag_lo0007"

calm_muts = mut_vec[(patient_df[mut_vec, calm] > 0) > 0]

calm_vec = rep(0, length(mut_vec))
names(calm_vec) = mut_vec
calm_vec[calm_muts] = 1

#spindled neural lesion annotation
snl = "PD51122ac_lo0013"

snl_muts = mut_vec[(patient_df[mut_vec, snl] > 0) > 0]

snl_vec = rep(0, length(mut_vec))
names(snl_vec) = mut_vec
snl_vec[snl_muts] = 1

ha = HeatmapAnnotation(cns_ecto = cns_ecto_vec,
                       pns_ecto = pns_ecto_vec, 
                       skin_non_cns_ecto = skin_non_cns_ecto_vec,
                       non_cns_endo = non_cns_endo_vec,
                       non_cns_meso = non_cns_meso_vec,
                       col = list(cns_ecto = bin_vec_cols,
                                  pns_ecto = bin_vec_cols,
                                  skin_non_cns_ecto = bin_vec_cols,
                                  non_cns_endo = bin_vec_cols,
                                  non_cns_meso = bin_vec_cols),
                        show_legend = c(F, F, F, F, F))

ha2 = HeatmapAnnotation(germ_layer = gl_vec,
                        tumour_root = tumour_root_vec,
                        CALM = calm_vec,
                        SNL = snl_vec,
                        col = list(tumour_root = bin_vec_cols,
                                   CALM = bin_vec_cols,
                                    SNL = bin_vec_cols,
                                   germ_layer = germ_layer_col_fun),
                        show_legend = c(T, F, F, F))

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/PD51122_shared_muts_heatmap_20221122.pdf", height = 4, width = 5, useDingbats = F)
Heatmap(cor(t(vaf)), 
        top_annotation = ha, 
        bottom_annotation = ha2,
        name = "Pearson correlation", 
        show_column_dend = F, 
        show_row_names = F, 
        show_column_names = F, 
        col = col_fun)
dev.off()

png("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/PD51122_shared_muts_heatmap_20221122.png", height = 4, width = 4.5, units = "in", res = 300)
Heatmap(cor(t(vaf)),
        name = "Pearson correlation", 
        show_column_dend = F, 
        show_row_names = F, 
        show_column_names = F,
        show_heatmap_legend = F,
        col = col_fun)
dev.off()

#clonal endoderm samples distort it
median(rowMeans(vaf[names(gl_vec[gl_vec == "mixed"]), ecto]))
median(rowMeans(vaf[names(gl_vec[gl_vec == "ectoderm"]), ecto]))

median(rowMeans(vaf[names(gl_vec[gl_vec == "mixed"]), endo]))
mean(rowMeans(vaf[names(gl_vec[gl_vec == "endoderm"]), endo]))

median(rowMeans(vaf[names(gl_vec[gl_vec == "mixed"]), meso]))
median(rowMeans(vaf[names(gl_vec[gl_vec == "mesoderm"]), meso]))

```

###PD50297 & PD51123 (EDF7)

####PD50297

```{r}

library(ComplexHeatmap)
library(circlize)

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/")

patient = "PD50297"

#read in dataframes
manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
patient_norm_manifest = manifest[manifest$keep == "Y" &
                            manifest$purity_trunk < 0.01 &
                            manifest$case.id == patient,]

table(patient_norm_manifest$bulk_tissue)
table(patient_norm_manifest$histo)

patient_df = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/", patient, "_norm_sample_shared_mut_df.txt"), header = T, sep = "\t", stringsAsFactors = F) #in a given lcm sample, does the variant occur at a sufficient depth when compared to error profiles created from the non-CNS LCM samples and bulk samples of the other two patients

vaf_input = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/", patient, "_VAF_merged_snps.txt"), header = T, sep = "\t", stringsAsFactors = F) #VAF of all called subs

vaf = vaf_input[row.names(patient_df),]

mut_vec = row.names(patient_df)

#germ layers
ecto = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Ectoderm",]$sample
endo = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Endoderm",]$sample
meso = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Mesoderm",]$sample

gl_vec = rep(NA, length(mut_vec))

ecto_muts = rowSums(patient_df[, ecto] > 0) > 0
endo_muts = rowSums(patient_df[, endo] > 0) > 0
meso_muts = rowSums(patient_df[, meso] > 0) > 0

gl_vec[ecto_muts] = "ectoderm"
gl_vec[endo_muts] = "endoderm"
gl_vec[meso_muts] = "mesoderm"
gl_vec[(ecto_muts & endo_muts & meso_muts) |
          (ecto_muts & endo_muts) |
          (endo_muts & meso_muts) |
          (ecto_muts & meso_muts)] = "mixed"
names(gl_vec) = mut_vec

#col functions
bin_vec_cols = c("white", "grey8")
names(bin_vec_cols) = c(0, 1)

germ_layer_col_fun = c("#75539e", "#1fd655", "#FFAE42", "grey85")
names(germ_layer_col_fun) = c("ectoderm", "endoderm", "mesoderm", "mixed")

max(cor(t(vaf))) #1
min(cor(t(vaf))) #-0.8104974

#col_fun = colorRamp2(c(-0.5, 0, 1), c("blue", "white", "red"))
col_fun = colorRamp2(c(-1, -0.5, 0, 0.5, 1), c("#1167B1", "#2a9df4", "white", "#C00000", "#800000"))

#skin_non_cns_ecto annotation
skin_non_cns_ecto = patient_norm_manifest[patient_norm_manifest$CNS == "N" &
                                            patient_norm_manifest$predominant_germ_layer == "Ectoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM" &
                                            patient_norm_manifest$bulk_tissue == "Skin",]$sample

skin_non_cns_ecto_muts = mut_vec[rowSums(patient_df[mut_vec, skin_non_cns_ecto] > 0) > 0]

skin_non_cns_ecto_vec = rep(0, length(mut_vec))
names(skin_non_cns_ecto_vec) = mut_vec
skin_non_cns_ecto_vec[skin_non_cns_ecto_muts] = 1

#pns_ecto annotation
pns_ecto = patient_norm_manifest[patient_norm_manifest$CNS == "N" &
                                            patient_norm_manifest$predominant_germ_layer == "Ectoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM" &
                                            patient_norm_manifest$histo %in% c("Nerve", "Ganglion", "Medulla"),]$sample

pns_ecto_muts = mut_vec[rowSums(patient_df[mut_vec, pns_ecto] > 0) > 0]

pns_ecto_vec = rep(0, length(mut_vec))
names(pns_ecto_vec) = mut_vec
pns_ecto_vec[pns_ecto_muts] = 1

#non cns endo annotation
non_cns_endo = patient_norm_manifest[patient_norm_manifest$CNS == "N" &
                                            patient_norm_manifest$predominant_germ_layer == "Endoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM",]$sample
non_cns_endo_muts = mut_vec[rowSums(patient_df[mut_vec, non_cns_endo] > 0) > 0]

non_cns_endo_vec = rep(0, length(mut_vec))
names(non_cns_endo_vec) = mut_vec
non_cns_endo_vec[non_cns_endo_muts] = 1

#non_cns_meso
non_cns_meso = patient_norm_manifest[patient_norm_manifest$CNS == "N" &
                                            patient_norm_manifest$predominant_germ_layer == "Mesoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM",]$sample
non_cns_meso_muts = mut_vec[rowSums(patient_df[mut_vec, non_cns_meso] > 0) > 0]

non_cns_meso_vec = rep(0, length(mut_vec))
names(non_cns_meso_vec) = mut_vec
non_cns_meso_vec[non_cns_meso_muts] = 1

#tumour root variant
snps = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars/", patient, "_put_tum_root_snp_combined_method_reviewed.txt"), header = T, sep = "\t", stringsAsFactors = F, row.names = 1)
snps_flt = row.names(snps[snps$Keep == "Y",])

tumour_root_vec = rep(0, length(mut_vec))
tumour_root_vec[mut_vec %in% snps_flt] = 1
names(tumour_root_vec) = mut_vec

#cns ecto annotation
cns_ecto = patient_norm_manifest[patient_norm_manifest$CNS == "Y" &
                                            patient_norm_manifest$predominant_germ_layer == "Ectoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM",]$sample

cns_ecto_muts = mut_vec[rowSums(patient_df[mut_vec, cns_ecto] > 0) > 0]

cns_ecto_vec = rep(0, length(mut_vec))
names(cns_ecto_vec) = mut_vec
cns_ecto_vec[cns_ecto_muts] = 1

ha = HeatmapAnnotation(cns_ecto = cns_ecto_vec,
                       pns_ecto = pns_ecto_vec, 
                       skin_non_cns_ecto = skin_non_cns_ecto_vec,
                       non_cns_endo = non_cns_endo_vec,
                       non_cns_meso = non_cns_meso_vec,
                       col = list(cns_ecto = bin_vec_cols,
                                  pns_ecto = bin_vec_cols,
                                  skin_non_cns_ecto = bin_vec_cols,
                                  non_cns_endo = bin_vec_cols,
                                  non_cns_meso = bin_vec_cols),
                        show_legend = c(F, F, F, F, F))

ha2 = HeatmapAnnotation(germ_layer = gl_vec,
                        tumour_root = tumour_root_vec,
                        col = list(tumour_root = bin_vec_cols,
                                   germ_layer = germ_layer_col_fun),
                        show_legend = c(T, F))

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/PD50297_shared_muts_heatmap_20221208.pdf", height = 4, width = 5, useDingbats = F)
Heatmap(cor(t(vaf)), 
        top_annotation = ha, 
        bottom_annotation = ha2,
        name = "Pearson correlation", 
        show_column_dend = F, 
        show_row_names = F, 
        show_column_names = F, 
        col = col_fun)
dev.off()

png("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/PD50297_shared_muts_heatmap_202212208.png", height = 4, width = 4.5, units = "in", res = 300)
Heatmap(cor(t(vaf)),
        name = "Pearson correlation", 
        show_column_dend = F, 
        show_row_names = F, 
        show_column_names = F,
        show_heatmap_legend = F,
        col = col_fun)
dev.off()
```

####PD51123

```{r}

library(ComplexHeatmap)
library(circlize)

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/")

patient = "PD51123"

#read in dataframes
manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
patient_norm_manifest = manifest[manifest$keep == "Y" &
                            manifest$purity_trunk < 0.01 &
                            manifest$case.id == patient,]

table(patient_norm_manifest$bulk_tissue)
table(patient_norm_manifest$histo)

patient_df = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/", patient, "_norm_sample_shared_mut_df.txt"), header = T, sep = "\t", stringsAsFactors = F) #in a given lcm sample, does the variant occur at a sufficient depth when compared to error profiles created from the non-CNS LCM samples and bulk samples of the other two patients

vaf_input = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/", patient, "_VAF_merged_snps.txt"), header = T, sep = "\t", stringsAsFactors = F) #VAF of all called subs

vaf = vaf_input[row.names(patient_df),]

mut_vec = row.names(patient_df)

#germ layers
ecto = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Ectoderm",]$sample
endo = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Endoderm",]$sample
meso = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Mesoderm",]$sample

gl_vec = rep(NA, length(mut_vec))

ecto_muts = rowSums(patient_df[, ecto] > 0) > 0
endo_muts = rowSums(patient_df[, endo] > 0) > 0
meso_muts = rowSums(patient_df[, meso] > 0) > 0

gl_vec[ecto_muts] = "ectoderm"
gl_vec[endo_muts] = "endoderm"
gl_vec[meso_muts] = "mesoderm"
gl_vec[(ecto_muts & endo_muts & meso_muts) |
          (ecto_muts & endo_muts) |
          (endo_muts & meso_muts) |
          (ecto_muts & meso_muts)] = "mixed"
names(gl_vec) = mut_vec

#col functions
bin_vec_cols = c("white", "grey8")
names(bin_vec_cols) = c(0, 1)

germ_layer_col_fun = c("#75539e", "#1fd655", "#FFAE42", "grey85")
names(germ_layer_col_fun) = c("ectoderm", "endoderm", "mesoderm", "mixed")

max(cor(t(vaf))) #1
min(cor(t(vaf))) #-0.5257019

#col_fun = colorRamp2(c(-0.5, 0, 1), c("blue", "white", "red"))
col_fun = colorRamp2(c(-1, -0.5, 0, 0.5, 1), c("#1167B1", "#2a9df4", "white", "#C00000", "#800000"))

#skin_non_cns_ecto annotation
skin_non_cns_ecto = patient_norm_manifest[patient_norm_manifest$CNS == "N" &
                                            patient_norm_manifest$predominant_germ_layer == "Ectoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM" &
                                            patient_norm_manifest$bulk_tissue == "Skin",]$sample

skin_non_cns_ecto_muts = mut_vec[rowSums(patient_df[mut_vec, skin_non_cns_ecto] > 0) > 0]

skin_non_cns_ecto_vec = rep(0, length(mut_vec))
names(skin_non_cns_ecto_vec) = mut_vec
skin_non_cns_ecto_vec[skin_non_cns_ecto_muts] = 1

#pns_ecto annotation
pns_ecto = patient_norm_manifest[patient_norm_manifest$CNS == "N" &
                                            patient_norm_manifest$predominant_germ_layer == "Ectoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM" &
                                            patient_norm_manifest$histo %in% c("Nerve", "Ganglion", "Medulla"),]$sample

pns_ecto_muts = mut_vec[rowSums(patient_df[mut_vec, pns_ecto] > 0) > 0]

pns_ecto_vec = rep(0, length(mut_vec))
names(pns_ecto_vec) = mut_vec
pns_ecto_vec[pns_ecto_muts] = 1

#non cns endo annotation
non_cns_endo = patient_norm_manifest[patient_norm_manifest$CNS == "N" &
                                            patient_norm_manifest$predominant_germ_layer == "Endoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM",]$sample
non_cns_endo_muts = mut_vec[rowSums(patient_df[mut_vec, non_cns_endo] > 0) > 0]

non_cns_endo_vec = rep(0, length(mut_vec))
names(non_cns_endo_vec) = mut_vec
non_cns_endo_vec[non_cns_endo_muts] = 1

#non_cns_meso
non_cns_meso = patient_norm_manifest[patient_norm_manifest$CNS == "N" &
                                            patient_norm_manifest$predominant_germ_layer == "Mesoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM",]$sample
non_cns_meso_muts = mut_vec[rowSums(patient_df[mut_vec, non_cns_meso] > 0) > 0]

non_cns_meso_vec = rep(0, length(mut_vec))
names(non_cns_meso_vec) = mut_vec
non_cns_meso_vec[non_cns_meso_muts] = 1

#tumour root variant
snps = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars/", patient, "_put_tum_root_snp_combined_method_reviewed.txt"), header = T, sep = "\t", stringsAsFactors = F, row.names = 1)
snps_flt = row.names(snps[snps$Keep == "Y",])

tumour_root_vec = rep(0, length(mut_vec))
tumour_root_vec[mut_vec %in% snps_flt] = 1
names(tumour_root_vec) = mut_vec

#cns ecto annotation
cns_ecto = patient_norm_manifest[patient_norm_manifest$CNS == "Y" &
                                            patient_norm_manifest$predominant_germ_layer == "Ectoderm" &
                                            patient_norm_manifest$experimental_arm == "WGS_LCM",]$sample

cns_ecto_muts = mut_vec[rowSums(patient_df[mut_vec, cns_ecto] > 0) > 0]

cns_ecto_vec = rep(0, length(mut_vec))
names(cns_ecto_vec) = mut_vec
cns_ecto_vec[cns_ecto_muts] = 1

ha = HeatmapAnnotation(cns_ecto = cns_ecto_vec,
                       pns_ecto = pns_ecto_vec, 
                       skin_non_cns_ecto = skin_non_cns_ecto_vec,
                       non_cns_endo = non_cns_endo_vec,
                       non_cns_meso = non_cns_meso_vec,
                       col = list(cns_ecto = bin_vec_cols,
                                  pns_ecto = bin_vec_cols,
                                  skin_non_cns_ecto = bin_vec_cols,
                                  non_cns_endo = bin_vec_cols,
                                  non_cns_meso = bin_vec_cols),
                        show_legend = c(F, F, F, F, F))

ha2 = HeatmapAnnotation(germ_layer = gl_vec,
                        tumour_root = tumour_root_vec,
                        col = list(tumour_root = bin_vec_cols,
                                   germ_layer = germ_layer_col_fun),
                        show_legend = c(T, F))

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/PD51123_shared_muts_heatmap_20221208.pdf", height = 4, width = 5, useDingbats = F)
Heatmap(cor(t(vaf)), 
        top_annotation = ha, 
        bottom_annotation = ha2,
        name = "Pearson correlation", 
        show_column_dend = F, 
        show_row_names = F, 
        show_column_names = F, 
        col = col_fun)
dev.off()

png("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/PD51123_shared_muts_heatmap_20221208.png", height = 4, width = 4.5, units = "in", res = 300)
Heatmap(cor(t(vaf)),
        name = "Pearson correlation", 
        show_column_dend = F, 
        show_row_names = F, 
        show_column_names = F,
        show_heatmap_legend = F,
        col = col_fun)
dev.off()

```

##Fig. 3c - spatially constrained expansions

###PD51122

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/")

patient = "PD51122"

#read in dataframes
manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
patient_norm_manifest = manifest[manifest$keep == "Y" &
                            manifest$purity_trunk < 0.01 &
                            manifest$case.id == patient,]

table(patient_norm_manifest$bulk_tissue)
table(patient_norm_manifest$histo)

patient_df = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/", patient, "_norm_sample_shared_mut_df.txt"), header = T, sep = "\t", stringsAsFactors = F) #in a given lcm sample, does the variant occur at a sufficient depth when compared to error profiles created from the non-CNS LCM samples and bulk samples of the other two patients

vaf_input = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/", patient, "_VAF_merged_snps.txt"), header = T, sep = "\t", stringsAsFactors = F) #VAF of all called subs

vaf = vaf_input[row.names(patient_df),]

mut_vec = row.names(patient_df)

#germ layers
ecto = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Ectoderm",]$sample
endo = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Endoderm",]$sample
meso = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Mesoderm",]$sample

gl_vec = rep(NA, length(mut_vec))

ecto_muts = rowSums(patient_df[, ecto] > 0) > 0
endo_muts = rowSums(patient_df[, endo] > 0) > 0
meso_muts = rowSums(patient_df[, meso] > 0) > 0

gl_vec[ecto_muts] = "ectoderm"
gl_vec[endo_muts] = "endoderm"
gl_vec[meso_muts] = "mesoderm"
gl_vec[(ecto_muts & endo_muts & meso_muts) |
          (ecto_muts & endo_muts) |
          (endo_muts & meso_muts) |
          (ecto_muts & meso_muts)] = "mixed"
names(gl_vec) = mut_vec

Heatmap(patient_df[names(gl_vec[gl_vec == "mesoderm"]), manifest[manifest$sample %in% colnames(patient_df)[colSums(patient_df[names(gl_vec[gl_vec == "mesoderm"]), ]) > 0],]$sample])
Heatmap(patient_df[names(gl_vec[gl_vec == "ectoderm"]), manifest[manifest$sample %in% colnames(patient_df)[colSums(patient_df[names(gl_vec[gl_vec == "ectoderm"]), ]) > 0],]$sample])
Heatmap(patient_df[names(gl_vec[gl_vec == "endoderm"]), manifest[manifest$sample %in% colnames(patient_df)[colSums(patient_df[names(gl_vec[gl_vec == "endoderm"]), ]) > 0],]$sample])

```

###PD50297

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/")

patient = "PD50297"

#read in dataframes
manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
patient_norm_manifest = manifest[manifest$keep == "Y" &
                            manifest$purity_trunk < 0.01 &
                            manifest$case.id == patient,]

table(patient_norm_manifest$bulk_tissue)
table(patient_norm_manifest$histo)

patient_df = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/mut_cluster_analysis/", patient, "_norm_sample_shared_mut_df.txt"), header = T, sep = "\t", stringsAsFactors = F) #in a given lcm sample, does the variant occur at a sufficient depth when compared to error profiles created from the non-CNS LCM samples and bulk samples of the other two patients

vaf_input = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/", patient, "_VAF_merged_snps.txt"), header = T, sep = "\t", stringsAsFactors = F) #VAF of all called subs

vaf = vaf_input[row.names(patient_df),]

mut_vec = row.names(patient_df)

#germ layers
ecto = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Ectoderm",]$sample
endo = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Endoderm",]$sample
meso = patient_norm_manifest[patient_norm_manifest$experimental_arm == "WGS_LCM" & patient_norm_manifest$predominant_germ_layer == "Mesoderm",]$sample

gl_vec = rep(NA, length(mut_vec))

ecto_muts = rowSums(patient_df[, ecto] > 0) > 0
endo_muts = rowSums(patient_df[, endo] > 0) > 0
meso_muts = rowSums(patient_df[, meso] > 0) > 0

gl_vec[ecto_muts] = "ectoderm"
gl_vec[endo_muts] = "endoderm"
gl_vec[meso_muts] = "mesoderm"
gl_vec[(ecto_muts & endo_muts & meso_muts) |
          (ecto_muts & endo_muts) |
          (endo_muts & meso_muts) |
          (ecto_muts & meso_muts)] = "mixed"
names(gl_vec) = mut_vec

Heatmap(patient_df[names(gl_vec[gl_vec == "endoderm"]), manifest[manifest$sample %in% colnames(patient_df)[colSums(patient_df[names(gl_vec[gl_vec == "endoderm"]), ]) > 0],]$sample])

endo_samples = manifest[manifest$sample %in% colnames(patient_df)[colSums(patient_df[names(gl_vec[gl_vec == "endoderm"]), ]) > 0],]$sample
Heatmap(patient_df[names(gl_vec[gl_vec == "endoderm"]),][rowSums(patient_df[names(gl_vec[gl_vec == "endoderm"]), endo_samples[!grepl("PD50297g", endo_samples)]]) == 0, endo_samples[grepl("PD50297g", endo_samples)]])

Heatmap(patient_df[names(gl_vec[gl_vec == "mesoderm"]), manifest[manifest$sample %in% colnames(patient_df)[colSums(patient_df[names(gl_vec[gl_vec == "mesoderm"]), ]) > 0],]$sample])
Heatmap(patient_df[names(gl_vec[gl_vec == "ectoderm"]), manifest[manifest$sample %in% colnames(patient_df)[colSums(patient_df[names(gl_vec[gl_vec == "ectoderm"]), ]) > 0],]$sample])

```

##Fig. 3d & EDF8 - distribution of tumour root mutations across CNS

###Get pseudobulks

To increase depth to improve specificity of detection for mutations shared between tumour and normal. Going to try again with only samples where <1% tumour infiltration is suspected. This is will lower overall tumour infiltration and hopefully preserve some bulk samples for downstream analyses.

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup

export PATH=/software/R-4.1.0/bin:${PATH}
#export R_HOME=$(R RHOME)
#export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"

for patient in PD50297 PD51122 PD51123;
do
bsub -q normal -o ${patient}_pseudo_non_tumour.out -e ${patient}_pseudo_non_tumour.err -n 4 -R 'select[mem>=60000] rusage[mem=60000] span[hosts=1]' -M60000 /software/R-4.1.0/bin/Rscript ~/scripts/non_tumour_pseudo_bulk_count_20220925.R -p $patient
done

```

```{r}

#~/scripts/non_tumour_pseudo_bulk_count_20220925.R

library(optparse)

#setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup")
#patient = "PD51123"

option_list = list(
  make_option(c("-p", "--patient"), type="character", default=NULL, help="patient name", metavar="character"))

opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

patient = opt$patient

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
manifest_flt = manifest[manifest$keep == "Y" & manifest$case.id == patient & manifest$purity_trunk < 0.01, ]

NR = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/", patient, "_NR_final_merged_snps.txt"), header = T, sep = "\t", stringsAsFactors = F)
NV = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/", patient, "_NV_final_merged_snps.txt"), header = T, sep = "\t", stringsAsFactors = F)

bulk_samples = unique(manifest_flt$bulk_sample)

merged_NV = data.frame(matrix(nrow = nrow(NV), ncol = length(bulk_samples)), row.names = row.names(NV))
merged_NR = data.frame(matrix(nrow = nrow(NV), ncol = length(bulk_samples)), row.names = row.names(NV))

colnames(merged_NV) = colnames(merged_NR) = bulk_samples

for(mut in row.names(merged_NV)){
  
  for(sample in colnames(merged_NV)){
    
    all_samples = manifest_flt[manifest_flt$bulk_sample == sample,]$sample
    
    if(length(all_samples) > 1){
      
      merged_NV[mut, sample] = as.numeric(rowSums(NV[mut, all_samples]))
      merged_NR[mut, sample] = as.numeric(rowSums(NR[mut, all_samples]))
      
    }
    
    if(length(all_samples) == 1){
      
      merged_NV[mut, sample] = as.numeric(NV[mut, all_samples])
      merged_NR[mut, sample] = as.numeric(NR[mut, all_samples])
      
    }
    
  }
  
}

write.table(merged_NV, paste0(patient, "_merged_bulk_filtered_snp_NV_non_tumour_only.txt"), col.names = T, row.names = T, sep = "\t", quote = F)
write.table(merged_NR, paste0(patient, "_merged_bulk_filtered_snp_NR_non_tumour_only.txt"), col.names = T, row.names = T, sep = "\t", quote = F)

```

###Estimate tumour involvement per normal tissue pseudobulk

###PD51122

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/")

patient = "PD51122"
truncal_cluster = "6_9"

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
manifest_flt = manifest[manifest$keep == "Y" & manifest$case.id == patient & manifest$purity_trunk < 0.01, ]

samples = manifest[manifest$case.id == patient & manifest$battenberg_purity >= 0.4 & !is.na(manifest$battenberg_purity), ]$sample

#get variants in trunk
vars = read.table(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient, '/md_out/', patient, '_allClusterassignmentsFromParallelRuns_clusters_reviewed.txt'), header = T, sep = '\t', stringsAsFactors = F)

cluster_vars = vars[vars$cluster.no == truncal_cluster,]
cluster_vars$ID = paste0(cluster_vars$chr, "_", cluster_vars$pos)

#identify which truncal variants cover two chromomes copies in 2+0 sites
dp_files = list.files(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient), "allDirichletProcessInfo_fromMultipleSamplesWithoutChromosomeLosses.txt", full.names = T)

trunc_pre_dup = list()
for(file in dp_files){
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  sample = unlist(strsplit(basename(file), "_allDirichletProcessInfo_fromMultipleSamplesWithoutChromosomeLosses.txt"))
  data$ID = paste0(data$chr, "_", data$pos)
  trunc_pre_dup[[sample]] = data[data$ID %in% cluster_vars$ID & data$subclonal.CN == 2 & is.na(data$nMaj2) & data$nMin1 == 0 & data$no.chrs.bearing.mut == 2,]$ID
}

hi_conf_loci = Reduce(intersect, trunc_pre_dup)
hi_conf_loci = paste0("chr", hi_conf_loci)
length(hi_conf_loci) #60

#read in VAF dataframe
NR = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_merged_bulk_filtered_snp_NR_non_tumour_only.txt"), header = T, sep = "\t", stringsAsFactors = F)
NV = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_merged_bulk_filtered_snp_NV_non_tumour_only.txt"), header = T, sep = "\t", stringsAsFactors = F)

NR[NR == 0] = 1

VAF = NV / NR

VAF_filter = VAF
VAF_filter$ID = paste0(unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 1)), "_", unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 2)))

mean_no_outliers <- function(x, remove_na = TRUE) { #early embryonic variants might skew overall calculation
  qnt <- quantile(x, probs = c(.25, .75), na.rm = remove_na)
  val <- 1.5 * IQR(x, na.rm = remove_na)
  y <- x
  y[x < (qnt[1] - val)] <- NA
  y[x > (qnt[2] + val)] <- NA
  z = mean(y, na.rm = remove_na)
  return(z)
}

tum_infiltration_est = data.frame(apply(VAF[VAF_filter$ID %in% hi_conf_loci, ], 2, function(x) mean_no_outliers(x)))
tum_infiltration_est$PD = row.names(tum_infiltration_est)
colnames(tum_infiltration_est)[1] = c("purity_trunc")
tum_infiltration_est = tum_infiltration_est[, c(2,1)]

tum_infiltration_est$bulk_tissue = NA
tum_infiltration_est$location = NA
tum_infiltration_est$total_depth = NA
tum_infiltration_est$CNS = NA

for(i in 1:nrow(tum_infiltration_est)){
  tum_infiltration_est$bulk_tissue[i] = manifest_flt[manifest_flt$bulk_sample == tum_infiltration_est$PD[i], ]$bulk_tissue[1]
  tum_infiltration_est$location[i] = manifest_flt[manifest_flt$bulk_sample == tum_infiltration_est$PD[i], ]$location[1]
  tum_infiltration_est$total_depth[i] = sum(manifest_flt[manifest_flt$bulk_sample == tum_infiltration_est$PD[i], ]$picard_median_coverage)
  tum_infiltration_est$CNS[i] = manifest_flt[manifest_flt$bulk_sample == tum_infiltration_est$PD[i], ]$CNS[1]
}

write.table(tum_infiltration_est, paste0(patient, "_pseudo_bulk_no_tumour_metadata.txt"), sep = '\t', quote = F, col.names = T, row.names = F)

```

###PD50297 & PD51123

```{r}

#PD50297
patient = "PD50297"
truncal_cluster = 9

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
manifest_flt = manifest[manifest$keep == "Y" & manifest$case.id == patient & manifest$purity_trunk < 0.01, ]

samples = manifest[manifest$case.id == patient & manifest$battenberg_purity >= 0.4 & !is.na(manifest$battenberg_purity), ]$sample

#get variants in trunk
vars = read.table(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient, '/md_out/', patient, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)

cluster_vars = vars[vars$cluster.no == truncal_cluster,]
cluster_vars$ID = paste0(cluster_vars$chr, "_", cluster_vars$pos)

#identify which truncal variants occur at 1+1 sites
dp_files = list.files(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient), "allDirichletProcessInfo_fromMultipleSamplesWithoutChromosomeLosses.txt", full.names = T)

trunc_pre_dup = list()
for(file in dp_files){
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  sample = unlist(strsplit(basename(file), "_allDirichletProcessInfo_fromMultipleSamplesWithoutChromosomeLosses.txt"))
  data$ID = paste0(data$chr, "_", data$pos)
  trunc_pre_dup[[sample]] = data[data$ID %in% cluster_vars$ID & data$subclonal.CN == 2 & is.na(data$nMaj2) & data$nMin1 == 1 & data$nMaj1 == 1,]$ID
}

hi_conf_loci = Reduce(intersect, trunc_pre_dup)
hi_conf_loci = paste0("chr", hi_conf_loci)
length(hi_conf_loci) #385

#read in VAF dataframe
NR = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_merged_bulk_filtered_snp_NR_non_tumour_only.txt"), header = T, sep = "\t", stringsAsFactors = F)
NV = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_merged_bulk_filtered_snp_NV_non_tumour_only.txt"), header = T, sep = "\t", stringsAsFactors = F)

NR[NR == 0] = 1

VAF = NV / NR

VAF_filter = VAF
VAF_filter$ID = paste0(unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 1)), "_", unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 2)))

mean_no_outliers <- function(x, remove_na = TRUE) { #early embryonic variants might skew overall calculation
  qnt <- quantile(x, probs = c(.25, .75), na.rm = remove_na)
  val <- 1.5 * IQR(x, na.rm = remove_na)
  y <- x
  y[x < (qnt[1] - val)] <- NA
  y[x > (qnt[2] + val)] <- NA
  z = mean(y, na.rm = remove_na)
  return(z)
}

tum_infiltration_est = data.frame(apply(VAF[VAF_filter$ID %in% hi_conf_loci, ], 2, function(x) 2 * mean_no_outliers(x))) #x2 at 1+1 sites for cellularity
tum_infiltration_est$PD = row.names(tum_infiltration_est)
colnames(tum_infiltration_est)[1] = c("purity_trunc")
tum_infiltration_est = tum_infiltration_est[, c(2,1)]

tum_infiltration_est$bulk_tissue = NA
tum_infiltration_est$location = NA
tum_infiltration_est$total_depth = NA
tum_infiltration_est$CNS = NA

for(i in 1:nrow(tum_infiltration_est)){
  tum_infiltration_est$bulk_tissue[i] = manifest_flt[manifest_flt$bulk_sample == tum_infiltration_est$PD[i], ]$bulk_tissue[1]
  tum_infiltration_est$location[i] = manifest_flt[manifest_flt$bulk_sample == tum_infiltration_est$PD[i], ]$location[1]
  tum_infiltration_est$total_depth[i] = sum(manifest_flt[manifest_flt$bulk_sample == tum_infiltration_est$PD[i], ]$picard_median_coverage)
  tum_infiltration_est$CNS[i] = manifest_flt[manifest_flt$bulk_sample == tum_infiltration_est$PD[i], ]$CNS[1]
}

write.table(tum_infiltration_est, paste0(patient, "_pseudo_bulk_no_tumour_metadata.txt"), sep = '\t', quote = F, col.names = T, row.names = F)

#PD51123
patient = "PD51123"
truncal_cluster = 2

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20220717.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
manifest_flt = manifest[manifest$keep == "Y" & manifest$case.id == patient & manifest$purity_trunk < 0.01, ]

samples = manifest[manifest$case.id == patient & manifest$battenberg_purity >= 0.4 & !is.na(manifest$battenberg_purity), ]$sample

#get variants in trunk
vars = read.table(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient, '/md_out/', patient, '_allClusterassignmentsFromParallelRuns.txt'), header = T, sep = '\t', stringsAsFactors = F)

cluster_vars = vars[vars$cluster.no == truncal_cluster,]
cluster_vars$ID = paste0(cluster_vars$chr, "_", cluster_vars$pos)

#identify which truncal variants occur at 1+1 sites
dp_files = list.files(paste0('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/nd_dpclust/01_Input/', patient), "allDirichletProcessInfo_fromMultipleSamplesWithoutChromosomeLosses.txt", full.names = T)

trunc_pre_dup = list()
for(file in dp_files){
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  sample = unlist(strsplit(basename(file), "_allDirichletProcessInfo_fromMultipleSamplesWithoutChromosomeLosses.txt"))
  data$ID = paste0(data$chr, "_", data$pos)
  trunc_pre_dup[[sample]] = data[data$ID %in% cluster_vars$ID & data$subclonal.CN == 2 & is.na(data$nMaj2) & data$nMin1 == 1 & data$nMaj1 == 1,]$ID
}

hi_conf_loci = Reduce(intersect, trunc_pre_dup)
hi_conf_loci = paste0("chr", hi_conf_loci)
length(hi_conf_loci) #260

#read in VAF dataframe
NR = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_merged_bulk_filtered_snp_NR_non_tumour_only.txt"), header = T, sep = "\t", stringsAsFactors = F)
NV = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_merged_bulk_filtered_snp_NV_non_tumour_only.txt"), header = T, sep = "\t", stringsAsFactors = F)

NR[NR == 0] = 1

VAF = NV / NR

VAF_filter = VAF
VAF_filter$ID = paste0(unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 1)), "_", unlist(lapply(strsplit(row.names(VAF_filter), '_'), "[[", 2)))

mean_no_outliers <- function(x, remove_na = TRUE) { #early embryonic variants might skew overall calculation
  qnt <- quantile(x, probs = c(.25, .75), na.rm = remove_na)
  val <- 1.5 * IQR(x, na.rm = remove_na)
  y <- x
  y[x < (qnt[1] - val)] <- NA
  y[x > (qnt[2] + val)] <- NA
  z = mean(y, na.rm = remove_na)
  return(z)
}

tum_infiltration_est = data.frame(apply(VAF[VAF_filter$ID %in% hi_conf_loci, ], 2, function(x) 2 * mean_no_outliers(x))) #x2 at 1+1 sites for cellularity
tum_infiltration_est$PD = row.names(tum_infiltration_est)
colnames(tum_infiltration_est)[1] = c("purity_trunc")
tum_infiltration_est = tum_infiltration_est[, c(2,1)]

tum_infiltration_est$bulk_tissue = NA
tum_infiltration_est$location = NA
tum_infiltration_est$total_depth = NA
tum_infiltration_est$CNS = NA

for(i in 1:nrow(tum_infiltration_est)){
  tum_infiltration_est$bulk_tissue[i] = manifest_flt[manifest_flt$bulk_sample == tum_infiltration_est$PD[i], ]$bulk_tissue[1]
  tum_infiltration_est$location[i] = manifest_flt[manifest_flt$bulk_sample == tum_infiltration_est$PD[i], ]$location[1]
  tum_infiltration_est$total_depth[i] = sum(manifest_flt[manifest_flt$bulk_sample == tum_infiltration_est$PD[i], ]$picard_median_coverage)
  tum_infiltration_est$CNS[i] = manifest_flt[manifest_flt$bulk_sample == tum_infiltration_est$PD[i], ]$CNS[1]
}

write.table(tum_infiltration_est, paste0(patient, "_pseudo_bulk_no_tumour_metadata.txt"), sep = '\t', quote = F, col.names = T, row.names = F)

```

###Plot distribution of variants

####PD50297

```{r}

patient = "PD50297"

#read in sample manifest
manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
patient_norm_manifest = manifest[manifest$keep == "Y" &
                            manifest$purity_trunk < 0.01 &
                            manifest$case.id == patient,]

#read in previously created pseudobulk data from normal tissue
patient_manifest = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_pseudo_bulk_no_tumour_metadata.txt"), header = T, sep = '\t', stringsAsFactors = F, quote = '')
patient_manifest_flt = patient_manifest[patient_manifest$purity_trunc < 0.01,]
row.names(patient_manifest_flt) = patient_manifest_flt$PD

#tumour root
snps = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars/", patient, "_put_tum_root_snp_combined_method_reviewed.txt"), header = T, sep = "\t", stringsAsFactors = F, row.names = 1)
snps_flt = row.names(snps[snps$Keep == "Y",])

#create vaf per region
NR = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_merged_bulk_filtered_snp_NR_non_tumour_only.txt"), header = T, sep = "\t", stringsAsFactors = F)
NV = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_merged_bulk_filtered_snp_NV_non_tumour_only.txt"), header = T, sep = "\t", stringsAsFactors = F)

patient_manifest_flt_2 = patient_manifest_flt[patient_manifest_flt$PD %in% c("PD50297ar", "PD50297aq", "PD50297as", "PD50297ae", "PD50297ac", "PD50297ak", "PD50297al", "PD50297ay", "PD50297ab", "PD50297z", "PD50297af", "PD50297ah", "PD50297b"),]
patient_manifest_flt_2$region = NA
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297ar", "PD50297aq"),]$region = "left_cerebral_cortex"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297as", "PD50297ae"),]$region = "right_cerebral_cortex"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297ac", "PD50297ak"),]$region = "left_cerebellum"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297al", "PD50297ay"),]$region = "right_cerebellum"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297ab"),]$region = "pineal_gland"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297z"),]$region = "pituitary_gland"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297af"),]$region = "left_svz"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297ah"),]$region = "right_svz"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297b"),]$region = "blood"

for(region in unique(patient_manifest_flt_2$region)){
  temp = weighted.mean(patient_manifest_flt_2[patient_manifest_flt_2$region == region,]$purity_trunc, patient_manifest_flt_2[patient_manifest_flt_2$region == region,]$total_depth)
  print(paste0(region, " = ", temp, " (", temp <0.005, ")"))
}

#all <0.005

for(region in unique(patient_manifest_flt_2$region)){
  temp = sum(patient_manifest_flt_2[patient_manifest_flt_2$region == region,]$total_depth)
  print(paste0(region, " = ", temp))
}

merged_NV = data.frame(left_cerebral_cortex = rowSums(NV[, c("PD50297ar", "PD50297aq")]),
                       right_cerebral_cortex = rowSums(NV[, c("PD50297as", "PD50297ae")]),
                       left_cerebellum = rowSums(NV[, c("PD50297ac", "PD50297ak")]),
                       right_cerebellum = rowSums(NV[, c("PD50297al", "PD50297ay")]),
                       pineal_gland = NV[, "PD50297ab"],
                       pituitary_gland = NV[, "PD50297z"],
                       left_svz = NV[, "PD50297af"],
                       right_svz = NV[, "PD50297ah"],
                       blood = NV[, "PD50297b"])
merged_NR = data.frame(left_cerebral_cortex = rowSums(NR[, c("PD50297ar", "PD50297aq")]),
                       right_cerebral_cortex = rowSums(NR[, c("PD50297as", "PD50297ae")]),
                       left_cerebellum = rowSums(NR[, c("PD50297ac", "PD50297ak")]),
                       right_cerebellum = rowSums(NR[, c("PD50297al", "PD50297ay")]),
                       pineal_gland = NR[, "PD50297ab"],
                       pituitary_gland = NR[, "PD50297z"],
                       left_svz = NR[, "PD50297af"],
                       right_svz = NR[, "PD50297ah"],
                       blood = NR[, "PD50297b"])

merged_NR[merged_NR == 0] = 1
merged_VAF = merged_NV / merged_NR
merged_VAF_flt = merged_VAF[snps_flt,]
merged_VAF_flt$mut = row.names(merged_VAF_flt)

library(reshape2)
library(ggplot2)

merged_VAF_flt.m = reshape2::melt(merged_VAF_flt, id.vars = "mut")
merged_VAF_flt.m$mut = factor(merged_VAF_flt.m$mut, levels = names(sort(apply(merged_VAF[snps_flt, ], 1, mean), decreasing = T)))

merged_VAF_flt.m$region = NA
merged_VAF_flt.m[grepl("cerebral", merged_VAF_flt.m$variable),]$region = "cerebral cortex"
merged_VAF_flt.m[grepl("cerebellum", merged_VAF_flt.m$variable),]$region = "cerebellum"
merged_VAF_flt.m[grepl("gland", merged_VAF_flt.m$variable),]$region = "other CNS"
merged_VAF_flt.m[grepl("svz", merged_VAF_flt.m$variable),]$region = "subventricular zone"
merged_VAF_flt.m[grepl("blood", merged_VAF_flt.m$variable),]$region = "blood"

merged_VAF_flt.m$position = NA
merged_VAF_flt.m[grepl("left", merged_VAF_flt.m$variable),]$position = "left"
merged_VAF_flt.m[grepl("right", merged_VAF_flt.m$variable),]$position = "right"
merged_VAF_flt.m[merged_VAF_flt.m$variable == "pituitary_gland",]$position = "pituitary gland"
merged_VAF_flt.m[merged_VAF_flt.m$variable == "pineal_gland",]$position = "pineal gland"
merged_VAF_flt.m[merged_VAF_flt.m$variable == "blood",]$position = "blood"

merged_VAF_flt.m$region = factor(merged_VAF_flt.m$region, levels = c("cerebral cortex", "subventricular zone", "cerebellum", "other CNS", "blood"))
aggregate(value ~ region + position, merged_VAF_flt.m[!merged_VAF_flt.m$mut %in% c("chrX_52010308_C_T", "chr17_67362034_C_A", "chr4_68023594_C_T", "chr8_87155015_G_A"),], mean)

merged_VAF_flt.m$value = round(merged_VAF_flt.m$value, digits = 3)

merged_VAF_flt.m$is.zero = ifelse(merged_VAF_flt.m$value == 0, T, F)

pd50297_merged_VAF_flt.m = merged_VAF_flt.m

ggplot(merged_VAF_flt.m, mapping = aes(x = mut, y = value)) +
  geom_line(mapping = aes(col = position, group = position)) +
  geom_point(mapping = aes(fill = is.zero), pch = 21, size = 2) +
  theme_bw() +
  facet_grid(region ~ ., switch = "y") +
  scale_y_log10(position = "right", limits = c(1e-3, 1)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face = "bold"), 
        strip.background=element_rect(fill = NA),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "grey95"),
        strip.text.y.left = element_text(angle = 0, hjust = 1),
        strip.background.y = element_blank()) +
  xlab("") +
  ylab("VAF") +
  coord_cartesian(clip = 'off') +
  scale_fill_manual(values = c("white", "black"), guide = "none")

```

#####Per mutations stats

```{r}

patient = "PD50297"

#read in sample manifest
manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
patient_norm_manifest = manifest[manifest$keep == "Y" &
                            manifest$purity_trunk < 0.01 &
                            manifest$case.id == patient,]

#read in previously created pseudobulk data from normal tissue
patient_manifest = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_pseudo_bulk_no_tumour_metadata.txt"), header = T, sep = '\t', stringsAsFactors = F, quote = '')
patient_manifest_flt = patient_manifest[patient_manifest$purity_trunc < 0.01,]
row.names(patient_manifest_flt) = patient_manifest_flt$PD

#tumour root
snps = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars/", patient, "_put_tum_root_snp_combined_method_reviewed.txt"), header = T, sep = "\t", stringsAsFactors = F, row.names = 1)
snps_flt = row.names(snps[snps$Keep == "Y",])

#create vaf per region
NR = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_merged_bulk_filtered_snp_NR_non_tumour_only.txt"), header = T, sep = "\t", stringsAsFactors = F)
NV = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_merged_bulk_filtered_snp_NV_non_tumour_only.txt"), header = T, sep = "\t", stringsAsFactors = F)

patient_manifest_flt_2 = patient_manifest_flt[patient_manifest_flt$PD %in% c("PD50297ar", "PD50297aq", "PD50297as", "PD50297ae", "PD50297ac", "PD50297ak", "PD50297al", "PD50297ay", "PD50297ab", "PD50297z", "PD50297af", "PD50297ah", "PD50297b"),]
patient_manifest_flt_2$region = NA
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297ar", "PD50297aq"),]$region = "left_cerebral_cortex"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297as", "PD50297ae"),]$region = "right_cerebral_cortex"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297ac", "PD50297ak"),]$region = "left_cerebellum"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297al", "PD50297ay"),]$region = "right_cerebellum"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297ab"),]$region = "pineal_gland"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297z"),]$region = "pituitary_gland"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297af"),]$region = "left_svz"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297ah"),]$region = "right_svz"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD50297b"),]$region = "blood"

for(region in unique(patient_manifest_flt_2$region)){
  temp = weighted.mean(patient_manifest_flt_2[patient_manifest_flt_2$region == region,]$purity_trunc, patient_manifest_flt_2[patient_manifest_flt_2$region == region,]$total_depth)
  print(paste0(region, " = ", temp, " (", temp <0.005, ")"))
}

#all <0.005

merged_NV = data.frame(left_cerebral_cortex = rowSums(NV[, c("PD50297ar", "PD50297aq")]),
                       right_cerebral_cortex = rowSums(NV[, c("PD50297as", "PD50297ae")]),
                       left_cerebellum = rowSums(NV[, c("PD50297ac", "PD50297ak")]),
                       right_cerebellum = rowSums(NV[, c("PD50297al", "PD50297ay")]),
                       pineal_gland = NV[, "PD50297ab"],
                       pituitary_gland = NV[, "PD50297z"],
                       left_svz = NV[, "PD50297af"],
                       right_svz = NV[, "PD50297ah"],
                       blood = NV[, "PD50297b"])
merged_NR = data.frame(left_cerebral_cortex = rowSums(NR[, c("PD50297ar", "PD50297aq")]),
                       right_cerebral_cortex = rowSums(NR[, c("PD50297as", "PD50297ae")]),
                       left_cerebellum = rowSums(NR[, c("PD50297ac", "PD50297ak")]),
                       right_cerebellum = rowSums(NR[, c("PD50297al", "PD50297ay")]),
                       pineal_gland = NR[, "PD50297ab"],
                       pituitary_gland = NR[, "PD50297z"],
                       left_svz = NR[, "PD50297af"],
                       right_svz = NR[, "PD50297ah"],
                       blood = NR[, "PD50297b"])

merged_NV_flt = merged_NV[snps_flt,]
merged_NR_flt = merged_NR[snps_flt,]

merged_df_stat = data.frame(mutID = row.names(merged_NV_flt), l_r_cerebrum = NA, l_r_svz = NA, l_r_cerebellum = NA)

for(i in 1:nrow(merged_df_stat)){
  
  #cerebrum
  xl1 = merged_NR_flt[i, "left_cerebral_cortex"] - merged_NV_flt[i, "left_cerebral_cortex"]
  xr1 = merged_NR_flt[i, "right_cerebral_cortex"] - merged_NV_flt[i, "right_cerebral_cortex"]
  xl2 = merged_NV_flt[i, "left_cerebral_cortex"]
  xr2 = merged_NV_flt[i, "right_cerebral_cortex"]

  tbl = matrix(c(xl1, xl2, xr1, xr2),
        nrow = 2,
        dimnames = list(status = c("wt", "mt"),
                       side = c("left", "right")))
  
  res = fisher.test(tbl)
  merged_df_stat[i,]$l_r_cerebrum = res$p.value
  
  #svz
  xl1 = merged_NR_flt[i, "left_svz"] - merged_NV_flt[i, "left_svz"]
  xr1 = merged_NR_flt[i, "right_svz"] - merged_NV_flt[i, "right_svz"]
  xl2 = merged_NV_flt[i, "left_svz"]
  xr2 = merged_NV_flt[i, "right_svz"]

  tbl = matrix(c(xl1, xl2, xr1, xr2),
        nrow = 2,
        dimnames = list(status = c("wt", "mt"),
                       side = c("left", "right")))
  
  res = fisher.test(tbl)
  merged_df_stat[i,]$l_r_svz = res$p.value
  
  #cerebellum
  xl1 = merged_NR_flt[i, "left_cerebellum"] - merged_NV_flt[i, "left_cerebellum"]
  xr1 = merged_NR_flt[i, "right_cerebellum"] - merged_NV_flt[i, "right_cerebellum"]
  xl2 = merged_NV_flt[i, "left_cerebellum"]
  xr2 = merged_NV_flt[i, "right_cerebellum"]

  tbl = matrix(c(xl1, xl2, xr1, xr2),
        nrow = 2,
        dimnames = list(status = c("wt", "mt"),
                       side = c("left", "right")))
  
  res = fisher.test(tbl)
  merged_df_stat[i,]$l_r_cerebellum = res$p.value
  
}

row.names(merged_df_stat) = merged_df_stat$mutID
merged_df_stat[levels(pd50297_merged_VAF_flt.m$mut), ]

```

####PD51122

```{r}

patient = "PD51122"

#read in sample manifest
manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
patient_norm_manifest = manifest[manifest$keep == "Y" &
                            manifest$purity_trunk < 0.01 &
                            manifest$case.id == patient,]

#read in previously created pseudobulk data from normal tissue
patient_manifest = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_pseudo_bulk_no_tumour_metadata.txt"), header = T, sep = '\t', stringsAsFactors = F, quote = '')
patient_manifest_flt = patient_manifest[patient_manifest$purity_trunc < 0.01,]
row.names(patient_manifest_flt) = patient_manifest_flt$PD

#tumour root
snps = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars/", patient, "_put_tum_root_snp_combined_method_reviewed.txt"), header = T, sep = "\t", stringsAsFactors = F, row.names = 1)
snps_flt = row.names(snps[snps$Keep == "Y",])

#create vaf per region
NR = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_merged_bulk_filtered_snp_NR_non_tumour_only.txt"), header = T, sep = "\t", stringsAsFactors = F)
NV = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_merged_bulk_filtered_snp_NV_non_tumour_only.txt"), header = T, sep = "\t", stringsAsFactors = F)

patient_manifest_flt_2 = patient_manifest_flt[patient_manifest_flt$PD %in% c("PD51122e", "PD51122g", "PD51122f", "PD51122h", "PD51122b", "PD51122p", "PD51122i", "PD51122n", "PD51122d", "PD51122af", "PD51122v", "PD51122w", "PD51122q"),]
patient_manifest_flt_2$region = NA
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51122e", "PD51122g"),]$region = "left_cerebral_cortex"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51122f", "PD51122h"),]$region = "right_cerebral_cortex"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51122b"),]$region = "left_cerebellum"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51122p"),]$region = "right_cerebellum"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51122i", "PD51122n"),]$region = "thalamus"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51122d"),]$region = "pons"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51122af"),]$region = "pituitary_gland"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51122v", "PD51122w"),]$region = "spinal_cord"
#patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51122w"),]$region = "lower_spinal_cord"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51122q"),]$region = "blood"

for(region in unique(patient_manifest_flt_2$region)){
  temp = weighted.mean(patient_manifest_flt_2[patient_manifest_flt_2$region == region,]$purity_trunc, patient_manifest_flt_2[patient_manifest_flt_2$region == region,]$total_depth)
  print(paste0(region, " = ", temp, " (", temp <0.005, ")"))
}
#thalamus >0.005

for(region in unique(patient_manifest_flt_2$region)){
  temp = sum(patient_manifest_flt_2[patient_manifest_flt_2$region == region,]$total_depth)
  print(paste0(region, " = ", temp))
}

merged_NV = data.frame(left_cerebral_cortex = rowSums(NV[, c("PD51122e", "PD51122g")]),
                       right_cerebral_cortex = rowSums(NV[, c("PD51122f", "PD51122h")]),
                       left_cerebellum = NV[, c("PD51122b")],
                       right_cerebellum = NV[, c("PD51122p")],
                       #thalamus = rowSums(NV[, c("PD51122i", "PD51122n")]),
                       pons = NV[, "PD51122d"],
                       pituitary_gland = NV[, "PD51122af"],
                       spinal_cord = rowSums(NV[, c("PD51122v", "PD51122w")]),
                       #lower_spinal_cord = NV[, "PD51122w"],
                       blood = NV[, "PD51122q"])
merged_NR = data.frame(left_cerebral_cortex = rowSums(NR[, c("PD51122e", "PD51122g")]),
                       right_cerebral_cortex = rowSums(NR[, c("PD51122f", "PD51122h")]),
                       left_cerebellum = NR[, c("PD51122b")],
                       right_cerebellum = NR[, c("PD51122p")],
                       #thalamus = rowSums(NR[, c("PD51122i", "PD51122n")]),
                       pons = NR[, "PD51122d"],
                       pituitary_gland = NR[, "PD51122af"],
                       spinal_cord = rowSums(NR[, c("PD51122v", "PD51122w")]),
                       #lower_spinal_cord = NR[, "PD51122w"],
                       blood = NR[, "PD51122q"])

merged_NR[merged_NR == 0] = 1
merged_VAF = merged_NV / merged_NR
merged_VAF_flt = merged_VAF[snps_flt,]
merged_VAF_flt$mut = row.names(merged_VAF_flt)

library(reshape2)
library(ggplot2)

merged_VAF_flt.m = reshape2::melt(merged_VAF_flt, id.vars = "mut")
merged_VAF_flt.m$mut = factor(merged_VAF_flt.m$mut, levels = names(sort(apply(merged_VAF[snps_flt, ], 1, mean), decreasing = T)))

merged_VAF_flt.m$region = NA
merged_VAF_flt.m[grepl("cerebral", merged_VAF_flt.m$variable),]$region = "cerebral cortex"
merged_VAF_flt.m[grepl("cerebellum", merged_VAF_flt.m$variable),]$region = "cerebellum"
merged_VAF_flt.m[grepl("gland|pons", merged_VAF_flt.m$variable),]$region = "other CNS"
#merged_VAF_flt.m[grepl("thalamus", merged_VAF_flt.m$variable),]$region = "thalamus"
merged_VAF_flt.m[grepl("spinal_cord", merged_VAF_flt.m$variable),]$region = "spinal_cord"
merged_VAF_flt.m[grepl("blood", merged_VAF_flt.m$variable),]$region = "blood"

merged_VAF_flt.m$position = NA
merged_VAF_flt.m[grepl("left", merged_VAF_flt.m$variable),]$position = "left"
merged_VAF_flt.m[grepl("right", merged_VAF_flt.m$variable),]$position = "right"
merged_VAF_flt.m[merged_VAF_flt.m$variable == "pituitary_gland",]$position = "pituitary gland"
#merged_VAF_flt.m[merged_VAF_flt.m$variable == "thalamus",]$position = "thalamus"
merged_VAF_flt.m[merged_VAF_flt.m$variable == "pons",]$position = "pons"
merged_VAF_flt.m[merged_VAF_flt.m$variable == "spinal_cord",]$position = "spinal_cord"
#merged_VAF_flt.m[merged_VAF_flt.m$variable == "lower_spinal_cord",]$position = "lower"
merged_VAF_flt.m[merged_VAF_flt.m$variable == "blood",]$position = "blood"

merged_VAF_flt.m$region = factor(merged_VAF_flt.m$region, levels = c("cerebral cortex", "cerebellum", "other CNS", "spinal_cord", "blood"))
merged_VAF_flt.m$value = round(merged_VAF_flt.m$value, digits = 3)

merged_VAF_flt.m$is.zero = ifelse(merged_VAF_flt.m$value == 0, T, F)

pd51122_merged_VAF_flt.m = merged_VAF_flt.m

ggplot(merged_VAF_flt.m, mapping = aes(x = mut, y = value)) +
  geom_line(mapping = aes(col = position, group = position)) +
  geom_point(mapping = aes(fill = is.zero), pch = 21, size = 2) +
  theme_bw() +
  facet_grid(region ~ ., switch = "y") +
  scale_y_log10(position = "right", limits = c(1e-3, 1)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face = "bold"), 
        strip.background=element_rect(fill = NA),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "grey95"),
        strip.text.y.left = element_text(angle = 0, hjust = 1),
        strip.background.y = element_blank()) +
  xlab("") +
  ylab("VAF") +
  coord_cartesian(clip = 'off') +
  scale_fill_manual(values = c("white", "black"), guide = "none")

```

####PD51123

```{r}

patient = "PD51123"

#read in sample manifest
manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
patient_norm_manifest = manifest[manifest$keep == "Y" &
                            manifest$purity_trunk < 0.01 &
                            manifest$case.id == patient,]

#read in previously created pseudobulk data from normal tissue
patient_manifest = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_pseudo_bulk_no_tumour_metadata.txt"), header = T, sep = '\t', stringsAsFactors = F, quote = '')
patient_manifest_flt = patient_manifest[patient_manifest$purity_trunc < 0.01,]
row.names(patient_manifest_flt) = patient_manifest_flt$PD

#tumour root
snps = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars/", patient, "_put_tum_root_snp_combined_method_reviewed.txt"), header = T, sep = "\t", stringsAsFactors = F, row.names = 1)
snps_flt = row.names(snps[snps$Keep == "Y",])

#create vaf per region
NR = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_merged_bulk_filtered_snp_NR_non_tumour_only.txt"), header = T, sep = "\t", stringsAsFactors = F)
NV = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/merged_pileup/", patient, "_merged_bulk_filtered_snp_NV_non_tumour_only.txt"), header = T, sep = "\t", stringsAsFactors = F)

patient_manifest_flt_2 = patient_manifest_flt[patient_manifest_flt$PD %in% c("PD51123f", "PD51123s", "PD51123w", "PD51123z", "PD51123b", "PD51123e", "PD51123j", "PD51123k", "PD51123d", "PD51123u", "PD51123v", "PD51123y", "PD51123t", "PD51123r", "PD51123aa", "PD51123h"),]
patient_manifest_flt_2$region = NA
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51123f", "PD51123s", "PD51123w", "PD51123z"),]$region = "left_cerebral_cortex"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51123b", "PD51123e", "PD51123j", "PD51123k"),]$region = "right_cerebral_cortex"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51123d"),]$region = "left_cerebellum"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51123u"),]$region = "right_cerebellum"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51123v", "PD51122n"),]$region = "thalamus"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51123y"),]$region = "pituitary_gland"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51123t", "PD51123r"),]$region = "spinal_cord"
#patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51123r"),]$region = "lower_spinal_cord"
patient_manifest_flt_2[patient_manifest_flt_2$PD %in% c("PD51123aa", "PD51123h"),]$region = "skin"

for(region in unique(patient_manifest_flt_2$region)){
  temp = weighted.mean(patient_manifest_flt_2[patient_manifest_flt_2$region == region,]$purity_trunc, patient_manifest_flt_2[patient_manifest_flt_2$region == region,]$total_depth)
  print(paste0(region, " = ", temp, " (", temp <0.005, ")"))
}
#all <0.005

for(region in unique(patient_manifest_flt_2$region)){
  temp = sum(patient_manifest_flt_2[patient_manifest_flt_2$region == region,]$total_depth)
  print(paste0(region, " = ", temp))
}

merged_NV = data.frame(left_cerebral_cortex = rowSums(NV[, c("PD51123f", "PD51123s", "PD51123w", "PD51123z")]),
                       right_cerebral_cortex = rowSums(NV[, c("PD51123b", "PD51123e", "PD51123j", "PD51123k")]),
                       left_cerebellum = NV[, c("PD51123d")],
                       right_cerebellum = NV[, c("PD51123u")],
                       thalamus = NV[, c("PD51123v")],
                       pituitary_gland = NV[, "PD51123y"],
                       spinal_cord = rowSums(NV[, c("PD51123t", "PD51123r")]),
                       #lower_spinal_cord = NV[, c("PD51123r")],
                       skin = rowSums(NV[, c("PD51123aa", "PD51123h")]))
merged_NR = data.frame(left_cerebral_cortex = rowSums(NR[, c("PD51123f", "PD51123s", "PD51123w", "PD51123z")]),
                       right_cerebral_cortex = rowSums(NR[, c("PD51123b", "PD51123e", "PD51123j", "PD51123k")]),
                       left_cerebellum = NR[, c("PD51123d")],
                       right_cerebellum = NR[, c("PD51123u")],
                       thalamus = NR[, c("PD51123v")],
                       pituitary_gland = NR[, "PD51123y"],
                       spinal_cord = rowSums(NR[, c("PD51123t", "PD51123r")]),
                       #lower_spinal_cord = NR[, c("PD51123r")],
                       skin = rowSums(NR[, c("PD51123aa", "PD51123h")]))

merged_NR[merged_NR == 0] = 1
merged_VAF = merged_NV / merged_NR
merged_VAF_flt = merged_VAF[snps_flt,]
merged_VAF_flt$mut = row.names(merged_VAF_flt)

library(reshape2)
library(ggplot2)

merged_VAF_flt.m = reshape2::melt(merged_VAF_flt, id.vars = "mut")
merged_VAF_flt.m$mut = factor(merged_VAF_flt.m$mut, levels = names(sort(apply(merged_VAF[snps_flt, ], 1, mean), decreasing = T)))

merged_VAF_flt.m$region = NA
merged_VAF_flt.m[grepl("cerebral", merged_VAF_flt.m$variable),]$region = "cerebral cortex"
merged_VAF_flt.m[grepl("cerebellum", merged_VAF_flt.m$variable),]$region = "cerebellum"
merged_VAF_flt.m[grepl("gland", merged_VAF_flt.m$variable),]$region = "other CNS"
merged_VAF_flt.m[grepl("thalamus", merged_VAF_flt.m$variable),]$region = "other CNS"
merged_VAF_flt.m[grepl("spinal_cord", merged_VAF_flt.m$variable),]$region = "spinal_cord"
merged_VAF_flt.m[grepl("skin", merged_VAF_flt.m$variable),]$region = "skin"

merged_VAF_flt.m$position = NA
merged_VAF_flt.m[grepl("left", merged_VAF_flt.m$variable),]$position = "left"
merged_VAF_flt.m[grepl("right", merged_VAF_flt.m$variable),]$position = "right"
merged_VAF_flt.m[merged_VAF_flt.m$variable == "pituitary_gland",]$position = "pituitary gland"
merged_VAF_flt.m[merged_VAF_flt.m$variable == "thalamus",]$position = "thalamus"
merged_VAF_flt.m[merged_VAF_flt.m$variable == "spinal_cord",]$position = "spinal_cord"
#merged_VAF_flt.m[merged_VAF_flt.m$variable == "lower_spinal_cord",]$position = "lower"
merged_VAF_flt.m[merged_VAF_flt.m$variable == "skin",]$position = "skin"

merged_VAF_flt.m$region = factor(merged_VAF_flt.m$region, levels = c("cerebral cortex", "cerebellum", "other CNS", "spinal_cord", "skin"))
merged_VAF_flt.m$value = round(merged_VAF_flt.m$value, digits = 3)

merged_VAF_flt.m$is.zero = ifelse(merged_VAF_flt.m$value == 0, T, F)

pd51123_merged_VAF_flt.m = merged_VAF_flt.m

ggplot(merged_VAF_flt.m, mapping = aes(x = mut, y = value)) +
  geom_line(mapping = aes(col = position, group = position), alpha = 0.5) +
  geom_point(mapping = aes(fill = is.zero), pch = 21, size = 2) +
  theme_bw() +
  facet_grid(region ~ ., switch = "y") +
  scale_y_log10(position = "right", limits = c(1e-3, 1)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face = "bold"), 
        strip.background=element_rect(fill = NA),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "grey95"),
        strip.text.y.left = element_text(angle = 0, hjust = 1),
        strip.background.y = element_blank()) +
  xlab("") +
  ylab("VAF") +
  coord_cartesian(clip = 'off') +
  scale_fill_manual(values = c("white", "black"), guide = "none")

```

####Per germ layer stats (normal LCM)

```{r}

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

non_tumour_lcm_manifest = manifest[manifest$keep == "Y" &
                            manifest$purity_trunk < 0.01 &
                            manifest$experimental_arm == "WGS_LCM" & 
                            !manifest$sample %in% c("PD51122ac_lo0013", "PD51122ag_lo0007"),]
table(non_tumour_lcm_manifest$predominant_germ_layer)

#all subs NR and NV
pd50297_nr = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/PD50297_NR_merged_snps.txt", header = T, sep = "\t", stringsAsFactors = F)
pd50297_nv = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/PD50297_NV_merged_snps.txt", header = T, sep = "\t", stringsAsFactors = F)
pd51122_nr = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/PD51122_NR_merged_snps.txt", header = T, sep = "\t", stringsAsFactors = F)
pd51122_nv = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/PD51122_NV_merged_snps.txt", header = T, sep = "\t", stringsAsFactors = F)
pd51123_nr = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/PD51123_NR_merged_snps.txt", header = T, sep = "\t", stringsAsFactors = F)
pd51123_nv = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data_mk2/merged_final_subs/PD51123_NV_merged_snps.txt", header = T, sep = "\t", stringsAsFactors = F)

#tumour root
pd50297_snps = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars/PD50297_put_tum_root_snp_combined_method_reviewed.txt"), header = T, sep = "\t", stringsAsFactors = F, row.names = 1)
pd50297_snps_flt = row.names(pd50297_snps[pd50297_snps$Keep == "Y",])
pd51122_snps = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars/PD51122_put_tum_root_snp_combined_method_reviewed.txt"), header = T, sep = "\t", stringsAsFactors = F, row.names = 1)
pd51122_snps_flt = row.names(pd51122_snps[pd51122_snps$Keep == "Y",])
pd51123_snps = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tum_root_vars/PD51123_put_tum_root_snp_combined_method_reviewed.txt"), header = T, sep = "\t", stringsAsFactors = F, row.names = 1)
pd51123_snps_flt = row.names(pd51123_snps[pd51123_snps$Keep == "Y",])

#subset
pd50297_ecto = non_tumour_lcm_manifest[non_tumour_lcm_manifest$case.id == "PD50297" & non_tumour_lcm_manifest$predominant_germ_layer == "Ectoderm",]$sample
pd50297_endo = non_tumour_lcm_manifest[non_tumour_lcm_manifest$case.id == "PD50297" & non_tumour_lcm_manifest$predominant_germ_layer == "Endoderm",]$sample
pd50297_meso = non_tumour_lcm_manifest[non_tumour_lcm_manifest$case.id == "PD50297" & non_tumour_lcm_manifest$predominant_germ_layer == "Mesoderm",]$sample

pd50297_root_df = data.frame(ecto = rowSums(pd50297_nv[pd50297_snps_flt, pd50297_ecto]) / rowSums(pd50297_nr[pd50297_snps_flt, pd50297_ecto]),
                             endo = rowSums(pd50297_nv[pd50297_snps_flt, pd50297_endo]) / rowSums(pd50297_nr[pd50297_snps_flt, pd50297_endo]),
                             meso = rowSums(pd50297_nv[pd50297_snps_flt, pd50297_meso]) / rowSums(pd50297_nr[pd50297_snps_flt, pd50297_meso]))

pd51122_ecto = non_tumour_lcm_manifest[non_tumour_lcm_manifest$case.id == "PD51122" & non_tumour_lcm_manifest$predominant_germ_layer == "Ectoderm",]$sample
pd51122_endo = non_tumour_lcm_manifest[non_tumour_lcm_manifest$case.id == "PD51122" & non_tumour_lcm_manifest$predominant_germ_layer == "Endoderm",]$sample
pd51122_meso = non_tumour_lcm_manifest[non_tumour_lcm_manifest$case.id == "PD51122" & non_tumour_lcm_manifest$predominant_germ_layer == "Mesoderm",]$sample

pd51122_root_df = data.frame(ecto = rowSums(pd51122_nv[pd51122_snps_flt, pd51122_ecto]) / rowSums(pd51122_nr[pd51122_snps_flt, pd51122_ecto]),
                             endo = rowSums(pd51122_nv[pd51122_snps_flt, pd51122_endo]) / rowSums(pd51122_nr[pd51122_snps_flt, pd51122_endo]),
                             meso = rowSums(pd51122_nv[pd51122_snps_flt, pd51122_meso]) / rowSums(pd51122_nr[pd51122_snps_flt, pd51122_meso]))

pd51123_ecto = non_tumour_lcm_manifest[non_tumour_lcm_manifest$case.id == "PD51123" & non_tumour_lcm_manifest$predominant_germ_layer == "Ectoderm",]$sample
pd51123_endo = non_tumour_lcm_manifest[non_tumour_lcm_manifest$case.id == "PD51123" & non_tumour_lcm_manifest$predominant_germ_layer == "Endoderm",]$sample
pd51123_meso = non_tumour_lcm_manifest[non_tumour_lcm_manifest$case.id == "PD51123" & non_tumour_lcm_manifest$predominant_germ_layer == "Mesoderm",]$sample

pd51123_root_df = data.frame(ecto = rowSums(pd51123_nv[pd51123_snps_flt, pd51123_ecto]) / rowSums(pd51123_nr[pd51123_snps_flt, pd51123_ecto]),
                             endo = 0,
                             meso = rowSums(pd51123_nv[pd51123_snps_flt, pd51123_meso]) / rowSums(pd51123_nr[pd51123_snps_flt, pd51123_meso])) #no endo sampling

pd50297_root_df$case.id = "PD50297"
pd51122_root_df$case.id = "PD51122"
pd51123_root_df$case.id = "PD51123"

all_root_df = rbind(pd50297_root_df, pd51122_root_df, pd51123_root_df)
all_root_df$mutID = row.names(all_root_df)

library(reshape2)

all_root_df.m = reshape2::melt(all_root_df, id.vars = c("case.id", "mutID"))
all_root_df.m$mutID = factor(all_root_df.m$mutID, levels = c(c(levels(pd50297_merged_VAF_flt.m$mut), levels(pd51122_merged_VAF_flt.m$mut), levels(pd51123_merged_VAF_flt.m$mut))))
all_root_df.m$value = round(all_root_df.m$value, digits = 3)
all_root_df.m$is.zero = ifelse(all_root_df.m$value == 0, T, F)

germ_layer_col_fun = c("#75539e", "#1fd655", "#FFAE42", "grey85")
names(germ_layer_col_fun) = c("ectoderm", "endoderm", "mesoderm", "mixed")

#axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face = "bold")

p5 = ggplot(all_root_df.m, mapping = aes(x = mutID, y = value)) +
  geom_line(mapping = aes(col = variable, group = variable), alpha = 1, size = 1.5) +
  geom_point(mapping = aes(fill = is.zero), pch = 21, size = 2) +
  theme_bw() +
  facet_grid(. ~ case.id, scales = "free", space = "free") +
  scale_y_log10(position = "right", limits = c(1e-3, 1)) +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        strip.background=element_rect(fill = NA),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "grey97"),
        strip.text.y.left = element_text(angle = 0, hjust = 1),
        strip.background.y = element_blank(),
        plot.margin = unit(c(0, 0, 0, 0), "cm"))  +
  xlab("") +
  ylab("VAF") +
  coord_cartesian(clip = 'off') +
  scale_fill_manual(values = c("white", "black"), guide = "none") +
  scale_color_manual(values = c("#75539e", "#1fd655", "#FFAE42"), guide = "none")

```


####Plot

```{r}

library(ggplot2)
library(ggpubr)
library(cowplot)

cols = c("#e83b3d", "#88d0ed", "#999999", "#FF007F", "#999999", "#999999", "#999999", "#999999", "#999999")
names(cols) = c("left", "right", "spinal_cord", "pituitary gland", "pineal gland", "thalamus", "pons", "blood", "skin")

pd50297_merged_VAF_flt.m$position = factor(pd50297_merged_VAF_flt.m$position, levels = c("left", "right", "upper", "lower", "pituitary gland", "pineal gland", "thalamus", "pons", "blood", "skin"))
pd51122_merged_VAF_flt.m$position = factor(pd51122_merged_VAF_flt.m$position, levels = c("left", "right", "upper", "lower", "pituitary gland", "pineal gland", "thalamus", "pons", "blood", "skin"))
pd51123_merged_VAF_flt.m$position = factor(pd51123_merged_VAF_flt.m$position, levels = c("left", "right", "upper", "lower", "pituitary gland", "pineal gland", "thalamus", "pons", "blood", "skin"))

p1 = ggplot(pd50297_merged_VAF_flt.m, mapping = aes(x = mut, y = value)) +
      geom_line(mapping = aes(col = position, group = position), alpha = 1, size = 1.5) +
      geom_point(mapping = aes(fill = is.zero), pch = 21, size = 2) +
      theme_bw() +
      facet_grid(region ~ ., switch = "y") +
      scale_y_log10(position = "right", limits = c(1e-3, 1)) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face = "bold"), 
          strip.background=element_rect(fill = NA),
          panel.border = element_blank(),
          panel.background = element_rect(fill = "grey97"),
          strip.text.y.left = element_blank(),
          strip.background.y = element_blank(),
          plot.margin = unit(c(0, 0, 0, 0), "cm"),
          axis.text.y = element_blank(),
          axis.ticks.y = element_blank())  +
    xlab("") +
    ylab("") +
    coord_cartesian(clip = 'off') +
    scale_fill_manual(values = c("white", "black"), guide = "none") +
    scale_color_manual(values = cols, guide = "none")

p2 = ggplot(pd51122_merged_VAF_flt.m, mapping = aes(x = mut, y = value)) +
        geom_line(mapping = aes(col = position, group = position), alpha = 1, size = 1.5) +
        geom_point(mapping = aes(fill = is.zero), pch = 21, size = 2) +
        theme_bw() +
        facet_grid(region ~ ., switch = "y") +
        scale_y_log10(position = "right", limits = c(1e-3, 1)) +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face = "bold"), 
              strip.background=element_rect(fill = NA),
              panel.border = element_blank(),
              panel.background = element_rect(fill = "grey97"),
              strip.text.y.left = element_blank(),
              strip.background.y = element_blank(),
              plot.margin = unit(c(0, 0, 0, 0), "cm"),
              axis.text.y = element_blank(),
              axis.ticks.y = element_blank())  +
        xlab("") +
        ylab("") +
        coord_cartesian(clip = 'off') +
        scale_fill_manual(values = c("white", "black"), guide = "none") +
        scale_color_manual(values = cols, guide = "none")

p3 = ggplot(pd51123_merged_VAF_flt.m, mapping = aes(x = mut, y = value)) +
        geom_line(mapping = aes(col = position, group = position), alpha = 1, size = 1.5) +
        geom_point(mapping = aes(fill = is.zero), pch = 21, size = 2) +
        theme_bw() +
        facet_grid(region ~ ., switch = "y") +
        scale_y_log10(position = "right", limits = c(1e-3, 1)) +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face = "bold"), 
              strip.background=element_rect(fill = NA),
              panel.border = element_blank(),
              panel.background = element_rect(fill = "grey97"),
              strip.text.y.left = element_blank(),
              strip.background.y = element_blank(),
              plot.margin = unit(c(0, 0, 0, 0), "cm"))  +
        xlab("") +
        ylab("VAF") +
        coord_cartesian(clip = 'off') +
        scale_fill_manual(values = c("white", "black"), guide = "none") +
        scale_color_manual(values = cols, guide = "none")

p4 = plot_grid(p1, p2, p3, ncol = 3, rel_widths = c(14.2, 4, 5.8))

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/tum_root_plot_20221123.pdf", height = 6, width = 5, useDingbats = F)
plot_grid(p5, p4, ncol = 1, align = "h", axis = "lr", rel_heights = c(1, 5))
dev.off()

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/PD50297_tum_root_plot_20221123.pdf", height = 5, width = 4, useDingbats = F)
ggplot(pd50297_merged_VAF_flt.m, mapping = aes(x = mut, y = value)) +
  geom_line(mapping = aes(col = position, group = position), alpha = 1, size = 1.5) +
  geom_point(mapping = aes(fill = is.zero), pch = 21, size = 2) +
  theme_bw() +
  facet_grid(region ~ ., switch = "y") +
  scale_y_log10(position = "right", limits = c(1e-3, 1)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face = "bold"), 
        strip.background=element_rect(fill = NA),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "grey97"),
        strip.text.y.left = element_text(angle = 0, hjust = 1),
        strip.background.y = element_blank(),
        plot.margin = unit(c(0, 0, 0, 0), "cm"))  +
  xlab("") +
  ylab("VAF") +
  coord_cartesian(clip = 'off') +
  scale_fill_manual(values = c("white", "black"), guide = "none") +
  scale_color_manual(values = cols, guide = "none")
dev.off()

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/PD51122_tum_root_plot_20221123.pdf", height = 5, width = 2.14, useDingbats = F)
ggplot(pd51122_merged_VAF_flt.m, mapping = aes(x = mut, y = value)) +
  geom_line(mapping = aes(col = position, group = position), alpha = 1, size = 1.5) +
  geom_point(mapping = aes(fill = is.zero), pch = 21, size = 2) +
  theme_bw() +
  facet_grid(region ~ ., switch = "y") +
  scale_y_log10(position = "right", limits = c(1e-3, 1)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face = "bold"), 
        strip.background=element_rect(fill = NA),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "grey97"),
        strip.text.y.left = element_text(angle = 0, hjust = 1),
        strip.background.y = element_blank(),
        plot.margin = unit(c(0, 0, 0, 0), "cm"))  +
  xlab("") +
  ylab("VAF") +
  coord_cartesian(clip = 'off') +
  scale_fill_manual(values = c("white", "black"), guide = "none") +
  scale_color_manual(values = cols, guide = "none")
dev.off()

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/PD51123_tum_root_plot_20221123.pdf", height = 5, width = 2.275, useDingbats = F)
ggplot(pd51123_merged_VAF_flt.m, mapping = aes(x = mut, y = value)) +
  geom_line(mapping = aes(col = position, group = position), alpha = 1, size = 1.5) +
  geom_point(mapping = aes(fill = is.zero), pch = 21, size = 2) +
  theme_bw() +
  facet_grid(region ~ ., switch = "y") +
  scale_y_log10(position = "right", limits = c(1e-3, 1)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face = "bold"), 
        strip.background=element_rect(fill = NA),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "grey97"),
        strip.text.y.left = element_text(angle = 0, hjust = 1),
        strip.background.y = element_blank(),
        plot.margin = unit(c(0, 0, 0, 0), "cm"))  +
  xlab("") +
  ylab("VAF") +
  coord_cartesian(clip = 'off') +
  scale_fill_manual(values = c("white", "black"), guide = "none") +
  scale_color_manual(values = cols, guide = "none")
dev.off()

```

##HQ tumour stats

```{r}

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

hq_tumour_samples = manifest[manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$average.reads.per.chromosome.copy.tumour >= 5 & (manifest$CNAqc_check == "PASS" | manifest$ascatPCA_gof >= 95), ]$sample #gold standard samples

#total
table(substr(hq_tumour_samples, 0, 7))
#PD50297 PD51122 PD51123 
#    47      69      36
sum(table(substr(hq_tumour_samples, 0, 7))) #152

#lcm
table(substr(hq_tumour_samples[grepl("lo", hq_tumour_samples)], 0, 7))
#PD50297 PD51122 PD51123 
#    41      62      31 
sum(table(substr(hq_tumour_samples[grepl("lo", hq_tumour_samples)], 0, 7))) #134

#bulk
table(substr(hq_tumour_samples[!grepl("lo", hq_tumour_samples)], 0, 7))
#PD50297 PD51122 PD51123 
#     6       7       5 
sum(table(substr(hq_tumour_samples[!grepl("lo", hq_tumour_samples)], 0, 7))) #18

```

##Fig. 4d - PDGFRA copy number plots

###Calculate local coverage

```{r}

library(biomaRt)

gene = "PDGFRA"

#  load biomart
ensembl = useMart("ensembl")
ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")
getAtt <- c('chromosome_name', 'start_position', 'end_position',
            'strand','exon_chrom_start','exon_chrom_end', 'rank')

# find exon location
elocs <- getBM(attributes=c(getAtt),filters=c("hgnc_symbol","with_ccds"), value=list(gene,TRUE),mart=ensembl)
min(elocs$exon_chrom_start) #54229293
max(elocs$exon_chrom_end) #54298245

```

Create a bed file with the genome in 500bp windows

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/localised_copy_number_calling

bedtools makewindows \
-g /lustre/scratch119/casm/team294rr/rs30/External_Databases/References/GRCh38_full_analysis_set_plus_decoy_hla/chrom.sizes.txt \
-w 500 \
-i winnum > 500bp.window.bed

for SAMPLE in PD51123o_lo0008 PD51123o_lo0012 PD51123q;
do
#echo $SAMPLE
bsub -q normal -o ${SAMPLE}_pdgra.out -e ${SAMPLE}_pdgra.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 ~/scripts/segment_cov_20221031.sh chr4 54229000 54298500 $SAMPLE PD51123aa3
done

```

```{bash}

#!/bin/bash
#segment_cov_20221031.sh

#bsub -q yesterday -Is -n 4 -R 'select[mem>=80000] rusage[mem=80000] span[hosts=1]' -M80000 bash
#cd /lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/localised_copy_number_calling

#chr="chr4"
#start=51000000
#end=56000000
#tumour="PD51122ai"
#normal="PD51122q"

chr=$1
start=$2
end=$3
tumour=$4
normal=$5

#get per bp coverage
samtools depth \
-a \
-r ${chr}:${start}-${end} \
-q 25 \
-Q 30 \
-o ${tumour}.tumour.${chr}.${start}.${end}.seg.txt \
/nfs/cancer_ref01/nst_links/live/2571/${tumour}/${tumour}.sample.dupmarked.bam

samtools depth \
-a \
-r ${chr}:${start}-${end} \
-q 25 \
-Q 30 \
-o ${tumour}.normal.${chr}.${start}.${end}.seg.txt \
/nfs/cancer_ref01/nst_links/live/2571/${normal}/${normal}.sample.dupmarked.bam

#create bed files
awk -F "\t" 'BEGIN {OFS = FS} {print $1, $2 -1, $2, $3}' ${tumour}.tumour.${chr}.${start}.${end}.seg.txt > ${tumour}.tumour.${chr}.${start}.${end}.seg.bed

awk -F "\t" 'BEGIN {OFS = FS} {print $1, $2 -1, $2, $3}' ${tumour}.normal.${chr}.${start}.${end}.seg.txt > ${tumour}.normal.${chr}.${start}.${end}.seg.bed

#calculate average per 500bp bin
bedtools map -a 500bp.window.bed -b ${tumour}.tumour.${chr}.${start}.${end}.seg.bed -c 4 -o mean > ${tumour}.tumour.whole.genome.bins_averaged.txt

infile="${tumour}.tumour.whole.genome.bins_averaged.txt"
awk -F "\t" 'BEGIN {OFS = FS} { if($5 != ".") { print $1, $2, $3, $4, $5} }' $infile > ${tumour}.tumour.${chr}.${start}.${end}.seg.bins_averaged.txt

bedtools map -a 500bp.window.bed -b ${tumour}.normal.${chr}.${start}.${end}.seg.bed -c 4 -o mean > ${tumour}.normal.whole.genome.bins_averaged.txt

infile="${tumour}.normal.whole.genome.bins_averaged.txt"
awk -F "\t" 'BEGIN {OFS = FS} { if($5 != ".") { print $1, $2, $3, $4, $5} }' $infile > ${tumour}.normal.${chr}.${start}.${end}.seg.bins_averaged.txt

```

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/localised_copy_number_calling")

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

tumour1_bins = read.table("PD51123o_lo0008.tumour.chr4.54229000.54298500.seg.bins_averaged.txt", header = F, sep = "\t", stringsAsFactors = F)
names(tumour1_bins) = c("chr", "start", "end", "winnum", "coverage")
tumour1_bins$plot_coord = apply(tumour1_bins[, 2:3], 1, mean)
tumour1_bins$sample = "PD51123o_lo0008"

tumour2_bins = read.table("PD51123o_lo0012.tumour.chr4.54229000.54298500.seg.bins_averaged.txt", header = F, sep = "\t", stringsAsFactors = F)
names(tumour2_bins) = c("chr", "start", "end", "winnum", "coverage")
tumour2_bins$plot_coord = apply(tumour2_bins[, 2:3], 1, mean)
tumour2_bins$sample = "PD51123o_lo0012"

normal_bins = read.table("PD51123o_lo0008.normal.chr4.54229000.54298500.seg.bins_averaged.txt", header = F, sep = "\t", stringsAsFactors = F)
names(normal_bins) = c("chr", "start", "end", "winnum", "coverage")
normal_bins$plot_coord = apply(normal_bins[, 2:3], 1, mean)
normal_bins$sample = "PD51123aa3"

all_cov = rbind(tumour1_bins, tumour2_bins, normal_bins)

all_cov$sample = factor(all_cov$sample, levels = c("PD51123o_lo0008", "PD51123o_lo0012", "PD51123aa3"))

#PDGFRA coordinates
min(elocs$start_position) #54229280
max(elocs$end_position) #54298245

#PDGFRA breakpoints
svs = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/structural_variants/paeds_autopsy_annotated_svs_uniq_svs_labelled_purity_annotated_20221206.txt", header = T, sep = "\t", stringsAsFactors = F)
svs[svs$gene == "PDGFRA" & svs$gene2 == "PDGFRA" & svs$sample %in% c("PD51123o_lo0008", "PD51123o_lo0012"), c(1:6, 10)]
#chro1   start1     end1      chro2   start2      end2          sample
#chr4    54253544   54253548  chr4    54268457    54268461      PD51123o_lo0008
#chr4    54267788   54267791  chr4    54272641    54272644      PD51123o_lo0012

#PD51123o_lo0008
within_del_cov = mean(all_cov[all_cov$sample == "PD51123o_lo0008" & all_cov$start >= 54253544 & all_cov$end <= 54268461,]$coverage) #31.06579
outside_del_cov = mean(all_cov[all_cov$sample == "PD51123o_lo0008" & ((all_cov$start >= 54229280 & all_cov$end <= 54253544) | (all_cov$start >= 54268461 & all_cov$end <= 54298245)),]$coverage) #408.9732
tumour_purity = manifest[manifest$sample == "PD51123o_lo0008",]$purity_trunk
tumour_cn = 18 #outside deletion
normal_cn = 2

tum_cov = outside_del_cov * (tumour_purity * tumour_cn / (tumour_purity * tumour_cn + (1 - tumour_purity) * normal_cn)) #407.4018
norm_cov = outside_del_cov - tum_cov #1.571355, should be consistent across the autosomal genome
tum_cov_per_cn = tum_cov / tumour_cn
tum_cov_within = within_del_cov - norm_cov
tum_cov_within / tum_cov_per_cn #1.303135

#PD51123o_lo0012
within_del_cov = mean(all_cov[all_cov$sample == "PD51123o_lo0012" & all_cov$start >= 54267788 & all_cov$end <= 54272644,]$coverage)  #65.77267
outside_del_cov = mean(all_cov[all_cov$sample == "PD51123o_lo0012" & ((all_cov$start >= 54229280 & all_cov$end <= 54267788) | (all_cov$start >= 54272644 & all_cov$end <= 54298245)),]$coverage) #732.1705
tumour_purity = manifest[manifest$sample == "PD51123o_lo0012",]$purity_trunk
tumour_cn = 33 #outside deletion
normal_cn = 2

tum_cov = outside_del_cov * (tumour_purity * tumour_cn / (tumour_purity * tumour_cn + (1 - tumour_purity) * normal_cn)) #724.5423
norm_cov = outside_del_cov - tum_cov #7.62821, should be consistent across the autosomal genome
tum_cov_per_cn = tum_cov / tumour_cn
tum_cov_within = within_del_cov - norm_cov
tum_cov_within / tum_cov_per_cn #2.648247

```

###Plot

Create localised copy number plots using calculations above

```{r}

#PDGFRA coordinates
min(elocs$start_position) #54229280
max(elocs$end_position) #54298245

#PDGFRA breakpoints
svs = read.table("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/structural_variants/paeds_autopsy_annotated_svs_uniq_svs_labelled_purity_annotated_20221206.txt", header = T, sep = "\t", stringsAsFactors = F)
svs[svs$gene == "PDGFRA" & svs$gene2 == "PDGFRA" & svs$sample %in% c("PD51123o_lo0008", "PD51123o_lo0012"), c(1:6, 10)]
#chro1   start1     end1      chro2   start2      end2          sample
#chr4    54253544   54253548  chr4    54268457    54268461      PD51123o_lo0008
#chr4    54267788   54267791  chr4    54272641    54272644      PD51123o_lo0012

#create df
est_cn_df = data.frame(sample = c("PD51123o_lo0008", "PD51123o_lo0008", "PD51123o_lo0008", "PD51123o_lo0012", "PD51123o_lo0012", "PD51123o_lo0012", "PD51123p", "PD51123aa3"), 
                       start = c(54228500, 54253544, 54268461, 54228500, 54267788, 54272644, 54228500, 54228500), 
                       end = c(54253543, 54268460, 54298501, 54267787, 54272643, 54298501, 54298501, 54298501), 
                       tot_cn = c(18, 1, 18, 33, 3, 33, 19, 2))
est_cn_df$sample = factor(est_cn_df$sample, levels = c("PD51123o_lo0008", "PD51123o_lo0012", "PD51123p", "PD51123aa3"))

p2 = ggplot(est_cn_df) +
  geom_segment(mapping = aes(x = start, xend = end, y = tot_cn, yend = tot_cn), size = 2) +
  facet_grid(sample ~ .) +
  theme_bw() +
  scale_x_continuous(expand = c(0,0), limits = c(54228500, 54298501)) +
  ylab("Total copy number") +
  xlab("Chromosome 4 position") +
  ylim(0, 40)

p1 = ggplot() +
  geom_line(mapping = aes(x = c(54229280, 54298245), y = c(0.5, 0.5)), size = 2) +
  geom_segment(elocs, mapping = aes(x = exon_chrom_start, xend = exon_chrom_end, y = 0.5, yend = 0.5), size = 10) +
  scale_x_continuous(expand = c(0,0), limits = c(54228500, 54298501)) +
  ylim(0.45, 0.55) +
  theme_void()

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/pdgra_deletions_20221211.pdf", width = 4, height = 4, useDingbats = F)
plot_grid(p1, p2, ncol = 1, rel_heights = c(0.05, 0.95), align = "v", axis = "lr")
dev.off()

write.table(est_cn_df, "/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/pdgra_deletions_data_20221222.txt", col.names = T, row.names = F, sep = "\t", quote = F)

```

##Fig. 4e, 4f & EDF 9-10 - effects of radiotherapy on tumour diversity

###EDF 9 - indel to sub ratio

```{r}

library(ggplot2)
library(ggbeeswarm)

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')

manifest_flt = manifest[manifest$indel_burden >= 10 & manifest$keep == "Y" & 
                          (manifest$average.reads.per.chromosome.copy.tumour >= 5 |
                             manifest$histo %in% c("Crypt", "Epithelium", "Epidermis")),] #no skin samples with enough indels

manifest_flt$annot = NA
manifest_flt[manifest_flt$average.reads.per.chromosome.copy.tumour >= 5 & manifest_flt$case.id == "PD50297",]$annot = "PD50297 tumour"
manifest_flt[manifest_flt$average.reads.per.chromosome.copy.tumour >= 5 & manifest_flt$case.id == "PD51122",]$annot = "PD51122 tumour"
manifest_flt[manifest_flt$average.reads.per.chromosome.copy.tumour >= 5 & manifest_flt$case.id == "PD51123",]$annot = "PD51123 tumour"
manifest_flt[manifest_flt$bulk_tissue %in% c("Appendix", "Large intestine"),]$annot = "Crypt(s), large intestine"
manifest_flt[manifest_flt$bulk_tissue %in% c("Small intestine"),]$annot = "Crypt(s), small intestine"

table(manifest_flt$annot)

manifest_flt$indel_sub_ratio = manifest_flt$indel_burden / manifest_flt$sub_burden

manifest_flt$annot = factor(manifest_flt$annot, levels = c("Crypt(s), small intestine", "Crypt(s), large intestine", "PD51122 tumour", "PD51123 tumour", "PD50297 tumour"))

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/indes2subs_20221222.pdf", width = 2.8, height = 4, useDingbats = F)
ggplot(manifest_flt) +
  geom_quasirandom(mapping = aes(x = annot, y = indel_sub_ratio), size = 2, col = "#999999") +
  geom_boxplot(mapping = aes(x = annot, y = indel_sub_ratio), fill = NA, outlier.shape = NA) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  xlab("") +
  ylab("Indels / substitutions") +
  ylim(0, 0.2)
dev.off()

```

###Fig. 4e & 4f & EDF 10

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/indels/04_bbinom")

indel.files = list.files(".", ".pindel.HP.9.read.direction.germline.beta.binom.unk.amb.min.depth.filtered.txt")

all_indels = data.frame()
for(file in indel.files){
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F, colClasses = c("character", "character", "character", "numeric", "character", "character", "numeric", "numeric", "numeric", "character", "character", "character", "character", "character", "character"))
  if(unique(data$Sample) != "PD50297o_lo0013"){
    print(unique(data$Sample))
    all_indels = rbind(all_indels, data)
  }
}

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
manifest$del2ins = NA

for(i in 1:nrow(manifest)){
  
  if(manifest$keep[i] == "Y" & manifest$sample[i] %in% unique(all_indels$Sample)){
    
    print(manifest$sample[i])
    sample_indels = all_indels[all_indels$Sample == manifest$sample[i],]
    manifest$del2ins[i] = nrow(sample_indels[sample_indels$Mutation == "Del",]) / nrow(sample_indels[sample_indels$Mutation == "Ins",])
    
  }
  
}

manifest_flt = manifest[manifest$indel_burden >= 10 & manifest$keep == "Y" & 
                          (manifest$average.reads.per.chromosome.copy.tumour >= 5 |
                             manifest$histo %in% c("Crypt", "Epithelium", "Epidermis")),] #no skin samples with enough indels

manifest_flt$annot = NA
manifest_flt[manifest_flt$average.reads.per.chromosome.copy.tumour >= 5 & manifest_flt$case.id == "PD50297",]$annot = "PD50297 tumour"
manifest_flt[manifest_flt$average.reads.per.chromosome.copy.tumour >= 5 & manifest_flt$case.id == "PD51122",]$annot = "PD51122 tumour"
manifest_flt[manifest_flt$average.reads.per.chromosome.copy.tumour >= 5 & manifest_flt$case.id == "PD51123",]$annot = "PD51123 tumour"
manifest_flt[manifest_flt$bulk_tissue %in% c("Appendix", "Large intestine"),]$annot = "Crypt(s), large intestine"
manifest_flt[manifest_flt$bulk_tissue %in% c("Small intestine"),]$annot = "Crypt(s), small intestine"

table(manifest_flt$annot)

manifest_flt$annot = factor(manifest_flt$annot, levels = c("Crypt(s), small intestine", "Crypt(s), large intestine", "PD51122 tumour", "PD51123 tumour", "PD50297 tumour"))

library(ggplot2)
library(ggbeeswarm)
library(cowplot)
library(ggpubr)

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/del2ins_20221220.pdf", width = 2.8, height = 4, useDingbats = F)
ggplot(manifest_flt) +
  geom_quasirandom(mapping = aes(x = annot, y = del2ins), size = 2, col = "#999999") +
  geom_boxplot(mapping = aes(x = annot, y = del2ins), fill = NA, outlier.shape = NA) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
        panel.grid = element_blank()) +
  xlab("") +
  ylab("Small deletions / insertions") +
  ylim(0, 10)
dev.off()

wilcox.test(manifest_flt[manifest_flt$annot == "PD50297 tumour",]$del2ins, 
            manifest_flt[manifest_flt$annot == "PD51122 tumour",]$del2ins,
            alternative = "two.sided") #W = 3558, p-value = 3.063e-14
wilcox.test(manifest_flt[manifest_flt$annot == "PD51123 tumour",]$del2ins, 
            manifest_flt[manifest_flt$annot == "PD51122 tumour",]$del2ins,
            alternative = "two.sided") #W = 2365, p-value = 2.606e-05

manifest_flt$sample = factor(manifest_flt$sample, levels = manifest_flt[order(manifest_flt$del2ins),]$sample)
median_per_bulk = aggregate(del2ins ~ bulk_sample, manifest_flt, median)
manifest_flt$bulk_sample = factor(manifest_flt$bulk_sample, levels = median_per_bulk[order(median_per_bulk$del2ins),]$bulk_sample)

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/del2ins_tumour_ratio_20221111.pdf", height = 5, width = 10, useDingbats = F)
ggplot(manifest_flt[manifest_flt$annot %in% c("PD50297 tumour", "PD51122 tumour", "PD51123 tumour"),]) +
  geom_quasirandom(mapping = aes(x = bulk_sample, y = del2ins, col = average.reads.per.chromosome.copy.tumour), size = 2) +
  geom_boxplot(mapping = aes(x = bulk_sample, y = del2ins), fill = NA, outlier.shape = NA) +
  scale_color_gradient(low = "#88D0ED", high = "#e83b3d", limits = c(5, 30)) +
  theme_bw() +
  ylim(0, 10) +
  xlab("") +
  ylab("Small deletions / insertions") +
  facet_grid(. ~ case.id, space = "free_x", scales = "free_x") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
        plot.margin = unit(c(0, 0, 0, 0), "cm"),
        strip.background = element_rect(fill = "black"),
        strip.text = element_text(colour = "white"))
dev.off()

#Measure clonal diversity
setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/subs/07_indel")

#manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
manifest_flt = manifest[manifest$indel_burden >= 10 & 
                           manifest$keep == "Y" &
                           manifest$average.reads.per.chromosome.copy.tumour >= 5 &
                           !is.na(manifest$average.reads.per.chromosome.copy.tumour),]
tumour_bulk_samples = manifest[manifest$keep == "Y" & manifest$purity_trunk >= 0.4 & manifest$experimental_arm == "WGS_bulk",]$sample
sum(!tumour_bulk_samples %in% manifest_flt$sample) #0

all_subs = data.frame()
for(sample in manifest_flt$sample){
  data = read.table(paste0(sample, ".standard.gnomAD.AF.0.001.PON.read.direction.low_cov.bq.mq.binom.bbinom.5bp.indel.filtered.tier.annot.txt"), header = T, sep = '\t', stringsAsFactors = F)
    print(unique(data$Sample))
    all_subs = rbind(all_subs, data)
}

pd50297_shared_muts = names(table(all_subs[all_subs$Sample %in% tumour_bulk_samples[grepl("PD50297", tumour_bulk_samples)],]$mutID))[table(all_subs[all_subs$Sample %in% tumour_bulk_samples[grepl("PD50297", tumour_bulk_samples)],]$mutID) == length(tumour_bulk_samples[grepl("PD50297", tumour_bulk_samples)])]
pd51122_shared_muts = names(table(all_subs[all_subs$Sample %in% tumour_bulk_samples[grepl("PD51122", tumour_bulk_samples)],]$mutID))[table(all_subs[all_subs$Sample %in% tumour_bulk_samples[grepl("PD51122", tumour_bulk_samples)],]$mutID) == length(tumour_bulk_samples[grepl("PD51122", tumour_bulk_samples)])]
pd51123_shared_muts = names(table(all_subs[all_subs$Sample %in% tumour_bulk_samples[grepl("PD51123", tumour_bulk_samples)],]$mutID))[table(all_subs[all_subs$Sample %in% tumour_bulk_samples[grepl("PD51123", tumour_bulk_samples)],]$mutID) == length(tumour_bulk_samples[grepl("PD51123", tumour_bulk_samples)])]

manifest_flt$frac_trunk = NA

for(i in 1:nrow(manifest_flt)){
  
  if(manifest_flt$case.id[i] == "PD50297") manifest_flt$frac_trunk[i] = nrow(all_subs[all_subs$Sample == manifest_flt$sample[i] & all_subs$mutID %in% pd50297_shared_muts,]) / nrow(all_subs[all_subs$Sample == manifest_flt$sample[i],])
  if(manifest_flt$case.id[i] == "PD51122") manifest_flt$frac_trunk[i] = nrow(all_subs[all_subs$Sample == manifest_flt$sample[i] & all_subs$mutID %in% pd51122_shared_muts,]) / nrow(all_subs[all_subs$Sample == manifest_flt$sample[i],])
  if(manifest_flt$case.id[i] == "PD51123") manifest_flt$frac_trunk[i] = nrow(all_subs[all_subs$Sample == manifest_flt$sample[i] & all_subs$mutID %in% pd51123_shared_muts,]) / nrow(all_subs[all_subs$Sample == manifest_flt$sample[i],])
  
}

pdf("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/draft_figures/subs_vs_del2ins_ratio_20221205.pdf", height = 3, width = 3, useDingbats = F)
ggplot() +
  geom_point(manifest_flt, mapping = aes(x = del2ins, y = frac_trunk, fill = case.id, col = case.id), pch = 21, cex = 100, stroke = NA, alpha = 0.9) +
  theme_bw()+
  scale_fill_manual(values = c("dodgerblue", "black", "#F9B233"), guide = "none")+
  scale_color_manual(values = alpha(c("dodgerblue", "black", "#F9B233"), 0), guide = "none") +
  stat_cor(manifest_flt[manifest_flt$case.id == "PD50297",], mapping = aes(x = del2ins, y = frac_trunk), method = "pearson", label.x = 6, label.y = 0.75, col = "dodgerblue", digits = 3) +
  stat_cor(manifest_flt[manifest_flt$case.id == "PD51122",], mapping = aes(x = del2ins, y = frac_trunk), method = "pearson", label.x = 6, label.y = 0.7, col = "black") +
  stat_cor(manifest_flt[manifest_flt$case.id == "PD51123",], mapping = aes(x = del2ins, y = frac_trunk), method = "pearson", label.x = 6, label.y = 0.65, col = "#F9B233", digits = 3) +
  ylim(0, 1) +
  ylab("Fraction of substitutions found in shared trunk") +
  xlab("Small deletions / insertions")
dev.off()

cor(manifest_flt[manifest_flt$case.id == "PD50297",]$frac_trunk, manifest_flt[manifest_flt$case.id == "PD50297",]$del2ins, method = "pearson") #-0.714767
cor(manifest_flt[manifest_flt$case.id == "PD51122",]$frac_trunk, manifest_flt[manifest_flt$case.id == "PD51122",]$del2ins, method = "pearson") #-0.3762564
cor(manifest_flt[manifest_flt$case.id == "PD51123",]$frac_trunk, manifest_flt[manifest_flt$case.id == "PD51123",]$del2ins, method = "pearson") #-0.7368382

```


###Radiotherapy and sub mutations

```{r}

all_subs = read.table("/lustre/scratch119/realdata/mdt1/team274/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/subs/07_indel/all_sample_subs_distribution_annotated_20221201.txt", header = T, sep = "\t", stringsAsFactors = F)

sub_groups <- aggregate(mutID ~ Case_ID + mut_status, all_subs[all_subs$pure_tumour_mut == T,], unique)
names(sub_groups$mutID) = apply(sub_groups[,1:2], 1, function(x) paste(x, collapse = "_"))

count_trinuc_context_38(sub_groups$mutID$PD50297_multi)
count_trinuc_context_38(sub_groups$mutID$PD50297_bulk)
count_trinuc_context_38(sub_groups$mutID$PD50297_sample)
count_trinuc_context_38(sub_groups$mutID$PD51122_multi)
count_trinuc_context_38(sub_groups$mutID$PD51122_bulk)
count_trinuc_context_38(sub_groups$mutID$PD51122_sample)
count_trinuc_context_38(sub_groups$mutID$PD51123_multi)
count_trinuc_context_38(sub_groups$mutID$PD51123_bulk)
count_trinuc_context_38(sub_groups$mutID$PD51123_sample)

```

###Indel mutational spectra by tumour per developmental stage

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/03_DNA_analyses/tumour_indel_clonality")

manifest = read.table('/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/00_supplementary_data/wgs_paeds_pm_sample_metadata_20221015.txt', header = T, sep = '\t', stringsAsFactors = F, quote = '')
bulk_tumour_samples = manifest[!is.na(manifest$battenberg_purity) & manifest$battenberg_purity >= 0.4, ]$sample


patients = c("PD50297", "PD51122", "PD51123")

indel_list = list()
for(patient in patients){
  
  patient_bulk = bulk_tumour_samples[grepl(patient, bulk_tumour_samples)]
  
  all_indels = data.frame()
  for(bulk in patient_bulk){
    data = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/paed_brain/autopsy/wgs_data/02_DNA_processing/variant_calls/indels/04_bbinom/", bulk, ".pindel.HP.9.read.direction.germline.beta.binom.unk.amb.min.depth.filtered.txt"), header = T, sep = "\t", stringsAsFactors = F)
    data$Wildtype = as.character(data$Wildtype)
    data$Mutant = as.character(data$Mutant)
    all_indels = rbind(all_indels, data)
    
  }
  
  all_indels$mutID = paste(all_indels$Chr, all_indels$Position, all_indels$Wildtype, all_indels$Mutant, sep = "_")
  
  root = names(table(all_indels$mutID))[table(all_indels$mutID) == length(patient_bulk)] #shared between all samples
  early = names(table(all_indels$mutID))[table(all_indels$mutID) > 1 & table(all_indels$mutID) < length(patient_bulk)]  #shared between subsets of samples
  late =names(table(all_indels$mutID))[table(all_indels$mutID) == 1] #private to a single sample
  
  indel_list[[paste0(patient, "_root")]] = root
  indel_list[[paste0(patient, "_early")]] = early
  indel_list[[paste0(patient, "_late")]] = late
    
}

library(MutationalPatterns)
library(GenomeInfoDb)
library("BSgenome.Hsapiens.UCSC.hg38")

element_lengths = unlist(lapply(indel_list, length))

indels = data.frame(matrix(unlist(lapply(indel_list, function(x) strsplit(x, "_"))), ncol = 4, byrow = T))
names(indels) = c("Chr", "Start", "Ref", "Alt")
indels$id = unlist(sapply(names(element_lengths), function(x) rep(x, element_lengths[x])))
indels_flt = indels[nchar(indels$Ref) != nchar(indels$Alt),] #remove indels where ref and alt are the same length, 950 in PD50297
indels_flt$End = as.numeric(indels_flt$Start) + (nchar(indels_flt$Ref) - 1)

grange_obj = makeGRangesListFromDataFrame(indels_flt, split.field = "id", keep.extra.columns = T, ignore.strand = T, seqnames.field = "Chr",
                                                start.field = "Start", end.field = "End")
GenomeInfoDb::genome(grange_obj) = "hg38"
seqlengths(grange_obj) <- seqlengths(BSgenome.Hsapiens.UCSC.hg38@seqinfo)[1:24]
indel_grl <- get_indel_context(grange_obj, "hg38")
indel_counts <- count_indel_contexts(indel_grl)
indel_counts <- indel_counts[, c("PD51122_root", "PD51122_early", "PD51122_late", "PD50297_root", "PD50297_early", "PD50297_late", "PD51123_root", "PD51123_early", "PD51123_late")]

pdf("indel_spectra_by_stage_20230219.pdf", height = 6, width = 10)
plot_indel_contexts(indel_counts, condensed = TRUE) + theme(strip.text.y = element_text(angle = 0))
dev.off()

```
