---
  title: "Placental mutational landscape"
author: "Thomas R W Oliver"
output: 
  html_document:
  toc: true
toc_depth: 2
number_sections: true
editor_options:
  chunk_output_type: console
---

#00_SUPPLEMENTARY_DATA

```{r}

#final list of included samples
manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)
old_manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml_draft_20220711/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)

manifest = read.table("/lustre/scratch126/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)
old_manifest = read.table("/lustre/scratch126/casm/team274sb/to3/placenta_ml_draft_20220711/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)

table(old_manifest$batch)
 #1   2   3 
 #24 134 258 

```

#01_QC

##Verifybamid

First, check that none of the included samples are contaminated, both the umbilical cord anor their matched umbilical cord or maternal sample.

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/verifybamid

#placental biopsies
infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt'
grep -v "Placenta_PD" $infile | cut -f1,3 > placenta_sample_list.txt

#umbilical cord
infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt'
grep -v "Placenta_PD" $infile | cut -f11,14 | sort | uniq >> placenta_sample_list.txt

#run
cut -f1,2 placenta_sample_list.txt | while read -r SAMPLE PROJECT ;  do echo "/lustre/scratch119/casm/team268im/fa8/VerifyBamId-farm5/VerifyBamID --Epsilon 1e-12 --UDPath /lustre/scratch119/casm/team294rr/rs30/git_repositories/VerifyBamID/resource/1000g.phase3.100k.b37.vcf.gz.dat.UD --BedPath /lustre/scratch119/casm/team294rr/rs30/git_repositories/VerifyBamID/resource/1000g.phase3.100k.b37.vcf.gz.dat.bed --MeanPath /lustre/scratch119/casm/team294rr/rs30/git_repositories/VerifyBamID/resource/1000g.phase3.100k.b37.vcf.gz.dat.mu --Reference /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/genome.fa --Output /lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/verifybamid/"$SAMPLE".verifybamid --BamFile /nfs/cancer_ref01/nst_links/live/$PROJECT/$SAMPLE/$SAMPLE.sample.dupmarked.bam " | bsub -J verifybamid-"$SAMPLE" -o verifybamid-"$SAMPLE".out -e verifybamid-"$SAMPLE".err -q long -n 4 -R 'select[mem>=16000] rusage[mem=16000] span[hosts=1]' -M16000 -env "all" ; done

#maternal samples
infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt'
grep -v "Placenta_PD" $infile | cut -f15,18 | grep -v "-" | sort | uniq > maternal_sample_list.txt

#run
cut -f1,2 maternal_sample_list.txt | while read -r SAMPLE PROJECT ;  do echo "/lustre/scratch119/casm/team268im/fa8/VerifyBamId-farm5/VerifyBamID --Epsilon 1e-12 --UDPath /lustre/scratch119/casm/team294rr/rs30/git_repositories/VerifyBamID/resource/1000g.phase3.100k.b37.vcf.gz.dat.UD --BedPath /lustre/scratch119/casm/team294rr/rs30/git_repositories/VerifyBamID/resource/1000g.phase3.100k.b37.vcf.gz.dat.bed --MeanPath /lustre/scratch119/casm/team294rr/rs30/git_repositories/VerifyBamID/resource/1000g.phase3.100k.b37.vcf.gz.dat.mu --nce /lustre/scratch119/casm/team78pipelines/nce/human/GRCh37d5/genome.fa --Output /lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/verifybamid/"$SAMPLE".verifybamid --BamFile /nfs/cancer_ref01/nst_links/live/$PROJECT/$SAMPLE/$SAMPLE.sample.dupmarked.bam " | bsub -J verifybamid-"$SAMPLE" -o verifybamid-"$SAMPLE".out -e verifybamid-"$SAMPLE".err -q normal -n 4 -R 'select[mem>=16000] rusage[mem=16000] span[hosts=1]' -M16000 -env "all" ; done

#infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt'
#for sample in $(cat placenta_sample_list.txt | cut -f1);
#do
#cp /lustre/scratch119/casm/team274sb/to3/placenta_ml_draft_20220711/01_QC/verifybamid/${sample}* .
#echo $sample
#done

#for sample in $(cat maternal_sample_list.txt | cut -f1);
#do
#cp /lustre/scratch119/casm/team274sb/to3/placenta_ml_draft_20220711/01_QC/verifybamid/${sample}* .
#echo $sample
#done

```

```{r}

setwd('/lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/verifybamid')

all_v_files = list.files(path = '.', pattern = '.verifybamid.selfSM')

v_df = c()
for(i in all_v_files){
  sample = unlist(strsplit(i, '.verifybamid.selfSM'))
  v_sample = read.table(i, header = T, comment.char = '', sep = '\t', stringsAsFactors = F)
  v_df = rbind(v_df, c(sample, v_sample$FREEMIX))
}

v_df = data.frame(v_df)
names(v_df) = c('sample', 'freemix')

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

sum(!manifest$Placenta_PD %in% v_df$sample) #0
sum(!manifest$Cord_PD %in% v_df$sample) #0
sum(!manifest$Maternal_PD %in% v_df$sample) #12, all where the maternal sample is missing

manifest$Biopsy_freemix = NA
manifest$Cord_freemix = NA
manifest$Maternal_freemix = NA

for(i in 1:nrow(manifest)){
  if(manifest$Placenta_PD[i] %in% v_df$sample) manifest$Biopsy_freemix[i] = v_df[v_df$sample == manifest$Placenta_PD[i],]$freemix
  if(manifest$Cord_PD[i] %in% v_df$sample) manifest$Cord_freemix[i] = v_df[v_df$sample == manifest$Cord_PD[i],]$freemix
  if(manifest$Maternal_PD[i] %in% v_df$sample) manifest$Maternal_freemix[i] = v_df[v_df$sample == manifest$Maternal_PD[i],]$freemix
}

write.table(manifest, '/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt', col.names = T, row.names = F, quote = F, sep = '\t')

```

```{r}

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

manifest$Biopsy_freemix = as.numeric(manifest$Biopsy_freemix)
manifest$Cord_freemix = as.numeric(manifest$Cord_freemix)
manifest$Maternal_freemix = as.numeric(manifest$Maternal_freemix)

omit_placenta = manifest[manifest$Biopsy_freemix > 0.01 & !is.na(manifest$Biopsy_freemix), c("Placenta_PD", "Biopsy_freemix")]$Placenta_PD
omit_cord = unique(manifest[manifest$Cord_freemix > 0.01 & !is.na(manifest$Cord_freemix), c("Cord_PD", "Cord_freemix")])$Cord_PD
omit_maternal = unique(manifest[manifest$Maternal_freemix > 0.01 & !is.na(manifest$Maternal_freemix), c("Maternal_PD", "Maternal_freemix")])$Maternal_PD

#Placenta_PD Biopsy_freemix
#PD45547b      0.0113674  placental biopsy, in original paper
#PD45556e      0.0125352  placental biopsy, in original paper
#PD45581e      0.0403400  placental biospy, in original paper with trisomic 10
#PD45559f2      0.4423430 membrane, new
#PD45585c      0.0171571  placental biopsy, new
#PD45590f      0.2955070  membrane, new
#PD51922e      0.0113040  placental biopsy, new

#Cord_PD Cord_freemix
#PD42165b    0.0156394  cord, in original paper too
#PD45551f    0.0133509  cord, in original paper too
#PD45565f    0.0212583  cord, in original paper
#PD45570f    0.0319331  cord, in original paper
#PD51904b    0.0135413  cord, new

#Maternal_PD Maternal_freemix
#PD45655b     0.0137554
#PD45673b     0.0235980

```

##Check for maternal contamination

###Copy over caveman files for maternal samples

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/maternal_snps

#copy over maternal caveman files
infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
awk -F "\t" '{ if($15 != "-" && $15 != "Maternal_PD") { print $15, $18} }' $infile | sort | uniq | while read -r -a tmp;
do
#echo ${tmp[0]} #PD ID
#echo ${tmp[1]} #canapps project
cp /nfs/cancer_ref01/nst_links/live/${tmp[1]}/${tmp[0]}/${tmp[0]}.caveman_c.annot.vcf.gz .
done

#Check that none of the vcfs have been truncated and are missing chromosomes
ls *.caveman_c.annot.vcf.gz | while read FILE ; do zgrep "^#" -v "$FILE" | cut -f1 | uniq -c > qc_chrom_counts/"$FILE".counts.txt ; done
wc -l qc_chrom_counts/*.counts.txt #none are truncated 27/4/21

```

###Extract maternal SNPs

```{bash}

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/maternal_snps

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
awk -F "\t" '{ if($15 != "-") { print $15} }' $infile | sort | uniq | while read -r -a tmp;
do
echo ${tmp[0]}
bsub -q small -o ${tmp[0]}.out -e ${tmp[0]}.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 Rscript --no-save --no-restore --verbose /lustre/scratch119/casm/team274sb/to3/placenta_ml/mat_contamination/mat_snp_filter.R -g female -v ${tmp[0]}.caveman_c.annot.vcf.gz
done

#bsub -q small -o PD42143b.out -e PD42143b.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 Rscript --no-save --no-restore --verbose /lustre/scratch119/casm/team274sb/to3/placenta_ml/mat_contamination/mat_snp_filter.R -g female -v PD42143b.caveman_c.annot.vcf.gz

```

```{r}

#mat_snp_filter.R
#Pipeline to extract the SNPs from maternal genome and write a loci for cgpvaf

#read in library
library(VariantAnnotation)
library(optparse)

option_list = list(
  make_option(c("-g", "--gender"), type="character", default=NULL, help="Gender of sample", metavar="character"),
  make_option(c("-v", "--vcf"), type="character", default=NULL, help="VCF file with mutation data", metavar="character"))

opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

gender = opt$gender
vcf_file = opt$vcf

exact.binomial=function(gender, data, cutoff=-5){
  # Modification to code written by tc16
  # Function to filter out germline variants based on unmatched
  # variant calls of multiple samples from same individual (aggregate coverage
  # ideally >150 or so, but will work with less). NV is matrix of reads supporting 
  # variant and NR the matrix with total depth (samples as columns, mutations rows, 
  # with rownames as chr_pos_ref_alt or equivalent). Returns a logical vector, 
  # TRUE if mutation is likely to be germline.
  
  XY_chromosomal = grepl("X|Y",rownames(data))
  autosomal = !XY_chromosomal
  
  if(gender=="female"){
    NV_vec = data$NV
    NR_vec = data$NR
    pval = rep(1,length(NV_vec))
    for (n in 1:length(NV_vec)){
      if(NR_vec[n]>0){
        pval[n] = binom.test(x=NV_vec[n],
                             n=NR_vec[n],
                             p=0.5,alt='less')$p.value
      }
      if (n%%1000==0){
        print(n)
      }
    }
  }
  # For male, split test in autosomal and XY chromosomal part
  if(gender=="male"){
    pval=rep(1,nrow(data$NV))
    NV_vec = data$NV[autosomal]
    NR_vec = data$NR[autosomal]
    pval_auto = rep(1,sum(autosomal))
    pval_XY = rep(1,sum(XY_chromosomal))
    
    for (n in 1:sum(autosomal)){
      if(NR_vec[n]>0){
        pval_auto[n] = binom.test(x=NV_vec[n],
                                  n=NR_vec[n],
                                  p=0.5,alt='less')$p.value
      }
      if (n%%1000==0){
        print(n)
      }
    }
    NV_vec = data$NV[XY_chromosomal]
    NR_vec = data$NR[XY_chromosomal]
    for (n in 1:sum(XY_chromosomal)){
      if(NR_vec[n]>0){
        pval_XY[n] = binom.test(x=NV_vec[n],
                                n=NR_vec[n],
                                p=0.95,alt='less')$p.value
      }
      if (n%%1000==0){
        print(n)
      }
    }
    pval[autosomal]=pval_auto
    pval[XY_chromosomal]=pval_XY
  }
  qval = p.adjust(pval,method="BH")
  germline = log10(qval)>cutoff
  return(germline)
}

#gender = 'female'
#vcf_file = 'PD42143b.caveman_c.annot.vcf.gz'

vcf = readVcf(vcf_file)
sample = unlist(strsplit(vcf_file, '.caveman_c.annot.vcf.gz'))
reads = cbind((geno(vcf)$FAZ)[,2]+(geno(vcf)$RAZ)[,2],(geno(vcf)$FCZ)[,2]+(geno(vcf)$RCZ)[,2],
                (geno(vcf)$FGZ)[,2]+(geno(vcf)$RGZ)[,2],(geno(vcf)$FTZ)[,2]+(geno(vcf)$RTZ)[,2])
var.df = data.frame(chr = as.character(seqnames(rowRanges(vcf))),
                    pos = start(ranges(rowRanges(vcf))),
                    ref = as.character(ref(vcf)),
                    alt = unlist(CharacterList(alt(vcf))))
var.df$NR=rowSums(reads)
var.df$NV=NA
colnames(reads)=c("A","C","G","T")
for (k in c("A","C","G","T")){
  var.df$NV[var.df$alt==k] = reads[var.df$alt==k,k]
}
var.df$VAF = var.df$NV/ var.df$NR
var.df$ASMD = info(vcf)$ASMD
var.df$CLPM = info(vcf)$CLPM
var.df$FILTER = rowRanges(vcf)$FILTER
var.df.filtered = var.df[(var.df$ASMD >= 140) & (var.df$CLPM == 0) & (var.df$FILTER == 'PASS') & (var.df$NR >= 10) & (var.df$NR <= 100),]
row.names(var.df.filtered) = paste(var.df.filtered$chr, var.df.filtered$pos, var.df.filtered$ref, var.df.filtered$alt, sep = '_')
germline = exact.binomial(gender = 'female', data = var.df.filtered, cutoff = -5)
var.df.filtered.germline = var.df.filtered[germline,]
write.table(var.df.filtered.germline[,1:4], paste0(sample, '.pass.filtered.germline.txt'), row.names = F, col.names = F, sep = '\t', quote = F)

```

###Symlink bam files to directory

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/maternal_cgpvaf

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml_draft_20220711/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
#infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
awk -F "\t" '{ if($15 != "-" && $15 != "Maternal_PD") { print $1, $3} }' $infile > all_placenta_samples.txt #placental biopsies from placentas with maternal DNA available
awk -F "\t" '{ if($15 != "-" && $15 != "Maternal_PD") { print $11, $14} }' $infile | sort | uniq >> all_placenta_samples.txt #cord DNA from placentas with maternal DNA available
awk -F "\t" '{ if($15 != "-" && $15 != "Maternal_PD") { print $15, $18} }' $infile | sort | uniq >> all_placenta_samples.txt #maternal DNA available

wc -l all_placenta_samples.txt #626

while read line;
do 
set $line
#echo $1
#echo $2
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam.bai
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam.bas
done < all_placenta_samples.txt

ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam
ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam.bai
ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam.bas

```

###Generate list of samples per maternal sample

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/maternal_cgpvaf

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'

for cord_sample in $(awk -F "\t" '{ if($15 != "-" && $15 != "Maternal_PD") { print $11} }' $infile | uniq);
do

patient1=${cord_sample:0:7}
echo ${cord_sample} > ${patient1}_samples.txt

maternal_sample=$(grep ${cord_sample} $infile | awk -F "\t" '{ print $13 }' | uniq)
echo ${maternal_sample} >> ${patient1}_samples.txt

  for placenta_sample in $(grep ${cord_sample} $infile | awk -F "\t" '{ print $1 }');
  do
    echo ${placenta_sample} >> ${patient1}_samples.txt
  done

  patient2=${placenta_sample:0:7}
  mv ${patient1}_samples.txt ${patient2}_samples.txt

done

```

###Run cgpvaf

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/maternal_cgpvaf

module load cgpVAFcommand

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/Placenta_bulk_landscape_project_metadata_20211112.txt'

for f in $(ls PD*_samples.txt);
do

patient=${f%_samples.txt}
#echo $patient

declare -a sampleArray=()

while read line;
do
  set $line
  sampleArray+=($1)
done < ${patient}_samples.txt
 
#echo "${sampleArray[@]}"   

cord_sample=$(head -1 ${patient}_samples.txt)
maternal_sample=$(grep ${cord_sample} $infile | cut -f13 | uniq)

#echo $cord_sample
#echo $maternal_sample

mkdir -p ${patient}
  
bsub -q normal -o ${patient}/%J.${patient}.out -e ${patient}/%J.${patient}.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 perl /software/CASM/modules/installs/vafcorrect/shim-bin/cgpVaf.pl -d ./ -o ${patient} -a snp -mq 30 -bq 25 -g /lustre/scratch119/casm/team78pipelines/nce/human/GRCh37d5/genome.fa -be .sample.dupmarked.bam -hdr /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/maternal_snps/${maternal_sample}.pass.filtered.germline.txt -nn PDv37is -tn ${sampleArray[@]}

done

#bsub -q normal -o PD42145/%J.PD42145.out -e PD42145/%J.PD42145.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 perl /software/CASM/modules/installs/vafcorrect/shim-bin/cgpVaf.pl -d ./ -o PD42145 -a snp -mq 30 -bq 25 -g /lustre/scratch119/casm/team78pipelines/nce/human/GRCh37d5/genome.fa -be .sample.dupmarked.bam -hdr /lustre/scratch119/casm/team78pipelines/canpipe/live/ref/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/maternal_snps/PD42143b.pass.filtered.germline.txt -nn PDv37is -tn PD42145b PD42143b PD42145c PD42145d PD42145e PD42145f

```

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/maternal_cgpvaf

ls PD*/*tmp*
ls PD*/*_snp_vaf.tsv | wc -l # 111/111!

cp PD*/*_snp_vaf.tsv /lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/mat_contamination/tc16_filters

```

###tc16 filtering

```{bash}

cd  /lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/mat_contamination/tc16_filters

#remove header info
for f in $(ls *_snp_vaf.tsv);
do
filename=$(basename -- "$f")
filename="${filename%_snp_vaf.tsv}"
#echo $filename
grep -v "##" $f > ${filename}_snp_vaf_nohash.tsv
done

```

####Run maternal SNP filtering and generate basic plots

```{bash}

export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/mat_contamination/tc16_filters

for file in $(ls *_snp_vaf_nohash.tsv);
do
stem=${file%_snp_vaf_nohash.tsv}
#echo $stem
bsub -q normal -o $stem.out -e $stem.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 /software/R-3.6.1/bin/Rscript ~/scripts/placenta_maternal_filtering_20220506.R -f $file
done

```

```{r}

#~/scripts/placenta_maternal_filtering_20220506.R

#additional functions
#-------------------------------------------------
# Filtering germline substitutions using an 
# exact binomial test
# Tim Coorens - January 2019
#-------------------------------------------------

options(stringsAsFactors = F)

#-------------------------------------------------
# Functions
#-------------------------------------------------

#1. Exact binomial test - modified by to3 to remove sex parameter as we will apply it to autosomal variants only 29/4/21

exact.binomial=function(NV, NR, cutoff=-5){
  # Function to filter out germline variants based on unmatched
  # variant calls of multiple samples from same individual (aggregate coverage
  # ideally >150 or so, but will work with less). NV is matrix of reads supporting 
  # variants and NR the matrix with total depth (samples as columns, mutations rows, 
  # with rownames as chr_pos_ref_alt or equivalent). Returns a logical vector, 
  # TRUE if mutation is likely to be germline.
  
    NV_vec = rowSums(NV)
    NR_vec = rowSums(NR)
    pval = rep(1,length(NV_vec))
    for (n in 1:length(NV_vec)){
      if(NR_vec[n]>0){
        pval[n] = binom.test(x=NV_vec[n],
                             n=NR_vec[n],
                             p=0.5,alt='less')$p.value
      }
      if (n%%1000==0){
        print(n)
      }
    }
  qval = p.adjust(pval,method="BH")
  germline = log10(qval)>cutoff
  return(germline)
}


#-------------------------------------------------
# Beta-binomial overdispersion - filtering out artefacts
# Tim Coorens - April 2018
#-------------------------------------------------
require(VGAM)
estimateRho_gridml = function(NV_vec,NR_vec) {
  #Make sure depth is non-zero
  NV_vec=NV_vec[NR_vec>0]
  NR_vec=NR_vec[NR_vec>0]
  
  # Function to estimate maximum likelihood value of rho for beta-binomial
  rhovec = 10^seq(-6,-0.05,by=0.05) # rho will be bounded within 1e-6 and 0.89
  mu=sum(NV_vec)/sum(NR_vec)
  ll = sapply(rhovec, function(rhoj) sum(dbetabinom(x=NV_vec, size=NR_vec, rho=rhoj, prob=mu, log=T)))
  return(rhovec[ll==max(ll)][1])
}

beta.binom.filter = function(NR,NV,cutoff=0.1, binom.pval=F,pval.cutoff=0.05){
  # Function to apply beta-binomial filter for artefacts. Works best on sets of
  # clonal samples (ideally >10 or so). As before, takes NV and NR as input. 
  # Optionally calculates pvalue of likelihood beta-binomial with estimated rho
  # fits better than binomial. This was supposed to protect against low-depth variants,
  # but use with caution. Returns logical vector with good variants = TRUE
  
  rho_est = pval = rep(NA,nrow(NR))
  for (k in 1:nrow(NR)){
    rho_est[k]=estimateRho_gridml(NV_vec = as.numeric(NV[k,]),
                                  NR_vec = as.numeric(NR[k,]))
    if (k%%1000==0){
      print(k)
    }
  }
  flt_rho=rho_est>=cutoff
  return(rho_est)
}

#run
library(ggplot2)
library(optparse)

option_list = list(
  make_option(c("-f", "--text_file"), type="character", default=NULL, help="Text file containing cgpVAF output", metavar="character"))

opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

i = opt$text_file
#i = "PDv37is_PD51926b_snp_vaf_nohash.tsv"

#read in sample, coerce to correct format
#remove sex chromosomes to avoid comparisons between male fetus and maternal blood being calculated separately - there should still be plenty of SNPs to work with.
#setwd('/lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/mat_contamination/tc16_filters')
#file.list = list.files('.', pattern = '_snp_vaf_nohash.tsv') #read in matrix containing variant depth (NV) and total depth (NR)
#file.list = file.list[grepl("PD5302", file.list)]

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt", header = T, sep = '\t', stringsAsFactors = F)

#for(i in file.list){
data <- read.table(i, comment.char="",header = T)
matched_normal = unlist(strsplit(unlist(strsplit(i, 'PDv37is_'))[2], '_snp_vaf_nohash.tsv'))
Muts = paste(data$Chrom, data$Pos, data$Ref, data$Alt, sep="_")
Genotype = data[, grepl("VAF",colnames(data))&colnames(data) != "PDv37is_VAF"]
NR = data[, grepl("DEP",colnames(data))&colnames(data) != "PDv37is_DEP"]
NV = data[, grepl("MTR",colnames(data))&colnames(data) != "PDv37is_MTR"]
colnames(Genotype) = colnames(NR) = colnames(NV) = gsub("_VAF", "", colnames(Genotype))
rownames(Genotype) = rownames(NV) = rownames(NR) = Muts
  
patient = substr(colnames(Genotype)[2], 0, 7)

#remove sex chromosome from equation
Muts = Muts[!grepl("X|Y",Muts)]
Genotype = Genotype[row.names(Genotype) %in% Muts,]
NR = NR[row.names(NR) %in% Muts,]
NV = NV[row.names(NV) %in% Muts,]

#remove variants that are uniformly low or high coverage across all samples as they are likely to be mapping artefacts
Depth_filter = (rowMeans(NR) > 10 & rowMeans(NR) < 100)
NR = NR[Depth_filter, ]
NV = NV[Depth_filter, ]

germline = exact.binomial(NV = NV, NR = NR, cutoff = -5) #determine which variants are germline

write.table(row.names(NR)[germline], paste0(patient,"_snp_variants_shared_with_mat.txt"), row.names = F, col.names = F, quote = F)
write.table(row.names(NR)[!germline], paste0(patient,"_snp_variants_not_shared_with_mat.txt"), row.names = F, col.names = F, quote =F)

NR_flt=NR[!germline, ]
NV_flt=NV[!germline, ]
  
NR_flt[NR_flt==0] = 1 #turn sites with no coverage in a sample to a depth of 1 read to prevent NA values downstream
  
rho_est = beta.binom.filter(NR = NR_flt,NV = NV_flt) #remove variants seen at similar VAF across every sample
flt_rho = log10(rho_est) < (-1)
rho_filtered_out = rownames(NR_flt)[flt_rho]

write.table(rho_filtered_out, paste0(patient, "_bbinom_filtered_out_snp.txt"))
write.table(rho_est, paste0(patient, "_rho_est_snp.txt"))

NR_flt_2 = NR_flt[!rownames(NR_flt) %in% rho_filtered_out,]
NV_flt_2 = NV_flt[!rownames(NV_flt) %in% rho_filtered_out,]
  
VAF = NV_flt_2 / NR_flt_2
VAF$mut = row.names(VAF)
Vaf.m = reshape2::melt(VAF, id.vars = 'mut')
Vaf.m$variable = as.character(Vaf.m$variable)
Vaf.m$source = ifelse(Vaf.m$variable %in% manifest$Maternal_PD, 'mother', "placenta")
  
p = ggplot(Vaf.m) + geom_boxplot(mapping = aes(x = variable, y = value, fill = source)) + theme_bw() + xlab('') + ylab('VAF') + labs(title = patient) + theme(plot.margin = unit(c(1,1,1,1), "cm"), plot.title = element_text(face = "bold", hjust = 0.5, size = 18), legend.position = 'none', axis.text.x = element_text(angle = 90, vjust = 0.5)) + scale_fill_manual(values = c("red", "grey"))
ggsave(paste0(patient, '_unique_maternal_SNPs_plot.pdf'), p, device = 'pdf')

write.table(NR_flt_2, paste0(patient,"_NR_filtered_all_snp.txt"))
write.table(NV_flt_2, paste0(patient,"_NV_filtered_all_snp.txt"))

#print(patient)
#}

```

####Final maternal contamination estimates and composite plot

```{r}

setwd('/lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/mat_contamination/tc16_filters')

all_nr_files = list.files(path = '.', pattern = '_NR_filtered_all_snp.txt')
all_nv_files = list.files(path = '.', pattern = '_NV_filtered_all_snp.txt')

est_mat = c()
for(i in all_nr_files){
  sample = unlist(strsplit(i, '_NR_filtered_all_snp.txt'))
  NR = read.table(i, header = T, sep = ' ', stringsAsFactors = F)
  NR_nonzero = NR
  NR_nonzero[NR_nonzero == 0] = 1
  NV = read.table(paste0(sample, '_NV_filtered_all_snp.txt'), header = T, sep = ' ', stringsAsFactors = F)
  VAF = NV / NR_nonzero
  est_mat = c(est_mat, 2 * apply(VAF, 2, median)) #double to get cellular %, not VAF
}

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest$Biopsy_est_mat_contamination = NA
manifest$Cord_est_mat_contamination = NA

for(i in 1:nrow(manifest)){
  manifest$Biopsy_est_mat_contamination[i] = est_mat[manifest$Placenta_PD[i]]
  manifest$Cord_est_mat_contamination[i] = est_mat[manifest$Cord_PD[i]]
}

write.table(manifest, '/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt', col.names = T, row.names = F, quote = F, sep = '\t')

#compare freemix and maternal contamination estimates
contamination.df = data.frame(sample = c(manifest$Placenta_PD, manifest[!duplicated(manifest$Cord_PD),]$Cord_PD), freemix = c(manifest$Biopsy_freemix, manifest[!duplicated(manifest$Cord_PD),]$Cord_freemix), maternal_contamination_est = c(manifest$Biopsy_est_mat_contamination, manifest[!duplicated(manifest$Cord_PD),]$Cord_est_mat_contamination))

library(ggplot2)

ggplot(contamination.df, mapping = aes(x = freemix, y = maternal_contamination_est, label = sample)) + geom_point() + theme_linedraw() + geom_text(aes(label=ifelse( maternal_contamination_est > 0.01,as.character(sample),'')),hjust=0.5,vjust=0) + geom_abline(slope = 1, intercept = 0, col = "red") + geom_smooth(method='lm', formula= y~x) + ylab("Estimated maternal contamination") + xlab("Estimated total contamination")

contamination.df[contamination.df$maternal_contamination_est == 0 & contamination.df$freemix >= 0.01,] #five samples have contamination just above 1% that cannot be accounted for by maternal contamination
#sample   freemix maternal_contamination_est
#PD45547b 0.0113674                          0
#PD45556e 0.0125352                          0
#PD51922e 0.0113040                          0
#PD45551f 0.0133509                          0
#PD51904b 0.0135413                          0

```

##Check insert sizes

###Symlink bam files to directory

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/insert_size_metrics

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml_draft_20220711/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
#infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
grep -v Placenta_PD $infile | cut -f1,3 > all_placenta_samples.txt
grep -v Placenta_PD $infile | cut -f11,14 | sort | uniq >> all_placenta_samples.txt

#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"
#awk -F "\t" '{ if($23 == "Yes") { print $1, $3} }' $infile > all_placenta_samples.txt
#awk -F "\t" '{ if($23 == "Yes") { print $9, $12} }' $infile | sort | uniq >> all_placenta_samples.txt

wc -l all_placenta_samples.txt #all 530

while read line;
do 
set $line
#echo $1
#echo $2
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam.bai
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam.bas
done < all_placenta_samples.txt

```

###Run picard tools

```{bash}

#picard_insert_size_metrics.sh
BAMNAME=$1
SAMPLE=$2

java -jar /lustre/scratch119/casm/team274sb/to3/paed_brain/software/picard.jar CollectInsertSizeMetrics \
      I=$BAMNAME \
      O=${SAMPLE}_insert_size_metrics.txt \
      H=${SAMPLE}_insert_size_histogram.pdf \
      M=0.5

```


```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/insert_size_metrics

mkdir logs

module load jre/1.8.0_131
#insert size metrics has a R dependency apparently...
export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"

MEM=20000
CORE=1

#bam="PD51927f.sample.dupmarked.bam"
#sample="PD51927f"
for bam in $(ls *.sample.dupmarked.bam);
do
#echo $bam
sample=${bam%.sample.dupmarked.bam}
#echo $sample
bsub -J"$sample" -M"$MEM" -R"select[mem>$MEM] rusage[mem=$MEM] span[hosts=1]" -n $CORE -q normal \
	-e logs/stderr.$sample.%J -o logs/stdout.$sample.%J \
	"./picard_insert_size_metrics.sh $bam $sample"
done


#PD51891b hit time limit
bam="PD51891b.sample.dupmarked.bam"
sample="PD51891b"

bsub -J"$sample" -M"$MEM" -R"select[mem>$MEM] rusage[mem=$MEM] span[hosts=1]" -n $CORE -q long \
	-e logs/stderr.$sample.%J -o logs/stdout.$sample.%J \
	"./picard_insert_size_metrics.sh $bam $sample"

```

```{bash}

#file="PD51927e_insert_size_metrics.txt"
#sample="PD51927e"
for file in $(ls *_insert_size_metrics.txt);
do
sample=${file%_insert_size_metrics.txt}
#echo $sample
head -n 8 $file | tail -n 2 > ${sample}.insert.size.summary.txt
done

```

###Append to manifest

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/insert_size_metrics")

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest$bx_mean_insert_size = NA

#bx = "PD51924c"
for(bx in manifest$Placenta_PD){
  data = read.table(paste0(bx, ".insert.size.summary.txt"), header = T, sep = '\t', stringsAsFactors = F)
  manifest[manifest$Placenta_PD == bx,]$bx_mean_insert_size = data$MEAN_INSERT_SIZE
}

library(ggplot2)
library(ggpubr)

manifest$batch = NA

manifest[grepl("PD421", manifest$Placenta_PD) & !grepl("PD42145", manifest$Placenta_PD),]$batch = 1
manifest[grepl("PD455", manifest$Placenta_PD) & manifest$Placenta_canapps_project == 2058,]$batch = 2
manifest[grepl("PD51", manifest$Placenta_PD) | manifest$Placenta_canapps_project == 2633,]$batch = 3

manifest$cord_mean_insert_size = NA

for(bx in manifest$Cord_PD){
  data = read.table(paste0(bx, ".insert.size.summary.txt"), header = T, sep = '\t', stringsAsFactors = F)
  manifest[manifest$Cord_PD == bx,]$cord_mean_insert_size = data$MEAN_INSERT_SIZE
}

write.table(manifest, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", col.names = T, row.names = F, sep = '\t', quote = F)

```

##Mosdepth

###Run for whole genome

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/mosdepth

#placental biopsies only
infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt'
grep -v "Placenta_PD" $infile | cut -f1,3 > placenta_sample_list.txt

while read line;
do
set $line
echo /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam >> placenta_bam_list.txt
done < placenta_sample_list.txt

while read line;
do
set $line
sampleID=$(basename $1 | sed "s/[.].*//")
#echo $1
#echo $sampleID
/nfs/users/nfs_t/to3/scripts/run_mosdepth.sh $sampleID $1 "0:1:4:100:" /lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/mosdepth ; 
done < placenta_bam_list.txt

module load bedtools

ls *.quantized.quantized.bed.gz | while read FILE ; do zgrep -P "\tCALLABLE|HIGH_COVERAGE$" "$FILE" | bedtools intersect -a - -b /lustre/scratch119/casm/team294rr/rs30/External_Databases/References/GRCh37d5/chrom.sizes.bed -u | bedtools subtract -a - -b /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/caveman/flagging/simple_repeats.bed | bedtools subtract -a - -b /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/caveman/flagging/hi_seq_depth.bed | bedtools subtract -a - -b /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/caveman/flagging/centromeric_repeats.bed | bedtools subtract -a - -b /lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/LCR-hs37d5.bed | bedtools subtract -a - -b /lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/hg19-blacklist.v2.chr.renamed.bed > ${FILE%.bed.gz}.callable.calledContigs.subtracted_ignored_regions.bed ; done

ls *.quantized.quantized.callable.calledContigs.subtracted_ignored_regions.bed | while read FILE ; do perl /lustre/scratch119/casm/team294rr/rs30/repositories/Baylor/scripts/BED_stats/get_BED-stats.v2.pl "$FILE" ; done

```

####Run mosdepth for autosomal genome

Used as a standardised way across sexes to estimate mutations per Mb

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/01_QC/mosdepth

for f in $(ls *.quantized.quantized.callable.calledContigs.subtracted_ignored_regions.bed);
do
filename=$(basename -- "$f")
filename="${filename%.quantized.quantized.callable.calledContigs.subtracted_ignored_regions.bed}"
echo $filename
awk -F "\t" '{ if(($1 != "X") && ($1 != "Y")) { print } }' ${filename}.quantized.quantized.callable.calledContigs.subtracted_ignored_regions.bed > ${filename}.quantized.quantized.callable.calledContigs.subtracted_ignored_regions.autosomalchr_only.bed
done

ls *.quantized.quantized.callable.calledContigs.subtracted_ignored_regions.autosomalchr_only.bed | while read FILE ; do perl /lustre/scratch119/casm/team294rr/rs30/repositories/Baylor/scripts/BED_stats/get_BED-stats.v2.pl "$FILE" ; done

```

##QC plots

```{r}

old_manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml_draft_20220711/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)

mean(old_manifest[old_manifest$batch == 3,]$bx_mean_insert_size) #392.781
mean(old_manifest[old_manifest$batch %in% c(1, 2),]$bx_mean_insert_size) #514.7338

library(ggplot2)
library(cowplot)

pdf("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/figures/contamination_comparison_thesis_20230215.pdf", width = 4, height = 4)
ggplot(old_manifest) +
  geom_point(mapping = aes(x = Biopsy_freemix, y = Biopsy_est_mat_contamination)) +
  theme_bw() +
  xlim(0, 0.5) +
  ylim(0, 0.5) +
  xlab("Fraction contamination (verifyBamID2)") +
  ylab("Fraction contamination (maternal)")
dev.off()

cor(old_manifest[!is.na(old_manifest$Biopsy_est_mat_contamination),]$Biopsy_freemix, y = old_manifest[!is.na(old_manifest$Biopsy_est_mat_contamination),]$Biopsy_est_mat_contamination, method = "pearson") #0.9955691

ggplot(old_manifest) +
  geom_histogram(mapping = aes(x = sub_burden), bins = 100) +
  theme_pubr() +
  coord_cartesian(expand = F) +
  xlab("Substitution burden")

sum(old_manifest$sub_burden > 500) #5
sum(old_manifest$sub_burden > 1000) #1

old_manifest$batch2 = old_manifest$batch
old_manifest[old_manifest$batch2 == 2,]$batch2 = 1
old_manifest[old_manifest$batch2 == 1,]$batch2 = "Original"
old_manifest[old_manifest$batch2 == 3,]$batch2 = "Additional"

#burden vs VAF
p1 = ggplot(old_manifest) +
  geom_point(mapping = aes(y = sub_burden, x = sub_median_VAF, col = batch2), show.legend = FALSE) +
  theme_bw() +
  xlab("Median substitution VAF") +
  ylab("Substitution burden") +
  xlim(0, 0.5) +
  scale_color_manual(values = c(Additional = "grey60", Original = "black")) +
  theme()

#burden vs Coverage
p2 = ggplot(old_manifest) +
  geom_point(mapping = aes(y = sub_burden, x = biopsy_coverage, col = batch2), show.legend = FALSE) +
  theme_bw() +
  xlab("Sample coverage") +
  ylab("Substitution burden") +
  scale_color_manual(values = c(Additional = "grey60", Original = "black"))

#burden vs contamination estimate
p3 = ggplot(old_manifest) +
  geom_point(mapping = aes(y = sub_burden, x = Biopsy_freemix, col = batch2), show.legend = FALSE) +
  theme_bw() +
  xlab("Fraction contamination (verifyBamID2)") +
  ylab("Substitution burden") +
  scale_color_manual(values = c(Additional = "grey60", Original = "black")) +
  xlim(0, 0.5)

#insert size vs burden
p4 = ggplot(old_manifest) +
  geom_point(mapping = aes(x = bx_mean_insert_size, y = sub_burden, col = batch2), show.legend = FALSE) +
  theme_bw() +
  xlab("Mean insert size") +
  ylab("Substitution burden") +
  geom_vline(xintercept = 320, col = "red",lty = 2) +
  scale_color_manual(values = c(Additional = "grey60", Original = "black"))

pdf("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/figures/figure2.3_20230215.pdf", width = 8, height = 8)
plot_grid(p1, p2, p3, p4, ncol = 2)
dev.off()

```


#02_VARIANT_FILTERING

##Subs - matched biopsies

Putative variants called by CaVEMan.

###PASS, ASMD & CLPM filtering

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/01_standard

module load vcftools

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
grep -v "Placenta_PD" $infile | awk -F "\t" '{ { print $1, $3} }' | while read -r -a tmp;
do
#echo ${tmp[0]} #PD ID
#echo ${tmp[1]} #canapps project
bcftools filter -i 'FILTER == "PASS" & INFO/ASMD >= 140 & INFO/CLPM == 0' /nfs/cancer_ref01/nst_links/live/${tmp[1]}/${tmp[0]}/${tmp[0]}.caveman_c.annot.vcf.gz > ${tmp[0]}.caveman_c.annot.asmd.clpm.pass.filtered.vcf
done

#PD51879c, PD51926c and PD51926e annot files copied over later, 27/5/22

#check all run matched
for f in $(ls *.caveman_c.annot.asmd.clpm.pass.filtered.vcf);
do
less $f | grep -m1 "TUMOUR" >> matched_sample_list.txt
less $f | grep -m1 "NORMAL" >> matched_sample_list.txt
done

grep "PDv37is" matched_sample_list.txt | wc -l #0, no run unmatched, 27/5/22
ls *.caveman_c.annot.asmd.clpm.pass.filtered.vcf | wc -l #416 correct, 27/5/22

#Check that none of the vcfs have been truncated and are missing chromosomes
ls *.caveman_c.annot.asmd.clpm.pass.filtered.vcf | while read FILE ; do grep "^#" -v "$FILE" | cut -f1 | uniq -c > ../qc_chrom_counts/"$FILE".counts.txt ; done 
wc -l ../qc_chrom_counts/*.counts.txt #lmost all >20 27/5/22 but heavy filtering applied to low burden samples so missing calls on some chromosomes expected

```

###Maternal SNP filtering

Remove any substitution called in mother to eliminate any maternal contamination, where maternal contamination data is available

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/01_standard

module load bedtools

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
for f in $(ls PD*.caveman_c.annot.asmd.clpm.pass.filtered.vcf);
do
sample=${f%.caveman_c.annot.asmd.clpm.pass.filtered.vcf}
maternal=$(grep -w $sample $infile | cut -f15 | uniq)
if [[ "$maternal" != "-" ]]; 
then 
bedtools intersect -header -v -a $f -b /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/maternal_snps/${maternal}.caveman_c.annot.vcf.gz > ../02_maternal/${sample}.caveman_c.annot.asmd.clpm.pass.maternal.filtered.vcf
fi
done

#no maternal sample for the following:
#PD51883b
#PD51883c
#PD51883d
#PD51883e
#PD51883f
#PD51890b
#PD51890c
#PD51890d
#PD51890e
#PD51890f
#PD51920b
#PD51920c
#PD51920d
#PD51920e
#PD51920f

for f in $(ls PD*.caveman_c.annot.asmd.clpm.pass.filtered.vcf);
do
sample=${f%.caveman_c.annot.asmd.clpm.pass.filtered.vcf}
maternal=$(grep -w $sample $infile | cut -f15 | uniq)
if [[ "$maternal" == "-" ]]; 
then 
cp $f ../02_maternal
fi
done

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/02_maternal

for file in $(ls *filtered.vcf);
do
echo $file
grep -v "#" $file | wc -l
done

```

###Blacklists

Remove variants called in hg19 blacklist sites.

Change chromosome labelling in bed file.

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data

cut -c 4- hg19-blacklist.v2.bed > hg19-blacklist.v2.chr.renamed.bed
#cut -c 4- hs_hg19_repeatMasker.bed > hs_hg19_repeatMasker.chr.renamed.bed

```

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/02_maternal

module load bedtools

for f in $(ls *.filtered.vcf);
do
sample=${f%.filtered.vcf}
#echo $sample
bedtools intersect -header -v -a $f -b /lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/LCR-hs37d5.bed /lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/hg19-blacklist.v2.chr.renamed.bed > ../03_blacklist/${sample}.ENCODE.LCR.blacklist.filtered.vcf
done

for f in $(ls *vcf);
do
echo $f
grep -v "#" $f | wc -l
done

```

###cgpVAF MQ30

####Symlink bam files to directory

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq30_cgpvaf

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
grep -v Placenta_PD $infile | cut -f1,3 > all_placenta_samples.txt
grep -v Placenta_PD $infile | cut -f11,14 | sort | uniq >> all_placenta_samples.txt

#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"
#awk -F "\t" '{ if($23 == "Yes") { print $1, $3} }' $infile > all_placenta_samples.txt
#awk -F "\t" '{ if($23 == "Yes") { print $9, $12} }' $infile | sort | uniq >> all_placenta_samples.txt

wc -l all_placenta_samples.txt #all 530

while read line;
do 
set $line
#echo $1
#echo $2
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam.bai
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam.bas
done < all_placenta_samples.txt

ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam
ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam.bai
ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam.bas

```

####Generate bed files

Done per patient. Ignore warnings in PD421 cases regarding missing samples - run matched this time and those were the matched samples

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq30_cgpvaf

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"
#for cord_sample in $(awk -F "\t" '{ if($23 == "Yes") { print $9} }' $infile | uniq);
for cord_sample in $(grep -v Placenta_PD $infile | cut -f11 | uniq);
do

patient1=${cord_sample:0:7}
echo ${cord_sample} > ${patient1}_samples.txt

  for placenta_sample in $(grep ${cord_sample} $infile | awk -F "\t" '{ print $1 }');
  do
    echo ${placenta_sample} >> ${patient1}_samples.txt
  done

  patient2=${placenta_sample:0:7}
  mv ${patient1}_samples.txt ${patient2}_samples.txt

  while read line;
  do
    set $line
    grep -vh '#' /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/03_blacklist/${1}*.filtered.vcf | cut -f 1,2,4,5 >> ${patient2}_subs.bed 
  done < ${patient2}_samples.txt

less ${patient2}_subs.bed | sort -k1,1 -k2,2n -k3,4 | uniq > ${patient2}_subs_unique.bed
mv ${patient2}_subs_unique.bed ${patient2}_subs.bed

done

```

####Run cgpvaf

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq30_cgpvaf

module load cgpVAFcommand

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"

for f in $(grep -v 'Placenta_PD' $infile | cut -f1 | awk '{ $1 = substr($1, 1, 7) } 1' | sort | uniq);
do
  declare -a sampleArray=()
  #echo $1
  
  while read line;
  do
  set $line
  sampleArray+=($1)
    
  done < ${f}_samples.txt
  
  #echo "${sampleArray[@]}"
  
  mkdir -p ${f}
  
  bsub -q normal -o ${f}/%J.$f.out -e ${f}/%J.$f.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 perl /software/CASM/modules/installs/vafcorrect/shim-bin/cgpVaf.pl -d ./ -o $f -a snp -mq 30 -bq 25 -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be .sample.dupmarked.bam -hdr /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b ${f}_subs.bed -nn PDv37is -tn ${sampleArray[@]}
done

```

###cgpVAF MQ0

####Symlink bam files to directory

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq0_cgpvaf

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
grep -v Placenta_PD $infile | cut -f1,3 > all_placenta_samples.txt
grep -v Placenta_PD $infile | cut -f11,14 | sort | uniq >> all_placenta_samples.txt

#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"
#awk -F "\t" '{ if($23 == "Yes") { print $1, $3} }' $infile > all_placenta_samples.txt
#awk -F "\t" '{ if($23 == "Yes") { print $9, $12} }' $infile | sort | uniq >> all_placenta_samples.txt

wc -l all_placenta_samples.txt #all 530

while read line;
do 
set $line
#echo $1
#echo $2
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam.bai
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam.bas
done < all_placenta_samples.txt

ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam
ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam.bai
ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam.bas

```

####Generate bed files

Done per patient. Ignore warnings in PD421 cases regarding missing samples - run matched this time and those were the matched samples

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq0_cgpvaf

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"
#for cord_sample in $(awk -F "\t" '{ if($23 == "Yes") { print $9} }' $infile | uniq);
for cord_sample in $(grep -v Placenta_PD $infile | cut -f11 | uniq);
do

patient1=${cord_sample:0:7}
echo ${cord_sample} > ${patient1}_samples.txt

  for placenta_sample in $(grep ${cord_sample} $infile | awk -F "\t" '{ print $1 }');
  do
    echo ${placenta_sample} >> ${patient1}_samples.txt
  done

  patient2=${placenta_sample:0:7}
  mv ${patient1}_samples.txt ${patient2}_samples.txt

  while read line;
  do
    set $line
    grep -vh '#' /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/03_blacklist/${1}*.filtered.vcf | cut -f 1,2,4,5 >> ${patient2}_subs.bed 
  done < ${patient2}_samples.txt

less ${patient2}_subs.bed | sort -k1,1 -k2,2n -k3,4 | uniq > ${patient2}_subs_unique.bed
mv ${patient2}_subs_unique.bed ${patient2}_subs.bed

done

```

####Run cgpvaf

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq0_cgpvaf

module load cgpVAFcommand

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"

for f in $(grep -v 'Placenta_PD' $infile | cut -f1 | awk '{ $1 = substr($1, 1, 7) } 1' | sort | uniq);
do
  declare -a sampleArray=()
  #echo $1
  
  while read line;
  do
  set $line
  sampleArray+=($1)
    
  done < ${f}_samples.txt
  
  #echo "${sampleArray[@]}"
  
  mkdir -p ${f}
  
  bsub -q normal -o ${f}/%J.$f.out -e ${f}/%J.$f.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 perl /software/CASM/modules/installs/vafcorrect/shim-bin/cgpVaf.pl -d ./ -o $f -a snp -mq 0 -bq 25 -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be .sample.dupmarked.bam -hdr /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b ${f}_subs.bed -nn PDv37is -tn ${sampleArray[@]}
done

```

###BQ + MQ filter

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq30_filter

cp ../mq30_cgpvaf/PD*/*_snp_vaf.tsv .

#remove header info
for f in $(ls *_snp_vaf.tsv);
do
filename=$(basename -- "$f")
filename="${filename%_snp_vaf.tsv}"
#echo $filename
grep -v "##" $f > ${filename}_snp_vaf_nohash.tsv
done

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq0_filter

cp ../mq0_cgpvaf/PD*/*_snp_vaf.tsv .

#remove header info
for f in $(ls *_snp_vaf.tsv);
do
filename=$(basename -- "$f")
filename="${filename%_snp_vaf.tsv}"
#echo $filename
grep -v "##" $f > ${filename}_snp_vaf_nohash.tsv
done

```

####Run

```{r}

#setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/bq_mq_filter")
setwd("/lustre/scratch126/casm/team274sb/to3/placenta_ml_draft_20220711/02_Variant_filtering/matched_ssms/bq_mq_filter")

#manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest = read.table("/lustre/scratch126/casm/team274sb/to3/placenta_ml_draft_20220711/00_Supplementary_data/placenta_ml_metadata_20220505.txt", header = T, sep = '\t', stringsAsFactors = F)

#mq30_files = list.files("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq30_filter", "_snp_vaf_nohash.tsv")
#mq0_files = list.files("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq0_filter", "_snp_vaf_nohash.tsv") #same name
mq30_files = list.files("/lustre/scratch126/casm/team274sb/to3/placenta_ml_draft_20220711/02_Variant_filtering/matched_ssms/mq30_filter", "_snp_vaf_nohash.tsv")
mq0_files = list.files("/lustre/scratch126/casm/team274sb/to3/placenta_ml_draft_20220711/02_Variant_filtering/matched_ssms/mq0_filter", "_snp_vaf_nohash.tsv") #same name

#txt_file = "PDv37is_PD51883b_snp_vaf_nohash.tsv"
for(txt_file in mq30_files){
  
  #mq30 bq25
  data = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq30_filter/", txt_file), comment.char = "", header = T)
  patient = substr(unlist(strsplit(txt_file, 'PDv37is_'))[2], 0, 7)
  cord = unique(manifest[grepl(patient, manifest$Cord_PD), ]$Cord_PD)

  #extract data frames of interest
  Muts = paste(data$Chrom,data$Pos,data$Ref,data$Alt,sep="_")
  Genotype = data[,grepl("VAF",colnames(data))&colnames(data)!="PDv37is_VAF"]
  NR = data[,grepl("DEP",colnames(data))&colnames(data)!="PDv37is_DEP"]
  NV = data[,grepl("MTR",colnames(data))&colnames(data)!="PDv37is_MTR"]
  colnames(Genotype)=colnames(NR)=colnames(NV)=gsub("_VAF","",colnames(Genotype))
  rownames(Genotype)=rownames(NV)=rownames(NR)=Muts
  
  NR_mq30 = NR[, colnames(NR) != cord]
  NV_mq30 = NV[, colnames(NV) != cord]
  
  #mq0 bq25
  data = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq0_filter/", txt_file), comment.char = "", header = T)

  #extract data frames of interest
  Muts = paste(data$Chrom,data$Pos,data$Ref,data$Alt,sep="_")
  Genotype = data[,grepl("VAF",colnames(data))&colnames(data)!="PDv37is_VAF"]
  NR = data[,grepl("DEP",colnames(data))&colnames(data)!="PDv37is_DEP"]
  NV = data[,grepl("MTR",colnames(data))&colnames(data)!="PDv37is_MTR"]
  colnames(Genotype)=colnames(NR)=colnames(NV)=gsub("_VAF","",colnames(Genotype))
  rownames(Genotype)=rownames(NV)=rownames(NR)=Muts
  
  NR_mq0 = NR[, colnames(NR) != cord]
  NV_mq0 = NV[, colnames(NV) != cord]
  
  #bq filter
  low_bq_loci = row.names(NV_mq30[rowSums(NV_mq30 > 3) == 0, ]) #at least 4 supporting reads in at least one sample needed
  good_bq_loci = row.names(NV_mq30)[!row.names(NV_mq30) %in% low_bq_loci]
  
  write.table(low_bq_loci, paste0(patient, "_low_bq_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  write.table(good_bq_loci, paste0(patient, "_good_bq_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  
  NR_mq30_flt = NR_mq30[good_bq_loci,]
  NV_mq30_flt = NV_mq30[good_bq_loci,]
  NR_mq0_flt = NR_mq0[good_bq_loci,]
  NV_mq0_flt = NV_mq0[good_bq_loci,]
  
  #mq filter
  low_mq_loci = row.names(NR_mq30_flt)[rowMeans(NR_mq30_flt / NR_mq0_flt) < 0.9] #an average of >90% of all reads at that locus have MQ >30
  good_mq_loci = row.names(NR_mq30_flt)[!row.names(NR_mq30_flt) %in% low_mq_loci]
  
  write.table(low_mq_loci, paste0(patient, "_low_mq_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  
  #remaining vars
  write.table(good_mq_loci, paste0(patient, "_good_subs.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  
}

```

Calculate the fraction of subs that miss this MQ threshold

```{r}

setwd("/lustre/scratch126/casm/team274sb/to3/placenta_ml_draft_20220711/02_Variant_filtering/matched_ssms/bq_mq_filter")

good_subs = list.files(".", "_good_subs.txt")
placentas = unlist(strsplit(good_subs, "_good_subs.txt"))

df = data.frame(matrix(nrow = length(placentas), ncol = 3, dimnames = list(placentas, c("Placenta", "low_mq_vars", "hi_mq_vars"))))

df$Placenta = placentas
for(i in 1:nrow(df)){
  
  df$low_mq_vars[i] = length(read.table(paste0(df$Placenta[i], "_low_mq_loci.txt"), header = F)[,1])
  df$hi_mq_vars[i] = length(read.table(paste0(df$Placenta[i], "_good_subs.txt"), header = F)[,1])
  
}

df$prop = df$low_mq_vars / (df$low_mq_vars + df$hi_mq_vars)

hist(df$prop)
summary(df$prop)
#Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
#0.005865 0.014332 0.023149 0.051979 0.067044 0.342576 
sum(df$prop >= 0.2) #4

```

####Intersect calls with the original caveman files

Remove problematic line that prevents us writing out the vcf

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/01_standard

for f in $(ls *.caveman_c.annot.asmd.clpm.pass.filtered.vcf);
do
filename=$(basename -- "$f")
sample="${filename%.caveman_c.annot.asmd.clpm.pass.filtered.vcf}"
#echo $f
#echo $sample
grep -v "##vcfProcessLog=" $f > ${sample}.edited.vcf
done

```

#####BQ + MQ

```{r}

setwd('/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/01_standard')

library(VariantAnnotation)

patients = unlist(strsplit(list.files("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/bq_mq_filter", "_good_subs.txt"), "_good_subs.txt"))
all_vcfs = list.files(path = '.', '.edited.vcf')
#manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt", header = T, sep = '\t', stringsAsFactors = F)

for(patient in patients){
  filtered_muts = read.table(paste0('/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/bq_mq_filter/', patient, '_good_subs.txt'), header = F, sep = '\t', stringsAsFactors = F)[,1]
  patient_vcfs = all_vcfs[grepl(paste0(unique(substr(manifest[grepl(patient, manifest$Cord_PD),]$Cord_PD, 0, 7)), "|", unique(substr(manifest[grepl(patient, manifest$Cord_PD),]$Placenta_PD, 0, 7))), all_vcfs)]
  for(vcf in patient_vcfs){
    sample = strsplit(vcf, '\\.')[[1]][1]
    data = VariantAnnotation::readVcf(vcf)
    chr = as.character(seqnames(rowRanges(data)))
    pos = start(ranges(rowRanges(data)))
    ref = as.character(ref(data))
    alt = unlist(CharacterList(alt(data)))
    
    info(data)$mut_ID = paste(chr, pos, ref, alt, sep = '_')
    info(header(data)) = rbind(info(header(data)), DataFrame(Number = '.', Type = 'String', Description = 'Unique mutation ID', row.names = 'mut_ID'))
    filtered_data = data[info(data)$mut_ID %in% filtered_muts]
    
    alt = unlist(CharacterList(alt(filtered_data)))
    fs = vector()
    rs = vector()
    for(i in 1:length(alt)){
      fs_count = geno(filtered_data)[[paste0("F", alt[i], "Z")]][i,"TUMOUR"]
      rs_count = geno(filtered_data)[[paste0("R", alt[i], "Z")]][i,"TUMOUR"]
      fs = c(fs, fs_count)
      rs = c(rs, rs_count)
    }
    
    info(filtered_data)$FS = fs
    info(filtered_data)$RS = rs
    info(header(filtered_data)) = rbind(info(header(filtered_data)), DataFrame(Number = '1', Type = 'Integer', Description = 'Forward strand mutant allele count', row.names = 'FS'))
    info(header(filtered_data)) = rbind(info(header(filtered_data)), DataFrame(Number = '1', Type = 'Integer', Description = 'Reverse strand mutant allele count', row.names = 'RS'))
    
    writeVcf(filtered_data, filename = paste0("../04_bq_mq/", sample, '.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.filtered.vcf'))
    
  }
}

```

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/04_bq_mq

for file in $(ls *.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.filtered.vcf);
do
echo $file
grep -v "#" $file | wc -l
done

```

######Convert to text file

######Function

```{r}

library(VariantAnnotation)

annotated_subs_text_file = function(vcf_file){
    vcf = readVcf(vcf_file)
    
    if(nrow(vcf) > 0){
      sample = unlist(strsplit(vcf_file, '.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.filtered.vcf'))
      case = substr(sample, 0, 7)
      reads = cbind((geno(vcf)$FAZ)[,2]+(geno(vcf)$RAZ)[,2],(geno(vcf)$FCZ)[,2]+(geno(vcf)$RCZ)[,2],
                  (geno(vcf)$FGZ)[,2]+(geno(vcf)$RGZ)[,2],(geno(vcf)$FTZ)[,2]+(geno(vcf)$RTZ)[,2])
      var.df = data.frame(Chr = as.character(seqnames(rowRanges(vcf))),
                      Position = start(ranges(rowRanges(vcf))),
                      Wildtype = as.character(ref(vcf)),
                      Mutant = unlist(CharacterList(alt(vcf))))
      var.df$Total_read_depth=rowSums(reads)
      var.df$Variant_read_depth=NA
      var.df$Case_ID = case
      var.df$Sample = sample
      var.df = var.df[, c(7, 8, 1:6)]
      colnames(reads)=c("A","C","G","T")
      for (k in c("A","C","G","T")){
        var.df$Variant_read_depth[var.df$Mutant==k] = reads[var.df$Mutant==k,k]
      }
  
      var.df$VAF = var.df$Variant_read_depth/ var.df$Total_read_depth
      var.df$Mutation = info(vcf)$VT
      vagrent_annots = lapply(strsplit(info(vcf)$VW, '\\|'), `length<-`, max(lengths(strsplit(info(vcf)$VW, '\\|')))) #make all elements in list the same length for subsetting purposes
      
      if(max(unlist(lapply(vagrent_annots, function(x) length(x)))) == 1){ #where none land in gene footprints and single NA values are found in each case
        var.df$Gene = NA
        var.df$Transcript = NA
        var.df$Codon = NA
        var.df$Protein = NA
      } else {
        var.df$Gene = sapply(vagrent_annots, '[[', 1)
        var.df$Transcript = sapply(vagrent_annots, '[[', 3)
        var.df$Codon = sapply(vagrent_annots, '[[', 4)
        var.df$Protein = sapply(vagrent_annots, '[[', 5)
      }
      
      var.df$Consequence = unlist(lapply(info(vcf)$VC, function(x) if(identical(x, character(0))) NA_character_ else x))
      
      write.table(var.df, paste0(sample, '.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.filtered.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
      
    }
}

```

######Run

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/04_bq_mq")

vcf_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.filtered.vcf")

for(file in vcf_files){
  annotated_subs_text_file(file)
}

```

###Direction filter

At least one read supporting variant in both directions.

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/04_bq_mq

module load bcftools

#bq mq
for f in $(ls *.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.filtered.vcf);
do
filename=$(basename -- "$f")
sample="${filename%.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.filtered.vcf}"
bcftools filter -e 'INFO/FS = 0 | INFO/RS = 0' $f > ../05_direction/${sample}.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.filtered.vcf
done

```

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/05_direction

#bq mq
for file in $(ls PD*.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.filtered.vcf);
do
echo $file
grep -v "#" $file | wc -l
done

```

###Indel site filtering

A few subs near to indels have snuck through. Filter against bed file of 5bp window around indels within the unmatched pindel file.

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/05_direction

module load bedtools

#bq mq
for f in $(ls PD*.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.filtered.vcf);
do
filename=$(basename -- "$f")
sample="${filename%.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.filtered.vcf}"
bedtools intersect -v -header -a ${sample}.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.filtered.vcf -b /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/indels/${sample}.all_indels.5bp.window.bed > ../06_indel/${sample}.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.filtered.vcf
done

```

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/06_indel

#bq mq
for file in $(ls PD*.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.filtered.vcf);
do
echo $file
grep -v "#" $file | wc -l
done

```

###Convert to text file

####Function

```{r}

library(VariantAnnotation)

annotated_subs_text_file_bq_mq = function(vcf_file){
    vcf = readVcf(vcf_file)
    
    if(nrow(vcf) > 0){
      sample = unlist(strsplit(vcf_file, '.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.filtered.vcf'))
      case = substr(sample, 0, 7)
      reads = cbind((geno(vcf)$FAZ)[,2]+(geno(vcf)$RAZ)[,2],(geno(vcf)$FCZ)[,2]+(geno(vcf)$RCZ)[,2],
                  (geno(vcf)$FGZ)[,2]+(geno(vcf)$RGZ)[,2],(geno(vcf)$FTZ)[,2]+(geno(vcf)$RTZ)[,2])
      var.df = data.frame(Chr = as.character(seqnames(rowRanges(vcf))),
                      Position = start(ranges(rowRanges(vcf))),
                      Wildtype = as.character(ref(vcf)),
                      Mutant = unlist(CharacterList(alt(vcf))))
      var.df$Total_read_depth=rowSums(reads)
      var.df$Variant_read_depth=NA
      var.df$Case_ID = case
      var.df$Sample = sample
      var.df = var.df[, c(7, 8, 1:6)]
      colnames(reads)=c("A","C","G","T")
      for (k in c("A","C","G","T")){
        var.df$Variant_read_depth[var.df$Mutant==k] = reads[var.df$Mutant==k,k]
      }
  
      var.df$VAF = var.df$Variant_read_depth/ var.df$Total_read_depth
      var.df$Mutation = info(vcf)$VT
      vagrent_annots = lapply(strsplit(info(vcf)$VW, '\\|'), `length<-`, max(lengths(strsplit(info(vcf)$VW, '\\|')))) #make all elements in list the same length for subsetting purposes
      
      if(max(unlist(lapply(vagrent_annots, function(x) length(x)))) == 1){ #where none land in gene footprints and single NA values are found in each case
        var.df$Gene = NA
        var.df$Transcript = NA
        var.df$Codon = NA
        var.df$Protein = NA
      } else {
        var.df$Gene = sapply(vagrent_annots, '[[', 1)
        var.df$Transcript = sapply(vagrent_annots, '[[', 3)
        var.df$Codon = sapply(vagrent_annots, '[[', 4)
        var.df$Protein = sapply(vagrent_annots, '[[', 5)
      }
      
      var.df$Consequence = unlist(lapply(info(vcf)$VC, function(x) if(identical(x, character(0))) NA_character_ else x))
      
      write.table(var.df, paste0(sample, '.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.filtered.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
      
    }
}

```

####Run

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/06_indel")

vcf_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.filtered.vcf")

for(file in vcf_files){
  annotated_subs_text_file_bq_mq(file)
}

```

###Correct PD45581c chr10 calls

chr10 trisomy with maternal UPD in cord - hence calling paternal chr10 SNPs as somatic when run matched.

```{r}

#bq mq
subs = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/06_indel/PD45581c.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.filtered.txt", header = T, sep = '\t', stringsAsFactors = F)

chr10_pval = rep(NA, length(subs[subs$Chr == 10,]$Variant_read_depth))
for(i in 1:length(chr10_pval)){
  chr10_pval[i] = binom.test(subs[subs$Chr == 10,]$Variant_read_depth[i], subs[subs$Chr == 10,]$Total_read_depth[i], p = 1/3, alt = "less")$p.value
}

qval = p.adjust(chr10_pval, method = "BH")
germline = log10(qval) > -3

subs$keep = "Y"
subs[subs$Chr == 10,][germline,]$keep = "N"

subs_flt = subs[subs$keep == "Y",]

write.table(subs_flt[, 1:15], "/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/06_indel/PD45581c.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.filtered.txt", col.names = T, row.names = F, sep = '\t', quote = F)

```

###Recurrent artefact filter

The latest batch of samples in particular is of lower DNA quality with some recurrent artefacts (across placentas) that should be removed.

```{r}

#bq mq
setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/06_indel")

txt_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.filtered.txt")

all_subs = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  all_subs = rbind(all_subs, data)
}

all_subs$mut_ID = paste(all_subs$Chr, all_subs$Position, all_subs$Wildtype, all_subs$Mutant, sep = "_")

#recurrent artefacts?
recurrent_vars = all_subs[!duplicated(all_subs[, c("Case_ID", "mut_ID")]),]$mut_ID[duplicated(all_subs[!duplicated(all_subs[, c("Case_ID", "mut_ID")]),]$mut_ID)] #452, 1377 if MQ filter wasn't applied
length(recurrent_vars) #452
length(unique(recurrent_vars)) #376, some repeated more than once

#probably a more elegent way of writing it in retrospect 26/1/23
recurrent_vars2 = names(table(all_subs[!duplicated(all_subs[, c("Case_ID", "mut_ID")]),]$mut_ID))[table(all_subs[!duplicated(all_subs[, c("Case_ID", "mut_ID")]),]$mut_ID) > 1]
length(recurrent_vars2) #376
sum(recurrent_vars2 %in% recurrent_vars) #376 perfect match

temp = all_subs[all_subs$mut_ID %in% recurrent_vars2,]
temp[order(temp$mut_ID),]

library(GenomicRanges)
library(Rsamtools)
library(MASS)
library(optparse)

genomeFile <- "/nfs/cancer_ref01/Homo_sapiens/37/genome.fa"

input = matrix(unlist(strsplit(recurrent_vars, "_")), ncol = 4, byrow = T)
  
mutations <- input
mutations <- as.data.frame(mutations, stringsAsFactors = F)
colnames(mutations) <- c("chr","pos","ref","mut")
mutations$pos <- as.numeric(mutations$pos)
mutations <- mutations[(mutations$ref %in% c("A","C","G","T")) & (mutations$mut %in% c("A","C","G","T")) & mutations$chr %in% c(1:22,"X","Y"),]
mutations$trinuc_ref <- as.vector(scanFa(genomeFile, GRanges(mutations$chr, IRanges(as.numeric(mutations$pos)-1, 
                                                                                      as.numeric(mutations$pos)+1))))
  
# 2. Annotating the mutation from the pyrimidine base
ntcomp <- c(T="A",G="C",C="G",A="T")
mutations$sub <- paste(mutations$ref, mutations$mut,sep=">")
mutations$trinuc_ref_py <- mutations$trinuc_ref
for (j in 1:nrow(mutations)) {
  if (mutations$ref[j] %in% c("A","G")) { # Purine base
    mutations$sub[j] = paste(ntcomp[mutations$ref[j]],ntcomp[mutations$mut[j]],sep=">")
    mutations$trinuc_ref_py[j] = paste(ntcomp[rev(strsplit(mutations$trinuc_ref[j],split="")[[1]])],collapse="")
  }
}
  
# 3. Counting subs
freqs = table(paste(mutations$sub,paste(substr(mutations$trinuc_ref_py,1,1),substr(mutations$trinuc_ref_py,3,3),sep="-"),sep=","))
sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")
freqs_full = freqs[full_vec]; freqs_full[is.na(freqs_full)] = 0; names(freqs_full) = full_vec
  
xstr = paste(substr(full_vec,5,5), substr(full_vec,1,1), substr(full_vec,7,7), sep="")
  
colvec = rep(c("dodgerblue","black","red","grey70","olivedrab3","plum2"),each=16)
y = freqs_full; maxy = max(y)
h = barplot(y, las=2, col=colvec, border=NA, ylim=c(0,maxy*1.5), space=1, cex.names=0.6, names.arg=xstr, ylab="Number of mutations")
for (j in 1:length(sub_vec)) {
  xpos = h[c((j-1)*16+1,j*16)]
  rect(xpos[1]-0.5, maxy*1.2, xpos[2]+0.5, maxy*1.3, border=NA, col=colvec[j*16])
  text(x=mean(xpos), y=maxy*1.3, pos=3, label=sub_vec[j])
}    

table(mutations$sub)
table(all_subs[all_subs$mut_ID %in% recurrent_vars,]$Consequence) #no coding variants

#now filter
txt_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.filtered.txt")

for(file in txt_files){
  sample = unlist(strsplit(file, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$mut_ID = paste(data$Chr, data$Position, data$Wildtype, data$Mutant, sep = "_")
  data_flt = data[!data$mut_ID %in% recurrent_vars,]
  write.table(data_flt, paste0("../07_recurrent/", sample, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt"), col.names = T, row.names = F, sep = '\t', quote = F)
}

```

##Subs - unmatched cord

Putative variants called by CaVEMan.

###PASS, ASMD & CLPM filtering
```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/subs/01_standard

module load bcftools

#placental biopsies
infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/Placenta_bulk_landscape_project_metadata_20211112.txt'
awk -F "\t" '{ if($23 == "Yes") { print $1, $3} }' $infile| while read -r -a tmp;
do
#echo ${tmp[0]} #PD ID
#echo ${tmp[1]} #canapps project
bcftools filter -i 'FILTER == "PASS" & INFO/ASMD >= 140 & INFO/CLPM == 0' /nfs/cancer_ref01/nst_links/live/${tmp[1]}/${tmp[0]}/${tmp[0]}.caveman_c.annot.vcf.gz > ${tmp[0]}.caveman_c.annot.asmd.clpm.pass.filtered.vcf
done

#new placental biopsies 6/5/22
infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
awk -F "\t" '{ if($24 == "No") { print $1, $3} }' $infile | while read -r -a tmp;
do
#echo ${tmp[0]} #PD ID
#echo ${tmp[1]} #canapps project
bcftools filter -i 'FILTER == "PASS" & INFO/ASMD >= 140 & INFO/CLPM == 0' /nfs/cancer_ref01/nst_links/live/${tmp[1]}/${tmp[0]}/${tmp[0]}.caveman_c.annot.vcf.gz > ${tmp[0]}.caveman_c.annot.asmd.clpm.pass.filtered.vcf
done

#cord biopsies
infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/Placenta_bulk_landscape_project_metadata_20211112.txt'
awk -F "\t" '{ if($23 == "Yes") { print $9, $12} }' $infile | sort | uniq | while read -r -a tmp;
do
#echo ${tmp[0]} #PD ID
#echo ${tmp[1]} #canapps project
bcftools filter -i 'FILTER == "PASS" & INFO/ASMD >= 140 & INFO/CLPM == 0' /nfs/cancer_ref01/nst_links/live/${tmp[1]}/${tmp[0]}/${tmp[0]}.caveman_c.annot.vcf.gz > ${tmp[0]}.caveman_c.annot.asmd.clpm.pass.filtered.vcf
done

#new cord biopsy 6/5/22
infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
awk -F "\t" '{ if($24 == "No") { print $11, $14} }' $infile | sort | uniq | while read -r -a tmp;
do
#echo ${tmp[0]} #PD ID
#echo ${tmp[1]} #canapps project
bcftools filter -i 'FILTER == "PASS" & INFO/ASMD >= 140 & INFO/CLPM == 0' /nfs/cancer_ref01/nst_links/live/${tmp[1]}/${tmp[0]}/${tmp[0]}.caveman_c.annot.vcf.gz > ${tmp[0]}.caveman_c.annot.asmd.clpm.pass.filtered.vcf
done

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/Placenta_bulk_landscape_project_metadata_20211112.txt'
grep "PD51922\\|PD51912\\|PD51914\\|PD51916"  $infile | awk -F "\t" '{ if($23 == "Yes") { print $1, $3} }' | while read -r -a tmp;
do
#echo ${tmp[0]} #PD ID
#echo ${tmp[1]} #canapps project
bcftools filter -i 'FILTER == "PASS" & INFO/ASMD >= 140 & INFO/CLPM == 0' /nfs/cancer_ref01/nst_links/live/${tmp[1]}/${tmp[0]}/${tmp[0]}.caveman_c.annot.vcf.gz > ${tmp[0]}.caveman_c.annot.asmd.clpm.pass.filtered.vcf
done

#check all run unmatched
for f in $(ls *.caveman_c.annot.asmd.clpm.pass.filtered.vcf);
do
less $f | grep -m1 "TUMOUR" >> matched_sample_list.txt
less $f | grep -m1 "NORMAL" >> matched_sample_list.txt
done

grep "PDv37is" matched_sample_list.txt | wc -l #530
ls *.caveman_c.annot.asmd.clpm.pass.filtered.vcf | wc -l #530, looks like we have unmatched for all, 6/5/22

#Check that none of the vcfs have been truncated and are missing chromosomes
ls *.caveman_c.annot.asmd.clpm.pass.filtered.vcf | while read FILE ; do grep "^#" -v "$FILE" | cut -f1 | uniq -c > ../qc_chrom_counts/"$FILE".counts.txt ; done 
wc -l ../qc_chrom_counts/*.counts.txt #none are truncated 15/11/21

```

###Maternal SNP filtering

Remove any substitution called in mother to eliminate any maternal contamination, where maternal contamination data is available

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/subs/01_standard

module load bedtools

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
for f in $(ls PD*.caveman_c.annot.asmd.clpm.pass.filtered.vcf);
do
sample=${f%.caveman_c.annot.asmd.clpm.pass.filtered.vcf}
maternal=$(grep -w $sample $infile | cut -f15 | uniq)
if [[ "$maternal" != "-" ]]; 
then 
bedtools intersect -header -v -a $f -b /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/maternal_snps/${maternal}.caveman_c.annot.vcf.gz > ../02_maternal/${sample}.caveman_c.annot.asmd.clpm.pass.maternal.filtered.vcf
fi
done

#no maternal sample for the following:
#PD51883b
#PD51883c
#PD51883d
#PD51883e
#PD51883f
#PD51890b
#PD51890c
#PD51890d
#PD51890e
#PD51890f
#PD51920b
#PD51920c
#PD51920d
#PD51920e
#PD51920f

for f in $(ls PD*.caveman_c.annot.asmd.clpm.pass.filtered.vcf);
do
sample=${f%.caveman_c.annot.asmd.clpm.pass.filtered.vcf}
maternal=$(grep -w $sample $infile | cut -f15 | uniq)
if [[ "$maternal" == "-" ]]; 
then 
cp $f ../02_maternal
fi
done

```

###Indel site filtering

Remove any subs called at the site of an indel - previous work demonstrated that this method tends to let a few subs called at the sites of germline indels slip through. This will pre-emptively ensure that this is not the case.

First, copy over the pindel VCF files
```{bash}

#generate bed file of indel regions to avoid
cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/indels/01_pindel

#placental biopsies
infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/Placenta_bulk_landscape_project_metadata_20211112.txt'
awk -F "\t" '{ if($23 == "Yes") { print $1, $3} }' $infile| while read -r -a tmp;
do
#echo ${tmp[0]} #PD ID
#echo ${tmp[1]} #canapps project
cp /nfs/cancer_ref01/nst_links/live/${tmp[1]}/${tmp[0]}/${tmp[0]}.pindel.annot.vcf.gz .
done

#placental biopsies 6/5/22
infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
awk -F "\t" '{ if($24 == "No") { print $1, $3} }' $infile| while read -r -a tmp;
do
#echo ${tmp[0]} #PD ID
#echo ${tmp[1]} #canapps project
cp /nfs/cancer_ref01/nst_links/live/${tmp[1]}/${tmp[0]}/${tmp[0]}.pindel.annot.vcf.gz .
done

#cord biopsies
infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/Placenta_bulk_landscape_project_metadata_20211112.txt'
awk -F "\t" '{ if($23 == "Yes") { print $9, $12} }' $infile | sort | uniq | while read -r -a tmp;
do
#echo ${tmp[0]} #PD ID
#echo ${tmp[1]} #canapps project
cp /nfs/cancer_ref01/nst_links/live/${tmp[1]}/${tmp[0]}/${tmp[0]}.pindel.annot.vcf.gz .
done

#cord biopsy 6/5/22
cp /nfs/cancer_ref01/nst_links/live/2058/PD42145b/PD42145b.pindel.annot.vcf.gz .

#check all run unmatched
for f in $(ls *.pindel.annot.vcf.gz);
do
#echo $f
gunzip $f
done

for f in $(ls *.pindel.annot.vcf);
do
zless $f | grep -m1 "TUMOUR" >> matched_sample_list.txt
zless $f | grep -m1 "NORMAL" >> matched_sample_list.txt
done

grep "PDv37is" matched_sample_list.txt | wc -l #530
ls *.pindel.annot.vcf | wc -l #530, looks like we have unmatched for all, 6/5/22

#Check that none of the vcfs have been truncated and are missing chromosomes
ls *.pindel.annot.vcf | while read FILE ; do grep "^#" -v "$FILE" | cut -f1 | uniq -c > ../qc_chrom_counts/"$FILE".counts.txt ; done 
wc -l ../qc_chrom_counts/*.pindel.annot.vcf.counts.txt #none are truncated 6/5/22

```

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/indels/01_pindel

bsub -q normal -o bed_file_5bp.out -e bed_file_5bp.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 bash ../bed_file_5bp_20220323.sh

#bed_file_5bp_20220323.sh
for f in $(ls *.pindel.annot.vcf.gz);
do
filename=$(basename -- "$f")
sample="${filename%.pindel.annot.vcf.gz}"
#zless ${sample}.pindel.annot.vcf.gz | awk '! /\#/' | awk '{if(length($4) > length($5)) print $1"\t"($2-1)"\t"($2+length($4)-1); else print $1"\t"($2-1)"\t"($2+length($5)-1)}' > ${sample}.all_indels.bed
zless ${sample}.pindel.annot.vcf.gz | awk '! /\#/' | awk '{if(length($4) > length($5)) print $1"\t"($2-6)"\t"($2+length($4)+4); else print $1"\t"($2-6)"\t"($2+length($5)+4)}' > ${sample}.all_indels.5bp.window.bed
done

#now intersect with subs
cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/subs/02_maternal

module load bedtools

#for those without maternal samples
for f in $(ls PD*.caveman_c.annot.asmd.clpm.pass.filtered.vcf);
do
filename=$(basename -- "$f")
sample="${filename%.caveman_c.annot.asmd.clpm.pass.filtered.vcf}"
bedtools intersect -v -header -a ${sample}.caveman_c.annot.asmd.clpm.pass.filtered.vcf -b ../../indels/${sample}.all_indels.5bp.window.bed > ../03_indel/${sample}.caveman_c.annot.asmd.clpm.pass.indel.filtered.vcf
done

#for those with maternal samples
for f in $(ls PD*.caveman_c.annot.asmd.clpm.pass.maternal.filtered.vcf);
do
filename=$(basename -- "$f")
sample="${filename%.caveman_c.annot.asmd.clpm.pass.maternal.filtered.vcf}"
bedtools intersect -v -header -a ${sample}.caveman_c.annot.asmd.clpm.pass.maternal.filtered.vcf -b ../../indels/${sample}.all_indels.5bp.window.bed > ../03_indel/${sample}.caveman_c.annot.asmd.clpm.pass.maternal.indel.filtered.vcf
done

```

###Germline, coverage & recurrent artefact filter

####Symlink bam files to directory

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cgpvaf

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
grep -v Placenta_PD $infile | cut -f1,3 > all_placenta_samples.txt
grep -v Placenta_PD $infile | cut -f11,14 | sort | uniq >> all_placenta_samples.txt

#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"
#awk -F "\t" '{ if($23 == "Yes") { print $1, $3} }' $infile > all_placenta_samples.txt
#awk -F "\t" '{ if($23 == "Yes") { print $9, $12} }' $infile | sort | uniq >> all_placenta_samples.txt

wc -l all_placenta_samples.txt #all 530

while read line;
do 
set $line
#echo $1
#echo $2
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam.bai
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam.bas
done < all_placenta_samples.txt

ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam
ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam.bai
ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam.bas

```

####Generate bed files

Done per patient

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cgpvaf

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"
#for cord_sample in $(awk -F "\t" '{ if($23 == "Yes") { print $9} }' $infile | uniq);
for cord_sample in $(grep -v Placenta_PD $infile | cut -f11 | uniq);
do

patient1=${cord_sample:0:7}
echo ${cord_sample} > ${patient1}_samples.txt

  for placenta_sample in $(grep ${cord_sample} $infile | awk -F "\t" '{ print $1 }');
  do
    echo ${placenta_sample} >> ${patient1}_samples.txt
  done

  patient2=${placenta_sample:0:7}
  mv ${patient1}_samples.txt ${patient2}_samples.txt

  while read line;
  do
    set $line
    grep -vh '#' /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/subs/03_indel/${1}*.indel.filtered.vcf | cut -f 1,2,4,5 >> ${patient2}_subs.bed 
  done < ${patient2}_samples.txt

less ${patient2}_subs.bed | sort -k1,1 -k2,2n -k3,4 | uniq > ${patient2}_subs_unique.bed
mv ${patient2}_subs_unique.bed ${patient2}_subs.bed

done

```

####Run cgpvaf

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cgpvaf

module load cgpVAFcommand

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"

for f in $(grep -v 'Placenta_PD' $infile | cut -f1 | awk '{ $1 = substr($1, 1, 7) } 1' | sort | uniq);
do
  declare -a sampleArray=()
  #echo $1
  
  while read line;
  do
  set $line
  sampleArray+=($1)
    
  done < ${f}_samples.txt
  
  #echo "${sampleArray[@]}"
  
  mkdir -p ${f}
  
  bsub -q normal -o ${f}/%J.$f.out -e ${f}/%J.$f.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 perl /software/CASM/modules/installs/vafcorrect/shim-bin/cgpVaf.pl -d ./ -o $f -a snp -mq 30 -bq 25 -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be .sample.dupmarked.bam -hdr /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b ${f}_subs.bed -nn PDv37is -tn ${sampleArray[@]}
done

```

Copy over the per case pileup files
```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/tc16_filtering

cp ../cgpvaf/PD*/*_snp_vaf.tsv /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/tc16_filtering

#remove header info
for f in $(ls *_snp_vaf.tsv);
do
filename=$(basename -- "$f")
filename="${filename%_snp_vaf.tsv}"
#echo $filename
grep -v "##" $f > ${filename}_snp_vaf_nohash.tsv
done

#list of samples with large scale CN changes that will profoundly affect germline filtering
#nano samples_with_cnv.txt
#PD42154b3
#PD45581c
#PD45581e
#PD45581f
#PD51880e
#PD51899b

```

####Run filter

Get the cgpVAF pileup from unmatched run and interrogate for variants called in cord

```{r}

#cord_bulk_mut_filter_20220623.R
library(VariantAnnotation)
library(optparse)

source("/nfs/users/nfs_t/to3/scripts/germline_exact_binom.R")

option_list = list(
  make_option(c("-t", "--text_file"), type="character", default=NULL, help="Text file containing patient putative calls", metavar="character"))

opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

#setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/tc16_filtering")
txt_file = opt$text_file
#txt_file = "PDv37is_PD42157b_snp_vaf_nohash.tsv"

#read in metadata
manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)
contam_samples = c(manifest[manifest$Biopsy_freemix >= 0.01 & !is.na(manifest$Biopsy_freemix),]$Placenta_PD, unique(manifest[manifest$Cord_freemix >= 0.01 & !is.na(manifest$Cord_freemix),]$Cord_PD)) #list of samples with over 1% contamination, likely maternal

#read in relevant files
data = read.table(txt_file, comment.char = "", header = T)
#samples_with_CNVs = read.table(cna_file, header = F, stringsAsFactors = F)[,1]
#samples_with_CNVs = sort(unique(c(svs_flt$file_sample, ascatPCA_flt$sample, battenberg_flt$sample))) #copy number changes

patient = substr(unlist(strsplit(txt_file, 'PDv37is_'))[2], 0, 7)
cord = unique(manifest[grepl(patient, manifest$Cord_PD),]$Cord_PD)

#extract data frames of interest
Muts = paste(data$Chrom,data$Pos,data$Ref,data$Alt,sep="_")
Genotype = data[,grepl("VAF",colnames(data))&colnames(data)!="PDv37is_VAF"]
NR = data[,grepl("DEP",colnames(data))&colnames(data)!="PDv37is_DEP"]
NV = data[,grepl("MTR",colnames(data))&colnames(data)!="PDv37is_MTR"]
colnames(Genotype)=colnames(NR)=colnames(NV)=gsub("_VAF","",colnames(Genotype))
rownames(Genotype)=rownames(NV)=rownames(NR)=Muts
write.table(Muts, paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/germline_filter/", patient,"_input_muts.txt"), col.names = F, row.names = F, sep ='\t', quote = F)

#establish depth of sequencing across autosomal and sex chromosomes
XY_chromosomal = grepl("X|Y",Muts)
autosomal = !XY_chromosomal
xy_depth=mean(rowMeans(NR[XY_chromosomal,]))
autosomal_depth=mean(rowMeans(NR[autosomal,]))

#list of samples
samples = colnames(NR)

#samples without a copy number change and <1% maternal contaminaton
#noCNVs=!samples%in% unique(c(contam_samples, samples_with_CNVs))
noContam = !samples %in% contam_samples #only on this basis - will remove variants that are germline but at sites of somatic CN change later

#establish gender based on the relative coverage of X|Y vs autosomal chromosomes
gender='male'
if(xy_depth>0.8*autosomal_depth) gender='female'

#need two samples without 
if(sum(noContam) >= 2){
  
  #total depth filter
  if(gender=="male") Depth_filter = (rowMeans(NR[,noContam]) > 10 & rowMeans(NR[,noContam]) < 250 & autosomal) | (rowMeans(NR[,noContam]) > 10 & rowMeans(NR[,noContam]) < 125 & XY_chromosomal)
  if(gender=="female") Depth_filter = (rowMeans(NR[,noContam]) > 10 & rowMeans(NR[,noContam]) < 250) #changed 13/5/22 as some umbilical cord samples are much more deeply sequenced

  #apply total depth filter to data frames
  NR = NR[Depth_filter,]
  NV = NV[Depth_filter,]
  Muts = Muts[Depth_filter]
  Genotype = Genotype[Depth_filter,]

  avg_total_seq_depth = sum(colMeans(NR[,noContam]))

}

if(sum(noContam) >= 2 & avg_total_seq_depth >= 100){ #need at least 2 samples that can undertake germline filtering
  
  #minimum variant depth filter in umbilical cord
  Depth_filter_2 = NV[, cord] >= 4 #4 or more HQ reads
  
  #apply variant depth filter to data frames
  NR = NR[Depth_filter_2,]
  NV = NV[Depth_filter_2,]
  Muts = Muts[Depth_filter_2]
  Genotype = Genotype[Depth_filter_2,]

  #run germline filter
  germline=exact.binomial(gender = gender, NV = NV[,noContam], NR = NR[,noContam], cutoff = -3) #determine which variants are germline, changed cutoff 23/6/22 as it is too stringent for tissues with such little sampling

  #write out germline and putative somatic variants
  write.table(row.names(NR)[germline], paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/germline_filter/", patient,"_germline_snp_variants.txt"), row.names = F, col.names = F, quote=F)
  write.table(row.names(NR)[!germline], paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/germline_filter/", patient,"_somatic_snp_variants.txt"), row.names = F, col.names = F, quote=F)
  write.table(median(NR[germline & !grepl("X|Y", row.names(NR)), cord]), paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/germline_filter/", patient,"_cord_median_germline_auto_snp_depth.txt"), row.names = F, col.names = F, quote=F)
  write.table(median(NR[germline & grepl("X|Y", row.names(NR)), cord]), paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/germline_filter/", patient,"_cord_median_germline_sex_snp_depth.txt"), row.names = F, col.names = F, quote=F)
  
}

```

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/tc16_filtering

bsub -q normal -o %J.out -e %J.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 bash run_cord_bulk_sub_filter_20220623.sh

```

```{bash}

#run_cord_bulk_sub_filter_20220623.sh
export PATH=/software/R-3.6.1/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="~/custom_libraries/R/x86_64-pc-linux-gnu-library/3.6_dw"

for f in $(ls *_snp_vaf_nohash.tsv);
do
/software/R-3.6.1/bin/Rscript ~/scripts/cord_bulk_mut_filter_20220623.R -t $f
done

```

####Identify recurrent calls

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/germline_filter")

input_mut_files = list.files(".", "_input_muts.txt")

all_vars = c()

for(file in input_mut_files){
  data = read.table(file, header = F, sep = '\t', stringsAsFactors = F)[,1]
  all_vars = c(all_vars, data)
}

recurrent_vars = unique(all_vars[duplicated(all_vars)])

write.table(recurrent_vars, "placenta_recurrent_vars_20220623.txt", col.names = F, row.names = F, sep = '\t', quote = F)

```

###Intersect calls with the original caveman files

Remove problematic line that prevents us writing out the vcf

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/subs/01_standard

for f in $(ls *.caveman_c.annot.asmd.clpm.pass.filtered.vcf);
do
filename=$(basename -- "$f")
sample="${filename%.caveman_c.annot.asmd.clpm.pass.filtered.vcf}"
#echo $f
#echo $sample
grep -v "##vcfProcessLog=" $f > ${sample}.edited.vcf
done

```

```{r}

setwd('/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/subs/01_standard')

library(VariantAnnotation)

patients = unlist(strsplit(list.files("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/germline_filter", "_somatic_snp_variants.txt"), "_somatic_snp_variants.txt"))
all_vcfs = list.files(path = '.', '.edited.vcf')
manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)
recurrent_vars = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/germline_filter/placenta_recurrent_vars_20220623.txt", header = F, sep = '\t', stringsAsFactors = F)[,1]

for(patient in patients){
  filtered_muts = read.table(paste0('/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/germline_filter/', patient, '_somatic_snp_variants.txt'), header = F, sep = '\t', stringsAsFactors = F)[,1]
  patient_vcfs = all_vcfs[grepl(paste0(unique(substr(manifest[grepl(patient, manifest$Cord_PD),]$Cord_PD, 0, 7)), "|", unique(substr(manifest[grepl(patient, manifest$Cord_PD),]$Placenta_PD, 0, 7))), all_vcfs)]
  for(vcf in patient_vcfs){
    sample = strsplit(vcf, '\\.')[[1]][1]
    data = VariantAnnotation::readVcf(vcf)
    chr = as.character(seqnames(rowRanges(data)))
    pos = start(ranges(rowRanges(data)))
    ref = as.character(ref(data))
    alt = unlist(CharacterList(alt(data)))
    
    info(data)$mut_ID = paste(chr, pos, ref, alt, sep = '_')
    info(header(data)) = rbind(info(header(data)), DataFrame(Number = '.', Type = 'String', Description = 'Unique mutation ID', row.names = 'mut_ID'))
    filtered_data = data[info(data)$mut_ID %in% filtered_muts & !info(data)$mut_ID %in% recurrent_vars]
    
    alt = unlist(CharacterList(alt(filtered_data)))
    fs = vector()
    rs = vector()
    for(i in 1:length(alt)){
      fs_count = geno(filtered_data)[[paste0("F", alt[i], "Z")]][i,"TUMOUR"]
      rs_count = geno(filtered_data)[[paste0("R", alt[i], "Z")]][i,"TUMOUR"]
      fs = c(fs, fs_count)
      rs = c(rs, rs_count)
    }
    
    info(filtered_data)$FS = fs
    info(filtered_data)$RS = rs
    info(header(filtered_data)) = rbind(info(header(filtered_data)), DataFrame(Number = '1', Type = 'Integer', Description = 'Forward strand mutant allele count', row.names = 'FS'))
    info(header(filtered_data)) = rbind(info(header(filtered_data)), DataFrame(Number = '1', Type = 'Integer', Description = 'Reverse strand mutant allele count', row.names = 'RS'))
    
    writeVcf(filtered_data, filename = paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/01_binom/", sample, '.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.filtered.vcf'))
    
  }
}

```

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/01_binom/

#for file in $(ls *.caveman_c.annot.asmd.clpm.pass.maternal.depth.binom.filtered.vcf);
#do
#sample=${file%.caveman_c.annot.asmd.clpm.pass.maternal.depth.binom.filtered.vcf}
#mv $file $sample.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.filtered.vcf
#done

for file in $(ls *.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.filtered.vcf);
do
echo $file
grep -v "#" $file | wc -l
done

```

###PON filter


```{bash}

#in a screen
bsub -Is -q yesterday -R "select[mem>100000] rusage[mem=100000] span[hosts=1]" -M 100000 bash

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/01_binom/

module load bedtools

for f in $(ls PD*.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.filtered.vcf);
do
filename=$(basename -- "$f")
sample="${filename%.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.filtered.vcf}"
bedtools intersect -v -header -a ${sample}.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.filtered.vcf -b /lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/Mutect2-WGS-panel-b37.vcf > ../02_pon/${sample}.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.pon.filtered.vcf
done

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/02_pon

for f in $(ls *.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.pon.filtered.vcf);
do
echo $f
grep -v "#" $f | wc -l
done

```

###Common SNP filter

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/02_pon

module load bedtools

for f in $(ls *.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.pon.filtered.vcf);
do
sample=${f%.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.pon.filtered.vcf}
echo $sample
bsub -q normal -o %J.out -e %J.err -n 4 -R 'select[mem>=80000] rusage[mem=80000] span[hosts=1]' -M80000 bash common_snp_filter_20220620.sh $f $sample
done

cd ../03_snp

for f in $(ls *vcf);
do
echo $f
grep -v "#" $f | wc -l
done

```

```{bash}

#common_snp_filter_20220620.sh
FILE=$1
SAMPLE=$2

bedtools intersect -header -v -a $FILE -b /lustre/scratch119/casm/team294rr/External_Databases/Variants/dbSNP_v151/All_20180423.common.vcf.gz > ../03_snp/${SAMPLE}.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.pon.commonSNP.filtered.vcf

```

###Blacklists

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data")

seg_dups = read.table("ucsc_segmental_duplications_grch37.tsv", header = T, sep = '\t', stringsAsFactors = F,comment.char = "")

seg_dups.lower = seg_dups[, 1:3]
seg_dups.upper = seg_dups[, 7:9]

names(seg_dups.lower) = names(seg_dups.upper) = c("chr", "start", "end")

seg_dups.all = rbind(seg_dups.lower, seg_dups.upper)
seg_dups.all$chr = unlist(strsplit(seg_dups.all$chr, "chr"))[c(F, T)]
seg_dups.all = seg_dups.all[seg_dups.all$chr %in% c(1:22, "X", "Y"),]

write.table(seg_dups.all, "ucsc_segmental_duplications_grch37.bed", col.names = F, row.names = F, sep = '\t', quote = F)

```


```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/03_snp

module load bedtools

for f in $(ls *.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.pon.commonSNP.filtered.vcf);
do
sample=${f%.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.pon.commonSNP.filtered.vcf}
#echo $sample
bedtools intersect -header -v -a $f -b /lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/LCR-hs37d5.bed /lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/hg19-blacklist.v2.chr.renamed.bed /lustre/scratch119/casm/team294rr/rs30/External_Databases/References/GRCh37d5/PAR_regions.bed /lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/ucsc_segmental_duplications_grch37.bed > ../04_blacklist/${sample}.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.ENCODE.LCR.PAR_region.blacklist.filtered.vcf
done

#cd ../04_blacklist

#for f in $(ls *vcf);
#do
#echo $f
#grep -v "#" $f | wc -l
#done

```

###Direction filter

At least one read supporting variant in both directions.

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/04_blacklist

module load bcftools

for f in $(ls *.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.ENCODE.LCR.PAR_region.blacklist.filtered.vcf);
do
filename=$(basename -- "$f")
sample="${filename%.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.ENCODE.LCR.PAR_region.blacklist.filtered.vcf}"
bcftools filter -e 'INFO/FS = 0 | INFO/RS = 0' $f > ../05_direction/${sample}.caveman_c.annot.asmd.clpm.pass.maternal.cord.depth.vaf.binom.recurrent_var.ENCODE.LCR.PAR_region.blacklist.direction.filtered.vcf
done

cd ../05_direction
for file in $(ls PD*.vcf);
do
echo $file
grep -v "#" $file | wc -l
done


```

###MQ filter

####cgpVAF MQ30

#####Symlink bam files to directory

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq30_cgpvaf

#infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt'
grep -v Placenta_PD $infile | cut -f1,3 > all_placenta_samples.txt
grep -v Placenta_PD $infile | cut -f11,14 | sort | uniq >> all_placenta_samples.txt

#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"
#awk -F "\t" '{ if($23 == "Yes") { print $1, $3} }' $infile > all_placenta_samples.txt
#awk -F "\t" '{ if($23 == "Yes") { print $9, $12} }' $infile | sort | uniq >> all_placenta_samples.txt

wc -l all_placenta_samples.txt #all 530

while read line;
do 
set $line
#echo $1
#echo $2
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam.bai
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam.bas
done < all_placenta_samples.txt

ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam
ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam.bai
ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam.bas

```

#####Generate bed files

Done per patient. Ignore warnings in PD421 cases regarding missing samples - run matched this time and those were the matched samples. Some samples weren't put through these pipeline.

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq30_cgpvaf

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt'
#infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"
#for cord_sample in $(awk -F "\t" '{ if($23 == "Yes") { print $9} }' $infile | uniq);
for cord_sample in $(grep -v Placenta_PD $infile | cut -f11 | uniq);
do

patient1=${cord_sample:0:7}
echo ${cord_sample} > ${patient1}_samples.txt

  for placenta_sample in $(grep ${cord_sample} $infile | awk -F "\t" '{ print $1 }');
  do
    echo ${placenta_sample} >> ${patient1}_samples.txt
  done

  patient2=${placenta_sample:0:7}
  mv ${patient1}_samples.txt ${patient2}_samples.txt

  while read line;
  do
    set $line
    grep -vh '#' /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/05_direction/${1}*.filtered.vcf | cut -f 1,2,4,5 >> ${patient2}_subs.bed 
  done < ${patient2}_samples.txt

less ${patient2}_subs.bed | sort -k1,1 -k2,2n -k3,4 | uniq > ${patient2}_subs_unique.bed
mv ${patient2}_subs_unique.bed ${patient2}_subs.bed

done

```

#####Run cgpvaf

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq30_cgpvaf

module load cgpVAFcommand

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt'
#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"

for f in $(grep -v 'Placenta_PD' $infile | cut -f1 | awk '{ $1 = substr($1, 1, 7) } 1' | sort | uniq);
do
  declare -a sampleArray=()
  #echo $1
  
  while read line;
  do
  set $line
  sampleArray+=($1)
    
  done < ${f}_samples.txt
  
  #echo "${sampleArray[@]}"
  
  mkdir -p ${f}
  
  bsub -q normal -o ${f}/%J.$f.out -e ${f}/%J.$f.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 perl /software/CASM/modules/installs/vafcorrect/shim-bin/cgpVaf.pl -d ./ -o $f -a snp -mq 30 -bq 25 -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be .sample.dupmarked.bam -hdr /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b ${f}_subs.bed -nn PDv37is -tn ${sampleArray[@]}
done

```

####cgpVAF MQ0

#####Symlink bam files to directory

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq0_cgpvaf

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt'
grep -v Placenta_PD $infile | cut -f1,3 > all_placenta_samples.txt
grep -v Placenta_PD $infile | cut -f11,14 | sort | uniq >> all_placenta_samples.txt

#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"
#awk -F "\t" '{ if($23 == "Yes") { print $1, $3} }' $infile > all_placenta_samples.txt
#awk -F "\t" '{ if($23 == "Yes") { print $9, $12} }' $infile | sort | uniq >> all_placenta_samples.txt

wc -l all_placenta_samples.txt #all 530

while read line;
do 
set $line
#echo $1
#echo $2
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam.bai
ln -sf /nfs/cancer_ref01/nst_links/live/${2}/${1}/${1}.sample.dupmarked.bam.bas
done < all_placenta_samples.txt

ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam
ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam.bai
ln -sf /nfs/cancer_ref01/nst_links/live/1548/PDv37is/PDv37is.sample.dupmarked.bam.bas

```

#####Generate bed files

Done per patient. Ignore warnings in PD421 cases regarding missing samples - run matched this time and those were the matched samples

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq0_cgpvaf

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt'
#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"
#for cord_sample in $(awk -F "\t" '{ if($23 == "Yes") { print $9} }' $infile | uniq);
for cord_sample in $(grep -v Placenta_PD $infile | cut -f11 | uniq);
do

patient1=${cord_sample:0:7}
echo ${cord_sample} > ${patient1}_samples.txt

  for placenta_sample in $(grep ${cord_sample} $infile | awk -F "\t" '{ print $1 }');
  do
    echo ${placenta_sample} >> ${patient1}_samples.txt
  done

  patient2=${placenta_sample:0:7}
  mv ${patient1}_samples.txt ${patient2}_samples.txt

  while read line;
  do
    set $line
    grep -vh '#' /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/05_direction/${1}*.filtered.vcf | cut -f 1,2,4,5 >> ${patient2}_subs.bed 
  done < ${patient2}_samples.txt

less ${patient2}_subs.bed | sort -k1,1 -k2,2n -k3,4 | uniq > ${patient2}_subs_unique.bed
mv ${patient2}_subs_unique.bed ${patient2}_subs.bed

done

```

#####Run cgpvaf

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq0_cgpvaf

module load cgpVAFcommand

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt'
#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"

for f in $(grep -v 'Placenta_PD' $infile | cut -f1 | awk '{ $1 = substr($1, 1, 7) } 1' | sort | uniq);
do
  declare -a sampleArray=()
  #echo $1
  
  while read line;
  do
  set $line
  sampleArray+=($1)
    
  done < ${f}_samples.txt
  
  #echo "${sampleArray[@]}"
  
  mkdir -p ${f}
  
  bsub -q normal -o ${f}/%J.$f.out -e ${f}/%J.$f.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 perl /software/CASM/modules/installs/vafcorrect/shim-bin/cgpVaf.pl -d ./ -o $f -a snp -mq 0 -bq 25 -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be .sample.dupmarked.bam -hdr /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b ${f}_subs.bed -nn PDv37is -tn ${sampleArray[@]}
done

```

####MQ filter applied

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq30_filter

cp ../mq30_cgpvaf/PD*/*_snp_vaf.tsv .

#remove header info
for f in $(ls *_snp_vaf.tsv);
do
filename=$(basename -- "$f")
filename="${filename%_snp_vaf.tsv}"
#echo $filename
grep -v "##" $f > ${filename}_snp_vaf_nohash.tsv
done

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq0_filter

cp ../mq0_cgpvaf/PD*/*_snp_vaf.tsv .

#remove header info
for f in $(ls *_snp_vaf.tsv);
do
filename=$(basename -- "$f")
filename="${filename%_snp_vaf.tsv}"
#echo $filename
grep -v "##" $f > ${filename}_snp_vaf_nohash.tsv
done

```

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq_filter")

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)
contam_samples = c(manifest[manifest$Biopsy_freemix >= 0.01 & !is.na(manifest$Biopsy_freemix),]$Placenta_PD, unique(manifest[manifest$Cord_freemix >= 0.01 & !is.na(manifest$Cord_freemix),]$Cord_PD)) #list of samples with over 1% contamination, likely maternal

mq30_files = list.files("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq30_filter", "_snp_vaf_nohash.tsv")
mq0_files = list.files("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq0_filter", "_snp_vaf_nohash.tsv") #same name

svs = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/Placenta_AS_filtered_Mar8_2022_Annotated.txt", header = T, sep = '\t', stringsAsFactors = F)
svs_flt = svs[svs$PASS_FAIL == "PASS" & svs$bkdist > 1e5 & svs$file_sample %in% unique(c(manifest$Placenta_PD, manifest$Cord_PD)), ] #108 breakpoints

ascatPCA = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/ascatPCA/Placenta_CNV_ascatPCA_20220322_additional_samples_20220513_copy_number.txt", header = T, sep = '\t') 
ascatPCA_flt = ascatPCA[ascatPCA$length > 1e5 & ascatPCA$sample %in% unique(c(manifest$Placenta_PD, manifest$Cord_PD)),] #62 abnormal segments

battenberg = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/Battenberg/placenta_cnas_20220324_reviewed_20220426_additional_20220511.txt", header = T, sep = '\t')
battenberg_flt = battenberg[battenberg$consensus_review == "Yes" & battenberg$length > 1e5 & battenberg$sample %in% unique(c(manifest$Placenta_PD, manifest$Cord_PD)),] #44 abnormal segments

samples_with_CNVs = sort(unique(c(svs_flt$file_sample, ascatPCA_flt$sample, battenberg_flt$sample))) #copy number changes over 100kb

#txt_file = "PDv37is_PD45568f_snp_vaf_nohash.tsv"
#txt_file = "PDv37is_PD51913b_snp_vaf_nohash.tsv"
#txt_file = "PDv37is_PD42149b_snp_vaf_nohash.tsv"
for(txt_file in mq30_files){
  
  #mq30 bq25
  data = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq30_filter/", txt_file), comment.char = "", header = T)
  patient = substr(unlist(strsplit(txt_file, 'PDv37is_'))[2], 0, 7)
  cord = unique(manifest[grepl(patient, manifest$Cord_PD), ]$Cord_PD)

  #extract data frames of interest
  Muts = paste(data$Chrom,data$Pos,data$Ref,data$Alt,sep="_")
  Genotype = data[,grepl("VAF",colnames(data))&colnames(data)!="PDv37is_VAF"]
  NR = data[,grepl("DEP",colnames(data))&colnames(data)!="PDv37is_DEP"]
  NV = data[,grepl("MTR",colnames(data))&colnames(data)!="PDv37is_MTR"]
  colnames(Genotype)=colnames(NR)=colnames(NV)=gsub("_VAF","",colnames(Genotype))
  rownames(Genotype)=rownames(NV)=rownames(NR)=Muts
  
  NR_mq30 = NR
  NV_mq30 = NV
  
  NR_mq30_nonzero = NR_mq30
  NR_mq30_nonzero[NR_mq30_nonzero == 0] = 1
  
  #list of samples
  samples = colnames(NR)

  #samples without a copy number change and <1% maternal contaminaton
  noCNVs=!samples %in% unique(c(contam_samples, samples_with_CNVs))
  
  #mq0 bq25
  data = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq0_filter/", txt_file), comment.char = "", header = T)

  #extract data frames of interest
  Muts = paste(data$Chrom,data$Pos,data$Ref,data$Alt,sep="_")
  Genotype = data[,grepl("VAF",colnames(data))&colnames(data)!="PDv37is_VAF"]
  NR = data[,grepl("DEP",colnames(data))&colnames(data)!="PDv37is_DEP"]
  NV = data[,grepl("MTR",colnames(data))&colnames(data)!="PDv37is_MTR"]
  colnames(Genotype)=colnames(NR)=colnames(NV)=gsub("_VAF","",colnames(Genotype))
  rownames(Genotype)=rownames(NV)=rownames(NR)=Muts
  
  NR_mq0 = NR
  NV_mq0 = NV
  
  #mq filter
  low_mq_loci = row.names(NR_mq30)[rowMeans(NR_mq30 / NR_mq0) < 0.9]
  good_mq_loci = row.names(NR_mq30)[!row.names(NR_mq30) %in% low_mq_loci]
  
  write.table(good_mq_loci, paste0(patient, "_good_mq_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  write.table(low_mq_loci, paste0(patient, "_low_mq_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  #write.table((NV_mq30 / NR_mq30_nonzero)[good_mq_loci,], paste0(patient, "_good_mq_loci_vaf.txt"), sep = '\t', quote = F, col.names = T, row.names = T)
  
  #germline dup site SNP removal
  NR_mq30_flt = NR_mq30[good_mq_loci,]
  NV_mq30_flt = NV_mq30[good_mq_loci,]
  
  gender = tolower(unique(manifest[manifest$Cord_PD == cord,]$Sex))
  #cord_coverage = unique(manifest[manifest$Cord_PD == cord,]$cord_coverage)
  cord_coverage = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/germline_filter/", patient, "_cord_median_germline_auto_snp_depth.txt"), sep = '\t', header = F, stringsAsFactors = F)[,1]
  
  if(gender == "female"){
    NR_mq30_flt=NR_mq30_flt[!grepl("Y", row.names(NR_mq30_flt)),]
    NV_mq30_flt=NV_mq30_flt[!grepl("Y", row.names(NV_mq30_flt)),]
  }
  
  NR_mq30_flt$cord_rdr = NA
  
  for(i in 1:nrow(NR_mq30_flt)){
    chr = unlist(lapply(strsplit(row.names(NR_mq30_flt)[i], "_"), "[[", 1))
    if(gender == "female"){
      NR_mq30_flt$cord_rdr[i] = NR_mq30_flt[i, cord] / cord_coverage
    }
    if(gender == "male" & chr %in% c(1:22)){
      NR_mq30_flt$cord_rdr[i] = NR_mq30_flt[i, cord] / cord_coverage
    }
    if(gender == "male" & chr %in% c("X", "Y")){
      NR_mq30_flt$cord_rdr[i] = NR_mq30_flt[i, cord] / (cord_coverage / 2)
    }
  }
  
  NR_mq30_flt$exp_cn = 2
  
  if(gender == "male" & nrow(NR_mq30_flt[grepl("X|Y", row.names(NR_mq30_flt)),]) > 0) NR_mq30_flt[grepl("X|Y", row.names(NR_mq30_flt)),]$exp_cn = 1
  NR_mq30_flt$seg_copies = round(NR_mq30_flt$cord_rdr * NR_mq30_flt$exp_cn) #estimate the total germline copies of locus
  
  min_depth_var = row.names(NR_mq30_flt)[NR_mq30_flt$seg_copies >= 1] #remove remaining v. low coverage areas
  
  NV_mq30_flt_2 = NV_mq30_flt[min_depth_var, ]
  NR_mq30_flt_2 = NR_mq30_flt[min_depth_var, ]
  
  #apply exact binomial test to assess whether variants at likely germline duplication sites are germline or not
  pval_vec = rep(NA, nrow(NV_mq30_flt_2))
  names(pval_vec) = row.names(NV_mq30_flt_2)
  
  NV_vec = rowSums(NV_mq30_flt_2[, grepl("PD", colnames(NV_mq30_flt_2))][, noCNVs]) #no somatic copy number or contamination
  NR_vec = rowSums(NR_mq30_flt_2[, grepl("PD", colnames(NR_mq30_flt_2))][, noCNVs])
  for (n in 1:length(pval_vec)){
    if(NR_vec[n]>0){
      pval_vec[n] = binom.test(x = NV_vec[n],
                                n = NR_vec[n],
                                p = 1 / NR_mq30_flt_2$seg_copies[n], alt = 'less')$p.value
  
    }
  }
  
  qval = p.adjust(pval_vec,method="BH")
  germline_cn = log10(qval) > -3
  
  write.table(row.names(NR_mq30_flt_2)[germline_cn], paste0(patient, "_germline_dup_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  write.table(row.names(NR_mq30_flt_2)[!germline_cn], paste0(patient, "_somatic_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  
  NR_mq30_flt_2_nonzero = NR_mq30_flt_2[, samples]
  NR_mq30_flt_2_nonzero[NR_mq30_flt_2_nonzero == 0] = 1
  VAF_final = (NV_mq30_flt_2 / NR_mq30_flt_2_nonzero)[!germline_cn,]
  VAF_final_sort = VAF_final[order(VAF_final[,cord], decreasing = T),]
  VAF_final_shared = VAF_final[rowSums(VAF_final[, noCNVs] > 0) == sum(noCNVs),]
  
  write.table(VAF_final, paste0(patient, "_final_VAF.txt"), col.names = T, sep = '\t', row.names = T, quote = F)
  write.table(VAF_final_sort, paste0(patient, "_final_VAF_cord_sorted.txt"), col.names = T, sep = '\t', row.names = T, quote = F)
  write.table(VAF_final_shared, paste0(patient, "_final_VAF_shared.txt"), col.names = T, sep = '\t', row.names = T, quote = F)
  
}

```

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq_filter")

source("/nfs/users/nfs_t/to3/scripts/beta_binom_flt.R")

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)
contam_samples = c(manifest[manifest$Biopsy_freemix >= 0.01 & !is.na(manifest$Biopsy_freemix),]$Placenta_PD, unique(manifest[manifest$Cord_freemix >= 0.01 & !is.na(manifest$Cord_freemix),]$Cord_PD)) #list of samples with over 1% contamination, likely maternal

mq30_files = list.files("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq30_filter", "_snp_vaf_nohash.tsv")
mq0_files = list.files("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq0_filter", "_snp_vaf_nohash.tsv") #same name

#txt_file = "PDv37is_PD45568f_snp_vaf_nohash.tsv"
#txt_file = "PDv37is_PD51913b_snp_vaf_nohash.tsv"
#txt_file = "PDv37is_PD42153b_snp_vaf_nohash.tsv"
for(txt_file in mq30_files){
  
  #mq30 bq25
  data = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq30_filter/", txt_file), comment.char = "", header = T)
  patient = substr(unlist(strsplit(txt_file, 'PDv37is_'))[2], 0, 7)
  cord = unique(manifest[grepl(patient, manifest$Cord_PD), ]$Cord_PD)

  #extract data frames of interest
  Muts = paste(data$Chrom,data$Pos,data$Ref,data$Alt,sep="_")
  Genotype = data[,grepl("VAF",colnames(data))&colnames(data)!="PDv37is_VAF"]
  NR = data[,grepl("DEP",colnames(data))&colnames(data)!="PDv37is_DEP"]
  NV = data[,grepl("MTR",colnames(data))&colnames(data)!="PDv37is_MTR"]
  colnames(Genotype)=colnames(NR)=colnames(NV)=gsub("_VAF","",colnames(Genotype))
  rownames(Genotype)=rownames(NV)=rownames(NR)=Muts
  
  NR_mq30 = NR
  NV_mq30 = NV
  
  NR_mq30_nonzero = NR_mq30
  NR_mq30_nonzero[NR_mq30_nonzero == 0] = 1
  
  #mq0 bq25
  data = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq0_filter/", txt_file), comment.char = "", header = T)

  #extract data frames of interest
  Muts = paste(data$Chrom,data$Pos,data$Ref,data$Alt,sep="_")
  Genotype = data[,grepl("VAF",colnames(data))&colnames(data)!="PDv37is_VAF"]
  NR = data[,grepl("DEP",colnames(data))&colnames(data)!="PDv37is_DEP"]
  NV = data[,grepl("MTR",colnames(data))&colnames(data)!="PDv37is_MTR"]
  colnames(Genotype)=colnames(NR)=colnames(NV)=gsub("_VAF","",colnames(Genotype))
  rownames(Genotype)=rownames(NV)=rownames(NR)=Muts
  
  NR_mq0 = NR
  NV_mq0 = NV
  
  #mq filter
  low_mq_loci = row.names(NR_mq30)[rowMeans(NR_mq30 / NR_mq0) < 0.95] #changed from 0.9 28/6/22
  good_mq_loci = row.names(NR_mq30)[!row.names(NR_mq30) %in% low_mq_loci]
  
  write.table(good_mq_loci, paste0(patient, "_good_mq_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  write.table(low_mq_loci, paste0(patient, "_low_mq_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  #write.table((NV_mq30 / NR_mq30_nonzero)[good_mq_loci,], paste0(patient, "_good_mq_loci_vaf.txt"), sep = '\t', quote = F, col.names = T, row.names = T)
  
  #germline dup site SNP removal
  NR_mq30_flt = NR_mq30[good_mq_loci,]
  NV_mq30_flt = NV_mq30[good_mq_loci,]
  
  gender = tolower(unique(manifest[manifest$Cord_PD == cord,]$Sex))
  #cord_coverage = unique(manifest[manifest$Cord_PD == cord,]$cord_coverage)
  cord_auto_coverage = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/germline_filter/", patient, "_cord_median_germline_auto_snp_depth.txt"), sep = '\t', header = F, stringsAsFactors = F)[,1]
  cord_sex_coverage = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/germline_filter/", patient, "_cord_median_germline_sex_snp_depth.txt"), sep = '\t', header = F, stringsAsFactors = F)[,1]
  
  if(gender == "female"){
    NR_mq30_flt=NR_mq30_flt[!grepl("Y", row.names(NR_mq30_flt)),]
    NV_mq30_flt=NV_mq30_flt[!grepl("Y", row.names(NV_mq30_flt)),]
  }
  
  
  
  pois_pval = rep(NA, nrow(NR_mq30_flt))
  for(i in 1:length(pois_pval)){
    if(!grepl("X|Y", row.names(NR_mq30_flt)[i]) | gender == "female"){ #autosomal | female
      pois_pval[i] = ppois(q = NR_mq30_flt[i, cord] - 1, lambda = cord_auto_coverage, lower.tail = F) #- 1 so it can be the depth achieved or high
    }
    if(grepl("X|Y", row.names(NR_mq30_flt)[i])){ #male sex chromsome
      pois_pval[i] = ppois(q = NR_mq30_flt[i, cord] - 1, lambda = cord_sex_coverage, lower.tail = F) #correct for lower coverage
    }
    
  }
  pois_qval = p.adjust(pois_pval, method = "BH")
  
  norm_cov = log10(pois_qval) > -3
  #temp = (log10(pois_qval) > -5 & log10(pois_qval) < -3)
  #(NV_mq30_flt/NR_mq30_flt)[temp,]
  
  write.table(row.names(NR_mq30_flt)[!norm_cov], paste0(patient, "_germline_dup_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  write.table(row.names(NR_mq30_flt)[norm_cov], paste0(patient, "_somatic_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  
  NR_mq30_flt_2 = NR_mq30_flt[norm_cov,]
  NV_mq30_flt_2 = NV_mq30_flt[norm_cov,]
  
  NR_mq30_flt_2_nonzero = NR_mq30_flt_2
  NR_mq30_flt_2_nonzero[NR_mq30_flt_2_nonzero == 0] = 1
  
  #beta binom annotation
  shared_muts = row.names(NV_mq30_flt_2[rowSums(NV_mq30_flt_2[, !colnames(NV_mq30_flt_2) %in% contam_samples] > 0) >= 2,]) #at least one read in two or more samples that are not contaminated
  
  #beta binomial fitting - exclude contaminated samples
  rho_est = beta.binom.filter(NR = NR_mq30_flt_2_nonzero[shared_muts, !colnames(NR_mq30_flt_2_nonzero) %in% contam_samples], NV = NV_mq30_flt_2[shared_muts, !colnames(NV_mq30_flt_2) %in% contam_samples])

  flt_rho = rho_est < 0.1

  #flt_rho = log10(rho_est) < (-1)
  rho_filtered_out = rownames(NR_mq30_flt_2_nonzero[shared_muts,])[flt_rho]

  write.table(rho_filtered_out, paste0(patient, "_bbinom_filtered_snv.txt"))
  write.table(rho_est, paste0(patient, "_rho_est_shared_snv.txt"))

  #filtered data frames
  NR_mq30_flt_3 = NR_mq30_flt_2
  NV_mq30_flt_3 = NV_mq30_flt_2
  
  NR_mq30_flt_3$beta.binom = NV_mq30_flt_3$beta.binom = "FAIL"
  
  NR_mq30_flt_3[!rownames(NR_mq30_flt_3) %in% rho_filtered_out, ]$beta.binom = "PASS"
  NV_mq30_flt_3[!rownames(NV_mq30_flt_3) %in% rho_filtered_out, ]$beta.binom = "PASS"
  
  VAF_mq30_flt_3 = NV_mq30_flt_3
  
  VAF_mq30_flt_3[, 1:(ncol(NV_mq30_flt_3) - 1)] = VAF_mq30_flt_3[, 1:(ncol(NV_mq30_flt_3) - 1)] / NR_mq30_flt_3[, 1:(ncol(NR_mq30_flt_3) - 1)]
  
  VAF_final_sort = VAF_mq30_flt_3[order(VAF_mq30_flt_3[,cord], decreasing = T),]

  write.table(VAF_mq30_flt_3, paste0(patient, "_final_VAF.txt"), col.names = T, sep = '\t', row.names = T, quote = F)
  write.table(VAF_final_sort, paste0(patient, "_final_VAF_cord_sorted.txt"), col.names = T, sep = '\t', row.names = T, quote = F)
  
}

```

###Annotate VAF matrices with CN status 

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq_filter")

#read in metadata
manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)

svs = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/Placenta_AS_filtered_Mar8_2022_Annotated.txt", header = T, sep = '\t', stringsAsFactors = F)
svs_flt = svs[svs$PASS_FAIL == "PASS" & svs$bkdist > 0 & svs$file_sample %in% unique(c(manifest$Placenta_PD, manifest$Cord_PD)), ] #328 breakpoints

ascatPCA = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/ascatPCA/Placenta_CNV_ascatPCA_20220322_additional_samples_20220513_copy_number.txt", header = T, sep = '\t') 
ascatPCA_flt = ascatPCA[ascatPCA$sample %in% unique(c(manifest$Placenta_PD, manifest$Cord_PD)),] #101 abnormal segments

battenberg = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/Battenberg/placenta_cnas_20220324_reviewed_20220426_additional_20220511.txt", header = T, sep = '\t')
battenberg_flt = battenberg[battenberg$consensus_review == "Yes" & battenberg$sample %in% unique(c(manifest$Placenta_PD, manifest$Cord_PD)),] #67 abnormal segments

#concatenate cna region calls
svs_flt_coord = svs_flt[, c(3,4,8, 50)]
ascatPCA_flt_coord = ascatPCA_flt[, c(1:3, 6)]
battenberg_flt_coord = battenberg_flt[, c(1:3, 6)]
names(svs_flt_coord) = names(ascatPCA_flt_coord) = names(battenberg_flt_coord) = c("chr", "start", "end", "sample")
placenta_cna = rbind(svs_flt_coord, ascatPCA_flt_coord, battenberg_flt_coord)

#read in VAF matrices
vaf_files = list.files(".", "_final_VAF_cord_sorted.txt")

for(file in vaf_files){
  patient = unlist(strsplit(file, "_final_VAF_cord_sorted.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  samples = colnames(data)
  data$match_cna = F
  mut_df = data.frame(matrix(unlist(strsplit(row.names(data), "_")), ncol = 4, byrow = T))
  mut_df$X2 = as.numeric(mut_df$X2)
  
  for(i in 1:nrow(mut_df)){
    if(nrow(placenta_cna[placenta_cna$chr == mut_df$X1[i] & 
                         placenta_cna$start <= mut_df$X2[i] & 
                         placenta_cna$end >= mut_df$X2[i] & 
                         placenta_cna$sample %in% samples,]) > 0){
      data$match_cna[i] = T
                         }
  }
  
  print(data[data$match_cna == T,])

  write.table(data, paste0(patient, "_final_VAF_cord_sorted_cna_annot.txt"), col.names = T, row.names = T, sep = '\t', quote = F)
  
}

```

###Add complete annotations from old run

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq_filter")

new_files = list.files(".", "_final_VAF_cord_sorted_cna_annot.txt")

#file = "PD42145_final_VAF_cord_sorted_cna_annot.txt"
for(file in new_files){
  
  patient = unlist(strsplit(file, "_final_VAF_cord_sorted_cna_annot.txt"))

  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  data$Keep = NA
  data$Notes = NA
  
  if(file.exists(paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq_filter/old_20220622_2/rough_annot/", patient, "_final_VAF_cord_sorted_cna_annot.txt"))){
      
    old_data = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/unmatched_cord/mq_filter/old_20220622_2/rough_annot/", patient, "_final_VAF_cord_sorted_cna_annot.txt"), header = T, sep = '\t', stringsAsFactors = F, quote = "", row.names = 1)
    
    for(i in 1:nrow(data)){
      
      if(row.names(data)[i] %in% row.names(old_data)){
        data[i,]$Keep = old_data[row.names(data)[i],]$Keep
        data[i,]$Notes = old_data[row.names(data)[i],]$Notes
      }
      
    }
    
    
  }

  
  write.table(data, file, sep = '\t', quote = F, col.names = T, row.names = T)
  
}

```

##Indels - matched

Initial calls generated using pindel

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/01_pindel

#placental biopsies
infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
grep -v "Placenta_PD" $infile | awk -F "\t" '{ { print $1, $3} }' | while read -r -a tmp;
do
cp /nfs/cancer_ref01/nst_links/live/${tmp[1]}/${tmp[0]}/${tmp[0]}.pindel.annot.vcf.gz .
done


cp /nfs/cancer_ref01/nst_links/live/2633/PD51879c/PD51879c.pindel.annot.vcf.gz .
#PD51879c missing 27/5/22

```

###Homopolymer & PASS filtering

Filter original pindel files to only include those that pass pindel and are not 1bp events occurring on homopolymer runs 9bp or greater.

```{bash}

bsub -Is -q yesterday -R "select[mem>100000] rusage[mem=100000] span[hosts=1]" -M 100000 bash

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/01_pindel

for file in $(ls *.pindel.annot.vcf.gz);
do
gunzip $file
done

for f in $(ls *.pindel.annot.vcf);
do
less $f | grep -m1 "TUMOUR" >> matched_sample_list.txt
less $f | grep -m1 "NORMAL" >> matched_sample_list.txt
done

grep "PDv37is" matched_sample_list.txt | wc -l #0, no run unmatched, 27/5/22
ls *.pindel.annot.vcf | wc -l #416, none missing, 27/5/22

#Check that none of the vcfs have been truncated and are missing chromosomes
ls *.pindel.annot.vcf | while read FILE ; do grep "^#" -v "$FILE" | cut -f1 | uniq -c > ../qc_chrom_counts/"$FILE".counts.txt ; done 
wc -l ../qc_chrom_counts/*.counts.txt #none truncated 27/5/22

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/01_pindel

declare -a sampleArray=()

for f in $(ls *.pindel.annot.vcf);
do
  filename=$(basename -- "$f")
  sample="${filename%.pindel.annot.vcf}"
  #echo $sample
  sampleArray+=($sample)
done

#echo ${sampleArray[@]}

cd /nfs/users/nfs_t/to3/scripts/pindel-filtering

#standard
for f in ${sampleArray[@]};
do
echo $f
./bsub_homopolymer_filter_args_lustre.sh /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/01_pindel 9 $f PASS
done

mv /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/01_pindel/*.pindel.HP.9.vcf /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/02_hp9

```

###Quality filtering

Standard filters to all samples

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/02_hp9

module load bcftools

for f in $(ls *.pindel.HP.9.vcf);
do
filename=$(basename -- "$f")
sample="${filename%.pindel.HP.9.vcf}"
bcftools filter -i 'QUAL >= 300' $f > ../03_qual/${sample}.pindel.annot.qual.pass.HP.9.filtered.vcf
done

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/03_qual

for f in $(ls *.pindel.annot.qual.pass.HP.9.filtered.vcf);
do
filename=$(basename -- "$f")
sample="${filename%.pindel.annot.qual.pass.HP.9.filtered.vcf}"
echo $sample
grep -v "#" $f | wc -l
done

```

###Maternal indel filtering

Remove any indel called in mother to eliminate any maternal contamination / germline, where maternal data is available

Remove indel sites in placenta

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/03_qual

module load bedtools

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt'
#infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt'
for f in $(ls PD*.pindel.annot.qual.pass.HP.9.filtered.vcf);
do
sample=${f%.pindel.annot.qual.pass.HP.9.filtered.vcf}
maternal=$(grep -w $sample $infile | cut -f15 | uniq)
if [[ "$maternal" != "-" ]]; 
then 
bedtools intersect -header -v -a $f -b /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/maternal_indels/${maternal}.pindel.annot.vcf.gz > ../04_maternal/${sample}.pindel.annot.qual.pass.HP.9.maternal.filtered.vcf
fi
done

#no maternal sample for the following:
#PD51883b
#PD51883c
#PD51883d
#PD51883e
#PD51883f
#PD51890b
#PD51890c
#PD51890d
#PD51890e
#PD51890f
#PD51920b
#PD51920c
#PD51920d
#PD51920e
#PD51920f

for f in $(ls PD*.pindel.annot.qual.pass.HP.9.filtered.vcf);
do
sample=${f%.pindel.annot.qual.pass.HP.9.filtered.vcf}
maternal=$(grep -w $sample $infile | cut -f15 | uniq)
if [[ "$maternal" == "-" ]]; 
then 
cp $f ../04_maternal
fi
done

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/04_maternal
for f in $(ls *vcf);
do
echo $f
grep -v "#" $f | wc -l
done

```

###Blacklists

Remove variants called in hg19 blacklist sites.

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/04_maternal

module load bedtools

for f in $(ls *.filtered.vcf);
do
sample=${f%.filtered.vcf}
#echo $sample
bedtools intersect -header -v -a $f -b /lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/LCR-hs37d5.bed /lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/hg19-blacklist.v2.chr.renamed.bed > ../05_blacklist/${sample}.ENCODE.LCR.blacklist.filtered.vcf
done

for f in $(ls *vcf);
do
echo $f
grep -v "#" $f | wc -l
done

```

###cgpVAF MQ30

####Generate bed files

Done per patient. Ignore warnings in PD421 cases regarding missing samples - run matched this time and those were the matched samples

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq30_cgpvaf

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt'
#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"
#for cord_sample in $(awk -F "\t" '{ if($23 == "Yes") { print $9} }' $infile | uniq);
for cord_sample in $(grep -v Placenta_PD $infile | cut -f11 | uniq);
do

patient1=${cord_sample:0:7}
echo ${cord_sample} > ${patient1}_samples.txt

  for placenta_sample in $(grep ${cord_sample} $infile | awk -F "\t" '{ print $1 }');
  do
    echo ${placenta_sample} >> ${patient1}_samples.txt
  done

  patient2=${placenta_sample:0:7}
  mv ${patient1}_samples.txt ${patient2}_samples.txt

  while read line;
  do
    set $line
    grep -vh '#' /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/05_blacklist/${1}*.filtered.vcf | cut -f 1,2,4,5 >> ${patient2}_indels.bed 
  done < ${patient2}_samples.txt

less ${patient2}_indels.bed | sort -k1,1 -k2,2n -k3,4 | uniq > ${patient2}_indels_unique.bed
mv ${patient2}_indels_unique.bed ${patient2}_indels.bed

done

```

####Run cgpvaf

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq30_cgpvaf

module load cgpVAFcommand

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt'
#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"

for f in $(grep -v 'Placenta_PD' $infile | cut -f1 | awk '{ $1 = substr($1, 1, 7) } 1' | sort | uniq);
do
  declare -a sampleArray=()
  #echo $1
  
  while read line;
  do
  set $line
  sampleArray+=($1)
    
  done < ${f}_samples.txt
  
  #echo "${sampleArray[@]}"
  
  #mkdir -p ${f}
  
  bsub -q normal -o ${f}/%J.$f.out -e ${f}/%J.$f.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 perl /software/CASM/modules/installs/vafcorrect/shim-bin/cgpVaf.pl -d ./ -o $f -a indel -mq 30 -bq 25 -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be .sample.dupmarked.bam -hdr /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b ${f}_indels.bed -nn PDv37is -tn ${sampleArray[@]}
done

```

###cgpVAF MQ0

####Generate bed files

Done per patient. Ignore warnings in PD421 cases regarding missing samples - run matched this time and those were the matched samples

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq0_cgpvaf

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt'
#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"
#for cord_sample in $(awk -F "\t" '{ if($23 == "Yes") { print $9} }' $infile | uniq);
for cord_sample in $(grep -v Placenta_PD $infile | cut -f11 | uniq);
do

patient1=${cord_sample:0:7}
echo ${cord_sample} > ${patient1}_samples.txt

  for placenta_sample in $(grep ${cord_sample} $infile | awk -F "\t" '{ print $1 }');
  do
    echo ${placenta_sample} >> ${patient1}_samples.txt
  done

  patient2=${placenta_sample:0:7}
  mv ${patient1}_samples.txt ${patient2}_samples.txt

  while read line;
  do
    set $line
    grep -vh '#' /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/05_blacklist/${1}*.filtered.vcf | cut -f 1,2,4,5 >> ${patient2}_indels.bed 
  done < ${patient2}_samples.txt

less ${patient2}_indels.bed | sort -k1,1 -k2,2n -k3,4 | uniq > ${patient2}_indels_unique.bed
mv ${patient2}_indels_unique.bed ${patient2}_indels.bed

done

```

####Run cgpvaf

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq0_cgpvaf

module load cgpVAFcommand

infile='/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt'
#infile="/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt"

for f in $(grep -v 'Placenta_PD' $infile | cut -f1 | awk '{ $1 = substr($1, 1, 7) } 1' | sort | uniq);
do
  declare -a sampleArray=()
  #echo $1
  
  while read line;
  do
  set $line
  sampleArray+=($1)
    
  done < ${f}_samples.txt
  
  #echo "${sampleArray[@]}"
  
  #mkdir -p ${f}
  
  bsub -q normal -o ${f}/%J.$f.out -e ${f}/%J.$f.err -n 4 -R 'select[mem>=20000] rusage[mem=20000] span[hosts=1]' -M20000 perl /software/CASM/modules/installs/vafcorrect/shim-bin/cgpVaf.pl -d ./ -o $f -a indel -mq 0 -bq 25 -g /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa -be .sample.dupmarked.bam -hdr /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/shared/ucscHiDepth_0.01_mrg1000_no_exon_coreChrs.bed.gz -bo 1 -b ${f}_indels.bed -nn PDv37is -tn ${sampleArray[@]}
done

```

###Depth + Amb/Unk filter

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq30_filter

cp ../mq30_cgpvaf/PD*/*_indel_vaf.tsv .

#remove header info
for f in $(ls *_indel_vaf.tsv);
do
filename=$(basename -- "$f")
filename="${filename%_indel_vaf.tsv}"
#echo $filename
grep -v "##" $f > ${filename}_indel_vaf_nohash.tsv
done

```

####Run

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/bq_mq_filter")

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)

mq30_files = list.files("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq30_filter", "_indel_vaf_nohash.tsv")
#mq0_files = list.files("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq0_filter", "_snp_vaf_nohash.tsv") #same name

#txt_file = "PDv37is_PD51883b_indel_vaf_nohash.tsv"
for(txt_file in mq30_files){
  
  #mq30 bq25
  data = read.table(paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/mq30_filter/", txt_file), comment.char = "", header = T)
  patient = substr(unlist(strsplit(txt_file, 'PDv37is_'))[2], 0, 7)
  cord = unique(manifest[grepl(patient, manifest$Cord_PD), ]$Cord_PD)

  #extract data frames of interest
  Muts = paste(data$Chrom,data$Pos,data$Ref,data$Alt,sep="_")
  Genotype = data[,grepl("VAF",colnames(data))&colnames(data)!="PDv37is_VAF"]
  NR = data[,grepl("DEP",colnames(data))&colnames(data)!="PDv37is_DEP"]
  NV = data[,grepl("MTR",colnames(data))&colnames(data)!="PDv37is_MTR"]
  amb = data[,grepl("AMB",colnames(data))&colnames(data)!="PDv37is_AMB"]
  unk = data[,grepl("UNK",colnames(data))&colnames(data)!="PDv37is_UNK"]
  
  colnames(Genotype)=colnames(NR)=colnames(NV)=gsub("_VAF","",colnames(Genotype))
  rownames(Genotype)=rownames(NV)=rownames(NR)=Muts
  colnames(amb)=colnames(unk)=colnames(Genotype)
  rownames(amb)=rownames(unk)=rownames(Genotype)
  
  NR_nonzero = NR
  NR_nonzero[NR_nonzero == 0] = 1
  
  NR_mq30 = NR[, colnames(NR) != cord]
  NV_mq30 = NV[, colnames(NV) != cord]
  
  #bq filter
  low_bq_loci = row.names(NV_mq30[rowSums(NV_mq30 > 4) == 0, ]) #at least 5 supporting reads in at least one sample needed
  good_bq_loci = row.names(NV_mq30)[!row.names(NV_mq30) %in% low_bq_loci]
  
  write.table(low_bq_loci, paste0(patient, "_low_indel_depth_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  write.table(good_bq_loci, paste0(patient, "_good_indel_depth_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  
  NR_mq30_flt = NR_mq30[good_bq_loci,]
  NV_mq30_flt = NV_mq30[good_bq_loci,]
  
  NR_mq30_flt_nonzero = NR_mq30_flt
  NR_mq30_flt_nonzero[NR_mq30_flt_nonzero == 0] = 1
  
  vaf_mq30_flt = NV_mq30_flt / NR_mq30_flt_nonzero
  
  #mq filter
  bad_reads_vaf = (amb[row.names(NR_mq30_flt), colnames(NR_mq30_flt)] + unk[row.names(NR_mq30_flt), colnames(NR_mq30_flt)]) / NR_mq30_flt_nonzero[row.names(NR_mq30_flt),] #estimate fraction of reads at locus that are unknown or ambiguous status
  bad_sites = row.names(bad_reads_vaf[rowMeans(bad_reads_vaf) > 0.5,]) #where more than 50% of the reads are ambiguous or unknown
  good_sites = row.names(bad_reads_vaf)[!row.names(bad_reads_vaf) %in% bad_sites]
  
  good_sites_flt = good_sites[(NV / NR_nonzero)[good_sites, cord] < 0.2]
  
  write.table(bad_sites, paste0(patient, "_poor_amb_unk_indel_depth_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  write.table(good_sites_flt, paste0(patient, "_good_amb_unk_indel_depth_loci.txt"), sep = '\t', quote = F, col.names = F, row.names = F)
  
}

```

####Intersect final calls with the pindel files

```{r}

setwd('/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/05_blacklist')

library(VariantAnnotation)

patients = unlist(strsplit(list.files("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/bq_mq_filter", "_good_amb_unk_indel_depth_loci.txt"), "_good_amb_unk_indel_depth_loci.txt"))
all_vcfs = list.files(path = '.', '.filtered.vcf')
manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)

for(patient in patients){
  filtered_muts = read.table(paste0('/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/bq_mq_filter/', patient, '_good_amb_unk_indel_depth_loci.txt'), header = F, sep = '\t', stringsAsFactors = F)[,1]
  patient_vcfs = all_vcfs[grepl(paste0(unique(substr(manifest[grepl(patient, manifest$Cord_PD),]$Cord_PD, 0, 7)), "|", unique(substr(manifest[grepl(patient, manifest$Cord_PD),]$Placenta_PD, 0, 7))), all_vcfs)]
  for(vcf in patient_vcfs){
    sample = strsplit(vcf, '\\.')[[1]][1]
    data = VariantAnnotation::readVcf(vcf)
    chr=as.character(seqnames(rowRanges(data)))
    pos=start(ranges(rowRanges(data)))
    ref=as.character(ref(data))
    alt=as.character(unlist(CharacterList(alt(data))))
    
    info(data)$mut_ID = paste(chr, pos, ref, alt, sep = '_')
    info(header(data)) = rbind(info(header(data)), DataFrame(Number = '.', Type = 'String', Description = 'Unique mutation ID', row.names = 'mut_ID'))
    filtered_data = data[info(data)$mut_ID %in% filtered_muts]
    
    writeVcf(filtered_data, filename = paste0("../06_depth_amb_unk/", sample, '.pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.filtered.vcf'))
    
  }
}

```

```{bash}

cd /scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/06_depth_amb_unk

for file in $(ls *.pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.filtered.vcf);
do
echo $file
grep -v "#" $file | wc -l
done

```

###Direction filter

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/06_depth_amb_unk

module load bcftools

for f in $(ls *.pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.filtered.vcf);
do
filename=$(basename -- "$f")
sample="${filename%.pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.filtered.vcf}"
bcftools filter -i '(FORMAT/PU[1] > 0) & (FORMAT/NU[1] > 0)' $f > ../07_direction/${sample}.pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.filtered.vcf
done

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/07_direction

for file in $(ls *.pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.filtered.vcf);
do
echo $file
grep -v "#" $file | wc -l
done


```

###Convert to text files

####Function

```{r}

pindel_vcf_to_text_file = function(vcf_file){
  vcf = readVcf(vcf_file)
  if(nrow(vcf) > 0){
    sample = unlist(strsplit(vcf_file, '.pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.filtered.vcf'))
    var.df = data.frame(Chr=as.character(seqnames(rowRanges(vcf))),
                  Position=start(ranges(rowRanges(vcf))),
                  Wildtype=as.character(ref(vcf)),
                  Mutant=as.character(unlist(CharacterList(alt(vcf)))),
                  Variant_read_depth=geno(vcf)$FC[,2],
                  Total_read_depth=geno(vcf)$FD[,2])
    var.df$VAF = var.df$Variant_read_depth / var.df$Total_read_depth
    var.df$Mutation = info(vcf)$VT
    vagrent_annots = lapply(strsplit(info(vcf)$VW, '\\|'), `length<-`, max(lengths(strsplit(info(vcf)$VW, '\\|')))) #make all elements in list the same length for subsetting purposes
    if(max(unlist(lapply(vagrent_annots, function(x) length(x)))) == 1){ #where none land in gene footprints and single NA values are found in each case
      var.df$Gene = NA
      var.df$Transcript = NA
      var.df$Codon = NA
      var.df$Protein = NA
    } else {
      var.df$Gene = sapply(vagrent_annots, '[[', 1)
      var.df$Transcript = sapply(vagrent_annots, '[[', 3)
      var.df$Codon = sapply(vagrent_annots, '[[', 4)
      var.df$Protein = sapply(vagrent_annots, '[[', 5)
    }

    conseq_list = lapply(info(vcf)$VC, function(x) if(identical(x, character(0))) NA_character_ else x)
    conseq_vec = c()
    for(i in 1:length(conseq_list)){
      conseq_vec = c(conseq_vec, paste(unlist(conseq_list[i]), collapse = ';'))
    }
  
    var.df$Consequence = conseq_vec
    var.df$Case_ID = substr(sample, 0, 7)
    var.df$Sample = sample
  
    write.table(var.df[, c("Case_ID", "Sample", "Chr", "Position", "Wildtype", "Mutant", "Total_read_depth", "Variant_read_depth", "VAF", "Mutation", "Gene",      "Transcript", "Codon", "Protein", "Consequence")], paste0(sample, '.pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.filtered.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
  }
}

```

####Run

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/07_direction")

vcf_files = list.files(".", ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.filtered.vcf")

for(file in vcf_files){
  pindel_vcf_to_text_file(file)
}

```

###Correct PD45581c chr10 calls

chr10 trisomy with maternal UPD in cord - hence calling paternal chr10 germline indels as somatic when run matched.

```{r}

indels = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/07_direction/PD45581c.pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.filtered.txt", header = T, sep = '\t', stringsAsFactors = F)

chr10_pval = rep(NA, length(indels[indels$Chr == 10,]$Variant_read_depth))
for(i in 1:length(chr10_pval)){
  chr10_pval[i] = binom.test(indels[indels$Chr == 10,]$Variant_read_depth[i], indels[indels$Chr == 10,]$Total_read_depth[i], p = 1/3, alt = "less")$p.value
}

qval = p.adjust(chr10_pval, method = "BH")
germline = log10(qval) > -3

indels$keep = "Y"
indels[indels$Chr == 10,][germline,]$keep = "N"

indels_flt = indels[indels$keep == "Y",]

write.table(indels_flt[, 1:15], "/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/07_direction/PD45581c.pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.filtered.txt", col.names = T, row.names = F, sep = '\t', quote = F)

```

###Recurrent artefact filter

The latest batch of samples in particular is of lower DNA quality with some recurrent artefacts (across placentas) that should be removed.

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/07_direction/")

txt_files = list.files(".", ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.filtered.txt")

all_indels = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  all_indels = rbind(all_indels, data)
}

all_indels$mut_ID = paste(all_indels$Chr, all_indels$Position, all_indels$Wildtype, all_indels$Mutant, sep = "_")

#recurrent artefacts?
recurrent_vars = all_indels[!duplicated(all_indels[, c("Case_ID", "mut_ID")]),]$mut_ID[duplicated(all_indels[!duplicated(all_indels[, c("Case_ID", "mut_ID")]),]$mut_ID)] #65, previously 1377

length(recurrent_vars) #65
length(unique(recurrent_vars)) #52, some repeated more than once

#probably a more elegent way of writing it in retrospect 27/1/23
recurrent_vars2 = names(table(all_indels[!duplicated(all_indels[, c("Case_ID", "mut_ID")]),]$mut_ID))[table(all_indels[!duplicated(all_indels[, c("Case_ID", "mut_ID")]),]$mut_ID) > 1]
length(recurrent_vars2) #52
sum(recurrent_vars2 %in% recurrent_vars) #52 perfect match

temp = all_indels[all_indels$mut_ID %in% recurrent_vars2,]
temp[order(temp$mut_ID),]

table(all_indels[all_indels$mut_ID %in% recurrent_vars,]$Consequence) #no coding variants

#now filter
txt_files = list.files(".", ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.filtered.txt")

for(file in txt_files){
  sample = unlist(strsplit(file, ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$mut_ID = paste(data$Chr, data$Position, data$Wildtype, data$Mutant, sep = "_")
  data_flt = data[!data$mut_ID %in% recurrent_vars,]
  write.table(data_flt, paste0("../08_recurrent/", sample, ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.recurrent_vars.filtered.txt"), col.names = T, row.names = F, sep = '\t', quote = F)
}

```

##Copy number aberrations

###ascatPCA

####Restrict to final list of samples

```{r}

ascat = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/ascatPCA/Placenta_CNV_ascatPCA_20220322_additional_samples_20220513_copy_number.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

ascat_flt = ascat[ascat$sample %in% manifest[manifest$Other != "Membrane", ]$Placenta_PD, ]

write.table(ascat_flt, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/ascatPCA/placenta_ascatPCA_calls_final_20220712.txt", col.names = T, row.names = F, sep = '\t', quote = F)

```

###Battenberg

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/Battenberg")

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220322.txt", header = T, sep = '\t', stringsAsFactors = F)

contaminated_bx = manifest[manifest$Biopsy_freemix > 0.01,]$Placenta_PD

pp_files = list.files(".", "_purity_ploidy.txt")

purity_ploidy = data.frame()
for(file in pp_files){
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  sample = unlist(strsplit(file, "_purity_ploidy.txt"))
  data$sample = sample
  purity_ploidy = rbind(purity_ploidy, data) #either there are at least subclones, both of which represent more than 10% of cells or there is a single clonal call that is not 1+1
}

hist(purity_ploidy[!purity_ploidy$sample %in% contaminated_bx, ]$ploidy, breaks = 100)
purity_ploidy[purity_ploidy$ploidy == min(purity_ploidy[!purity_ploidy$sample %in% contaminated_bx, ]$ploidy),]
#purity   ploidy      psi   sample
#0.22 1.955168 2.688005 PD45583b

purity_ploidy[!purity_ploidy$sample %in% contaminated_bx & purity_ploidy$ploidy > 2.5, ] #uneven coverage resulting in abnormal calls with equal maj and min alleles

bb_files = list.files(".", "_subclones.txt")

abnormal_cn = data.frame()
for(file in bb_files){
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  sample = unlist(strsplit(file, "_subclones.txt"))
  data$sample = sample
  abnormal_cn = rbind(abnormal_cn, data[(!is.na(data$frac2_A) & (data$frac1_A > 0.1 & data$frac2_A > 0.1)) | ((data$nMaj1_A != 1 | data$nMin1_A != 1) & data$frac1_A > 0.1),]) #either there are at least subclones, both of which represent more than 10% of cells or there is a single clonal call that is not 1+1
}

table(abnormal_cn[!abnormal_cn$sample %in% c(contaminated_bx), ]$sample) #inflated abnormal calls in PD45583b with poor fit, likely due to uneven coverage

nrow(abnormal_cn[!abnormal_cn$sample %in% c(contaminated_bx, "PD45583b"), ]) #117 calls

final_abnormal_calls = abnormal_cn[!abnormal_cn$sample %in% c(contaminated_bx, "PD45583b"), ]

final_abnormal_calls.annot = data.frame()
for(sample in unique(final_abnormal_calls$sample)){
  
  baf_file = read.table(paste0("/lustre/scratch119/realdata/mdt1/team274/na15/Placenta/results/phased_normalized_battenberg/", sample, "/", sample, ".BAFsegmented.txt"), header = T, sep = '\t', stringsAsFactors = F)
  
  sample_cnas = final_abnormal_calls[final_abnormal_calls$sample == sample,]
  sample_cnas$informative_snps = NA
  
  for(i in 1:nrow(sample_cnas)){
    sample_cnas$informative_snps[i] = nrow(baf_file[baf_file$Chromosome == sample_cnas$chr[i] & baf_file$Position >= sample_cnas$startpos[i] & baf_file$Position <= sample_cnas$endpos[i], ])
  }
  
  sample_cnas$length = sample_cnas$endpos - sample_cnas$startpos
  final_abnormal_calls.annot = rbind(final_abnormal_calls.annot, sample_cnas)
  
}

write.table(final_abnormal_calls.annot, "placenta_cnas_20220324.txt", col.names = T, row.names = F, sep = '\t', quote = F)

library(ggplot2)

ggplot(final_abnormal_calls.annot) + geom_point(mapping = aes(x = length, y = informative_snps)) + theme_linedraw() + xlab("Segment length (bp)") + ylab("Number of informative SNPs in segment")

#review last batch of samples 11/5/22

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/Battenberg/batch_20220511")

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220505.txt", header = T, sep = '\t', stringsAsFactors = F)

contaminated_bx = manifest[manifest$Biopsy_freemix > 0.01,]$Placenta_PD

pp_files = list.files(".", "_purity_ploidy.txt")

purity_ploidy = data.frame()
for(file in pp_files){
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  sample = unlist(strsplit(file, "_purity_ploidy.txt"))
  data$sample = sample
  purity_ploidy = rbind(purity_ploidy, data) 
}

hist(purity_ploidy[!purity_ploidy$sample %in% contaminated_bx, ]$ploidy, breaks = 100)
purity_ploidy[purity_ploidy$ploidy == min(purity_ploidy[!purity_ploidy$sample %in% contaminated_bx, ]$ploidy),]
#purity   ploidy      psi   sample
#0.95 1.999992 2.251988 PD45573h

purity_ploidy[!purity_ploidy$sample %in% contaminated_bx & purity_ploidy$ploidy > 2.5, ] #uneven coverage resulting in abnormal calls with equal maj and min alleles
#none

bb_files = list.files(".", "_subclones.txt")

abnormal_cn = data.frame()
for(file in bb_files){
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)
  sample = unlist(strsplit(file, "_subclones.txt"))
  data$sample = sample
  abnormal_cn = rbind(abnormal_cn, data[(!is.na(data$frac2_A) & (data$frac1_A > 0.1 & data$frac2_A > 0.1)) | ((data$nMaj1_A != 1 | data$nMin1_A != 1) & data$frac1_A > 0.1),]) #either there are at least subclones, both of which represent more than 10% of cells or there is a single clonal call that is not 1+1
}

table(abnormal_cn[!abnormal_cn$sample %in% c(contaminated_bx), ]$sample) #a single call in the new samples

final_abnormal_calls = abnormal_cn

final_abnormal_calls.annot = data.frame()
for(sample in unique(final_abnormal_calls$sample)){
  
  baf_file = read.table(paste0("/lustre/scratch119/casm/team274sb/na15/Placenta/results/battenberg/", sample, "/", sample, ".BAFsegmented.txt"), header = T, sep = '\t', stringsAsFactors = F)
  
  sample_cnas = final_abnormal_calls[final_abnormal_calls$sample == sample,]
  sample_cnas$informative_snps = NA
  
  for(i in 1:nrow(sample_cnas)){
    sample_cnas$informative_snps[i] = nrow(baf_file[baf_file$Chromosome == sample_cnas$chr[i] & baf_file$Position >= sample_cnas$startpos[i] & baf_file$Position <= sample_cnas$endpos[i], ])
  }
  
  sample_cnas$length = sample_cnas$endpos - sample_cnas$startpos
  final_abnormal_calls.annot = rbind(final_abnormal_calls.annot, sample_cnas)
  
}

write.table(final_abnormal_calls.annot, "placenta_cnas_20220511.txt", col.names = T, row.names = F, sep = '\t', quote = F)

```

####Restrict to final list of samples

```{r}

battenberg = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/Battenberg/placenta_cnas_20220324_reviewed_20220426_additional_20220511.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

battenberg_flt = battenberg[battenberg$consensus_review == "Yes" & battenberg$sample %in% manifest[manifest$Other != "Membrane", ]$Placenta_PD, ]

write.table(battenberg_flt, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/Battenberg/placenta_battenberg_calls_final_20220712.txt", col.names = T, row.names = F, sep = '\t', quote = F)

```


##Structural variants

###BRASS

####Restrict to final list of samples

```{r}

brass = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/Placenta_AS_filtered_Mar8_2022_Annotated.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

brass_flt = brass[brass$PASS_FAIL == "PASS" & brass$file_sample %in% manifest[manifest$Other != "Membrane", ]$Placenta_PD, ]

write.table(brass_flt, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/placenta_brass_calls_final_20220712.txt", col.names = T, row.names = F, sep = '\t', quote = F)

```


###Retrotranposon activity

```{r}

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)
rt_calls = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/placenta_RT_calls_20220617.txt", header = T, sep = '\t', stringsAsFactors = F)

rt_calls_flt = rt_calls[rt_calls$sample %in% manifest$Placenta_PD,]

write.table(rt_calls_flt, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/placenta_RT_calls_20220617_filtered_samples.txt", col.names = T, row.names = F, sep = '\t', quote = F)

rt_coords = as.data.frame(rbind(as.matrix(rt_calls_flt[, c(1:2, 4)]), as.matrix(rt_calls_flt[, c(1, 3, 4)])))
rt_coords_flt = rt_coords[!is.na(rt_coords$start),]

sv_calls = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/Placenta_AS_filtered_Mar8_2022_Annotated.txt", header = T, sep = '\t', stringsAsFactors = F)
sv_calls$near_rt = NA

#i = 59
for(i in 1:nrow(sv_calls)){
  if(sv_calls$file_sample[i] %in% rt_coords_flt$sample){
    sample_rt_calls = rt_coords_flt[rt_coords_flt$sample == sv_calls$file_sample[i],]
    
    if(nrow(sample_rt_calls[(sample_rt_calls$chro == sv_calls$chr1[i] & sample_rt_calls$start >= sv_calls$start1[i] - 100 & sample_rt_calls$start <= sv_calls$start1[i] + 100) | 
                            (sample_rt_calls$chro == sv_calls$chr1[i] & sample_rt_calls$start >= sv_calls$end1[i] - 100 & sample_rt_calls$start <= sv_calls$end1[i] + 100) |
                            (sample_rt_calls$chro == sv_calls$chr2[i] & sample_rt_calls$start >= sv_calls$start2[i] - 100 & sample_rt_calls$start <= sv_calls$start2[i] + 100) |
                            (sample_rt_calls$chro == sv_calls$chr2[i] & sample_rt_calls$start >= sv_calls$end2[i] - 100 & sample_rt_calls$start <= sv_calls$end2[i] + 100), ]) > 0){
      sv_calls$near_rt[i] = "Y"
    } else {
      
        sv_calls$near_rt[i] = "N"
        
      }
    
  }
  
  if(!sv_calls$file_sample[i] %in% rt_coords_flt$sample){
    sv_calls$near_rt[i] = "N"
  }
  
}

#one SV nearby failed SV calling so it doesn't matter

```

####Restrict to final list of samples

```{r}

rt = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/placenta_RT_calls_20220617_filtered_samples.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

rt_flt = rt[rt$sample %in% manifest[manifest$Other != "Membrane", ]$Placenta_PD, ]

write.table(rt_flt, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/placenta_rt_calls_final_20220712.txt", col.names = T, row.names = F, sep = '\t', quote = F)

```

#03_DNA_ANALYSES

##Add burdens to metadata

###Subs

```{r}

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/07_recurrent")

txt_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt")

all_subs = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  all_subs = rbind(all_subs, data)
}

manifest$sub_burden = NA
manifest$sub_auto_burden = NA
manifest$sub_highVAF_burden = NA
manifest$sub_median_VAF = NA

for(i in 1:nrow(manifest)){
  if(manifest$Placenta_PD[i] %in% unique(all_subs$Sample)){
    manifest$sub_burden[i] = nrow(all_subs[all_subs$Sample == manifest$Placenta_PD[i],])
    manifest$sub_auto_burden[i] = nrow(all_subs[all_subs$Sample == manifest$Placenta_PD[i] & all_subs$Chr %in% c(1:22),])
    manifest$sub_highVAF_burden[i] = nrow(all_subs[all_subs$Sample == manifest$Placenta_PD[i] & all_subs$Chr %in% c(1:22) & all_subs$VAF >= 0.2,])
    manifest$sub_median_VAF[i] = median(all_subs[all_subs$Sample == manifest$Placenta_PD[i] & all_subs$Chr %in% c(1:22),]$VAF)
  }
}

write.table(manifest, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", col.names = T, row.names = F, sep = '\t', quote = F)

```

###Indels

```{r}

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/08_recurrent")

txt_files = list.files(".", ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.recurrent_vars.filtered.txt")

all_indels = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.recurrent_vars.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  data = data[data$VAF >= 0.1,] #added 16/6/22 due to an excess of low VAF calls
  all_indels = rbind(all_indels, data)
}

manifest$indel_burden = NA
manifest$indel_highVAF_burden = NA
manifest$indel_median_VAF = NA

for(i in 1:nrow(manifest)){
    manifest$indel_burden[i] = nrow(all_indels[all_indels$Sample == manifest$Placenta_PD[i],])
    manifest$indel_highVAF_burden[i] = nrow(all_indels[all_indels$Sample == manifest$Placenta_PD[i] & all_indels$Chr %in% c(1:22) & all_indels$VAF >= 0.2,])
    manifest$indel_median_VAF[i] = median(all_indels[all_indels$Sample == manifest$Placenta_PD[i] & all_indels$Chr %in% c(1:22),]$VAF)
}

write.table(manifest, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", col.names = T, row.names = F, sep = '\t', quote = F)

```

###Structural variants

```{r}

rt_calls = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/placenta_rt_calls_final_20220712.txt", header = T, sep = '\t', stringsAsFactors = F)
svs_flt = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/placenta_brass_calls_final_20220712.txt", header = T, sep = '\t', stringsAsFactors = F)
svs_flt = svs_flt[svs_flt$bkdist == -1 | svs_flt$bkdist >= 1000,]
manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

#old = manifest 1 deletion was removed

manifest$rt_count = NA
manifest$del_sv_count = NA
manifest$inv_sv_count = NA
manifest$dup_sv_count = NA
manifest$tra_sv_count = NA

for(i in 1:nrow(manifest)){
  manifest$rt_count[i] = nrow(rt_calls[rt_calls$sample == manifest$Placenta_PD[i], ])
  manifest$del_sv_count[i] = nrow(svs_flt[svs_flt$file_sample == manifest$Placenta_PD[i] & svs_flt$svclass == "deletion",])
  manifest$inv_sv_count[i] = nrow(svs_flt[svs_flt$file_sample == manifest$Placenta_PD[i] & svs_flt$svclass == "inversion",])
  manifest$dup_sv_count[i] = nrow(svs_flt[svs_flt$file_sample == manifest$Placenta_PD[i] & svs_flt$svclass == "tandem-duplication",])
  manifest$tra_sv_count[i] = nrow(svs_flt[svs_flt$file_sample == manifest$Placenta_PD[i] & svs_flt$svclass == "translocation",])
}

manifest[manifest$Other == "Membrane",]$rt_count = NA
manifest[manifest$Other == "Membrane",]$del_sv_count = NA
manifest[manifest$Other == "Membrane",]$inv_sv_count = NA
manifest[manifest$Other == "Membrane",]$dup_sv_count = NA
manifest[manifest$Other == "Membrane",]$tra_sv_count = NA

write.table(manifest, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", col.names = T, row.names = F, sep = '\t', quote = F)

```

##SV burden analysis

```{r}

library(ggplot2)

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

manifest$total_sv_burden = rowSums(manifest[, c("del_sv_count", "inv_sv_count", "dup_sv_count", "tra_sv_count")])

svs_flt = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/placenta_brass_calls_final_20220712.txt", header = T, sep = '\t', stringsAsFactors = F)
svs_flt = read.table("/lustre/scratch126/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/placenta_brass_calls_final_20220712.txt", header = T, sep = '\t', stringsAsFactors = F)
svs_flt = svs_flt[(svs_flt$bkdist == -1 | svs_flt$bkdist >= 1000) & svs_flt$sample %in% manifest[manifest$Other != "Membrane",]$Placenta_PD,]

#check for shared SVs
#merge all SVs for a placenta whose lower and upper breakpoints fall within a 1kb window

svs_flt$case.id = substr(svs_flt$sample, 0, 7)

merged_sv_annot = list()
for(case in unique(svs_flt$case.id)){
  
  case_svs = svs_flt[svs_flt$case.id == case,]
  case_svs = case_svs[order(case_svs$chr1, case_svs$start1, case_svs$end1),]
  
  #lower bk
  case_svs$start1_500bp = case_svs$start1 - 500
  case_svs$end1_500bp = case_svs$end1 + 500
  temp1 = makeGRangesFromDataFrame(case_svs[, c("chr1", "start1_500bp", "end1_500bp")], start.field = "start1_500bp", end.field = "end1_500bp", seqnames.field = "chr1")
  temp1.ov = findOverlaps(temp1)
  ov.vec1 = temp1.ov@from[!duplicated(temp1.ov@to)]
  case_svs$bk1 = paste(case_svs[ov.vec1,]$chr1, case_svs[ov.vec1,]$start1_500bp, case_svs[ov.vec1,]$end1_500bp, sep = "_")
  
  case_svs = case_svs[order(case_svs$chr2, case_svs$start2, case_svs$end2),]
  
  #upper bk
  case_svs$start2_500bp = case_svs$start2 - 500
  case_svs$end2_500bp = case_svs$end2 + 500
  temp2 = makeGRangesFromDataFrame(case_svs[, c("chr2", "start2_500bp", "end2_500bp")], start.field = "start2_500bp", end.field = "end2_500bp", seqnames.field = "chr2")
  temp2.ov = findOverlaps(temp2)
  ov.vec2 = temp2.ov@from[!duplicated(temp2.ov@to)]
  case_svs$bk2 = paste(case_svs[ov.vec2,]$chr2, case_svs[ov.vec2,]$start2_500bp, case_svs[ov.vec2,]$end2_500bp, sep = "_")
  
  case_svs = case_svs[order(case_svs$chr1, case_svs$start1, case_svs$end1),]
  
  print(sum(case_svs$chr1 != unlist(lapply(strsplit(case_svs$bk1, "_"), "[[", 1)))) #0 - correct order
  print(sum(case_svs$chr2 != unlist(lapply(strsplit(case_svs$bk2, "_"), "[[", 1)))) #0 - correct order
  
  #merge
  case_svs$sv_id = paste(case_svs$bk1, case_svs$svclass, case_svs$bk2, sep = "_")
  #case_svs2 = case_svs[, c("sample", "case.id", "sv_id", "VAF")]
  
  merged_sv_annot[[case]] = case_svs
  
}

merged_sv_annot_df = data.frame()

#case.id = "PD42145"
for(case.id in names(merged_sv_annot)){
  
  merged_sv_annot_df = rbind(merged_sv_annot_df, merged_sv_annot[[case.id]])
  
}

nrow(svs_flt) #287
nrow(merged_sv_annot_df) #287

nrow(merged_sv_annot_df[!duplicated(merged_sv_annot_df$sv_id),]) #282

merged_sv_annot_df[merged_sv_annot_df$sv_id %in% merged_sv_annot_df[duplicated(merged_sv_annot_df$sv_id),]$sv_id,]

table(unlist(lapply(strsplit(unique(merged_sv_annot_df$sv_id), "_"), "[[", 4)))
#deletion          inversion tandem-duplication      translocation 
#229                 17                 34                  2 

table(unlist(lapply(strsplit(unique(merged_sv_annot_df$sv_id), "_"), "[[", 4))) / length(unique(merged_sv_annot_df$sv_id))
#deletion          inversion          tandem-duplication      translocation 
#0.812056738       0.060283688        0.120567376             0.007092199 

summary(merged_sv_annot_df[!duplicated(merged_sv_annot_df$sv_id) & merged_sv_annot_df$bkdist != -1,]$bkdist)
#Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
#1253     16770     60726    545111    138728 116191987

fs_bed = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/hg37_fragile_sites.bed", header = F, sep = "\t", stringsAsFactors = F)
fs_bed = read.table("/lustre/scratch126/casm/team274sb/to3/placenta_ml/99_Reference_data/hg37_fragile_sites.bed", header = F, sep = "\t", stringsAsFactors = F)

lower_bk = makeGRangesFromDataFrame(merged_sv_annot_df[, c("chr1", "start1", "end1")], start.field = "start1", end.field = "end1", seqnames.field = "chr1")
upper_bk = makeGRangesFromDataFrame(merged_sv_annot_df[, c("chr2", "start2", "end2")], start.field = "start2", end.field = "end2", seqnames.field = "chr2")
fragile_sites = makeGRangesFromDataFrame(fs_bed[, c("V1", "V2", "V3")], start.field = "V2", end.field = "V3", seqnames.field = "V1")

lower_overlaps = findOverlaps(lower_bk, fragile_sites)
merged_sv_annot_df$lower_bk_fragile = F
merged_sv_annot_df[lower_overlaps@from,]$lower_bk_fragile = T

upper_overlaps = findOverlaps(upper_bk, fragile_sites)
merged_sv_annot_df$upper_bk_fragile = F
merged_sv_annot_df[upper_overlaps@from,]$upper_bk_fragile = T

nrow(merged_sv_annot_df[!duplicated(merged_sv_annot_df$sv_id) & (merged_sv_annot_df$lower_bk_fragile == T | merged_sv_annot_df$upper_bk_fragile == T),]) #154

nrow(merged_sv_annot_df[!duplicated(merged_sv_annot_df$sv_id) & (merged_sv_annot_df$lower_bk_fragile == T | merged_sv_annot_df$upper_bk_fragile == T),]) / length(unique(merged_sv_annot_df$sv_id))

table(merged_sv_annot_df[!duplicated(merged_sv_annot_df$sv_id) & (merged_sv_annot_df$lower_bk_fragile == T | merged_sv_annot_df$upper_bk_fragile == T),]$svclass) / table(merged_sv_annot_df[!duplicated(merged_sv_annot_df$sv_id),]$svclass)

merged_sv_annot_df[merged_sv_annot_df$lower_bk_fragile == F & merged_sv_annot_df$upper_bk_fragile == F,]

table(merged_sv_annot_df[!duplicated(merged_sv_annot_df$sv_id) & (merged_sv_annot_df$lower_bk_fragile == T | merged_sv_annot_df$upper_bk_fragile == T),]$svclass) / table(merged_sv_annot_df[!duplicated(merged_sv_annot_df$sv_id),]$svclass)

```


##Identify driver point mutations

```{r}

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/07_recurrent")

txt_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt")

all_subs = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  all_subs = rbind(all_subs, data)
}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/08_recurrent")

txt_files = list.files(".", ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.recurrent_vars.filtered.txt")

all_indels = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.recurrent_vars.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  data = data[data$VAF >= 0.1,] #added 16/6/22 due to an excess of low VAF calls
  all_indels = rbind(all_indels, data)
}

all_muts = rbind(all_subs, all_indels)

cosmic_v94 <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/cosmic_v94_cancer_gene_census.csv', sep = ',', header = T, stringsAsFactors = F)
cancer_genes <- unlist(strsplit(paste(cosmic_v94$Synonyms, collapse = ","), ","))

all_muts_cancer_genes = all_muts[all_muts$Gene %in% cancer_genes,]
table(all_muts_cancer_genes$Consequence)

write.table(all_muts_cancer_genes[all_muts_cancer_genes$Consequence %in% c("5prime_UTR_ess_splice", "frameshift", "missense", "nonsense"),], "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/blk_placenta_put_point_muts_20230201.txt", col.names = T, row.names = F, sep = "\t", quote = F)

#none look convincing cancer drivers 4/2/23

```

##Identify putative drivers in CNAs


```{r}

#COSMIC consensus list of genes
cosmic_v94 <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/cosmic_v94_cancer_gene_census.csv', sep = ',', header = T, stringsAsFactors = F)
cosmic_v94 = cosmic_v94[!grepl(":-", cosmic_v94$Genome.Location),]

#get GRCh37 build genomic coordinates
library(biomaRt)
ensembl37 <- useEnsembl(biomart = 'genes', 
                       dataset = 'hsapiens_gene_ensembl',
                       version = 'GRCh37')
gene_coord = getBM(attributes=c('hgnc_symbol','chromosome_name','start_position','end_position'), mart = ensembl37)
gene_coord = gene_coord[gene_coord$hgnc_symbol != "",]
gene_coord = gene_coord[!grepl("^MIR", gene_coord$hgnc_symbol),]
gene_coord = gene_coord[gene_coord$chromosome_name %in% c(1:22, "X", "Y"),]
gene_coord = gene_coord[gene_coord$hgnc_symbol != "UGT2A1",] #last duplicated entry
row.names(gene_coord) = gene_coord$hgnc_symbol

#add coords to cosmic data
cosmic_v94 = cosmic_v94[cosmic_v94$Gene.Symbol %in% gene_coord$hgnc_symbol,]
nrow(cosmic_v94) #707
cosmic_v94$chr_h19 = gene_coord[cosmic_v94$Gene.Symbol,]$chromosome_name
cosmic_v94$start_h19 = gene_coord[cosmic_v94$Gene.Symbol,]$start_position
cosmic_v94$end_h19 = gene_coord[cosmic_v94$Gene.Symbol,]$end_position

oncogenes = cosmic_v94[grepl('oncogene', cosmic_v94$Role.in.Cancer) | grepl('Dom', cosmic_v94$Molecular.Genetics),]
rcgenes = cosmic_v94[grepl('TSG', cosmic_v94$Role.in.Cancer) | grepl('Rec', cosmic_v94$Molecular.Genetics),]

#read in my data
manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

ascat_flt = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/ascatPCA/placenta_ascatPCA_calls_final_20220712.txt", header = T, sep = "\t", stringsAsFactors = F, quote="")
battenberg_flt = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/Battenberg/placenta_battenberg_calls_final_20220712.txt", header = T, sep = "\t", stringsAsFactors = F, quote="")

#annotate with genes
ascat_flt$oncogene = NA
ascat_flt$tsg = NA
battenberg_flt$oncogene = NA
battenberg_flt$tsg = NA

for(i in 1:nrow(ascat_flt)){
  
  if(nrow(oncogenes[oncogenes$chr_h19 == ascat_flt$chr[i] & oncogenes$end_h19 > ascat_flt$start[i] & oncogenes$start_h19 < ascat_flt$end[i],]) > 0){
    ascat_flt$oncogene[i] = paste0(oncogenes[oncogenes$chr_h19 == ascat_flt$chr[i] & oncogenes$end_h19 > ascat_flt$start[i] & oncogenes$start_h19 < ascat_flt$end[i],]$Gene.Symbol, collapse = ",")
  }
  
  if(nrow(rcgenes[rcgenes$chr_h19 == ascat_flt$chr[i] & rcgenes$end_h19 > ascat_flt$start[i] & rcgenes$start_h19 < ascat_flt$end[i],]) > 0){
    ascat_flt$tsg[i] = paste0(rcgenes[rcgenes$chr_h19 == ascat_flt$chr[i] & rcgenes$end_h19 > ascat_flt$start[i] & rcgenes$start_h19 < ascat_flt$end[i],]$Gene.Symbol, collapse = ",")
  }

}

for(i in 1:nrow(battenberg_flt)){

  if(nrow(oncogenes[oncogenes$chr_h19 == battenberg_flt$chr[i] & oncogenes$end_h19 > battenberg_flt$start[i] & oncogenes$start_h19 < battenberg_flt$end[i],]) > 0){
    battenberg_flt$oncogene[i] = paste0(oncogenes[oncogenes$chr_h19 == battenberg_flt$chr[i] & oncogenes$end_h19 > battenberg_flt$start[i] & oncogenes$start_h19 < battenberg_flt$end[i],]$Gene.Symbol, collapse = ",")
  }
  
  if(nrow(rcgenes[rcgenes$chr_h19 == battenberg_flt$chr[i] & rcgenes$end_h19 > battenberg_flt$start[i] & rcgenes$start_h19 < battenberg_flt$end[i],]) > 0){
    battenberg_flt$tsg[i] = paste0(rcgenes[rcgenes$chr_h19 == battenberg_flt$chr[i] & rcgenes$end_h19 > battenberg_flt$start[i] & rcgenes$start_h19 < battenberg_flt$end[i],]$Gene.Symbol, collapse = ",")
  }
  
}

write.table(ascat_flt, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/ascatPCA/placenta_ascatPCA_calls_final_20220712_cancer_gene_annot_20230204.txt", col.names = T, row.names = F, sep = '\t', quote = F)
write.table(battenberg_flt, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/Battenberg/placenta_battenberg_calls_final_20220712_cancer_gene_annot_20230204.txt", col.names = T, row.names = F, sep = '\t', quote = F)

#2 11p UPD events

```

##Identify 11p CNAs

###Binomial test

Allele counts for mother and father using maternal SNPs per sample

```{r}

library(ggplot2)
library(zoo)

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/locus_11p15.5_upd/maternal_snps")

allelecount_files = list.files(".", "alleleCounts.tab")
samples = unlist(lapply(strsplit(allelecount_files, "_"), "[[", 1))

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest$binom.pval.11p15.5.bx = NA
manifest$binom.p.success.11p15.5.bx = NA
manifest$binom.pval.11p15.5.cord = NA
manifest$binom.p.success.11p15.5.cord = NA

manifest$Placenta_PD[!manifest$Placenta_PD %in% samples] #3 placentas without maternal DNA

#IGF2 11:2150342..2170833
#H19 chr11:2016406-2019065

#file = "PD51880e_alleleCounts.tab"
for(file in allelecount_files){
  
  sample = unlist(strsplit(file, "_"))[1]
  
  if(sample %in% manifest$Placenta_PD){
    
    print(sample)
    
    data = read.table(file, header = T, sep = "\t", stringsAsFactors = F, comment.char = "")
    data_flt = data[data$Chromosome == 11 & data$Position < 2800000,]
    data_flt2 = data[data$Chromosome == 11,]
  
    data_flt2$mutCountT1_af = data_flt2$mutCountT1 / (data_flt2$mutCountT1 + data_flt2$mutCountT2)
    data_flt2$mutCountT2_af = data_flt2$mutCountT2 / (data_flt2$mutCountT1 + data_flt2$mutCountT2)
    data_flt2$mutCountN1_af = data_flt2$mutCountN1 / (data_flt2$mutCountN1 + data_flt2$mutCountN2)
    data_flt2$mutCountN2_af = data_flt2$mutCountN2 / (data_flt2$mutCountN1 + data_flt2$mutCountN2)
    data_flt2$mutCountT1_af_rollmean = rollmean(data_flt2$mutCountT1_af, 20, fill = NA)
    data_flt2$mutCountT2_af_rollmean = rollmean(data_flt2$mutCountT2_af, 20, fill = NA)
    data_flt2$mutCountN1_af_rollmean = rollmean(data_flt2$mutCountN1_af, 20, fill = NA)
    data_flt2$mutCountN2_af_rollmean = rollmean(data_flt2$mutCountN2_af, 20, fill = NA)
    
    bx_res = binom.test(x = sum(data_flt$mutCountT1), n = sum(data_flt$mutCountT1) + sum(data_flt$mutCountT2), p = 0.5, alternative = "two.sided")
  
    manifest[manifest$Placenta_PD == sample, ]$binom.pval.11p15.5.bx = bx_res$p.value
    manifest[manifest$Placenta_PD == sample, ]$binom.p.success.11p15.5.bx = as.numeric(bx_res$estimate)
    
    png(paste0(sample, "_chr11_phased_snps_20230204.png"), width = 8, height = 3, units = "in", res = 800)
    p = ggplot(data_flt2) +
      geom_point(mapping = aes(x = Position, y = mutCountT1_af), alpha = 0) +
      xlim(0, max(data_flt2$Position)) +
      ylim(0, 1) +
      theme_bw() +
      geom_rect(mapping = aes(xmin = 0, xmax = 2800000, ymin = 0, ymax = 1),
                color="transparent",
                fill = "yellow") +
      geom_point(mapping = aes(x = Position, y = mutCountT1_af), col = "red", pch = ".", alpha = 0.5) +
      geom_point(mapping = aes(x = Position, y = mutCountT2_af), col = "blue", pch = ".", alpha = 0.5) +
      #geom_line(mapping = aes(x = Position, y = mutCountT1_af_rollmean), col = "red") +
      #geom_line(mapping = aes(x = Position, y = mutCountT2_af_rollmean), col = "dodgerblue") +
      coord_cartesian(expand = F) +
      xlab("Chromosome 11 position") +
      ylab("Allele fraction") +
      ggtitle(label = sample)
    print(p)
    dev.off()
    
    png(paste0(sample, "_chr11_phased_snps_rollmean_20230204.png"), width = 8, height = 3, units = "in", res = 800)
    p = ggplot(data_flt2) +
      geom_point(mapping = aes(x = Position, y = mutCountT1_af), alpha = 0) +
      xlim(0, max(data_flt2$Position)) +
      ylim(0, 1) +
      theme_bw() +
      geom_rect(mapping = aes(xmin = 0, xmax = 2800000, ymin = 0, ymax = 1),
                color="transparent",
                fill = "yellow") +
      #geom_point(mapping = aes(x = Position, y = mutCountT1_af), col = "red", pch = ".", alpha = 0.3) +
      #geom_point(mapping = aes(x = Position, y = mutCountT2_af), col = "dodgerblue", pch = ".", alpha = 0.3) +
      geom_line(mapping = aes(x = Position, y = mutCountT1_af_rollmean), col = "red") +
      geom_line(mapping = aes(x = Position, y = mutCountT2_af_rollmean), col = "dodgerblue") +
      coord_cartesian(expand = F) +
      xlab("Chromosome 11 position") +
      ylab("Allele fraction") +
      ggtitle(label = sample)
    print(p)
    dev.off()
    
    cord = manifest[manifest$Placenta_PD == sample, ]$Cord_PD
    
    png(paste0(cord, "_chr11_phased_snps_20230204.png"), width = 8, height = 3, units = "in", res = 800)
    p = ggplot(data_flt2) +
      geom_point(mapping = aes(x = Position, y = mutCountN1_af), alpha = 0) +
      xlim(0, max(data_flt2$Position)) +
      ylim(0, 1) +
      theme_bw() +
      geom_rect(mapping = aes(xmin = 0, xmax = 2800000, ymin = 0, ymax = 1),
                color="transparent",
                fill = "yellow") +
      geom_point(mapping = aes(x = Position, y = mutCountN1_af), col = "red", pch = ".", alpha = 0.3) +
      geom_point(mapping = aes(x = Position, y = mutCountN2_af), col = "blue", pch = ".", alpha = 0.3) +
      #geom_line(mapping = aes(x = Position, y = mutCountN1_af_rollmean), col = "red") +
      #geom_line(mapping = aes(x = Position, y = mutCountN2_af_rollmean), col = "dodgerblue") +
      coord_cartesian(expand = F) +
      xlab("Chromosome 11 position") +
      ylab("Allele fraction") +
      ggtitle(label = cord)
    print(p)
    dev.off()
    
    png(paste0(cord, "_chr11_phased_snps_rollmean_20230204.png"), width = 8, height = 3, units = "in", res = 800)
    p = ggplot(data_flt2) +
      geom_point(mapping = aes(x = Position, y = mutCountN1_af), alpha = 0) +
      xlim(0, max(data_flt2$Position)) +
      ylim(0, 1) +
      theme_bw() +
      geom_rect(mapping = aes(xmin = 0, xmax = 2800000, ymin = 0, ymax = 1),
                color="transparent",
                fill = "yellow") +
      #geom_point(mapping = aes(x = Position, y = mutCountN1_af), col = "darkred", pch = ".", alpha = 0.3) +
      #geom_point(mapping = aes(x = Position, y = mutCountN2_af), col = "blue", pch = ".", alpha = 0.3) +
      geom_line(mapping = aes(x = Position, y = mutCountN1_af_rollmean), col = "red") +
      geom_line(mapping = aes(x = Position, y = mutCountN2_af_rollmean), col = "dodgerblue") +
      coord_cartesian(expand = F) +
      xlab("Chromosome 11 position") +
      ylab("Allele fraction") +
      ggtitle(label = cord)
    print(p)
    dev.off()
    
    cord_res = binom.test(x = sum(data_flt$mutCountN1), n = sum(data_flt$mutCountN1) + sum(data_flt$mutCountN2), p = 0.5, alternative = "two.sided")
  
    manifest[manifest$Placenta_PD == sample, ]$binom.pval.11p15.5.cord = cord_res$p.value
    manifest[manifest$Placenta_PD == sample, ]$binom.p.success.11p15.5.cord = as.numeric(cord_res$estimate)
    
  }
  
}

bx_pval = manifest[, c("Placenta_PD", "binom.pval.11p15.5.bx")]
cord_pval = manifest[, c("Cord_PD", "binom.pval.11p15.5.cord")]
bx_pval = bx_pval[!duplicated(bx_pval) & !is.na(bx_pval$binom.pval.11p15.5.bx),]
cord_pval = cord_pval[!duplicated(cord_pval) & !is.na(cord_pval$binom.pval.11p15.5.cord),]
colnames(bx_pval) = colnames(cord_pval) = c("ID", "pval")

all_pval = rbind(bx_pval, cord_pval)
all_pval$qval = p.adjust(all_pval$pval, method = "fdr")

manifest$binom.qval.11p15.5.bx = NA
manifest$binom.qval.11p15.5.cord = NA

for(i in 1:nrow(manifest)){
  
  if(manifest$Placenta_PD[i] %in% all_pval$ID){
    manifest$binom.qval.11p15.5.bx[i] = all_pval[all_pval$ID == manifest$Placenta_PD[i],]$qval
  }
  
  if(manifest$Cord_PD[i] %in% all_pval$ID){
    manifest$binom.qval.11p15.5.cord[i] = all_pval[all_pval$ID == manifest$Cord_PD[i],]$qval
  }
  
}

hist(log10(manifest$binom.qval.11p15.5.bx), breaks = 200, xlim = c(-50, 0))
abline(v = -8, col = "red")
hist(log10(manifest$binom.qval.11p15.5.cord), breaks = 200, xlim = c(-50, 0))
abline(v = -8, col = "red")

manifest[!is.na(manifest$binom.qval.11p15.5.bx) & manifest$Biopsy_est_mat_contamination == 0 & manifest$Biopsy_freemix < 0.01,][order(manifest[!is.na(manifest$binom.qval.11p15.5.bx),]$binom.qval.11p15.5.bx), c(1:2)]

manifest[!is.na(manifest$binom.qval.11p15.5.bx) & 
           log10(manifest$binom.qval.11p15.5.bx) < -8 &
           manifest$Biopsy_est_mat_contamination == 0 & 
           manifest$Biopsy_freemix < 0.01,]$Placenta_PD
#"PD42154b3" "PD51873e"  "PD51876c"  "PD51876d"  "PD51876e"  "PD51876f"  "PD51880e"  "PD51882c"  "PD51900e" 

manifest[!is.na(manifest$binom.qval.11p15.5.cord) & 
           log10(manifest$binom.qval.11p15.5.cord) < -8 &
           manifest$Cord_est_mat_contamination == 0 & 
           manifest$Cord_freemix < 0.01,]$Cord_PD
#"PD45554f" "PD45554f" "PD51876b" "PD51876b" "PD51876b" "PD51876b" "PD51899b" "PD51899b" "PD51899b" "PD51899b"

write.table(manifest, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", col.names = T, row.names = F, sep = '\t', quote = F)

#could this just be due to chance? Pick 1000 random SNPs from other chromosomes to check per sample
manifest$binom.pval.rand.bx = NA
manifest$binom.p.success.rand.bx = NA
manifest$binom.pval.rand.cord = NA
manifest$binom.p.success.rand.cord = NA

set.seed(42)

#file = "PD51880e_alleleCounts.tab"
for(file in allelecount_files){
  
  sample = unlist(strsplit(file, "_"))[1]
  
  if(sample %in% manifest$Placenta_PD){
    
    print(sample)
    
    data = read.table(file, header = T, sep = "\t", stringsAsFactors = F, comment.char = "")
    data_flt = data[data$Chromosome %in% c(1:10, 12:22),]
    rand_rows = sample(1:nrow(data_flt), 1000, replace = FALSE) #similar number of SNPs from other chromosomes
    data_flt = data_flt[rand_rows,]
    
    bx_res = binom.test(x = sum(data_flt$mutCountT1), n = sum(data_flt$mutCountT1) + sum(data_flt$mutCountT2), p = 0.5, alternative = "two.sided")
  
    manifest[manifest$Placenta_PD == sample, ]$binom.pval.rand.bx = bx_res$p.value
    manifest[manifest$Placenta_PD == sample, ]$binom.p.success.rand.bx = as.numeric(bx_res$estimate)
    
    cord = manifest[manifest$Placenta_PD == sample, ]$Cord_PD
    
    cord_res = binom.test(x = sum(data_flt$mutCountN1), n = sum(data_flt$mutCountN1) + sum(data_flt$mutCountN2), p = 0.5, alternative = "two.sided")
  
    manifest[manifest$Placenta_PD == sample, ]$binom.pval.rand.cord = cord_res$p.value
    manifest[manifest$Placenta_PD == sample, ]$binom.p.success.rand.cord = as.numeric(cord_res$estimate)
    
  }
  
}

#correct the fact the p value has been calculated multiple times per cord biopsy
cord_pval = manifest[, c("Cord_PD", "binom.pval.rand.cord")]
cord_pval = cord_pval[!duplicated(cord_pval$Cord_PD),]

for(i in 1:nrow(manifest)){
    
  manifest$binom.pval.rand.cord[i] = cord_pval[cord_pval$Cord_PD == manifest$Cord_PD[i],]$binom.pval.rand.cord
  
}

#now apply multiple hypothesis correction
bx_pval = manifest[, c("Placenta_PD", "binom.pval.rand.bx")]
cord_pval = manifest[, c("Cord_PD", "binom.pval.rand.cord")]
bx_pval = bx_pval[!duplicated(bx_pval) & !is.na(bx_pval$binom.pval.rand.bx),]
cord_pval = cord_pval[!duplicated(cord_pval) & !is.na(cord_pval$binom.pval.rand.cord),]
colnames(bx_pval) = colnames(cord_pval) = c("ID", "pval")

all_pval = rbind(bx_pval, cord_pval)
all_pval$qval = p.adjust(all_pval$pval, method = "fdr")

manifest$binom.qval.rand.bx = NA
manifest$binom.qval.rand.cord = NA

for(i in 1:nrow(manifest)){
  
  if(manifest$Placenta_PD[i] %in% all_pval$ID){
    manifest$binom.qval.rand.bx[i] = all_pval[all_pval$ID == manifest$Placenta_PD[i],]$qval
  }
  
  if(manifest$Cord_PD[i] %in% all_pval$ID){
    manifest$binom.qval.rand.cord[i] = all_pval[all_pval$ID == manifest$Cord_PD[i],]$qval[1]
  }
  
}

min(manifest[manifest$Biopsy_est_mat_contamination == 0 & 
           manifest$Biopsy_freemix < 0.01,]$binom.qval.rand.bx, na.rm = T) #1.847134e-05
min(manifest[manifest$Cord_est_mat_contamination == 0 & 
           manifest$Cord_freemix < 0.01,]$binom.qval.rand.cord, na.rm = T) #0.01372039

manifest[!is.na(manifest$binom.qval.rand.bx) & 
           log10(manifest$binom.qval.rand.bx) < -8 &
           manifest$Biopsy_est_mat_contamination == 0 & 
           manifest$Biopsy_freemix < 0.01,]$Placenta_PD
 
manifest[!is.na(manifest$binom.qval.rand.cord) & 
           log10(manifest$binom.qval.rand.cord) < -8 &
           manifest$Cord_est_mat_contamination == 0 & 
           manifest$Cord_freemix < 0.01,]$Cord_PD

#Using a similar number of SNPs, random sampling does not falsely call any splits using the same filtering criteria

manifest$pat.11p15.5.pop = NA

manifest[!is.na(manifest$binom.qval.11p15.5.bx) & 
           log10(manifest$binom.qval.11p15.5.bx) < -8 &
           manifest$Biopsy_est_mat_contamination == 0 & 
           manifest$Biopsy_freemix < 0.01,]$pat.11p15.5.pop =
  1 - ((1 - manifest[!is.na(manifest$binom.qval.11p15.5.bx) & 
                       log10(manifest$binom.qval.11p15.5.bx) < -8 &
                       manifest$Biopsy_est_mat_contamination == 0 &
                       manifest$Biopsy_freemix < 0.01,]$binom.p.success.11p15.5.bx) * 2)

manifest[!is.na(manifest$pat.11p15.5.pop),]

manifest$pat.11p15.5.pop.cord = NA

manifest[!is.na(manifest$binom.qval.11p15.5.cord) & 
           log10(manifest$binom.qval.11p15.5.cord) < -8 &
           manifest$Cord_est_mat_contamination == 0 & 
           manifest$Cord_freemix < 0.01,]$pat.11p15.5.pop.cord =
  1 - ((1 - manifest[!is.na(manifest$binom.qval.11p15.5.cord) & 
                       log10(manifest$binom.qval.11p15.5.cord) < -8 &
                       manifest$Cord_est_mat_contamination == 0 &
                       manifest$Cord_freemix < 0.01,]$binom.p.success.11p15.5.cord) * 2)

manifest[!is.na(manifest$pat.11p15.5.pop.cord),]

colnames(manifest)[63] = "pat.11p15.5.pop.bx"

write.table(manifest, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", col.names = T, row.names = F, sep = '\t', quote = F)

```

Double check that my assumption about about CN-LOH is correct.

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/locus_11p15.5_upd/maternal_snps")

#PD42154b3 and PD51880e are the two largest changes splits in parental alleles observed - can we see a drop in coverage there associated with the breakpoints?

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

sample = "PD42154b3"
data = read.table(paste0(sample, "_alleleCounts.tab"), header = T, sep = "\t", stringsAsFactors = F, comment.char = "")
data_flt = data[data$Chromosome == 11,]
data_flt$rollmean_cov = rollmean(data_flt$mutCountT1 + data_flt$mutCountT2, 20, fill = T)

ggplot(data_flt) +
  geom_line(mapping = aes(x = Position, y = rollmean_cov))

sample = "PD51880e"
data = read.table(paste0(sample, "_alleleCounts.tab"), header = T, sep = "\t", stringsAsFactors = F, comment.char = "")
data_flt = data[data$Chromosome == 11,]
data_flt$rollmean_cov = rollmean(data_flt$mutCountT1 + data_flt$mutCountT2, 20, fill = T)

ggplot(data_flt) +
  geom_line(mapping = aes(x = Position, y = rollmean_cov))

#neither or the largest events look like they have a stepwise change in coverage to accompany the allelic imbalance

```

####Corrections plot

```{r}

manifest = read.table("/lustre/scratch126/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

bx_qval = manifest[, c("Placenta_PD", "binom.qval.11p15.5.bx")]
cord_qval = manifest[, c("Cord_PD", "binom.qval.11p15.5.cord")]
bx_qval = bx_qval[!duplicated(bx_qval) & !is.na(bx_qval$binom.qval.11p15.5.bx),]
cord_qval = cord_qval[!duplicated(cord_qval) & !is.na(cord_qval$binom.qval.11p15.5.cord),]
colnames(bx_qval) = colnames(cord_qval) = c("ID", "qval")
all_qval = rbind(bx_qval, cord_qval)

pdf("/lustre/scratch126/casm/team274sb/to3/placenta_ml/03_DNA_analyses/figures/11p_upd_log10_q_vals.pdf", width = 7, height = 6, useDingbats = F)
hist(log10(all_qval$qval), breaks = 200, xlim = c(-50, 0), xlab = "log10(q value)", main = "")
abline(v = -8, col = "red")
dev.off()

```

####Thesis plot figure 2.9

```{r}

library(ggplot2)
library(reshape2)
library(cowplot)

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/locus_11p15.5_upd/maternal_snps")

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

sample = "PD42154b3"

plot_phased_11p_af_bx = function(sample){
  
  data = read.table(paste0(sample, "_alleleCounts.tab"), header = T, sep = "\t", stringsAsFactors = F, comment.char = "")
  data_flt = data[data$Chromosome == 11,]
  data_flt$pat_af = data_flt$mutCountT1 / (data_flt$mutCountT1 + data_flt$mutCountT2) 
  data_flt$mat_af = data_flt$mutCountT2 / (data_flt$mutCountT1 + data_flt$mutCountT2) 

  data_flt.m = reshape2::melt(data_flt[, c("Position", "pat_af", "mat_af")], id.vars = "Position")
  data_flt.m = data_flt.m[order(data_flt.m$Position),]

  p = ggplot(data_flt.m) +
        xlim(0, max(data_flt.m$Position) / 1e6) +
        ylim(0, 1) +
        theme_bw() +
        geom_rect(mapping = aes(xmin = 0, xmax = 2800000 / 1e6, ymin = 0, ymax = 1),
                  color="transparent",
                  fill = "yellow") +
        geom_point(mapping = aes(x = Position / 1e6, y = value, col = variable), pch = ".", alpha = 0.3, show.legend = F) +
        scale_color_manual(values = c(pat_af = "red", mat_af = "blue")) +
        coord_cartesian(expand = F) +
        xlab("Chromosome 11 position (Mb)") +
        ylab("Allele fraction")
  return(p)
  
}

plot_phased_11p_af_cord = function(sample){
  
  data = read.table(paste0(sample, "_alleleCounts.tab"), header = T, sep = "\t", stringsAsFactors = F, comment.char = "")
  data_flt = data[data$Chromosome == 11,]
  data_flt$pat_af = data_flt$mutCountN1 / (data_flt$mutCountN1 + data_flt$mutCountN2) 
  data_flt$mat_af = data_flt$mutCountN2 / (data_flt$mutCountN1 + data_flt$mutCountN2) 

  data_flt.m = reshape2::melt(data_flt[, c("Position", "pat_af", "mat_af")], id.vars = "Position")
  data_flt.m = data_flt.m[order(data_flt.m$Position),]

  p = ggplot(data_flt.m) +
        xlim(0, max(data_flt.m$Position) / 1e6) +
        ylim(0, 1) +
        theme_bw() +
        geom_rect(mapping = aes(xmin = 0, xmax = 2800000 / 1e6, ymin = 0, ymax = 1),
                  color="transparent",
                  fill = "yellow") +
        geom_point(mapping = aes(x = Position / 1e6, y = value, col = variable), pch = ".", alpha = 0.3, show.legend = F) +
        scale_color_manual(values = c(pat_af = "red", mat_af = "blue")) +
        coord_cartesian(expand = F) +
        xlab("Chromosome 11 position (Mb)") +
        ylab("Allele fraction")
  return(p)
  
}


p1 <- plot_phased_11p_af_bx(sample = "PD42154b3")
p2 <- plot_phased_11p_af_bx(sample = "PD51880e")
p3 <- plot_phased_11p_af_cord(sample = "PD51899c")


pdf("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/figures/11p_upd_call_examples.pdf", width = 7, height = 6, useDingbats = F)
plot_grid(p1, p2, p3, ncol = 1)
dev.off()

```


###Phased Battenberg

To minimise low fraction, false positive calls, we restricted our initial search of Battenberg to the clones >10% of cells. Now we know about the recurrence of 11p UPD, how many did we miss in those calls?

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/cnas/Battenberg")

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

txt_files = list.files(".", "_subclones.txt")

put_cnas = data.frame()
#file = "PD51880e_subclones.txt"
for(file in txt_files){
  
  data = read.table(file, header = T, sep = "\t", stringsAsFactors = F)
  sample = unlist(strsplit(file, "_"))[1]
  data$sample = sample
  
  if(sample %in% manifest[manifest$Biopsy_est_mat_contamination == 0 &
                          manifest$Biopsy_freemix < 0.01,]$Placenta_PD){
    
    if(nrow(data[data$chr == 11 & (!is.na(data$nMaj2_A) | data$nMin1_A != 1),]) > 0){
    
    put_cnas = rbind(put_cnas, data[data$chr == 11 & (!is.na(data$nMaj2_A) | data$nMin1_A != 1),])
    
  }
    
  }
  
}

put_cnas$sample #"PD42154b3" "PD51876d"  "PD51880e"  "PD51881d"  "PD51881e"  "PD51881f"  "PD51882c"  "PD51900e"  "PD51918e"

```

##Identify mutations in major clone

Binomial mixture model

```{r}

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/07_recurrent")

txt_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt")

all_subs = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  all_subs = rbind(all_subs, data)
}

all_subs_auto = manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/07_recurrent")

txt_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt")

all_subs = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  all_subs = rbind(all_subs, data)
}

all_subs_auto = all_subs[all_subs$Chr %in% c(1:22),]

source("/nfs/users/nfs_t/to3/scripts/binom_mix_model.R")

manifest$maj_clone1_sub_burden = NA
manifest$maj_clone1_sub_vaf = NA
manifest$maj_clone2_sub_burden = NA
manifest$maj_clone2_sub_vaf = NA
manifest$maj_clone_total_sub_burden = NA

set.seed(42)

#sample = "PD51877c"
for(sample in manifest$Placenta_PD){
  
  print(sample)
  #subset to sample mutations
  sample_df = all_subs_auto[all_subs_auto$Sample == sample,]
  
  #input vectors
  NV_vec = sample_df$Variant_read_depth
  NR_vec = sample_df$Total_read_depth
  
  #binomial mixture model
  res = binom_mix(NV_vec, NR_vec, nrange=1:3, mode='Full')
  
  pdf(paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/binom_mixture_model/", sample, "_binom_mix_vaf_plot.pdf"), width = 6, height = 4)
  
  p = hist(NV_vec/NR_vec, breaks = 50, xlim=c(0, 1) , col='lightgray', freq = F, xlab = "Variant Allele Frequency", main = sample, cex.main = 1)
  y_coord = max(p$density) - 0.5
  y_intv = y_coord / 5
  text(y = y_coord, x = 0.9, label = 'Data')
  segments(lwd = 2, lty = 'solid', col = 'black', y0 = y_coord + 0.25, x0 = 0.85, x1 = 0.95)

  cols=c("red", "blue", "darkgreen", "magenta", "brown")
  for (i in 1:res$n){
    depth=rpois(n=5000,lambda=median(NR_vec))
    sim_NV=unlist(lapply(depth,rbinom,n=1,prob=res$p[i]))
    sim_VAF=sim_NV/depth
    dens=density(sim_VAF)
    lines(x=dens$x,y=res$prop[i]*dens$y,lwd=2,lty='dashed',col=cols[i])
    y_coord=y_coord-y_intv/2
    text(y=y_coord,x=0.9,label=paste0("p",i,": ",round(res$p[i],digits=2)))
    segments(lwd=2,lty='dashed',col=cols[i],y0=y_coord+y_intv/4,x0=0.85,x1=0.95)
  }
  
  dev.off()
  
  maj_clone_no = which(res$p > 0.25)
  
  if(length(maj_clone_no) == 1){
    
    manifest[manifest$Placenta_PD == sample, ]$maj_clone1_sub_burden = sum(res$Which_cluster == maj_clone_no)
    manifest[manifest$Placenta_PD == sample, ]$maj_clone1_sub_vaf = max(res$p)
    
    manifest[manifest$Placenta_PD == sample, ]$maj_clone_total_sub_burden = sum(res$Which_cluster == maj_clone_no)
    
  }
  
  if(length(maj_clone_no) == 2){
    
    manifest[manifest$Placenta_PD == sample, ]$maj_clone1_sub_burden = sum(res$Which_cluster == maj_clone_no[1])
    manifest[manifest$Placenta_PD == sample, ]$maj_clone2_sub_burden = sum(res$Which_cluster == maj_clone_no[2])
    manifest[manifest$Placenta_PD == sample, ]$maj_clone1_sub_vaf = res$p[maj_clone_no[1]]
    manifest[manifest$Placenta_PD == sample, ]$maj_clone2_sub_vaf = res$p[maj_clone_no[2]]
    
    manifest[manifest$Placenta_PD == sample, ]$maj_clone_total_sub_burden = sum(res$Which_cluster == maj_clone_no[1]) + sum(res$Which_cluster == maj_clone_no[2])
    
  }
}

manifest[is.na(manifest$maj_clone1_sub_burden),]$Placenta_PD

manifest[is.na(manifest$maj_clone_total_sub_burden),]$maj_clone_total_sub_burden = 0

#add sensitivity estimate for calling mutations in dominant clone
manifest$maj_clone_sensitivity = NA

set.seed(42)
for(i in 1:nrow(manifest)){
  
  if(manifest$maj_clone_total_sub_burden[i] == 0){
    
    manifest$maj_clone_sensitivity[i] = NA
    
  } else {
    
    if(is.na(manifest$maj_clone2_sub_burden[i])){
      
      manifest$maj_clone_sensitivity[i] = mean(unlist(lapply(rpois(n=100000,lambda = manifest$biopsy_coverage[i]),function(x) pbinom(q=3, size=x, p = manifest$maj_clone1_sub_vaf[i],lower.tail = F))))
      
    } else {
      
      weighted_vaf = weighted.mean(c(manifest$maj_clone1_sub_vaf[i], manifest$maj_clone2_sub_vaf[i]), c(manifest$maj_clone1_sub_burden[i], manifest$maj_clone2_sub_burden[i]))
      manifest$maj_clone_sensitivity[i] = mean(unlist(lapply(rpois(n=100000,lambda = manifest$biopsy_coverage[i]),function(x) pbinom(q=3, size=x, p = weighted_vaf,lower.tail = F))))
      
    }
    
  }
  
}

sum(is.na(manifest$maj_clone_sensitivity)) #30
sum(!is.na(manifest$maj_clone_sensitivity)) #343

min(manifest$maj_clone_sensitivity, na.rm = T) #0.9232551
manifest[manifest$maj_clone_sensitivity == min(manifest$maj_clone_sensitivity, na.rm = T) & !is.na(manifest$maj_clone_sensitivity),]
summary(manifest$maj_clone_sensitivity)
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#0.9233  0.9983  0.9996  0.9978  0.9999  1.0000      30 

manifest$maj_clone_total_sub_burden_sens_corrected = manifest$maj_clone_total_sub_burden / manifest$maj_clone_sensitivity

manifest[is.na(manifest$maj_clone_total_sub_burden_sens_corrected),]$maj_clone_total_sub_burden_sens_corrected = 0

write.table(manifest, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", col.names = T, row.names = F, sep = '\t', quote = F)

table(manifest$clin_group)

manifest_bg = manifest[manifest$clin_group %in% 1:9,] #5 or more samples

table(manifest_bg$clin_group)
#  1   2   3   4   5   6   7   8   9 
# 17  14  32  15  16  18  14 196  40 

kruskal.test(manifest_bg$maj_clone_total_sub_burden_sens_corrected, manifest_bg$clin_group_relabel) #Kruskal-Wallis chi-squared = 4.263, df = 3, p-value = 0.2344

library(ggplot2)
ggplot(manifest_bg) +
  geom_boxplot(mapping = aes(x = clin_group, y = maj_clone_total_sub_burden))

ggplot(manifest) +
  geom_point(mapping = aes(x = biopsy_coverage, y = maj_clone_total_sub_burden))

```

##Signature extraction

###mSigHDP v2.0.1

Run using a previous version of the manifest - looks like it inclued the same samples so we are OK.

```{r}

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)
old_manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml_draft_20220711/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)

sum(!old_manifest[old_manifest$bx_mean_insert_size >= 320 & old_manifest$Other != "Membrane",]$Placenta_PD %in% manifest$Placenta_PD) #0
manifest[!manifest$Placenta_PD %in% old_manifest[old_manifest$bx_mean_insert_size >= 320 & old_manifest$Other != "Membrane",]$Placenta_PD,] #membranes only

```


####Generate input data frame

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/07_recurrent")

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)

#check that the filtered subs have at least 100 subs left per placenta
txt_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt")

all_subs = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  all_subs = rbind(all_subs, data)
}

all_subs_flt = all_subs[all_subs$Sample %in% manifest[manifest$bx_mean_insert_size >= 320 & manifest$Other != "Membrane",]$Placenta_PD, ]

all_subs_flt_uniq = all_subs_flt[!duplicated(all_subs_flt[, c("Case_ID", "mut_ID")]),]

min(table(all_subs_flt_uniq$Case_ID)) #131, we are all good to go

for(i in 1:nrow(manifest)){
  if(manifest$bx_mean_insert_size[i] >= 320 & manifest$Other[i] != "Membrane"){
    system(paste0("cp ", manifest$Placenta_PD[i], ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt /lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/signature_extraction/mSigHDP/01_Input"))
  }
}

```

Function to generate the counts per trinucleotide context

```{r}

count_muts_96 = function(mutations, genomeFile = "/nfs/cancer_ref01/Homo_sapiens/37/genome.fa"){
  library("GenomicRanges")
  library("Rsamtools")
  library("MASS")
  colnames(mutations) = c("chr","pos","ref","mut")
  mutations = mutations[(mutations$ref %in% c("A","C","G","T")) & (mutations$mut %in% c("A","C","G","T")) & mutations$chr %in% c(1:22,"X","Y"),]
  mutations$trinuc_ref = as.vector(scanFa(genomeFile, GRanges(mutations$chr, IRanges(as.numeric(mutations$pos)-1,
                                                                                     as.numeric(mutations$pos)+1))))
 
  # 2. Annotating the mutation from the pyrimidine base
  ntcomp = c(T="A",G="C",C="G",A="T")
  mutations$sub = paste(mutations$ref,mutations$mut,sep=">")
  mutations$trinuc_ref_py = mutations$trinuc_ref
  for (j in 1:nrow(mutations)) {
    if (mutations$ref[j] %in% c("A","G")) { # Purine base
      mutations$sub[j] = paste(ntcomp[mutations$ref[j]],ntcomp[mutations$mut[j]],sep=">")
      mutations$trinuc_ref_py[j] = paste(ntcomp[rev(strsplit(mutations$trinuc_ref[j],split="")[[1]])],collapse="")
    }
  }
  freqs = table(paste(mutations$sub,paste(substr(mutations$trinuc_ref_py,1,1),substr(mutations$trinuc_ref_py,3,3),sep="-"),sep=","))
  sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
  ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
  full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")
  freqs_full = freqs[full_vec]; freqs_full[is.na(freqs_full)] = 0; names(freqs_full) = full_vec
  return(freqs_full)
}

```

Generate input text files - unique subs per patient to avoid double counting.

```{r}

library(cosmicsig)

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/signature_extraction/mSigHDP/01_Input")

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)

mut_txt_files = list.files('.', pattern = '.caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt')

#list of unique subs per placenta
all_subs = data.frame()
for(file in mut_txt_files){
  sample = unlist(strsplit(file, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  all_subs = rbind(all_subs, data)
}

All_var = all_subs[, c(1, 3:6)][!duplicated(all_subs[, c(1, 3:6)]),]
colnames(All_var) = c("sample", "chr","pos","ref","mut")

write.table(All_var, 'uniq_subs_placenta_20220628.txt', col.names = T, row.names = F, sep = '\t', quote = F)

samples=unique(All_var$sample)
mut_count = matrix(0,nrow=length(samples),ncol=96)

#create dataframe of context counts per patient so as not to duplicate the counting of individual subs
for (k in 1:length(samples)){
  mut_count[k,]=count_muts_96(All_var[All_var$sample==samples[k],2:5])
  print(k)
}

mut_count = data.frame(mut_count)
row.names(mut_count) = samples

sub_vec = c("C>A","C>G","C>T","T>A","T>C","T>G")
ctx_vec = paste(rep(c("A","C","G","T"),each=4),rep(c("A","C","G","T"),times=4),sep="-")
full_vec = paste(rep(sub_vec,each=16),rep(ctx_vec,times=6),sep=",")

names(mut_count) = full_vec

colnames(mut_count) = paste0(substr(colnames(mut_count), 5, 5), substr(colnames(mut_count), 1, 1), substr(colnames(mut_count), 7, 7), substr(colnames(mut_count), 3, 3))

mut_count.t = as.data.frame(t(mut_count)) 

sig_order <- names(cosmicsig::COSMIC_v3.2$signature$GRCh37$SBS96[  , "SBS1"]) #order of contexts expected trinucleotide context followed by mutant allele

mut_count.t = mut_count.t[sig_order,]

write.table(mut_count.t, "trinuc_mat_input.txt", col.names = T, row.names = T, sep = '\t', quote = F)

```

####Run

```{bash}

export PATH=/software/R-4.1.0/bin:${PATH}
export R_HOME=$(R RHOME)
export R_LIBS_USER="/lustre/scratch119/casm/team274sb/to3/placenta_ml/ZZ_Libraries/R4.1_msighdp"

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/signature_extraction/mSigHDP/01_Input

bsub -o ../02_Output/%J.out -e ../02_Output/%J.err -q normal -R'select[mem>100000] rusage[mem=100000]' -M100000 -n 20 /software/R-4.1.0/bin/Rscript /lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/signature_extraction/mSigHDP/msighdp_standard_parameters_20220515.R

#bsub -Is -q normal -R "select[mem>100000] rusage[mem=100000] span[hosts=1]" -M 100000 -n 20 R


```

```{r}

#/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/signature_extraction/mSigHDP/msighdp_standard_parameters_20220515.R

library(mSigHdp)
library(ICAMS)
library(cosmicsig)

input_data = read.table("trinuc_mat_input.txt", header = T, sep = '\t', stringsAsFactors = F)
input_data <- ICAMS::as.catalog(input_data, catalog.type = "counts")

results <- mSigHdp::RunHdpxParallel(
  input.catalog        = input_data,
  out.dir              = "../02_Output",
  num.child.process    = 20, 
  CPU.cores            = 20,
  seedNumber           = 42,
  K.guess              = 10, #3 found in the original study, increased for large cohort size
  burnin               = 5000,
  burnin.multiplier    = 20,
  post.n               = 200, 
  post.space           = 100, 
  multi.types          = FALSE, #treated as one "tumour" type 
  overwrite            = TRUE,
  gamma.alpha          = 1,
  gamma.beta           = 20, 
  high.confidence.prop = 0.9,
  checkpoint           = TRUE,
  verbose              = TRUE) 

save(results, file = "../02_Output/multi.chains.Rdata")

```

####Deconvolute signature

```{r}

#Deconvolute HDP signatures into reference signatures and arrive at final set
options(stringsAsFactors = F)
#library(hdp)
library(RColorBrewer)
library(lsa)
library(lattice)

setwd('/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/signature_extraction/mSigHDP/02_Output')

mut.cols = rep(c("dodgerblue","black","red","grey70","olivedrab3","plum2"),each=16)

#Load mSigHdp signatures
hdp_sigs = read.table("extracted.signatures.csv", header = T, sep = ',')
row.names(hdp_sigs) = paste0(substr(hdp_sigs$Trinucleotide, 1, 1), "[", hdp_sigs$Mutation.type, "]", substr(hdp_sigs$Trinucleotide, 3, 3))
ref = read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/COSMIC_v3.2_SBS_GRCh37.txt', header = T, sep = '\t', stringsAsFactors = F, row.names = 1)

#Assess cosine similarities for all unassigned signatures (X0, N1, N2)
cosine_matrix=data.frame(matrix(nrow=ncol(as.matrix(hdp_sigs[, c('hdp.1')])), ncol=ncol(ref)))
rownames(cosine_matrix)=c('hdp.1')
colnames(cosine_matrix)=colnames(ref)

for (n in 1:nrow(cosine_matrix)) {
  for (m in 1:ncol(cosine_matrix)) {
    cosine_matrix[n,m] <- cosine(x=hdp_sigs[,rownames(cosine_matrix)[n]],
                                 y=ref[,colnames(cosine_matrix)[m]])
  }
}

write.table(cosine_matrix, "Cosine_similarities_unassigned_to_priors.txt",sep="\t",quote=F)

#plot output
pdf("cosine_similarities_unassigned_to_priors.pdf", height=5, width=15)
color.palette = colorRampPalette(c("white", "orange", "purple"))
levelplot(t(cosine_matrix[dim(cosine_matrix)[1]:1,]),col.regions=color.palette, aspect="fill", scales=list(x=list(rot=90)))
dev.off()

gdsigs = c("SBS1", "SBS5", "SBS18")

signatures=t(ref[,gdsigs])
sample_list= c('hdp.1', 'hdp.1') #keep twice to avoid accidental conversion to vector
profiles=hdp_sigs[,sample_list]

signature_fraction = matrix(NA, nrow=nrow(signatures),ncol=length(sample_list))
rownames(signature_fraction) = rownames(signatures)
colnames(signature_fraction) = sample_list
maxiter <- 1000

for (j in 1:length(sample_list)) {
  freqs = profiles[,j]
  freqs[is.na(freqs)] = 0
  # EM algorithm with to estimate the signature contribution
  alpha = runif(nrow(signatures)); alpha=alpha/sum(alpha) # Random start (seems to give ~identical results)
  for (iter in 1:maxiter) {
    contr = t(array(alpha,dim=c(nrow(signatures),96))) * t(signatures)
    probs = contr/array(rowSums(contr),dim=dim(contr))
    probs = probs * freqs
    old_alpha = alpha
    alpha = colSums(probs)/sum(probs)
    if (sum(abs(alpha-old_alpha))<1e-5) {
      break
    }
  }
  # Saving the signature contributions for the sample
  print(j/length(sample_list))
  signature_fraction[,j] = alpha
}

#Plot initial deconvolution and save results
pdf("Deconvolution_msighdp_sigs_after_prior_R1.pdf", height=5, width=10)
color.palette = colorRampPalette(c("white", "orange", "purple"))
levelplot((signature_fraction[nrow(signature_fraction):1,]),col.regions=color.palette, aspect="fill", scales=list(x=list(rot=90)))
dev.off()

#write.table(signature_fraction[,'N1'], "hdp_unassigned_sig_broken_down_into_cosmic_sigs.txt", sep="\t", col.names=T, row.names = T, quote=F)

sigs_deconv_R2=list()

for(n in 1:length(sample_list)){
  sigs_deconv_R2[[n]]=rownames(signature_fraction)[signature_fraction[,n]>0]
}

names(sigs_deconv_R2)=colnames(signature_fraction)

#No signature has one major contributor only

sigs_to_deconv=names(sigs_deconv_R2)[unlist(lapply(sigs_deconv_R2,length))>1]

ref_sigs_R2=sort(unique(unlist(sigs_deconv_R2)))
signature_fractionR2=matrix(NA,ncol=length(sigs_to_deconv),nrow=length(ref_sigs_R2))
rownames(signature_fractionR2)=ref_sigs_R2
colnames(signature_fractionR2)=sigs_to_deconv

#repeat the deconvolution with the identified constitutive signatures
n = 1
for(s in sigs_to_deconv){
  gdsigs <- sigs_deconv_R2[[s]]
  signatures <- t(ref[,gdsigs])
  
  signature_fraction = matrix(NA,nrow=nrow(signatures),ncol=length(sample_list))
  rownames(signature_fraction) = rownames(signatures)
  colnames(signature_fraction) = sample_list
  maxiter <- 1000
  
  freqs = profiles[,s]
  freqs[is.na(freqs)] = 0
  
  alpha = runif(nrow(signatures)); alpha=alpha/sum(alpha) # Random start (seems to give ~identical results)
  for (iter in 1:maxiter) {
    contr = t(array(alpha,dim=c(nrow(signatures),96))) * t(signatures)
    probs = contr/array(rowSums(contr),dim=dim(contr))
    probs = probs * freqs
    old_alpha = alpha
    alpha = colSums(probs)/sum(probs)
    if (sum(abs(alpha-old_alpha))<1e-5) {
      break
    }
  }
  # Saving the signature contributions for the sample
  signature_fractionR2[gdsigs,n] = alpha
  n=n+1
  reconsbs <- rep(0,96)
  for (g in gdsigs) {
    reconsbs=reconsbs+(ref[,g]*alpha[g])
  }
  cosine_reconst=cosine(x=reconsbs, y=hdp_sigs[,s])
  print(paste0(s,": ",cosine_reconst))
  pdf(paste0("HDP_",s,"_reconstitution2.pdf"))
  par(mfrow=c(length(alpha)+2,1))
  par(mar=c(1,2,4,1))
  barplot(hdp_sigs[,s], col=mut.cols, main=paste0("HDP ",s),names.arg="")
  barplot(reconsbs, col=mut.cols, main=paste0("Reconstituted ",s," cosine similarity to original: ", round(cosine_reconst, digits=2)))
  for (g in gdsigs) {
    barplot(ref[,g], col=mut.cols, main=paste0("PCAWG ", g, " accounts for ", round(alpha[g], digits=2)))
  }
  dev.off()
}

```

####Reassign known signatures back to samples

#####deconstructSigs

```{r}

library(deconstructSigs)
library(VariantAnnotation)
library(BSgenome.Hsapiens.UCSC.hg19)
library(BSgenome)

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/signature_extraction/deconstructSigs")

input = read.table('/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/signature_extraction/mSigHDP/01_Input/uniq_subs_placenta_20220628.txt', header = T, sep = '\t', stringsAsFactors = F)
prior_counts = read.table('/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/signature_extraction/mSigHDP/01_Input/trinuc_mat_input.txt', header = T, sep = '\t', stringsAsFactors = F)

names(input) = c('id', "chr", "pos", "ref", "alt")
input$chr = paste0('chr', input$chr)
sigs.input <- mut.to.sigs.input(mut.ref = input, 
                                sample.id = "id", 
                                chr = "chr", 
                                pos = "pos", 
                                ref = "ref", 
                                alt = "alt", bsg = BSgenome.Hsapiens.UCSC.hg19)

sum(rowSums(sigs.input) != colSums(prior_counts)) #0 matches previous
ref = t(read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/COSMIC_v3.2_SBS_GRCh37.txt', header = T, sep = '\t', stringsAsFactors = F, row.names = 1))
ref = as.data.frame(ref[c("SBS1", "SBS5", "SBS18"),])

#default
sigs = data.frame()
for(placenta in row.names(sigs.input)){
  output = whichSignatures(tumor.ref = sigs.input, 
                         sample.id = placenta,
                         signatures.ref = ref, 
                         contexts.needed = TRUE,
                         tri.counts.method = 'default', 
                         signature.cutoff = 0.01)
  sigs = rbind(sigs, output$weights)
}

write.table(sigs, "placenta_sigs_mapped_20220628.txt", col.names = T, row.names = T, sep = '\t', quote = F)

```

#####Plot

```{r}

#setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/signature_extraction/deconstructSigs")
setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml_draft_20220711/03_DNA_analyses/signature_extraction/deconstructSigs")

sigs = read.table("placenta_sigs_mapped_20220628.txt", header = T, sep = '\t', stringsAsFactors = F)
#manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest$case.id = substr(manifest$Placenta_PD, 0, 7)

sigs$Unknown = 1 - round(rowSums(sigs), digits = 3)
rowSums(sigs) #now all equal 1

sigs$clin_group = NA
sigs$clin_group2 = NA
for(i in 1:nrow(sigs)){
  sigs$clin_group[i] = unique(manifest[manifest$case.id == row.names(sigs)[i],]$clin_group_relabel)
  sigs$clin_group2[i] = unique(manifest[manifest$case.id == row.names(sigs)[i],]$clin_group)
}

sigs$clin_group2 = ifelse(sigs$clin_group2 == "8", "Normal", "Abnormal")
sigs$clin_group2 = factor(sigs$clin_group2, levels = c("Normal", "Abnormal"))
sigs$clin_group = factor(sigs$clin_group, levels = c("healthy_control", "big_baby_placenta", "small_baby", "pre_eclampsia"))

library(ggplot2)
library(ggbeeswarm)
library(ggpubr)
library(reshape2)

sigs$case.id = row.names(sigs)

summary(sigs$SBS18)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#0.2458  0.4630  0.5133  0.5179  0.5701  0.8835
 
sigs.m = reshape2::melt(sigs[, c(1:4,6,7)], id.vars = c("case.id", "clin_group2"))
#sigs.m$clin_group = factor(sigs.m$clin_group, levels = c("healthy_control", "big_baby_placenta", "small_baby", "pre_eclampsia"))
sigs.m$case.id = factor(sigs.m$case.id, levels = sigs[order(sigs$SBS18),]$case.id)
sigs.m$variable = factor(sigs.m$variable, levels = c("Unknown", "SBS18", "SBS5", "SBS1"))

pdf("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/figures/bulk_placenta_sigs_thesis_20230216.pdf", width = 6, height = 4)
ggplot(sigs.m) +
  geom_bar(mapping = aes(x = case.id, y = value, fill = variable), position = "stack", stat = "identity", show.legend = F, width = 1) +
  facet_grid(. ~ clin_group2, scales = "free_x", space = "free_x") +
  theme_bw() +
  coord_cartesian(expand = F) +
  theme(axis.text = element_blank(), ) +
  xlab("Placentas") +
  ylab("Proportion") +
  scale_fill_manual(values = c(SBS1 = "#666666", SBS5 = "white", SBS18 = "#1b9dd9", Unknown = "black"))
dev.off()

```

#####Plot fraction compared to normal adult tissues

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml_draft_20220711/03_DNA_analyses/signature_extraction/deconstructSigs")

sigs = read.table("placenta_sigs_mapped_20220628.txt", header = T, sep = '\t', stringsAsFactors = F)
#manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220610.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest$case.id = substr(manifest$Placenta_PD, 0, 7)

sigs$Unknown = 1 - round(rowSums(sigs), digits = 3)
rowSums(sigs) #now all equal 1

sigs$clin_group = NA
for(i in 1:nrow(sigs)){
  sigs$clin_group[i] = unique(manifest[manifest$case.id == row.names(sigs)[i],]$clin_group_relabel)
}

sigs$clin_group = factor(sigs$clin_group, levels = c("healthy_control", "big_baby_placenta", "small_baby", "pre_eclampsia"))
sigs$case.id = row.names(sigs)

pb_meta = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/panbody_ref_metadata.txt", header = T, sep = "\t", stringsAsFactors = F)
pb_meta_flt = pb_meta[pb_meta$SBS_Count >= 100 & rowSums(pb_meta[,25:39]) != 0,] #have HDP results and at least 100 substitutions
nrow(pb_meta_flt) #408

pb_meta[pb_meta$SBS_Count >= 100 & rowSums(pb_meta[,25:39]) == 0,]$Sample
pb_meta[pb_meta$SBS_Count < 100 & rowSums(pb_meta[,25:39]) != 0,]$Sample
colnames(pb_meta[,25:39])

summary(pb_meta_flt$HDP_SBS18)
#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#0.00000 0.01855 0.03911 0.05347 0.08947 0.24267

placenta_df = sigs[, c("case.id", "SBS18")]
placenta_df$histo = "placenta"
pb_df = pb_meta_flt[, c("Sample", "HDP_SBS18", "TissueType4")]
colnames(pb_df) = colnames(placenta_df)

combined_df = rbind(placenta_df, pb_df)

bg_tissues = names(table(combined_df$histo))[table(combined_df$histo) >= 5]

bg_combined_df = combined_df[combined_df$histo %in% bg_tissues,]
median_sbs18_df = aggregate(SBS18 ~ histo, bg_combined_df, median)
bg_combined_df$histo = factor(bg_combined_df$histo, levels = median_sbs18_df[order(median_sbs18_df$SBS18),]$histo)


library(ggplot2)
library(ggbeeswarm)
library(ggpubr)

pdf("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/figures/bulk_placenta_vs_panbody_sigs_thesis_20230216.pdf", width = 6, height = 4)
ggplot(bg_combined_df) +
  geom_boxplot(mapping = aes(x = histo, y = SBS18)) +
  theme_bw() +
  xlab("") +
  ylab("Proportion of susbtitutions attributed to SBS18") +
  theme(axis.text.x = element_text(vjust = 0.5, hjust = 1, angle = 90))
dev.off()

```

##Plot sub type & mutational spectra by VAF bin

```{r}

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

#read in subs
setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/07_recurrent")

txt_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt")

all_subs = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  all_subs = rbind(all_subs, data)
}

all_vars_bx_flt = all_subs[all_subs$sample %in% manifest[manifest$Other != "Membrane",]$Placenta_PD, ]

#get aggregate VAF for shared variants
avg_vaf = aggregate(VAF ~ mut_ID, all_vars_bx_flt, median)

all_vars_bx_flt$avg_vaf = NA

for(i in 1:nrow(all_vars_bx_flt)){
  
  all_vars_bx_flt$avg_vaf[i] = avg_vaf[avg_vaf$mut_ID == all_vars_bx_flt$mut_ID[i],]$VAF
  
}

#all muts are unique per placenta so this should work

all_vars_bx_flt = all_vars_bx_flt[all_vars_bx_flt$Chr %in% c(1:22),]
all_vars_bx_flt = all_vars_bx_flt[!duplicated(all_vars_bx_flt$mut_ID),]

#VAF bins
all_vars_bx_flt$vaf_bin = NA
#all_vars_bx_flt[all_vars_bx_flt$VAF < 0.1, ]$vaf_bin = "0_0.1"
#all_vars_bx_flt[all_vars_bx_flt$VAF >= 0.1 & all_vars_bx_flt$VAF < 0.25, ]$vaf_bin = "0.1_0.25"
#all_vars_bx_flt[all_vars_bx_flt$VAF >= 0.25, ]$vaf_bin = "0.25_1"

all_vars_bx_flt[all_vars_bx_flt$VAF < 0.1, ]$vaf_bin = "0_0.1"
all_vars_bx_flt[all_vars_bx_flt$VAF >= 0.1, ]$vaf_bin = "0.1_1"

all_subs_bx_flt = all_vars_bx_flt

ntcomp <- c(T="A",G="C",C="G",A="T")
all_subs_bx_flt$sub <- paste(all_subs_bx_flt$Wildtype, all_subs_bx_flt$Mutant,sep=">")
for (j in 1:nrow(all_subs_bx_flt)) {
  if (all_subs_bx_flt$Wildtype[j] %in% c("A","G")) { # Purine base
    all_subs_bx_flt$sub[j] = paste(ntcomp[all_subs_bx_flt$Wildtype[j]],ntcomp[all_subs_bx_flt$Mutant[j]],sep=">")
  }
}

freq = table(all_subs_bx_flt$sub, all_subs_bx_flt$vaf_bin)
freq_prop = data.frame(t(t(freq) / colSums(freq)))
names(freq_prop) = c("sub_type", "vaf", "prop")
#freq_prop$vaf = factor(freq_prop$vaf, levels = c("0_0.1", "0.1_0.25", "0.25_1"))
freq_prop$vaf = factor(freq_prop$vaf, levels = c("0_0.1", "0.1_1"))

library(reshape2)
library(ggplot2)
library(ggpubr)

ggplot(freq_prop) + 
  geom_bar(mapping = aes(x = vaf, y = prop, fill = sub_type), stat = "identity", position = "stack") +
  coord_cartesian(expand = F) +
  theme_pubr() +
  theme(panel.background = element_rect(colour = "black"))

colSums(table(all_subs_bx_flt$sub, all_subs_bx_flt$vaf_bin))

library(MutationalPatterns)
library(GenomeInfoDb)
library("BSgenome.Hsapiens.UCSC.hg19")

subs_input = all_subs_bx_flt[, c("Chr", "Position", "Wildtype", "Mutant", "vaf_bin")]
names(subs_input) = c("Chr", "Position", "Ref", "Alt", "filter")
subs_input$Chr = paste0("chr", subs_input$Chr)

grange_obj = makeGRangesListFromDataFrame(subs_input, 
                                          split.field = "filter", 
                                          keep.extra.columns = T, 
                                          ignore.strand = T, 
                                          seqnames.field = "Chr",
                                          start.field = "Position", 
                                          end.field = "Position")
GenomeInfoDb::genome(grange_obj) = "hg19"
seqlengths(grange_obj) <- seqlengths(BSgenome.Hsapiens.UCSC.hg19@seqinfo)[1:22]

type_occurrences <- mut_type_occurrences(grange_obj, "hg19")
mut_mat <- mut_matrix(vcf_list = grange_obj, ref_genome = "hg19")

plot_96_profile(mut_mat)

input = all_subs_bx_flt[, c("vaf_bin", "Chr", "Position", "Wildtype", "Mutant")]

names(input) = c('id', "chr", "pos", "ref", "alt")
input$chr = paste0('chr', input$chr)
sigs.input <- mut.to.sigs.input(mut.ref = input, 
                                sample.id = "id", 
                                chr = "chr", 
                                pos = "pos", 
                                ref = "ref", 
                                alt = "alt", bsg = BSgenome.Hsapiens.UCSC.hg19)

ref = t(read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/COSMIC_v3.2_SBS_GRCh37.txt', header = T, sep = '\t', stringsAsFactors = F, row.names = 1))
ref = as.data.frame(ref[c("SBS1", "SBS5", "SBS18"),])

#default
vaf_sigs = data.frame()
for(vaf in row.names(sigs.input)){
  output = whichSignatures(tumor.ref = sigs.input, 
                         sample.id = vaf,
                         signatures.ref = ref, 
                         contexts.needed = TRUE,
                         tri.counts.method = 'default', 
                         signature.cutoff = 0.01)
  vaf_sigs = rbind(vaf_sigs, output$weights)
}

vaf_sigs
#           SBS1      SBS5     SBS18
#0.1_1 0.05180819 0.4239939 0.5241979
#0_0.1 0.00000000 0.4254212 0.5745788

```

##Stats for thesis

```{r}

library(ggplot2)
library(lme4)

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

manifest = read.table("/lustre/scratch126/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

nrow(manifest[manifest$Other != "Membrane",]) #371
length(unique(manifest[manifest$Maternal_PD != "-",]$Cord_PD)) #109
summary(manifest$biopsy_coverage)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#26.33   37.11   41.58   41.77   46.29   63.26 

summary(manifest[manifest$Other != "Membrane",]$sub_burden)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#31.0   104.0   128.0   130.2   152.0   237.0 

summary(manifest[manifest$Other != "Membrane",]$indel_burden)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#0.000   7.000   9.000   9.938  13.000  25.000

summary(manifest[manifest$Other != "Membrane",]$sub_median_VAF)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#0.1053  0.1988  0.2410  0.2402  0.2793  0.4524

summary(manifest[manifest$Other == "Membrane",]$sub_burden)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#8.00   16.25   24.50   24.50   32.75   41.00 

summary(manifest[manifest$Other == "Membrane",]$indel_burden)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#2.00    3.25    4.50    4.50    5.75    7.00 

summary(manifest[manifest$Other == "Membrane",]$sub_median_VAF)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#0.1122  0.1126  0.1130  0.1130  0.1134  0.1138

summary(manifest[manifest$Other != "Membrane",]$maj_clone_total_sub_burden_sens_corrected)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#0.00   35.01   56.38   57.81   78.09  186.09 

#manifest[manifest$maj_clone_total_sub_burden_sens_corrected > 180,]

#clonal muts plot
table(manifest[manifest$Other != "Membrane",]$clin_group)


manifest_bg = manifest[manifest$clin_group %in% 1:9 & manifest$Other != "Membrane",] #5 or more samples

table(manifest$clin_group)
#1    10     2 2,4,7     3   3,4     4     5     6     7     8     9 
#17     4    14     3    32     4    15    16    18    14   196    40 

table(manifest_bg$clin_group)
# 1   2   3   4   5   6   7   8   9 
#16  14  31  15  16  18  14 196  40 

kruskal.test(manifest_bg$maj_clone_total_sub_burden_sens_corrected, manifest_bg$clin_group) #Kruskal-Wallis chi-squared = 12.111, df = 8, p-value = 0.1463

summary(manifest_bg$maj_clone_total_sub_burden / manifest_bg$sub_auto_burden)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#0.0000  0.3268  0.4932  0.4714  0.6513  1.0000 

library(ggplot2)

pdf("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/figures/clonal_subs_thesis_20230216.pdf", width = 4, height = 4)
ggplot(manifest_bg) +
  geom_boxplot(mapping = aes(x = clin_group, y = maj_clone_total_sub_burden)) +
  theme_bw() +
  ylab("Major clone substitution burden") +
  xlab("Clinical group")
dev.off()

cor(manifest$biopsy_coverage, manifest$maj_clone_total_sub_burden) #0.1248709

ggplot(manifest) +
  geom_point(mapping = aes(x = biopsy_coverage, y = maj_clone_total_sub_burden))

library(ggplot2)

ggplot(manifest) +
  geom_boxplot(mapping = aes(x = clin_group, y = maj_clone_total_sub_burden / sub_auto_burden))

plot(manifest$clin_group, manifest$maj_clone_total_sub_burden_sens_corrected / manifest$sub_burden)




case_id_clin_group.lme =lmer(formula = "maj_clone_total_sub_burden_sens_corrected ~ 1 + clin_group + (1 | Cord_PD)", data = manifest_bg, REML = F)
summary(case_id_clin_group.lme)

case_id_clin_group.lme =lmer(formula = "maj_clone_total_sub_burden_sens_corrected ~ 1 + clin_group + (1 | Cord_PD)", data = manifest_bg, REML = F)
case_id.lme = lmer(formula = "maj_clone_total_sub_burden_sens_corrected ~ 1 + (1 | Cord_PD)", data = manifest_bg, REML = F)
anova(case_id_clin_group.lme, case_id.lme)

null.lm = lm(formula = "maj_clone_total_sub_burden_sens_corrected ~ 1", data = manifest_bg)
case_id.lme = lmer(formula = "maj_clone_total_sub_burden_sens_corrected ~ 1 + (1 | Cord_PD)", data = manifest_bg, REML = F)
case_id_clin_group.lme =lmer(formula = "maj_clone_total_sub_burden_sens_corrected ~ 1 + clin_group + (1 | Cord_PD)", data = manifest_bg, REML = F)

anova(null.lm, case_fixed.lme, case_fixed_clin_group.lme)
#                         npar    AIC    BIC  logLik deviance Chisq Df Pr(>Chisq)
#case_fixed.lme               3 3565.6 3577.2 -1779.8   3559.6                    
#case_fixed_clin_group.lme   11 3570.2 3613.0 -1774.1   3548.2 11.33  8     0.1837

```

##Regression analysis of burdens by clinical group

```{r}

library(lme4)
library(MASS)

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest_flt = manifest[manifest$Other != "Membrane",]

glmer.nb(sub_highVAF_burden ~ clin_group_relabel + biopsy_coverage + (1 | Cord_PD), data = manifest_flt)

```


##Selection analysis

Regions I have masked:
/lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/caveman/flagging/simple_repeats.bed 
/lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/caveman/flagging/hi_seq_depth.bed /lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/caveman/flagging/centromeric_repeats.bed
/lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/LCR-hs37d5.bed 
/lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/hg19-blacklist.v2.chr.renamed.bed

###trinucleotide dNdS

####Modify refCDS file

```{r}

library(GenomicRanges)
library(dndscv)
library(Rsamtools)
library(dplyr)

data("refcds_hg19", package = "dndscv")
gene_list = sapply(RefCDS, function(x) x$gene_name)
genes = gene_list
chrs   = unlist(sapply(RefCDS,function(x) rep(x$chr,nrow(x$intervals_cds))))

starts = unlist(sapply(RefCDS,function(x) x$intervals_cds[,1]))
ends   = unlist(sapply(RefCDS,function(x) x$intervals_cds[,2]))
bed    = data.frame(chr=chrs,start=starts,end=ends)
bed$start = bed$start - 4 # 1-based
bed$end   = bed$end   + 4
refcds_gr = GRanges(bed$chr, IRanges(start=bed$start,end=bed$end))

#load masks
caveman_1 = read.table("/lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/caveman/flagging/simple_repeats.bed", header = F, sep = '\t', stringsAsFactors = F)
caveman_2 = read.table("/lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/caveman/flagging/hi_seq_depth.bed", header = F, sep = '\t', stringsAsFactors = F)
caveman_3 = read.table("/lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/caveman/flagging/centromeric_repeats.bed", header = F, sep = '\t', stringsAsFactors = F)
lcr_mask = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/LCR-hs37d5.bed", header = F, sep = '\t', stringsAsFactors = F)
encode_mask = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/99_Reference_data/hg19-blacklist.v2.chr.renamed.bed", header = F, sep = '\t', stringsAsFactors = F)[, 1:3]

#combine masks
bed_mask       = rbind(caveman_1, caveman_2, caveman_3, lcr_mask, encode_mask)
masks_gr       = GRanges(bed_mask[,1],IRanges(bed_mask[,2]+1,bed_mask[,3])) # make it 1-based

# Intersect masks with the RefCDS intervals
intersection = as.data.frame(GenomicRanges::intersect(refcds_gr,masks_gr))

# Expand intersection into 1-base intervals
chrs = unlist(sapply(c(1:nrow(intersection)), function(x) rep(intersection[x,"seqnames"], intersection[x,"end"]-intersection[x,"start"]+1)))
poss = unlist(sapply(c(1:nrow(intersection)), function(x) c(intersection[x,"start"]:intersection[x,"end"])))
to_filter = data.frame(chr=chrs,pos=poss)

# For each mask site, get the base at that site:
to_filter$ref = as.vector(scanFa("/lustre/scratch119/casm/team78pipelines/reference/human/GRCh37d5/genome.fa", GRanges(to_filter$chr, IRanges(start=to_filter$pos, end=to_filter$pos))))

# Expand to all possible subs (and then remove cases with N in ref or with ref = mut)
chrs = rep(to_filter$chr,each=4)
poss = rep(to_filter$pos,each=4)
refs = rep(to_filter$ref,each=4)
muts = rep(c("A","C","G","T"),nrow(to_filter))
mut_matrix = data.frame(sampleID="none",chr=chrs,pos=poss,ref=refs,mut=muts)
mut_matrix = mut_matrix[which(mut_matrix$ref %in% c("A","C","G","T")),]
mut_matrix = mut_matrix[which(mut_matrix$ref != mut_matrix$mut),]
mut_matrix$chr = gsub("chr","",mut_matrix$chr)

# The run dndscv with that mutation matrix
dndsout = dndscv(mut_matrix, cv=NULL, max_coding_muts_per_sample=Inf, max_muts_per_gene_per_sample=Inf, outmats=T)

# Modify the RefCDS and save it
for (j in 1:length(RefCDS)) {
  RefCDS[[j]]$L = RefCDS[[j]]$L - dndsout$N[,,j]
}

output_file = "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/RefCDS_GRCh37_masked.Rdat"

save(RefCDS, file=output_file)
save.image("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/IMAGE.RefCDS_GRCh37_masked.Rdat")
# Done!

```

####run dNdS

```{r}

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

#read in subs
setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/07_recurrent")

txt_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt")

all_subs = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  all_subs = rbind(all_subs, data)
}

#read in indels
setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/08_recurrent")

txt_files = list.files(".", ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.recurrent_vars.filtered.txt")

all_indels = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.recurrent_vars.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  data = data[data$VAF >= 0.1,] #added 16/6/22 due to an excess of low VAF calls
  all_indels = rbind(all_indels, data)
}

#combine
all_vars_bx = rbind(all_subs, all_indels)
all_vars_bx_flt = all_vars_bx[all_vars_bx$sample %in% manifest[manifest$Other != "Membrane",]$Placenta_PD, ]

all_vars_bx_flt_uniq = all_vars_bx_flt[!duplicated(all_vars_bx_flt[, c(1, 3, 4, 5, 6)]), c(1, 3, 4, 5, 6)]
#nrow(all_vars_bx_flt_uniq[all_vars_bx_flt_uniq$Wildtype == TRUE,]) #0
#nrow(all_vars_bx_flt_uniq[all_vars_bx_flt_uniq$Mutant == TRUE,]) #0

names(all_vars_bx_flt_uniq) = c("sampleID", "chr", "pos", "ref", "mut")

write.table(all_vars_bx_flt_uniq, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/all_placenta_muts_uniq_20220713.txt", col.names = T, row.names = F, sep = '\t', quote = F)

library("dndscv")
library("seqinr")
library("Biostrings")
library("MASS")
library("GenomicRanges")
library(dplyr)

dnds_input <- all_vars_bx_flt_uniq

dndsout = dndscv(dnds_input, outmats = T)

dndsout$globaldnds #theta = 0.637 subs, indels 21.5
#     name      mle     cilow   cihigh
#wmis wmis 1.188118 0.9472763 1.490193
#wnon wnon 1.867743 1.2092228 2.884883
#wspl wspl 1.634753 0.8327936 3.208978
#wtru wtru 1.800057 1.2255772 2.643821
#wall wall 1.236594 0.9890311 1.546123

print(head(dndsout$sel_cv), digits = 3) #none significant after correction for multiple hypothesis testing

dndsout = dndscv(dnds_input,
                 refdb = "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/RefCDS_GRCh37_masked.Rdat",
                 cv = NULL,
                 max_coding_muts_per_sample = Inf,
                 max_muts_per_gene_per_sample = Inf,
                 outmats = T)

dndsout$globaldnds #theta = 0.617 subs, indels 21.5
#     name      mle     cilow   cihigh
#wmis wmis 1.185471 0.9451446 1.486907
#wnon wnon 1.858670 1.2038034 2.869781
#wspl wspl 1.617866 0.8244857 3.174696
#wtru wtru 1.788523 1.2180747 2.626123
#wall wall 1.233579 0.9866062 1.542376

print(head(dndsout$sel_cv), digits = 3) #none significant after correction for multiple hypothesis testing

saveRDS(dndsout, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/dndsoutput_20220713.rds")

#trial for each substitution type in turn
##C>A
dnds_input <- all_vars_bx_flt_uniq[(all_vars_bx_flt_uniq$ref == "C" & all_vars_bx_flt_uniq$mut == "A") |
                                    (all_vars_bx_flt_uniq$ref == "G" & all_vars_bx_flt_uniq$mut == "T"),]

dndsout = dndscv(dnds_input,
                 refdb = "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/RefCDS_GRCh37_masked.Rdat",
                 cv = NULL,
                 max_coding_muts_per_sample = Inf,
                 max_muts_per_gene_per_sample = Inf,
                 outmats = T)

dndsout$globaldnds #theta = 27.5
#     name      mle     cilow   cihigh
#wmis wmis 1.271356 0.7756057 2.083978
#wnon wnon 1.651924 0.7932596 3.440052
#wspl wspl 1.660113 0.4428677 6.223023
#wtru wtru 1.653342 0.8256128 3.310922
#wall wall 1.313394 0.8066104 2.138582

##C>G
dnds_input <- all_vars_bx_flt_uniq[(all_vars_bx_flt_uniq$ref == "C" & all_vars_bx_flt_uniq$mut == "G") |
                                    (all_vars_bx_flt_uniq$ref == "G" & all_vars_bx_flt_uniq$mut == "C"),]

dndsout = dndscv(dnds_input,
                 refdb = "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/RefCDS_GRCh37_masked.Rdat",
                 cv = NULL,
                 max_coding_muts_per_sample = Inf,
                 max_muts_per_gene_per_sample = Inf,
                 outmats = T)

dndsout$globaldnds #theta = 11.9
#     name          mle      cilow    cihigh
#wmis wmis 5.638201e-01 0.32018877 0.9928302
#wnon wnon 8.216950e-09 0.00000000       Inf
#wspl wspl 3.815052e-01 0.04699188 3.0972633
#wtru wtru 2.343576e-01 0.02947042 1.8636815
#wall wall 5.535541e-01 0.31442875 0.9745359

##C>T
dnds_input <- all_vars_bx_flt_uniq[(all_vars_bx_flt_uniq$ref == "C" & all_vars_bx_flt_uniq$mut == "T") |
                                    (all_vars_bx_flt_uniq$ref == "G" & all_vars_bx_flt_uniq$mut == "A"),]

dndsout = dndscv(dnds_input,
                 refdb = "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/RefCDS_GRCh37_masked.Rdat",
                 cv = NULL,
                 max_coding_muts_per_sample = Inf,
                 max_muts_per_gene_per_sample = Inf,
                 outmats = T)

dndsout$globaldnds #theta = 0.243
#     name      mle     cilow   cihigh
#wmis wmis 1.484652 1.0740091 2.052302
#wnon wnon 2.200681 1.1748671 4.122166
#wspl wspl 2.327970 0.7860236 6.894762
#wtru wtru 2.228920 1.2699099 3.912155
#wall wall 1.547653 1.1264597 2.126335

##T>A
dnds_input <- all_vars_bx_flt_uniq[(all_vars_bx_flt_uniq$ref == "T" & all_vars_bx_flt_uniq$mut == "A") |
                                    (all_vars_bx_flt_uniq$ref == "A" & all_vars_bx_flt_uniq$mut == "T"),]

dndsout = dndscv(dnds_input,
                 refdb = "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/RefCDS_GRCh37_masked.Rdat",
                 cv = NULL,
                 max_coding_muts_per_sample = Inf,
                 max_muts_per_gene_per_sample = Inf,
                 outmats = T)

dndsout$globaldnds #theta = 7.04
#     name          mle     cilow    cihigh
#wmis wmis 2.899429e+00 0.3589349  23.42120
#wnon wnon 1.821384e+01 1.5375648 215.75929
#wspl wspl 9.238979e-08 0.0000000       Inf
#wtru wtru 1.338536e+01 1.1500665 155.78913
#wall wall 3.150370e+00 0.3958010  25.07531

##T>C
dnds_input <- all_vars_bx_flt_uniq[(all_vars_bx_flt_uniq$ref == "T" & all_vars_bx_flt_uniq$mut == "C") |
                                    (all_vars_bx_flt_uniq$ref == "A" & all_vars_bx_flt_uniq$mut == "G"),]

dndsout = dndscv(dnds_input,
                 refdb = "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/RefCDS_GRCh37_masked.Rdat",
                 cv = NULL,
                 max_coding_muts_per_sample = Inf,
                 max_muts_per_gene_per_sample = Inf,
                 outmats = T)

dndsout$globaldnds #theta = 6.96
#     name        mle     cilow    cihigh
#wmis wmis  0.7391607 0.3589409  1.522141
#wnon wnon 13.9161510 0.0000000       Inf
#wspl wspl  2.2301941 0.4881684 10.188626
#wtru wtru  2.2301941 0.4881684 10.188626
#wall wall  0.8023372 0.3963800  1.624060

##T>G
dnds_input <- all_vars_bx_flt_uniq[(all_vars_bx_flt_uniq$ref == "T" & all_vars_bx_flt_uniq$mut == "G") |
                                    (all_vars_bx_flt_uniq$ref == "A" & all_vars_bx_flt_uniq$mut == "C"),]

dndsout = dndscv(dnds_input,
                 refdb = "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/RefCDS_GRCh37_masked.Rdat",
                 cv = NULL,
                 max_coding_muts_per_sample = Inf,
                 max_muts_per_gene_per_sample = Inf,
                 outmats = T)

dndsout$globaldnds #theta = 5.89e-05
#     name          mle     cilow       cihigh
#wmis wmis 1.527953e+00 0.3095537     7.541954
#wnon wnon 1.460988e+02 0.9195342 23212.689731
#wspl wspl 2.281149e-08 0.0000000          Inf
#wtru wtru 4.462446e+00 0.2882032    69.095092
#wall wall 1.403725e+00 0.2953568     6.671397

```

####Run dNdS by VAF

```{r}

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

#read in subs
setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/07_recurrent")

txt_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt")

all_subs = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  all_subs = rbind(all_subs, data)
}

#read in indels
setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/08_recurrent")

txt_files = list.files(".", ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.recurrent_vars.filtered.txt")

all_indels = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.recurrent_vars.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  data = data[data$VAF >= 0.1,] #added 16/6/22 due to an excess of low VAF calls
  all_indels = rbind(all_indels, data)
}

#combine
all_vars_bx = rbind(all_subs, all_indels)
all_vars_bx_flt = all_vars_bx[all_vars_bx$sample %in% manifest[manifest$Other != "Membrane",]$Placenta_PD, ]

#high VAF
all_vars_bx_flt_high = all_vars_bx_flt[all_vars_bx_flt$VAF >= 0.2,]
all_vars_bx_flt_high_uniq = all_vars_bx_flt_high[!duplicated(all_vars_bx_flt_high[, c(1, 3, 4, 5, 6)]), c(1, 3, 4, 5, 6)]

names(all_vars_bx_flt_high_uniq) = c("sampleID", "chr", "pos", "ref", "mut")

library("dndscv")
library("seqinr")
library("Biostrings")
library("MASS")
library("GenomicRanges")
library(dplyr)

dnds_input <- all_vars_bx_flt_high_uniq

dndsout = dndscv(dnds_input,
                 refdb = "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/RefCDS_GRCh37_masked.Rdat",
                 cv = NULL,
                 max_coding_muts_per_sample = Inf,
                 max_muts_per_gene_per_sample = Inf,
                 outmats = T)

dndsout$globaldnds #theta = 0.299 subs, indels 11.5
#     name      mle     cilow   cihigh
#wmis wmis 1.194984 0.8874561 1.609077
#wnon wnon 2.267122 1.3142551 3.910839
#wspl wspl 2.257557 0.9964668 5.114634
#wtru wtru 2.264447 1.3990237 3.665213
#wall wall 1.274374 0.9508145 1.708040

saveRDS(dndsout, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/dndsoutput_highVAF_20220713.rds")

#low VAF
all_vars_bx_flt_low = all_vars_bx_flt[all_vars_bx_flt$VAF < 0.2,]
all_vars_bx_flt_low_uniq = all_vars_bx_flt_low[!duplicated(all_vars_bx_flt_low[, c(1, 3, 4, 5, 6)]), c(1, 3, 4, 5, 6)]

names(all_vars_bx_flt_low_uniq) = c("sampleID", "chr", "pos", "ref", "mut")

library("dndscv")
library("seqinr")
library("Biostrings")
library("MASS")
library("GenomicRanges")
library(dplyr)

dnds_input <- all_vars_bx_flt_low_uniq

dndsout = dndscv(dnds_input,
                 refdb = "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/RefCDS_GRCh37_masked.Rdat",
                 cv = NULL,
                 max_coding_muts_per_sample = Inf,
                 max_muts_per_gene_per_sample = Inf,
                 outmats = T)

dndsout$globaldnds #theta = 63.2 subs, 10.4 indels
#     name       mle     cilow   cihigh
#wmis wmis 1.1800764 0.8323867 1.672997
#wnon wnon 1.3638373 0.6569447 2.831368
#wspl wspl 0.9296693 0.2709364 3.189993
#wtru wtru 1.2310025 0.6417357 2.361357
#wall wall 1.1874990 0.8409067 1.676944

saveRDS(dndsout, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/dndsoutput_lowVAF_20220713.rds")

```

####Plot dNdS values

```{r}

all_dnds = readRDS("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/dndsoutput_20220713.rds")
high_vaf_dnds = readRDS("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/dndsoutput_highVAF_20220713.rds")
low_vaf_dnds = readRDS("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/dndsoutput_lowVAF_20220713.rds")

all_global = all_dnds$globaldnds
all_global$group = "all"
high_global = high_vaf_dnds$globaldnds
high_global$group = "high"
low_global = low_vaf_dnds$globaldnds
low_global$group = "low"

total_global = rbind(all_global, high_global, low_global)

library(ggplot2)
library(ggpubr)

ggplot(total_global) +
  geom_point(mapping = aes(x = group, y = mle, col = group)) +
  geom_linerange(mapping = aes(x = group, ymin = cilow, ymax = cihigh, col = group)) +
  facet_grid(. ~ name) +
  theme_pubr() +
  geom_hline(yintercept = 1, lty = 2) +
  xlab("") +
  theme(axis.text.x = element_blank()) +
  scale_color_manual(labels = c("All", "Early (VAF >=0.2)", "Late (VAF <0.2)"), values = c("black", "dodgerblue", "red"))

```

###DDD gene intersection

```{r}

#Get DD2GP genes

ddg2p = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/DDG2P_2_11_2021.csv", sep = ',', header = T, stringsAsFactors = F)
table(ddg2p$mutation.consequence)
colnames(ddg2p)[1] = "gene"
ddg2p_mono = unique(ddg2p[ddg2p$allelic.requirement %in% c("monoallelic","x-linked dominant","hemizygous","imprinted", "biallelic,monoallelic","mitochondrial") &  ddg2p$DDD.category %in% c("probable","confirmed","both DD and IF"), ]$gene)

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)

#read in subs
setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/07_recurrent")

txt_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt")

all_subs = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  all_subs = rbind(all_subs, data)
}

#read in indels
setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/indels/08_recurrent")

txt_files = list.files(".", ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.recurrent_vars.filtered.txt")

all_indels = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".pindel.annot.qual.pass.HP.9.maternal.ENCODE.LCR.blacklist.depth.unk.amb.direction.recurrent_vars.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  data = data[data$VAF >= 0.1,] #added 16/6/22 due to an excess of low VAF calls
  all_indels = rbind(all_indels, data)
}

#combine
all_vars_bx = rbind(all_subs, all_indels)
all_vars_bx_flt = all_vars_bx[all_vars_bx$sample %in% manifest[manifest$Other != "Membrane",]$Placenta_PD, ]

table(all_vars_bx_flt$Consequence)

ddd_ix = all_vars_bx_flt[all_vars_bx_flt$Gene %in% ddg2p_mono & all_vars_bx_flt$Consequence %in% c("ess_splice", "frameshift", "inframe", "missense", "nonsense", "5prime_UTR_ess_splice", "nc_ess_splice"),]
ddd_ix$mutID = paste(ddd_ix$Chr, ddd_ix$Position, ddd_ix$Wildtype, ddd_ix$Mutant, sep = "_")

nrow(ddd_ix) #28
length(unique(ddd_ix$mutID)) #28, not repeated across multiple biopsies

ddd_ix$ddd_mutation.consequence = NA
ddd_ix$ddd_disease.name = NA
for(i in 1:nrow(ddd_ix)){
  ddd_ix$ddd_mutation.consequence[i] = paste(ddg2p[ddg2p$gene == ddd_ix$Gene[i],]$mutation.consequence, collapse = ", ")
  ddd_ix$ddd_disease.name[i] =paste(ddg2p[ddg2p$gene == ddd_ix$Gene[i],]$disease.name, collapse = ", ")
}

write.table(ddd_ix, "/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/ddd_dominant_gene_intersection_placenta_20220713.txt", col.names = T, row.names = T, sep = '\t', quote = F)

```

###Genes within TD/DEL

Review the variants called across all samples
```{r}

svs_flt = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/placenta_brass_calls_final_20220712.txt", header = T, sep = '\t', stringsAsFactors = F)

#annotate genes contained within deletions and tandem-duplications
##generate bed file

annot_sv <- svs_flt[svs_flt$svclass %in% c('deletion', 'tandem-duplication'),] #269 SVs

write.table(annot_sv[, c('chr1', 'start1', 'end2')], '/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/placenta_dup_del_coord_20220712.bed', col.names = F, row.names = F, quote = F, sep = '\t')

```

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/

module load bedtools
~/scripts/intersect_intv.sh placenta_dup_del_coord_20220712.bed placenta_sv_genes_20220712.txt

```

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/")

sv_genes <- read.table('placenta_sv_genes_20220712.txt', header = T, stringsAsFactors = F, sep = '\t')
sv_genes <- sv_genes[!(duplicated(sv_genes)),]
sv_genes$Chr <- unlist(lapply(strsplit(sv_genes$Chromosome, ':'), "[[", 1))
sv_genes$Gene.Start <- as.numeric(unlist(lapply(strsplit(unlist(lapply(strsplit(sv_genes$Chromosome, ':'), "[[", 2)), "-"), "[[", 1)))
sv_genes$Gene.End <- as.numeric(unlist(lapply(strsplit(unlist(lapply(strsplit(sv_genes$Chromosome, ':'), "[[", 2)), "-"), "[[", 2))) #these coordinates are for the exonic regions of genes

reassembled = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/structural_variants/placenta_brass_calls_final_20220712.txt", header = T, sep = '\t', stringsAsFactors = F)

reassembled$Genes.in.dup.del <- NA
annot_sv$Genes.in.dup.del <- NA

for(i in 1:nrow(sv_genes)){
  if(length(annot_sv[annot_sv$chr1 == sv_genes$Chr[i] & #if there are SVs containing the gene of interest
                     annot_sv$start1 < sv_genes$Gene.End[i] & 
                     annot_sv$end2 > sv_genes$Gene.Start[i],]$Genes.in.dup.del)){
    annot_sv[annot_sv$chr1 == sv_genes$Chr[i] & 
               annot_sv$start1 < sv_genes$Gene.End[i] & 
               annot_sv$end2 > sv_genes$Gene.Start[i],]$Genes.in.dup.del = paste(annot_sv[annot_sv$chr1 == sv_genes$Chr[i] & 
                                                                                            annot_sv$start1 < sv_genes$Gene.End[i] & 
                                                                                            annot_sv$end2 > sv_genes$Gene.Start[i],]$Genes.in.dup.del, sv_genes$GeneSymbol[i], sep = ',')
  }
}

##now remove aberrant NAs after paste function
max(nchar(annot_sv$Genes.in.dup.del), na.rm = T) #129
annot_sv[!is.na(annot_sv$Genes.in.dup.del),]$Genes.in.dup.del <- substr(annot_sv[!is.na(annot_sv$Genes.in.dup.del),]$Genes.in.dup.del, 4, 130) 

for(i in 1:nrow(reassembled)){
  if(reassembled$chr1[i] %in% annot_sv$chr1 & 
     reassembled$start1[i] %in% annot_sv$start1 & 
     reassembled$end1[i] %in% annot_sv$end1 & 
     reassembled$chr2[i] %in% annot_sv$chr2 & 
     reassembled$start2[i] %in% annot_sv$start2 & 
     reassembled$end2[i] %in% annot_sv$end2 & 
     reassembled$sample[i] %in% annot_sv$sample){
    reassembled$Genes.in.dup.del[i] = annot_sv[annot_sv$chr1 == reassembled$chr1[i] & 
                                                 annot_sv$start1 == reassembled$start1[i] & 
                                                 annot_sv$end1 == reassembled$end1[i] & 
                                                 annot_sv$chr2 == reassembled$chr2[i] & 
                                                 annot_sv$start2 == reassembled$start2[i] & 
                                                 annot_sv$end2 == reassembled$end2[i] & 
                                                 annot_sv$sample == reassembled$sample[i],]$Genes.in.dup.del[1]
  }
}

reassembled$Genes.at.breakpoints = NA

for(i in 1:nrow(reassembled)){
  reassembled$Genes.at.breakpoints[i] = paste(unique(c(reassembled$gene1[i], reassembled$gene2[i]))[!grepl("_", unique(c(reassembled$gene1[i], reassembled$gene2[i])))], collapse = ",")
}

#check if breakpoints generally or regions within intrachromosomal SVs are cancer genes
##Intersect with COSMIC v94 consensus cancer gene list
cosmic_v94 <- read.table('/lustre/scratch119/casm/team294rr/to3/testes/tumour/manuscript/nature_resub/reference_datasets/cosmic_v94_cancer_gene_census.csv', sep = ',', header = T, stringsAsFactors = F)
cancer_genes <- unlist(strsplit(paste(cosmic_v94$Synonyms, collapse = ","), ","))

##add flagged cancer genes to list 
reassembled$Cancer.gene <- NA

for(i in 1:nrow(reassembled)){
  gene_list = unlist(strsplit(reassembled$Genes.in.dup.del[i], ','))
  gene_list = c(gene_list, reassembled$gene1[i], reassembled$gene2[i])
  gene_list = unique(gene_list)
  gene_list = cancer_genes[which(as.vector(cancer_genes) %in% as.vector(gene_list))]
  if(length(gene_list)){
    reassembled$Cancer.gene[i] = paste(gene_list, collapse = ',')
  }
  else {
    reassembled$Cancer.gene[i] = NA
  }
}

reassembled$Cancer.gene

#annotate whether the cancer genes are dominant or recessive
reassembled$Dom.Rec.COSMICv94 <- NA
for(i in 1:nrow(reassembled)){
  if(!is.na(reassembled$Cancer.gene[i])){
    gene.list <- unlist(strsplit(reassembled$Cancer.gene[i], ','))
    reassembled$Dom.Rec.COSMICv94[i] = paste(cosmic_v94[cosmic_v94$Gene.Symbol %in% gene.list, ]$Molecular.Genetics, collapse = ',')
  }
}

#annotate if in monoallelic DDD genes

#Get DD2GP genes

ddg2p = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/DDG2P_2_11_2021.csv", sep = ',', header = T, stringsAsFactors = F)
table(ddg2p$mutation.consequence)
colnames(ddg2p)[1] = "gene"
ddg2p_mono = unique(ddg2p[ddg2p$allelic.requirement %in% c("monoallelic","x-linked dominant","hemizygous","imprinted", "biallelic,monoallelic","mitochondrial") &  ddg2p$DDD.category %in% c("probable","confirmed","both DD and IF"), ]$gene)

reassembled$DDD.mono.genes <- NA
for(i in 1:nrow(reassembled)){
  gene_list = unlist(strsplit(reassembled$Genes.in.dup.del[i], ','))
  gene_list = c(gene_list, reassembled$gene1[i], reassembled$gene2[i])
  gene_list = unique(gene_list)
  gene_list = ddg2p_mono[which(as.vector(ddg2p_mono) %in% as.vector(gene_list))]
  if(length(gene_list)){
    reassembled$DDD.mono.genes[i] = paste(gene_list, collapse = ',')
  }
  else {
    reassembled$DDD.mono.genes[i] = NA
  }
}

write.table(reassembled, '/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/placenta_brass_calls_final_20220712_annot.txt', col.names = T, row.names = F, quote = F, sep = '\t')

```

##Rainfall plot of all unique subs

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/rainfall_plot")

all_vars_bx_flt_uniq = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/all_placenta_muts_uniq_20220713.txt", header = T, sep = '\t', stringsAsFactors = F)

#all subs together
data = all_vars_bx_flt_uniq
data$ID = "."
data.2 = data[, c("chr", "pos", "ID", "ref", "mut")]
names(data.2) = c("#CHROM", "POS", "ID", "REF", "ALT")
data.2$`#CHROM` = factor(data.2$`#CHROM`, levels = c(1:22, "X", "Y"))
data.3 = data.2[order(data.2$`#CHROM`, data.2$POS, decreasing = F),]
data.3 = data.3[nchar(data.3$REF) == 1 & nchar(data.3$ALT) == 1, ] #remove indels
write.table(data.3, "all_placenta_uniq_flt_subs.vcf", col.names = T, row.names = F, sep = '\t', quote = F)

muts = paste(data.3$`#CHROM`, data.3$POS, data.3$REF, data.3$ALT, sep = "_")
muts[duplicated(muts)]

```

Add extra header lines

```{bash}

cp vcf_header_lines.txt all_placenta_uniq_flt_subs_header.vcf
cat all_placenta_uniq_flt_subs.vcf >> all_placenta_uniq_flt_subs_header.vcf

```

```{r}

library(GenomicRanges)
library(MutationalPatterns)
ref_genome <- "BSgenome.Hsapiens.UCSC.hg19"
library(BSgenome.Hsapiens.UCSC.hg19)

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)
#manifest$case.id = substr(manifest$Placenta_PD, 0, 7)
#clinical.df = manifest[, c("case.id", "clin_group_relabel")]
#clinical.df = clinical.df[!duplicated(clinical.df),]
#row.names(clinical.df) = clinical.df$case.id

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/rainfall_plot")

vcf_files <- list.files(".", pattern = "all_placenta_uniq_flt_subs_header.vcf", full.names = F)
sample_names = "all"
grl <- read_vcfs_as_granges(vcf_files, sample_names, ref_genome) #it has removed DBS and a single SBS that is called at the same site as another

c_a_grl = grl[[1]][grepl("C\\/A", names(grl[[1]])) | grepl("G\\/T", names(grl[[1]])),]

#chr = substr(as.vector(grl$all@seqnames), 4, 5)
#pos = start(grl$all@ranges)
#ref = as.vector(grl$all$REF)
#alt = as.vector(unlist(grl$all$ALT))

#muts.2 = paste(chr, pos, ref, alt, sep = "_")
#muts.2[duplicated(muts.2)]

#muts[!muts %in% muts.2] #6_13184428_G_T stands out, removed for unclear reasons
#diff(data.3$POS[muts %in% muts[!muts %in% muts.2]])

#group = clinical.df[sample_names,]$clin_group_relabel

# Define autosomal chromosomes
chromosomes <- seqnames(get(ref_genome))[1:22]

# Make a rainfall plot
plot_rainfall(grl[[1]],
  title = names(grl[1]),
  chromosomes = chromosomes, cex = 1.5, ylim = 1e+09
)

plot_rainfall(c_a_grl,
  title = "C_A",
  chromosomes = chromosomes, cex = 1.5, ylim = 1e+09
)

#examine in more detail, DBS excluded
chr = substr(as.vector(grl$all@seqnames), 4, 5)
pos = start(grl$all@ranges)
ref = as.vector(grl$all$REF)
alt = as.vector(unlist(grl$all$ALT))

muts.df = data.frame(chr = chr,
                     pos = pos,
                     ref = ref,
                     alt = alt)

muts.df$diff = NA
for(chr in muts.df$chr){
  diff_vec = c(NA, diff(muts.df[muts.df$chr == chr, ]$pos))
  muts.df[muts.df$chr == chr, ]$diff = diff_vec
}

mut_set.1 = as.numeric(row.names(muts.df[!is.na(muts.df$diff) & muts.df$diff < 100 & muts.df$chr %in% c(1:22),])) - 1 #preceding mutation
mut_set.2 = as.numeric(row.names(muts.df[!is.na(muts.df$diff) & muts.df$diff < 100 & muts.df$chr %in% c(1:22),])) #proceeding mutation

muts_set.all = sort(unique(c(mut_set.1, mut_set.2)))

hotspot.subs = muts.df[muts_set.all, ]
hs.muts = paste(hotspot.subs$chr, hotspot.subs$pos, hotspot.subs$ref, hotspot.subs$alt, sep = "_")

#read in subs
setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/07_recurrent")

txt_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt")

all_subs = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  all_subs = rbind(all_subs, data)
}

table(all_subs[all_subs$mut_ID %in% hs.muts,]$Consequence)

```

##Indel spectra per clinical group

```{r}

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/indel_spectra")

all_vars_bx_flt_uniq = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/driver_analysis/all_placenta_muts_uniq_20220713.txt", header = T, sep = '\t', stringsAsFactors = F)

for(placenta in unique(all_vars_bx_flt_uniq$sampleID)){
  data = all_vars_bx_flt_uniq[all_vars_bx_flt_uniq$sampleID == placenta, ]
  data$ID = "."
  data.2 = data[, c("chr", "pos", "ID", "ref", "mut")]
  names(data.2) = c("#CHROM", "POS", "ID", "REF", "ALT")
  data.2$`#CHROM` = factor(data.2$`#CHROM`, levels = c(1:22, "X", "Y"))
  data.3 = data.2[order(data.2$`#CHROM`, data.2$POS, decreasing = F),]
  data.3 = data.3[nchar(data.3$REF) > 1 | nchar(data.3$ALT) > 1, ] #indels only
  
  if(nrow(data.3) > 0){
    write.table(data.3, paste0(placenta, "_uniq_flt_indel.vcf"), col.names = T, row.names = F, sep = '\t', quote = F)
  }
}

```

Add extra header lines

```{bash}

cd /lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/indel_spectra

for file in $(ls *_uniq_flt_indel.vcf);
do
sed -i '1i ##fileformat=VCFv4.1' $file 
done

```

```{r}

library(GenomicRanges)
library(MutationalPatterns)
ref_genome <- "BSgenome.Hsapiens.UCSC.hg19"
library(BSgenome.Hsapiens.UCSC.hg19)

manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest$case.id = substr(manifest$Placenta_PD, 0, 7)
clinical.df = manifest[, c("case.id", "clin_group_relabel")]
clinical.df = clinical.df[!duplicated(clinical.df),]
row.names(clinical.df) = clinical.df$case.id

setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/indel_spectra")

vcf_files <- list.files(".", pattern = "_uniq_flt_indel.vcf", full.names = F)
sample_names = substr(vcf_files, 0, 7)
grl <- read_vcfs_as_granges(vcf_files, sample_names, ref_genome, type = "indel")
group = clinical.df[sample_names,]$clin_group_relabel

#get indel spectra
indel_grl <- get_indel_context(grl, ref_genome)
indel_counts <- count_indel_contexts(indel_grl)
indel_counts.df = data.frame(indel_counts)
indel_counts.df$healthy = rowSums(indel_counts.df[, colnames(indel_counts.df) %in% clinical.df[clinical.df$clin_group_relabel == "healthy_control",]$case.id])
indel_counts.df$small_baby = rowSums(indel_counts.df[, colnames(indel_counts.df) %in% clinical.df[clinical.df$clin_group_relabel == "small_baby",]$case.id])
indel_counts.df$big_baby_placenta = rowSums(indel_counts.df[, colnames(indel_counts.df) %in% clinical.df[clinical.df$clin_group_relabel == "big_baby_placenta",]$case.id])
indel_counts.df$pre_eclampsia = rowSums(indel_counts.df[, colnames(indel_counts.df) %in% clinical.df[clinical.df$clin_group_relabel == "pre_eclampsia",]$case.id])
indel_counts.groups.mat = as.matrix(indel_counts.df[, 113:116])

plot_indel_contexts(indel_counts.groups.mat, condensed = TRUE, )

```

##Shared subs between biopsies

```{r}

#setwd("/lustre/scratch119/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/07_recurrent")
setwd("/lustre/scratch126/casm/team274sb/to3/placenta_ml/02_Variant_filtering/matched_ssms/subs/07_recurrent")

#manifest = read.table("/lustre/scratch119/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest = read.table("/lustre/scratch126/casm/team274sb/to3/placenta_ml/00_Supplementary_data/placenta_ml_metadata_20220711.txt", header = T, sep = '\t', stringsAsFactors = F)
manifest_flt = manifest[manifest$Other != "Membrane",]
manifest_flt$case.id = substr(manifest_flt$Placenta_PD, 0, 7)

txt_files = list.files(".", ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt")

all_subs = data.frame()
for(file in txt_files){
  sample = unlist(strsplit(file, ".caveman_c.annot.asmd.clpm.pass.maternal.ENCODE.LCR.blacklist.bq.mq.direction.indel.recurrent_var.filtered.txt"))
  data = read.table(file, header = T, sep = '\t', stringsAsFactors = F)  
  data$sample = sample
  all_subs = rbind(all_subs, data)
}

all_subs_flt = all_subs[all_subs$Sample %in% manifest_flt$Placenta_PD, ] #only HQ, placental samples

#check that no mutations are called in different placentas, shouldn't be the case after filtering shared variants
max(table(all_subs_flt[!duplicated(all_subs_flt$Case_ID),]$mut_ID)) #1 correct

#identify mutations called in multiple placental biopsies
shared_muts = unique(all_subs_flt$mut_ID[duplicated(all_subs_flt$mut_ID)])
unique_muts = all_subs_flt$mut_ID[!all_subs_flt$mut_ID %in% shared_muts]

length(c(shared_muts, unique_muts)) #47388
nrow(all_subs_flt) #48289, some duplicated multiple times

#placenta = "PD42145"
for(placenta in unique(all_subs_flt$Case_ID)){
  placenta_subs = all_subs_flt[all_subs_flt$Case_ID == placenta,]
  bx = manifest_flt[manifest_flt$case.id == placenta,]$Placenta_PD
  intersect_df = data.frame(matrix(nrow = length(bx), ncol = length(bx), dimnames = list(bx, bx)))
  
  for(i in 1:nrow(intersect_df)){
    for(j in 1:ncol(intersect_df)){
      intersect_df[i, j] = length(intersect(placenta_subs[placenta_subs$Sample == row.names(intersect_df)[i],]$mut_ID, placenta_subs[placenta_subs$Sample == colnames(intersect_df)[j],]$mut_ID))
    }
  }
  
  write.table(intersect_df, paste0("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/sharedness_assessment/", placenta, "_called_subs_intersect.txt"), col.names = T, row.names = T, sep = '\t', quote = F)
  
}

placenta_df = data.frame(CaseID = unique(manifest_flt$case.id), 
                         no.subs.shared = NA,
                         no.subs.shared.2bx = NA,
                         no.subs.shared.3bx = NA,
                         no.subs.shared.4bx = NA,
                         no.subs.total = NA,
                         no.bx = NA)
all_subs_flt_shared = all_subs_flt[all_subs_flt$mut_ID %in% shared_muts,]

for(i in 1:nrow(placenta_df)){
  
  placenta_df$no.subs.shared[i] = length(unique(all_subs_flt_shared[all_subs_flt_shared$Case_ID == placenta_df$CaseID[i],]$mut_ID))
  placenta_df$no.subs.shared.2bx[i] = sum(table(all_subs_flt_shared[all_subs_flt_shared$Case_ID == placenta_df$CaseID[i],]$mut_ID) == 2)
  placenta_df$no.subs.shared.3bx[i] = sum(table(all_subs_flt_shared[all_subs_flt_shared$Case_ID == placenta_df$CaseID[i],]$mut_ID) == 3)
  placenta_df$no.subs.shared.4bx[i] = sum(table(all_subs_flt_shared[all_subs_flt_shared$Case_ID == placenta_df$CaseID[i],]$mut_ID) == 4)
  placenta_df$no.subs.total[i] = length(unique(all_subs_flt[all_subs_flt$Case_ID == placenta_df$CaseID[i],]$mut_ID))
  placenta_df$no.bx[i] = nrow(manifest_flt[manifest_flt$case.id == placenta_df$CaseID[i],])
  
}

sum(placenta_df$no.bx) #371, correct
table(placenta_df$no.bx)

library(ggplot2)

ggplot(placenta_df) + 
  geom_boxplot(mapping = aes(x = factor(no.bx), y = no.subs.shared))

length(placenta_df[placenta_df$no.bx == 4,]$no.subs.shared) #66
sum(placenta_df[placenta_df$no.bx == 4,]$no.subs.shared == 0) #13
sum(placenta_df[placenta_df$no.bx == 4,]$no.subs.shared <= 5) #40
sum(placenta_df[placenta_df$no.bx == 4,]$no.subs.shared > 25) #5

table(placenta_df$no.bx)
#1  2  3  4 
#2 27 17 66 

placenta_df[placenta_df$no.subs.shared > 25 & placenta_df$no.bx == 4,]

placenta_df[placenta_df$no.subs.shared.3bx > 0,]

sum(placenta_df$no.subs.shared.2bx) #634
sum(placenta_df$no.subs.shared) #754

summary(placenta_df[placenta_df$no.bx == 4,]$no.subs.shared)

pdf("/lustre/scratch119/casm/team274sb/to3/placenta_ml/03_DNA_analyses/figures/four_bx_placenta_shared_thesis_20230216.pdf", width = 4, height = 4)
ggplot(placenta_df[placenta_df$no.bx == 4,]) + 
  geom_histogram(mapping = aes(x = no.subs.shared)) +
  theme_bw() +
  ylab("# placentas") +
  xlab("Number of substitutions shared between two or more biopsies")
dev.off()

nrow(placenta_df[placenta_df$no.bx == 4,]) #66

```
